{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import openai\n",
    "import pandas as pd\n",
    "import tiktoken as tk\n",
    "import tenacity\n",
    "from tqdm import tqdm\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import multiprocessing\n",
    "from colorama import Fore, Style\n",
    "import os\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1837, 20)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['numLines', 'imports', 'numParams', 'signature', 'lineStart',\n",
       "       'filePath', 'numChars', 'params', 'additionalImports', 'lineEnd',\n",
       "       'content', 'numCommentChars', 'numCommentLines', 'intraClassFieldsUsed',\n",
       "       'function_id', 'parentClass', 'comment', 'packageName', 'returnType',\n",
       "       'hash'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "initial_java_functions = \"/home/ubuntu/Bio-Code-Eval/Bio-Code-Eval-new/1-select-functions/filtered_functions_step_1_java.json\"\n",
    "\n",
    "df = pd.read_json(initial_java_functions)\n",
    "\n",
    "print(df.shape)\n",
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['totalNumLines', 'imports', 'numParameters', 'signature', 'lineStart',\n",
      "       'filePath', 'numChar', 'params', 'addImports', 'lineEnd', 'content',\n",
      "       'numCommentChars', 'numCommentLines', 'intraClassFieldsUsed',\n",
      "       'function_id', 'parentClass', 'comments', 'packageName', 'returnType',\n",
      "       'hash'],\n",
      "      dtype='object')\n",
      "(1837, 20)\n"
     ]
    }
   ],
   "source": [
    "# Renaming everything to spec\n",
    "\n",
    "df = df.rename(columns = {'numLines': 'totalNumLines', \n",
    "                          'numParams': 'numParameters', \n",
    "                          'numChars' : 'numChar', \n",
    "                          'additionalImports': 'addImports',\n",
    "                          'comment': 'comments'\n",
    "                          })\n",
    "\n",
    "print(df.columns)\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "totalNumLines                                                          12\n",
      "imports                 {'classCount': 0, 'imports': [], 'asteriskCoun...\n",
      "numParameters                                                           1\n",
      "signature               public void cleanJob(File jobDir) throws IOExc...\n",
      "lineStart                                                             110\n",
      "filePath                /home/ubuntu/repos/TGAC,RAMPART/src/main/java/...\n",
      "numChar                                                                 0\n",
      "params                                                      [File jobDir]\n",
      "addImports                                                             {}\n",
      "lineEnd                                                               122\n",
      "content                 public void cleanJob(File jobDir) throws IOExc...\n",
      "numCommentChars                                                         2\n",
      "numCommentLines                                                         5\n",
      "intraClassFieldsUsed    public static final String DESC = \"Cleans a RA...\n",
      "function_id                                                    1160665474\n",
      "parentClass                                                        PARSED\n",
      "comments                /**\\n     * Cleans a RAMPART job directory of ...\n",
      "packageName                                       uk.ac.tgac.rampart.util\n",
      "returnType                                                           void\n",
      "hash                                                  4293468902114874299\n",
      "Name: 71, dtype: object\n",
      "Index(['totalNumLines', 'imports', 'numParameters', 'signature', 'lineStart',\n",
      "       'filePath', 'numChar', 'params', 'addImports', 'lineEnd', 'content',\n",
      "       'numCommentChars', 'numCommentLines', 'intraClassFieldsUsed',\n",
      "       'function_id', 'parentClass', 'comments', 'packageName', 'returnType',\n",
      "       'hash'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(df.iloc[0])\n",
    "print(df.columns)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['imports'] = df['imports'].astype(str)\n",
    "df['addImports'] = df['addImports'].astype(str)\n",
    "df['comments'] = df['comments'].astype(str)\n",
    "df['signature'] = df['signature'].astype(str)\n",
    "df['content'] = df['content'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "<class 'str'>\n",
      "<class 'str'>\n",
      "<class 'str'>\n",
      "<class 'str'>\n"
     ]
    }
   ],
   "source": [
    "print(type(df['imports'].iloc[0]))\n",
    "print(type(df['addImports'].iloc[0]))\n",
    "print(type(df['comments'].iloc[0]))\n",
    "print(type(df['signature'].iloc[0]))\n",
    "print(type(df['content'].iloc[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Cleans a RAMPART job directory of any known temporary information.  This will not delete any information not in\n",
      "     * the MECQ, MASS, AMP or REPORT directories.\n",
      "     * @param jobDir The RAMPART job directory to clean\n",
      "     * @throws java.io.IOException Thrown if there were an issues cleaning the directory\n",
      "     */\n",
      "public void cleanJob(File jobDir) throws IOException\n",
      "**************************************\n",
      "public void cleanJob(File jobDir) throws IOException {\n",
      "    if (this.verbose) {\n",
      "        System.out.print(\"Cleaning RAMPART job directory: \" + jobDir.getAbsolutePath() + \" ...\");\n",
      "    }\n",
      "    FileUtils.deleteDirectory(new File(jobDir, \"rampart_out\"));\n",
      "    FileUtils.deleteDirectory(new File(jobDir, \"wait_logs\"));\n",
      "    if (this.verbose) {\n",
      "        System.out.println(\" done.\");\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "df['prompt'] = df['imports'].str.cat([df['addImports'], df['comments'], df['signature']], sep='\\n') + '\\n**************************************\\n' + df['content']\n",
    "print(df.iloc[0]['prompt'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_tokens(s):\n",
    "    return len(s.split())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['good_length'] = df['prompt'].apply(lambda x: count_tokens(x.split('**************************************')[0]) < 4096)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       totalNumLines                                            imports   \n",
      "71                12  {'classCount': 0, 'imports': [], 'asteriskCoun...  \\\n",
      "86                10  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "87                13  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "90                11  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "91                11  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "...              ...                                                ...   \n",
      "58158              8  {'classCount': 2, 'imports': [{'fieldCount': 0...   \n",
      "58159             12  {'classCount': 2, 'imports': [{'fieldCount': 0...   \n",
      "58221             11  {'classCount': 0, 'imports': [{'fieldCount': 2...   \n",
      "58222             10  {'classCount': 0, 'imports': [{'fieldCount': 2...   \n",
      "58241              9  {'classCount': 2, 'imports': [{'fieldCount': 8...   \n",
      "\n",
      "       numParameters                                          signature   \n",
      "71                 1  public void cleanJob(File jobDir) throws IOExc...  \\\n",
      "86                 1                       public T create(String name)   \n",
      "87                 2  public T create(String name, ConanExecutorServ...   \n",
      "90                 0                 public List<String> listServices()   \n",
      "91                 0                       public List<T> getServices()   \n",
      "...              ...                                                ...   \n",
      "58158              3  public void initialize(final QualityYieldMetri...   \n",
      "58159              2  public void collectMetrics(final JavaRDD<GATKR...   \n",
      "58221              1  public static Map<String, Byte> constructLibra...   \n",
      "58222              1  private static Map<String, Short> getHeaderRea...   \n",
      "58241              0   protected String[] customCommandLineValidation()   \n",
      "\n",
      "       lineStart                                           filePath  numChar   \n",
      "71           110  /home/ubuntu/repos/TGAC,RAMPART/src/main/java/...        0  \\\n",
      "86            58  /home/ubuntu/repos/TGAC,RAMPART/src/main/java/...        0   \n",
      "87            76  /home/ubuntu/repos/TGAC,RAMPART/src/main/java/...        0   \n",
      "90           125  /home/ubuntu/repos/TGAC,RAMPART/src/main/java/...        0   \n",
      "91           142  /home/ubuntu/repos/TGAC,RAMPART/src/main/java/...        0   \n",
      "...          ...                                                ...      ...   \n",
      "58158         30  /home/ubuntu/repos/broadinstitute,gatk/src/mai...        0   \n",
      "58159         45  /home/ubuntu/repos/broadinstitute,gatk/src/mai...        0   \n",
      "58221        206  /home/ubuntu/repos/broadinstitute,gatk/src/mai...        0   \n",
      "58222        222  /home/ubuntu/repos/broadinstitute,gatk/src/mai...        0   \n",
      "58241        390  /home/ubuntu/repos/broadinstitute,gatk/src/mai...        0   \n",
      "\n",
      "                                                  params   \n",
      "71                                         [File jobDir]  \\\n",
      "86                                         [String name]   \n",
      "87               [String name, ConanExecutorService ces]   \n",
      "90                                                    []   \n",
      "91                                                    []   \n",
      "...                                                  ...   \n",
      "58158  [final QualityYieldMetricsArgumentCollection i...   \n",
      "58159  [final JavaRDD<GATKRead> filteredReads, final ...   \n",
      "58221                       [final SAMFileHeader header]   \n",
      "58222                       [final SAMFileHeader header]   \n",
      "58241                                                 []   \n",
      "\n",
      "                                              addImports  lineEnd  ...   \n",
      "71                                                    {}      122  ...  \\\n",
      "86                                                    {}       68  ...   \n",
      "87                                                    {}       89  ...   \n",
      "90                                                    {}      136  ...   \n",
      "91                                                    {}      153  ...   \n",
      "...                                                  ...      ...  ...   \n",
      "58158  {'QualityYieldMetrics': 'public class QualityY...       38  ...   \n",
      "58159  {'QualityYieldMetrics': 'public class QualityY...       57  ...   \n",
      "58221                                                 {}      217  ...   \n",
      "58222                                                 {}      232  ...   \n",
      "58241                                                 {}      399  ...   \n",
      "\n",
      "      numCommentLines                               intraClassFieldsUsed   \n",
      "71                  5  public static final String DESC = \"Cleans a RA...  \\\n",
      "86                  4  private ServiceLoader<T> loader;\\nprivate Clas...   \n",
      "87                  5  private ServiceLoader<T> loader;\\nprivate Clas...   \n",
      "90                  3  private ServiceLoader<T> loader;\\nprivate Clas...   \n",
      "91                  3  private ServiceLoader<T> loader;\\nprivate Clas...   \n",
      "...               ...                                                ...   \n",
      "58158               2  private static final long serialVersionUID = 1...   \n",
      "58159               4  private static final long serialVersionUID = 1...   \n",
      "58221               2  public static final String OPTICAL_DUPLICATE_T...   \n",
      "58222               2  public static final String OPTICAL_DUPLICATE_T...   \n",
      "58241               3  private static final long serialVersionUID = 1...   \n",
      "\n",
      "       function_id parentClass   \n",
      "71      1160665474      PARSED  \\\n",
      "86       684336824      PARSED   \n",
      "87      1229318016      PARSED   \n",
      "90        40356599      PARSED   \n",
      "91     -1945730202      PARSED   \n",
      "...            ...         ...   \n",
      "58158    404558657      PARSED   \n",
      "58159   1225186912      PARSED   \n",
      "58221   -575023434      PARSED   \n",
      "58222   1173260070      PARSED   \n",
      "58241    706803885      PARSED   \n",
      "\n",
      "                                                comments   \n",
      "71     /**\\n     * Cleans a RAMPART job directory of ...  \\\n",
      "86     /**\\n     * Creates a new instance of the serv...   \n",
      "87     /**\\n     * Creates a new instance of the serv...   \n",
      "90     /**\\n     * Returns a list of a strings descri...   \n",
      "91     /**\\n     * Returns a list of the requested se...   \n",
      "...                                                  ...   \n",
      "58158  /**\\n     * Initialize the collector with inpu...   \n",
      "58159  /**\\n     * Do the actual metrics collection o...   \n",
      "58221  /**\\n     * Method which generates a map of th...   \n",
      "58222  /**\\n     * Method which generates a map of th...   \n",
      "58241  /**\\n     * mode adjustments\\n     * @return e...   \n",
      "\n",
      "                                             packageName         returnType   \n",
      "71                               uk.ac.tgac.rampart.util               void  \\\n",
      "86                               uk.ac.tgac.rampart.util                  T   \n",
      "87                               uk.ac.tgac.rampart.util                  T   \n",
      "90                               uk.ac.tgac.rampart.util       List<String>   \n",
      "91                               uk.ac.tgac.rampart.util            List<T>   \n",
      "...                                                  ...                ...   \n",
      "58158  org.broadinstitute.hellbender.tools.spark.pipe...               void   \n",
      "58159  org.broadinstitute.hellbender.tools.spark.pipe...               void   \n",
      "58221  org.broadinstitute.hellbender.tools.spark.tran...   Map<String,Byte>   \n",
      "58222  org.broadinstitute.hellbender.tools.spark.tran...  Map<String,Short>   \n",
      "58241  org.broadinstitute.hellbender.tools.spark.tran...           String[]   \n",
      "\n",
      "                      hash                                             prompt   \n",
      "71     4293468902114874299  {'classCount': 0, 'imports': [], 'asteriskCoun...  \\\n",
      "86    -6087990313142724704  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "87    -7755862905863672092  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "90     2306136299996287996  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "91    -5402760501751281215  {'classCount': 0, 'imports': [], 'asteriskCoun...   \n",
      "...                    ...                                                ...   \n",
      "58158 -2296360894438706670  {'classCount': 2, 'imports': [{'fieldCount': 0...   \n",
      "58159 -1437108828658389746  {'classCount': 2, 'imports': [{'fieldCount': 0...   \n",
      "58221  2041963366887340468  {'classCount': 0, 'imports': [{'fieldCount': 2...   \n",
      "58222  5424899629802517197  {'classCount': 0, 'imports': [{'fieldCount': 2...   \n",
      "58241 -5237748791198521558  {'classCount': 2, 'imports': [{'fieldCount': 8...   \n",
      "\n",
      "       good_length  \n",
      "71            True  \n",
      "86            True  \n",
      "87            True  \n",
      "90            True  \n",
      "91            True  \n",
      "...            ...  \n",
      "58158         True  \n",
      "58159         True  \n",
      "58221         True  \n",
      "58222         True  \n",
      "58241        False  \n",
      "\n",
      "[1837 rows x 22 columns]\n"
     ]
    }
   ],
   "source": [
    "print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Cleans a RAMPART job directory of any known temporary information.  This will not delete any information not in\n",
      "     * the MECQ, MASS, AMP or REPORT directories.\n",
      "     * @param jobDir The RAMPART job directory to clean\n",
      "     * @throws java.io.IOException Thrown if there were an issues cleaning the directory\n",
      "     */\n",
      "public void cleanJob(File jobDir) throws IOException\n",
      "**************************************\n",
      "public void cleanJob(File jobDir) throws IOException {\n",
      "    if (this.verbose) {\n",
      "        System.out.print(\"Cleaning RAMPART job directory: \" + jobDir.getAbsolutePath() + \" ...\");\n",
      "    }\n",
      "    FileUtils.deleteDirectory(new File(jobDir, \"rampart_out\"));\n",
      "    FileUtils.deleteDirectory(new File(jobDir, \"wait_logs\"));\n",
      "    if (this.verbose) {\n",
      "        System.out.println(\" done.\");\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "print(df.iloc[0]['prompt'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dataframe now has generated prompts, now to filter prompts based on keywords\n",
    "for_keyword_filtering = df\n",
    "keywords = \"/home/ubuntu/Bio-Code-Eval/Bio-Code-Eval-new/2-filter-and-annotate/dictionary.csv\"\n",
    "\n",
    "# Read the words from the CSV file into a list\n",
    "words = []\n",
    "with open(keywords, 'r', newline='') as f:\n",
    "    reader = csv.reader(f)\n",
    "    next(reader)  # Skip the header\n",
    "    for row in reader:\n",
    "        words.append(row[0])  # Add the word from the first column of each row\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "133504\n"
     ]
    }
   ],
   "source": [
    "print(len(words))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def word_in_dictionary(input_string, dictionary): \n",
    "    for word in dictionary:\n",
    "        if word in input_string:\n",
    "            return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26/1837\n",
      "102/1837\n",
      "115/1837\n",
      "125/1837\n",
      "136/1837\n",
      "139/1837\n",
      "179/1837\n",
      "218/1837\n",
      "266/1837\n",
      "286/1837\n",
      "300/1837\n",
      "302/1837\n",
      "326/1837\n",
      "328/1837\n",
      "369/1837\n",
      "477/1837\n",
      "493/1837\n",
      "535/1837\n",
      "544/1837\n",
      "660/1837\n",
      "705/1837\n",
      "717/1837\n",
      "720/1837\n",
      "728/1837\n",
      "772/1837\n",
      "777/1837\n",
      "785/1837\n",
      "809/1837\n",
      "833/1837\n",
      "894/1837\n",
      "919/1837\n",
      "931/1837\n",
      "936/1837\n",
      "975/1837\n",
      "978/1837\n",
      "1042/1837\n",
      "1087/1837\n",
      "1126/1837\n",
      "1135/1837\n",
      "1154/1837\n",
      "1176/1837\n",
      "1204/1837\n",
      "1300/1837\n",
      "1332/1837\n",
      "1345/1837\n",
      "1350/1837\n",
      "1375/1837\n",
      "1402/1837\n",
      "1408/1837\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[39mfor\u001b[39;00m index, row \u001b[39min\u001b[39;00m for_keyword_filtering\u001b[39m.\u001b[39miterrows():\n\u001b[1;32m      4\u001b[0m     input_string \u001b[39m=\u001b[39m row[\u001b[39m'\u001b[39m\u001b[39mprompt\u001b[39m\u001b[39m'\u001b[39m]\n\u001b[0;32m----> 6\u001b[0m     \u001b[39mif\u001b[39;00m word_in_dictionary(input_string, words):\n\u001b[1;32m      7\u001b[0m         post_filtering_by_kw \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mconcat([post_filtering_by_kw, for_keyword_filtering\u001b[39m.\u001b[39mloc[[index]]], axis\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m)\n\u001b[1;32m      8\u001b[0m         track \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m \u001b[39m1\u001b[39m\n",
      "Cell \u001b[0;32mIn[14], line 3\u001b[0m, in \u001b[0;36mword_in_dictionary\u001b[0;34m(input_string, dictionary)\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mword_in_dictionary\u001b[39m(input_string, dictionary): \n\u001b[1;32m      2\u001b[0m     \u001b[39mfor\u001b[39;00m word \u001b[39min\u001b[39;00m dictionary:\n\u001b[0;32m----> 3\u001b[0m         \u001b[39mif\u001b[39;00m word \u001b[39min\u001b[39;00m input_string:\n\u001b[1;32m      4\u001b[0m             \u001b[39mreturn\u001b[39;00m \u001b[39mTrue\u001b[39;00m\n\u001b[1;32m      5\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mFalse\u001b[39;00m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "post_filtering_by_kw = pd.DataFrame()\n",
    "track = 0\n",
    "for index, row in for_keyword_filtering.iterrows():\n",
    "    input_string = row['prompt']\n",
    "    \n",
    "    if word_in_dictionary(input_string, words):\n",
    "        post_filtering_by_kw = pd.concat([post_filtering_by_kw, for_keyword_filtering.loc[[index]]], axis=0)\n",
    "        track += 1\n",
    "        if index % 25 == 0:\n",
    "            print(str(track) + '/' + str(len(for_keyword_filtering)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1837, 22)\n"
     ]
    }
   ],
   "source": [
    "print(post_filtering_by_kw.shape)\n",
    "input_for_gpt = post_filtering_by_kw\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "api_key = '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'\n",
    "enc = tk.encoding_for_model('gpt-3.5-turbo')\n",
    "\n",
    "data = input_for_gpt\n",
    "token_limit = 3500\n",
    "openai.api_key = api_key\n",
    "\n",
    "new_df = pd.DataFrame()\n",
    "gpt_outputs = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original Input Shape: (1837, 22)\n"
     ]
    }
   ],
   "source": [
    "print(\"Original Input Shape: \" + str(data.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1243, 23)\n",
      "Index(['totalNumLines', 'imports', 'numParameters', 'signature', 'lineStart',\n",
      "       'filePath', 'numChar', 'params', 'addImports', 'lineEnd', 'content',\n",
      "       'numCommentChars', 'numCommentLines', 'intraClassFieldsUsed',\n",
      "       'function_id', 'parentClass', 'comments', 'packageName', 'returnType',\n",
      "       'hash', 'prompt', 'good_length', 'Token Length'],\n",
      "      dtype='object')\n",
      "<class 'numpy.int64'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_702009/3989393089.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data['Token Length'] = tk_length\n",
      "/tmp/ipykernel_702009/3989393089.py:7: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  data['Token Length'] = data['Token Length'].astype(int)\n"
     ]
    }
   ],
   "source": [
    "tk_length = []\n",
    "for index, row in data.iterrows():\n",
    "    token = enc.encode(row['prompt'])\n",
    "    tk_length.append(len(token))\n",
    "\n",
    "data['Token Length'] = tk_length\n",
    "data['Token Length'] = data['Token Length'].astype(int)\n",
    "data = data[data['Token Length'] < 3500]\n",
    "\n",
    "# Filtered Data by Tokens\n",
    "print(data.shape)\n",
    "print(data.columns)\n",
    "print(type(data.iloc[3]['Token Length']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SVInterval {\\n    private final int contig;;\\n    private final int start;;\\n    private final int end;;\\n    public SVInterval(final int contig, final int start, final int end);\\n    public SVInterval(final int contig, final int start, final int end, final SVIntervalConstructorArgsValidator argsValidator);\\n    private SVInterval(final Kryo kryo, final Input input);\\n    private void serialize(final Kryo kryo, final Output output)\\n    public int getContig()\\n    public int getStart()\\n    public int getEnd()\\n    public SimpleInterval toSimpleInterval(final SAMSequenceDictionary sequenceDictionary)\\n    public int getLength()\\n    public SVLocation getStartLocation()\\n    public boolean overlaps(final SVInterval that)\\n    public boolean isDisjointFrom(final SVInterval that)\\n    public boolean isUpstreamOf(final SVInterval that)\\n    public boolean contains(final SVInterval that)\\n    public int gapLen(final SVInterval that)\\n    public int overlapLen(final SVInterval that)\\n    public SVInterval join(final SVInterval that)\\n    public SVInterval intersect(final SVInterval that)\\n    public int midpoint()\\n    public boolean equals(final Object obj)\\n    public int hashCode()\\n    public int compareTo(final SVInterval that)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/SVInterval.java', 'methodCount': 23, 'lineCount': 28}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class SVUtils {\\n    public static final String GATKSV_CONTIG_ALIGNMENTS_READ_GROUP_ID = \"GATKSVContigAlignments\";;\\n    public static String getSampleId(final SAMFileHeader header)\\n    public static SAMRecordComparator getSamRecordComparator(final SAMFileHeader.SortOrder sortOrder)\\n    public static List<String> getAttributeAsStringList(final VariantContext vc, final String attributeKey)\\n    public static Stream<String> getAttributeAsStringStream(final VariantContext vc, final String attributeKey)\\n    public static SimpleInterval makeOneBpInterval(final String chr, final int pos)\\n    public static Set<String> getCanonicalChromosomes(final String nonCanonicalContigNamesFile, @Nonnull final SAMSequenceDictionary dictionary)\\n    public static Integer getContigIDFromName(final String contigName, final SAMSequenceDictionary sequenceDictionary)\\n    public static SVInterval locatableToSVInterval(final Locatable loc, final SAMSequenceDictionary sequenceDictionary)\\n    public static int hashMapCapacity(final int nElements)\\n    public static int iteratorSize(final Iterator<T> itr)\\n    public static int iteratorSize(final LongIterator itr)\\n    public static Iterator<T> singletonIterator(final T t)\\n    public static Collection<SVKmer> uniquify(final Collection<SVKmer> coll1, final Collection<SVKmer> coll2)\\n    public static List<T> concatenateLists(final List<T> list1, final List<T> list2)\\n    public static Collector<T, ?, ArrayList<T>> arrayListCollector(final int size)\\n    public static EnumMap<T, Long> getZeroInitializedEnumMap(final Class<T> clazz)\\n    private static final long FNV64_DEFAULT_SEED = new BigInteger(\"14695981039346656037\").longValue();;\\n    public static long fnvLong64(final long toHash)\\n    public static long fnvLong64(long start, final long toHash)\\n    public static long fnvByteArray64(final byte[] toHash)\\n    public static long fnvByteArray64(long start, final byte[] toHash)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/tools/spark/sv/utils/SVUtils.java', 'methodCount': 20, 'lineCount': 24}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class IntHistogram {\\n    // the final slot (i.e., counts[counts.length-1]) accumulates the count of all observations greater than\\n// the maximum tracked value -- it\\'s the \"overflow\" bin\\nprivate final long[] counts;;\\n    private long totalObservations;;\\n    public IntHistogram(final int maxTrackedValue);\\n    private IntHistogram(final long[] counts, final long totalObservations);\\n    private IntHistogram(final Kryo kryo, final Input input);\\n    private void serialize(final Kryo kryo, final Output output)\\n    public void addObservation(final int observedValue)\\n    public void addObservations(final int observedValue, final long nObservations)\\n    public void addObservations(final IntHistogram intHistogram)\\n    public int getMaximumTrackedValue()\\n    public long getTotalObservations()\\n    public long getNObservations(final int observedValue)\\n    public void clear()\\n    public IntHistogram trim()\\n    public CDF getCDF()\\n    public AbstractIntegerDistribution empiricalDistribution(final int smoothing)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/tools/spark/utils/IntHistogram.java', 'methodCount': 14, 'lineCount': 18}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface GATKRead {\\n     String getName()\\n     int getFlags()\\n     void setName(final String name)\\n     int getLength()\\n     boolean isEmpty()\\n     void setPosition(final String contig, final int start)\\n     void setPosition(final Locatable locatable)\\n     String getAssignedContig()\\n     int getAssignedStart()\\n     int getUnclippedStart()\\n     int getUnclippedEnd()\\n     int getSoftStart()\\n     int getSoftEnd()\\n     int getAdaptorBoundary()\\n     String getMateContig()\\n     int getMateStart()\\n     void setMatePosition(final String contig, final int start)\\n     void setMatePosition(final Locatable locatable)\\n     int getFragmentLength()\\n     void setFragmentLength(final int fragmentLength)\\n     int getMappingQuality()\\n     void setMappingQuality(final int mappingQuality)\\n     byte[] getBases()\\n     byte[] getBasesNoCopy()\\n     int copyBases(final int offset, final byte[] destination, final int destinationOffset, final int maxLength)\\n     int copyBaseQualities(final int offset, final byte[] destination, final int destinationOffset, final int maxLength)\\n     boolean hasBases()\\n     boolean hasBaseQualities()\\n     byte getBase(final int i)\\n     String getBasesString()\\n     void setBases(final byte[] bases)\\n     byte[] getBaseQualities()\\n     byte[] getBaseQualitiesNoCopy()\\n     int getBaseQualityCount()\\n     byte getBaseQuality(final int i)\\n     void setBaseQualities(final byte[] baseQualities)\\n     Cigar getCigar()\\n     List<CigarElement> getCigarElements()\\n     CigarElement getCigarElement(final int i)\\n     int numCigarElements()\\n     void setCigar(final Cigar cigar)\\n     void setCigar(final String cigarString)\\n     String getReadGroup()\\n     void setReadGroup(final String readGroupID)\\n     boolean isPaired()\\n     void setIsPaired(final boolean isPaired)\\n     boolean isProperlyPaired()\\n     void setIsProperlyPaired(final boolean isProperlyPaired)\\n     boolean isUnmapped()\\n     void setIsUnmapped()\\n     boolean isUnplaced()\\n     void setIsUnplaced()\\n     boolean mateIsUnmapped()\\n     void setMateIsUnmapped()\\n     boolean mateIsUnplaced()\\n     void setMateIsUnplaced()\\n     boolean isReverseStrand()\\n     void setIsReverseStrand(final boolean isReverseStrand)\\n     boolean mateIsReverseStrand()\\n     void setMateIsReverseStrand(final boolean mateIsReverseStrand)\\n     boolean isFirstOfPair()\\n     void setIsFirstOfPair()\\n     boolean isSecondOfPair()\\n     void setIsSecondOfPair()\\n     boolean isSecondaryAlignment()\\n     void setIsSecondaryAlignment(final boolean isSecondaryAlignment)\\n     boolean isSupplementaryAlignment()\\n     void setIsSupplementaryAlignment(final boolean isSupplementaryAlignment)\\n     Optional<SamPairUtil.PairOrientation> getPairOrientation()\\n     boolean failsVendorQualityCheck()\\n     void setFailsVendorQualityCheck(final boolean failsVendorQualityCheck)\\n     boolean isDuplicate()\\n     void setIsDuplicate(final boolean isDuplicate)\\n     boolean hasAttribute(final String attributeName)\\n     Integer getAttributeAsInteger(final String attributeName)\\n     Float getAttributeAsFloat(final String attributeName)\\n     String getAttributeAsString(final String attributeName)\\n     byte[] getAttributeAsByteArray(final String attributeName)\\n     Object getTransientAttribute(final Object key)\\n     Optional<T> getOptionalTransientAttribute(final Object key, final Class<T> clazz)\\n     void setAttribute(final String attributeName, final Integer attributeValue)\\n     void setAttribute(final String attributeName, final String attributeValue)\\n     void setAttribute(final String attributeName, final byte[] attributeValue)\\n    public void setTransientAttribute(final Object key, final Object value)\\n     void clearAttribute(final String attributeName)\\n     void clearAttributes()\\n     void clearTransientAttribute(final String attributeName)\\n     GATKRead copy()\\n     GATKRead deepCopy()\\n     SAMRecord convertToSAMRecord(final SAMFileHeader header)\\n     String getSAMString()\\n    public void reverseComplement()\\n     String commonToString()\\n     void hardClipAttributes(final int newStart, final int newLength, int originalLength)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java', 'methodCount': 94, 'lineCount': 96}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'BreakpointEvidence': 'public class BreakpointEvidence {\\n    private static final SVInterval.Serializer intervalSerializer = new SVInterval.Serializer();;\\n    private final SVInterval location;;\\n    private final int weight;;\\n    // this piece of evidence is consistent with enough other evidence to be taken seriously\\nprivate boolean validated;;\\n    public BreakpointEvidence(final SVInterval location, final int weight, final boolean validated);\\n    protected BreakpointEvidence(final Kryo kryo, final Input input);\\n    public SVInterval getLocation()\\n    public int getWeight()\\n    public boolean isValidated()\\n    public void setValidated(final boolean validated)\\n    public Boolean isEvidenceUpstreamOfBreakpoint()\\n    public boolean hasDistalTargets(final ReadMetadata readMetadata, final int minEvidenceMapQ)\\n    public List<StrandedInterval> getDistalTargets(final ReadMetadata readMetadata, final int minEvidenceMapq)\\n    public String toString()\\n    public String stringRep(final ReadMetadata readMetadata, final int minEvidenceMapq)\\n     boolean equalFields(final BreakpointEvidence that)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    protected void serialize(final Kryo kryo, final Output output)\\n     static SVInterval fixedWidthInterval(final int contigID, final int contigOffset, final int offsetUncertainty)\\n}\\n'}\n",
      "/**\n",
      "     * Used in production to gather evidence derived from K-S calculation on windows.\n",
      "     */\n",
      "public void testReadAndGatherEvidence(final GATKRead read, final List<BreakpointEvidence> evidenceList)\n",
      "**************************************\n",
      "/**\n",
      " * Used in production to gather evidence derived from K-S calculation on windows.\n",
      " */\n",
      "public void testReadAndGatherEvidence(final GATKRead read, final List<BreakpointEvidence> evidenceList) {\n",
      "    if (!filter.isTemplateLenTestable(read))\n",
      "        return;\n",
      "    if (!isSameBlock(read)) {\n",
      "        checkHistograms(evidenceList);\n",
      "        advanceBlock(read);\n",
      "    }\n",
      "    addObservation(read);\n",
      "}\n",
      "2644\n",
      "210\n",
      "71\n",
      "86\n",
      "87\n",
      "90\n",
      "91\n",
      "523\n",
      "540\n",
      "544\n",
      "554\n",
      "569\n",
      "580\n",
      "583\n",
      "617\n",
      "635\n",
      "644\n",
      "666\n",
      "670\n",
      "672\n",
      "799\n",
      "823\n",
      "836\n",
      "843\n",
      "854\n",
      "881\n",
      "900\n",
      "902\n",
      "903\n",
      "904\n",
      "914\n",
      "928\n",
      "929\n",
      "936\n",
      "946\n",
      "1087\n",
      "1124\n",
      "1141\n",
      "1211\n",
      "1230\n",
      "1587\n",
      "1588\n",
      "1665\n",
      "1671\n",
      "1804\n",
      "2393\n",
      "2394\n",
      "2472\n",
      "2571\n",
      "2645\n",
      "2679\n",
      "2680\n",
      "2681\n",
      "2690\n",
      "2718\n",
      "2740\n",
      "2768\n",
      "2805\n",
      "2827\n",
      "2863\n",
      "2891\n",
      "2910\n",
      "2921\n",
      "2940\n",
      "2959\n",
      "2964\n",
      "2993\n",
      "3006\n",
      "3007\n",
      "3012\n",
      "3013\n",
      "3141\n",
      "3169\n",
      "3791\n",
      "3827\n",
      "3828\n",
      "4021\n",
      "4458\n",
      "4474\n",
      "4531\n",
      "4595\n",
      "4602\n",
      "4683\n",
      "4926\n",
      "4970\n",
      "4992\n",
      "5437\n",
      "5468\n",
      "6106\n",
      "6114\n",
      "6244\n",
      "6309\n",
      "6665\n",
      "6700\n",
      "6706\n",
      "6721\n",
      "6811\n",
      "6813\n",
      "6815\n",
      "6817\n",
      "6819\n",
      "6821\n",
      "6823\n",
      "6825\n",
      "6827\n",
      "6830\n",
      "6832\n",
      "6844\n",
      "6845\n",
      "6858\n",
      "6868\n",
      "6871\n",
      "6884\n",
      "6921\n",
      "6950\n",
      "6973\n",
      "6974\n",
      "6975\n",
      "7013\n",
      "7014\n",
      "7021\n",
      "7045\n",
      "7083\n",
      "7101\n",
      "7162\n",
      "7165\n",
      "7194\n",
      "7198\n",
      "7199\n",
      "7216\n",
      "7260\n",
      "7396\n",
      "7423\n",
      "7438\n",
      "7457\n",
      "7463\n",
      "7495\n",
      "7496\n",
      "7533\n",
      "7534\n",
      "7551\n",
      "7552\n",
      "7590\n",
      "7620\n",
      "7622\n",
      "7629\n",
      "7631\n",
      "7641\n",
      "7642\n",
      "7644\n",
      "7649\n",
      "7657\n",
      "7659\n",
      "7660\n",
      "7665\n",
      "7673\n",
      "7674\n",
      "7675\n",
      "7676\n",
      "7677\n",
      "7696\n",
      "7704\n",
      "7730\n",
      "7735\n",
      "7759\n",
      "7761\n",
      "7763\n",
      "7772\n",
      "7780\n",
      "7782\n",
      "7788\n",
      "7831\n",
      "7834\n",
      "7839\n",
      "7844\n",
      "7845\n",
      "7846\n",
      "7906\n",
      "7951\n",
      "8018\n",
      "8039\n",
      "8044\n",
      "8067\n",
      "8068\n",
      "8116\n",
      "8127\n",
      "8132\n",
      "8171\n",
      "8185\n",
      "8198\n",
      "8208\n",
      "8209\n",
      "8277\n",
      "8278\n",
      "8287\n",
      "8291\n",
      "8300\n",
      "8326\n",
      "8417\n",
      "8420\n",
      "8487\n",
      "8502\n",
      "8548\n",
      "8564\n",
      "8585\n",
      "8595\n",
      "8653\n",
      "8687\n",
      "8910\n",
      "8914\n",
      "9033\n",
      "9074\n",
      "9078\n",
      "9088\n",
      "9103\n",
      "9108\n",
      "9110\n",
      "9158\n",
      "9164\n",
      "9166\n",
      "9281\n",
      "9309\n",
      "9374\n",
      "9445\n",
      "9465\n",
      "9484\n",
      "9486\n",
      "9493\n",
      "9512\n",
      "9513\n",
      "9514\n",
      "9515\n",
      "9516\n",
      "9544\n",
      "9583\n",
      "9584\n",
      "9594\n",
      "9597\n",
      "9625\n",
      "9626\n",
      "9945\n",
      "9946\n",
      "9947\n",
      "9948\n",
      "9949\n",
      "9988\n",
      "9999\n",
      "10001\n",
      "10033\n",
      "10039\n",
      "10048\n",
      "10068\n",
      "10164\n",
      "10233\n",
      "10242\n",
      "10247\n",
      "10291\n",
      "10319\n",
      "10429\n",
      "10448\n",
      "10513\n",
      "10556\n",
      "10575\n",
      "10580\n",
      "10625\n",
      "10679\n",
      "10812\n",
      "10827\n",
      "10832\n",
      "10834\n",
      "10836\n",
      "10837\n",
      "10846\n",
      "10851\n",
      "10854\n",
      "10860\n",
      "10870\n",
      "10882\n",
      "10891\n",
      "10912\n",
      "10927\n",
      "10936\n",
      "10937\n",
      "10940\n",
      "10943\n",
      "10990\n",
      "10997\n",
      "10998\n",
      "11025\n",
      "11097\n",
      "11125\n",
      "11132\n",
      "11698\n",
      "11895\n",
      "12169\n",
      "12171\n",
      "12172\n",
      "12174\n",
      "12230\n",
      "12231\n",
      "12473\n",
      "13064\n",
      "13193\n",
      "13578\n",
      "13701\n",
      "13702\n",
      "13708\n",
      "13709\n",
      "13710\n",
      "13711\n",
      "13712\n",
      "13717\n",
      "13787\n",
      "13867\n",
      "13934\n",
      "13941\n",
      "13942\n",
      "13958\n",
      "14453\n",
      "15158\n",
      "15341\n",
      "15354\n",
      "15423\n",
      "15481\n",
      "15494\n",
      "15532\n",
      "15576\n",
      "15587\n",
      "15598\n",
      "15624\n",
      "15625\n",
      "15626\n",
      "15671\n",
      "15679\n",
      "15754\n",
      "15824\n",
      "15883\n",
      "15885\n",
      "15891\n",
      "15892\n",
      "15928\n",
      "15930\n",
      "16052\n",
      "16090\n",
      "16178\n",
      "16181\n",
      "16236\n",
      "16379\n",
      "16473\n",
      "16480\n",
      "16506\n",
      "16509\n",
      "16529\n",
      "16530\n",
      "16531\n",
      "16582\n",
      "16668\n",
      "16707\n",
      "16785\n",
      "16953\n",
      "16956\n",
      "16987\n",
      "16988\n",
      "17099\n",
      "17218\n",
      "19171\n",
      "19172\n",
      "19176\n",
      "19515\n",
      "19519\n",
      "19556\n",
      "19560\n",
      "19567\n",
      "19568\n",
      "19580\n",
      "19622\n",
      "19623\n",
      "19626\n",
      "19640\n",
      "19644\n",
      "19703\n",
      "19704\n",
      "19712\n",
      "19714\n",
      "19765\n",
      "19772\n",
      "19773\n",
      "19783\n",
      "19786\n",
      "19787\n",
      "19788\n",
      "19803\n",
      "19883\n",
      "19890\n",
      "19891\n",
      "19896\n",
      "19905\n",
      "19912\n",
      "19917\n",
      "19921\n",
      "19922\n",
      "19924\n",
      "19929\n",
      "20004\n",
      "20013\n",
      "20089\n",
      "20212\n",
      "20214\n",
      "20221\n",
      "20222\n",
      "20223\n",
      "20224\n",
      "20225\n",
      "20226\n",
      "20230\n",
      "20239\n",
      "20240\n",
      "20247\n",
      "20257\n",
      "20258\n",
      "20263\n",
      "20280\n",
      "20289\n",
      "20294\n",
      "20437\n",
      "20454\n",
      "20461\n",
      "20493\n",
      "20502\n",
      "20503\n",
      "20504\n",
      "20505\n",
      "20515\n",
      "20516\n",
      "20519\n",
      "20548\n",
      "20756\n",
      "20760\n",
      "20761\n",
      "20764\n",
      "20765\n",
      "20768\n",
      "20769\n",
      "20772\n",
      "20773\n",
      "20776\n",
      "20777\n",
      "20780\n",
      "20781\n",
      "20784\n",
      "20785\n",
      "20787\n",
      "20825\n",
      "20986\n",
      "21511\n",
      "21618\n",
      "21626\n",
      "21627\n",
      "21631\n",
      "21633\n",
      "21693\n",
      "21700\n",
      "21711\n",
      "21782\n",
      "21796\n",
      "21797\n",
      "21818\n",
      "21824\n",
      "21829\n",
      "21839\n",
      "21840\n",
      "21841\n",
      "21859\n",
      "21888\n",
      "21899\n",
      "21908\n",
      "21910\n",
      "21928\n",
      "21958\n",
      "21978\n",
      "22002\n",
      "22029\n",
      "22037\n",
      "22041\n",
      "22063\n",
      "22082\n",
      "22185\n",
      "22407\n",
      "22448\n",
      "22507\n",
      "23168\n",
      "23183\n",
      "23189\n",
      "23190\n",
      "23191\n",
      "23197\n",
      "23198\n",
      "23216\n",
      "23227\n",
      "23265\n",
      "23295\n",
      "23302\n",
      "23382\n",
      "23774\n",
      "23785\n",
      "23890\n",
      "23911\n",
      "23920\n",
      "23922\n",
      "23931\n",
      "24021\n",
      "24036\n",
      "24438\n",
      "24511\n",
      "24575\n",
      "24583\n",
      "24592\n",
      "24616\n",
      "24705\n",
      "24707\n",
      "24709\n",
      "24846\n",
      "24848\n",
      "25027\n",
      "25091\n",
      "25093\n",
      "25103\n",
      "25108\n",
      "25134\n",
      "25161\n",
      "25218\n",
      "25251\n",
      "25258\n",
      "25308\n",
      "25309\n",
      "25373\n",
      "25477\n",
      "25671\n",
      "25672\n",
      "25707\n",
      "25743\n",
      "25746\n",
      "25752\n",
      "25757\n",
      "25758\n",
      "25760\n",
      "25825\n",
      "25847\n",
      "25915\n",
      "25925\n",
      "25932\n",
      "25956\n",
      "25958\n",
      "26010\n",
      "26012\n",
      "26039\n",
      "26172\n",
      "26177\n",
      "26245\n",
      "26265\n",
      "26303\n",
      "26338\n",
      "26339\n",
      "26412\n",
      "26424\n",
      "26510\n",
      "26511\n",
      "26682\n",
      "26697\n",
      "26754\n",
      "26757\n",
      "26840\n",
      "26843\n",
      "27151\n",
      "27250\n",
      "27287\n",
      "27324\n",
      "27355\n",
      "27383\n",
      "27401\n",
      "27402\n",
      "27418\n",
      "27499\n",
      "27562\n",
      "27577\n",
      "27622\n",
      "27623\n",
      "27651\n",
      "27689\n",
      "27713\n",
      "27714\n",
      "27815\n",
      "27824\n",
      "27829\n",
      "27851\n",
      "27925\n",
      "27927\n",
      "27956\n",
      "27962\n",
      "27965\n",
      "27968\n",
      "27988\n",
      "27990\n",
      "28021\n",
      "28046\n",
      "28052\n",
      "28171\n",
      "28176\n",
      "28202\n",
      "28209\n",
      "28211\n",
      "28212\n",
      "28239\n",
      "28240\n",
      "28255\n",
      "28256\n",
      "28258\n",
      "28286\n",
      "28297\n",
      "28300\n",
      "28322\n",
      "28326\n",
      "28327\n",
      "28330\n",
      "28335\n",
      "28339\n",
      "28340\n",
      "28342\n",
      "28344\n",
      "28346\n",
      "28355\n",
      "28381\n",
      "28484\n",
      "28485\n",
      "28487\n",
      "28582\n",
      "28583\n",
      "28606\n",
      "28683\n",
      "28684\n",
      "28685\n",
      "28688\n",
      "28689\n",
      "28691\n",
      "28692\n",
      "28693\n",
      "28706\n",
      "28709\n",
      "28711\n",
      "28743\n",
      "28746\n",
      "28823\n",
      "28826\n",
      "28827\n",
      "28896\n",
      "28977\n",
      "28994\n",
      "29001\n",
      "29069\n",
      "29070\n",
      "29072\n",
      "29074\n",
      "29077\n",
      "29215\n",
      "29216\n",
      "29217\n",
      "29219\n",
      "29220\n",
      "29221\n",
      "29222\n",
      "29223\n",
      "29224\n",
      "29236\n",
      "29335\n",
      "29350\n",
      "29362\n",
      "29373\n",
      "29418\n",
      "29444\n",
      "29530\n",
      "29531\n",
      "29532\n",
      "29541\n",
      "29544\n",
      "29547\n",
      "29565\n",
      "29583\n",
      "29584\n",
      "29592\n",
      "29595\n",
      "29601\n",
      "29614\n",
      "29619\n",
      "29620\n",
      "29621\n",
      "29622\n",
      "29629\n",
      "29630\n",
      "29631\n",
      "29650\n",
      "29651\n",
      "29652\n",
      "29653\n",
      "29654\n",
      "29655\n",
      "29662\n",
      "29713\n",
      "29715\n",
      "29747\n",
      "29757\n",
      "29786\n",
      "29800\n",
      "29815\n",
      "29864\n",
      "29866\n",
      "29868\n",
      "29875\n",
      "29878\n",
      "29879\n",
      "29884\n",
      "29891\n",
      "29901\n",
      "29907\n",
      "29908\n",
      "29943\n",
      "29968\n",
      "29978\n",
      "29989\n",
      "29992\n",
      "30003\n",
      "30209\n",
      "30215\n",
      "30217\n",
      "30223\n",
      "30231\n",
      "30234\n",
      "30237\n",
      "30238\n",
      "30239\n",
      "30242\n",
      "30243\n",
      "30245\n",
      "30246\n",
      "30247\n",
      "30255\n",
      "30258\n",
      "30288\n",
      "30298\n",
      "30311\n",
      "30339\n",
      "30356\n",
      "30366\n",
      "30391\n",
      "30424\n",
      "30425\n",
      "30429\n",
      "30448\n",
      "30450\n",
      "30456\n",
      "30474\n",
      "30521\n",
      "30544\n",
      "30547\n",
      "30548\n",
      "30549\n",
      "30551\n",
      "30577\n",
      "30638\n",
      "30641\n",
      "30686\n",
      "30710\n",
      "30842\n",
      "30860\n",
      "30884\n",
      "30912\n",
      "31088\n",
      "31123\n",
      "31351\n",
      "31371\n",
      "31415\n",
      "31421\n",
      "31424\n",
      "31455\n",
      "31461\n",
      "31468\n",
      "31564\n",
      "31580\n",
      "31611\n",
      "31613\n",
      "31632\n",
      "31633\n",
      "31664\n",
      "31686\n",
      "31687\n",
      "31724\n",
      "31776\n",
      "31855\n",
      "31859\n",
      "31863\n",
      "31867\n",
      "31898\n",
      "31948\n",
      "32012\n",
      "32060\n",
      "32077\n",
      "32081\n",
      "32163\n",
      "32226\n",
      "32286\n",
      "32404\n",
      "32459\n",
      "32521\n",
      "32543\n",
      "32552\n",
      "32558\n",
      "32560\n",
      "32564\n",
      "32565\n",
      "32567\n",
      "32568\n",
      "32573\n",
      "32575\n",
      "32577\n",
      "32578\n",
      "32579\n",
      "32612\n",
      "32637\n",
      "32667\n",
      "32673\n",
      "32702\n",
      "32706\n",
      "32730\n",
      "32740\n",
      "32789\n",
      "32805\n",
      "32840\n",
      "32865\n",
      "32920\n",
      "33039\n",
      "33173\n",
      "33213\n",
      "33278\n",
      "33323\n",
      "33353\n",
      "33354\n",
      "33360\n",
      "33362\n",
      "33370\n",
      "33416\n",
      "33492\n",
      "33538\n",
      "33561\n",
      "33587\n",
      "33593\n",
      "33634\n",
      "33646\n",
      "33648\n",
      "33663\n",
      "33668\n",
      "33691\n",
      "33719\n",
      "33720\n",
      "33733\n",
      "33743\n",
      "33753\n",
      "33788\n",
      "33800\n",
      "33805\n",
      "33808\n",
      "33815\n",
      "33816\n",
      "33840\n",
      "33879\n",
      "33889\n",
      "33969\n",
      "33979\n",
      "34005\n",
      "34006\n",
      "34078\n",
      "34124\n",
      "34156\n",
      "34409\n",
      "34411\n",
      "34412\n",
      "34430\n",
      "34499\n",
      "34663\n",
      "34706\n",
      "35046\n",
      "35124\n",
      "35755\n",
      "35962\n",
      "36006\n",
      "36073\n",
      "36194\n",
      "36440\n",
      "36457\n",
      "36459\n",
      "36718\n",
      "36945\n",
      "36960\n",
      "36971\n",
      "37029\n",
      "37030\n",
      "37222\n",
      "37223\n",
      "37225\n",
      "37290\n",
      "37296\n",
      "37378\n",
      "38166\n",
      "38168\n",
      "38515\n",
      "39824\n",
      "40129\n",
      "40184\n",
      "40215\n",
      "40285\n",
      "40286\n",
      "40288\n",
      "40406\n",
      "40447\n",
      "40546\n",
      "40962\n",
      "41106\n",
      "41120\n",
      "41134\n",
      "41156\n",
      "41587\n",
      "41655\n",
      "41656\n",
      "41674\n",
      "41678\n",
      "41680\n",
      "41687\n",
      "41727\n",
      "41736\n",
      "41831\n",
      "42022\n",
      "42078\n",
      "42243\n",
      "42253\n",
      "42384\n",
      "42387\n",
      "42388\n",
      "42389\n",
      "42494\n",
      "42502\n",
      "42539\n",
      "42581\n",
      "42582\n",
      "42668\n",
      "42671\n",
      "42683\n",
      "42694\n",
      "42870\n",
      "42876\n",
      "42878\n",
      "42880\n",
      "42902\n",
      "42908\n",
      "42931\n",
      "42976\n",
      "43028\n",
      "43058\n",
      "43112\n",
      "43113\n",
      "43163\n",
      "43201\n",
      "43208\n",
      "43249\n",
      "43264\n",
      "43274\n",
      "43283\n",
      "43338\n",
      "43615\n",
      "43630\n",
      "43640\n",
      "43677\n",
      "43713\n",
      "43715\n",
      "43725\n",
      "43767\n",
      "43771\n",
      "43774\n",
      "43778\n",
      "43791\n",
      "43828\n",
      "43835\n",
      "43836\n",
      "43840\n",
      "43859\n",
      "43862\n",
      "43863\n",
      "43927\n",
      "43928\n",
      "43929\n",
      "43948\n",
      "44092\n",
      "44108\n",
      "44111\n",
      "44162\n",
      "44177\n",
      "44212\n",
      "44231\n",
      "44285\n",
      "44311\n",
      "44324\n",
      "44350\n",
      "44388\n",
      "44389\n",
      "44428\n",
      "44462\n",
      "44484\n",
      "44503\n",
      "44512\n",
      "44540\n",
      "44542\n",
      "44543\n",
      "44544\n",
      "44550\n",
      "44552\n",
      "44611\n",
      "44633\n",
      "44726\n",
      "44765\n",
      "44861\n",
      "45096\n",
      "45111\n",
      "45113\n",
      "45117\n",
      "45209\n",
      "45291\n",
      "45339\n",
      "45369\n",
      "45371\n",
      "45407\n",
      "45567\n",
      "45570\n",
      "45576\n",
      "45600\n",
      "45601\n",
      "45608\n",
      "45893\n",
      "46026\n",
      "46113\n",
      "46164\n",
      "46265\n",
      "46266\n",
      "46346\n",
      "46369\n",
      "46375\n",
      "46412\n",
      "46414\n",
      "46422\n",
      "46423\n",
      "46525\n",
      "46528\n",
      "46554\n",
      "46569\n",
      "46602\n",
      "46682\n",
      "46737\n",
      "46749\n",
      "46758\n",
      "46852\n",
      "46902\n",
      "46932\n",
      "46933\n",
      "47156\n",
      "47177\n",
      "47178\n",
      "47188\n",
      "47219\n",
      "47239\n",
      "47269\n",
      "47281\n",
      "47320\n",
      "47321\n",
      "47322\n",
      "47408\n",
      "47415\n",
      "47434\n",
      "47437\n",
      "47460\n",
      "47462\n",
      "47538\n",
      "47548\n",
      "47571\n",
      "47733\n",
      "47748\n",
      "47751\n",
      "47765\n",
      "47766\n",
      "47883\n",
      "47887\n",
      "47978\n",
      "47997\n",
      "48020\n",
      "48033\n",
      "48088\n",
      "48089\n",
      "48097\n",
      "48102\n",
      "48214\n",
      "48215\n",
      "48236\n",
      "48351\n",
      "48352\n",
      "48355\n",
      "48359\n",
      "48374\n",
      "48393\n",
      "48395\n",
      "48396\n",
      "48397\n",
      "48402\n",
      "48403\n",
      "48409\n",
      "48413\n",
      "48445\n",
      "48457\n",
      "48466\n",
      "48475\n",
      "48534\n",
      "48550\n",
      "48657\n",
      "48702\n",
      "48712\n",
      "48723\n",
      "48725\n",
      "48738\n",
      "48750\n",
      "48763\n",
      "48772\n",
      "48797\n",
      "48837\n",
      "48864\n",
      "49033\n",
      "49225\n",
      "49234\n",
      "49286\n",
      "49309\n",
      "49312\n",
      "49395\n",
      "49537\n",
      "49636\n",
      "49655\n",
      "49659\n",
      "49669\n",
      "49820\n",
      "49825\n",
      "49840\n",
      "49874\n",
      "49877\n",
      "49887\n",
      "49888\n",
      "50064\n",
      "50245\n",
      "50339\n",
      "50369\n",
      "50410\n",
      "50461\n",
      "50466\n",
      "50517\n",
      "50520\n",
      "50530\n",
      "50534\n",
      "50670\n",
      "50676\n",
      "50693\n",
      "50756\n",
      "50770\n",
      "50775\n",
      "50791\n",
      "50866\n",
      "50867\n",
      "50981\n",
      "51032\n",
      "51034\n",
      "51039\n",
      "51040\n",
      "51065\n",
      "51069\n",
      "51075\n",
      "51107\n",
      "51229\n",
      "51235\n",
      "51251\n",
      "51296\n",
      "51300\n",
      "51321\n",
      "51322\n",
      "51326\n",
      "51335\n",
      "51441\n",
      "51459\n",
      "51611\n",
      "51789\n",
      "51803\n",
      "51914\n",
      "51934\n",
      "51947\n",
      "51948\n",
      "52333\n",
      "52807\n",
      "52820\n",
      "53018\n",
      "53019\n",
      "53231\n",
      "53489\n",
      "54361\n",
      "54820\n",
      "54939\n",
      "54943\n",
      "55169\n",
      "55191\n",
      "55209\n",
      "55210\n",
      "55220\n",
      "55229\n",
      "55230\n",
      "55273\n",
      "55644\n",
      "55649\n",
      "56106\n",
      "56436\n",
      "56476\n",
      "56477\n",
      "56483\n",
      "56486\n",
      "56489\n",
      "56499\n",
      "56604\n",
      "57263\n",
      "57280\n",
      "57281\n",
      "57304\n",
      "57319\n",
      "57461\n",
      "57625\n",
      "57779\n",
      "57834\n",
      "58158\n",
      "58159\n",
      "57625    {'classCount': 0, 'imports': [{'fieldCount': 1...\n",
      "57779    {'classCount': 0, 'imports': [{'fieldCount': 3...\n",
      "57834    {'classCount': 0, 'imports': [{'fieldCount': 1...\n",
      "58158    {'classCount': 2, 'imports': [{'fieldCount': 0...\n",
      "58159    {'classCount': 2, 'imports': [{'fieldCount': 0...\n",
      "Name: prompt, dtype: object\n",
      "57625    1171\n",
      "57779    2644\n",
      "57834    3391\n",
      "58158    2083\n",
      "58159    2024\n",
      "Name: Token Length, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(str(data.loc[57779]['prompt']))\n",
    "print((data.loc[57779]['Token Length']))\n",
    "print((data.iloc[0]['Token Length']))\n",
    "for i in range(len(data)):\n",
    "    print(data.index[i])\n",
    "\n",
    "print(data.tail(5)['prompt'])\n",
    "print(data.tail(5)['Token Length'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 1/1243 [00:01<37:43,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 2/1243 [00:03<34:43,  1.68s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 3/1243 [00:05<38:33,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "87\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Creates a new instance of the service with the specified name if it can be found, otherwise return null\n",
      "     * @param name The identifier name for the service to create\n",
      "     * @param ces The conan executor service which should be injected into the created service\n",
      "     * @return A new service, with conan executor service injected inside it\n",
      "     */\n",
      "public T create(String name, ConanExecutorService ces)\n",
      "**************************************\n",
      "public T create(String name, ConanExecutorService ces) {\n",
      "    for (T service : loader) {\n",
      "        if (this.acceptsIdentifier(name, service)) {\n",
      "            ((AbstractConanProcess) service).setConanProcessService(ces.getConanProcessService());\n",
      "            return service;\n",
      "        }\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "192\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 4/1243 [00:07<41:18,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "90\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 5/1243 [00:09<39:31,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "91\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   0%|          | 6/1243 [00:11<37:51,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "523\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 7/1243 [00:12<35:46,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "540\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 8/1243 [00:14<38:27,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "544\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 8}\n",
      "{'MultiTypeTree': 'public class MultiTypeTree {\\n    /*\\n     * Inputs:\\n     */\\npublic Input<String> typeLabelInput = new Input<>(\"typeLabel\", \"Label for type traits (default \\'type\\')\", \"type\");;\\n    public Input<TraitSet> typeTraitInput = new Input<>(\"typeTrait\", \"Type trait set.  Used only by BEAUti.\");;\\n    public Input<TypeSet> typeSetInput = new Input<>(\"typeSet\", \"Type set input.\");;\\n    /*\\n     * Non-input fields:\\n     */\\nprotected String typeLabel;;\\n    protected TraitSet typeTraitSet;;\\n    protected TypeSet typeSet;;\\n    public MultiTypeTree();\\n    public MultiTypeTree(Node rootNode);\\n    public void initAndValidate()\\n    protected void processTraits(List<TraitSet> traitList)\\n    public TraitSet getTypeTrait()\\n    public TypeSet getTypeSet()\\n    public boolean hasTypeTrait()\\n    public void setTypeTrait(TraitSet traitSet)\\n    public String getTypeLabel()\\n    public final void initArrays()\\n    private void listNodes(MultiTypeNode node, MultiTypeNode[] nodes)\\n    public MultiTypeTree copy()\\n    public void assignFrom(StateNode other)\\n    public void assignFromFragile(StateNode other)\\n    private void assignFromFragileHelper(int iStart, int iEnd, Node[] otherNodes)\\n    public boolean isValid()\\n    private boolean timesAreValid(Node node)\\n    private boolean typesAreValid(Node node)\\n    public Tree getFlattenedTree(boolean useTypeStrings)\\n    public void initFromFlatTree(Tree flatTree, boolean takeNrsFromFlatTree)\\n    private int numberInternalNodes(Node node, int nextNr)\\n    public int getTotalNumberOfChanges()\\n    public String toString()\\n    protected void store()\\n    private void storeNodes(int iStart, int iEnd)\\n    public void init(PrintStream printStream)\\n    public void log(long i, PrintStream printStream)\\n    public void close(PrintStream printStream)\\n    public void fromXML(org.w3c.dom.Node node)\\n}\\n'}\n",
      "/**\n",
      "     * Deep copy, returns a completely new multi-type tree.\n",
      "     *\n",
      "     * @return a deep copy of this multi-type tree\n",
      "     */\n",
      "public MultiTypeTree copy()\n",
      "**************************************\n",
      "@Override\n",
      "public MultiTypeTree copy() {\n",
      "    MultiTypeTree tree = new MultiTypeTree();\n",
      "    tree.ID = ID;\n",
      "    tree.index = index;\n",
      "    tree.root = root.copy();\n",
      "    tree.nodeCount = nodeCount;\n",
      "    tree.internalNodeCount = internalNodeCount;\n",
      "    tree.leafNodeCount = leafNodeCount;\n",
      "    tree.typeLabel = typeLabel;\n",
      "    return tree;\n",
      "}\n",
      "627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 9/1243 [00:16<36:47,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "554\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 10/1243 [00:18<36:16,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "569\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 11/1243 [00:20<38:45,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "580\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 12/1243 [00:21<37:04,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "583\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 13/1243 [00:23<35:40,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "617\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class MultiTypeNode {\\n    // Type metadata:\\nint nTypeChanges = 0;;\\n    List<Integer> changeTypes = new ArrayList<Integer>();;\\n    List<Double> changeTimes = new ArrayList<Double>();;\\n    int nodeType = 0;;\\n    public int getChangeCount()\\n    public int getChangeType(int idx)\\n    public double getChangeTime(int idx)\\n    public int getFinalType()\\n    public double getFinalChangeTime()\\n    public int getNodeType()\\n    public void setNodeType(int nodeType)\\n    public void addChange(int newType, double time)\\n    public void clearChanges()\\n    public void setChangeTime(int idx, double newTime)\\n    public void setChangeType(int idx, int newType)\\n    public void truncateChanges(int newNChanges)\\n    public void insertChange(int idx, int newType, double newTime)\\n    public void removeChange(int idx)\\n    public MultiTypeNode shallowCopy()\\n    public MultiTypeNode copy()\\n    public void assignFrom(Node[] nodes, Node node)\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,MultiTypeTree/src/multitypetree/evolution/tree/MultiTypeNode.java', 'methodCount': 17, 'lineCount': 23}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Choose a change type uniformly at random from the types not found\n",
      "     * in the this.illegalTypes set.\n",
      "     * \n",
      "     * @return chosen type\n",
      "     */\n",
      "private int selectLegalChangeType()\n",
      "**************************************\n",
      "private int selectLegalChangeType() {\n",
      "    int n = Randomizer.nextInt(migModel.getNTypes() - illegalTypes.size());\n",
      "    int changeType;\n",
      "    for (changeType = 0; changeType < migModel.getNTypes(); changeType++) {\n",
      "        if (illegalTypes.contains(changeType))\n",
      "            continue;\n",
      "        if (n == 0)\n",
      "            break;\n",
      "        n -= 1;\n",
      "    }\n",
      "    return changeType;\n",
      "}\n",
      "476\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 14/1243 [00:25<38:35,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "635\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|          | 15/1243 [00:27<40:25,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "644\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|         | 16/1243 [00:30<41:38,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "666\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|         | 17/1243 [00:32<41:58,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "670\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   1%|         | 18/1243 [00:33<38:57,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "672\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class MultiTypeNode {\\n    // Type metadata:\\nint nTypeChanges = 0;;\\n    List<Integer> changeTypes = new ArrayList<Integer>();;\\n    List<Double> changeTimes = new ArrayList<Double>();;\\n    int nodeType = 0;;\\n    public int getChangeCount()\\n    public int getChangeType(int idx)\\n    public double getChangeTime(int idx)\\n    public int getFinalType()\\n    public double getFinalChangeTime()\\n    public int getNodeType()\\n    public void setNodeType(int nodeType)\\n    public void addChange(int newType, double time)\\n    public void clearChanges()\\n    public void setChangeTime(int idx, double newTime)\\n    public void setChangeType(int idx, int newType)\\n    public void truncateChanges(int newNChanges)\\n    public void insertChange(int idx, int newType, double newTime)\\n    public void removeChange(int idx)\\n    public MultiTypeNode shallowCopy()\\n    public MultiTypeNode copy()\\n    public void assignFrom(Node[] nodes, Node node)\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,MultiTypeTree/src/multitypetree/evolution/tree/MultiTypeNode.java', 'methodCount': 17, 'lineCount': 23}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Obtain joint probability of typing along branches between srcNode and\n",
      "     * the root, the sister of srcNode and the root, and the node type of the\n",
      "     * root.\n",
      "     *\n",
      "     * @param srcNode\n",
      "     * @return\n",
      "     */\n",
      "protected double getRootBranchTypeProb(Node srcNode)\n",
      "**************************************\n",
      "protected double getRootBranchTypeProb(Node srcNode) {\n",
      "    double logProb = 0.0;\n",
      "    Node srcNodeP = srcNode.getParent();\n",
      "    Node srcNodeS = getOtherChild(srcNodeP, srcNode);\n",
      "    logProb += Math.log(1.0 / migModel.getNTypes());\n",
      "    logProb += getBranchTypeProb(srcNode);\n",
      "    logProb += getBranchTypeProb(srcNodeS);\n",
      "    return logProb;\n",
      "}\n",
      "501\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 19/1243 [00:35<39:55,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "799\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 20/1243 [00:38<41:24,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "823\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 21/1243 [00:40<42:02,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "836\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 22/1243 [00:41<39:33,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "843\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 23/1243 [00:43<39:48,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "854\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class Tuple {\\n    private Tuple();\\n    public static Tuple2<T1, T2> tuple(T1 f, T2 s)\\n    public static Tuple3<T1, T2, T3> tuple(T1 f, T2 s, T3 t)\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/collection/Tuple.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ReferenceResource {\\n    /**\\n * Fasta files store in thread local variables to avoid multithreading issues.\\n */\\nprivate ThreadLocal<Map<String, IndexedFastaSequenceFile>> threadLocalFastaFiles = ThreadLocal.withInitial(HashMap::new);;\\n    private synchronized IndexedFastaSequenceFile fetchFasta(String file)\\n    public String[] retrieveSubSeq(String fasta, String chr, int start, int end)\\n    public Reference getReference(Region region)\\n    public Reference getReference(Region region, int extension, Reference ref)\\n    private Reference addPositionsToSeedSequence(Reference ref, int sequenceStart, int i, String keySequence)\\n    public static boolean isLoaded(String chr, int sequenceStart, int sequenceEnd, Reference reference)\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/data/ReferenceResource.java', 'methodCount': 6, 'lineCount': 9}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class Region {\\n    /**\\n * Chromosome name\\n */\\npublic final String chr;;\\n    /**\\n * Region start position\\n */\\npublic final int start;;\\n    /**\\n * Region end position\\n */\\npublic final int end;;\\n    /**\\n * Gene name\\n */\\npublic final String gene;;\\n    /**\\n * Position to start looking for variants (for amplicon based calling)\\n */\\npublic final int insertStart;;\\n    /**\\n * Position to end looking for variants (for amplicon based calling)\\n */\\npublic final int insertEnd;;\\n    public Region(String chr, int start, int end, String gene);\\n    public Region(String chr, int start, int end, String gene, int insertStart, int insertEnd);\\n    public static Region newModifiedRegion(Region region, int changedStart, int changedEnd)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String printRegion()\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/data/Region.java', 'methodCount': 7, 'lineCount': 15}, {'fieldCount': 10, 'classCount': 1, 'methodsOnly': 'public class GlobalReadOnlyScope {\\n    private volatile static GlobalReadOnlyScope instance;;\\n    public static GlobalReadOnlyScope instance()\\n    public static synchronized void init(Configuration conf, Map<String, Integer> chrLengths, String sample, String samplem, String ampliconBasedCalling, Map<String, Integer> adaptorForward, Map<String, Integer> adaptorReverse)\\n    private volatile static AbstractMode mode;;\\n    public static AbstractMode getMode()\\n    public static synchronized void setMode(AbstractMode runMode)\\n    public static synchronized void clear()\\n    public final Configuration conf;;\\n    public final Map<String, Integer> chrLengths;;\\n    public final String sample;;\\n    public final String samplem;;\\n    public final String ampliconBasedCalling;;\\n    public final PrinterType printerTypeOut;;\\n    public final Map<String, Integer> adaptorForward;;\\n    public final Map<String, Integer> adaptorReverse;;\\n    public GlobalReadOnlyScope(Configuration conf, Map<String, Integer> chrLengths, String sample, String samplem, String ampliconBasedCalling, Map<String, Integer> adaptorForward, Map<String, Integer> adaptorReverse);\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/data/scopedata/GlobalReadOnlyScope.java', 'methodCount': 6, 'lineCount': 18}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class RegionMissedSourceException {\\n    public final static String RegionSourceMissedMessage = \"The required BED file or region missed, please, set it \" + \"with path to BED or with -R option.\";;\\n    public RegionMissedSourceException();\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/exception/RegionMissedSourceException.java', 'methodCount': 1, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 10}\n",
      "{}\n",
      "/**\n",
      "     * Read map of chromosome lengths\n",
      "     * @param bam BAM file name\n",
      "     * @return Map of chromosome lengths. Key - chromosome name, value - length\n",
      "     * @throws IOException if BAM/SAM file can't be opened\n",
      "     */\n",
      "public static Map<String, Integer> readChr(String bam) throws IOException\n",
      "**************************************\n",
      "public static Map<String, Integer> readChr(String bam) throws IOException {\n",
      "    try (SamReader reader = SamReaderFactory.makeDefault().open(new File(bam))) {\n",
      "        SAMFileHeader header = reader.getFileHeader();\n",
      "        Map<String, Integer> chrs = new HashMap<>();\n",
      "        for (SAMSequenceRecord record : header.getSequenceDictionary().getSequences()) {\n",
      "            record.getSequenceLength();\n",
      "            String sn = record.getSequenceName();\n",
      "            int ln = record.getSequenceLength();\n",
      "            chrs.put(sn, ln);\n",
      "        }\n",
      "        return chrs;\n",
      "    }\n",
      "}\n",
      "1348\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 24/1243 [00:45<40:19,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "881\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 25/1243 [00:47<38:03,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "900\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 26/1243 [00:49<39:35,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "902\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 27/1243 [00:51<37:52,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "903\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 28/1243 [00:53<37:15,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "904\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 75, 'classCount': 1, 'methodsOnly': 'public class Configuration {\\n    public static final String HG19 = \"/ngs/reference_data/genomes/Hsapiens/hg19/seq/hg19.fa\";;\\n    public static final String HG38 = \"/ngs/reference_data/genomes/Hsapiens/hg38/seq/hg38.fa\";;\\n    public static final String MM10 = \"/ngs/reference_data/genomes/Mmusculus/mm10/seq/mm10.fa\";;\\n    /**\\n * Print a header row describing columns\\n */\\n//-h\\npublic boolean printHeader;;\\n    /**\\n * The delimiter for split region_info\\n */\\n// -d\\npublic String delimiter;;\\n    /**\\n * Path to bed file with regions\\n */\\npublic String bed;;\\n    /**\\n * The number of nucleotide to extend for each segment\\n */\\n// -x\\npublic int numberNucleotideToExtend;;\\n    /**\\n * Indicate whether is zero-based coordinates, as IGV does\\n * When use -R option, it is set to false\\n */\\n// -z\\npublic Boolean zeroBased;;\\n    /**\\n * Indicate it\\'s amplicon based calling.  Reads don\\'t map to the amplicon will be skipped.\\n * A read pair is considered belonging the amplicon if the edges are less than int bp to the amplicon,\\n * and overlap fraction is at least float. Default: 10:0.95\\n */\\n//-a\\npublic String ampliconBasedCalling;;\\n    //-c\\npublic int columnForChromosome = -1;;\\n    public RegionBuilder.BedRowFormat bedRowFormat;;\\n    /**\\n * The regular expression to extract sample name from bam filenames.\\n */\\n// -n\\npublic String sampleNameRegexp;;\\n    /**\\n * The sample name to be used directly\\n */\\n//-N\\npublic String sampleName;;\\n    /**\\n * The reference fasta\\n */\\n// -G\\npublic String fasta;;\\n    /**\\n * The indexed BAM file name(s)\\n */\\n//-b\\npublic BamNames bam;;\\n    /**\\n * For downsampling fraction\\n */\\n//-Z\\npublic Double downsampling;;\\n    // -C\\npublic boolean chromosomeNameIsNumber;;\\n    /**\\n * If set, reads with mapping quality less than INT will be filtered and ignored\\n */\\n//-Q\\npublic Integer mappingQuality;;\\n    /**\\n * Indicate to remove duplicated reads\\n */\\n//-t\\npublic boolean removeDuplicatedReads;;\\n    /**\\n * If set, reads with mismatches more than INT will be filtered and ignored\\n */\\n//-m, default = 8\\npublic int mismatch;;\\n    /**\\n * Verbose mode. Will output variant calling process.\\n */\\n//-y\\npublic boolean y;;\\n    /**\\n * The phred score for a base to be considered a good call\\n */\\n// -q, default = 22.5\\npublic double goodq;;\\n    /**\\n * Extension of bp to look for mismatches after insertion or deletion\\n */\\n// -X, default 2\\npublic int vext = 2;;\\n    /**\\n * Trim bases after [INT] bases in the reads\\n */\\n// -T\\npublic int trimBasesAfter = 0;;\\n    /**\\n * Indicate whether to perform local realignment\\n */\\n// -k, default false\\npublic boolean performLocalRealignment;;\\n    /**\\n * The indel size\\n */\\n// -I, default 50\\npublic int indelsize = 50;;\\n    /**\\n * The cutoff to decide whether a position has read strand bias\\n */\\npublic double bias = 0.05d;;\\n    /**\\n * The minimum reads for bias calculation. Default: 2 $minb\\n */\\n// -B\\npublic int minBiasReads = 2;;\\n    /**\\n * The minimum # of variance reads. Default: 2. If -p, it is set to 0.\\n */\\n// -r\\npublic int minr = 2;;\\n    /**\\n * Debug mode. Will print some error messages and append full genotype at the end.\\n */\\n// -D\\npublic boolean debug = false;;\\n    /**\\n * The threshold for allele frequency. If -p it is set to -1.\\n */\\n// -f\\npublic double freq = 0.01;;\\n    /**\\n * Indicate to move indels to 3-prime if alternative alignment can be achieved.\\n */\\n//-3\\npublic boolean moveIndelsTo3 = false;;\\n    /**\\n * The hexical to filter reads.\\n */\\n//-F\\npublic String samfilter = \"0x504\";;\\n    /**\\n * chr:start[-end]. If end is omitted, then a single position.\\n */\\n//-R\\npublic String regionOfInterest;;\\n    /**\\n * The read position filter. Default: 5\\n */\\n// -P\\npublic int readPosFilter = 5;;\\n    /**\\n * The Qratio of (good_quality_reads)/(bad_quality_reads+0.5)\\n */\\n// -o\\npublic double qratio = 1.5;;\\n    /**\\n * The minimum mean mapping quality to be considered. Default: 0.\\n */\\n// -O\\npublic double mapq = 0;;\\n    /**\\n * Do pileup regardless the frequency\\n */\\n// -p\\npublic boolean doPileup = false;;\\n    /**\\n * The lowest allele frequency in normal sample allowed for a putative somatic mutations. Default: 0.05.\\n */\\n// -V\\npublic double lofreq = 0.05d;;\\n    /**\\n * Any base with quality &lt;=10 will be consider low quality in soft-clipped seq and extension will stop.\\n */\\npublic static final int LOWQUAL = 10;;\\n    /**\\n * The minimum matches for a read to be considered\\n */\\n// -M\\npublic int minmatch = 0;;\\n    /**\\n *  Output splicing read counts\\n */\\n// -i\\npublic boolean outputSplicing = false;;\\n    /**\\n * How strict to be when reading a SAM or BAM.\\n */\\n// -VS\\npublic ValidationStringency validationStringency = ValidationStringency.LENIENT;;\\n    /**\\n * Include Ns in the total depth calculation.\\n */\\n// -K\\npublic boolean includeNInTotalDepth = false;;\\n    /**\\n * Indicate unique mode, which when mate pairs overlap,\\n * the overlapping part will be counted only once using forward read only.\\n */\\n// -u\\npublic boolean uniqueModeAlignmentEnabled = false;;\\n    /**\\n * Indicate unique mode, which when mate pairs overlap,\\n * the overlapping part will be counted only once using first read only.\\n */\\n// -UN\\npublic boolean uniqueModeSecondInPairEnabled = false;;\\n    /**\\n * Threads count to use in multithreading mode\\n */\\n//-th\\npublic int threads;;\\n    /**\\n * The larger seed size\\n */\\npublic static final int SEED_1 = 17;;\\n    /**\\n * The smaller seed size\\n */\\npublic static final int SEED_2 = 12;;\\n    /**\\n * The adaptor size\\n */\\npublic static final int ADSEED = 6;;\\n    /**\\n * Indicate to turn off chimeric reads filtering.  Chimeric reads are artifacts from library construction,\\n * where a read can be split into two segments, each will be aligned within 1-2 read length distance,\\n * but in opposite direction.\\n */\\n// --chimeric\\npublic boolean chimeric = false;;\\n    /**\\n * Turn off structural variant calling when set to true\\n */\\n//-U\\npublic boolean disableSV = false;;\\n    /**\\n * Turn on deleting of duplicate variants that can appear due to VarDict linear work on regions.\\n */\\npublic boolean deleteDuplicateVariants = false;;\\n    /**\\n * Applying Fisher exact test on forward and reverse counts of variant.\\n */\\npublic boolean fisher = false;;\\n    /**\\n * The minimum distance between two SV clusters in term of read length\\n */\\npublic static final double MINSVCDIST = 1.5;;\\n    /**\\n * The minimum position in read for mapping quality in SV analysis\\n */\\npublic static final int MINMAPBASE = 15;;\\n    /**\\n * The minimum distance between start position and end of SV structure in inter-chr translocation\\n */\\npublic static final int MINSVPOS = 25;;\\n    /**\\n * Mean Insert size\\n */\\n//-w\\npublic int INSSIZE = 300;;\\n    /**\\n * Insert std\\n */\\n//-W\\npublic int INSSTD = 100;;\\n    /**\\n * Insert std amount\\n */\\n//-A\\npublic int INSSTDAMT = 4;;\\n    /**\\n * The minimum structural variant length to be presented using &lt;DEL&gt; &lt;DUP&gt; &lt;INV&gt; &lt;INS&gt;, etc.\\n */\\n//-L\\npublic int SVMINLEN = 1000;;\\n    /**\\n * Max Structure variant size to be called in realignment step\\n */\\npublic static final int SVMAXLEN = 150000;;\\n    /**\\n * the flanking sequence length for SV\\n */\\npublic static final int SVFLANK = 50;;\\n    /**\\n * The minimum mapping quality when structural variant is only supported by discordant pairs\\n */\\npublic static final int DISCPAIRQUAL = 35;;\\n    public static final int EXTENSION = 5000;;\\n    public static final String DEFAULT_AMPLICON_PARAMETERS = \"10:0.95\";;\\n    /**\\n * Default reference extension $REFEXT\\n */\\npublic int referenceExtension = 1200;;\\n    /**\\n * Default printer for variants - system.out\\n */\\npublic PrinterType printerType = PrinterType.OUT;;\\n    /**\\n * Exception counter\\n */\\npublic AtomicInteger exceptionCounter = new AtomicInteger(0);;\\n    /**\\n * Maximum of exception to continue work\\n */\\npublic static int MAX_EXCEPTION_COUNT = 10;;\\n    /**\\n * List of adaptor sequences\\n */\\npublic List<String> adaptor = new ArrayList<>();;\\n    /**\\n * In CRISPR mode, the minimum amount in bp that a read needs to overlap with cutting site.\\n */\\npublic int crisprFilteringBp = 0;;\\n    /**\\n * The genomic position that CRISPR/Cas9 suppose to cut, typically 3bp from the PAM NGG site and within the guide.\\n */\\npublic int crisprCuttingSite = 0;;\\n    /**\\n * The variant frequency threshold to determine variant as good in case of monomer MSI\\n */\\n// -mfreq\\npublic double monomerMsiFrequency = 0.25d;;\\n    /**\\n * The variant frequency threshold to determine variant as good in case of non-monomer MSI\\n */\\n// -nmfreq\\npublic double nonMonomerMsiFrequency = 0.1d;;\\n    public boolean isColumnForChromosomeSet()\\n    public boolean isDownsampling()\\n    public boolean hasMappingQuality()\\n    public boolean isZeroBasedDefined()\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/Configuration.java', 'methodCount': 4, 'lineCount': 81}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class VariationMap {\\n    public SV sv;;\\n    public static SV getSV(Map<Integer, VariationMap<String, Variation>> hash, int start)\\n    public static void removeSV(Map<Integer, VariationMap<String, Variation>> hash, int start)\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/collection/VariationMap.java', 'methodCount': 2, 'lineCount': 5}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Construct reference sequence\n",
      "     * @param baseToPosition map of reference bases    $ref\n",
      "     * @param from start position\n",
      "     * @param to end position\n",
      "     * @param seq sequence to go through found in findconseq()\n",
      "     * @param EXTRA extra subsequence from findMatch()\n",
      "     * @return reference sequence starting at from and ending at to\n",
      "     */\n",
      "public static String joinRefFor5Lgins(Map<Integer, Character> baseToPosition, int from, int to, String seq, String EXTRA)\n",
      "**************************************\n",
      "public static String joinRefFor5Lgins(Map<Integer, Character> baseToPosition, int from, int to, String seq, String EXTRA) {\n",
      "    StringBuilder sb = new StringBuilder();\n",
      "    for (int i = from; i <= to; i++) {\n",
      "        if (to - i < seq.length() - EXTRA.length()) {\n",
      "            sb.append(charAt(seq, to - i + EXTRA.length()));\n",
      "        } else {\n",
      "            Character ch = baseToPosition.get(i);\n",
      "            if (ch != null) {\n",
      "                sb.append(ch);\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return sb.toString();\n",
      "}\n",
      "3080\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 29/1243 [00:53<31:18,  1.55s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "914\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 30/1243 [00:55<31:34,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "928\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   2%|         | 31/1243 [00:57<31:49,  1.58s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "929\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 32/1243 [00:59<33:36,  1.66s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "936\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 33/1243 [01:00<33:03,  1.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "946\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class Region {\\n    /**\\n * Chromosome name\\n */\\npublic final String chr;;\\n    /**\\n * Region start position\\n */\\npublic final int start;;\\n    /**\\n * Region end position\\n */\\npublic final int end;;\\n    /**\\n * Gene name\\n */\\npublic final String gene;;\\n    /**\\n * Position to start looking for variants (for amplicon based calling)\\n */\\npublic final int insertStart;;\\n    /**\\n * Position to end looking for variants (for amplicon based calling)\\n */\\npublic final int insertEnd;;\\n    public Region(String chr, int start, int end, String gene);\\n    public Region(String chr, int start, int end, String gene, int insertStart, int insertEnd);\\n    public static Region newModifiedRegion(Region region, int changedStart, int changedEnd)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String printRegion()\\n}\\n', 'filePath': '/home/ubuntu/repos/AstraZeneca-NGS,VarDictJava/src/main/java/com/astrazeneca/vardict/data/Region.java', 'methodCount': 7, 'lineCount': 15}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Method calls on cycles where position and records are processing to skip number of Exceptions set in\n",
      "     * Configuration.MAX_EXCEPTION_COUNT to try continue work if not critical exception occurs and get partial results.\n",
      "     * Each Exception printed in STDERR. If limit of exceptions reaches, VarDict stops with last re-throwed Exception.\n",
      "     * @param exception exception to check on limit exceeded\n",
      "     * @param place characterizes designation of the place where Exception occurs (position, SAM record, etc)\n",
      "     * @param placeDef specific place (chromosome, name of record, etc)\n",
      "     * @param region region from BED file/-R option\n",
      "     */\n",
      "public static void printExceptionAndContinue(Exception exception, String place, String placeDef, Region region)\n",
      "**************************************\n",
      "public static void printExceptionAndContinue(Exception exception, String place, String placeDef, Region region) {\n",
      "    String firstPart = \"There was Exception while processing \" + place + \" on \" + placeDef;\n",
      "    String secondPart = \". The processing will be continued from the next \" + place + \".\";\n",
      "    if (region != null) {\n",
      "        System.err.println(firstPart + \" on region \" + region.printRegion() + secondPart);\n",
      "        exception.printStackTrace();\n",
      "    } else {\n",
      "        System.err.println(firstPart + \" but region is undefined\" + secondPart);\n",
      "        exception.printStackTrace();\n",
      "    }\n",
      "    int currentCount = instance().conf.exceptionCounter.incrementAndGet();\n",
      "    if (currentCount > Configuration.MAX_EXCEPTION_COUNT) {\n",
      "        System.err.println(\"VarDictJava fails (there were \" + instance().conf.exceptionCounter.get() + \" continued exceptions during the run).\");\n",
      "        throw new RuntimeException(exception);\n",
      "    }\n",
      "}\n",
      "673\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 34/1243 [01:02<33:04,  1.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1087\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 35/1243 [01:03<33:03,  1.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1124\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 36/1243 [01:05<34:54,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1141\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 37/1243 [01:07<34:15,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1211\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 38/1243 [01:09<33:54,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1230\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "\t * @param matrix\n",
      "\t *            the matrix to set\n",
      "\t */\n",
      "public void set(int i, int j, double matrix)\n",
      "**************************************\n",
      "public void set(int i, int j, double matrix) {\n",
      "    if (matrix < 0) {\n",
      "        matrix = 0;\n",
      "    }\n",
      "    if (inRange(i) && inRange(j) && i != j) {\n",
      "        this.matrix[i][j] = matrix;\n",
      "        this.matrix[j][i] = matrix;\n",
      "    }\n",
      "}\n",
      "137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 39/1243 [01:11<35:54,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1587\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 40/1243 [01:12<34:57,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1588\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 41/1243 [01:14<35:45,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1665\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 42/1243 [01:16<38:11,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1671\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   3%|         | 43/1243 [01:18<36:29,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1804\n",
      "{'classCount': 3, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class BlastGraph {\\n    /**\\n * map gi with HitVertex object used by getVertex() and\\n * containVertex() methods\\n */\\nprotected Map<String, HitVertex> hitVertices;;\\n    protected ArrayList<Set<HitVertex>> subSetList;;\\n    /**\\n */\\nprivate static final long serialVersionUID = -2280763146567617357L;;\\n    // Comparator to sort graph by number of vertices\\npublic static final ByVerticesComparator byVertices = new ByVerticesComparator();;\\n    public BlastGraph();\\n    public boolean addVertex(HitVertex vertex)\\n    public HitVertex getVertex(String id)\\n    public boolean containsVertex(String id)\\n    public double getCoverage(ValueEdge edge)\\n    public double getCoverage2(ValueEdge edge)\\n    public double getScoreDensity(ValueEdge edge)\\n    public double getPercentageIdentity(ValueEdge edge)\\n    public double getPercentagePositive(ValueEdge edge)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByEvalue(double evalue, int index)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByCoverage(double coverage, int index)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByCoverage2(double coverage, int index)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByCoverageLength(double coverageLength, int index)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByScoreDensity(double scoreDensity, int index)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByPercentageIdentity(double percentageIdentity, int index)\\n    public Map<ValueEdge, Pair<HitVertex>> filtByPercentagePositive(double percentagePositive, int index)\\n    public void generateSubSet()\\n    public ArrayList<Set<HitVertex>> getSubSet()\\n    public double getConserveness()\\n    public void union(BlastGraph<HitVertex, ValueEdge> graphToUnion)\\n    public void empty()\\n    public void copy(BlastGraph<HitVertex, ValueEdge> graphToCopy)\\n    public void save(File gmlFile) throws IOException\\n    public void readGraph(File gmlFile) throws IOException, ParserConfigurationException, SAXException, GraphIOException\\n    public void createFromBlast(File blastXmlFile) throws BioException, IOException, SAXException\\n}\\n', 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/BlastGraph.java', 'methodCount': 25, 'lineCount': 31}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': \"public class HitVertex {\\n    // protein.fasta info\\nprivate String id;;\\n    private String accession;;\\n    private int length;;\\n    // gene index or orf nummber in a genome\\nprivate int index;;\\n    // 1 for forward, -1 for reverse, 0 for unknown or\\nprivate int strand;;\\n    // doesn't matter\\n// gene location in a genome\\nprivate Segment location;;\\n    private String genomeAcc;;\\n    private Taxon taxon;;\\n    private String description;;\\n    private String organism;;\\n    // Other information\\nprivate HashMap<String, String> attributes;;\\n    public HitVertex(String id, String accession, int length, int index, int strand, Segment location, String genomeAcc, Taxon taxon, String description, String organism);\\n    public HitVertex(String id);\\n    public String getAccession()\\n    public String getId()\\n    public void setId(String id)\\n    public void setAccession(String accession)\\n    public int getLength()\\n    public void setLength(int length)\\n    public String getDescription()\\n    public void setDescription(String description)\\n    public String getOrganism()\\n    public void setOrganism(String organism)\\n    public Map<String, String> getAllAttributes()\\n    public int getStrand()\\n    public void setStrand(int strand)\\n    public int getIndex()\\n    public void setIndex(int index)\\n    public Segment getLocation()\\n    public void setLocation(Segment location)\\n    public String getGenomeAcc()\\n    public void setGenomeAcc(String genomeAcc)\\n    public Taxon getTaxon()\\n    public void setTaxon(Taxon taxon)\\n    public String getAttribute(String key)\\n    public void putAttribute(String key, String value)\\n    public boolean contains(String string)\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n}\\n\", 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/HitVertex.java', 'methodCount': 28, 'lineCount': 41}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class Segment {\\n    int start, stop;;\\n    public final static ByStartComparator byStart = new ByStartComparator();;\\n    public Segment(int start, int stop);\\n    public Segment(String segment);\\n    public int getStart()\\n    public void setStart(int start)\\n    public int getStop()\\n    public void setStop(int stop)\\n    public static int segmentCoverage(Segment[] segments)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/Segment.java', 'methodCount': 8, 'lineCount': 12}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class SubHit {\\n    private double expectValue;;\\n    private double score;;\\n    private int numberOfIdentities;;\\n    private int numberOfPositives;;\\n    private double percentageIdentity;;\\n    private int alignmentSize;;\\n    private int querySequenceStart;;\\n    private int querySequenceEnd;;\\n    private int subjectSequenceStart;;\\n    private int subjectSequenceEnd;;\\n    //Comparator\\npublic static final ByScoreComparator byScore = new ByScoreComparator();;\\n    public SubHit(double expectValue, double score, int numberOfIdentities, int numberOfPositives, double percentageIdentity, int alignmentSize, int querySequenceStart, int querySequenceEnd, int subjectSequenceStart, int subjectSequenceEnd);\\n    public double getExpectValue()\\n    public void setExpectValue(double expectValue)\\n    public double getScore()\\n    public void setScore(double score)\\n    public int getNumberOfIdentities()\\n    public void setNumberOfIdentities(int numberOfIdentities)\\n    public int getNumberOfPositives()\\n    public void setNumberOfPositives(int numberOfPositives)\\n    public double getPercentageIdentity()\\n    public void setPercentageIdentity(double percentageIdentity)\\n    public int getAlignmentSize()\\n    public void setAlignmentSize(int alignmentSize)\\n    public int getQuerySequenceStart()\\n    public void setQuerySequenceStart(int querySequenceStart)\\n    public int getQuerySequenceEnd()\\n    public void setQuerySequenceEnd(int querySequenceEnd)\\n    public int getSubjectSequenceStart()\\n    public void setSubjectSequenceStart(int subjectSequenceStart)\\n    public int getSubjectSequenceEnd()\\n    public void setSubjectSequenceEnd(int subjectSequenceEnd)\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n}\\n', 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/SubHit.java', 'methodCount': 23, 'lineCount': 36}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class Taxon {\\n    private List<String> taxons;;\\n    public Taxon(List<String> taxons);\\n    public Taxon(String taxons);\\n    public String getTaxon()\\n    public String getParentTaxon()\\n    public String getFullTaxon()\\n    public String getTaxon(int i)\\n    public int getLevels()\\n    public String getTop()\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/Taxon.java', 'methodCount': 9, 'lineCount': 12}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ValueEdge {\\n    private double expectValue;;\\n    private double score;;\\n    private int numSubHit;;\\n    private SubHit[] subHit;;\\n    public ValueEdge(double expectValue, double score, int numSubHit, SubHit[] subHit);\\n    public double getExpectValue()\\n    public void setExpectValue(double expectValue)\\n    public double getScore()\\n    public void setScore(double score)\\n    public int getNumSubHit()\\n    public int getQueryCoverage()\\n    public int getSubjectCoverage()\\n    public void setNumSubHit(int numSubHit)\\n    public SubHit[] getSubHit()\\n    public void setSubHit(SubHit[] subHit)\\n}\\n', 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/ValueEdge.java', 'methodCount': 11, 'lineCount': 17}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Global {\\n    public static final BlastGraph<HitVertex, ValueEdge> graph = new BlastGraph<HitVertex, ValueEdge>();;\\n    public static final String APP_VERSION = \"BlastGraph v1.0beta\";;\\n    public static final Preferences PREFERENCES = Preferences.userRoot().node(\"BlastGraph\" + Global.APP_VERSION);;\\n    private static final String SEPARATOR = \"::!!::\";;\\n    public static final WorkStatus WORK_STATUS = new WorkStatus();;\\n    public static final HashMap<String, String> SETTING = new HashMap<String, String>();;\\n    public static final FrameProxy APP_FRAME_PROXY = new FrameProxy();;\\n    public static final CommandManager COMMAND_MANAGER = new CommandManager();;\\n    public static String getSeparator(char separator)\\n    public static String getSeparator()\\n    public static String getAppPath(Class cls)\\n}\\n', 'filePath': '/home/ubuntu/repos/bigwiv,BlastGraph/src/org/bigwiv/blastgraph/global/Global.java', 'methodCount': 3, 'lineCount': 13}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 19}\n",
      "{}\n",
      "/**\n",
      "\t * Write\n",
      "\t * \n",
      "\t * @param file\n",
      "\t * @param vv\n",
      "\t * @throws IOException\n",
      "\t */\n",
      "public static void saveImage(File imageFile, String format, VisualizationViewer<HitVertex, ValueEdge> vv) throws IOException\n",
      "**************************************\n",
      "public static void saveImage(File imageFile, String format, VisualizationViewer<HitVertex, ValueEdge> vv) throws IOException {\n",
      "    vv.setDoubleBuffered(false);\n",
      "    int width = vv.getWidth();\n",
      "    int height = vv.getHeight();\n",
      "    BufferedImage bi = new BufferedImage(width, height, BufferedImage.TYPE_INT_RGB);\n",
      "    Graphics2D graphics = bi.createGraphics();\n",
      "    vv.paint(graphics);\n",
      "    graphics.dispose();\n",
      "    ImageIO.write(bi, format, imageFile);\n",
      "    vv.setDoubleBuffered(true);\n",
      "}\n",
      "2439\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 44/1243 [01:20<38:22,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2393\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 45/1243 [01:22<37:11,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2394\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 46/1243 [01:24<35:57,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2472\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 47/1243 [01:26<37:37,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2571\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 48/1243 [01:27<36:27,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2645\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SimpleInterval1D {\\n    private final static Logger log = Logger.getLogger(SimpleInterval1D.class.getName());;\\n    // left endpoint, inclusive\\npublic final long left;;\\n    // right endpoint, inclusive\\npublic final long right;;\\n    public SimpleInterval1D(long left, long right);\\n    public SimpleInterval1D(Interval1D reg);\\n    public SimpleInterval1D(Random r);\\n    public SimpleInterval1D(Random r, int lower, int upper);\\n    public static void main(String[] args)\\n    public long length()\\n    public long getLeft()\\n    public long getRight()\\n    public boolean intersects(Interval1D that)\\n    public boolean intersects(Interval1D that, double reciprocalRatio)\\n    public boolean intersects(Interval1D that, double reciprocalRatio, int wiggle)\\n    public boolean contains(final long point)\\n    public int compareTo(final Interval1D that)\\n    public SimpleInterval1D union(final Interval1D that)\\n    public long getCenter()\\n    public String toString()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/intervalTree/SimpleInterval1D.java', 'methodCount': 18, 'lineCount': 23}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Alt {\\n    //TODO convert this class to immutable\\nprivate SymbolicAllele symbolicAllele;;\\n    private Breakend breakend;;\\n    private FlexSeq seq;;\\n    private int breakendCount;;\\n    private int symbolicAlleleCount;;\\n    public static Alt altFactory(String alt)\\n    public Alt();\\n    public Alt(SymbolicAllele s);\\n    public Alt(FlexSeq s);\\n    public Alt(Breakend b);\\n    public Alt copy()\\n    public int length()\\n    public boolean isSeq()\\n    public byte[] substring(int start, int end)\\n    public byte[] substring(int start)\\n    public byte byteAt(int i)\\n    public FlexSeq.Type getSeqType()\\n    public int getCopyNumber()\\n    public int getBreakendCount()\\n    public void setBreakendCount(int breakendCount)\\n    public int getSymbolicAlleleCount()\\n    public void setSymbolicAlleleCount(int symbolicAlleleCount)\\n    public SymbolicAllele getSymbolicAllele()\\n    public void setSymbolicAllele(SymbolicAllele symbolicAllele)\\n    public Breakend getBreakend()\\n    public void setBreakend(Breakend breakend)\\n    public FlexSeq getSeq()\\n    public void setSeq(FlexSeq seq)\\n    public byte[] getSequence()\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/types/variant/alt/Alt.java', 'methodCount': 24, 'lineCount': 31}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SimpleReference {\\n    private final static Logger log = Logger.getLogger(SimpleReference.class.getName());;\\n    // chr_idx -> reference_string\\nprivate final Map<ChrString, Sequence> data = new HashMap<>();;\\n    private final Map<ChrString, IndexedFastaSequenceFile> dataSources = new HashMap<>();;\\n    private long nonNCount = -1;;\\n    private String referenceFileName = null;;\\n    private boolean sequencesFullyLoaded = false;;\\n    public SimpleReference();\\n    public SimpleReference(String filename);\\n    public SimpleReference(final Collection<String> filenames);\\n    private void addReference(String filename)\\n    public Sequence getSequence(ChrString chr_name)\\n    public byte byteAt(ChrString chr_name, int loc)\\n    public char charAt(ChrString chr_name, int loc)\\n    public byte[] byteRange(ChrString chr_name, int start_loc, int end_loc)\\n    public int getRefLen(ChrString chr_name)\\n    public int getNumContigs()\\n    private void loadAllSequences()\\n    public long getNumNonNBases()\\n    public long getNumNonNBases(final File regions) throws IOException\\n    public Set<ChrString> keySet()\\n    public String getReferenceFileName()\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/util/SimpleReference.java', 'methodCount': 15, 'lineCount': 23}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class StringUtilities {\\n    public static String concatenateArray(final int[] a, final CharSequence delimiter)\\n    public static String concatenateArray(final String[] a, final CharSequence delimiter)\\n    public static String concatenateArray(final ChrString[] a, final CharSequence delimiter)\\n    public static String[] fastSplit(final String s, final String delim)\\n    public static boolean isNonNegativeInteger(final String s)\\n    public static int parseInt(final String s)\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/util/StringUtilities.java', 'methodCount': 6, 'lineCount': 8}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 5}\n",
      "{'Variant': 'public class Variant {\\n    private final static Logger log = Logger.getLogger(Variant.class.getName());;\\n    // this is hopefully a unique index, for the split variants\\npublic int splitVariantIndex = 0;;\\n    // this is hopefully a unique index, for the whole variants\\npublic int wholeVariantIndex = 0;;\\n    // this is the type before the variant was split into canonical ones\\npublic VariantOverallType originalType = null;;\\n    // use a seed for reproducibility, should be an option or global\\nprivate Random rand = null;;\\n    private int pos = -1, referenceAlleleLength = -1;;\\n    private byte[] ref;;\\n    private ChrString chr;;\\n    private Alt[] alts;;\\n    // -1 for not avaliable\\nprivate byte maternal = 0, paternal = 0;;\\n    // Phasing\\nprivate boolean isPhased = false;;\\n    private String filter;;\\n    private String qual = \".\";;\\n    private String info;;\\n    private String varId;;\\n    // this is when the reference base is deleted\\n// if it is the same as the first alt base\\n//so refDeleted.length() <= 1 is always true?\\nprivate String refDeleted;;\\n    private String extraBase = \"\";;\\n    private String clippedSequence = \"\";;\\n    private ChrString[] chr2;;\\n    private int[] pos2;;\\n    private int[] end2;;\\n    //is sequence inverted? useful for interspersed dup, translocation dup\\nprivate Boolean isinv;;\\n    //translocation ID\\nprivate String traid;;\\n    //caching hashcode\\nprivate volatile int hashCode;;\\n    //when this variant is a composite variant, this field will store constitutional variants\\nprivate List<Variant> compositions;;\\n    //additional INFO fields\\n//3\\' end distance with a matching variant\\nprivate int threePrimeDistance = -1;;\\n    //5\\' end distance with a matching variant\\nprivate int fivePrimeDistance = -1;;\\n    //length difference with a matching variant\\nprivate int lengthDifference = -1;;\\n    //true if length is imprecise\\nprivate Boolean isLengthImprecise = false;;\\n    public List<Variant> getCompositions()\\n    public void setCompositions(List<Variant> compositions)\\n    public Variant(final Random rand);\\n    public boolean isInversed()\\n    private Variant(Builder builder);\\n    public Variant(final Variant var);\\n    public ChrString getChr()\\n    public int getPos()\\n    public boolean setNovelPosition(final int pos, final SimpleReference ref)\\n    public void setVarID(final String id)\\n    public int getReferenceAlleleLength()\\n    public int maternal()\\n    public int paternal()\\n    public byte[] insertion(final int ind)\\n    public ChrString getChr2(final int ind)\\n    public int getPos2(final int ind)\\n    public int getEnd2(final int ind)\\n    public ChrString[] getAllChr2()\\n    public int[] getAllPos2()\\n    public int[] getAllEnd2()\\n    public int getEnd()\\n    public String getTraid()\\n    public int getInsertionLength(final int ind)\\n    public int maxLen(final int ind)\\n    public int maxLen()\\n    public int minLen()\\n    public SimpleInterval1D getAlternativeAlleleInterval(final int ind)\\n    public SimpleInterval1D getVariantInterval(final int ind)\\n    public SimpleInterval1D getVariantInterval(final int ind, final boolean ignoreInsertionLength)\\n    public SimpleInterval1D getGenotypeUnionAlternativeInterval()\\n    public SimpleInterval1D getGenotypeUnionVariantInterval()\\n    public boolean isSmallVariant(int genotype, int cutoff, boolean ignoreInsertionLength)\\n    public Genotypes getGenotypes()\\n    public int getAllele(final int parent)\\n    public int getMinMatchLength()\\n    public void setAllele(final int parent, final byte allele)\\n    public int variantBases()\\n    public VariantType getType(final int ind)\\n    public VariantOverallType getType()\\n    public Alt getAlt(final int ind)\\n    public void setAlt(final int ind, Alt alt)\\n    public void setThreePrimeDistance(int threePrimeDistance)\\n    public void setFivePrimeDistance(int fivePrimeDistance)\\n    public void setLengthDifference(int lengthDifference)\\n    public String getFilter()\\n    public String getQual()\\n    public String getInfo()\\n    public boolean isPhased()\\n    public boolean isRef()\\n    public String getVariantId()\\n    public byte[] getReference()\\n    public String getReferenceString()\\n    public byte[] getRef_deleted()\\n    public int getCN(final int ind)\\n    public boolean hasCN()\\n    public String alternativeAlleleString()\\n    public int getNumberOfAlternativeAlleles()\\n    public void randomizeHaplotype()\\n    public void randomizeGenotype(GenderType gender)\\n    public boolean isAltACTGN()\\n    public boolean isHom()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public int compareTo(final Variant other)\\n    public List<Integer> getSVLEN()\\n    public String getLengthString()\\n    public void calculateExtraBase(final Sequence refSeq)\\n    public String toString()\\n    public String toString(final int paternal, final int maternal)\\n    private List<String> rmDupInfo(List<String> fields)\\n    public void output(PrintWriter p)\\n    public byte getGoodPaternal()\\n    public byte getGoodMaternal()\\n    public String getExtraBase()\\n    public Boolean isLengthImprecise()\\n    public void setLengthImprecise(Boolean lengthImprecise)\\n}\\n'}\n",
      "/**\n",
      "     * get maximum length of a variant across\n",
      "     * genotypes and reference and alternative alleles\n",
      "     * @return\n",
      "     */\n",
      "public int maxLen()\n",
      "**************************************\n",
      "public int maxLen() {\n",
      "    if (compositions == null) {\n",
      "        return Math.max(maxLen(getAllele(0)), maxLen(getAllele(1)));\n",
      "    } else {\n",
      "        if (compositions.isEmpty())\n",
      "            return 0;\n",
      "        int maxLen = compositions.get(0).maxLen();\n",
      "        for (Variant c : compositions) {\n",
      "            maxLen = Math.max(maxLen, c.maxLen());\n",
      "        }\n",
      "        return maxLen;\n",
      "    }\n",
      "}\n",
      "2591\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 49/1243 [01:30<39:13,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2679\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 50/1243 [01:32<40:14,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2680\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 51/1243 [01:34<41:57,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2681\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 52/1243 [01:36<41:54,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2690\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 53/1243 [01:38<42:34,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2718\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class VarSimTool {\\n    public final String VERSION = \"VarSim\" + \" \" + getClass().getPackage().getImplementationVersion();;\\n    @Option(name = \"-version\", usage = \"Print version\", help = true)\\nprotected boolean printVersion = false;;\\n    @Option(name = \"-h\", usage = \"Print help message\", help = true, aliases = { \"-help\" })\\nprotected boolean help = false;;\\n    protected String command = \"\";;\\n    protected String description = \"\";;\\n    public VarSimTool(final String command, final String description);\\n    public abstract void run(String[] args) throws IOException\\n    public String getCommand()\\n    public String getDescription()\\n    public void printUsage(final CmdLineParser parser)\\n    public boolean parseArguments(final String[] args)\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/VarSimTool.java', 'methodCount': 6, 'lineCount': 13}, {'fieldCount': 12, 'classCount': 1, 'methodsOnly': 'public class FlexSeq {\\n    //TODO make all fields final if possible\\npublic static Set<Byte> LEGAL_SEQUENCE_CHARS = Stream.of((byte) \\'A\\', (byte) \\'T\\', (byte) \\'C\\', (byte) \\'G\\', (byte) \\'a\\', (byte) \\'t\\', (byte) \\'c\\', (byte) \\'g\\', (byte) \\'N\\', (byte) \\'n\\').collect(Collectors.toCollection(HashSet<Byte>::new));;\\n    Type type;;\\n    int length;;\\n    int copyNumber = 1;;\\n    byte[] sequence;;\\n    String variantId = \"\";;\\n    ChrString chr2;;\\n    int pos2;;\\n    int end2;;\\n    int referenceAlleleLength;;\\n    boolean isinv;;\\n    volatile int hashCode;;\\n    private FlexSeq(Builder b);\\n    public FlexSeq();\\n    public FlexSeq(final byte seq);\\n    public FlexSeq(final byte[] seq);\\n    public FlexSeq(final Type t, int len);\\n    public FlexSeq(final Type t, final int len, final int copy_num);\\n    public FlexSeq(final FlexSeq b);\\n    public byte byteAt(final int i)\\n    public byte[] substring(final int beginIndex)\\n    public byte[] substring(final int beginIndex, final int endIndex)\\n    public int length()\\n    public boolean isInversed()\\n    public int varLength()\\n    private void checkSequence(byte[] s)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public int getCopyNumber()\\n    public Type getType()\\n    public String getVariantId()\\n    public byte[] getSequence()\\n    public boolean isSeq()\\n    public String toString()\\n    public ChrString getChr2()\\n    public int getPos2()\\n    public int getEnd2()\\n    public int getReferenceAlleleLength()\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/types/FlexSeq.java', 'methodCount': 26, 'lineCount': 40}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/types/variant/VariantOverallType.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class SampleParams {\\n    public int addedNum;;\\n    public int seenNum;;\\n    public SampleParams();\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/types/SampleParams.java', 'methodCount': 1, 'lineCount': 5}, {'fieldCount': 29, 'classCount': 1, 'methodsOnly': 'public class Variant {\\n    private final static Logger log = Logger.getLogger(Variant.class.getName());;\\n    // this is hopefully a unique index, for the split variants\\npublic int splitVariantIndex = 0;;\\n    // this is hopefully a unique index, for the whole variants\\npublic int wholeVariantIndex = 0;;\\n    // this is the type before the variant was split into canonical ones\\npublic VariantOverallType originalType = null;;\\n    // use a seed for reproducibility, should be an option or global\\nprivate Random rand = null;;\\n    private int pos = -1, referenceAlleleLength = -1;;\\n    private byte[] ref;;\\n    private ChrString chr;;\\n    private Alt[] alts;;\\n    // -1 for not avaliable\\nprivate byte maternal = 0, paternal = 0;;\\n    // Phasing\\nprivate boolean isPhased = false;;\\n    private String filter;;\\n    private String qual = \".\";;\\n    private String info;;\\n    private String varId;;\\n    // this is when the reference base is deleted\\n// if it is the same as the first alt base\\n//so refDeleted.length() <= 1 is always true?\\nprivate String refDeleted;;\\n    private String extraBase = \"\";;\\n    private String clippedSequence = \"\";;\\n    private ChrString[] chr2;;\\n    private int[] pos2;;\\n    private int[] end2;;\\n    //is sequence inverted? useful for interspersed dup, translocation dup\\nprivate Boolean isinv;;\\n    //translocation ID\\nprivate String traid;;\\n    //caching hashcode\\nprivate volatile int hashCode;;\\n    //when this variant is a composite variant, this field will store constitutional variants\\nprivate List<Variant> compositions;;\\n    //additional INFO fields\\n//3\\' end distance with a matching variant\\nprivate int threePrimeDistance = -1;;\\n    //5\\' end distance with a matching variant\\nprivate int fivePrimeDistance = -1;;\\n    //length difference with a matching variant\\nprivate int lengthDifference = -1;;\\n    //true if length is imprecise\\nprivate Boolean isLengthImprecise = false;;\\n    public List<Variant> getCompositions()\\n    public void setCompositions(List<Variant> compositions)\\n    public Variant(final Random rand);\\n    public boolean isInversed()\\n    private Variant(Builder builder);\\n    public Variant(final Variant var);\\n    public ChrString getChr()\\n    public int getPos()\\n    public boolean setNovelPosition(final int pos, final SimpleReference ref)\\n    public void setVarID(final String id)\\n    public int getReferenceAlleleLength()\\n    public int maternal()\\n    public int paternal()\\n    public byte[] insertion(final int ind)\\n    public ChrString getChr2(final int ind)\\n    public int getPos2(final int ind)\\n    public int getEnd2(final int ind)\\n    public ChrString[] getAllChr2()\\n    public int[] getAllPos2()\\n    public int[] getAllEnd2()\\n    public int getEnd()\\n    public String getTraid()\\n    public int getInsertionLength(final int ind)\\n    public int maxLen(final int ind)\\n    public int maxLen()\\n    public int minLen()\\n    public SimpleInterval1D getAlternativeAlleleInterval(final int ind)\\n    public SimpleInterval1D getVariantInterval(final int ind)\\n    public SimpleInterval1D getVariantInterval(final int ind, final boolean ignoreInsertionLength)\\n    public SimpleInterval1D getGenotypeUnionAlternativeInterval()\\n    public SimpleInterval1D getGenotypeUnionVariantInterval()\\n    public boolean isSmallVariant(int genotype, int cutoff, boolean ignoreInsertionLength)\\n    public Genotypes getGenotypes()\\n    public int getAllele(final int parent)\\n    public int getMinMatchLength()\\n    public void setAllele(final int parent, final byte allele)\\n    public int variantBases()\\n    public VariantType getType(final int ind)\\n    public VariantOverallType getType()\\n    public Alt getAlt(final int ind)\\n    public void setAlt(final int ind, Alt alt)\\n    public void setThreePrimeDistance(int threePrimeDistance)\\n    public void setFivePrimeDistance(int fivePrimeDistance)\\n    public void setLengthDifference(int lengthDifference)\\n    public String getFilter()\\n    public String getQual()\\n    public String getInfo()\\n    public boolean isPhased()\\n    public boolean isRef()\\n    public String getVariantId()\\n    public byte[] getReference()\\n    public String getReferenceString()\\n    public byte[] getRef_deleted()\\n    public int getCN(final int ind)\\n    public boolean hasCN()\\n    public String alternativeAlleleString()\\n    public int getNumberOfAlternativeAlleles()\\n    public void randomizeHaplotype()\\n    public void randomizeGenotype(GenderType gender)\\n    public boolean isAltACTGN()\\n    public boolean isHom()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public int compareTo(final Variant other)\\n    public List<Integer> getSVLEN()\\n    public String getLengthString()\\n    public void calculateExtraBase(final Sequence refSeq)\\n    public String toString()\\n    public String toString(final int paternal, final int maternal)\\n    private List<String> rmDupInfo(List<String> fields)\\n    public void output(PrintWriter p)\\n    public byte getGoodPaternal()\\n    public byte getGoodMaternal()\\n    public String getExtraBase()\\n    public Boolean isLengthImprecise()\\n    public void setLengthImprecise(Boolean lengthImprecise)\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/types/variant/Variant.java', 'methodCount': 76, 'lineCount': 107}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Alt {\\n    //TODO convert this class to immutable\\nprivate SymbolicAllele symbolicAllele;;\\n    private Breakend breakend;;\\n    private FlexSeq seq;;\\n    private int breakendCount;;\\n    private int symbolicAlleleCount;;\\n    public static Alt altFactory(String alt)\\n    public Alt();\\n    public Alt(SymbolicAllele s);\\n    public Alt(FlexSeq s);\\n    public Alt(Breakend b);\\n    public Alt copy()\\n    public int length()\\n    public boolean isSeq()\\n    public byte[] substring(int start, int end)\\n    public byte[] substring(int start)\\n    public byte byteAt(int i)\\n    public FlexSeq.Type getSeqType()\\n    public int getCopyNumber()\\n    public int getBreakendCount()\\n    public void setBreakendCount(int breakendCount)\\n    public int getSymbolicAlleleCount()\\n    public void setSymbolicAlleleCount(int symbolicAlleleCount)\\n    public SymbolicAllele getSymbolicAllele()\\n    public void setSymbolicAllele(SymbolicAllele symbolicAllele)\\n    public Breakend getBreakend()\\n    public void setBreakend(Breakend breakend)\\n    public FlexSeq getSeq()\\n    public void setSeq(FlexSeq seq)\\n    public byte[] getSequence()\\n}\\n', 'filePath': '/home/ubuntu/repos/bioinform,varsim/src/main/java/com/bina/varsim/types/variant/alt/Alt.java', 'methodCount': 24, 'lineCount': 31}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{'SampleParams': 'public class SampleParams {\\n    public int addedNum;;\\n    public int seenNum;;\\n    public SampleParams();\\n}\\n'}\n",
      "/**\n",
      "     * This is for sampling without replacement. If the genotype is not sampled, it is replaced with zero\n",
      "     *\n",
      "     * @param geno       previous genotype\n",
      "     * @param seenAdded params for sampling\n",
      "     * @param numSample number that we want to sample\n",
      "     * @param numTotal  total number of such variants\n",
      "     * @param outputAll Don't sample, just output everything\n",
      "     * @return\n",
      "     */\n",
      "public byte sampleGenotype(byte geno, SampleParams seenAdded, int numSample, int numTotal, boolean outputAll)\n",
      "**************************************\n",
      "public byte sampleGenotype(byte geno, SampleParams seenAdded, int numSample, int numTotal, boolean outputAll) {\n",
      "    if (!outputAll) {\n",
      "        if (seenAdded.addedNum < numSample) {\n",
      "            double randNum = rand.nextDouble();\n",
      "            if ((numTotal - seenAdded.seenNum) * randNum >= (numSample - seenAdded.addedNum)) {\n",
      "                seenAdded.seenNum++;\n",
      "                return 0;\n",
      "            } else {\n",
      "                seenAdded.seenNum++;\n",
      "                seenAdded.addedNum++;\n",
      "                return geno;\n",
      "            }\n",
      "        }\n",
      "    } else {\n",
      "        return geno;\n",
      "    }\n",
      "    return 0;\n",
      "}\n",
      "2883\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 54/1243 [01:40<36:56,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2740\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   4%|         | 55/1243 [01:42<38:17,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2768\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 56/1243 [01:44<39:54,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2805\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 57/1243 [01:46<37:21,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2827\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 58/1243 [01:48<38:20,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2863\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{'Conversion': \"public class Conversion {\\n    protected ConversionGraph acg;;\\n    /**\\n * Nodes below branches to which recombinant edge connects.\\n */\\nprotected Node node1, node2;;\\n    /**\\n * Heights on branches at which recombinant edge connects.\\n */\\nprotected double height1, height2;;\\n    /**\\n * Range of nucleotides affected by homologous gene conversion.\\n */\\nprotected int startSite, endSite;;\\n    /**\\n * Locus with which conversion is associated.\\n */\\nprotected Locus locus;;\\n    /**\\n * Used by ACGAnnotator to incoroporate additional metadata into\\n * the summary ACG.\\n */\\npublic String newickMetaDataBottom, newickMetaDataMiddle, newickMetaDataTop;;\\n    /**\\n * Used by ACGAnnotator to keep track of which conversions belong to which\\n * ACGs. (The acg reference can't be used, as ACGAnnotator avoids keeping\\n * multiple ACGs in memory simultaneously.)\\n */\\npublic int acgIndex;;\\n    public Conversion();\\n    public Conversion(Node node1, double height1, Node node2, double height2, int startSite, int endSite, ConversionGraph acg, Locus locus);\\n    public Node getNode1()\\n    public Node getNode2()\\n    public double getHeight1()\\n    public double getHeight2()\\n    public int getStartSite()\\n    public void setStartSite(int startSite)\\n    public int getEndSite()\\n    public void setEndSite(int endSite)\\n    public void setNode1(Node node1)\\n    public void setNode2(Node node2)\\n    public void setHeight1(double height1)\\n    public void setHeight2(double height2)\\n    public Locus getLocus()\\n    public void setLocus(Locus locus)\\n    public int getSiteCount()\\n    public boolean isValid()\\n    public void setConversionGraph(ConversionGraph acg)\\n    public void startEditing()\\n    public Conversion getCopy()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n\"}\n",
      "/**\n",
      "     * Obtain new recombination with exactly the same\n",
      "     * field values as this one.\n",
      "     *\n",
      "     * @return copy of Conversion object\n",
      "     */\n",
      "public Conversion getCopy()\n",
      "**************************************\n",
      "public Conversion getCopy() {\n",
      "    Conversion copy = new Conversion();\n",
      "    copy.locus = locus;\n",
      "    copy.acg = acg;\n",
      "    copy.startSite = startSite;\n",
      "    copy.endSite = endSite;\n",
      "    copy.node1 = node1;\n",
      "    copy.node2 = node2;\n",
      "    copy.height1 = height1;\n",
      "    copy.height2 = height2;\n",
      "    return copy;\n",
      "}\n",
      "643\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 59/1243 [01:50<40:01,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2891\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 60/1243 [01:52<43:34,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2910\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 61/1243 [01:55<43:53,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2921\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 62/1243 [01:57<44:09,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2940\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 63/1243 [01:59<44:22,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2959\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 7, 'classCount': 1, 'methodsOnly': \"public class Conversion {\\n    protected ConversionGraph acg;;\\n    /**\\n * Nodes below branches to which recombinant edge connects.\\n */\\nprotected Node node1, node2;;\\n    /**\\n * Heights on branches at which recombinant edge connects.\\n */\\nprotected double height1, height2;;\\n    /**\\n * Range of nucleotides affected by homologous gene conversion.\\n */\\nprotected int startSite, endSite;;\\n    /**\\n * Locus with which conversion is associated.\\n */\\nprotected Locus locus;;\\n    /**\\n * Used by ACGAnnotator to incoroporate additional metadata into\\n * the summary ACG.\\n */\\npublic String newickMetaDataBottom, newickMetaDataMiddle, newickMetaDataTop;;\\n    /**\\n * Used by ACGAnnotator to keep track of which conversions belong to which\\n * ACGs. (The acg reference can't be used, as ACGAnnotator avoids keeping\\n * multiple ACGs in memory simultaneously.)\\n */\\npublic int acgIndex;;\\n    public Conversion();\\n    public Conversion(Node node1, double height1, Node node2, double height2, int startSite, int endSite, ConversionGraph acg, Locus locus);\\n    public Node getNode1()\\n    public Node getNode2()\\n    public double getHeight1()\\n    public double getHeight2()\\n    public int getStartSite()\\n    public void setStartSite(int startSite)\\n    public int getEndSite()\\n    public void setEndSite(int endSite)\\n    public void setNode1(Node node1)\\n    public void setNode2(Node node2)\\n    public void setHeight1(double height1)\\n    public void setHeight2(double height2)\\n    public Locus getLocus()\\n    public void setLocus(Locus locus)\\n    public int getSiteCount()\\n    public boolean isValid()\\n    public void setConversionGraph(ConversionGraph acg)\\n    public void startEditing()\\n    public Conversion getCopy()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/Conversion.java', 'methodCount': 24, 'lineCount': 33}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class ConversionGraph {\\n    /**\\n * Unlike Trees, Conversion graphs require an alignment (or at least\\n * the length of an alignment) to be specified so that the regions of\\n * the alignment affected by recombination events can be recorded.\\n */\\npublic Input<List<Locus>> lociInput = new Input<>(\"locus\", \"Locus associated with graph.\", new ArrayList<>());;\\n    public Input<String> fromStringInput = new Input<>(\"fromString\", \"Initialise ARG from string representation.\");;\\n    public Input<String> fromExtNewickInput = new Input<>(\"extendedNewick\", \"Initialise ARG from extended Newick representation.\");;\\n    public Input<Boolean> wholeLocusConversionsInput = new Input<>(\"wholeLocusConversionsOnly\", \"Force region boundaries to coincide with locus boundaries.\", false);;\\n    /**\\n * List of recombinations on graph.\\n */\\nprotected Map<Locus, List<Conversion>> convs;;\\n    protected Map<Locus, List<Conversion>> storedConvs;;\\n    protected final static List<Conversion> emptyConvList = new ArrayList<>();;\\n    /**\\n * Event and region lists.\\n */\\nprotected Map<Locus, RegionList> regionLists;;\\n    protected CFEventList cfEventList;;\\n    protected List<Locus> loci, convertibleLoci;;\\n    protected int totalConvertibleSequenceLength;;\\n    public void initAndValidate()\\n    public int getTotalConvertibleSequenceLength()\\n    public Locus getLocusByID(String id)\\n    public List<Locus> getConvertibleLoci()\\n    public boolean wholeLocusModeOn()\\n    public void setWholeLocusMode(boolean newVal)\\n    public void addConversion(Conversion conv)\\n    public void deleteConversion(Conversion conv)\\n    public List<Conversion> getConversions(Locus locus)\\n    public int getConvCount(Locus locus)\\n    public int getTotalConvCount()\\n    public int getConversionIndex(Conversion conv)\\n    public List<Region> getRegions(Locus locus)\\n    public int getRegionCount(Locus locus)\\n    public int getUselessConvCount()\\n    public List<CFEventList.Event> getCFEvents()\\n    public double getClonalFrameLength()\\n    public boolean isInvalid()\\n    public String toString()\\n    public String toStringOld()\\n    public void fromStringOld(String str)\\n    public void fromXML(final org.w3c.dom.Node node)\\n    public ConversionGraph copy()\\n    private void generalAssignFrom(StateNode other, boolean fragileAssignment)\\n    public void assignFrom(StateNode other)\\n    public void assignFromFragile(StateNode other)\\n    public String getExtendedNewick()\\n    public String getExtendedNewick(boolean computeAffectedSites)\\n    public String getTrimmedExtendedNewick()\\n    private String extendedNewickTraverse(Node node, AffectedSiteList asList, boolean intraCFOnly)\\n    public void fromExtendedNewick(String string)\\n    public void fromExtendedNewick(String string, boolean numbered, int nodeNumberoffset)\\n    protected void store()\\n    public void restore()\\n    public void startEditing(Operator operator)\\n    public boolean clonalFrameIsDirty()\\n    public void init(PrintStream out)\\n    public void log(long nSample, PrintStream out)\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/ConversionGraph.java', 'methodCount': 38, 'lineCount': 51}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Locus {\\n    public Input<Integer> siteCountInput = new Input<>(\"siteCount\", \"Number of sites contained in alignments associated with this locus\");;\\n    public Input<Alignment> alignmentInput = new Input<>(\"alignment\", \"Initialize locus using this alignment.\", Input.Validate.XOR, siteCountInput);;\\n    public Input<Boolean> conversionsAllowedInput = new Input<>(\"conversionsAllowed\", \"If false, no conversions will be allowed on this locus. (Default true.)\", true);;\\n    protected int siteCount;;\\n    protected Alignment alignment;;\\n    public Locus();\\n    public Locus(String name, Alignment alignment);\\n    public Locus(String name, int siteCount);\\n    public void initAndValidate()\\n    public int getSiteCount()\\n    public Alignment getAlignment()\\n    public boolean hasAlignment()\\n    public boolean conversionsAllowed()\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/Locus.java', 'methodCount': 8, 'lineCount': 15}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Assemble list of bitSets for this ACG.\n",
      "     */\n",
      "public BitSet[] getBitSets(ConversionGraph acg)\n",
      "**************************************\n",
      "/**\n",
      " * Assemble list of bitSets for this ACG.\n",
      " */\n",
      "public BitSet[] getBitSets(ConversionGraph acg) {\n",
      "    if (bitSets == null)\n",
      "        bitSets = new BitSet[acg.getNodeCount()];\n",
      "    applyToClades(acg.getRoot(), (cladeNode, bits) -> {\n",
      "        bitSets[cladeNode.getNr()] = bits;\n",
      "        return null;\n",
      "    });\n",
      "    return bitSets;\n",
      "}\n",
      "1713\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 64/1243 [02:02<44:58,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2964\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 65/1243 [02:03<41:46,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2993\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 66/1243 [02:05<38:57,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3006\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 67/1243 [02:07<39:29,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3007\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   5%|         | 68/1243 [02:09<41:04,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3012\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 7, 'classCount': 1, 'methodsOnly': \"public class Conversion {\\n    protected ConversionGraph acg;;\\n    /**\\n * Nodes below branches to which recombinant edge connects.\\n */\\nprotected Node node1, node2;;\\n    /**\\n * Heights on branches at which recombinant edge connects.\\n */\\nprotected double height1, height2;;\\n    /**\\n * Range of nucleotides affected by homologous gene conversion.\\n */\\nprotected int startSite, endSite;;\\n    /**\\n * Locus with which conversion is associated.\\n */\\nprotected Locus locus;;\\n    /**\\n * Used by ACGAnnotator to incoroporate additional metadata into\\n * the summary ACG.\\n */\\npublic String newickMetaDataBottom, newickMetaDataMiddle, newickMetaDataTop;;\\n    /**\\n * Used by ACGAnnotator to keep track of which conversions belong to which\\n * ACGs. (The acg reference can't be used, as ACGAnnotator avoids keeping\\n * multiple ACGs in memory simultaneously.)\\n */\\npublic int acgIndex;;\\n    public Conversion();\\n    public Conversion(Node node1, double height1, Node node2, double height2, int startSite, int endSite, ConversionGraph acg, Locus locus);\\n    public Node getNode1()\\n    public Node getNode2()\\n    public double getHeight1()\\n    public double getHeight2()\\n    public int getStartSite()\\n    public void setStartSite(int startSite)\\n    public int getEndSite()\\n    public void setEndSite(int endSite)\\n    public void setNode1(Node node1)\\n    public void setNode2(Node node2)\\n    public void setHeight1(double height1)\\n    public void setHeight2(double height2)\\n    public Locus getLocus()\\n    public void setLocus(Locus locus)\\n    public int getSiteCount()\\n    public boolean isValid()\\n    public void setConversionGraph(ConversionGraph acg)\\n    public void startEditing()\\n    public Conversion getCopy()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/Conversion.java', 'methodCount': 24, 'lineCount': 33}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class ConversionGraph {\\n    /**\\n * Unlike Trees, Conversion graphs require an alignment (or at least\\n * the length of an alignment) to be specified so that the regions of\\n * the alignment affected by recombination events can be recorded.\\n */\\npublic Input<List<Locus>> lociInput = new Input<>(\"locus\", \"Locus associated with graph.\", new ArrayList<>());;\\n    public Input<String> fromStringInput = new Input<>(\"fromString\", \"Initialise ARG from string representation.\");;\\n    public Input<String> fromExtNewickInput = new Input<>(\"extendedNewick\", \"Initialise ARG from extended Newick representation.\");;\\n    public Input<Boolean> wholeLocusConversionsInput = new Input<>(\"wholeLocusConversionsOnly\", \"Force region boundaries to coincide with locus boundaries.\", false);;\\n    /**\\n * List of recombinations on graph.\\n */\\nprotected Map<Locus, List<Conversion>> convs;;\\n    protected Map<Locus, List<Conversion>> storedConvs;;\\n    protected final static List<Conversion> emptyConvList = new ArrayList<>();;\\n    /**\\n * Event and region lists.\\n */\\nprotected Map<Locus, RegionList> regionLists;;\\n    protected CFEventList cfEventList;;\\n    protected List<Locus> loci, convertibleLoci;;\\n    protected int totalConvertibleSequenceLength;;\\n    public void initAndValidate()\\n    public int getTotalConvertibleSequenceLength()\\n    public Locus getLocusByID(String id)\\n    public List<Locus> getConvertibleLoci()\\n    public boolean wholeLocusModeOn()\\n    public void setWholeLocusMode(boolean newVal)\\n    public void addConversion(Conversion conv)\\n    public void deleteConversion(Conversion conv)\\n    public List<Conversion> getConversions(Locus locus)\\n    public int getConvCount(Locus locus)\\n    public int getTotalConvCount()\\n    public int getConversionIndex(Conversion conv)\\n    public List<Region> getRegions(Locus locus)\\n    public int getRegionCount(Locus locus)\\n    public int getUselessConvCount()\\n    public List<CFEventList.Event> getCFEvents()\\n    public double getClonalFrameLength()\\n    public boolean isInvalid()\\n    public String toString()\\n    public String toStringOld()\\n    public void fromStringOld(String str)\\n    public void fromXML(final org.w3c.dom.Node node)\\n    public ConversionGraph copy()\\n    private void generalAssignFrom(StateNode other, boolean fragileAssignment)\\n    public void assignFrom(StateNode other)\\n    public void assignFromFragile(StateNode other)\\n    public String getExtendedNewick()\\n    public String getExtendedNewick(boolean computeAffectedSites)\\n    public String getTrimmedExtendedNewick()\\n    private String extendedNewickTraverse(Node node, AffectedSiteList asList, boolean intraCFOnly)\\n    public void fromExtendedNewick(String string)\\n    public void fromExtendedNewick(String string, boolean numbered, int nodeNumberoffset)\\n    protected void store()\\n    public void restore()\\n    public void startEditing(Operator operator)\\n    public boolean clonalFrameIsDirty()\\n    public void init(PrintStream out)\\n    public void log(long nSample, PrintStream out)\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/ConversionGraph.java', 'methodCount': 38, 'lineCount': 51}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Locus {\\n    public Input<Integer> siteCountInput = new Input<>(\"siteCount\", \"Number of sites contained in alignments associated with this locus\");;\\n    public Input<Alignment> alignmentInput = new Input<>(\"alignment\", \"Initialize locus using this alignment.\", Input.Validate.XOR, siteCountInput);;\\n    public Input<Boolean> conversionsAllowedInput = new Input<>(\"conversionsAllowed\", \"If false, no conversions will be allowed on this locus. (Default true.)\", true);;\\n    protected int siteCount;;\\n    protected Alignment alignment;;\\n    public Locus();\\n    public Locus(String name, Alignment alignment);\\n    public Locus(String name, int siteCount);\\n    public void initAndValidate()\\n    public int getSiteCount()\\n    public Alignment getAlignment()\\n    public boolean hasAlignment()\\n    public boolean conversionsAllowed()\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/Locus.java', 'methodCount': 8, 'lineCount': 15}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class Region {\\n    public final int leftBoundary, rightBoundary;;\\n    public final Set<Conversion> activeConversions;;\\n    final int hashCodeCached;;\\n    public Region(int leftBoundary, int rightBoundary, Set<Conversion> activeConversions);\\n    public int getRegionLength()\\n    public boolean isClonalFrame()\\n    public String toString()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/tgvaughan,bacter/src/bacter/Region.java', 'methodCount': 6, 'lineCount': 11}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Obtain mean length of converted edges in ACG.\n",
      "     * \n",
      "     * @param acg\n",
      "     * @return mean length, or NaN if ACG has no converted edges\n",
      "     */\n",
      "public static double getMeanEdgeLength(ConversionGraph acg)\n",
      "**************************************\n",
      "public static double getMeanEdgeLength(ConversionGraph acg) {\n",
      "    if (acg.getTotalConvCount() < 1)\n",
      "        return Double.NaN;\n",
      "    double mean = 0.0;\n",
      "    for (Locus locus : acg.getConvertibleLoci()) {\n",
      "        for (Conversion conv : acg.getConversions(locus)) mean += conv.getHeight2() - conv.getHeight1();\n",
      "    }\n",
      "    mean /= acg.getTotalConvCount();\n",
      "    return mean;\n",
      "}\n",
      "1899\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 69/1243 [02:11<39:54,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3013\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 70/1243 [02:13<39:35,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3141\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 71/1243 [02:15<38:08,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3169\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 72/1243 [02:17<39:26,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3791\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 73/1243 [02:19<37:58,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3827\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class Util {\\n    public static TreeNode parseNewick(String newickString, TreeNode root)\\n    public static Set<T> difference(Set<T> set1, Set<T> set2)\\n    public static Set<T> union(Set<T> set1, Set<T> set2)\\n    public static Set<T> intersection(Set<T> set1, Set<T> set2)\\n    public static Set<T> subtract(Set<T> set1, Set<T> set2)\\n    public static List<Integer> sortedOrder(List<T> c, boolean ascending)\\n    public static List<T> enlist(T... objects)\\n    public static Map<T, Double> average(Map<T, Collection<? extends Number>> toAverage)\\n    public static List<Double> rescale(List<? extends Number> numbers)\\n    public static void save(Object obj, String fileName) throws FileNotFoundException, IOException\\n    public static void save(Object obj, OutputStream os) throws FileNotFoundException, IOException\\n    public static Object load(String fileName) throws ClassNotFoundException, FileNotFoundException, IOException\\n    public static double logSum(List<Double> logNumbers)\\n    public static double logSubtract(double log_x, double log_y)\\n    public static double sum(Collection<? extends Number> numbers)\\n    public static void remove(Collection<T> collection, Collection<T> toRemove)\\n    public static List<T> removeAll(Collection<T> collection, T toRemove)\\n    public static double sse(Collection<? extends Number> numbers)\\n    public static double sum_squares(Collection<? extends Number> numbers)\\n    public static double var(Collection<? extends Number> numbers)\\n    public static double cov(List<? extends Number> numbers_x, List<? extends Number> numbers_y)\\n    public static double entropy(List<Double> p_i)\\n    public static double JSD(List<Double>... p_i)\\n    public static double pearson_correlation(List<S> numbers_x, List<T> numbers_y)\\n    public static List<Double> rank(List<T> to_rank)\\n    public static double spearman_correlation(List<S> numbers_x, List<T> numbers_y)\\n    public static double mutual_information(List<Double> prob_xy, List<Double> prob_x, List<Double> prob_y)\\n    public static double r_squared(List<? extends Number> numbers_x, List<? extends Number> numbers_y)\\n    public static double stdev(Collection<? extends Number> numbers)\\n    public static double stderr(Collection<? extends Number> numbers)\\n    public static double mean(Collection<? extends Number> numbers)\\n    public static T quantile(Collection<T> numbers, double quantile)\\n    public static double median(Collection<? extends Number> numbers)\\n    public static double product(Collection<? extends Number> numbers)\\n    public static List<Byte> list(byte[] bytes)\\n    public static List<Short> list(short[] shorts)\\n    public static List<Integer> list(int[] ints)\\n    public static List<Long> list(long[] longs)\\n    public static List<Float> list(float[] floats)\\n    public static List<Double> list(double[] doubles)\\n    public static List<Character> list(char[] chars)\\n    public static List<Boolean> list(boolean[] booleans)\\n    public static Queue<T> queue(T... objects)\\n    public static List<T> list(T... objects)\\n    public static List<T> list(Collection<T> elements)\\n    public static Set<T> set(T... elements)\\n    public static Set<T> set(Collection<T> elements)\\n    public static List list()\\n    public static int baseValue(List<T> values, List<T> bitString)\\n    public static Map<S, Collection<T>> mapInverse(Map<T, S> map)\\n    public static Map<S, T> mapInvert(Map<T, S> map)\\n    public static List<T> select(T[] array, List<Integer> indices)\\n    public static int occurencesOf(Collection<T> c, T obj)\\n    public static List<Integer> indicesOf(List<T> list, T obj)\\n    public static List<T> select(List<T> list, List<Integer> indices)\\n    public static Collection<T> combine(Collection<T>... collections)\\n    public static Collection<T> combine(Collection<? extends Collection<T>> collections)\\n    public static List<T> reverse(List<T> toReverse)\\n    public static T copy(T t)\\n    public static List<T> cast(Class<? extends T> clazz, Collection<?> c)\\n    public static Map<S, T> map(List<S> keys, List<T> values)\\n    public static List<List<T>> enumerateObjects(List<T> possibleValues, int listSize)\\n    public static T sample(Collection<T> states)\\n    public static List<T> sample(List<Double> probabilities, List<T> states, int number)\\n    public static T sample(List<Double> probabilities, List<T> states)\\n    public static List<Double> normalize(List<? extends Number> probabilities)\\n    public static List<Double> standardize(List<? extends Number> numbers)\\n    public static List<Double> scale(List<? extends Number> numbers, double min, double max)\\n    public static List<Integer> seq(int j)\\n    public static List<Integer> seq(int start, int end)\\n    public static Double floor(Double[] items, double key)\\n    public static Double ceiling(Double[] items, double key)\\n    public static void printf(String string, Object... objects)\\n    public static T getObj(Collection<T> collection, T toGet)\\n    public static List<T> get(Map<S, T> map, Collection<S> list)\\n    public static double logit(double x, double center, double scale)\\n    public static String sprintf(String format, Object... toPrint)\\n    public static T get(Collection<T> collection, T toGet)\\n    public static List<Integer> rank(List<S> values, boolean ascending)\\n    public static List<T> sort(List<T> listToSort, List<S> values, boolean ascending)\\n    public static List<Tuple<S, T>> sortTuple(List<T> listToSort, List<S> values, boolean ascending)\\n    public static Color hsba(Number h, Number s, Number b, Number a)\\n    public static Color hsb(Number h, Number s, Number b)\\n    public static boolean notNull(Object... objects)\\n    public static List<T> repeat(T t, int size)\\n    public static List<T> filter(Collection<T> toFilter, Criteria<T> criteria)\\n    public static List<T> evaluate(Collection<S> domain, Function<S, T> function)\\n    public static List<T> convolve(List<S> values, Function<List<S>, T> function, int bandWidth)\\n    public static List<Double> convolution(List<S> values, Function<List<S>, T> function, int bandWidth, double[] p)\\n    public static boolean anySatisfy(Iterable<T> toCheck, Criteria<T> criteria)\\n    public static boolean allSatisfy(Collection<T> toCheck, Criteria<T> criteria)\\n    public static List<Integer> convertNumber(Collection<S> numbers, int i)\\n    public static List<Byte> convertNumber(Collection<S> numbers, byte b)\\n    public static List<Double> convertNumber(Collection<S> numbers, double b)\\n    public static List<Short> convertNumber(Collection<S> numbers, short s)\\n    public static List<Long> convertNumber(Collection<S> numbers, long l)\\n    public static List<Float> convertNumber(Collection<S> numbers, float f)\\n    public static List<Color> colorGradient(int n)\\n    public static List<Color> colorGradient(int n, double a)\\n    public static String wrap(String string, int width)\\n    public static void copy(File src, File dest) throws IOException\\n    public static List<T> bin(List<S> list, int binWidth, Function<List<S>, T> binningFunction)\\n    public static void initialize(T[] array, Instantiator<T> instantiator, Object... objects)\\n}\\n', 'filePath': '/home/ubuntu/repos/shenkers,isoscm/IsoSCM/src/util/Util.java', 'methodCount': 103, 'lineCount': 105}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'IntervalNode': 'public class IntervalNode {\\n    /**\\n */\\nprivate static final long serialVersionUID = 1L;;\\n    /**\\n * Left sub-tree.\\n */\\nIntervalNode<S> left;;\\n    /**\\n * Right sub-tree.\\n */\\nIntervalNode<S> right;;\\n    /**\\n * Parent tree.\\n */\\nIntervalNode<S> parent;;\\n    /**\\n * Skew factor.\\n */\\nSkew skew;;\\n    int maxEnd;;\\n    int start;;\\n    int end;;\\n    boolean rebalance;;\\n    public S data;;\\n     IntervalNode(int start, int end, final IntervalNode<S> parent);\\n     IntervalNode(int start, int end, final IntervalNode<S> parent, S data);\\n    public int getStart()\\n    public int getEnd()\\n     boolean overlaps(int start, int end)\\n    public List<IntervalNode<S>> getOverlappingIntervals(int start, int end)\\n    public List<IntervalNode<S>> getOverlappingIntervals(int start, int end, List<IntervalNode<S>> overlaps)\\n     int size()\\n     IntervalNode<S> getSmallest()\\n     IntervalNode<S> getLargest()\\n    public IntervalNode<S> getPrevious()\\n    public IntervalNode<S> getNext()\\n     IntervalNode<S> insert(int start, int end)\\n     IntervalNode<S> insert(int start, int end, S data)\\n    public void delete()\\n    private boolean rebalanceLeftGrown()\\n    private boolean rebalanceRightGrown()\\n    private boolean rebalanceLeftShrunk()\\n    private boolean rebalanceRightShrunk()\\n    private void rotateCW()\\n    private void rotateCCW()\\n    private boolean updateMaxEnd()\\n    private void propogateUpdate()\\n    public String toString()\\n    public String toNewickString()\\n}\\n'}\n",
      "/** Get the node whose element is not smaller than the reference object.\n",
      "\t * @param reference reference object (may not be in the tree)\n",
      "\t * @return the tree node containing the smallest element not smaller\n",
      "\t * than the reference object or null if either the tree is empty or\n",
      "\t * all its elements are smaller than the reference object\n",
      "\t * @see #getSmallest\n",
      "\t * @see #getLargest\n",
      "\t * @see #getNotLarger\n",
      "\t * @see Node#getPrevious\n",
      "\t * @see Node#getNext\n",
      "\t */\n",
      "public IntervalNode<S> getNotSmaller(int start)\n",
      "**************************************\n",
      "public IntervalNode<S> getNotSmaller(int start) {\n",
      "    IntervalNode<S> candidate = null;\n",
      "    for (IntervalNode<S> node = top; node != null; ) {\n",
      "        if (node.start < start) {\n",
      "            if (node.right == null) {\n",
      "                return candidate;\n",
      "            }\n",
      "            node = node.right;\n",
      "        } else {\n",
      "            candidate = node;\n",
      "            if (node.left == null) {\n",
      "                return candidate;\n",
      "            }\n",
      "            node = node.left;\n",
      "        }\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "2438\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 74/1243 [02:22<42:48,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3828\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 75/1243 [02:24<42:29,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 76/1243 [02:25<37:54,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4458\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 77/1243 [02:27<36:10,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4474\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 78/1243 [02:29<34:37,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4531\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "\t * Row normalize the current gene expression set. Row normalization involved\n",
      "\t * subtracting the mena of the row from each expression value in the row and\n",
      "\t * subsequently dividing it by the standard deviation of the expression row.\n",
      "\t *\n",
      "\t * @return an array of the row normalized values of the gene expression set.\n",
      "\t */\n",
      "public float[] rowNormalize()\n",
      "**************************************\n",
      "/**\n",
      " * Row normalize the current gene expression set. Row normalization involved\n",
      " * subtracting the mena of the row from each expression value in the row and\n",
      " * subsequently dividing it by the standard deviation of the expression row.\n",
      " *\n",
      " * @return an array of the row normalized values of the gene expression set.\n",
      " */\n",
      "public float[] rowNormalize() {\n",
      "    float[] normalize = new float[expression.length];\n",
      "    float mean = mean();\n",
      "    float std = std(mean);\n",
      "    if (std == 0.0) {\n",
      "        for (int i = 0; i < expression.length; i++) normalize[i] = 0.0f;\n",
      "    } else {\n",
      "        for (int i = 0; i < expression.length; i++) normalize[i] = (expression[i] - mean) / std;\n",
      "    }\n",
      "    return normalize;\n",
      "}\n",
      "284\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 79/1243 [02:31<36:38,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4595\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   6%|         | 80/1243 [02:33<38:08,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4602\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 81/1243 [02:35<36:17,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4683\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 82/1243 [02:37<40:28,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4926\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 83/1243 [02:39<37:46,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4970\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "\t * Get scores array (elements are in no particualr order)\n",
      "\t * \n",
      "\t * @return double[] scores\n",
      "\t */\n",
      "public synchronized double[] getScores()\n",
      "**************************************\n",
      "public synchronized double[] getScores() {\n",
      "    if (scores == null) {\n",
      "        Map<Integer, Double> gene2score = getGene2Score();\n",
      "        scores = new double[gene2score.size()];\n",
      "        int i = 0;\n",
      "        for (Double score : gene2score.values()) {\n",
      "            scores[i++] = score;\n",
      "        }\n",
      "    }\n",
      "    return scores;\n",
      "}\n",
      "150\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 84/1243 [02:41<41:21,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4992\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 85/1243 [02:43<40:31,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5437\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 86/1243 [02:45<40:22,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5468\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 87/1243 [02:47<37:12,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6106\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 88/1243 [02:49<35:32,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6114\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class Config {\\n    // C O N S T R U C T O R\\nprivate static final Map<ConfigKey, String> config = new HashMap<ConfigKey, String>();;\\n    public Config(String source) throws IOException, InvalidConfigException;\\n    public static String help()\\n    private static String getConfigFileAsString(String source) throws IOException, InvalidConfigException\\n    private static void colorNameToInt() throws InvalidColourException\\n    public static int get256Color(ConfigKey key) throws InvalidColourException\\n    public static String get(ConfigKey key)\\n    public static void set(ConfigKey key, String value) throws InvalidColourException\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/coloring/Config.java', 'methodCount': 7, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/coloring/ConfigKey.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidColourException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 1L;;\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/exceptions/InvalidColourException.java', 'methodCount': 0, 'lineCount': 3}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidCommandLineException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 1L;;\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/exceptions/InvalidCommandLineException.java', 'methodCount': 0, 'lineCount': 3}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidGenomicCoordsException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 1L;;\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/exceptions/InvalidGenomicCoordsException.java', 'methodCount': 0, 'lineCount': 3}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class Faidx {\\n    public Faidx(File fasta) throws IOException, UnindexableFastaFileException;\\n    private boolean isCompressed(File fasta) throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/faidx/Faidx.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 14}\n",
      "{}\n",
      "/** For debugging only \n",
      "\t * @throws IOException \n",
      "\t * @throws InvalidGenomicCoordsException */\n",
      "public String toStringVerbose(int windowSize) throws InvalidGenomicCoordsException, IOException\n",
      "**************************************\n",
      "public String toStringVerbose(int windowSize) throws InvalidGenomicCoordsException, IOException {\n",
      "    List<Double> mapping = seqFromToLenOut(this.getUserWindowSize());\n",
      "    String str = \"Genome coords: \" + from + \"-\" + to + \"; screen width: \" + mapping.size() + \"; scale: \" + this.getBpPerScreenColumn() + \" bp/column\" + \"; Mapping: \" + mapping;\n",
      "    str += \"\\n\";\n",
      "    str += this.toString();\n",
      "    return str;\n",
      "}\n",
      "802\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 89/1243 [02:51<39:33,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6244\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 90/1243 [02:53<37:28,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6309\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 91/1243 [02:55<37:56,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6665\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 92/1243 [02:57<39:52,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6700\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   7%|         | 93/1243 [02:59<40:00,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6706\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class Config {\\n    // C O N S T R U C T O R\\nprivate static final Map<ConfigKey, String> config = new HashMap<ConfigKey, String>();;\\n    public Config(String source) throws IOException, InvalidConfigException;\\n    public static String help()\\n    private static String getConfigFileAsString(String source) throws IOException, InvalidConfigException\\n    private static void colorNameToInt() throws InvalidColourException\\n    public static int get256Color(ConfigKey key) throws InvalidColourException\\n    public static String get(ConfigKey key)\\n    public static void set(ConfigKey key, String value) throws InvalidColourException\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/coloring/Config.java', 'methodCount': 7, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/coloring/ConfigKey.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidColourException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 1L;;\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/exceptions/InvalidColourException.java', 'methodCount': 0, 'lineCount': 3}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidGenomicCoordsException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 1L;;\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/exceptions/InvalidGenomicCoordsException.java', 'methodCount': 0, 'lineCount': 3}, {'fieldCount': 12, 'classCount': 1, 'methodsOnly': 'public class GenomicCoords {\\n    // Background character as spacer\\npublic static final String SPACER = \"-\";;\\n    // Char to use to mark region\\npublic static final String TICKED = \"*\";;\\n    private String chrom;;\\n    private Integer from;;\\n    private Integer to;;\\n    // Can be null\\nprivate SAMSequenceDictionary samSeqDict;;\\n    /**\\n * Size of the screen window\\n * Only user getUserWindowSize() to access it after init at constructor.\\n */\\nprivate String fastaFile = null;;\\n    // Source of the sequence dictionary. Can be path to file of fasta, bam, genome. Or genome tag e.g. hg19.\\nprivate String samSeqDictSource = null;;\\n    private byte[] refSeq = null;;\\n    public boolean isSingleBaseResolution = false;;\\n    private int terminalWidth;;\\n    private List<Double> mapping;;\\n    public GenomicCoords(String region, int terminalWidth, SAMSequenceDictionary samSeqDict, String fastaFile, boolean verbose) throws InvalidGenomicCoordsException, IOException;\\n    public GenomicCoords(String region, int terminalWidth, SAMSequenceDictionary samSeqDict, String fastaFile) throws InvalidGenomicCoordsException, IOException;\\n     GenomicCoords(int terminalWidth) throws InvalidGenomicCoordsException, IOException;\\n    private void update() throws InvalidGenomicCoordsException, IOException\\n    public void setGenome(List<String> input, boolean includeGenomeFile) throws IOException\\n    protected void setRefSeq() throws IOException, InvalidGenomicCoordsException\\n    public byte[] getRefSeq() throws IOException, InvalidGenomicCoordsException\\n    public byte[] getSequenceFromFasta() throws IOException\\n    private GenomicCoords parseStringToGenomicCoords(String x) throws InvalidGenomicCoordsException, IOException\\n    private String[] parseStringSpaceSep(String x) throws InvalidGenomicCoordsException\\n    private String[] parseStringColonSep(String x) throws InvalidGenomicCoordsException\\n    public void correctCoordsAgainstSeqDict(SAMSequenceDictionary samSeqDict) throws InvalidGenomicCoordsException, IOException\\n    public String toString()\\n    public String toStringRegion()\\n    private int getMidpoint()\\n    public void zoomOut() throws IOException, InvalidGenomicCoordsException\\n    public void zoomIn() throws IOException, InvalidGenomicCoordsException\\n    public void left() throws InvalidGenomicCoordsException, IOException\\n    public void right() throws InvalidGenomicCoordsException, IOException\\n    private List<Double> seqFromToLenOut(int size) throws InvalidGenomicCoordsException, IOException\\n    public double getBpPerScreenColumn() throws InvalidGenomicCoordsException, IOException\\n    public String getChromIdeogram(int nDist, boolean noFormat) throws InvalidGenomicCoordsException, IOException, InvalidColourException\\n    public String toStringVerbose(int windowSize) throws InvalidGenomicCoordsException, IOException\\n    public String printableGenomicRuler(int markDist, boolean noFormat) throws InvalidGenomicCoordsException, IOException, InvalidColourException\\n    public String printablePercentRuler(int markDist, boolean noFormat) throws InvalidGenomicCoordsException, IOException, InvalidColourException\\n    private String printRulerFromList(List<Double> marks, int markDist, int digits) throws InvalidGenomicCoordsException, IOException\\n    public String printableRefSeq(boolean noFormat) throws IOException, InvalidGenomicCoordsException, InvalidColourException\\n    private boolean setSamSeqDictFromAnySource(List<String> testfiles, boolean includeGenomeFile) throws IOException\\n    private boolean setSamSeqDictFromVCF(String vcf) throws MalformedURLException\\n    private boolean setSamSeqDictFromFasta(String fasta) throws IOException\\n    private boolean setSamSeqDictFromBam(String bamfile)\\n    private boolean setSamSeqDictFromGenomeFile(String genome) throws IOException\\n    public boolean equalCoords(GenomicCoords other)\\n    public boolean equalCoordsAndWindowSize(GenomicCoords other) throws InvalidGenomicCoordsException, IOException\\n    public Object clone()\\n    public void centerAndExtendGenomicCoords(GenomicCoords gc, int size, double slop) throws InvalidGenomicCoordsException, IOException\\n    private int getTerminalWidth()\\n    public List<Double> getMapping()\\n    public String getChrom()\\n    public Integer getFrom()\\n    public Integer getTo()\\n    public int getUserWindowSize() throws InvalidGenomicCoordsException, IOException\\n    public Integer getGenomicWindowSize() throws InvalidGenomicCoordsException\\n    public SAMSequenceDictionary getSamSeqDict()\\n    public void setSamSeqDict(SAMSequenceDictionary samSeqDict)\\n    public String getFastaFile()\\n    protected void setFastaFile(String fastaFile)\\n    public String getSamSeqDictSource()\\n    public void setSamSeqDictSource(String samSeqDictSource)\\n    private void extend(int left, int right, String refpoint) throws InvalidGenomicCoordsException, IOException\\n    public void cmdInputExtend(List<String> cmdInput) throws InvalidCommandLineException, InvalidGenomicCoordsException, IOException\\n    public void setTerminalWidth(int terminalWidth) throws InvalidGenomicCoordsException, IOException\\n    protected void setSingleBaseResolution() throws InvalidGenomicCoordsException, IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/samTextViewer/GenomicCoords.java', 'methodCount': 53, 'lineCount': 67}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    public static double round(double value, int places)\\n    public static File createTempFile(String prefix, String suffix, boolean deleteOnExit)\\n    public static BufferedReader reader(String fileOrUrl) throws MalformedURLException, IOException\\n    public static String templateNameFromSamReadName(String readName)\\n    public static boolean argListContainsFlag(List<String> argList, String flag)\\n    public static List<String> getNArgsForParam(List<String> argList, String param, int nargs) throws InvalidCommandLineException\\n    public static String getArgForParam(List<String> argList, String param, String defArg) throws InvalidCommandLineException\\n    public static void checkFasta(String fasta, int debug) throws IOException, UnindexableFastaFileException\\n    public static long getAlignedReadCount(String bam) throws IOException\\n    public static List<IntervalFeature> mergeIntervalFeatures(List<IntervalFeature> intervalList, boolean screenCoords) throws InvalidGenomicCoordsException, InvalidColourException\\n    public static boolean hasTabixIndex(String fileName) throws IOException\\n    public static String initRegionFromFile(String x) throws IOException, InvalidGenomicCoordsException, ClassNotFoundException, InvalidCommandLineException, InvalidRecordException, SQLException\\n    private static String initRegionFromBigBed(String bigBedFile) throws IOException\\n    private static String initRegionFromBigWig(String bigWigFile) throws IOException\\n    public static boolean bamHasIndex(String bam) throws IOException\\n    public static TrackFormat getFileTypeFromName(String fileName)\\n    public static List<List<T>> transpose(List<List<T>> table)\\n    public static List<Double> collapseValues(List<Double> values, List<Integer> valuePositions, List<Double> referencePositions)\\n    public static byte[] prepareRefSeq(String fasta, GenomicCoords gc) throws IOException\\n    private static int parseStringToIntWithUnits(String x)\\n    public static String parseConsoleInput(List<String> tokens, GenomicCoords gc) throws InvalidGenomicCoordsException, IOException, InvalidCommandLineException\\n    protected static String parseGoToRegion(String rawInput)\\n    public static boolean isInteger(String s)\\n    public static Double calculateAverage(List<Double> list)\\n    public static int getIndexOfclosestValue(double genomePos, List<Double> mapping)\\n    public static boolean allIsNaN(Iterable<Double> x)\\n    public static List<Double> seqFromToLenOut(double from, double to, int lengthOut)\\n    public static List<String> tabulateList(List<String> rawList, int terminalWidth, String columnSep)\\n    public static String[] roundToSignificantDigits(double x, double y, int nSignif)\\n    public static double roundToSignificantFigures(double num, int n)\\n    public static String parseIntToMetricSuffix(int x)\\n    public static boolean urlFileExists(String URLName)\\n    public static void addSourceName(List<String> inputFileList, List<String> newFileNames, int debug) throws InvalidCommandLineException, IOException\\n    public static TrackFormat sniffFile(File x) throws IOException\\n    public static String printSamSeqDict(SAMSequenceDictionary samSeqDict, int graphSize)\\n    public static int parseZoom(String cmdInput, int defaultInt) throws InvalidCommandLineException\\n    public static ArrayList<String> tokenize(String x, String delimiterString)\\n    public static String stripAnsiCodes(String x)\\n    public static String parseCmdinputToGetSnapshotFile(String cmdInput, GenomicCoords gc) throws IOException\\n    public static String tildeToHomeDir(String path)\\n    public static long countReadsInWindow(String bam, GenomicCoords gc, List<SamRecordFilter> filters) throws MalformedURLException\\n    public static Map<K, V> sortByValue(Map<K, V> map)\\n    public static TabixFormat trackFormatToTabixFormat(TrackFormat fmt)\\n    public static Float[] range(List<Float> y)\\n    public static String coordinatesToString(String chrom, Integer from, Integer to)\\n    public static List<String> globFiles(List<String> cmdInput) throws IOException\\n    private static List<Path> match(String glob, String location) throws IOException\\n    protected static List<String> checkUpdates(long timeout) throws IOException\\n    public static int versionCompare(String str1, String str2)\\n    public static boolean[] passAwkFilter(String[] rawLines, String awkScript) throws IOException\\n    public static String padEndMultiLine(String x, int size)\\n    public static List<String> parseStringCoordsToList(String region) throws InvalidGenomicCoordsException\\n    public static SamReader getSamReader(String workFilename) throws MalformedURLException\\n    public static List<Integer> indexOfCharsOnFormattedLine(String fmtString)\\n    public static int getTerminalWidth() throws IOException\\n    public static VCFHeader getVCFHeader(String source) throws MalformedURLException\\n    public static VCFHeaderVersion getVCFHeaderVersion(VCFHeader vcfHeader)\\n    public static void sortAndIndexSamOrBam(String inSamOrBam, String sortedBam, boolean deleteOnExit) throws IOException\\n    public static boolean equalReadNames(String readName, String readName2)\\n    private static String cleanSamReadName(String readName)\\n    public static List<String> vcfHeaderToStrings(VCFHeader header)\\n    public static String roundNumbers(final String x, int d, TrackFormat trackFormat)\\n    public static List<Float> winsor2(List<Float> x, double multiple)\\n    private static float medianAbsoluteDeviation(List<Float> x, float center)\\n    public static Boolean asBoolean(String x)\\n    public static List<String> suggestCommand(String hint, List<String> options)\\n    private static int levenshtein(String a, String b)\\n    public static boolean isOverlapping(int xStart, int xEnd, int yStart, int yEnd)\\n    public static String reformatFileName(String filename, boolean absolute)\\n    public static boolean[] matchByAwkOrRegex(String[] rawLines, String regex) throws IOException\\n    public static String quote(String x)\\n    public static TrackFormat guessTrackType(String file)\\n}\\n', 'filePath': '/home/ubuntu/repos/dariober,ASCIIGenome/src/main/java/samTextViewer/Utils.java', 'methodCount': 72, 'lineCount': 74}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "\t * Obtain the start position of the read on screen. Screen positions are 1-based. \n",
      "\t * @return\n",
      "\t * @throws IOException \n",
      "\t * @throws InvalidGenomicCoordsException \n",
      "\t */\n",
      "private void setTextStart() throws InvalidGenomicCoordsException, IOException\n",
      "**************************************\n",
      "private void setTextStart() throws InvalidGenomicCoordsException, IOException {\n",
      "    int alnStart;\n",
      "    if (this.showSoftClip) {\n",
      "        alnStart = this.getSoftUnclippedAlignmentStart(samRecord);\n",
      "    } else {\n",
      "        alnStart = samRecord.getAlignmentStart();\n",
      "    }\n",
      "    if (alnStart <= gc.getFrom()) {\n",
      "        this.textStart = 1;\n",
      "        return;\n",
      "    }\n",
      "    this.textStart = Utils.getIndexOfclosestValue(alnStart, gc.getMapping()) + 1;\n",
      "    return;\n",
      "}\n",
      "3179\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 94/1243 [03:02<40:40,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6721\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 95/1243 [03:03<37:38,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6811\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 96/1243 [03:05<36:06,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6813\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 97/1243 [03:07<34:40,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6815\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 98/1243 [03:09<36:12,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6817\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Determines whether a SAMRecord matches this filter\n",
      "     *\n",
      "     * @param record the SAMRecord to evaluate\n",
      "     *\n",
      "     * @return true if the SAMRecord matches the filter, otherwise false\n",
      "     */\n",
      "public boolean filterOut(final SAMRecord record)\n",
      "**************************************\n",
      "/**\n",
      " * Determines whether a SAMRecord matches this filter\n",
      " *\n",
      " * @param record the SAMRecord to evaluate\n",
      " *\n",
      " * @return true if the SAMRecord matches the filter, otherwise false\n",
      " */\n",
      "public boolean filterOut(final SAMRecord record) {\n",
      "    if (include) {\n",
      "        if (record.getMateUnmappedFlag()) {\n",
      "            return false;\n",
      "        }\n",
      "    } else {\n",
      "        // exclude\n",
      "        if (!record.getMateUnmappedFlag()) {\n",
      "            return false;\n",
      "        }\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "195\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 99/1243 [03:11<37:33,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6819\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 100/1243 [03:12<35:22,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6821\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 101/1243 [03:14<33:43,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6823\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 102/1243 [03:16<35:36,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6825\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 103/1243 [03:18<36:32,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6827\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Determines whether a SAMRecord matches this filter\n",
      "     *\n",
      "     * @param record the SAMRecord to evaluate\n",
      "     *\n",
      "     * @return true if the SAMRecord matches the filter, otherwise false\n",
      "     */\n",
      "public boolean filterOut(final SAMRecord record)\n",
      "**************************************\n",
      "/**\n",
      " * Determines whether a SAMRecord matches this filter\n",
      " *\n",
      " * @param record the SAMRecord to evaluate\n",
      " *\n",
      " * @return true if the SAMRecord matches the filter, otherwise false\n",
      " */\n",
      "public boolean filterOut(final SAMRecord record) {\n",
      "    if (include) {\n",
      "        if (record.getSupplementaryAlignmentFlag()) {\n",
      "            return false;\n",
      "        }\n",
      "    } else {\n",
      "        // exclude\n",
      "        if (!record.getSupplementaryAlignmentFlag()) {\n",
      "            return false;\n",
      "        }\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "195\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 104/1243 [03:20<34:36,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6830\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   8%|         | 105/1243 [03:21<33:18,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6832\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 106/1243 [03:23<35:28,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6844\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 107/1243 [03:26<37:02,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6845\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 108/1243 [03:28<40:42,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6858\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/mendel/ChromosomeType.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Genotype {\\n    public static final int NO_CALL = -1;;\\n    public static final int REF_CALL = 0;;\\n    /**\\n * List of allele numbers/calls\\n */\\nprivate final int[] alleleNumbers;;\\n    private final int ploidy;;\\n    private final boolean hasNoObservedCalls;;\\n    private final boolean isHet;;\\n    private final boolean isHomRef;;\\n    private final boolean isHomAlt;;\\n    public Genotype(Collection<Integer> alleleNumbers);\\n    private int[] findObservedCalls(int numCalls, int[] alleleNumbers)\\n    private boolean calculateIsHet(int ploidy, boolean hasNoObservedCalls, int[] alleleNumbers)\\n    private boolean calculateIsHomRef(boolean hasNoObservedCalls, int[] alleleNumbers)\\n    private boolean calculateIsHomAlt(boolean hasNoObservedCalls, int[] observedCalls)\\n    private boolean hasRefCall(int[] observedCalls)\\n    public ImmutableList<Integer> getAlleleNumbers()\\n    public int getPloidy()\\n    public boolean isDiploid()\\n    public boolean isMonoploid()\\n    public boolean isHet()\\n    public boolean isHomRef()\\n    public boolean isHomAlt()\\n    public boolean isNotObserved()\\n    public String toString()\\n    private String makeAlleleNumbersString()\\n    public int hashCode()\\n    public boolean equals(Object o)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/mendel/Genotype.java', 'methodCount': 18, 'lineCount': 28}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class GenotypeCalls {\\n    private final static Genotype GT_NO_CALL = new Genotype(ImmutableList.of(Genotype.NO_CALL));;\\n    /**\\n * Type of the chromosome that the variant lies on (autosomal, X-chromosomal, etc.)\\n */\\nprivate final ChromosomeType chromType;;\\n    /**\\n * Mapping from sample name to {@link Genotype}\\n */\\nprivate final ImmutableSortedMap<String, Genotype> sampleToGenotype;;\\n    /**\\n * List of sample names\\n */\\nprivate final ImmutableList<String> sampleNames;;\\n    /**\\n * A payload object for later easier reidentification\\n */\\nprivate final Object payload;;\\n    public GenotypeCalls(ChromosomeType chromType, Iterable<? extends Entry<String, Genotype>> sampleToGenotype);\\n    public GenotypeCalls(ChromosomeType chromType, Iterable<? extends Entry<String, Genotype>> sampleToGenotype, Object payload);\\n    public int getNSamples()\\n    public Genotype getGenotypeForSample(String sample)\\n    public Genotype getGenotypeBySampleNo(int sampleNo)\\n    public ChromosomeType getChromType()\\n    public ImmutableSortedMap<String, Genotype> getSampleToGenotype()\\n    public ImmutableList<String> getSampleNames()\\n    public Object getPayload()\\n    public String toString()\\n    public Iterator<Entry<String, Genotype>> iterator()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/mendel/GenotypeCalls.java', 'methodCount': 13, 'lineCount': 20}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MendelianInheritanceChecker {\\n    /**\\n * Pedigree to use for mendelian inheritance checking\\n */\\nfinal private Pedigree pedigree;;\\n    /**\\n * Helper for querying a pedigree\\n */\\nfinal private PedigreeQueryDecorator queryPed;;\\n    /**\\n * Mendelian compatibility checker for each sub mode of inheritance\\n */\\nfinal private ImmutableMap<SubModeOfInheritance, AbstractMendelianChecker> checkers;;\\n    public MendelianInheritanceChecker(Pedigree pedigree);\\n    public ImmutableMap<ModeOfInheritance, ImmutableList<GenotypeCalls>> checkMendelianInheritance(Collection<GenotypeCalls> calls, Collection<GenotypeCalls> recessiveCalls) throws IncompatiblePedigreeException\\n    public ImmutableMap<ModeOfInheritance, ImmutableList<GenotypeCalls>> checkMendelianInheritance(Collection<GenotypeCalls> calls) throws IncompatiblePedigreeException\\n    public ImmutableMap<SubModeOfInheritance, ImmutableList<GenotypeCalls>> checkMendelianInheritanceSub(Collection<GenotypeCalls> calls, Collection<GenotypeCalls> compHetRecessiveCalls) throws IncompatiblePedigreeException\\n    public ImmutableMap<SubModeOfInheritance, ImmutableList<GenotypeCalls>> checkMendelianInheritanceSub(Collection<GenotypeCalls> calls) throws IncompatiblePedigreeException\\n    public ImmutableList<GenotypeCalls> filterCompatibleRecords(Collection<GenotypeCalls> calls, ModeOfInheritance mode) throws IncompatiblePedigreeException\\n    public ImmutableList<GenotypeCalls> filterCompatibleRecordsSub(Collection<GenotypeCalls> calls, SubModeOfInheritance subMode) throws IncompatiblePedigreeException\\n    public Pedigree getPedigree()\\n    private boolean isCompatibleWithPedigree(GenotypeCalls calls)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/mendel/MendelianInheritanceChecker.java', 'methodCount': 9, 'lineCount': 14}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/pedigree/Disease.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': \"public class Person {\\n    /**\\n * the individual's name\\n */\\nprivate final String name;;\\n    /**\\n * the individual's father, or <code>null</code> if father is not in pedigree\\n */\\nprivate final Person father;;\\n    /**\\n * the individual's mother, or <code>null</code> if mother is not in pedigree\\n */\\nprivate final Person mother;;\\n    /**\\n * the individual's sex\\n */\\nprivate final Sex sex;;\\n    /**\\n * the individual's disease state\\n */\\nprivate final Disease disease;;\\n    /**\\n * extra fields from the PED file\\n */\\nprivate final ImmutableList<String> extraFields;;\\n    public Person(String name, Person father, Person mother, Sex sex, Disease disease, Collection<String> extraFields);\\n    public Person(String name, Person father, Person mother, Sex sex, Disease disease);\\n     Person(PedPerson pedPerson, PedFileContents pedFileContents, HashMap<String, Person> existing);\\n    public String getName()\\n    public Person getFather()\\n    public Person getMother()\\n    public Sex getSex()\\n    public Disease getDisease()\\n    public ImmutableList<String> getExtraFields()\\n    public boolean isFounder()\\n    public boolean isMale()\\n    public boolean isFemale()\\n    public boolean isAffected()\\n    public boolean isUnaffected()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n}\\n\", 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/pedigree/Person.java', 'methodCount': 17, 'lineCount': 25}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/pedigree/Sex.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "\t * @return whether <code>calls</code> is compatible with AD inheritance in the case of a single individual in the\n",
      "\t * pedigree\n",
      "\t */\n",
      "private boolean isCompatibleSingleton(GenotypeCalls calls)\n",
      "**************************************\n",
      "private boolean isCompatibleSingleton(GenotypeCalls calls) {\n",
      "    if (calls.getNSamples() == 0)\n",
      "        return false;\n",
      "    final Genotype gt = calls.getGenotypeBySampleNo(0);\n",
      "    if (pedigree.getMembers().get(0).getSex() == Sex.FEMALE) {\n",
      "        return calls.getGenotypeBySampleNo(0).isHet();\n",
      "    } else {\n",
      "        return (gt.isHet() || gt.isHomAlt());\n",
      "    }\n",
      "}\n",
      "2072\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 109/1243 [03:31<43:39,  2.31s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6868\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 110/1243 [03:32<39:16,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6871\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 111/1243 [03:35<39:29,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6884\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 112/1243 [03:37<39:23,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6921\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 113/1243 [03:39<40:51,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6950\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'GenomePosition': 'public class GenomePosition {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nprivate final ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the position on the chromosome\\n */\\nprivate final int pos;;\\n    public GenomePosition(ReferenceDictionary refDict, Strand strand, int chr, int pos);\\n    public GenomePosition(ReferenceDictionary refDict, Strand strand, int chr, int pos, PositionType positionType);\\n    public GenomePosition(GenomePosition other);\\n    public GenomePosition(GenomePosition other, Strand strand);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getPos()\\n    public GenomePosition withStrand(Strand strand)\\n    public boolean isLt(GenomePosition other)\\n    public boolean isLeq(GenomePosition other)\\n    public boolean isGt(GenomePosition other)\\n    public boolean isGeq(GenomePosition other)\\n    public boolean isEq(GenomePosition other)\\n    public int differenceTo(GenomePosition pos)\\n    public int differenceTo(GenomeInterval itv)\\n    public GenomePosition shifted(int delta)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomePosition other)\\n}\\n', 'GenomeInterval': 'public class GenomeInterval {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nfinal ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the begin position on the chromosome\\n */\\nprivate final int beginPos;;\\n    /**\\n * the end position on the chromosome\\n */\\nprivate final int endPos;;\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos);\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos, PositionType positionType);\\n    public GenomeInterval(GenomePosition pos, int length);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getBeginPos()\\n    public int getEndPos()\\n    public GenomeInterval withStrand(Strand strand)\\n    public GenomePosition getGenomeBeginPos()\\n    public GenomePosition getGenomeEndPos()\\n    public int length()\\n    public GenomeInterval intersection(GenomeInterval other)\\n    public GenomeInterval union(GenomeInterval other)\\n    public boolean isLeftOf(GenomePosition pos)\\n    public boolean isRightOf(GenomePosition pos)\\n    public boolean isLeftOfGap(GenomePosition pos)\\n    private GenomePosition ensureSameStrand(GenomePosition pos)\\n    public boolean isRightOfGap(GenomePosition pos)\\n    public boolean contains(GenomePosition pos)\\n    public boolean contains(GenomeInterval other)\\n    public GenomeInterval withMorePadding(int padding)\\n    public GenomeInterval withMorePadding(int paddingUpstream, int paddingDownstream)\\n    public boolean overlapsWith(GenomeInterval other)\\n    private GenomeInterval ensureSameStrand(GenomeInterval other)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomeInterval other)\\n}\\n'}\n",
      "/**\n",
      "\t * @return <code>true</code> if <code>pos</code> lies within an intron of {@link #transcript}\n",
      "\t */\n",
      "public boolean liesInIntron(GenomePosition pos)\n",
      "**************************************\n",
      "public boolean liesInIntron(GenomePosition pos) {\n",
      "    for (int i = 0; i + 1 < transcript.getExonRegions().size(); ++i) {\n",
      "        GenomeInterval intronRegion = transcript.intronRegion(i);\n",
      "        if (intronRegion.contains(pos))\n",
      "            return true;\n",
      "    }\n",
      "    return false;\n",
      "}\n",
      "1084\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 114/1243 [03:41<40:27,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6973\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 115/1243 [03:43<40:53,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6974\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 116/1243 [03:46<41:10,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6975\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 117/1243 [03:47<38:07,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7013\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:   9%|         | 118/1243 [03:50<39:41,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7014\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'GenomePosition': 'public class GenomePosition {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nprivate final ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the position on the chromosome\\n */\\nprivate final int pos;;\\n    public GenomePosition(ReferenceDictionary refDict, Strand strand, int chr, int pos);\\n    public GenomePosition(ReferenceDictionary refDict, Strand strand, int chr, int pos, PositionType positionType);\\n    public GenomePosition(GenomePosition other);\\n    public GenomePosition(GenomePosition other, Strand strand);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getPos()\\n    public GenomePosition withStrand(Strand strand)\\n    public boolean isLt(GenomePosition other)\\n    public boolean isLeq(GenomePosition other)\\n    public boolean isGt(GenomePosition other)\\n    public boolean isGeq(GenomePosition other)\\n    public boolean isEq(GenomePosition other)\\n    public int differenceTo(GenomePosition pos)\\n    public int differenceTo(GenomeInterval itv)\\n    public GenomePosition shifted(int delta)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomePosition other)\\n}\\n', 'GenomeInterval': 'public class GenomeInterval {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nfinal ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the begin position on the chromosome\\n */\\nprivate final int beginPos;;\\n    /**\\n * the end position on the chromosome\\n */\\nprivate final int endPos;;\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos);\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos, PositionType positionType);\\n    public GenomeInterval(GenomePosition pos, int length);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getBeginPos()\\n    public int getEndPos()\\n    public GenomeInterval withStrand(Strand strand)\\n    public GenomePosition getGenomeBeginPos()\\n    public GenomePosition getGenomeEndPos()\\n    public int length()\\n    public GenomeInterval intersection(GenomeInterval other)\\n    public GenomeInterval union(GenomeInterval other)\\n    public boolean isLeftOf(GenomePosition pos)\\n    public boolean isRightOf(GenomePosition pos)\\n    public boolean isLeftOfGap(GenomePosition pos)\\n    private GenomePosition ensureSameStrand(GenomePosition pos)\\n    public boolean isRightOfGap(GenomePosition pos)\\n    public boolean contains(GenomePosition pos)\\n    public boolean contains(GenomeInterval other)\\n    public GenomeInterval withMorePadding(int padding)\\n    public GenomeInterval withMorePadding(int paddingUpstream, int paddingDownstream)\\n    public boolean overlapsWith(GenomeInterval other)\\n    private GenomeInterval ensureSameStrand(GenomeInterval other)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomeInterval other)\\n}\\n'}\n",
      "/**\n",
      "\t * @param pos {@link GenomePosition} with to translate into {@link NucleotidePointLocation}\n",
      "\t * @return {@link NucleotidePointLocation}, given the transcript in {@link #transcript}.\n",
      "\t */\n",
      "public NucleotidePointLocation getNucleotidePointLocation(GenomePosition pos)\n",
      "**************************************\n",
      "/**\n",
      " * @param pos {@link GenomePosition} with to translate into {@link NucleotidePointLocation}\n",
      " * @return {@link NucleotidePointLocation}, given the transcript in {@link #transcript}.\n",
      " */\n",
      "public NucleotidePointLocation getNucleotidePointLocation(GenomePosition pos) {\n",
      "    // Guard against cases upstream/downstream of transcription region.\n",
      "    if (// upstream of transcription region\n",
      "    transcript.getTXRegion().isRightOf(pos))\n",
      "        return getCDNANucleotidePointLocationForUpstreamPos(pos);\n",
      "    else if (// downstream of transcription region\n",
      "    transcript.getTXRegion().isLeftOf(pos))\n",
      "        return getCDNANucleotidePointLocationForDownstreamPos(pos);\n",
      "    // The main difference now is between intronic and exonic regions.\n",
      "    if (soDecorator.liesInExon(new GenomeInterval(pos, 0)))\n",
      "        return getCDNANucleotidePointLocationForExonPos(pos);\n",
      "    else\n",
      "        return getCDNANucleotidePointLocationForIntronPos(pos);\n",
      "}\n",
      "1249\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 119/1243 [03:51<36:36,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 120/1243 [03:53<37:10,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7045\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 121/1243 [03:55<38:25,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7083\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 122/1243 [03:58<39:22,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7101\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 123/1243 [04:00<41:55,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7162\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'GenomeVariant': 'public class GenomeVariant {\\n    /**\\n * position of the change\\n */\\nprivate final GenomePosition pos;;\\n    /**\\n * nucleic acid reference string\\n */\\nprivate final String ref;;\\n    /**\\n * nucleic acid alternative string\\n */\\nprivate final String alt;;\\n    public GenomeVariant(GenomePosition pos, String ref, String alt);\\n    public GenomeVariant(GenomePosition pos, String ref, String alt, Strand strand) throws InvalidGenomeVariant;\\n    public static boolean wouldBeSymbolicAllele(String allele)\\n    public String getChrName()\\n    public GenomePosition getGenomePos()\\n    public int getPos()\\n    public String getRef()\\n    public String getAlt()\\n    public int getChr()\\n    public GenomeInterval getGenomeInterval()\\n    public GenomeVariant withStrand(Strand strand)\\n    public String toString()\\n    public GenomeVariantType getType()\\n    public boolean isTransition()\\n    public boolean isTransversion()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(Annotation other)\\n}\\n'}\n",
      "/**\n",
      "\t * Return modified transcript after applying a {@link GenomeVariant}.\n",
      "\t *\n",
      "\t * @param change {@link GenomeVariant} to apply to the transcript\n",
      "\t * @return transcript string with applied {@link GenomeVariant}\n",
      "\t */\n",
      "public String getTranscriptWithChange(GenomeVariant change)\n",
      "**************************************\n",
      "/**\n",
      " * Return modified transcript after applying a {@link GenomeVariant}.\n",
      " *\n",
      " * @param change {@link GenomeVariant} to apply to the transcript\n",
      " * @return transcript string with applied {@link GenomeVariant}\n",
      " */\n",
      "public String getTranscriptWithChange(GenomeVariant change) {\n",
      "    change = change.withStrand(transcript.getStrand());\n",
      "    switch(change.getType()) {\n",
      "        case SNV:\n",
      "        case INSERTION:\n",
      "            return getTranscriptWithPointInRefAffected(change);\n",
      "        case DELETION:\n",
      "        case BLOCK_SUBSTITUTION:\n",
      "            return getTranscriptWithRangeInRefAffected(change);\n",
      "        default:\n",
      "            throw new Error(\"Unhandled change type \" + change.getType());\n",
      "    }\n",
      "}\n",
      "561\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 124/1243 [04:03<43:36,  2.34s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7165\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 125/1243 [04:05<43:49,  2.35s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7194\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 126/1243 [04:07<42:12,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7198\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 127/1243 [04:09<41:18,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7199\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 128/1243 [04:11<40:55,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7216\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'TranscriptPosition': 'public class TranscriptPosition {\\n    /**\\n * the transcript that this position is relative to\\n */\\nprivate final TranscriptModel transcript;;\\n    /**\\n * the position within the transcript\\n */\\nprivate final int pos;;\\n    public TranscriptPosition(TranscriptModel transcript, int pos);\\n    public TranscriptPosition(TranscriptModel transcript, int pos, PositionType positionType);\\n    public TranscriptModel getTranscript()\\n    public int getPos()\\n    public TranscriptPosition shifted(int delta)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n}\\n', 'ProjectionException': 'public class ProjectionException {\\n    public static final long serialVersionUID = 1L;;\\n    public ProjectionException(String msg);\\n}\\n'}\n",
      "/**\n",
      "\t * @return the CDS transcript string\n",
      "\t */\n",
      "public String getCDSTranscript()\n",
      "**************************************\n",
      "public String getCDSTranscript() {\n",
      "    try {\n",
      "        TranscriptPosition tBeginPos = genomeToTranscriptPos(transcript.getCDSRegion().getGenomeBeginPos());\n",
      "        TranscriptPosition tEndPos = genomeToTranscriptPos(transcript.getCDSRegion().getGenomeEndPos());\n",
      "        return transcript.getTrimmedSequence().substring(tBeginPos.getPos(), tEndPos.getPos());\n",
      "    } catch (ProjectionException e) {\n",
      "        throw new Error(\"Bug: CDS begin/end must be translatable into transcript positions\");\n",
      "    }\n",
      "}\n",
      "412\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 129/1243 [04:13<37:02,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7260\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  10%|         | 130/1243 [04:15<37:19,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7396\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 131/1243 [04:17<35:35,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7423\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 132/1243 [04:19<36:52,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7438\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 133/1243 [04:21<37:41,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7457\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class IntervalArray {\\n    /**\\n * version number to use when serializing\\n */\\nprivate static final long serialVersionUID = 1L;;\\n    /**\\n * list of {@link Interval} objects, sorted by begin position\\n */\\nprivate final ImmutableList<Interval<T>> intervals;;\\n    /**\\n * list of {@link Interval} objects, sorted by end position\\n */\\nprivate final ImmutableList<Interval<T>> intervalsEnd;;\\n    public IntervalArray(Collection<T> elements, IntervalEndExtractor<T> extractor);\\n    public ImmutableList<Interval<T>> getIntervals()\\n    public ImmutableList<Interval<T>> getIntervalsEnd()\\n    public int size()\\n    public QueryResult findOverlappingWithPoint(int point)\\n    private T findRightNeighbor(int point)\\n    private T findLeftNeighbor(int point)\\n    private void findOverlappingWithPoint(int begin, int end, int center, int point, QueryResultBuilder result)\\n    public QueryResult findOverlappingWithInterval(int begin, int end)\\n    private void findOverlappingWithInterval(int begin, int end, int center, int iBegin, int iEnd, QueryResultBuilder result)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/impl/intervals/IntervalArray.java', 'methodCount': 10, 'lineCount': 15}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class TranscriptIntervalEndExtractor {\\n    public int getBegin(TranscriptModel transcript)\\n    public int getEnd(TranscriptModel transcript)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptIntervalEndExtractor.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class TranscriptModel {\\n    /**\\n * Accession number of the transcript (e.g., the UCSC knownGene id - uc011nca.2). The version number may be\\n * included.\\n */\\nprivate final String accession;;\\n    /**\\n * Gene symbol of the known Gene. Can be null for some genes. Note that in annovar, $name2 corresponds to the\\n * geneSymbol if available, otherwise the kgID is used.\\n */\\nprivate final String geneSymbol;;\\n    /**\\n * Genomic interval with transcript begin/end.\\n */\\nprivate final GenomeInterval txRegion;;\\n    /**\\n * Genomic interval with CDS begin/end.\\n *\\n * <b>Note</b> that in Jannovar, the CDS region includes the start and stop codon.\\n */\\nprivate final GenomeInterval cdsRegion;;\\n    /**\\n * Genomic intervals with the exons, order is dictated by strand of transcript.\\n */\\nprivate final ImmutableList<GenomeInterval> exonRegions;;\\n    /**\\n * cDNA sequence of the spliced RNA of this known gene transcript.\\n */\\nprivate final String sequence;;\\n    /**\\n * The gene ID, from Ensembl (<code>\"ENS[MUS]*G0+([0-9]+)\"</code>), Entrez (\"<code>ENTREZ([0-9]+)</code>\\n * \"), RefSeq (\"<code>gene([0-9]+)</code>\").\\n *\\n * <code>null</code> for no available gene ID.\\n */\\nprivate final String geneID;;\\n    /**\\n * Alternative gene IDs, as parsed from RefSeq GFF3 file\\n * <p>\\n * See {@link #getAltGeneIDs()} for more information\\n */\\nprivate final ImmutableSortedMap<String, String> altGeneIDs;;\\n    /**\\n * The transcript support level of the this transcript (the lower the better).\\n *\\n * @see TranscriptSupportLevels\\n * @see <a href=\"http://www.ensembl.org/Help/Glossary?id=492\">http://www.ensembl.org/Help/Glossary?id=492</a>\\n */\\nprivate final int transcriptSupportLevel;;\\n    /**\\n * The alignment of the transcript sequence to the genomic exon region.\\n */\\nprivate final Alignment seqAlignment;;\\n    /**\\n * Whether or not the transcript aligns with mismatches to the reference.\\n */\\nprivate final boolean hasSubstitutions;;\\n    /**\\n * Whether or not the transcript aligns with indels to the reference.\\n */\\nprivate final boolean hasIndels;;\\n    /**\\n * Class version (for serialization).\\n */\\nprivate static final long serialVersionUID = 4L;;\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasIndels, boolean hasSubstitutions);\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasSubstitutions, boolean hasIndels, Map<String, String> altGeneIDs, Alignment seqAlignment);\\n    public String getAccession()\\n    public String getGeneSymbol()\\n    public GenomeInterval getTXRegion()\\n    public GenomeInterval getCDSRegion()\\n    public ImmutableList<GenomeInterval> getExonRegions()\\n    public String getSequence()\\n    public String getTrimmedSequence()\\n    public Alignment getSeqAlignment()\\n    public String getGeneID()\\n    public ImmutableSortedMap<String, String> getAltGeneIDs()\\n    public int getTranscriptSupportLevel()\\n    public Strand getStrand()\\n    public int getChr()\\n    public boolean isCoding()\\n    public boolean isMitochondrial()\\n    public int cdsTranscriptLength()\\n    public int transcriptLength()\\n    public boolean isHasSubstitutions()\\n    public boolean isHasIndels()\\n    public GenomeInterval intronRegion(int i)\\n    private void checkForConsistency()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(TranscriptModel o)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptModel.java', 'methodCount': 27, 'lineCount': 42}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "\t * @param transcriptModels set of {@link TranscriptModel}s to build multi-mapping for\n",
      "\t * @return multi-mapping from gene symbol to {@link TranscriptModel}\n",
      "\t */\n",
      "private static ImmutableMultimap<String, TranscriptModel> makeTMByGeneSymbol(ImmutableList<TranscriptModel> transcriptModels)\n",
      "**************************************\n",
      "private static ImmutableMultimap<String, TranscriptModel> makeTMByGeneSymbol(ImmutableList<TranscriptModel> transcriptModels) {\n",
      "    ImmutableMultimap.Builder<String, TranscriptModel> builder = new ImmutableMultimap.Builder<String, TranscriptModel>();\n",
      "    for (TranscriptModel tm : transcriptModels) {\n",
      "        if (tm.getGeneSymbol() != null) {\n",
      "            builder.put(tm.getGeneSymbol(), tm);\n",
      "        }\n",
      "    }\n",
      "    return builder.build();\n",
      "}\n",
      "1735\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 134/1243 [04:23<35:07,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7463\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 135/1243 [04:25<35:42,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7495\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 136/1243 [04:28<41:49,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7496\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 137/1243 [04:30<41:51,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7533\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 138/1243 [04:32<41:02,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7534\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SVGenomeVariant {\\n    /**\\n * Genomic start/first position.\\n */\\nprotected final GenomePosition genomePos;;\\n    /**\\n * Genomic end/second position.\\n */\\nprotected final GenomePosition genomePos2;;\\n    /**\\n * Lower bound of confidence interval around {@link #genomePos}.\\n */\\nprotected final int posCILowerBound;;\\n    /**\\n * Upper bound of confidence interval around {@link #genomePos}.\\n */\\nprotected final int posCIUpperBound;;\\n    /**\\n * Lower bound of confidence interval around {@link #genomePos2}.\\n */\\nprotected final int pos2CILowerBound;;\\n    /**\\n * Upper bound of confidence interval around {@link #genomePos2}.\\n */\\nprotected final int pos2CIUpperBound;;\\n    public SVGenomeVariant(GenomePosition genomePos, GenomePosition genomePos2, int posCILowerBound, int posCIUpperBound, int pos2CILowerBound, int pos2CIUpperBound);\\n    public abstract SVGenomeVariant withStrand(Strand strand)\\n    public GenomeInterval getGenomeInterval()\\n    public GenomePosition getGenomePos()\\n    public GenomePosition getGenomePos2()\\n    public int getPosCILowerBound()\\n    public int getPosCIUpperBound()\\n    public int getPos2CILowerBound()\\n    public int getPos2CIUpperBound()\\n    public String getChrName()\\n    public int getChr()\\n    public int getPos()\\n    public String getChr2Name()\\n    public int getChr2()\\n    public int getPos2()\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/SVGenomeVariant.java', 'methodCount': 15, 'lineCount': 23}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class TranscriptModel {\\n    /**\\n * Accession number of the transcript (e.g., the UCSC knownGene id - uc011nca.2). The version number may be\\n * included.\\n */\\nprivate final String accession;;\\n    /**\\n * Gene symbol of the known Gene. Can be null for some genes. Note that in annovar, $name2 corresponds to the\\n * geneSymbol if available, otherwise the kgID is used.\\n */\\nprivate final String geneSymbol;;\\n    /**\\n * Genomic interval with transcript begin/end.\\n */\\nprivate final GenomeInterval txRegion;;\\n    /**\\n * Genomic interval with CDS begin/end.\\n *\\n * <b>Note</b> that in Jannovar, the CDS region includes the start and stop codon.\\n */\\nprivate final GenomeInterval cdsRegion;;\\n    /**\\n * Genomic intervals with the exons, order is dictated by strand of transcript.\\n */\\nprivate final ImmutableList<GenomeInterval> exonRegions;;\\n    /**\\n * cDNA sequence of the spliced RNA of this known gene transcript.\\n */\\nprivate final String sequence;;\\n    /**\\n * The gene ID, from Ensembl (<code>\"ENS[MUS]*G0+([0-9]+)\"</code>), Entrez (\"<code>ENTREZ([0-9]+)</code>\\n * \"), RefSeq (\"<code>gene([0-9]+)</code>\").\\n *\\n * <code>null</code> for no available gene ID.\\n */\\nprivate final String geneID;;\\n    /**\\n * Alternative gene IDs, as parsed from RefSeq GFF3 file\\n * <p>\\n * See {@link #getAltGeneIDs()} for more information\\n */\\nprivate final ImmutableSortedMap<String, String> altGeneIDs;;\\n    /**\\n * The transcript support level of the this transcript (the lower the better).\\n *\\n * @see TranscriptSupportLevels\\n * @see <a href=\"http://www.ensembl.org/Help/Glossary?id=492\">http://www.ensembl.org/Help/Glossary?id=492</a>\\n */\\nprivate final int transcriptSupportLevel;;\\n    /**\\n * The alignment of the transcript sequence to the genomic exon region.\\n */\\nprivate final Alignment seqAlignment;;\\n    /**\\n * Whether or not the transcript aligns with mismatches to the reference.\\n */\\nprivate final boolean hasSubstitutions;;\\n    /**\\n * Whether or not the transcript aligns with indels to the reference.\\n */\\nprivate final boolean hasIndels;;\\n    /**\\n * Class version (for serialization).\\n */\\nprivate static final long serialVersionUID = 4L;;\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasIndels, boolean hasSubstitutions);\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasSubstitutions, boolean hasIndels, Map<String, String> altGeneIDs, Alignment seqAlignment);\\n    public String getAccession()\\n    public String getGeneSymbol()\\n    public GenomeInterval getTXRegion()\\n    public GenomeInterval getCDSRegion()\\n    public ImmutableList<GenomeInterval> getExonRegions()\\n    public String getSequence()\\n    public String getTrimmedSequence()\\n    public Alignment getSeqAlignment()\\n    public String getGeneID()\\n    public ImmutableSortedMap<String, String> getAltGeneIDs()\\n    public int getTranscriptSupportLevel()\\n    public Strand getStrand()\\n    public int getChr()\\n    public boolean isCoding()\\n    public boolean isMitochondrial()\\n    public int cdsTranscriptLength()\\n    public int transcriptLength()\\n    public boolean isHasSubstitutions()\\n    public boolean isHasIndels()\\n    public GenomeInterval intronRegion(int i)\\n    private void checkForConsistency()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(TranscriptModel o)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptModel.java', 'methodCount': 27, 'lineCount': 42}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{'VCFSVAnnotationData': 'public class VCFSVAnnotationData {\\n    /**\\n * predicted effects\\n */\\npublic ImmutableSet<VariantEffect> effects = ImmutableSortedSet.<VariantEffect>of();;\\n    /**\\n * predicted impact\\n */\\npublic PutativeImpact impact = null;;\\n    /**\\n * symbol of affected gene\\n */\\npublic String geneSymbol = null;;\\n    /**\\n * ID of affected gene\\n */\\npublic String geneID = null;;\\n    /**\\n * type of the feature (<code>null</code> or <code>\"transcript\"</code>).\\n */\\npublic String featureType = null;;\\n    /**\\n * ID of the feature/transcript\\n */\\npublic String featureID = null;;\\n    /**\\n * bio type of the feature, one of \"Coding\" and \"Noncoding\".\\n */\\npublic String featureBioType = null;;\\n    /**\\n * whether or not the transcript is coding\\n */\\npublic boolean isCoding = false;;\\n    /**\\n * additional messages for the annotation\\n */\\npublic ImmutableSet<AnnotationMessage> messages = ImmutableSortedSet.<AnnotationMessage>of();;\\n    public void setTranscriptAndVariant(TranscriptModel tm, SVGenomeVariant variant)\\n    public Object[] toArray()\\n    public String toUnescapedString()\\n    private String escape(String str)\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "\t * Return string for {@code SVANNO} entry of the {@code INFO} VCF field.\n",
      "\t */\n",
      "public String toVCFSVAnnoString(boolean escape)\n",
      "**************************************\n",
      "public String toVCFSVAnnoString(boolean escape) {\n",
      "    VCFSVAnnotationData data = new VCFSVAnnotationData();\n",
      "    data.effects = effects;\n",
      "    data.impact = getPutativeImpact();\n",
      "    data.setTranscriptAndVariant(transcript, variant);\n",
      "    data.isCoding = (transcript == null) ? false : transcript.isCoding();\n",
      "    data.messages = messages;\n",
      "    if (escape)\n",
      "        return data.toString();\n",
      "    else\n",
      "        return data.toUnescapedString();\n",
      "}\n",
      "2055\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|         | 139/1243 [04:35<42:08,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7551\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|        | 140/1243 [04:37<41:03,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7552\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|        | 141/1243 [04:39<40:13,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7590\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  11%|        | 142/1243 [04:41<42:28,  2.32s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7620\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 143/1243 [04:43<41:41,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7622\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SVAnnotation {\\n    /**\\n * The DESCRIPTION string to use in the VCF header for VCFVariantAnnotation objects\\n */\\npublic static final String VCF_SVANN_DESCRIPTION_STRING = \"Functional SV Annotation:\\'Annotation|\" + \"Annotation_Impact|Gene_Name|Gene_ID|Feature_Type|Feature_ID|Transcript_BioType|ERRORS / WARNINGS / INFO\\'\";;\\n    private static final ImmutableSet<AnnotationMessage> EMPTY_ANNOTATION_MESSAGES = Sets.immutableEnumSet(EnumSet.noneOf(AnnotationMessage.class));;\\n    /**\\n * Annotated {@link SVGenomeVariant}.\\n */\\nprivate final SVGenomeVariant variant;;\\n    /**\\n * Annotate transcript.\\n */\\nprivate final TranscriptModel transcript;;\\n    /**\\n * The effects the variant is annotated with for the transcript.\\n */\\nprivate final ImmutableSet<VariantEffect> effects;;\\n    /**\\n * errors and warnings\\n */\\nprivate final ImmutableSet<AnnotationMessage> messages;;\\n    public SVAnnotation(Collection<AnnotationMessage> messages);\\n    public SVAnnotation(SVGenomeVariant variant, TranscriptModel transcriptModel, Collection<VariantEffect> effects);\\n    public SVGenomeVariant getVariant()\\n    public TranscriptModel getTranscript()\\n    public ImmutableSet<VariantEffect> getEffects()\\n    public PutativeImpact getPutativeImpact()\\n    public String toVCFSVAnnoString(boolean escape)\\n    public String toString()\\n    public VariantEffect getMostPathogenicVariantEffect()\\n    public int compareTo(SVAnnotation other)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/SVAnnotation.java', 'methodCount': 10, 'lineCount': 18}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/VariantEffect.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class GenomeInterval {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nfinal ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the begin position on the chromosome\\n */\\nprivate final int beginPos;;\\n    /**\\n * the end position on the chromosome\\n */\\nprivate final int endPos;;\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos);\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos, PositionType positionType);\\n    public GenomeInterval(GenomePosition pos, int length);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getBeginPos()\\n    public int getEndPos()\\n    public GenomeInterval withStrand(Strand strand)\\n    public GenomePosition getGenomeBeginPos()\\n    public GenomePosition getGenomeEndPos()\\n    public int length()\\n    public GenomeInterval intersection(GenomeInterval other)\\n    public GenomeInterval union(GenomeInterval other)\\n    public boolean isLeftOf(GenomePosition pos)\\n    public boolean isRightOf(GenomePosition pos)\\n    public boolean isLeftOfGap(GenomePosition pos)\\n    private GenomePosition ensureSameStrand(GenomePosition pos)\\n    public boolean isRightOfGap(GenomePosition pos)\\n    public boolean contains(GenomePosition pos)\\n    public boolean contains(GenomeInterval other)\\n    public GenomeInterval withMorePadding(int padding)\\n    public GenomeInterval withMorePadding(int paddingUpstream, int paddingDownstream)\\n    public boolean overlapsWith(GenomeInterval other)\\n    private GenomeInterval ensureSameStrand(GenomeInterval other)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomeInterval other)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/GenomeInterval.java', 'methodCount': 29, 'lineCount': 37}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class SVDuplication {\\n    public SVDuplication(GenomePosition genomePos, GenomePosition genomePos2, int posCILowerBound, int posCIUpperBound, int pos2CILowerBound, int pos2CIUpperBound);\\n    public SVDescription.Type getType()\\n    public SVDuplication withStrand(Strand strand)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/SVDuplication.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class TranscriptModel {\\n    /**\\n * Accession number of the transcript (e.g., the UCSC knownGene id - uc011nca.2). The version number may be\\n * included.\\n */\\nprivate final String accession;;\\n    /**\\n * Gene symbol of the known Gene. Can be null for some genes. Note that in annovar, $name2 corresponds to the\\n * geneSymbol if available, otherwise the kgID is used.\\n */\\nprivate final String geneSymbol;;\\n    /**\\n * Genomic interval with transcript begin/end.\\n */\\nprivate final GenomeInterval txRegion;;\\n    /**\\n * Genomic interval with CDS begin/end.\\n *\\n * <b>Note</b> that in Jannovar, the CDS region includes the start and stop codon.\\n */\\nprivate final GenomeInterval cdsRegion;;\\n    /**\\n * Genomic intervals with the exons, order is dictated by strand of transcript.\\n */\\nprivate final ImmutableList<GenomeInterval> exonRegions;;\\n    /**\\n * cDNA sequence of the spliced RNA of this known gene transcript.\\n */\\nprivate final String sequence;;\\n    /**\\n * The gene ID, from Ensembl (<code>\"ENS[MUS]*G0+([0-9]+)\"</code>), Entrez (\"<code>ENTREZ([0-9]+)</code>\\n * \"), RefSeq (\"<code>gene([0-9]+)</code>\").\\n *\\n * <code>null</code> for no available gene ID.\\n */\\nprivate final String geneID;;\\n    /**\\n * Alternative gene IDs, as parsed from RefSeq GFF3 file\\n * <p>\\n * See {@link #getAltGeneIDs()} for more information\\n */\\nprivate final ImmutableSortedMap<String, String> altGeneIDs;;\\n    /**\\n * The transcript support level of the this transcript (the lower the better).\\n *\\n * @see TranscriptSupportLevels\\n * @see <a href=\"http://www.ensembl.org/Help/Glossary?id=492\">http://www.ensembl.org/Help/Glossary?id=492</a>\\n */\\nprivate final int transcriptSupportLevel;;\\n    /**\\n * The alignment of the transcript sequence to the genomic exon region.\\n */\\nprivate final Alignment seqAlignment;;\\n    /**\\n * Whether or not the transcript aligns with mismatches to the reference.\\n */\\nprivate final boolean hasSubstitutions;;\\n    /**\\n * Whether or not the transcript aligns with indels to the reference.\\n */\\nprivate final boolean hasIndels;;\\n    /**\\n * Class version (for serialization).\\n */\\nprivate static final long serialVersionUID = 4L;;\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasIndels, boolean hasSubstitutions);\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasSubstitutions, boolean hasIndels, Map<String, String> altGeneIDs, Alignment seqAlignment);\\n    public String getAccession()\\n    public String getGeneSymbol()\\n    public GenomeInterval getTXRegion()\\n    public GenomeInterval getCDSRegion()\\n    public ImmutableList<GenomeInterval> getExonRegions()\\n    public String getSequence()\\n    public String getTrimmedSequence()\\n    public Alignment getSeqAlignment()\\n    public String getGeneID()\\n    public ImmutableSortedMap<String, String> getAltGeneIDs()\\n    public int getTranscriptSupportLevel()\\n    public Strand getStrand()\\n    public int getChr()\\n    public boolean isCoding()\\n    public boolean isMitochondrial()\\n    public int cdsTranscriptLength()\\n    public int transcriptLength()\\n    public boolean isHasSubstitutions()\\n    public boolean isHasIndels()\\n    public GenomeInterval intronRegion(int i)\\n    private void checkForConsistency()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(TranscriptModel o)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptModel.java', 'methodCount': 27, 'lineCount': 42}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "\t * Return augmented {@link effects} based on the transcript alone (coding/non-coding transcript).\n",
      "\t */\n",
      "private ImmutableSet<VariantEffect> buildEffectSet(Collection<VariantEffect> effects)\n",
      "**************************************\n",
      "private ImmutableSet<VariantEffect> buildEffectSet(Collection<VariantEffect> effects) {\n",
      "    final EnumSet<VariantEffect> tmpEffects;\n",
      "    if (transcript.isCoding()) {\n",
      "        tmpEffects = EnumSet.of(VariantEffect.CODING_TRANSCRIPT_VARIANT);\n",
      "    } else {\n",
      "        tmpEffects = EnumSet.of(VariantEffect.NON_CODING_TRANSCRIPT_VARIANT);\n",
      "    }\n",
      "    tmpEffects.addAll(effects);\n",
      "    return Sets.immutableEnumSet(tmpEffects);\n",
      "}\n",
      "2497\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 144/1243 [04:46<40:41,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7629\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 145/1243 [04:48<40:22,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7631\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 146/1243 [04:50<40:23,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7641\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 147/1243 [04:52<39:41,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7642\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 148/1243 [04:54<38:59,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7644\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SVAnnotation {\\n    /**\\n * The DESCRIPTION string to use in the VCF header for VCFVariantAnnotation objects\\n */\\npublic static final String VCF_SVANN_DESCRIPTION_STRING = \"Functional SV Annotation:\\'Annotation|\" + \"Annotation_Impact|Gene_Name|Gene_ID|Feature_Type|Feature_ID|Transcript_BioType|ERRORS / WARNINGS / INFO\\'\";;\\n    private static final ImmutableSet<AnnotationMessage> EMPTY_ANNOTATION_MESSAGES = Sets.immutableEnumSet(EnumSet.noneOf(AnnotationMessage.class));;\\n    /**\\n * Annotated {@link SVGenomeVariant}.\\n */\\nprivate final SVGenomeVariant variant;;\\n    /**\\n * Annotate transcript.\\n */\\nprivate final TranscriptModel transcript;;\\n    /**\\n * The effects the variant is annotated with for the transcript.\\n */\\nprivate final ImmutableSet<VariantEffect> effects;;\\n    /**\\n * errors and warnings\\n */\\nprivate final ImmutableSet<AnnotationMessage> messages;;\\n    public SVAnnotation(Collection<AnnotationMessage> messages);\\n    public SVAnnotation(SVGenomeVariant variant, TranscriptModel transcriptModel, Collection<VariantEffect> effects);\\n    public SVGenomeVariant getVariant()\\n    public TranscriptModel getTranscript()\\n    public ImmutableSet<VariantEffect> getEffects()\\n    public PutativeImpact getPutativeImpact()\\n    public String toVCFSVAnnoString(boolean escape)\\n    public String toString()\\n    public VariantEffect getMostPathogenicVariantEffect()\\n    public int compareTo(SVAnnotation other)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/SVAnnotation.java', 'methodCount': 10, 'lineCount': 18}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/VariantEffect.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class GenomeInterval {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nfinal ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the begin position on the chromosome\\n */\\nprivate final int beginPos;;\\n    /**\\n * the end position on the chromosome\\n */\\nprivate final int endPos;;\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos);\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos, PositionType positionType);\\n    public GenomeInterval(GenomePosition pos, int length);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getBeginPos()\\n    public int getEndPos()\\n    public GenomeInterval withStrand(Strand strand)\\n    public GenomePosition getGenomeBeginPos()\\n    public GenomePosition getGenomeEndPos()\\n    public int length()\\n    public GenomeInterval intersection(GenomeInterval other)\\n    public GenomeInterval union(GenomeInterval other)\\n    public boolean isLeftOf(GenomePosition pos)\\n    public boolean isRightOf(GenomePosition pos)\\n    public boolean isLeftOfGap(GenomePosition pos)\\n    private GenomePosition ensureSameStrand(GenomePosition pos)\\n    public boolean isRightOfGap(GenomePosition pos)\\n    public boolean contains(GenomePosition pos)\\n    public boolean contains(GenomeInterval other)\\n    public GenomeInterval withMorePadding(int padding)\\n    public GenomeInterval withMorePadding(int paddingUpstream, int paddingDownstream)\\n    public boolean overlapsWith(GenomeInterval other)\\n    private GenomeInterval ensureSameStrand(GenomeInterval other)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomeInterval other)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/GenomeInterval.java', 'methodCount': 29, 'lineCount': 37}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SVBreakend {\\n    /**\\n * Bases located in the reference at first/start pos.\\n */\\nprivate final String posRefBases;;\\n    /**\\n * Bases located in the reference at secon/end pos.\\n */\\nprivate final String pos2RefBases;;\\n    /**\\n * The side of the breakend.\\n */\\nprivate final Side side;;\\n    public SVBreakend(GenomePosition genomePos, GenomePosition genomePos2, int posCILowerBound, int posCIUpperBound, int pos2CILowerBound, int pos2CIUpperBound, String posRefBases, String pos2RefBases, Side side);\\n    public SVDescription.Type getType()\\n    public SVBreakend withStrand(Strand strand)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/SVBreakend.java', 'methodCount': 4, 'lineCount': 9}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class TranscriptModel {\\n    /**\\n * Accession number of the transcript (e.g., the UCSC knownGene id - uc011nca.2). The version number may be\\n * included.\\n */\\nprivate final String accession;;\\n    /**\\n * Gene symbol of the known Gene. Can be null for some genes. Note that in annovar, $name2 corresponds to the\\n * geneSymbol if available, otherwise the kgID is used.\\n */\\nprivate final String geneSymbol;;\\n    /**\\n * Genomic interval with transcript begin/end.\\n */\\nprivate final GenomeInterval txRegion;;\\n    /**\\n * Genomic interval with CDS begin/end.\\n *\\n * <b>Note</b> that in Jannovar, the CDS region includes the start and stop codon.\\n */\\nprivate final GenomeInterval cdsRegion;;\\n    /**\\n * Genomic intervals with the exons, order is dictated by strand of transcript.\\n */\\nprivate final ImmutableList<GenomeInterval> exonRegions;;\\n    /**\\n * cDNA sequence of the spliced RNA of this known gene transcript.\\n */\\nprivate final String sequence;;\\n    /**\\n * The gene ID, from Ensembl (<code>\"ENS[MUS]*G0+([0-9]+)\"</code>), Entrez (\"<code>ENTREZ([0-9]+)</code>\\n * \"), RefSeq (\"<code>gene([0-9]+)</code>\").\\n *\\n * <code>null</code> for no available gene ID.\\n */\\nprivate final String geneID;;\\n    /**\\n * Alternative gene IDs, as parsed from RefSeq GFF3 file\\n * <p>\\n * See {@link #getAltGeneIDs()} for more information\\n */\\nprivate final ImmutableSortedMap<String, String> altGeneIDs;;\\n    /**\\n * The transcript support level of the this transcript (the lower the better).\\n *\\n * @see TranscriptSupportLevels\\n * @see <a href=\"http://www.ensembl.org/Help/Glossary?id=492\">http://www.ensembl.org/Help/Glossary?id=492</a>\\n */\\nprivate final int transcriptSupportLevel;;\\n    /**\\n * The alignment of the transcript sequence to the genomic exon region.\\n */\\nprivate final Alignment seqAlignment;;\\n    /**\\n * Whether or not the transcript aligns with mismatches to the reference.\\n */\\nprivate final boolean hasSubstitutions;;\\n    /**\\n * Whether or not the transcript aligns with indels to the reference.\\n */\\nprivate final boolean hasIndels;;\\n    /**\\n * Class version (for serialization).\\n */\\nprivate static final long serialVersionUID = 4L;;\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasIndels, boolean hasSubstitutions);\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasSubstitutions, boolean hasIndels, Map<String, String> altGeneIDs, Alignment seqAlignment);\\n    public String getAccession()\\n    public String getGeneSymbol()\\n    public GenomeInterval getTXRegion()\\n    public GenomeInterval getCDSRegion()\\n    public ImmutableList<GenomeInterval> getExonRegions()\\n    public String getSequence()\\n    public String getTrimmedSequence()\\n    public Alignment getSeqAlignment()\\n    public String getGeneID()\\n    public ImmutableSortedMap<String, String> getAltGeneIDs()\\n    public int getTranscriptSupportLevel()\\n    public Strand getStrand()\\n    public int getChr()\\n    public boolean isCoding()\\n    public boolean isMitochondrial()\\n    public int cdsTranscriptLength()\\n    public int transcriptLength()\\n    public boolean isHasSubstitutions()\\n    public boolean isHasIndels()\\n    public GenomeInterval intronRegion(int i)\\n    private void checkForConsistency()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(TranscriptModel o)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptModel.java', 'methodCount': 27, 'lineCount': 42}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "\t * Return augmented {@code effects} based on the transcript alone (coding/non-coding transcript).\n",
      "\t */\n",
      "private ImmutableSet<VariantEffect> buildEffectSet(Collection<VariantEffect> effects)\n",
      "**************************************\n",
      "private ImmutableSet<VariantEffect> buildEffectSet(Collection<VariantEffect> effects) {\n",
      "    final EnumSet<VariantEffect> tmpEffects;\n",
      "    if (transcript.isCoding()) {\n",
      "        tmpEffects = EnumSet.of(VariantEffect.CODING_TRANSCRIPT_VARIANT);\n",
      "    } else {\n",
      "        tmpEffects = EnumSet.of(VariantEffect.NON_CODING_TRANSCRIPT_VARIANT);\n",
      "    }\n",
      "    tmpEffects.addAll(effects);\n",
      "    tmpEffects.add(VariantEffect.TRANSLOCATION);\n",
      "    tmpEffects.add(VariantEffect.STRUCTURAL_VARIANT);\n",
      "    return Sets.immutableEnumSet(tmpEffects);\n",
      "}\n",
      "2619\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 149/1243 [04:56<38:29,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7649\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 150/1243 [04:58<38:34,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7657\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 151/1243 [05:00<38:11,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7659\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 152/1243 [05:03<39:20,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7660\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 153/1243 [05:05<39:13,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7665\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SVAnnotation {\\n    /**\\n * The DESCRIPTION string to use in the VCF header for VCFVariantAnnotation objects\\n */\\npublic static final String VCF_SVANN_DESCRIPTION_STRING = \"Functional SV Annotation:\\'Annotation|\" + \"Annotation_Impact|Gene_Name|Gene_ID|Feature_Type|Feature_ID|Transcript_BioType|ERRORS / WARNINGS / INFO\\'\";;\\n    private static final ImmutableSet<AnnotationMessage> EMPTY_ANNOTATION_MESSAGES = Sets.immutableEnumSet(EnumSet.noneOf(AnnotationMessage.class));;\\n    /**\\n * Annotated {@link SVGenomeVariant}.\\n */\\nprivate final SVGenomeVariant variant;;\\n    /**\\n * Annotate transcript.\\n */\\nprivate final TranscriptModel transcript;;\\n    /**\\n * The effects the variant is annotated with for the transcript.\\n */\\nprivate final ImmutableSet<VariantEffect> effects;;\\n    /**\\n * errors and warnings\\n */\\nprivate final ImmutableSet<AnnotationMessage> messages;;\\n    public SVAnnotation(Collection<AnnotationMessage> messages);\\n    public SVAnnotation(SVGenomeVariant variant, TranscriptModel transcriptModel, Collection<VariantEffect> effects);\\n    public SVGenomeVariant getVariant()\\n    public TranscriptModel getTranscript()\\n    public ImmutableSet<VariantEffect> getEffects()\\n    public PutativeImpact getPutativeImpact()\\n    public String toVCFSVAnnoString(boolean escape)\\n    public String toString()\\n    public VariantEffect getMostPathogenicVariantEffect()\\n    public int compareTo(SVAnnotation other)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/SVAnnotation.java', 'methodCount': 10, 'lineCount': 18}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/annotation/VariantEffect.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class GenomeInterval {\\n    private static final long serialVersionUID = 2L;;\\n    /**\\n * reference dictionary to use for coordinate translation\\n */\\nfinal ReferenceDictionary refDict;;\\n    /**\\n * the strand that the position is located on\\n */\\nprivate final Strand strand;;\\n    /**\\n * the chromosome number, as index in chromosome dictionary\\n */\\nprivate final int chr;;\\n    /**\\n * the begin position on the chromosome\\n */\\nprivate final int beginPos;;\\n    /**\\n * the end position on the chromosome\\n */\\nprivate final int endPos;;\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos);\\n    public GenomeInterval(ReferenceDictionary refDict, Strand strand, int chr, int beginPos, int endPos, PositionType positionType);\\n    public GenomeInterval(GenomePosition pos, int length);\\n    public ReferenceDictionary getRefDict()\\n    public Strand getStrand()\\n    public int getChr()\\n    public int getBeginPos()\\n    public int getEndPos()\\n    public GenomeInterval withStrand(Strand strand)\\n    public GenomePosition getGenomeBeginPos()\\n    public GenomePosition getGenomeEndPos()\\n    public int length()\\n    public GenomeInterval intersection(GenomeInterval other)\\n    public GenomeInterval union(GenomeInterval other)\\n    public boolean isLeftOf(GenomePosition pos)\\n    public boolean isRightOf(GenomePosition pos)\\n    public boolean isLeftOfGap(GenomePosition pos)\\n    private GenomePosition ensureSameStrand(GenomePosition pos)\\n    public boolean isRightOfGap(GenomePosition pos)\\n    public boolean contains(GenomePosition pos)\\n    public boolean contains(GenomeInterval other)\\n    public GenomeInterval withMorePadding(int padding)\\n    public GenomeInterval withMorePadding(int paddingUpstream, int paddingDownstream)\\n    public boolean overlapsWith(GenomeInterval other)\\n    private GenomeInterval ensureSameStrand(GenomeInterval other)\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(GenomeInterval other)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/GenomeInterval.java', 'methodCount': 29, 'lineCount': 37}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class SVInversion {\\n    public SVInversion(GenomePosition genomePos, GenomePosition genomePos2, int posCILowerBound, int posCIUpperBound, int pos2CILowerBound, int pos2CIUpperBound);\\n    public SVDescription.Type getType()\\n    public SVInversion withStrand(Strand strand)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/SVInversion.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class TranscriptModel {\\n    /**\\n * Accession number of the transcript (e.g., the UCSC knownGene id - uc011nca.2). The version number may be\\n * included.\\n */\\nprivate final String accession;;\\n    /**\\n * Gene symbol of the known Gene. Can be null for some genes. Note that in annovar, $name2 corresponds to the\\n * geneSymbol if available, otherwise the kgID is used.\\n */\\nprivate final String geneSymbol;;\\n    /**\\n * Genomic interval with transcript begin/end.\\n */\\nprivate final GenomeInterval txRegion;;\\n    /**\\n * Genomic interval with CDS begin/end.\\n *\\n * <b>Note</b> that in Jannovar, the CDS region includes the start and stop codon.\\n */\\nprivate final GenomeInterval cdsRegion;;\\n    /**\\n * Genomic intervals with the exons, order is dictated by strand of transcript.\\n */\\nprivate final ImmutableList<GenomeInterval> exonRegions;;\\n    /**\\n * cDNA sequence of the spliced RNA of this known gene transcript.\\n */\\nprivate final String sequence;;\\n    /**\\n * The gene ID, from Ensembl (<code>\"ENS[MUS]*G0+([0-9]+)\"</code>), Entrez (\"<code>ENTREZ([0-9]+)</code>\\n * \"), RefSeq (\"<code>gene([0-9]+)</code>\").\\n *\\n * <code>null</code> for no available gene ID.\\n */\\nprivate final String geneID;;\\n    /**\\n * Alternative gene IDs, as parsed from RefSeq GFF3 file\\n * <p>\\n * See {@link #getAltGeneIDs()} for more information\\n */\\nprivate final ImmutableSortedMap<String, String> altGeneIDs;;\\n    /**\\n * The transcript support level of the this transcript (the lower the better).\\n *\\n * @see TranscriptSupportLevels\\n * @see <a href=\"http://www.ensembl.org/Help/Glossary?id=492\">http://www.ensembl.org/Help/Glossary?id=492</a>\\n */\\nprivate final int transcriptSupportLevel;;\\n    /**\\n * The alignment of the transcript sequence to the genomic exon region.\\n */\\nprivate final Alignment seqAlignment;;\\n    /**\\n * Whether or not the transcript aligns with mismatches to the reference.\\n */\\nprivate final boolean hasSubstitutions;;\\n    /**\\n * Whether or not the transcript aligns with indels to the reference.\\n */\\nprivate final boolean hasIndels;;\\n    /**\\n * Class version (for serialization).\\n */\\nprivate static final long serialVersionUID = 4L;;\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasIndels, boolean hasSubstitutions);\\n    public TranscriptModel(String accession, String geneSymbol, GenomeInterval txRegion, GenomeInterval cdsRegion, ImmutableList<GenomeInterval> exonRegions, String sequence, String geneID, int transcriptSupportLevel, boolean hasSubstitutions, boolean hasIndels, Map<String, String> altGeneIDs, Alignment seqAlignment);\\n    public String getAccession()\\n    public String getGeneSymbol()\\n    public GenomeInterval getTXRegion()\\n    public GenomeInterval getCDSRegion()\\n    public ImmutableList<GenomeInterval> getExonRegions()\\n    public String getSequence()\\n    public String getTrimmedSequence()\\n    public Alignment getSeqAlignment()\\n    public String getGeneID()\\n    public ImmutableSortedMap<String, String> getAltGeneIDs()\\n    public int getTranscriptSupportLevel()\\n    public Strand getStrand()\\n    public int getChr()\\n    public boolean isCoding()\\n    public boolean isMitochondrial()\\n    public int cdsTranscriptLength()\\n    public int transcriptLength()\\n    public boolean isHasSubstitutions()\\n    public boolean isHasIndels()\\n    public GenomeInterval intronRegion(int i)\\n    private void checkForConsistency()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public int compareTo(TranscriptModel o)\\n}\\n', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/reference/TranscriptModel.java', 'methodCount': 27, 'lineCount': 42}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "\t * Return augmented {@link effects} based on the transcript alone (coding/non-coding transcript).\n",
      "\t */\n",
      "private ImmutableSet<VariantEffect> buildEffectSet(Collection<VariantEffect> effects)\n",
      "**************************************\n",
      "private ImmutableSet<VariantEffect> buildEffectSet(Collection<VariantEffect> effects) {\n",
      "    final EnumSet<VariantEffect> tmpEffects;\n",
      "    if (transcript.isCoding()) {\n",
      "        tmpEffects = EnumSet.of(VariantEffect.CODING_TRANSCRIPT_VARIANT);\n",
      "    } else {\n",
      "        tmpEffects = EnumSet.of(VariantEffect.NON_CODING_TRANSCRIPT_VARIANT);\n",
      "    }\n",
      "    tmpEffects.addAll(effects);\n",
      "    return Sets.immutableEnumSet(tmpEffects);\n",
      "}\n",
      "2494\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 154/1243 [05:07<39:52,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7673\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  12%|        | 155/1243 [05:09<37:07,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7674\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 156/1243 [05:10<35:12,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7675\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 157/1243 [05:13<36:15,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7676\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 158/1243 [05:15<37:22,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7677\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/charite,jannovar/jannovar-core/src/main/java/de/charite/compbio/jannovar/Immutable.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{'Person': \"public class Person {\\n    /**\\n * the individual's name\\n */\\nprivate final String name;;\\n    /**\\n * the individual's father, or <code>null</code> if father is not in pedigree\\n */\\nprivate final Person father;;\\n    /**\\n * the individual's mother, or <code>null</code> if mother is not in pedigree\\n */\\nprivate final Person mother;;\\n    /**\\n * the individual's sex\\n */\\nprivate final Sex sex;;\\n    /**\\n * the individual's disease state\\n */\\nprivate final Disease disease;;\\n    /**\\n * extra fields from the PED file\\n */\\nprivate final ImmutableList<String> extraFields;;\\n    public Person(String name, Person father, Person mother, Sex sex, Disease disease, Collection<String> extraFields);\\n    public Person(String name, Person father, Person mother, Sex sex, Disease disease);\\n     Person(PedPerson pedPerson, PedFileContents pedFileContents, HashMap<String, Person> existing);\\n    public String getName()\\n    public Person getFather()\\n    public Person getMother()\\n    public Sex getSex()\\n    public Disease getDisease()\\n    public ImmutableList<String> getExtraFields()\\n    public boolean isFounder()\\n    public boolean isMale()\\n    public boolean isFemale()\\n    public boolean isAffected()\\n    public boolean isUnaffected()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n}\\n\"}\n",
      "/**\n",
      "\t * @return number of parents in pedigree\n",
      "\t */\n",
      "public int getNumberOfParents()\n",
      "**************************************\n",
      "public int getNumberOfParents() {\n",
      "    HashSet<String> parentNames = new HashSet<String>();\n",
      "    for (Person member : pedigree.getMembers()) {\n",
      "        if (member.getFather() != null)\n",
      "            parentNames.add(member.getFather().getName());\n",
      "        if (member.getMother() != null)\n",
      "            parentNames.add(member.getMother().getName());\n",
      "    }\n",
      "    return parentNames.size();\n",
      "}\n",
      "536\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 159/1243 [05:17<37:04,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7696\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 160/1243 [05:19<34:46,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7704\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 161/1243 [05:21<35:17,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7730\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 162/1243 [05:23<37:08,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7735\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 163/1243 [05:24<34:52,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7759\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "\t * (non-Javadoc)\n",
      "\t * \n",
      "\t * @see java.lang.Object#hashCode()\n",
      "\t */\n",
      "public int hashCode()\n",
      "**************************************\n",
      "@Override\n",
      "public int hashCode() {\n",
      "    final int prime = 31;\n",
      "    int result = 1;\n",
      "    result = prime * result + confidence;\n",
      "    result = prime * result + ((entity == null) ? 0 : entity.hashCode());\n",
      "    result = prime * result + geneID;\n",
      "    result = prime * result + ((location == null) ? 0 : location.hashCode());\n",
      "    result = prime * result + pmid;\n",
      "    result = prime * result + species;\n",
      "    return result;\n",
      "}\n",
      "166\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 164/1243 [05:26<34:58,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7761\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 165/1243 [05:28<34:53,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7763\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 166/1243 [05:30<35:41,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7772\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  13%|        | 167/1243 [05:33<36:24,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7780\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 168/1243 [05:35<37:22,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7782\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class MutationFinder {\\n    /*\\n     * Define the case-sensitive flag used for the regular_expression input file. This flag is appended to each regular-expression that is\\n     * case-sensitive\\n     */\\nprivate final String CASE_SENSITIVE = \"[CASE_SENSITIVE]\";;\\n    protected final static String MUT_RES = \"?P<mut_res>\";;\\n    protected final static String WT_RES = \"?P<wt_res>\";;\\n    protected final static String POS = \"?P<pos>\";;\\n    //Contains a list of ambiguous \"mutation\"-names (e.g., T47D generally refers to a cell-line and not to a mutation)\\nprotected final static Set<String> ambiguous = new HashSet<String>();;\\n    /*\\n     * regular_expressions: an interative set of regular expressions to be applied for extracting mutations. These are in the default python syntax\\n     * (i.e., perl regular expressions), with the single exception being that regular expressions which should be performed in a case sensitive manner\\n     * should be followed by the string \\'[CASE_SENSITIVE]\\', with no spaces between it and the regular expression. This can be a list, a file, or any\\n     * other object which supports iteration. For an example, you should refer to the regex.txt file in the MutationFinder directory.\\n     * \\n     * Since Java does not implement \"named groups\", we must process the Python regular expressions and store a mapping of the three components of a\\n     * point mutation to the groups that they are members of. These mappings are stored in the Map<String, Integer>, where the String is one of\\n     * MUT_RES, WT_RES, or POS, and the Integer represents its parenthetical group.\\n     */\\nprivate Map<MyPattern, Map<String, Integer>> regular_expressions;;\\n    public MutationFinder();\\n    public MutationFinder(String fileName);\\n    public MutationFinder(File file);\\n    public MutationFinder(Set<String> unprocessed_python_regexes);\\n    private void processPythonRegex(String regexStr, int line)\\n    private void loadRegularExpressionsFromFile(File file)\\n    private void loadRegularExpressionsFromJar(String file)\\n    private void loadRegularExpressionsFromStream(BufferedReader br)\\n    private void initializeAmbiguousMentions()\\n    public static Map<String, Integer> extractMappingsFromPythonRegex(String pythonRegex)\\n    public static String removeTagsFromPythonRegex(String regexStr)\\n    private static int countRegExParenthesesBeforeIndex(String regexStr, int index)\\n    private void postProcess(Map<Mutation, Set<int[]>> mutations)\\n    public Map<Mutation, Set<int[]>> extractMutations(String rawText) throws MutationException\\n    public static void main(String[] args)\\n}\\n', 'filePath': '/home/ubuntu/repos/rockt,SETH/src/main/java/edu/uchsc/ccp/nlp/ei/mutation/MutationFinder.java', 'methodCount': 15, 'lineCount': 23}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     *  (non-Javadoc)\n",
      "     * @see java.lang.Object#toString()\n",
      "     */\n",
      "public String toString()\n",
      "**************************************\n",
      "@Override\n",
      "public String toString() {\n",
      "    if (type == 'c' || type == 'g') {\n",
      "        if (insdeldup)\n",
      "            return type + \".\" + location + wildtype + mutation;\n",
      "        else\n",
      "            return type + \".\" + location + wildtype + \">\" + mutation;\n",
      "    } else\n",
      "        return type + \".\" + wildtype + location + mutation;\n",
      "}\n",
      "785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 169/1243 [05:37<36:05,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7788\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 170/1243 [05:39<37:14,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7831\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 171/1243 [05:41<38:21,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7834\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 172/1243 [05:44<41:34,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7839\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 173/1243 [05:46<40:49,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7844\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class MutationExtractor {\\n    /**\\n * This method simple fills the two mappings between amino acid one letter code and the amino acid three letter code\\n *\\n * build a mapping from three-letter code to one-letter code\\n */\\npublic static Map<String, String> populateAminoAcidThreeToOneLookupMap = null;;\\n    /**\\n * This method simple fills the two mappings between amino acid name and the amino acid one-letter code\\n *\\n * build a mapping from amino acid full name to one-letter code\\n */\\npublic static Map<String, String> populateAminoAcidNameToOneLookupMap = null;;\\n    public abstract Map<Mutation, Set<int[]>> extractMutations(String rawText) throws MutationException\\n    protected void warn(String message)\\n    protected void error(String message)\\n}\\n', 'filePath': '/home/ubuntu/repos/rockt,SETH/src/main/java/edu/uchsc/ccp/nlp/ei/mutation/MutationExtractor.java', 'methodCount': 3, 'lineCount': 7}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/rockt,SETH/src/main/java/seth/ner/wrapper/Type.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 3}\n",
      "{'UniprotFeature': 'public class UniprotFeature {\\n    private final static Logger logger = LoggerFactory.getLogger(UniprotFeature.class);;\\n    /**\\n * Gene identifier (Currently Entrez-Gene)\\n */\\nprivate final int geneId;;\\n    /**\\n * Modification type (e.g. signal peptide)\\n */\\nprivate final String modificationType;;\\n    /**\\n * Start location\\n */\\nprivate final int startLoc;;\\n    /**\\n * End location\\n */\\nprivate final int endLoc;;\\n    /**\\n *  Database connection\\n */\\nprivate static DatabaseConnection conn = null;;\\n    /**\\n *  Prepared statement for UniProt query\\n */\\nprivate static PreparedStatement uniprotQuery;;\\n    public UniprotFeature(int geneId, String modificationType, int startLoc, int endLoc);\\n    public static void init(DatabaseConnection mysql, String uniProt_Table) throws SQLException\\n    public static List<UniprotFeature> getFeatures(int geneID)\\n    public String toString()\\n    public int getGeneId()\\n    public String getModificationType()\\n    public int getStartLoc()\\n    public int getEndLoc()\\n}\\n', 'dbSNP': 'public class dbSNP {\\n    /**\\n * dbSNP internal id.\\n */\\nprotected int rsID;;\\n    /**\\n * gene associated with this dbSNP entry.\\n */\\nprotected int geneID;;\\n    /**\\n * known mutated amino acid residues for this SNP\\n */\\nprotected Set<String> residues;;\\n    /**\\n * Mutation has one wildtype\\n */\\nprotected String wildtype;;\\n    /**\\n * location of the mutation in terms of amino acids\\n */\\nprotected int aaPosition;;\\n    /**\\n * Different Strings describing the variations in <a href=\"http://www.hgvs.org/mutnomen/\"HGVS</a> nomenclature\\n */\\nprotected Set<HGVS> hgvs;;\\n    /**\\n * Used to split residues\\n */\\nprivate static final Pattern pattern = Pattern.compile(\"\\\\\\\\|\");;\\n    /**\\n * Database connection\\n */\\nprivate static DatabaseConnection conn = null;;\\n    /**\\n * Prepared statement for HGVS entries\\n */\\nprivate static PreparedStatement hgvsQuery;;\\n    /**\\n * Prepared statement for PSM entries\\n */\\nprivate static PreparedStatement snpQuery;;\\n    public static void init(DatabaseConnection mysql, String psmTable, String hgvs_view) throws SQLException\\n    public static List<dbSNP> getSNP(int geneID)\\n    public dbSNP();\\n    public dbSNP(int rsID, int geneID, Set<String> residues, int aaPosition, String wildtype);\\n    public int getRsID()\\n    public void setRsID(int rsID)\\n    public int getGeneID()\\n    public void setGeneID(int geneID)\\n    public Set<String> getResidues()\\n    public String getWildtype()\\n    public void setResidues(Set<String> residues)\\n    public int getAaPosition()\\n    public void setAaPosition(int aaPosition)\\n    public Set<HGVS> getHgvs()\\n    public void setHgvs(Set<HGVS> hgvs)\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Check if the {@link MutationMention} mention can be normalized to a PSM-candidate with a variable offset, derived from UniProt\n",
      "     *\n",
      "     * @param candidate     dbSNP candidate\n",
      "     * @param features List of possible UniProt features\n",
      "     * @return null if candidate can not be normalized, or the feature which was used to normalize the candidate\n",
      "     */\n",
      "private UniprotFeature normalizePSMVariableOffset(dbSNP candidate, List<UniprotFeature> features)\n",
      "**************************************\n",
      "private UniprotFeature normalizePSMVariableOffset(dbSNP candidate, List<UniprotFeature> features) {\n",
      "    for (UniprotFeature feature : features) {\n",
      "        if (feature.getGeneId() != candidate.getGeneID())\n",
      "            continue;\n",
      "        if (Integer.toString(candidate.getAaPosition() - feature.getEndLoc() + feature.getStartLoc() - 1).equals(position))\n",
      "            return feature;\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "1298\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 174/1243 [05:48<37:44,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7845\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 175/1243 [05:50<38:31,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7846\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 176/1243 [05:52<38:18,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7906\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 177/1243 [05:54<38:23,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7951\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 178/1243 [05:57<38:04,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8018\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 17, 'classCount': 1, 'methodsOnly': 'public class MutationMention {\\n    private final static Logger logger = LoggerFactory.getLogger(MutationMention.class);;\\n    /**\\n * Type of Mutation (e.g., substitution, insertion, ...)\\n */\\nprotected Type type;;\\n    /**\\n * Refers to which tool has been used for extraction (e.g., MutationFinder, SETH, ...)\\n */\\nprotected Tool tool;;\\n    /**\\n * Location in the text.\\n */\\nprotected EntityOffset location;;\\n    /**\\n * Text mention in the text.\\n */\\nprotected String text;;\\n    /**\\n * Used reference sequence (e.g. c., g., ...)\\n */\\nprotected String ref;;\\n    /**\\n * Wildtype residue.\\n */\\nprotected String wtResidue;;\\n    /**\\n * Mutated residue.\\n */\\nprotected String mutResidue;;\\n    /**\\n * Location of the mutation wrt residue or nucleotides\\n */\\nprotected String position;;\\n    /**\\n * Flag indicating if this SNP is a nucleotide sequence mutation (NSM).\\n */\\nprotected boolean nsm;;\\n    /**\\n * Flag indicating if this SNP is a protein sequence mutation (PSM).\\n */\\nprotected boolean psm;;\\n    /**\\n * Flag indicating if this SNP is ambiguous.\\n *  I.e., it can be a PSM or NSM\\n */\\nprotected boolean ambiguous;;\\n    /**\\n * Which regular expression has been used to extract this SNP from text (MutationFinder only)\\n */\\nprotected int patternId;;\\n    /**\\n * Pattern is used to extract wildtype-location-residue from a string e.g. A123T\\n */\\nprivate static final Pattern pattern = Pattern.compile(\"^([A-Z])([\\\\\\\\-\\\\\\\\+\\\\\\\\*]?[1-9][0-9]*[\\\\\\\\-\\\\\\\\+]?[0-9]*)([A-Z])$\");;\\n    /**\\n * Contains all ambigious characters used for nucleotides as well as amino acids\\n */\\nprivate static final ArrayList<String> atgc = new ArrayList<String>(Arrays.asList(\"A\", \"T\", \"G\", \"C\"));;\\n    /**\\n * List with normalized SNPs\\n */\\nprotected List<dbSNPNormalized> normalized;;\\n    /**\\n * Set of transcripts with which we successfully associated this Element\\n */\\nprotected Set<TranscriptNormalized> transcripts;;\\n    public void normalizeSequences(Set<Transcript> candidateTranscripts, List<UniprotFeature> features, boolean append)\\n    public void normalizeSNP(int id)\\n    public void normalizeSNP(List<dbSNP> candidates, List<UniprotFeature> features, boolean append)\\n    private static List<dbSNPNormalized> cleanResults(List<dbSNPNormalized> snpList)\\n    private boolean checkPSM_WildType(dbSNP candidate, boolean forward)\\n    private boolean checkPSM_Mutated(dbSNP candidate, boolean forward)\\n    private boolean normalizePSMSimple(dbSNP candidate)\\n    private boolean normalizePSMMethionine(dbSNP candidate)\\n    private UniprotFeature normalizePSMVariableOffset(dbSNP candidate, List<UniprotFeature> features)\\n    private boolean normalizePSMSimpleHGVS(dbSNP candidate)\\n    private boolean normalizePSMSimpleHGVSSwap(dbSNP candidate)\\n    private boolean normalizePSMMethionineHGVS(dbSNP candidate)\\n    private boolean normalizePSMMethionineSwapHGVS(dbSNP candidate)\\n    private boolean forwardNSMSimple(dbSNP candidate)\\n    private boolean reverseNSMSimple(dbSNP candidate)\\n    public static Map<Integer, List<MutationMention>> readMutations(String file) throws IOException\\n    public MutationMention();\\n    public MutationMention(String mutation);\\n    public MutationMention(EntityOffset location, String mutation);\\n    public MutationMention(int start, int end, String text, String ref, String location, String wild, String mutated, Type type, Tool tool);\\n    public EntityOffset getLocation()\\n    public int getStart()\\n    public int getEnd()\\n    private void setLocation(EntityOffset location)\\n    public String getWtResidue()\\n    public void setWtResidue(String wtResidue)\\n    public String getMutResidue()\\n    public void setMutResidue(String mutResidue)\\n    public String getPosition()\\n    public void setPosition(String position)\\n    public boolean isNsm()\\n    public void setNsm(boolean nsm)\\n    public boolean isPsm()\\n    public void setPsm(boolean psm)\\n    public boolean isAmbiguous()\\n    public void setAmbiguous(boolean ambiguous)\\n    public Type getType()\\n    public Tool getTool()\\n    public String getText()\\n    public String getRef()\\n    public List<dbSNPNormalized> getNormalized()\\n    public List<dbSNPNormalized> getBestNormalized()\\n    public Set<TranscriptNormalized> getTranscripts()\\n    public int getPatternId()\\n    public void setPatternId(int patternId)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toHGVS()\\n    private static String cleanSETHString(String mutation)\\n    private boolean isAminoAcid()\\n    private boolean isNucleotide()\\n    private boolean isAmgib()\\n    private void setMutationLocation()\\n    public String toString()\\n    public String toNormalized()\\n}\\n', 'filePath': '/home/ubuntu/repos/rockt,SETH/src/main/java/de/hu/berlin/wbi/objects/MutationMention.java', 'methodCount': 55, 'lineCount': 74}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Converts a scala {@link seth.ner.Mutation} object to a java {@link MutationMention}\n",
      "     * @param mutation A scala {@link seth.ner.Mutation} object\n",
      "     * @return The corresponding java {@link MutationMention} object\n",
      "     */\n",
      "private MutationMention convert(seth.ner.Mutation mutation)\n",
      "**************************************\n",
      "private MutationMention convert(seth.ner.Mutation mutation) {\n",
      "    return new MutationMention(mutation.start(), mutation.end(), mutation.text(), mutation.ref(), mutation.loc(), mutation.wild(), mutation.mutated(), mutation.typ(), MutationMention.Tool.SETH);\n",
      "}\n",
      "1450\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 179/1243 [05:59<38:05,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8039\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  14%|        | 180/1243 [06:01<38:12,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8044\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 181/1243 [06:03<37:57,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8067\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 182/1243 [06:05<37:39,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8068\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 183/1243 [06:07<34:45,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8116\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MathException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -9004610152740737812L;;\\n    /**\\n * Pattern used to build the message.\\n */\\nprivate final String pattern;;\\n    /**\\n * Arguments used to build the message.\\n */\\nprivate final Object[] arguments;;\\n    public MathException();\\n    public MathException(String pattern, Object... arguments);\\n    public MathException(Throwable rootCause);\\n    public MathException(Throwable rootCause, String pattern, Object... arguments);\\n    private static String translate(String s, Locale locale)\\n    public String getPattern()\\n    public Object[] getArguments()\\n    public String getMessage(final Locale locale)\\n    public String getMessage()\\n    public String getLocalizedMessage()\\n    public void printStackTrace()\\n    public void printStackTrace(PrintStream out)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MathException.java', 'methodCount': 12, 'lineCount': 17}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MathRuntimeException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -5128983364075381060L;;\\n    /**\\n * Pattern used to build the message.\\n */\\nprivate final String pattern;;\\n    /**\\n * Arguments used to build the message.\\n */\\nprivate final Object[] arguments;;\\n    public MathRuntimeException(final String pattern, final Object... arguments);\\n    public MathRuntimeException(final Throwable rootCause);\\n    public MathRuntimeException(final Throwable rootCause, final String pattern, final Object... arguments);\\n    private static String translate(final String s, final Locale locale)\\n    private static String buildMessage(final Locale locale, final String pattern, final Object... arguments)\\n    public String getPattern()\\n    public Object[] getArguments()\\n    public String getMessage(final Locale locale)\\n    public String getMessage()\\n    public String getLocalizedMessage()\\n    public void printStackTrace()\\n    public void printStackTrace(final PrintStream out)\\n    public static ArithmeticException createArithmeticException(final String pattern, final Object... arguments)\\n    public static ArrayIndexOutOfBoundsException createArrayIndexOutOfBoundsException(final String pattern, final Object... arguments)\\n    public static EOFException createEOFException(final String pattern, final Object... arguments)\\n    public static IOException createIOException(final Throwable rootCause)\\n    public static IllegalArgumentException createIllegalArgumentException(final String pattern, final Object... arguments)\\n    public static IllegalArgumentException createIllegalArgumentException(final Throwable rootCause)\\n    public static IllegalStateException createIllegalStateException(final String pattern, final Object... arguments)\\n    public static ConcurrentModificationException createConcurrentModificationException(final String pattern, final Object... arguments)\\n    public static NoSuchElementException createNoSuchElementException(final String pattern, final Object... arguments)\\n    public static NullPointerException createNullPointerException(final String pattern, final Object... arguments)\\n    public static ParseException createParseException(final int offset, final String pattern, final Object... arguments)\\n    public static RuntimeException createInternalError(final Throwable cause)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MathRuntimeException.java', 'methodCount': 24, 'lineCount': 29}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Access the initial domain value, based on <code>p</code>, used to\n",
      "     * bracket a CDF root.\n",
      "     *\n",
      "     * @param p the desired probability for the critical value\n",
      "     * @return initial domain value\n",
      "     */\n",
      "protected double getInitialDomain(double p)\n",
      "**************************************\n",
      "@Override\n",
      "protected double getInitialDomain(double p) {\n",
      "    if (p < .5) {\n",
      "        return mean * .5;\n",
      "    } else {\n",
      "        return mean;\n",
      "    }\n",
      "}\n",
      "863\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 184/1243 [06:09<35:51,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8127\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 185/1243 [06:11<36:30,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8132\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 186/1243 [06:13<34:25,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8171\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 187/1243 [06:15<35:15,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8185\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 188/1243 [06:17<35:23,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8198\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MathRuntimeException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -5128983364075381060L;;\\n    /**\\n * Pattern used to build the message.\\n */\\nprivate final String pattern;;\\n    /**\\n * Arguments used to build the message.\\n */\\nprivate final Object[] arguments;;\\n    public MathRuntimeException(final String pattern, final Object... arguments);\\n    public MathRuntimeException(final Throwable rootCause);\\n    public MathRuntimeException(final Throwable rootCause, final String pattern, final Object... arguments);\\n    private static String translate(final String s, final Locale locale)\\n    private static String buildMessage(final Locale locale, final String pattern, final Object... arguments)\\n    public String getPattern()\\n    public Object[] getArguments()\\n    public String getMessage(final Locale locale)\\n    public String getMessage()\\n    public String getLocalizedMessage()\\n    public void printStackTrace()\\n    public void printStackTrace(final PrintStream out)\\n    public static ArithmeticException createArithmeticException(final String pattern, final Object... arguments)\\n    public static ArrayIndexOutOfBoundsException createArrayIndexOutOfBoundsException(final String pattern, final Object... arguments)\\n    public static EOFException createEOFException(final String pattern, final Object... arguments)\\n    public static IOException createIOException(final Throwable rootCause)\\n    public static IllegalArgumentException createIllegalArgumentException(final String pattern, final Object... arguments)\\n    public static IllegalArgumentException createIllegalArgumentException(final Throwable rootCause)\\n    public static IllegalStateException createIllegalStateException(final String pattern, final Object... arguments)\\n    public static ConcurrentModificationException createConcurrentModificationException(final String pattern, final Object... arguments)\\n    public static NoSuchElementException createNoSuchElementException(final String pattern, final Object... arguments)\\n    public static NullPointerException createNullPointerException(final String pattern, final Object... arguments)\\n    public static ParseException createParseException(final int offset, final String pattern, final Object... arguments)\\n    public static RuntimeException createInternalError(final Throwable cause)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MathRuntimeException.java', 'methodCount': 24, 'lineCount': 29}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Access the initial domain value, based on <code>p</code>, used to\n",
      "     * bracket a CDF root.  This method is used by\n",
      "     * {@link #inverseCumulativeProbability(double)} to find critical values.\n",
      "     *\n",
      "     * @param p the desired probability for the critical value\n",
      "     * @return initial domain value\n",
      "     */\n",
      "protected double getInitialDomain(double p)\n",
      "**************************************\n",
      "@Override\n",
      "protected double getInitialDomain(double p) {\n",
      "    double ret;\n",
      "    if (p < .5) {\n",
      "        ret = median - scale;\n",
      "    } else if (p > .5) {\n",
      "        ret = median + scale;\n",
      "    } else {\n",
      "        ret = median;\n",
      "    }\n",
      "    return ret;\n",
      "}\n",
      "660\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 189/1243 [06:19<36:05,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8208\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 190/1243 [06:21<34:00,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8209\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 191/1243 [06:22<32:32,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8277\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  15%|        | 192/1243 [06:24<31:03,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8278\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 193/1243 [06:26<33:31,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8287\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MathException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -9004610152740737812L;;\\n    /**\\n * Pattern used to build the message.\\n */\\nprivate final String pattern;;\\n    /**\\n * Arguments used to build the message.\\n */\\nprivate final Object[] arguments;;\\n    public MathException();\\n    public MathException(String pattern, Object... arguments);\\n    public MathException(Throwable rootCause);\\n    public MathException(Throwable rootCause, String pattern, Object... arguments);\\n    private static String translate(String s, Locale locale)\\n    public String getPattern()\\n    public Object[] getArguments()\\n    public String getMessage(final Locale locale)\\n    public String getMessage()\\n    public String getLocalizedMessage()\\n    public void printStackTrace()\\n    public void printStackTrace(PrintStream out)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MathException.java', 'methodCount': 12, 'lineCount': 17}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MathRuntimeException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -5128983364075381060L;;\\n    /**\\n * Pattern used to build the message.\\n */\\nprivate final String pattern;;\\n    /**\\n * Arguments used to build the message.\\n */\\nprivate final Object[] arguments;;\\n    public MathRuntimeException(final String pattern, final Object... arguments);\\n    public MathRuntimeException(final Throwable rootCause);\\n    public MathRuntimeException(final Throwable rootCause, final String pattern, final Object... arguments);\\n    private static String translate(final String s, final Locale locale)\\n    private static String buildMessage(final Locale locale, final String pattern, final Object... arguments)\\n    public String getPattern()\\n    public Object[] getArguments()\\n    public String getMessage(final Locale locale)\\n    public String getMessage()\\n    public String getLocalizedMessage()\\n    public void printStackTrace()\\n    public void printStackTrace(final PrintStream out)\\n    public static ArithmeticException createArithmeticException(final String pattern, final Object... arguments)\\n    public static ArrayIndexOutOfBoundsException createArrayIndexOutOfBoundsException(final String pattern, final Object... arguments)\\n    public static EOFException createEOFException(final String pattern, final Object... arguments)\\n    public static IOException createIOException(final Throwable rootCause)\\n    public static IllegalArgumentException createIllegalArgumentException(final String pattern, final Object... arguments)\\n    public static IllegalArgumentException createIllegalArgumentException(final Throwable rootCause)\\n    public static IllegalStateException createIllegalStateException(final String pattern, final Object... arguments)\\n    public static ConcurrentModificationException createConcurrentModificationException(final String pattern, final Object... arguments)\\n    public static NoSuchElementException createNoSuchElementException(final String pattern, final Object... arguments)\\n    public static NullPointerException createNullPointerException(final String pattern, final Object... arguments)\\n    public static ParseException createParseException(final int offset, final String pattern, final Object... arguments)\\n    public static RuntimeException createInternalError(final Throwable cause)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MathRuntimeException.java', 'methodCount': 24, 'lineCount': 29}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class Beta {\\n    /**\\n * Maximum allowed numerical error.\\n */\\nprivate static final double DEFAULT_EPSILON = 10e-15;;\\n    private Beta();\\n    public static double regularizedBeta(double x, double a, double b) throws MathException\\n    public static double regularizedBeta(double x, double a, double b, double epsilon) throws MathException\\n    public static double regularizedBeta(double x, double a, double b, int maxIterations) throws MathException\\n    public static double regularizedBeta(double x, final double a, final double b, double epsilon, int maxIterations) throws MathException\\n    public static double logBeta(double a, double b)\\n    public static double logBeta(double a, double b, double epsilon, int maxIterations)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/special/Beta.java', 'methodCount': 7, 'lineCount': 10}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class Gamma {\\n    /**\\n * <a href=\"http://en.wikipedia.org/wiki/Euler-Mascheroni_constant\">Euler-Mascheroni constant</a>\\n *\\n * @since 2.0\\n */\\npublic static final double GAMMA = 0.577215664901532860606512090082;;\\n    /**\\n * Maximum allowed numerical error.\\n */\\nprivate static final double DEFAULT_EPSILON = 10e-15;;\\n    /**\\n * Lanczos coefficients\\n */\\nprivate static final double[] LANCZOS = { 0.99999999999999709182, 57.156235665862923517, -59.597960355475491248, 14.136097974741747174, -0.49191381609762019978, .33994649984811888699e-4, .46523628927048575665e-4, -.98374475304879564677e-4, .15808870322491248884e-3, -.21026444172410488319e-3, .21743961811521264320e-3, -.16431810653676389022e-3, .84418223983852743293e-4, -.26190838401581408670e-4, .36899182659531622704e-5 };;\\n    /**\\n * Avoid repeated computation of log of 2 PI in logGamma\\n */\\nprivate static final double HALF_LOG_2_PI = 0.5 * Math.log(2.0 * Math.PI);;\\n    // limits for switching algorithm in digamma\\n/**\\n * C limit.\\n */\\nprivate static final double C_LIMIT = 49;;\\n    /**\\n * S limit.\\n */\\nprivate static final double S_LIMIT = 1e-5;;\\n    private Gamma();\\n    public static double logGamma(double x)\\n    public static double regularizedGammaP(double a, double x) throws MathException\\n    public static double regularizedGammaP(double a, double x, double epsilon, int maxIterations) throws MathException\\n    public static double regularizedGammaQ(double a, double x) throws MathException\\n    public static double regularizedGammaQ(final double a, double x, double epsilon, int maxIterations) throws MathException\\n    public static double digamma(double x)\\n    public static double trigamma(double x)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/special/Gamma.java', 'methodCount': 8, 'lineCount': 16}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * {@inheritDoc}\n",
      "     */\n",
      "public double inverseCumulativeProbability(double p) throws MathException\n",
      "**************************************\n",
      "@Override\n",
      "public double inverseCumulativeProbability(double p) throws MathException {\n",
      "    if (p == 0) {\n",
      "        return 0;\n",
      "    } else if (p == 1) {\n",
      "        return 1;\n",
      "    } else {\n",
      "        return super.inverseCumulativeProbability(p);\n",
      "    }\n",
      "}\n",
      "1674\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 194/1243 [07:00<3:22:01, 11.55s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8291\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 195/1243 [07:02<2:32:34,  8.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8300\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 196/1243 [07:04<1:55:55,  6.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8326\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 197/1243 [07:06<1:32:58,  5.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8417\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 198/1243 [07:08<1:13:09,  4.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8420\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class FunctionEvaluationException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -4305020489115478365L;;\\n    /**\\n * Message for failed evaluation.\\n */\\nprivate static final String FAILED_EVALUATION_MESSAGE = \"evaluation failed for argument = {0}\";;\\n    /**\\n * Argument causing function evaluation failure\\n */\\nprivate double[] argument;;\\n    public FunctionEvaluationException(double argument);\\n    public FunctionEvaluationException(double argument, String pattern, Object... arguments);\\n    public FunctionEvaluationException(double[] argument, String pattern, Object... arguments);\\n    public FunctionEvaluationException(Throwable cause, double argument);\\n    public FunctionEvaluationException(Throwable cause, double[] argument);\\n    public FunctionEvaluationException(Throwable cause, double argument, String pattern, Object... arguments);\\n    public FunctionEvaluationException(Throwable cause, double[] argument, String pattern, Object... arguments);\\n    public double[] getArgument()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/FunctionEvaluationException.java', 'methodCount': 8, 'lineCount': 13}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MathRuntimeException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -5128983364075381060L;;\\n    /**\\n * Pattern used to build the message.\\n */\\nprivate final String pattern;;\\n    /**\\n * Arguments used to build the message.\\n */\\nprivate final Object[] arguments;;\\n    public MathRuntimeException(final String pattern, final Object... arguments);\\n    public MathRuntimeException(final Throwable rootCause);\\n    public MathRuntimeException(final Throwable rootCause, final String pattern, final Object... arguments);\\n    private static String translate(final String s, final Locale locale)\\n    private static String buildMessage(final Locale locale, final String pattern, final Object... arguments)\\n    public String getPattern()\\n    public Object[] getArguments()\\n    public String getMessage(final Locale locale)\\n    public String getMessage()\\n    public String getLocalizedMessage()\\n    public void printStackTrace()\\n    public void printStackTrace(final PrintStream out)\\n    public static ArithmeticException createArithmeticException(final String pattern, final Object... arguments)\\n    public static ArrayIndexOutOfBoundsException createArrayIndexOutOfBoundsException(final String pattern, final Object... arguments)\\n    public static EOFException createEOFException(final String pattern, final Object... arguments)\\n    public static IOException createIOException(final Throwable rootCause)\\n    public static IllegalArgumentException createIllegalArgumentException(final String pattern, final Object... arguments)\\n    public static IllegalArgumentException createIllegalArgumentException(final Throwable rootCause)\\n    public static IllegalStateException createIllegalStateException(final String pattern, final Object... arguments)\\n    public static ConcurrentModificationException createConcurrentModificationException(final String pattern, final Object... arguments)\\n    public static NoSuchElementException createNoSuchElementException(final String pattern, final Object... arguments)\\n    public static NullPointerException createNullPointerException(final String pattern, final Object... arguments)\\n    public static ParseException createParseException(final int offset, final String pattern, final Object... arguments)\\n    public static RuntimeException createInternalError(final Throwable cause)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MathRuntimeException.java', 'methodCount': 24, 'lineCount': 29}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class MaxIterationsExceededException {\\n    /**\\n * Serializable version identifier.\\n */\\nprivate static final long serialVersionUID = -7821226672760574694L;;\\n    /**\\n * Maximal number of iterations allowed.\\n */\\nprivate final int maxIterations;;\\n    public MaxIterationsExceededException(final int maxIterations);\\n    public MaxIterationsExceededException(final int maxIterations, final String pattern, final Object... arguments);\\n    public int getMaxIterations()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/MaxIterationsExceededException.java', 'methodCount': 3, 'lineCount': 7}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface UnivariateRealFunction {\\n     double value(double x) throws FunctionEvaluationException\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/apache/commons/math/analysis/UnivariateRealFunction.java', 'methodCount': 1, 'lineCount': 3}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * {@inheritDoc}\n",
      "     */\n",
      "protected void verifyIterationCount() throws IllegalArgumentException\n",
      "**************************************\n",
      "@Override\n",
      "protected void verifyIterationCount() throws IllegalArgumentException {\n",
      "    super.verifyIterationCount();\n",
      "    if (maximalIterationCount > 32) {\n",
      "        throw MathRuntimeException.createIllegalArgumentException(\"invalid iteration limits: min={0}, max={1}\", 0, 32);\n",
      "    }\n",
      "}\n",
      "1110\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 199/1243 [07:10<1:02:20,  3.58s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8487\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 200/1243 [07:12<55:11,  3.17s/it]  "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8502\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 201/1243 [07:14<49:44,  2.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8548\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 202/1243 [07:16<43:00,  2.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8564\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 203/1243 [07:18<40:44,  2.35s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8585\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'JSONException': 'public class JSONException {\\n    private static final long serialVersionUID = 1L;;\\n    public JSONException(String s);\\n}\\n', 'JSONStringer': \"public class JSONStringer {\\n    /**\\n * The output data, containing at most one top-level array or object.\\n */\\nfinal StringBuilder out = new StringBuilder();;\\n    /**\\n * Unlike the original implementation, this stack isn't limited to 20\\n * levels of nesting.\\n */\\nprivate final List<Scope> stack = new ArrayList<Scope>();;\\n    /**\\n * A string containing a full set of spaces for a single level of\\n * indentation, or null for no pretty printing.\\n */\\nprivate final String indent;;\\n    public JSONStringer();\\n     JSONStringer(int indentSpaces);\\n    public JSONStringer array() throws JSONException\\n    public JSONStringer endArray() throws JSONException\\n    public JSONStringer object() throws JSONException\\n    public JSONStringer endObject() throws JSONException\\n     JSONStringer open(Scope empty, String openBracket) throws JSONException\\n     JSONStringer close(Scope empty, Scope nonempty, String closeBracket) throws JSONException\\n    private Scope peek() throws JSONException\\n    private void replaceTop(Scope topOfStack)\\n    public JSONStringer value(Object value) throws JSONException\\n    public JSONStringer value(boolean value) throws JSONException\\n    public JSONStringer value(double value) throws JSONException\\n    public JSONStringer value(long value) throws JSONException\\n    private void string(String value)\\n    private void newline()\\n    public JSONStringer key(String name) throws JSONException\\n    private void beforeKey() throws JSONException\\n    private void beforeValue() throws JSONException\\n    public String toString()\\n}\\n\"}\n",
      "/**\n",
      "     * Closes the current scope by appending any necessary whitespace and the\n",
      "     * given bracket.\n",
      "     */\n",
      " JSONStringer close(Scope empty, Scope nonempty, String closeBracket) throws JSONException\n",
      "**************************************\n",
      "JSONStringer close(Scope empty, Scope nonempty, String closeBracket) throws JSONException {\n",
      "    Scope context = peek();\n",
      "    if (context != nonempty && context != empty) {\n",
      "        throw new JSONException(\"Nesting problem\");\n",
      "    }\n",
      "    stack.remove(stack.size() - 1);\n",
      "    if (context == nonempty) {\n",
      "        newline();\n",
      "    }\n",
      "    out.append(closeBracket);\n",
      "    return this;\n",
      "}\n",
      "538\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 204/1243 [07:20<36:43,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8595\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  16%|        | 205/1243 [07:22<37:06,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8653\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 206/1243 [07:23<33:58,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8687\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 207/1243 [07:26<35:04,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8910\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 208/1243 [07:28<35:12,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8914\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Input.java', 'methodCount': 34, 'lineCount': 44}, {'fieldCount': 17, 'classCount': 1, 'methodsOnly': 'public class Log {\\n    static public PrintStream nullStream = new PrintStream(new OutputStream() {\\n\\n    @Override\\n    public void write(byte[] b, int off, int len) throws IOException {\\n    }\\n\\n    @Override\\n    public void write(int b) throws IOException {\\n    }\\n});;\\n    static Level level = Level.info;;\\n    public static String[] values()\\n    static public PrintStream err;;\\n    public static void err(String msg)\\n    static public PrintStream warning;;\\n    public static void warning(String msg)\\n    static public PrintStream info;;\\n    public static void info(String msg)\\n    static public PrintStream debug;;\\n    public static void debug(String msg)\\n    static public PrintStream trace;;\\n    public static void trace(String msg)\\n    static private PrintStream errIfOpen;;\\n    static private PrintStream warningIfOpen;;\\n    static private PrintStream infoIfOpen;;\\n    static private PrintStream debugIfOpen;;\\n    static private PrintStream traceIfOpen;;\\n    final static public int ERROR = 0;;\\n    final static public int WARNING = 1;;\\n    final static public int INFO = 2;;\\n    final static public int DEBUG = 3;;\\n    final static public int TRACE = 4;;\\n    public static void setLevel(Level newLevel)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Log.java', 'methodCount': 7, 'lineCount': 26}, {'fieldCount': 32, 'classCount': 1, 'methodsOnly': 'public class Alignment {\\n    protected Class<?> mapType()\\n    protected final static String NUCLEOTIDE = \"nucleotide\";;\\n    protected final static String[] IMPLEMENTATION_DIR = { \"beast.base.evolution.datatype\" };;\\n    protected static TreeMap<String, DataType> types = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);;\\n    public static void findDataTypes()\\n    public static TreeMap<String, DataType> getTypes()\\n    public static DataType getDataTypeByName(String name)\\n    final public Input<List<Sequence>> sequenceInput = new Input<>(\"sequence\", \"sequence and meta data for particular taxon\", new ArrayList<>(), Validate.OPTIONAL);;\\n    final public Input<TaxonSet> taxonSetInput = new Input<>(\"taxa\", \"An optional taxon-set used only to sort the sequences into the same order as they appear in the taxon-set.\", new TaxonSet(), Validate.OPTIONAL);;\\n    final public Input<Integer> stateCountInput = new Input<>(\"statecount\", \"maximum number of states in all sequences\");;\\n    final public Input<String> dataTypeInput = new Input<>(\"dataType\", \"data type, one of \" + types, NUCLEOTIDE, types.keySet().toArray(new String[0]));;\\n    final public Input<DataType.Base> userDataTypeInput = new Input<>(\"userDataType\", \"non-standard, user specified data type, if specified \\'dataType\\' is ignored\");;\\n    final public Input<Boolean> stripInvariantSitesInput = new Input<>(\"strip\", \"sets weight to zero for sites that are invariant (e.g. all 1, all A or all unkown)\", false);;\\n    final public Input<String> siteWeightsInput = new Input<>(\"weights\", \"comma separated list of weights, one for each site in the sequences. If not specified, each site has weight 1\");;\\n    final public Input<Boolean> isAscertainedInput = new Input<>(\"ascertained\", \"is true if the alignment allows ascertainment correction, i.e., conditioning the \" + \"Felsenstein likelihood on excluding constant sites from the alignment\", false);;\\n    final public Input<Integer> excludefromInput = new Input<>(\"excludefrom\", \"first site to condition on, default 0\", 0);;\\n    final public Input<Integer> excludetoInput = new Input<>(\"excludeto\", \"last site to condition on (but excluding this site), default 0\", 0);;\\n    final public Input<Integer> excludeeveryInput = new Input<>(\"excludeevery\", \"interval between sites to condition on (default 1)\", 1);;\\n    public Input<Integer> m_includefrom = new Input<>(\"includefrom\", \"first site to condition on, default 0\", 0);;\\n    public Input<Integer> m_includeto = new Input<>(\"includeto\", \"last site to condition on, default 0\", 0);;\\n    public Input<Integer> m_includeevery = new Input<>(\"includeevery\", \"interval between sites to condition on (default 1)\", 1);;\\n    protected List<Sequence> sequences = new ArrayList<>();;\\n    protected List<String> taxaNames = new ArrayList<>();;\\n    protected List<Integer> stateCounts = new ArrayList<>();;\\n    protected int maxStateCount;;\\n    protected List<List<Integer>> counts = new ArrayList<>();;\\n    protected DataType m_dataType;;\\n    protected int[] patternWeight;;\\n    protected int[] siteWeights = null;;\\n    public List<double[][]> tipLikelihoods = new ArrayList<>();;\\n    protected boolean usingTipLikelihoods = false;;\\n    protected int[][] sitePatterns;;\\n    protected int[] patternIndex;;\\n    protected Set<Integer> excludedPatterns;;\\n    protected List<Integer> m_nIncluded;;\\n    public boolean isAscertained;;\\n    public Alignment();\\n    public Alignment(List<Sequence> sequences, Integer stateCount, String dataType);\\n    public Alignment(List<Sequence> sequences, String dataType);\\n    public void initAndValidate()\\n    protected void initDataType()\\n    private void initializeWithSequenceList(List<Sequence> sequences, boolean log)\\n    protected void sanityCheckCalcPatternsSetUpAscertainment(boolean log)\\n    public void sortByTaxonSet(TaxonSet toSortBy)\\n    protected void setupAscertainment()\\n     static String getSequence(Alignment data, int taxonIndex)\\n    public List<String> getTaxaNames()\\n    public List<Integer> getStateCounts()\\n    public List<List<Integer>> getCounts()\\n    public DataType getDataType()\\n    public int getTaxonCount()\\n    public int getNrTaxa()\\n    public int getTaxonIndex(String id)\\n    public int getPatternCount()\\n    public int[] getPattern(int patternIndex_)\\n    public int getPattern(int taxonIndex, int patternIndex_)\\n    public int getPatternWeight(int patternIndex_)\\n    public int getMaxStateCount()\\n    public int getPatternIndex(int site)\\n    public int getSiteCount()\\n    public int[] getWeights()\\n    protected void calcPatterns()\\n    protected void calcPatterns(boolean log)\\n    private long getTotalWeight()\\n    public String toString(boolean singleLine)\\n    public double[] getTipLikelihoods(int taxonIndex, int patternIndex_)\\n    public boolean[] getStateSet(int state)\\n     boolean isAmbiguousState(int state)\\n    public Set<Integer> getExcludedPatternIndices()\\n    public int getExcludedPatternCount()\\n    public double getAscertainmentCorrection(double[] patternLogProbs)\\n    public static void sortByTaxonName(List<Sequence> seqs)\\n    public String getSequenceAsString(String taxon)\\n    public double[] calcFrequencies()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/evolution/alignment/Alignment.java', 'methodCount': 42, 'lineCount': 76}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class CalculationNode {\\n    protected void store()\\n     final void checkDirtiness()\\n    protected boolean requiresRecalculation()\\n    protected void restore()\\n    protected void accept()\\n    public final boolean isDirtyCalculation()\\n    private boolean isDirty = false;;\\n    protected int getChecksum()\\n    protected int preOperatorChecksum;;\\n    public void storeChecksum()\\n    public boolean matchesOldChecksum()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/inference/CalculationNode.java', 'methodCount': 9, 'lineCount': 13}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class RealParameter {\\n    final public Input<Double> lowerValueInput = new Input<>(\"lower\", \"lower value for this parameter (default -infinity)\");;\\n    final public Input<Double> upperValueInput = new Input<>(\"upper\", \"upper value for this parameter (default +infinity)\");;\\n    public RealParameter();\\n    public RealParameter(final Double[] values);\\n    public RealParameter(final String value);\\n    public void initAndValidate()\\n     Double getMax()\\n     Double getMin()\\n    public Double getValue()\\n    public double getArrayValue()\\n    public double getArrayValue(final int index)\\n    public void log(final long sample, final PrintStream out)\\n    public int scale(final double scale)\\n     void fromXML(final int dimension, final String lower, final String upper, final String[] valuesString)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/inference/parameter/RealParameter.java', 'methodCount': 12, 'lineCount': 16}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * CalculationNode implementation *\n",
      "     */\n",
      "protected boolean requiresRecalculation()\n",
      "**************************************\n",
      "@Override\n",
      "protected boolean requiresRecalculation() {\n",
      "    boolean recalculates = false;\n",
      "    if (frequenciesInput.get().somethingIsDirty()) {\n",
      "        needsUpdate = true;\n",
      "        recalculates = true;\n",
      "    }\n",
      "    return recalculates;\n",
      "}\n",
      "2979\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 209/1243 [07:30<36:28,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9033\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 210/1243 [07:32<36:19,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9074\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 211/1243 [07:34<34:00,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9078\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 212/1243 [07:35<31:39,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9088\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 213/1243 [07:37<30:02,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9103\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Sets the currently updating node partials for node nodeIndex. This may\n",
      "     * need to repeatedly copy the partials for the different category partitions\n",
      "     */\n",
      "public void setCurrentNodePartials(int nodeIndex, double[] partials)\n",
      "**************************************\n",
      "public void setCurrentNodePartials(int nodeIndex, double[] partials) {\n",
      "    if (partials.length < partialsSize) {\n",
      "        int k = 0;\n",
      "        for (int i = 0; i < nrOfMatrices; i++) {\n",
      "            System.arraycopy(partials, 0, this.partials[currentPartialsIndex[nodeIndex]][nodeIndex], k, partials.length);\n",
      "            k += partials.length;\n",
      "        }\n",
      "    } else {\n",
      "        System.arraycopy(partials, 0, this.partials[currentPartialsIndex[nodeIndex]][nodeIndex], 0, partials.length);\n",
      "    }\n",
      "}\n",
      "213\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 214/1243 [07:38<29:01,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9108\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 215/1243 [07:39<25:52,  1.51s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9110\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 216/1243 [07:41<28:16,  1.65s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  17%|        | 217/1243 [07:44<30:54,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9164\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 218/1243 [07:45<27:00,  1.58s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9166\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Store current state\n",
      "     */\n",
      "public void restore()\n",
      "**************************************\n",
      "@Override\n",
      "public void restore() {\n",
      "    int[] iTmp1 = m_iCurrentMatrices;\n",
      "    m_iCurrentMatrices = m_iStoredMatrices;\n",
      "    m_iStoredMatrices = iTmp1;\n",
      "    int[] iTmp2 = m_iCurrentPartials;\n",
      "    m_iCurrentPartials = m_iStoredPartials;\n",
      "    m_iStoredPartials = iTmp2;\n",
      "}\n",
      "132\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 219/1243 [07:46<27:30,  1.61s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9281\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 220/1243 [07:49<30:28,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9309\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 221/1243 [07:51<32:05,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9374\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 222/1243 [07:53<33:01,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 223/1243 [07:54<31:44,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9465\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class BEASTObject {\\n    /**\\n * set of Objects that have this Object in one of its Inputs *\\n * @deprecate use getOuputs() or BEASTObject.getOuputs(object) instead\\n */\\n@Deprecated\\npublic Set<BEASTInterface> outputs = new HashSet<>();;\\n    /**\\n * cache collecting all Inputs and InputForAnnotatedConstrutors\\n * indexed through input name\\n */\\nprivate Map<String, Input<?>> inputcache;;\\n    public Set<BEASTInterface> getOutputs()\\n    public Map<String, Input<?>> getInputs()\\n    protected String ID;;\\n    public String getID()\\n    public void setID(final String ID)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/BEASTObject.java', 'methodCount': 5, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class HeapSort {\\n    public static void sort(@SuppressWarnings(\"rawtypes\") List<? extends Comparable> list, int[] indices)\\n    public static void sort(Comparable<?>[] array)\\n    public static void sort(Object[] array, Comparator<?> comparator)\\n    public static void sort(double[] array)\\n    public static void sortAbs(double[] array)\\n    public static void sort(double[] array, int[] indices)\\n    private static void adjust(@SuppressWarnings(\"rawtypes\") List<? extends Comparable> list, int[] indices, int lower, int upper)\\n    private static void adjust(@SuppressWarnings(\"rawtypes\") Comparable[] array, int lower, int upper)\\n    private static void adjust(Object[] array, @SuppressWarnings(\"rawtypes\") Comparator comparator, int lower, int upper)\\n    private static void adjust(double[] array, int lower, int upper)\\n    private static void adjustAbs(double[] array, int lower, int upper)\\n    private static void adjust(double[] array, int[] indices, int lower, int upper)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/util/HeapSort.java', 'methodCount': 12, 'lineCount': 14}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'Node': \"public class Node {\\n    /**\\n * label nr of node, used mostly when this is a leaf.\\n */\\nprotected int labelNr;;\\n    /**\\n * height of this node.\\n */\\nprotected double height = Double.MAX_VALUE;;\\n    /**\\n * Arbitrarily labeled metadata on this node. Not currently implemented as part of state!\\n */\\nprotected Map<String, Object> metaData = new TreeMap<>();;\\n    /**\\n * Length metadata for the edge above this node.  Not currently implemented as part of state!\\n */\\nprotected Map<String, Object> lengthMetaData = new TreeMap<>();;\\n    /**\\n * list of children of this node *\\n * Don't use m_left and m_right directly\\n * Use getChildCount() and getChild(x) or getChildren() instead\\n */\\nprotected List<Node> children = new ArrayList<>();;\\n    /**\\n * parent node in the beast.tree, null if root *\\n */\\nprotected Node parent = null;;\\n    /**\\n * status of this node after an operation is performed on the state *\\n */\\nprotected int isDirty = Tree.IS_CLEAN;;\\n    /**\\n * meta-data contained in square brackets in Newick *\\n */\\npublic String metaDataString, lengthMetaDataString;;\\n    /**\\n * The Tree that this node is a part of.\\n * This allows e.g. access to the State containing the Tree *\\n */\\nprotected Tree m_tree;;\\n    public Node();\\n    public Node(final String id);\\n    public void setTree(Tree tree)\\n    public Tree getTree()\\n    public void initAndValidate()\\n    public int getNr()\\n    public void setNr(final int labelIndex)\\n    public double getHeight()\\n    public double getDate()\\n    public void setHeight(final double height)\\n    public void setHeightDA(final double height)\\n    public final double getLength()\\n    public int isDirty()\\n    public void makeDirty(final int dirty)\\n    public void makeAllDirty(final int dirty)\\n    public Node getParent()\\n    public void setParent(final Node parent)\\n    public void setParent(final Node parent, final boolean inOperator)\\n     void setParentImmediate(final Node parent)\\n    public List<Node> getChildren()\\n    public List<Node> getChildrenMutable()\\n    public List<Node> getAllChildNodes()\\n    public void getAllChildNodes(final List<Node> childNodes)\\n    public List<Node> getAllChildNodesAndSelf()\\n    public void getAllChildNodesAndSelf(final List<Node> childNodes)\\n    public List<Node> getAllLeafNodes()\\n    public void getAllLeafNodes(final List<Node> leafNodes)\\n    public boolean isRoot()\\n    public boolean isLeaf()\\n    public void removeChild(final Node child)\\n    public void removeAllChildren(final boolean inOperator)\\n    public void addChild(final Node child)\\n    public int getNodeCount()\\n    public int getLeafNodeCount()\\n    public int getInternalNodeCount()\\n    public String toShortNewick(final boolean printInternalNodeNumbers)\\n     String toSortedNewick(final int[] maxNodeInClade)\\n    public String toSortedNewick(int[] maxNodeInClade, boolean printMetaData)\\n    public String toNewick(final List<String> labels)\\n    public String toNewick(boolean onlyTopology)\\n    public String toNewick()\\n    public String getNewickMetaData()\\n    public String getNewickLengthMetaData()\\n    public String toString(final List<String> labels)\\n    public String toString()\\n    public int sort()\\n    public int labelInternalNodes(int labelIndex)\\n    public Node copy()\\n    public void assignTo(final Node[] nodes)\\n    public void assignFrom(final Node[] nodes, final Node node)\\n    public void setMetaData(final String pattern, final Object value)\\n    public void removeMetaData(final String key)\\n    public void setLengthMetaData(String key, Object value)\\n    public Object getMetaData(final String pattern)\\n    public Object getLengthMetaData(String key)\\n    public Set<String> getMetaDataNames()\\n    public Set<String> getLengthMetaDataNames()\\n    public int scale(final double scale)\\n    protected void startEditing()\\n    public int getChildCount()\\n    public Node getChild(final int childIndex)\\n    public void setChild(final int childIndex, final Node node)\\n    public void setLeft(final Node leftChild)\\n    public Node getLeft()\\n    public void setRight(final Node rightChild)\\n    public Node getRight()\\n    public static Node connect(final Node left, final Node right, final double h)\\n    public boolean isDirectAncestor()\\n    public boolean isFake()\\n    public Node getDirectAncestorChild()\\n    public Node getNonDirectAncestorChild()\\n    public Node getFakeChild()\\n}\\n\"}\n",
      "/**\n",
      "     * Retrieve the true child node of this sampled ancestor.\n",
      "     *\n",
      "     * @return true child node\n",
      "     */\n",
      "public Node getNonDirectAncestorChild()\n",
      "**************************************\n",
      "public Node getNonDirectAncestorChild() {\n",
      "    if (!this.isFake()) {\n",
      "        return null;\n",
      "    }\n",
      "    if ((this.getLeft()).isDirectAncestor()) {\n",
      "        return getRight();\n",
      "    }\n",
      "    if ((this.getRight()).isDirectAncestor()) {\n",
      "        return getLeft();\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "1748\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 224/1243 [07:57<32:53,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9484\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 225/1243 [07:59<33:46,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9486\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 226/1243 [08:01<34:58,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9493\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 227/1243 [08:03<34:33,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9512\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 228/1243 [08:05<35:26,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9513\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Input.java', 'methodCount': 34, 'lineCount': 44}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class CalculationNode {\\n    protected void store()\\n     final void checkDirtiness()\\n    protected boolean requiresRecalculation()\\n    protected void restore()\\n    protected void accept()\\n    public final boolean isDirtyCalculation()\\n    private boolean isDirty = false;;\\n    protected int getChecksum()\\n    protected int preOperatorChecksum;;\\n    public void storeChecksum()\\n    public boolean matchesOldChecksum()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/inference/CalculationNode.java', 'methodCount': 9, 'lineCount': 13}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class HeapSort {\\n    public static void sort(@SuppressWarnings(\"rawtypes\") List<? extends Comparable> list, int[] indices)\\n    public static void sort(Comparable<?>[] array)\\n    public static void sort(Object[] array, Comparator<?> comparator)\\n    public static void sort(double[] array)\\n    public static void sortAbs(double[] array)\\n    public static void sort(double[] array, int[] indices)\\n    private static void adjust(@SuppressWarnings(\"rawtypes\") List<? extends Comparable> list, int[] indices, int lower, int upper)\\n    private static void adjust(@SuppressWarnings(\"rawtypes\") Comparable[] array, int lower, int upper)\\n    private static void adjust(Object[] array, @SuppressWarnings(\"rawtypes\") Comparator comparator, int lower, int upper)\\n    private static void adjust(double[] array, int lower, int upper)\\n    private static void adjustAbs(double[] array, int lower, int upper)\\n    private static void adjust(double[] array, int[] indices, int lower, int upper)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/util/HeapSort.java', 'methodCount': 12, 'lineCount': 14}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Returns the type of interval observed.\n",
      "     */\n",
      "public IntervalType getIntervalType(int i)\n",
      "**************************************\n",
      "@Override\n",
      "public IntervalType getIntervalType(int i) {\n",
      "    if (!intervalsKnown) {\n",
      "        calculateIntervals();\n",
      "    }\n",
      "    if (i >= intervalCount)\n",
      "        throw new IllegalArgumentException();\n",
      "    int numEvents = getCoalescentEvents(i);\n",
      "    if (numEvents > 0)\n",
      "        return IntervalType.COALESCENT;\n",
      "    else if (numEvents < 0)\n",
      "        return IntervalType.SAMPLE;\n",
      "    else\n",
      "        return IntervalType.NOTHING;\n",
      "}\n",
      "1389\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  18%|        | 229/1243 [08:07<35:44,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9514\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 230/1243 [08:09<35:26,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9515\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 231/1243 [08:11<35:41,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9516\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 232/1243 [08:14<36:13,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 233/1243 [08:15<33:57,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9583\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Input.java', 'methodCount': 34, 'lineCount': 44}, {'fieldCount': 17, 'classCount': 1, 'methodsOnly': 'public class Log {\\n    static public PrintStream nullStream = new PrintStream(new OutputStream() {\\n\\n    @Override\\n    public void write(byte[] b, int off, int len) throws IOException {\\n    }\\n\\n    @Override\\n    public void write(int b) throws IOException {\\n    }\\n});;\\n    static Level level = Level.info;;\\n    public static String[] values()\\n    static public PrintStream err;;\\n    public static void err(String msg)\\n    static public PrintStream warning;;\\n    public static void warning(String msg)\\n    static public PrintStream info;;\\n    public static void info(String msg)\\n    static public PrintStream debug;;\\n    public static void debug(String msg)\\n    static public PrintStream trace;;\\n    public static void trace(String msg)\\n    static private PrintStream errIfOpen;;\\n    static private PrintStream warningIfOpen;;\\n    static private PrintStream infoIfOpen;;\\n    static private PrintStream debugIfOpen;;\\n    static private PrintStream traceIfOpen;;\\n    final static public int ERROR = 0;;\\n    final static public int WARNING = 1;;\\n    final static public int INFO = 2;;\\n    final static public int DEBUG = 3;;\\n    final static public int TRACE = 4;;\\n    public static void setLevel(Level newLevel)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Log.java', 'methodCount': 7, 'lineCount': 26}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class TaxonSet {\\n    final public Input<Alignment> alignmentInput = new Input<>(\"alignment\", \"alignment where each sequence represents a taxon\");;\\n    final public Input<List<Taxon>> taxonsetInput = new Input<>(\"taxon\", \"list of taxa making up the set\", new ArrayList<>());;\\n    protected List<String> taxaNames;;\\n    protected List<Taxon> taxonList;;\\n    public TaxonSet();\\n    public TaxonSet(final List<Taxon> taxa);\\n    public TaxonSet(final Alignment alignment);\\n    public TaxonSet(final String id, final List<Taxon> taxa);\\n    public void initAndValidate()\\n    public Set<Taxon> getTaxonSet()\\n    public List<String> asStringList()\\n    public Set<String> getTaxaNames()\\n    public String getTaxonId(int taxonIndex)\\n    public int getTaxonIndex(String id)\\n    public Taxon getTaxon(String id)\\n    public boolean containsAny(final Collection<String> taxa)\\n    public boolean containsAll(final Collection<String> taxa)\\n    public boolean containsAny(final TaxonSet taxa)\\n    public boolean containsAll(final TaxonSet taxa)\\n    public int getTaxonCount()\\n    public int getNrTaxa()\\n    public String toString()\\n    public String toString(String indent)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/evolution/alignment/TaxonSet.java', 'methodCount': 19, 'lineCount': 25}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class BEASTClassLoader {\\n    // maps package name (String) to class loader (MultiParentURLClassLoader)\\n// used to find class loader for a package\\nstatic Map<String, MultiParentURLClassLoader> package2classLoaderMap = new HashMap<>();;\\n    // singleton class loader\\nstatic final public BEASTClassLoader classLoader = new BEASTClassLoader(new URL[0], BEASTClassLoader.class.getClassLoader());;\\n    // maps service (=class name) to service providers (=set of class names)\\nstatic private Map<String, Set<String>> services = new HashMap<>();;\\n    // maps service provider (=class name) to class loader for that provider (=class)\\nstatic private Map<String, ClassLoader> class2loaderMap = new HashMap<>();;\\n    static private Set<String> namespaces = new HashSet<>();;\\n    private BEASTClassLoader(URL[] urls, ClassLoader parent);\\n    public void addURL(URL url)\\n    public void addURL(URL url, String packageName, Map<String, Set<String>> services)\\n    public void addParent(String packageName, String parentPackage)\\n    public void addJar(String jarFile)\\n    public void addJar(String jarFile, String packageName)\\n    public static Class<?> forName(String className) throws ClassNotFoundException\\n    public static Class<?> forName(String className, String service) throws ClassNotFoundException\\n    public static Set<String> loadService(Class<?> service)\\n    public static void initServices()\\n    public static void initServices(String classPath)\\n    public static void addServices(String versionFile)\\n    public static Map<String, Set<String>> getServices()\\n    public void addServices(String packageName, Map<String, Set<String>> services)\\n    private static MultiParentURLClassLoader getClassLoader(String packageName)\\n    public static void addService(String service, String className, String packageName)\\n    public static void delService(Map<String, Set<String>> serviceMap, String packageName)\\n    public static String usesExistingNamespaces(Set<String> services)\\n    public static URL getResource(String packageName, String resourceName)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/pkgmgmt/BEASTClassLoader.java', 'methodCount': 19, 'lineCount': 26}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'Node': \"public class Node {\\n    /**\\n * label nr of node, used mostly when this is a leaf.\\n */\\nprotected int labelNr;;\\n    /**\\n * height of this node.\\n */\\nprotected double height = Double.MAX_VALUE;;\\n    /**\\n * Arbitrarily labeled metadata on this node. Not currently implemented as part of state!\\n */\\nprotected Map<String, Object> metaData = new TreeMap<>();;\\n    /**\\n * Length metadata for the edge above this node.  Not currently implemented as part of state!\\n */\\nprotected Map<String, Object> lengthMetaData = new TreeMap<>();;\\n    /**\\n * list of children of this node *\\n * Don't use m_left and m_right directly\\n * Use getChildCount() and getChild(x) or getChildren() instead\\n */\\nprotected List<Node> children = new ArrayList<>();;\\n    /**\\n * parent node in the beast.tree, null if root *\\n */\\nprotected Node parent = null;;\\n    /**\\n * status of this node after an operation is performed on the state *\\n */\\nprotected int isDirty = Tree.IS_CLEAN;;\\n    /**\\n * meta-data contained in square brackets in Newick *\\n */\\npublic String metaDataString, lengthMetaDataString;;\\n    /**\\n * The Tree that this node is a part of.\\n * This allows e.g. access to the State containing the Tree *\\n */\\nprotected Tree m_tree;;\\n    public Node();\\n    public Node(final String id);\\n    public void setTree(Tree tree)\\n    public Tree getTree()\\n    public void initAndValidate()\\n    public int getNr()\\n    public void setNr(final int labelIndex)\\n    public double getHeight()\\n    public double getDate()\\n    public void setHeight(final double height)\\n    public void setHeightDA(final double height)\\n    public final double getLength()\\n    public int isDirty()\\n    public void makeDirty(final int dirty)\\n    public void makeAllDirty(final int dirty)\\n    public Node getParent()\\n    public void setParent(final Node parent)\\n    public void setParent(final Node parent, final boolean inOperator)\\n     void setParentImmediate(final Node parent)\\n    public List<Node> getChildren()\\n    public List<Node> getChildrenMutable()\\n    public List<Node> getAllChildNodes()\\n    public void getAllChildNodes(final List<Node> childNodes)\\n    public List<Node> getAllChildNodesAndSelf()\\n    public void getAllChildNodesAndSelf(final List<Node> childNodes)\\n    public List<Node> getAllLeafNodes()\\n    public void getAllLeafNodes(final List<Node> leafNodes)\\n    public boolean isRoot()\\n    public boolean isLeaf()\\n    public void removeChild(final Node child)\\n    public void removeAllChildren(final boolean inOperator)\\n    public void addChild(final Node child)\\n    public int getNodeCount()\\n    public int getLeafNodeCount()\\n    public int getInternalNodeCount()\\n    public String toShortNewick(final boolean printInternalNodeNumbers)\\n     String toSortedNewick(final int[] maxNodeInClade)\\n    public String toSortedNewick(int[] maxNodeInClade, boolean printMetaData)\\n    public String toNewick(final List<String> labels)\\n    public String toNewick(boolean onlyTopology)\\n    public String toNewick()\\n    public String getNewickMetaData()\\n    public String getNewickLengthMetaData()\\n    public String toString(final List<String> labels)\\n    public String toString()\\n    public int sort()\\n    public int labelInternalNodes(int labelIndex)\\n    public Node copy()\\n    public void assignTo(final Node[] nodes)\\n    public void assignFrom(final Node[] nodes, final Node node)\\n    public void setMetaData(final String pattern, final Object value)\\n    public void removeMetaData(final String key)\\n    public void setLengthMetaData(String key, Object value)\\n    public Object getMetaData(final String pattern)\\n    public Object getLengthMetaData(String key)\\n    public Set<String> getMetaDataNames()\\n    public Set<String> getLengthMetaDataNames()\\n    public int scale(final double scale)\\n    protected void startEditing()\\n    public int getChildCount()\\n    public Node getChild(final int childIndex)\\n    public void setChild(final int childIndex, final Node node)\\n    public void setLeft(final Node leftChild)\\n    public Node getLeft()\\n    public void setRight(final Node rightChild)\\n    public Node getRight()\\n    public static Node connect(final Node left, final Node right, final double h)\\n    public boolean isDirectAncestor()\\n    public boolean isFake()\\n    public Node getDirectAncestorChild()\\n    public Node getNonDirectAncestorChild()\\n    public Node getFakeChild()\\n}\\n\"}\n",
      "/**\n",
      "     * @return a list of external (leaf) nodes contained in this tree\n",
      "     */\n",
      "public List<Node> getExternalNodes()\n",
      "**************************************\n",
      "@Override\n",
      "public List<Node> getExternalNodes() {\n",
      "    final ArrayList<Node> externalNodes = new ArrayList<>();\n",
      "    for (int i = 0; i < getNodeCount(); i++) {\n",
      "        final Node node = getNode(i);\n",
      "        if (node.isLeaf())\n",
      "            externalNodes.add(node);\n",
      "    }\n",
      "    return externalNodes;\n",
      "}\n",
      "3203\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 234/1243 [08:18<34:28,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9584\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 235/1243 [08:19<32:38,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9594\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 236/1243 [08:21<34:01,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9597\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 237/1243 [08:24<35:41,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9625\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 238/1243 [08:26<36:04,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9626\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Input.java', 'methodCount': 34, 'lineCount': 44}, {'fieldCount': 17, 'classCount': 1, 'methodsOnly': 'public class Log {\\n    static public PrintStream nullStream = new PrintStream(new OutputStream() {\\n\\n    @Override\\n    public void write(byte[] b, int off, int len) throws IOException {\\n    }\\n\\n    @Override\\n    public void write(int b) throws IOException {\\n    }\\n});;\\n    static Level level = Level.info;;\\n    public static String[] values()\\n    static public PrintStream err;;\\n    public static void err(String msg)\\n    static public PrintStream warning;;\\n    public static void warning(String msg)\\n    static public PrintStream info;;\\n    public static void info(String msg)\\n    static public PrintStream debug;;\\n    public static void debug(String msg)\\n    static public PrintStream trace;;\\n    public static void trace(String msg)\\n    static private PrintStream errIfOpen;;\\n    static private PrintStream warningIfOpen;;\\n    static private PrintStream infoIfOpen;;\\n    static private PrintStream debugIfOpen;;\\n    static private PrintStream traceIfOpen;;\\n    final static public int ERROR = 0;;\\n    final static public int WARNING = 1;;\\n    final static public int INFO = 2;;\\n    final static public int DEBUG = 3;;\\n    final static public int TRACE = 4;;\\n    public static void setLevel(Level newLevel)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Log.java', 'methodCount': 7, 'lineCount': 26}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class TaxonSet {\\n    final public Input<Alignment> alignmentInput = new Input<>(\"alignment\", \"alignment where each sequence represents a taxon\");;\\n    final public Input<List<Taxon>> taxonsetInput = new Input<>(\"taxon\", \"list of taxa making up the set\", new ArrayList<>());;\\n    protected List<String> taxaNames;;\\n    protected List<Taxon> taxonList;;\\n    public TaxonSet();\\n    public TaxonSet(final List<Taxon> taxa);\\n    public TaxonSet(final Alignment alignment);\\n    public TaxonSet(final String id, final List<Taxon> taxa);\\n    public void initAndValidate()\\n    public Set<Taxon> getTaxonSet()\\n    public List<String> asStringList()\\n    public Set<String> getTaxaNames()\\n    public String getTaxonId(int taxonIndex)\\n    public int getTaxonIndex(String id)\\n    public Taxon getTaxon(String id)\\n    public boolean containsAny(final Collection<String> taxa)\\n    public boolean containsAll(final Collection<String> taxa)\\n    public boolean containsAny(final TaxonSet taxa)\\n    public boolean containsAll(final TaxonSet taxa)\\n    public int getTaxonCount()\\n    public int getNrTaxa()\\n    public String toString()\\n    public String toString(String indent)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/evolution/alignment/TaxonSet.java', 'methodCount': 19, 'lineCount': 25}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class BEASTClassLoader {\\n    // maps package name (String) to class loader (MultiParentURLClassLoader)\\n// used to find class loader for a package\\nstatic Map<String, MultiParentURLClassLoader> package2classLoaderMap = new HashMap<>();;\\n    // singleton class loader\\nstatic final public BEASTClassLoader classLoader = new BEASTClassLoader(new URL[0], BEASTClassLoader.class.getClassLoader());;\\n    // maps service (=class name) to service providers (=set of class names)\\nstatic private Map<String, Set<String>> services = new HashMap<>();;\\n    // maps service provider (=class name) to class loader for that provider (=class)\\nstatic private Map<String, ClassLoader> class2loaderMap = new HashMap<>();;\\n    static private Set<String> namespaces = new HashSet<>();;\\n    private BEASTClassLoader(URL[] urls, ClassLoader parent);\\n    public void addURL(URL url)\\n    public void addURL(URL url, String packageName, Map<String, Set<String>> services)\\n    public void addParent(String packageName, String parentPackage)\\n    public void addJar(String jarFile)\\n    public void addJar(String jarFile, String packageName)\\n    public static Class<?> forName(String className) throws ClassNotFoundException\\n    public static Class<?> forName(String className, String service) throws ClassNotFoundException\\n    public static Set<String> loadService(Class<?> service)\\n    public static void initServices()\\n    public static void initServices(String classPath)\\n    public static void addServices(String versionFile)\\n    public static Map<String, Set<String>> getServices()\\n    public void addServices(String packageName, Map<String, Set<String>> services)\\n    private static MultiParentURLClassLoader getClassLoader(String packageName)\\n    public static void addService(String service, String className, String packageName)\\n    public static void delService(Map<String, Set<String>> serviceMap, String packageName)\\n    public static String usesExistingNamespaces(Set<String> services)\\n    public static URL getResource(String packageName, String resourceName)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/pkgmgmt/BEASTClassLoader.java', 'methodCount': 19, 'lineCount': 26}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'Node': \"public class Node {\\n    /**\\n * label nr of node, used mostly when this is a leaf.\\n */\\nprotected int labelNr;;\\n    /**\\n * height of this node.\\n */\\nprotected double height = Double.MAX_VALUE;;\\n    /**\\n * Arbitrarily labeled metadata on this node. Not currently implemented as part of state!\\n */\\nprotected Map<String, Object> metaData = new TreeMap<>();;\\n    /**\\n * Length metadata for the edge above this node.  Not currently implemented as part of state!\\n */\\nprotected Map<String, Object> lengthMetaData = new TreeMap<>();;\\n    /**\\n * list of children of this node *\\n * Don't use m_left and m_right directly\\n * Use getChildCount() and getChild(x) or getChildren() instead\\n */\\nprotected List<Node> children = new ArrayList<>();;\\n    /**\\n * parent node in the beast.tree, null if root *\\n */\\nprotected Node parent = null;;\\n    /**\\n * status of this node after an operation is performed on the state *\\n */\\nprotected int isDirty = Tree.IS_CLEAN;;\\n    /**\\n * meta-data contained in square brackets in Newick *\\n */\\npublic String metaDataString, lengthMetaDataString;;\\n    /**\\n * The Tree that this node is a part of.\\n * This allows e.g. access to the State containing the Tree *\\n */\\nprotected Tree m_tree;;\\n    public Node();\\n    public Node(final String id);\\n    public void setTree(Tree tree)\\n    public Tree getTree()\\n    public void initAndValidate()\\n    public int getNr()\\n    public void setNr(final int labelIndex)\\n    public double getHeight()\\n    public double getDate()\\n    public void setHeight(final double height)\\n    public void setHeightDA(final double height)\\n    public final double getLength()\\n    public int isDirty()\\n    public void makeDirty(final int dirty)\\n    public void makeAllDirty(final int dirty)\\n    public Node getParent()\\n    public void setParent(final Node parent)\\n    public void setParent(final Node parent, final boolean inOperator)\\n     void setParentImmediate(final Node parent)\\n    public List<Node> getChildren()\\n    public List<Node> getChildrenMutable()\\n    public List<Node> getAllChildNodes()\\n    public void getAllChildNodes(final List<Node> childNodes)\\n    public List<Node> getAllChildNodesAndSelf()\\n    public void getAllChildNodesAndSelf(final List<Node> childNodes)\\n    public List<Node> getAllLeafNodes()\\n    public void getAllLeafNodes(final List<Node> leafNodes)\\n    public boolean isRoot()\\n    public boolean isLeaf()\\n    public void removeChild(final Node child)\\n    public void removeAllChildren(final boolean inOperator)\\n    public void addChild(final Node child)\\n    public int getNodeCount()\\n    public int getLeafNodeCount()\\n    public int getInternalNodeCount()\\n    public String toShortNewick(final boolean printInternalNodeNumbers)\\n     String toSortedNewick(final int[] maxNodeInClade)\\n    public String toSortedNewick(int[] maxNodeInClade, boolean printMetaData)\\n    public String toNewick(final List<String> labels)\\n    public String toNewick(boolean onlyTopology)\\n    public String toNewick()\\n    public String getNewickMetaData()\\n    public String getNewickLengthMetaData()\\n    public String toString(final List<String> labels)\\n    public String toString()\\n    public int sort()\\n    public int labelInternalNodes(int labelIndex)\\n    public Node copy()\\n    public void assignTo(final Node[] nodes)\\n    public void assignFrom(final Node[] nodes, final Node node)\\n    public void setMetaData(final String pattern, final Object value)\\n    public void removeMetaData(final String key)\\n    public void setLengthMetaData(String key, Object value)\\n    public Object getMetaData(final String pattern)\\n    public Object getLengthMetaData(String key)\\n    public Set<String> getMetaDataNames()\\n    public Set<String> getLengthMetaDataNames()\\n    public int scale(final double scale)\\n    protected void startEditing()\\n    public int getChildCount()\\n    public Node getChild(final int childIndex)\\n    public void setChild(final int childIndex, final Node node)\\n    public void setLeft(final Node leftChild)\\n    public Node getLeft()\\n    public void setRight(final Node rightChild)\\n    public Node getRight()\\n    public static Node connect(final Node left, final Node right, final double h)\\n    public boolean isDirectAncestor()\\n    public boolean isFake()\\n    public Node getDirectAncestorChild()\\n    public Node getNonDirectAncestorChild()\\n    public Node getFakeChild()\\n}\\n\"}\n",
      "/**\n",
      "     * Adds a node to the end of the node array. nodeCount and leafNodeCount are recalculated.\n",
      "     * Use with care!\n",
      "     */\n",
      "public void addNode(final Node newNode)\n",
      "**************************************\n",
      "public void addNode(final Node newNode) {\n",
      "    final Node[] tmp = new Node[nodeCount + 1];\n",
      "    System.arraycopy(m_nodes, 0, tmp, 0, nodeCount);\n",
      "    tmp[nodeCount] = newNode;\n",
      "    newNode.setNr(nodeCount);\n",
      "    m_nodes = tmp;\n",
      "    nodeCount++;\n",
      "    if (newNode.getChildCount() > 0)\n",
      "        internalNodeCount++;\n",
      "    else\n",
      "        leafNodeCount++;\n",
      "}\n",
      "3238\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 239/1243 [08:28<33:00,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9945\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 240/1243 [08:29<31:57,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9946\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 241/1243 [08:31<30:37,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9947\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  19%|        | 242/1243 [08:33<29:53,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9948\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 243/1243 [08:34<28:51,  1.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9949\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Returns a uniform random permutation of ints 0,...,l-1\n",
      "     *\n",
      "     * @param l length of the array required.\n",
      "     */\n",
      "public int[] permuted(int l)\n",
      "**************************************\n",
      "public int[] permuted(int l) {\n",
      "    int[] array = new int[l];\n",
      "    for (int i = 0; i < l; i++) {\n",
      "        array[i] = i;\n",
      "    }\n",
      "    permute(array);\n",
      "    return array;\n",
      "}\n",
      "131\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 244/1243 [08:36<28:43,  1.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9988\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 245/1243 [08:38<30:39,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9999\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 246/1243 [08:40<32:36,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 247/1243 [08:42<33:26,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10033\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 248/1243 [08:45<35:18,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10039\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * compute covariance\n",
      "     *\n",
      "     * @param x      list of numbers\n",
      "     * @param y      list of numbers\n",
      "     * @param xmean  assumed mean of x\n",
      "     * @param ymean  assumed mean of y\n",
      "     * @param xstdev assumed stdev of x\n",
      "     * @param ystdev assumed stdev of y\n",
      "     * @return covariance of x and y\n",
      "     */\n",
      "public static double covariance(double[] x, double[] y, double xmean, double ymean, double xstdev, double ystdev)\n",
      "**************************************\n",
      "public static double covariance(double[] x, double[] y, double xmean, double ymean, double xstdev, double ystdev) {\n",
      "    if (x.length != y.length)\n",
      "        throw new IllegalArgumentException(\"x and y arrays must be same length!\");\n",
      "    int count = x.length;\n",
      "    double covar = 0.0;\n",
      "    for (int i = 0; i < x.length; i++) {\n",
      "        if (Double.isNaN(x[i]) || Double.isNaN(y[i])) {\n",
      "            count--;\n",
      "        } else {\n",
      "            covar += (x[i] - xmean) * (y[i] - ymean);\n",
      "        }\n",
      "    }\n",
      "    covar /= count;\n",
      "    covar /= (xstdev * ystdev);\n",
      "    return covar;\n",
      "}\n",
      "319\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 249/1243 [08:47<34:10,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10048\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 250/1243 [08:49<33:13,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10068\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 251/1243 [08:51<32:49,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10164\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 252/1243 [08:52<31:43,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10233\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 253/1243 [08:55<33:09,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10242\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class BEASTClassLoader {\\n    // maps package name (String) to class loader (MultiParentURLClassLoader)\\n// used to find class loader for a package\\nstatic Map<String, MultiParentURLClassLoader> package2classLoaderMap = new HashMap<>();;\\n    // singleton class loader\\nstatic final public BEASTClassLoader classLoader = new BEASTClassLoader(new URL[0], BEASTClassLoader.class.getClassLoader());;\\n    // maps service (=class name) to service providers (=set of class names)\\nstatic private Map<String, Set<String>> services = new HashMap<>();;\\n    // maps service provider (=class name) to class loader for that provider (=class)\\nstatic private Map<String, ClassLoader> class2loaderMap = new HashMap<>();;\\n    static private Set<String> namespaces = new HashSet<>();;\\n    private BEASTClassLoader(URL[] urls, ClassLoader parent);\\n    public void addURL(URL url)\\n    public void addURL(URL url, String packageName, Map<String, Set<String>> services)\\n    public void addParent(String packageName, String parentPackage)\\n    public void addJar(String jarFile)\\n    public void addJar(String jarFile, String packageName)\\n    public static Class<?> forName(String className) throws ClassNotFoundException\\n    public static Class<?> forName(String className, String service) throws ClassNotFoundException\\n    public static Set<String> loadService(Class<?> service)\\n    public static void initServices()\\n    public static void initServices(String classPath)\\n    public static void addServices(String versionFile)\\n    public static Map<String, Set<String>> getServices()\\n    public void addServices(String packageName, Map<String, Set<String>> services)\\n    private static MultiParentURLClassLoader getClassLoader(String packageName)\\n    public static void addService(String service, String className, String packageName)\\n    public static void delService(Map<String, Set<String>> serviceMap, String packageName)\\n    public static String usesExistingNamespaces(Set<String> services)\\n    public static URL getResource(String packageName, String resourceName)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/pkgmgmt/BEASTClassLoader.java', 'methodCount': 19, 'lineCount': 26}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 12}\n",
      "{'Input': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n'}\n",
      "/**\n",
      "     * get description of an input\n",
      "     *\n",
      "     * @param name of the input\n",
      "     * @return description of input\n",
      "     */\n",
      " String getTipText(final String name) throws IllegalArgumentException, IllegalAccessException\n",
      "**************************************\n",
      "default String getTipText(final String name) throws IllegalArgumentException, IllegalAccessException {\n",
      "    try {\n",
      "        Input<?> input = getInput(name);\n",
      "        if (input != null) {\n",
      "            return input.getTipText();\n",
      "        }\n",
      "    } catch (Exception e) {\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "1320\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  20%|        | 254/1243 [08:56<31:27,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10247\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 255/1243 [08:58<30:08,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10291\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 256/1243 [09:00<31:56,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10319\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 257/1243 [09:02<33:17,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10429\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 258/1243 [09:05<34:20,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10448\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Input.java', 'methodCount': 34, 'lineCount': 44}, {'fieldCount': 17, 'classCount': 1, 'methodsOnly': 'public class Log {\\n    static public PrintStream nullStream = new PrintStream(new OutputStream() {\\n\\n    @Override\\n    public void write(byte[] b, int off, int len) throws IOException {\\n    }\\n\\n    @Override\\n    public void write(int b) throws IOException {\\n    }\\n});;\\n    static Level level = Level.info;;\\n    public static String[] values()\\n    static public PrintStream err;;\\n    public static void err(String msg)\\n    static public PrintStream warning;;\\n    public static void warning(String msg)\\n    static public PrintStream info;;\\n    public static void info(String msg)\\n    static public PrintStream debug;;\\n    public static void debug(String msg)\\n    static public PrintStream trace;;\\n    public static void trace(String msg)\\n    static private PrintStream errIfOpen;;\\n    static private PrintStream warningIfOpen;;\\n    static private PrintStream infoIfOpen;;\\n    static private PrintStream debugIfOpen;;\\n    static private PrintStream traceIfOpen;;\\n    final static public int ERROR = 0;;\\n    final static public int WARNING = 1;;\\n    final static public int INFO = 2;;\\n    final static public int DEBUG = 3;;\\n    final static public int TRACE = 4;;\\n    public static void setLevel(Level newLevel)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Log.java', 'methodCount': 7, 'lineCount': 26}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class StateNode {\\n    /**\\n * Flag to indicate the StateNode is not constant.\\n * This is particularly useful for Beauti *\\n */\\nfinal public Input<Boolean> isEstimatedInput = new Input<>(\"estimate\", \"whether to estimate this item or keep constant to its initial value\", true);;\\n    public StateNode getCurrent()\\n    public StateNode getCurrentEditable(final Operator operator)\\n    public boolean somethingIsDirty()\\n    public void setSomethingIsDirty(final boolean isDirty)\\n    public abstract void setEverythingDirty(final boolean isDirty)\\n    public abstract StateNode copy()\\n    public abstract void assignTo(StateNode other)\\n    public abstract void assignFrom(StateNode other)\\n    public void assignFromWithoutID(StateNode other)\\n    public abstract void assignFromFragile(StateNode other)\\n    public final void toXML(PrintStream out)\\n    public final String toXML()\\n    private String normalise(String str)\\n    public abstract void fromXML(Node node)\\n    public abstract int scale(double scale)\\n    protected State state = null;;\\n    public State getState()\\n    public void setState(State newState)\\n    protected boolean hasStartedEditing = false;;\\n    public int index = -1;;\\n    public int getIndex()\\n    public void startEditing(final Operator operator)\\n    private boolean isCalledFromOperator(int level)\\n    protected abstract void store()\\n    public final void restoreStateNode()\\n    public abstract void restore()\\n    public boolean isEstimated()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/inference/StateNode.java', 'methodCount': 24, 'lineCount': 30}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Parameter {\\n     T getValue(int i)\\n     T getValue()\\n     void setValue(int i, T value)\\n     void setValue(T value)\\n     T getLower()\\n     void setLower(final T lower)\\n     T getUpper()\\n     void setUpper(final T upper)\\n     T[] getValues()\\n     String getID()\\n     int getMinorDimension1()\\n     int getMinorDimension2()\\n     int getColumnCount()\\n     int getRowCount()\\n     T getMatrixValue(int row, int column)\\n     T[] getRowValues(String key)\\n     T[] getRowValues(int row)\\n     String getKey(int i)\\n     T getValue(String key)\\n     List<String> getKeysList()\\n     void swap(int i, int j)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/inference/parameter/Parameter.java', 'methodCount': 21, 'lineCount': 23}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class RealParameter {\\n    final public Input<Double> lowerValueInput = new Input<>(\"lower\", \"lower value for this parameter (default -infinity)\");;\\n    final public Input<Double> upperValueInput = new Input<>(\"upper\", \"upper value for this parameter (default +infinity)\");;\\n    public RealParameter();\\n    public RealParameter(final Double[] values);\\n    public RealParameter(final String value);\\n    public void initAndValidate()\\n     Double getMax()\\n     Double getMin()\\n    public Double getValue()\\n    public double getArrayValue()\\n    public double getArrayValue(final int index)\\n    public void log(final long sample, final PrintStream out)\\n    public int scale(final double scale)\\n     void fromXML(final int dimension, final String lower, final String upper, final String[] valuesString)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/inference/parameter/RealParameter.java', 'methodCount': 12, 'lineCount': 16}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class Randomizer {\\n    private Randomizer();\\n    /**\\n * A random number generator that is initialized with the clock when this\\n * class is loaded into the JVM. Use this for all random numbers.\\n * Note: This method or getting random numbers in not thread-safe. Since\\n * MersenneTwisterFast is currently (as of 9/01) not synchronized using\\n * this function may cause concurrency issues. Use the static get methods of the\\n * MersenneTwisterFast class for access to a single instance of the class, that\\n * has synchronization.\\n */\\n//private static final MersenneTwisterFast random = MersenneTwisterFast.DEFAULT_INSTANCE;\\nfinal private static MersenneTwisterFast random = MersenneTwisterFast.DEFAULT_INSTANCE;;\\n    public static int randomChoice(double[] cf)\\n    public static int binarySearchSampling(double[] cpd)\\n    public static int randomChoicePDF(double[] pdf)\\n    public static double[] getNormalized(double[] array)\\n    public static double getTotal(double[] array, int start, int end)\\n    public static double getTotal(double[] array)\\n    public static long getSeed()\\n    public static void setSeed(long seed)\\n    public static byte nextByte()\\n    public static boolean nextBoolean()\\n    public static void nextBytes(byte[] bs)\\n    public static char nextChar()\\n    public static double nextGaussian()\\n    public static double nextGamma(double alpha, double lambda)\\n    public static long nextPoisson(double lambda)\\n    public static double nextDouble()\\n    public static double randomLogDouble()\\n    public static double nextExponential(double lambda)\\n    public static long nextGeometric(double p)\\n    public static float nextFloat()\\n    public static long nextLong()\\n    public static short nextShort()\\n    public static int nextInt()\\n    public static int nextInt(int n)\\n    public static double nextLogNormal(double M, double S, boolean meanInRealspace)\\n    public static double uniform(double low, double high)\\n    public static void shuffle(int[] array)\\n    public static void shuffle(int[] array, int numberOfShuffles)\\n    public static int[] shuffled(int l)\\n    public static int[] sampleIndicesWithReplacement(int l)\\n    public static void permute(int[] array)\\n    public static int[] permuted(int l)\\n    static int m_nIDNr = 0;;\\n    public static int nextIDNr()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/util/Randomizer.java', 'methodCount': 34, 'lineCount': 38}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * automatic parameter tuning *\n",
      "     */\n",
      "public void optimize(final double logAlpha)\n",
      "**************************************\n",
      "@Override\n",
      "public void optimize(final double logAlpha) {\n",
      "    if (optimiseInput.get()) {\n",
      "        double delta = calcDelta(logAlpha);\n",
      "        double scaleFactor = getCoercableParameterValue();\n",
      "        delta += Math.log(scaleFactor);\n",
      "        scaleFactor = Math.exp(delta);\n",
      "        setCoercableParameterValue(scaleFactor);\n",
      "    }\n",
      "}\n",
      "2745\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 259/1243 [09:07<35:22,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10513\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 260/1243 [09:09<35:38,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10556\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 261/1243 [09:11<35:52,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10575\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 262/1243 [09:14<36:26,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10580\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 263/1243 [09:16<36:25,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10625\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class JSONException {\\n    private static final long serialVersionUID = 1L;;\\n    public JSONException(String s);\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/json/JSONException.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class JSONObject {\\n    private static final Double NEGATIVE_ZERO = -0d;;\\n    private Object parent = null;;\\n    /**\\n * A sentinel value used to explicitly define a name with no value. Unlike\\n * {@code null}, names with this value:\\n * <ul>\\n *   <li>show up in the {@link #names} array\\n *   <li>show up in the {@link #keys} iterator\\n *   <li>return {@code true} for {@link #has(String)}\\n *   <li>do not throw on {@link #get(String)}\\n *   <li>are included in the encoded JSON string.\\n * </ul>\\n *\\n * <p>This value violates the general contract of {@link Object#equals} by\\n * returning true when compared to {@code null}. Its {@link #toString}\\n * method returns \"null\".\\n */\\npublic static final Object NULL = new Object() {\\n\\n    @Override\\n    public boolean equals(Object o) {\\n        // API specifies this broken equals implementation\\n        return o == this || o == null;\\n    }\\n\\n    @Override\\n    public String toString() {\\n        return \"null\";\\n    }\\n};;\\n    private final LinkedHashMap<String, Object> nameValuePairs;;\\n    public JSONObject();\\n    public JSONObject(Map copyFrom);\\n    public JSONObject(JSONTokener readFrom) throws JSONException;\\n    public JSONObject(String json) throws JSONException;\\n    public JSONObject(JSONObject copyFrom, String[] names) throws JSONException;\\n    public int length()\\n    public JSONObject put(String name, boolean value) throws JSONException\\n    public JSONObject put(String name, double value) throws JSONException\\n    public JSONObject put(String name, int value) throws JSONException\\n    public JSONObject put(String name, long value) throws JSONException\\n    public JSONObject put(String name, Object value) throws JSONException\\n    public JSONObject putOpt(String name, Object value) throws JSONException\\n    public JSONObject accumulate(String name, Object value) throws JSONException\\n    public JSONObject append(String name, Object value) throws JSONException\\n     String checkName(String name) throws JSONException\\n    public Object remove(String name)\\n    public boolean isNull(String name)\\n    public boolean has(String name)\\n    public Object get(String name) throws JSONException\\n    public Object opt(String name)\\n    public boolean getBoolean(String name) throws JSONException\\n    public boolean optBoolean(String name)\\n    public boolean optBoolean(String name, boolean fallback)\\n    public double getDouble(String name) throws JSONException\\n    public double optDouble(String name)\\n    public double optDouble(String name, double fallback)\\n    public int getInt(String name) throws JSONException\\n    public int optInt(String name)\\n    public int optInt(String name, int fallback)\\n    public long getLong(String name) throws JSONException\\n    public long optLong(String name)\\n    public long optLong(String name, long fallback)\\n    public String getString(String name) throws JSONException\\n    public String optString(String name)\\n    public String optString(String name, String fallback)\\n    public JSONArray getJSONArray(String name) throws JSONException\\n    public JSONArray optJSONArray(String name)\\n    public JSONObject getJSONObject(String name) throws JSONException\\n    public JSONObject optJSONObject(String name)\\n    public JSONArray toJSONArray(JSONArray names) throws JSONException\\n    public Iterator<String> keys()\\n    public Set<String> keySet()\\n    public JSONArray names()\\n    public String toString()\\n    public String toString(int indentSpaces) throws JSONException\\n     void writeTo(JSONStringer stringer) throws JSONException\\n    public static String numberToString(Number number) throws JSONException\\n    public static String quote(String data)\\n    public static Object wrap(Object o)\\n    public Object getParent()\\n    public void setParent(Object parent)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/json/JSONObject.java', 'methodCount': 51, 'lineCount': 57}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': \"public class JSONStringer {\\n    /**\\n * The output data, containing at most one top-level array or object.\\n */\\nfinal StringBuilder out = new StringBuilder();;\\n    /**\\n * Unlike the original implementation, this stack isn't limited to 20\\n * levels of nesting.\\n */\\nprivate final List<Scope> stack = new ArrayList<Scope>();;\\n    /**\\n * A string containing a full set of spaces for a single level of\\n * indentation, or null for no pretty printing.\\n */\\nprivate final String indent;;\\n    public JSONStringer();\\n     JSONStringer(int indentSpaces);\\n    public JSONStringer array() throws JSONException\\n    public JSONStringer endArray() throws JSONException\\n    public JSONStringer object() throws JSONException\\n    public JSONStringer endObject() throws JSONException\\n     JSONStringer open(Scope empty, String openBracket) throws JSONException\\n     JSONStringer close(Scope empty, Scope nonempty, String closeBracket) throws JSONException\\n    private Scope peek() throws JSONException\\n    private void replaceTop(Scope topOfStack)\\n    public JSONStringer value(Object value) throws JSONException\\n    public JSONStringer value(boolean value) throws JSONException\\n    public JSONStringer value(double value) throws JSONException\\n    public JSONStringer value(long value) throws JSONException\\n    private void string(String value)\\n    private void newline()\\n    public JSONStringer key(String name) throws JSONException\\n    private void beforeKey() throws JSONException\\n    private void beforeValue() throws JSONException\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/org/json/JSONStringer.java', 'methodCount': 20, 'lineCount': 25}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public interface BEASTInterface {\\n    final static String DEFEAULT_DESCRIPTION = \"Not documented!!!\";;\\n    public void initAndValidate()\\n    public String getID()\\n    public void setID(String ID)\\n    public Set<BEASTInterface> getOutputs()\\n    public Map<String, Input<?>> getInputs()\\n    public void init(final Object... objects)\\n    public void initByName(final Object... objects)\\n     static Set<BEASTInterface> getOutputs(Object object)\\n     String getDescription()\\n     Citation getCitation()\\n     List<Citation> getCitationList()\\n     String getCitations()\\n     String getCitations(final HashSet<String> citations, final HashSet<String> IDs)\\n     List<Input<?>> listInputs()\\n     void listAnnotatedInputs(Object o, List<Input<?>> inputs, Map<String, Input> inputNames)\\n     List<BEASTInterface> listActiveBEASTObjects()\\n     List<BEASTInterface> listActivePlugins() throws IllegalArgumentException, IllegalAccessException\\n     String getTipText(final String name) throws IllegalArgumentException, IllegalAccessException\\n     boolean isPrimitive(final String name)\\n     Object getInputValue(final String name)\\n     void setInputValue(final String name, final Object value)\\n     Input<?> getInput(final String name)\\n     void validateInputs()\\n     void getPredecessors(final List<BEASTInterface> predecessors)\\n     void determindClassOfInputs()\\n     boolean notCloneable()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/BEASTInterface.java', 'methodCount': 26, 'lineCount': 29}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class BEASTObject {\\n    /**\\n * set of Objects that have this Object in one of its Inputs *\\n * @deprecate use getOuputs() or BEASTObject.getOuputs(object) instead\\n */\\n@Deprecated\\npublic Set<BEASTInterface> outputs = new HashSet<>();;\\n    /**\\n * cache collecting all Inputs and InputForAnnotatedConstrutors\\n * indexed through input name\\n */\\nprivate Map<String, Input<?>> inputcache;;\\n    public Set<BEASTInterface> getOutputs()\\n    public Map<String, Input<?>> getInputs()\\n    protected String ID;;\\n    public String getID()\\n    public void setID(final String ID)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/BEASTObject.java', 'methodCount': 5, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Description.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class Input {\\n    /**\\n * input name, used for identification when getting/setting values of a plug-in *\\n */\\nString name = \"\";;\\n    /**\\n * short description of the function of this particular input *\\n */\\nString tipText = \"\";;\\n    /**\\n * value represented by this input\\n */\\nT value;;\\n    /**\\n * Type of T, automatically determined when setting a new value.\\n * Used for type checking.\\n */\\nprotected Class<?> theClass;;\\n    // (Q2R) I am surprised the default is not required ....\\nValidate rule = Validate.OPTIONAL;;\\n    /**\\n * used only if validation rule is XOR *\\n */\\nInput<?> other;;\\n    public T defaultValue;;\\n    /**\\n * Possible values for enumerations, e.g. if\\n * an input can be any of \"constant\", \"linear\", \"quadratic\"\\n * this array contains these values. Used for validation and user interfaces.\\n */\\npublic T[] possibleValues;;\\n    public Input();\\n    public Input(String name, String tipText);\\n    public Input(String name, String tipText, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue);\\n    public Input(String name, String tipText, T startValue, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, T startValue, Validate rule, Input<?> other, Class<?> theClass);\\n    public Input(String name, String tipText, T startValue, Validate rule);\\n    public Input(String name, String tipText, T startValue, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule);\\n    public Input(String name, String tipText, Validate rule, Class<?> type);\\n    public Input(String name, String tipText, Validate rule, Input<?> other);\\n    public Input(String name, String tipText, Validate rule, Input<?> other, Class<?> type);\\n    public Input(String name, String tipText, T startValue, T[] possibleValues);\\n    private void checkName()\\n    public String getName()\\n    public String getTipText()\\n    public String getHTMLTipText()\\n    public String getValueTipText()\\n    public Class<?> getType()\\n    public void setType(Class<?> theClass)\\n    public Validate getRule()\\n    public void setRule(final Validate rule)\\n    public final Input<?> getOther()\\n    public T get()\\n    public void setValue(final Object value, final BEASTInterface beastObject)\\n    public void set(final Object value)\\n    public boolean canSetValue(Object value, BEASTInterface beastObject)\\n    public void determineClass(final Object beastObject)\\n    private void setStringValue(final String stringValue, final BEASTInterface beastObject)\\n    private static void setStringValue(List list, String stringValue, Class<?> theClass)\\n    public void validate()\\n    public String toString()\\n    public static Object fromString(Object arg, Class<?> type)\\n}\\n', 'filePath': '/home/ubuntu/repos/CompEvol,beast2/src/beast/base/core/Input.java', 'methodCount': 34, 'lineCount': 44}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'StateNode': 'public class StateNode {\\n    /**\\n * Flag to indicate the StateNode is not constant.\\n * This is particularly useful for Beauti *\\n */\\nfinal public Input<Boolean> isEstimatedInput = new Input<>(\"estimate\", \"whether to estimate this item or keep constant to its initial value\", true);;\\n    public StateNode getCurrent()\\n    public StateNode getCurrentEditable(final Operator operator)\\n    public boolean somethingIsDirty()\\n    public void setSomethingIsDirty(final boolean isDirty)\\n    public abstract void setEverythingDirty(final boolean isDirty)\\n    public abstract StateNode copy()\\n    public abstract void assignTo(StateNode other)\\n    public abstract void assignFrom(StateNode other)\\n    public void assignFromWithoutID(StateNode other)\\n    public abstract void assignFromFragile(StateNode other)\\n    public final void toXML(PrintStream out)\\n    public final String toXML()\\n    private String normalise(String str)\\n    public abstract void fromXML(Node node)\\n    public abstract int scale(double scale)\\n    protected State state = null;;\\n    public State getState()\\n    public void setState(State newState)\\n    protected boolean hasStartedEditing = false;;\\n    public int index = -1;;\\n    public int getIndex()\\n    public void startEditing(final Operator operator)\\n    private boolean isCalledFromOperator(int level)\\n    protected abstract void store()\\n    public final void restoreStateNode()\\n    public abstract void restore()\\n    public boolean isEstimated()\\n}\\n'}\n",
      "/**\n",
      "     * return list of state nodes that this operator operates on.\n",
      "     * state nodes that are input to the operator but are never changed\n",
      "     * in a proposal should not be listed\n",
      "     */\n",
      "public List<StateNode> listStateNodes()\n",
      "**************************************\n",
      "public List<StateNode> listStateNodes() {\n",
      "    final List<StateNode> list = new ArrayList<>();\n",
      "    for (BEASTInterface o : listActiveBEASTObjects()) {\n",
      "        if (o instanceof StateNode) {\n",
      "            final StateNode stateNode = (StateNode) o;\n",
      "            if (stateNode.isEstimatedInput.get()) {\n",
      "                list.add(stateNode);\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return list;\n",
      "}\n",
      "3238\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|        | 264/1243 [09:18<35:42,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10679\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|       | 265/1243 [09:20<35:30,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10812\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|       | 266/1243 [09:22<35:41,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10827\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  21%|       | 267/1243 [09:25<35:50,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10832\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 268/1243 [09:26<33:12,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10834\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "\t * @param components double[][]  components obtained from constructor methods.\n",
      "\t */\n",
      "private void initialize(double[][] components)\n",
      "**************************************\n",
      "private void initialize(double[][] components) {\n",
      "    int n = components.length;\n",
      "    rows = new double[n][n];\n",
      "    for (int i = 0; i < n; i++) {\n",
      "        for (int j = 0; j < n; j++) rows[i][j] = components[i][j];\n",
      "    }\n",
      "    initialize();\n",
      "}\n",
      "133\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 269/1243 [09:28<31:31,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10836\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 270/1243 [09:30<29:50,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10837\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 271/1243 [09:32<31:04,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10846\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 272/1243 [09:33<29:32,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10851\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 273/1243 [09:35<28:18,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10854\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "\t * Finds the position of the largest pivot at step p.\n",
      "\t *\n",
      "\t * @param p int\tstep of pivoting.\n",
      "\t * @return int\n",
      "\t */\n",
      "private int largestPivot(int p)\n",
      "**************************************\n",
      "private int largestPivot(int p) {\n",
      "    double pivot = Math.abs(rows[p][p]);\n",
      "    int answer = p;\n",
      "    double x;\n",
      "    for (int i = p + 1; i < rows.length; i++) {\n",
      "        x = Math.abs(rows[i][p]);\n",
      "        if (x > pivot) {\n",
      "            answer = i;\n",
      "            pivot = x;\n",
      "        }\n",
      "    }\n",
      "    return answer;\n",
      "}\n",
      "166\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 274/1243 [09:36<27:25,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10860\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 275/1243 [09:38<27:34,  1.71s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10870\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 276/1243 [09:40<27:11,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10882\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 277/1243 [09:41<24:24,  1.52s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10891\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 278/1243 [09:42<24:17,  1.51s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10912\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'Matrix': 'public class Matrix {\\n    protected double[][] components;;\\n    protected LUPDecomposition lupDecomposition = null;;\\n    public Matrix(double[][] a);\\n    public Matrix(double[] a, int n, int m);\\n    public Matrix(int n, int m) throws NegativeArraySizeException;\\n    public void accumulate(Matrix a) throws IllegalDimension\\n    public Matrix add(Matrix a) throws IllegalDimension\\n    protected double[][] addComponents(Matrix a)\\n    public void clear()\\n    public int columns()\\n    public double component(int n, int m)\\n    public double determinant() throws IllegalDimension\\n    public boolean isPD() throws IllegalDimension\\n    public boolean equals(Matrix a)\\n    public Matrix inverse() throws ArithmeticException\\n    public boolean isSquare()\\n    protected LUPDecomposition lupDecomposition() throws IllegalDimension\\n    public Matrix product(double a)\\n    public Vector product(Vector v) throws IllegalDimension\\n    public Matrix product(Matrix a) throws IllegalDimension\\n    protected double[][] productComponents(double a)\\n    protected double[][] productComponents(Matrix a)\\n    public Matrix productWithTransposed(Matrix a) throws IllegalDimension\\n    protected double[][] productWithTransposedComponents(Matrix a)\\n    public int rows()\\n    protected Vector secureProduct(Vector v)\\n    protected Matrix secureProduct(Matrix a)\\n    protected Matrix secureSubtract(Matrix a)\\n    public Matrix subtract(Matrix a) throws IllegalDimension\\n    protected double[][] subtractComponents(Matrix a)\\n    public double[][] toComponents()\\n    public String toString()\\n    public String toStringOctave()\\n    public Matrix transpose()\\n    public SymmetricMatrix transposedProduct()\\n    public Matrix transposedProduct(Matrix a) throws IllegalDimension\\n    protected double[][] transposedProductComponents(Matrix a)\\n}\\n', 'IllegalDimension': 'public class IllegalDimension {\\n    /**\\n */\\nprivate static final long serialVersionUID = 8101918570370620261L;;\\n    public IllegalDimension();\\n    public IllegalDimension(String s);\\n}\\n'}\n",
      "/**\n",
      "\t * @param a MatrixAlgebra.Matrix\n",
      "\t * @throws dr.math.matrixAlgebra.IllegalDimension\n",
      "\t *          if the supplied matrix\n",
      "\t *          does not have the same dimensions.\n",
      "\t */\n",
      "public void accumulate(Matrix a) throws IllegalDimension\n",
      "**************************************\n",
      "/**\n",
      " * @param a MatrixAlgebra.Matrix\n",
      " * @throws dr.math.matrixAlgebra.IllegalDimension\n",
      " *          if the supplied matrix\n",
      " *          does not have the same dimensions.\n",
      " */\n",
      "public void accumulate(Matrix a) throws IllegalDimension {\n",
      "    if (a.rows() != rows() || a.columns() != columns())\n",
      "        throw new IllegalDimension(\"Operation error: cannot add a\" + a.rows() + \" by \" + a.columns() + \" matrix to a \" + rows() + \" by \" + columns() + \" matrix\");\n",
      "    int m = components[0].length;\n",
      "    for (int i = 0; i < components.length; i++) {\n",
      "        for (int j = 0; j < m; j++) components[i][j] += a.component(i, j);\n",
      "    }\n",
      "}\n",
      "694\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  22%|       | 279/1243 [09:44<24:45,  1.54s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10927\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 280/1243 [09:46<25:05,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10936\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 281/1243 [09:47<25:25,  1.59s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10937\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 282/1243 [09:49<25:30,  1.59s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10940\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 283/1243 [09:51<25:39,  1.60s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10943\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'Matrix': 'public class Matrix {\\n    protected double[][] components;;\\n    protected LUPDecomposition lupDecomposition = null;;\\n    public Matrix(double[][] a);\\n    public Matrix(double[] a, int n, int m);\\n    public Matrix(int n, int m) throws NegativeArraySizeException;\\n    public void accumulate(Matrix a) throws IllegalDimension\\n    public Matrix add(Matrix a) throws IllegalDimension\\n    protected double[][] addComponents(Matrix a)\\n    public void clear()\\n    public int columns()\\n    public double component(int n, int m)\\n    public double determinant() throws IllegalDimension\\n    public boolean isPD() throws IllegalDimension\\n    public boolean equals(Matrix a)\\n    public Matrix inverse() throws ArithmeticException\\n    public boolean isSquare()\\n    protected LUPDecomposition lupDecomposition() throws IllegalDimension\\n    public Matrix product(double a)\\n    public Vector product(Vector v) throws IllegalDimension\\n    public Matrix product(Matrix a) throws IllegalDimension\\n    protected double[][] productComponents(double a)\\n    protected double[][] productComponents(Matrix a)\\n    public Matrix productWithTransposed(Matrix a) throws IllegalDimension\\n    protected double[][] productWithTransposedComponents(Matrix a)\\n    public int rows()\\n    protected Vector secureProduct(Vector v)\\n    protected Matrix secureProduct(Matrix a)\\n    protected Matrix secureSubtract(Matrix a)\\n    public Matrix subtract(Matrix a) throws IllegalDimension\\n    protected double[][] subtractComponents(Matrix a)\\n    public double[][] toComponents()\\n    public String toString()\\n    public String toStringOctave()\\n    public Matrix transpose()\\n    public SymmetricMatrix transposedProduct()\\n    public Matrix transposedProduct(Matrix a) throws IllegalDimension\\n    protected double[][] transposedProductComponents(Matrix a)\\n}\\n'}\n",
      "/**\n",
      "\t * @param a MatrixAlgebra.Matrix\n",
      "\t * @return double[][]\tthe components of the product of the\n",
      "\t *         transpose of the receiver\n",
      "\t *         with the supplied matrix.\n",
      "\t */\n",
      "protected double[][] transposedProductComponents(Matrix a)\n",
      "**************************************\n",
      "protected double[][] transposedProductComponents(Matrix a) {\n",
      "    int p = this.rows();\n",
      "    int n = this.columns();\n",
      "    int m = a.columns();\n",
      "    double[][] newComponents = new double[n][m];\n",
      "    for (int i = 0; i < n; i++) {\n",
      "        for (int j = 0; j < m; j++) {\n",
      "            double sum = 0;\n",
      "            for (int k = 0; k < p; k++) sum += components[k][i] * a.components[k][j];\n",
      "            newComponents[i][j] = sum;\n",
      "        }\n",
      "    }\n",
      "    return newComponents;\n",
      "}\n",
      "604\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 284/1243 [09:52<25:50,  1.62s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10990\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 285/1243 [09:56<35:05,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10997\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 286/1243 [09:57<32:58,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10998\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 287/1243 [09:59<30:40,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11025\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 288/1243 [10:01<28:43,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11097\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * @return true if projectURL is valid\n",
      "     */\n",
      "public boolean isValidFormat()\n",
      "**************************************\n",
      "public boolean isValidFormat() {\n",
      "    try {\n",
      "        getProjectURL().toURI();\n",
      "        return true;\n",
      "    } catch (Exception e) {\n",
      "        return false;\n",
      "    }\n",
      "}\n",
      "92\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 289/1243 [10:02<27:28,  1.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11125\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 290/1243 [10:04<27:02,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11132\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 291/1243 [10:05<26:50,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11698\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  23%|       | 292/1243 [10:07<28:22,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11895\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 293/1243 [10:10<29:33,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12169\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "\t *  Returns total length of deletions and insertions for the input read. \n",
      "\t */\n",
      "public static int getNumIndelBases(SAMRecord read)\n",
      "**************************************\n",
      "public static int getNumIndelBases(SAMRecord read) {\n",
      "    int numIndelBases = 0;\n",
      "    for (CigarElement element : read.getCigar().getCigarElements()) {\n",
      "        if ((element.getOperator() == CigarOperator.D) || (element.getOperator() == CigarOperator.I)) {\n",
      "            numIndelBases += element.getLength();\n",
      "        }\n",
      "    }\n",
      "    return numIndelBases;\n",
      "}\n",
      "164\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 294/1243 [10:11<28:23,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12171\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 295/1243 [10:13<29:19,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12172\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 296/1243 [10:15<29:54,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12174\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 297/1243 [10:17<30:52,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12230\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 298/1243 [10:19<30:56,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12231\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "\t * Convert from phred64 to phred33\n",
      "\t */\n",
      "public String phred64ToPhred33(String phred64)\n",
      "**************************************\n",
      "public String phred64ToPhred33(String phred64) {\n",
      "    StringBuffer phred33 = new StringBuffer();\n",
      "    for (int i = 0; i < phred64.length(); i++) {\n",
      "        phred33.append((char) (phred64.charAt(i) - PHRED33_TO_PHRED64_DIFF));\n",
      "    }\n",
      "    return phred33.toString();\n",
      "}\n",
      "144\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 299/1243 [10:21<30:18,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 300/1243 [10:23<32:34,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13064\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 301/1243 [10:26<32:25,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13193\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 302/1243 [10:27<30:10,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13578\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 303/1243 [10:29<31:00,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13701\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class DoubleMatrixDataset {\\n    private static final org.apache.logging.log4j.Logger LOGGER = org.apache.logging.log4j.LogManager.getLogger(DoubleMatrixDataset.class);;\\n    protected DoubleMatrix2D matrix;;\\n    protected LinkedHashMap<R, Integer> hashRows;;\\n    protected LinkedHashMap<C, Integer> hashCols;;\\n    public DoubleMatrixDataset();\\n    public DoubleMatrixDataset(int rows, int columns);\\n    public DoubleMatrixDataset(LinkedHashMap<R, Integer> hashRows, LinkedHashMap<C, Integer> hashCols);\\n    public DoubleMatrixDataset(DoubleMatrix2D matrix, LinkedHashMap<R, Integer> hashRows, LinkedHashMap<C, Integer> hashCols);\\n    public DoubleMatrixDataset(DoubleMatrix2D matrix);\\n    public DoubleMatrixDataset(Collection<R> rowNames, Collection<C> colNames);\\n    public DoubleMatrixDataset(double[][] matrix, R[] rowNames, C[] colNames) throws Exception;\\n    public DoubleMatrixDataset(double[][] matrix, List<R> rowNames, List<C> colNames) throws Exception;\\n    public static DoubleMatrixDataset<String, String> loadDoubleData(String fileName) throws IOException, Exception\\n    public static List<String> readDoubleTextDataColNames(String fileName, char delimiter) throws IOException\\n    public static List<String> readDoubleTextDataRowNames(String fileName, char delimiter) throws IOException\\n    public static DoubleMatrixDataset<String, String> loadDoubleTextData(String fileName, char delimiter) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadSubsetOfRowsBinaryDoubleData(String fileName, String[] rowsToView) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadSubsetOfRowsBinaryDoubleData(String fileName, Collection<String> rowsToView) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadSubsetOfRowsBinaryDoubleData(String fileName, LinkedHashSet<String> rowsToView) throws IOException\\n    public static File unzipToTMP(File binaryGZipFile) throws IOException\\n    protected static DoubleMatrixDataset<String, String> loadSubsetOfRowsBinaryDoubleData(String fileName, LinkedHashSet<String> rowsToView, LinkedHashMap<String, Integer> originalRowMap, LinkedHashMap<String, Integer> originalColMap) throws IOException\\n    public static DoubleMatrixDataset<String, String> loadSubsetOfBinaryDoubleData(String fileName, Set<String> desiredRows, Set<String> desiredCols) throws IOException\\n    public static DoubleMatrixDataset<String, String> loadSubsetOfTextDoubleData(String fileName, char delimiter, Set<String> desiredRows, Set<String> desiredCols) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadSubsetOfTextDoubleData(String fileName, char delimiter, Set<String> desiredRows, Set<String> desiredCols, int linesToSkip) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadDoubleTextDoubleDataExlcudeCols(String fileName, char delimiter, HashSet<String> colsToExclude) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadDoubleTextDoubleDataExlcudeCols(String fileName, char delimiter, HashSet<String> colsToExclude, int linesToSkip) throws IOException, Exception\\n    public static DoubleMatrixDataset<String, String> loadTransEqtlExpressionMatrix(String fileName) throws FileNotFoundException, IOException\\n    public static DoubleMatrixDataset<String, String> loadDoubleBinaryData(String fileName) throws FileNotFoundException, IOException\\n    public void saveBinary(final String path) throws IOException\\n    public void saveBinary(final String path, final String datasetName, final String dataOnRows, final String dataOnCols) throws IOException\\n    public void saveBinaryOldFormat(String path) throws IOException\\n    protected static LinkedHashMap<String, Integer> loadIdentifiers(String filename) throws IOException\\n    protected static LinkedHashMap<String, Integer> loadIdentifiers(File file) throws IOException\\n    public void save(File file, String rowDescriptor) throws IOException\\n    public void save(String fileName) throws IOException\\n    public void save(File fileName) throws IOException\\n    public void save(String fileName, String rowDescriptor) throws IOException\\n    public void saveDice(String fileName) throws IOException\\n    public void saveDice(File fileName) throws IOException\\n    public void saveDice(String fileName, String rowDescriptor) throws IOException\\n    public void saveDice(File fileName, String rowDescriptor) throws IOException\\n    protected static byte[] intToByteArray(int value)\\n    protected static int byteArrayToInt(byte[] b)\\n    public int rows()\\n    public int columns()\\n    public LinkedHashMap<R, Integer> getHashRows()\\n    public void setHashRows(LinkedHashMap<R, Integer> hashRows)\\n    public LinkedHashMap<C, Integer> getHashCols()\\n    public void setHashCols(LinkedHashMap<C, Integer> hashCols)\\n    public ArrayList<R> getRowObjects()\\n    public LinkedHashMap<C, Integer> getHashColsCopy()\\n    public LinkedHashMap<R, Integer> getHashRowsCopy()\\n    public void setRowObjects(List<R> arrayList) throws Exception\\n    public ArrayList<C> getColObjects()\\n    public void setColObjects(List<C> arrayList) throws Exception\\n    public DoubleMatrix2D getMatrix()\\n    public void setMatrix(DoubleMatrix2D matrix)\\n    public void setMatrix(double[][] matrix)\\n    public double[][] getMatrixAs2dDoubleArray()\\n    public void orderOnColumnnames()\\n    public void orderOnRownames()\\n    public void reorderRows(LinkedHashMap<R, Integer> mappingIndex)\\n    public void reorderCols(LinkedHashMap<C, Integer> mappingIndex)\\n    public DoubleMatrixDataset<C, R> viewDice()\\n    private boolean compareHashCols(LinkedHashMap<C, Integer> mappingIndex, LinkedHashMap<C, Integer> originalHash)\\n    private boolean compareHashRows(LinkedHashMap<R, Integer> mappingIndex, LinkedHashMap<R, Integer> originalHash)\\n    public void setElement(R rowName, C columnName, double value)\\n    public double getElement(R rowName, C columnName)\\n    public DoubleMatrix1D getRow(R rowName)\\n    public DoubleMatrix1D getRow(int row)\\n    public DoubleMatrix1D getCol(C colName)\\n    public DoubleMatrix1D getCol(int col)\\n    public double getElement(int row, int column)\\n    public double getElementQuick(int row, int column)\\n    public void setElementQuick(int row, int column, double value)\\n    public boolean containsRow(R rowId)\\n    public boolean containsCol(C colId)\\n    public DoubleMatrixDataset<R, C> viewSelection(R[] rowsToView, C... colsToView)\\n    public DoubleMatrixDataset<R, C> viewSelection(Collection<R> rowsToView, Collection<C> colsToView)\\n    public DoubleMatrixDataset<R, C> viewSelection(LinkedHashSet<R> rowsToView, LinkedHashSet<C> colsToView)\\n    public DoubleMatrixDataset<R, C> viewRowSelection(LinkedHashSet<R> rowsToView)\\n    public DoubleMatrixDataset<R, C> viewRowSelection(R[] rowsToView)\\n    public DoubleMatrixDataset<R, C> viewRowSelection(Collection<R> rowsToView)\\n    public DoubleMatrix2D viewRowSelectionMatrix(Collection<R> rowsToView)\\n    public DoubleMatrixDataset<R, C> viewColSelection(C... colsToView)\\n    public DoubleMatrixDataset<R, C> viewColSelection(Collection<C> colsToView)\\n    public DoubleMatrixDataset<R, C> viewColSelection(LinkedHashSet<C> colsToView)\\n    public DoubleMatrix1D viewRow(R row)\\n    public DoubleMatrix1D viewRow(int row)\\n    public DoubleMatrix1D viewCol(C col)\\n    public DoubleMatrix1D viewCol(int col)\\n    public DoubleMatrix2D viewColAsMmatrix(int col)\\n    public DoubleMatrix2D viewRowAsMmatrix(int row)\\n    public DoubleMatrixDataset<C, C> calculateCorrelationMatrix()\\n    public DoubleMatrixDataset<C, C> calculateCorrelationMatrixOnNormalizedColumns()\\n    public DoubleMatrixDataset<C, C> calculateCovarianceMatrix()\\n    public int getRowIndex(R row)\\n    public int getColIndex(C col)\\n    public DoubleMatrixDataset<R, C> createRowForceNormalDuplicate()\\n    public DoubleMatrixDataset<R, C> createColumnForceNormalDuplicate()\\n    public void createColumnForceNormalInplace()\\n    public void normalizeRows()\\n    public void normalizeColumns()\\n    public void centerColumns()\\n    public void printMatrix()\\n    public void printSummary()\\n    public static DoubleMatrixDataset<String, String> correlateColumnsOf2ColumnNormalizedDatasets(DoubleMatrixDataset<String, String> d1, DoubleMatrixDataset<String, String> d2) throws Exception\\n    public static DoubleMatrixDataset<String, String> concatColumns(DoubleMatrixDataset<String, String> d1, DoubleMatrixDataset<String, String> d2)\\n    public DoubleMatrixDataset<R, C> duplicate()\\n}\\n', 'filePath': '/home/ubuntu/repos/molgenis,systemsgenetics/genetica-libraries/src/main/java/umcg/genetica/math/matrix2/DoubleMatrixDataset.java', 'methodCount': 105, 'lineCount': 111}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "\t * Lookup the pearson r matching a z-score.\n",
      "\t * @param z\n",
      "\t * @return\n",
      "\t */\n",
      "public double lookupRforZscore(double z)\n",
      "**************************************\n",
      "public double lookupRforZscore(double z) {\n",
      "    final long bin;\n",
      "    if (z >= maxZscore) {\n",
      "        bin = maxBin;\n",
      "    } else if (z < -maxZscore) {\n",
      "        bin = 0;\n",
      "    } else {\n",
      "        bin = Math.round(((z / maxZscore) + 1) * binsPerSide - halfStep);\n",
      "    }\n",
      "    return rLookupTable[(int) bin];\n",
      "}\n",
      "2127\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  24%|       | 304/1243 [10:32<33:39,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13702\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 305/1243 [10:34<33:37,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13708\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 306/1243 [10:36<33:23,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13709\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 307/1243 [10:38<30:53,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13710\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 308/1243 [10:39<29:14,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13711\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Calculates a weighted Z-score according to Whitlock's paper:\n",
      "     * http://www.ncbi.nlm.nih.gov/pubmed/16135132 Square root of the sample\n",
      "     * size is used as the weight for each test.\n",
      "     *\n",
      "     * @param zScores     Z-scores from individual tests\n",
      "     * @param sampleSizes sample sizes of these tests\n",
      "     * @return\n",
      "     */\n",
      "public static double getDatasetSizeWeightedZ(float[] zScores, int[] sampleSizes, double[] weights)\n",
      "**************************************\n",
      "public static double getDatasetSizeWeightedZ(float[] zScores, int[] sampleSizes, double[] weights) {\n",
      "    if (zScores.length != sampleSizes.length) {\n",
      "        throw new IllegalArgumentException(\"Zscores and sample sizes should have same length!\");\n",
      "    }\n",
      "    double weightedZ = 0;\n",
      "    double sampleSizeSum = 0;\n",
      "    for (int j = 0; j < zScores.length; j++) {\n",
      "        if (!Double.isNaN(zScores[j])) {\n",
      "            weightedZ += Math.sqrt(sampleSizes[j] * weights[j]) * zScores[j];\n",
      "            sampleSizeSum += (sampleSizes[j] * weights[j]);\n",
      "        }\n",
      "    }\n",
      "    weightedZ /= Math.sqrt(sampleSizeSum);\n",
      "    return weightedZ;\n",
      "}\n",
      "295\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 309/1243 [10:41<30:04,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13712\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 310/1243 [10:43<30:24,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13717\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 311/1243 [10:45<29:06,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13787\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 312/1243 [10:47<30:34,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13867\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 313/1243 [10:49<31:26,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13934\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Pair {\\n    private final L left;;\\n    private final R right;;\\n    private String sep = \"-\";;\\n    private SORTBY sorter = SORTBY.BOTH;;\\n    public Pair(L left, R right);\\n    public Pair(L left, R right, SORTBY sorter);\\n    public Pair(L left, R right, String sep);\\n    public Pair(L left, R right, String sep, SORTBY sorter);\\n    public void setSorter(SORTBY s)\\n    public L getLeft()\\n    public R getRight()\\n    public String toString()\\n    public int hashCode()\\n    public boolean equals(Object o)\\n    public int compareTo(Pair<L, R> toCompare)\\n}\\n', 'filePath': '/home/ubuntu/repos/molgenis,systemsgenetics/genetica-libraries/src/main/java/umcg/genetica/containers/Pair.java', 'methodCount': 11, 'lineCount': 17}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class BinaryFile {\\n    public static final boolean W = true;;\\n    public static final boolean R = false;;\\n    protected final DataOutputStream os;;\\n    protected DataInputStream is;;\\n    protected final File loc;;\\n    protected final boolean writeable;;\\n    private final OutputStream osh;;\\n    public BinaryFile(String loc, boolean mode) throws IOException;\\n    public BinaryFile(File loc, boolean mode) throws IOException;\\n    public BinaryFile(String loc, boolean mode, int buffersize) throws IOException;\\n    public BinaryFile(String loc, boolean mode, int buffersize, boolean useHash) throws IOException;\\n    public BinaryFile(File loc, boolean mode, int buffersize, boolean useHash) throws IOException;\\n    public void writeBytes(byte[] v) throws IOException\\n    public void writeInt(int v) throws IOException\\n    public void writeString(String s) throws IOException\\n    public void writeBool(boolean b) throws IOException\\n    public void writeFloat(float f) throws IOException\\n    public void writeDouble(double d) throws IOException\\n    public void writeLong(long l) throws IOException\\n    public int readBytes(byte[] b) throws IOException\\n    public int readInt() throws IOException, EOFException\\n    public boolean readBool() throws IOException, EOFException\\n    public String readString() throws IOException, EOFException\\n    public float readFloat() throws IOException, EOFException\\n    public double readDouble() throws IOException, EOFException\\n    public long readLong() throws IOException, EOFException\\n    public void close() throws IOException\\n    public void writeByte(byte b) throws IOException\\n    public void write(int b) throws IOException\\n    public byte[] getWrittenHash() throws IOException\\n    public int read(byte[] buffer) throws IOException\\n    public void write(byte[] buffer, int i, int len) throws IOException\\n    public void write(byte[] alleles0) throws IOException\\n    public int available() throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/molgenis,systemsgenetics/genetica-libraries/src/main/java/umcg/genetica/io/bin/BinaryFile.java', 'methodCount': 27, 'lineCount': 36}, {'fieldCount': 16, 'classCount': 1, 'methodsOnly': 'public class TextFile {\\n    // check 1,2,3\\npublic static final int DEFAULT_BUFFER_SIZE = 1024 * 64;;\\n    //Using the \\\\t from string.tab is technically not valid. I would not want to depend on this\\npublic static final Pattern tab = Pattern.compile(\"\\\\\\\\t\");;\\n    public static final Pattern space = Strings.space;;\\n    public static final Pattern colon = Strings.colon;;\\n    public static final Pattern semicolon = Strings.semicolon;;\\n    public static final Pattern comma = Strings.comma;;\\n    private boolean uselz4;;\\n    protected BufferedReader in;;\\n    protected File file;;\\n    public static final boolean W = true;;\\n    public static final boolean R = false;;\\n    protected BufferedWriter out;;\\n    protected boolean writeable;;\\n    protected static final String ENCODING = \"ISO-8859-1\";;\\n    private boolean gzipped;;\\n    private int buffersize;;\\n    public TextFile(String file, boolean mode) throws IOException;\\n    public TextFile(File file, boolean mode) throws IOException;\\n    public TextFile(File file, boolean mode, int buffersize) throws IOException;\\n    public TextFile(String file, boolean mode, int buffersize) throws IOException;\\n    public final void open() throws IOException\\n    public String readLine() throws IOException\\n    public void write(String line) throws IOException\\n    public void close() throws IOException\\n    public String[] readLineElems(Pattern p) throws IOException\\n    private Iterator<String[]> readLineElemsIterator(Pattern p) throws IOException\\n    public Iterable<String[]> readLineElemsIterable(Pattern p)\\n    public String[] readLineElemsReturnReference(Pattern p) throws IOException\\n    public String[] readLineElemsReturnObjects(Pattern p) throws IOException\\n    public int countLines() throws IOException\\n    public int countCols(Pattern p) throws IOException\\n    public String[] readAsArray() throws IOException\\n    public String[] readAsArray(int col, Pattern p) throws IOException\\n    public ArrayList<String> readAsArrayList() throws IOException\\n    public ArrayList<String> readAsArrayList(int col, Pattern p) throws IOException\\n    public void writeln(CharSequence csq) throws IOException\\n    public void writeln(String line) throws IOException\\n    public void writeln() throws IOException\\n    public void append(char c) throws IOException\\n    public void append(CharSequence csq) throws IOException\\n    public synchronized void writelnsynced(String str) throws IOException\\n    public void writelnTabDelimited(Object[] vals) throws IOException\\n    public void writelnDelimited(Object[] vals, Pattern p) throws IOException\\n    public Map<String, String> readAsHashMap(int col1, int col2) throws IOException\\n    public Map<String, String> readAsHashMap(int col1, int[] col2, String seperator) throws IOException\\n    public Map<String, String> readAsHashMap(int col1, int col2, Pattern p) throws IOException\\n    public Set<String> readAsSet(int col, Pattern p) throws IOException\\n    public void writeList(List l) throws IOException\\n    public String getFileName()\\n    public HashSet<Pair<String, String>> readAsPairs(int A, int B) throws IOException\\n    public void flush() throws IOException\\n    public HashSet<Triple<String, String, String>> readAsTriple(int A, int B, int C) throws IOException\\n    public Iterator<String> iterator()\\n}\\n', 'filePath': '/home/ubuntu/repos/molgenis,systemsgenetics/genetica-libraries/src/main/java/umcg/genetica/io/text/TextFile.java', 'methodCount': 37, 'lineCount': 55}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class Strings {\\n    public static final Pattern tab = Pattern.compile(\"\\\\t\");;\\n    public static final Pattern whitespace = Pattern.compile(\"\\\\\\\\s\");;\\n    public static final Pattern comma = Pattern.compile(\",\");;\\n    public static final Pattern semicolon = Pattern.compile(\";\");;\\n    public static final Pattern colon = Pattern.compile(\":\");;\\n    public static final Pattern pipe = Pattern.compile(\"\\\\\\\\|\");;\\n    public static final Pattern forwardslash = Pattern.compile(\"/\");;\\n    public static final Pattern backwardslash = Pattern.compile(\"\\\\\\\\\\\\\\\\\");;\\n    public static final Pattern dot = Pattern.compile(\"\\\\\\\\.\");;\\n    public static final Pattern space = Pattern.compile(\" \");;\\n    public static final Pattern dash = Pattern.compile(\"-\");;\\n    public static final Pattern equalssign = Pattern.compile(\"=\");;\\n    public static String concat(String[] s, Pattern t)\\n    public static String concat(Object[] s, Pattern t)\\n    public static String concat(double[] s, Pattern t, DecimalFormat f)\\n    public static String concat(double[] s, Pattern t)\\n    public static String concat(double[] s, DecimalFormat f, Pattern t)\\n    public static String concat(float[] s, DecimalFormat f, Pattern t)\\n    public static String concat(int[] s, Pattern t)\\n    public static String concat(List<String> s, Pattern t)\\n    public static String concat(String[] s, Pattern t, int start, int end)\\n    public static String[] split(String in)\\n    public static String reverse(String str)\\n    public static int[] countOccurrences(String ln, Pattern tab, String match)\\n    public static String[] subsplit(String ln, Pattern pattern, boolean[] colsToInclude)\\n    public static String[] subsplit(String ln, Pattern pattern, int offset, boolean[] colsToInclude)\\n    public static String[] subsplit(String ln, Pattern pattern, int offset, boolean[] colsToInclude, int nrCols)\\n    public static String[] subsplit(String ln, Pattern tab, int lowerBound, int upperBound)\\n    public static int countSeparatorOccurrences(String permln, Pattern tab)\\n    private static Map<String, String> cache;;\\n    public static synchronized boolean cacheContains(String s)\\n    public static synchronized String cache(String s)\\n    public static String concat(float[] s, Pattern t)\\n    public static String concat(String[] elems, boolean[] includeelem, Pattern tab)\\n    public static String concat(String[] s, Pattern t, boolean[] includeElem, String replaceNull)\\n}\\n', 'filePath': '/home/ubuntu/repos/molgenis,systemsgenetics/genetica-libraries/src/main/java/umcg/genetica/text/Strings.java', 'methodCount': 22, 'lineCount': 37}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 9}\n",
      "{}\n",
      "/**\n",
      "\t * Order columns\n",
      "\t */\n",
      "public void orderOnColumnnames()\n",
      "**************************************\n",
      "public void orderOnColumnnames() {\n",
      "    LinkedHashMap<C, Integer> newColHash = new LinkedHashMap<C, Integer>((int) Math.ceil(this.matrix.columns() / 0.75));\n",
      "    ArrayList<C> names = this.getColObjects();\n",
      "    Collections.sort(names);\n",
      "    int pos = 0;\n",
      "    for (C name : names) {\n",
      "        newColHash.put(name, pos);\n",
      "        pos++;\n",
      "    }\n",
      "    reorderCols(newColHash);\n",
      "}\n",
      "2274\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 314/1243 [10:52<32:49,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13941\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 315/1243 [10:54<31:22,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13942\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  25%|       | 316/1243 [10:55<29:29,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13958\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 317/1243 [10:57<28:17,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "14453\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 318/1243 [10:58<27:19,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15158\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 11}\n",
      "{}\n",
      "/**\n",
      "     * Close the file, and release any associated system resources.\n",
      "     *\n",
      "     * @exception IOException if an I/O error occurrs.\n",
      "     */\n",
      "public void close() throws IOException\n",
      "**************************************\n",
      "/**\n",
      " * Close the file, and release any associated system resources.\n",
      " *\n",
      " * @exception IOException if an I/O error occurrs.\n",
      " */\n",
      "public void close() throws IOException {\n",
      "    // If we are writing and the buffer has been modified, flush the contents\n",
      "    // of the buffer.\n",
      "    if ((mode | WRITE) > 0 && bufferModified) {\n",
      "        file.seek(bufferStart);\n",
      "        file.write(buffer, 0, (int) dataSize);\n",
      "    }\n",
      "    // Close the underlying file object.\n",
      "    file.close();\n",
      "}\n",
      "181\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 319/1243 [11:01<29:26,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15341\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 320/1243 [11:03<31:01,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15354\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 321/1243 [11:05<29:45,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15423\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 322/1243 [11:07<30:54,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15481\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 323/1243 [11:09<31:03,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15494\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ThrowableEntry {\\n    private String title;;\\n    private Throwable err;;\\n    public ThrowableEntry(MessageType type, String title, Throwable err);\\n    public String getMessage()\\n    public String getTitle()\\n    public void setTitle(String title)\\n    public Throwable getErr()\\n    public void setErr(Throwable err)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/component/exception/ThrowableEntry.java', 'methodCount': 6, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/component/message/MessageType.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Get spectrum by spectrum id\n",
      "     *\n",
      "     * @param spectrumId spectrum id\n",
      "     * @return Spectrum    spectrum\n",
      "     */\n",
      "private Spectrum getSpectrum(Comparable spectrumId)\n",
      "**************************************\n",
      "private Spectrum getSpectrum(Comparable spectrumId) {\n",
      "    Spectrum result = null;\n",
      "    try {\n",
      "        result = controller.getSpectrumById(spectrumId);\n",
      "    } catch (DataAccessException dex) {\n",
      "        String msg = \"Failed to retrieve data entry from data source\";\n",
      "        logger.error(msg, dex);\n",
      "        appContext.addThrowableEntry(new ThrowableEntry(MessageType.ERROR, msg, dex));\n",
      "    }\n",
      "    return result;\n",
      "}\n",
      "377\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 324/1243 [11:11<32:42,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15532\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 325/1243 [11:14<32:51,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15576\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 326/1243 [11:16<32:49,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15587\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 327/1243 [11:18<32:49,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15598\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 328/1243 [11:20<30:56,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15624\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class GUIBlocker {\\n    private final S source;;\\n    private final Scope scope;;\\n    private final T target;;\\n    public GUIBlocker(S src, Scope s, T tar);\\n    public final S getSource()\\n    public final Scope getScope()\\n    public final T getTarget()\\n    public abstract void block()\\n    public abstract void unblock()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/utils/GUIBlocker.java', 'methodCount': 6, 'lineCount': 11}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class PropertyChangeHelper {\\n    private final PropertyChangeSupport supporter;;\\n    public PropertyChangeHelper();\\n    public PropertyChangeHelper(Object source);\\n    public boolean hasPropertyChangeListener(PropertyChangeListener listener)\\n    public void addPropertyChangeListener(PropertyChangeListener listener)\\n    public void addPropertyChangeListener(String propName, PropertyChangeListener listener)\\n    public void removePropertyChangeListener(PropertyChangeListener listener)\\n    public void removePropertyChangeListener(String propName, PropertyChangeListener listener)\\n    public void removeAllPropertyChangeListeners()\\n    public PropertyChangeListener[] getPropertyChangeListeners()\\n    public PropertyChangeListener[] getPropertyChangeListeners(String propName)\\n    public void firePropertyChange(final PropertyChangeEvent event)\\n    public void firePropertyChange(String propName, Object oldValue, Object newValue)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/utils/PropertyChangeHelper.java', 'methodCount': 12, 'lineCount': 15}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ThreadCalculator {\\n    // This implies that the wait time and compute time are almost equal\\npublic static final float DEFAULT_RATIO_OF_WAIT_OVER_COMPUTE_TIME = 1.0f;;\\n    // This implies we want 80% CPU utilization\\npublic static final float DEFAULT_CPU_UTILIZATION = 0.8f;;\\n    public static int calculateNumberOfThreads()\\n    public static int calculateNumberOfThreads(float ratioOfWaitOverComputeTime)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/utils/ThreadCalculator.java', 'methodCount': 2, 'lineCount': 6}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 4}\n",
      "{'Task': 'public class Task {\\n    private static final Logger logger = LoggerFactory.getLogger(Task.class);;\\n    public static final String NAME_PROP = \"name\";;\\n    public static final String DESCRIPTION_PROP = \"description\";;\\n    public static final String COMPLETED_PROP = \"completed\";;\\n    public static final String GUI_BLOCKER_PROP = \"blocker\";;\\n    // task name\\n@GuardedBy(\"this\")\\nprivate String name;;\\n    // task description\\n@GuardedBy(\"this\")\\nprivate String description;;\\n    // lock for owners collection\\nprivate final Object ownersLock = new Object();;\\n    @GuardedBy(\"ownersLock\")\\nprivate final Collection<Object> owners;;\\n    // lock for task listeners collection\\nprivate final Object taskListenersLock = new Object();;\\n    @GuardedBy(\"taskListenersLock\")\\nprivate final Collection<TaskListener<T, V>> taskListeners;;\\n    // GUI blocker for the task for blocking GUI components\\n@GuardedBy(\"this\")\\nprivate GUIBlocker blocker;;\\n    public Task();\\n    public synchronized String getName()\\n    public void setName(String n)\\n    public synchronized String getDescription()\\n    public void setDescription(String desc)\\n    public synchronized GUIBlocker getGUIBlocker()\\n    public void setGUIBlocker(GUIBlocker bl)\\n    public final void addOwner(Object owner)\\n    public final List<Object> getOwners()\\n    public void removeByOwner(Object owner)\\n    public final boolean isPending()\\n    public final boolean isStarted()\\n    public synchronized boolean hasPropertyChangeListener(PropertyChangeListener listener)\\n    public void addTaskListener(TaskListener<T, V> listener)\\n    public void removeTaskListener(TaskListener<T, V> listener)\\n    public boolean hasTaskListener(TaskListener<T, V> listener)\\n    public Collection<TaskListener<T, V>> getTaskListeners()\\n    protected void process(List<V> values)\\n    protected final void done()\\n    protected abstract void finished()\\n    protected void failed(Throwable error)\\n    protected abstract void succeed(T results)\\n    protected abstract void cancelled()\\n    protected abstract void interrupted(InterruptedException iex)\\n    private void fireStartedListeners()\\n    private void fireProcessListeners(List<V> values)\\n    private void fireCompletionListeners()\\n    private void fireCancelledListeners()\\n    private void fireInterruptedListeners(InterruptedException iex)\\n    private void fireSucceedListeners(T result)\\n    private void fireFailedListeners(Throwable error)\\n    private void fireFinishedListeners()\\n    private void fireProgressListeners(int progress)\\n}\\n', 'TaskListener': 'public interface TaskListener {\\n     void started(TaskEvent<Void> event)\\n     void process(TaskEvent<List<V>> event)\\n     void finished(TaskEvent<Void> event)\\n     void failed(TaskEvent<Throwable> event)\\n     void succeed(TaskEvent<T> event)\\n     void cancelled(TaskEvent<Void> event)\\n     void interrupted(TaskEvent<InterruptedException> iex)\\n     void progress(TaskEvent<Integer> progress)\\n}\\n'}\n",
      "/**\n",
      "     * Return a list of Tasks which has the specified TaskListener.\n",
      "     *\n",
      "     * @param listener Task listener.\n",
      "     * @return List<Task>   a list of tasks.\n",
      "     */\n",
      "public List<Task> getTasks(TaskListener listener)\n",
      "**************************************\n",
      "@SuppressWarnings(\"unchecked\")\n",
      "public List<Task> getTasks(TaskListener listener) {\n",
      "    List<Task> ts = new ArrayList<>();\n",
      "    synchronized (tasksLock) {\n",
      "        for (Task task : tasks) {\n",
      "            if (task.hasTaskListener(listener)) {\n",
      "                ts.add(task);\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return ts;\n",
      "}\n",
      "1362\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  26%|       | 329/1243 [11:21<28:57,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15625\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 330/1243 [11:23<27:45,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15626\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 331/1243 [11:24<26:57,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15671\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 332/1243 [11:27<28:24,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15679\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 333/1243 [11:28<27:19,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15754\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Desktop {\\n    private static final Logger logger = Logger.getLogger(Desktop.class.getName());;\\n    private final Collection<ExitListener> exitListeners;;\\n    private static DesktopContext desktopContext;;\\n    private static Desktop desktop = null;;\\n    protected Desktop();\\n    public static synchronized void launch(final Class<T> appClass, final Class<V> appContextClass, final String[] args)\\n    private static V buildAppContext(Class<V> appContextClass) throws Exception\\n    private static T buildApp(final Class<T> appClass) throws Exception\\n    public static synchronized R getInstance(Class<R> appClass)\\n    public static synchronized Desktop getInstance()\\n    private static void checkAppLaunchState()\\n    public abstract void init(String[] args)\\n    public abstract void ready()\\n    public abstract void show()\\n    public abstract void postShow()\\n    public abstract void finish()\\n    public final void shutdown(EventObject event)\\n    public final DesktopContext getDesktopContext()\\n    public final Collection<ExitListener> getExitListeners()\\n    public final void addExitListeners(ExitListener listener)\\n    public final void removeExitListener(ExitListener listener)\\n    public abstract JFrame getMainComponent()\\n    public abstract JMenuBar getMenuBar()\\n    public abstract StatusBar getStatusBar()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/desktop/Desktop.java', 'methodCount': 20, 'lineCount': 26}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class DesktopContext {\\n    private static final Logger logger = Logger.getLogger(DesktopContext.class.getName());;\\n    /**\\n * reference to desktop object\\n */\\nprivate Desktop desktop = null;;\\n    /**\\n * task manager manages all the ongoing tasks\\n */\\nprivate final TaskManager taskManager;;\\n    /**\\n * property manager stores all the property assigned to this application\\n */\\nprivate final PropertyManager propertyManager;;\\n    /**\\n * ThrowableHandler stores all the throwables for this application\\n */\\nprivate final ThrowableHandler throwableHandler;;\\n    protected DesktopContext();\\n    public final synchronized Desktop getDesktop()\\n    public final synchronized void setDesktop(Desktop desktop)\\n    public final synchronized TaskManager getTaskManager()\\n    public final synchronized PropertyManager getPropertyManager()\\n    public final synchronized void addTask(Task task)\\n    public final synchronized void addTask(Task task, boolean notify)\\n    public final synchronized List<Task> getTask(TaskListener listener)\\n    public final synchronized List<Task> getTask(PropertyChangeListener listener)\\n    public final synchronized List<Task> getTask(Class<? extends Task> taskClass)\\n    public final synchronized boolean hasTask(Task task)\\n    public final synchronized boolean cancelTask(Task task, boolean interrupt)\\n    public final synchronized void removeTaskListener(TaskListener listener)\\n    public final synchronized Properties getProperties()\\n    public final synchronized String getProperty(String name)\\n    public final synchronized String getProperty(String name, Object[] args)\\n    public final synchronized void setProperty(String name, String value)\\n    public final synchronized void loadSystemProps(InputStream in) throws IOException\\n    public final synchronized void loadClusterProjectsProperties(InputStream in) throws IOException\\n    public final synchronized void loadClusterAssayProperties(InputStream in) throws IOException\\n    public final synchronized ThrowableHandler getThrowableHandler()\\n    public final synchronized List<ThrowableEntry> getAllThrowables()\\n    public final synchronized void addThrowableEntry(ThrowableEntry entry)\\n    public final synchronized void removeThrowableEntry(ThrowableEntry entry)\\n    public void removeAllThrowableEntries()\\n    public ClusterProjectProperties getProjectClusterProperties()\\n    public ClusterAssayProperties getProjectAssaysClusterProperties()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/desktop/DesktopContext.java', 'methodCount': 27, 'lineCount': 34}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class HttpUtilities {\\n    public static HttpURLConnection createHttpConnection(String url, String method) throws Exception\\n    public static void openURL(String url)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/url/HttpUtilities.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Show update dialog\n",
      "     */\n",
      "public static void showUpdateDialog()\n",
      "**************************************\n",
      "public static void showUpdateDialog() {\n",
      "    int option = JOptionPane.showConfirmDialog(null, \"<html><b>A new version of PRIDE Inspector is available</b>.<br><br> \" + \"Would you like to update?</html>\", \"Update Info\", JOptionPane.YES_NO_OPTION);\n",
      "    if (option == JOptionPane.YES_OPTION) {\n",
      "        DesktopContext context = Desktop.getInstance().getDesktopContext();\n",
      "        String website = context.getProperty(\"pride.inspector.download.website\");\n",
      "        HttpUtilities.openURL(website);\n",
      "    }\n",
      "}\n",
      "1143\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 334/1243 [11:30<28:31,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15824\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 335/1243 [11:32<27:05,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15883\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 336/1243 [11:34<28:10,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15885\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 337/1243 [11:36<27:11,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15891\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 338/1243 [11:38<29:36,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15892\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class DataAccessControllerPane {\\n    /**\\n * DataAccessController behind this panel\\n */\\nprotected DataAccessController controller = null;;\\n    /**\\n * A reference to parent component\\n */\\nprotected JComponent parentComponent;;\\n    /**\\n * reference to application context\\n */\\nprotected PrideInspectorContext appContext;;\\n    /**\\n * Title for the panel\\n */\\nprivate String title = null;;\\n    /**\\n * Icon to use when there is no ongoing background tasks associated with this panel\\n */\\nprivate Icon finalIcon = null;;\\n    /**\\n * Icon to use when there is ongoing background tasks associated with this panel\\n */\\nprivate Icon loadingIcon = null;;\\n    public DataAccessControllerPane(DataAccessController controller);\\n    public DataAccessControllerPane(DataAccessController controller, JComponent parentComponent);\\n    public DataAccessControllerPane(DataAccessController controller, JComponent parentComponent, String title);\\n    public JComponent getParentComponent()\\n    public void setParentComponent(JComponent parentComponent)\\n    public PrideInspectorContext getAppContext()\\n    public String getTitle()\\n    public void setTitle(String title)\\n    public Icon getIcon()\\n    public void setIcon(Icon icon)\\n    public Icon getLoadingIcon()\\n    public void setLoadingIcon(Icon icon)\\n    protected void setupMainPane()\\n    protected void addComponents()\\n    public void populate()\\n    public DataAccessController getController()\\n    public void propertyChange(PropertyChangeEvent evt)\\n    public void started(TaskEvent<Void> event)\\n    public void process(TaskEvent<List<V>> listTaskEvent)\\n    public void finished(TaskEvent<Void> event)\\n    public void failed(TaskEvent<Throwable> event)\\n    public void succeed(TaskEvent<T> tTaskEvent)\\n    public void cancelled(TaskEvent<Void> event)\\n    public void interrupted(TaskEvent<InterruptedException> iex)\\n    public void progress(TaskEvent<Integer> progress)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/component/DataAccessControllerPane.java', 'methodCount': 25, 'lineCount': 33}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class TableFactory {\\n    public static JTable createSpectrumTable()\\n    public static JTable createChromatogramTable()\\n    public static JXTable createProteinTable(final Collection<SearchEngineScoreCvTermReference> listProteinScores, final boolean hasProteinGroups, DataAccessController controller)\\n    public static JXTable createProteinTable(Collection<SearchEngineScoreCvTermReference> listProteinScores, DataAccessController controller)\\n    public static JXTable createPeptideSpeciesTable(int defaultRankingThreshold, double minDeltaMz, double maxDeltaMz)\\n    public static JTable createPeptideTable(TableModel tableModel, DataAccessController controller)\\n    public static JTable createPeptideTable(Collection<SearchEngineScoreCvTermReference> listPeptideScores, DataAccessController controller)\\n    public static JTable createPSMTable(Collection<SearchEngineScoreCvTermReference> listPeptideScores, int defaultRankingThreshold, DataAccessController controller)\\n    public static JTable createPTMTable()\\n    public static JTable createReferenceTable(Collection<Reference> references)\\n    public static JTable createParamTable(List<ParamGroup> paramGroups)\\n    public static JTable createParamTable(ParamGroup paramGroup)\\n    public static JTable createParamTable(Collection<Parameter> parameters)\\n    private static JTable createParamTable(ParamTableModel paramTableModel)\\n    public static JTable createContactTable(Collection<Person> contacts)\\n    public static JTable createQuantSampleTable(QuantitativeSample sample)\\n    public static JTable createQuantProteinTable(DataAccessController controller, TableModel tableModel)\\n    public static JTable createQuantProteinTable(DataAccessController controller, Collection<SearchEngineScoreCvTermReference> listProteinScores)\\n    public static JTable createQuantProteinTable(DataAccessController controller, Collection<SearchEngineScoreCvTermReference> listProteinScores, Map<Comparable, StudyVariable> studyVariables)\\n    public static JTable createQuantPeptideTable(DataAccessController controller, Collection<SearchEngineScoreCvTermReference> listPeptideScores, Map<Comparable, StudyVariable> studyVariables)\\n    public static JTable createQuantPeptideTable(DataAccessController controller, Collection<SearchEngineScoreCvTermReference> listPeptideScores)\\n    private static JXTable createDefaultJXTable(TableModel tableModel)\\n    private static ProteinSortableTreeTable createDefaultJXTreeTable(SortableProteinTreeTableModel tableModel, boolean hasProteinGropus)\\n    private static void configureTreeTable(JXTreeTable table)\\n    private static void configureTable(JXTable table)\\n    public static MouseAdapter createHeaderMouseAdapter(final JXTreeTable treeTbl)\\n    public static JXTable createProjectDetailTable()\\n    public static JXTable createAssayDetailTable()\\n    public static JTable createAssayFileDownloadTable()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/component/table/TableFactory.java', 'methodCount': 29, 'lineCount': 31}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Build meta data pane, this panel displays the identification type, search engine and search database\n",
      "     *\n",
      "     * @return JPanel   meta data pane\n",
      "     */\n",
      "private JPanel buildMetaDataPane()\n",
      "**************************************\n",
      "private JPanel buildMetaDataPane() {\n",
      "    JPanel metaDataPanel = new JPanel();\n",
      "    metaDataPanel.setOpaque(false);\n",
      "    metaDataPanel.setLayout(new FlowLayout(FlowLayout.LEFT));\n",
      "    JLabel tableLabel = new JLabel(\"<html><b>Samples</b></html>\");\n",
      "    metaDataPanel.add(tableLabel);\n",
      "    return metaDataPanel;\n",
      "}\n",
      "1332\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 339/1243 [11:39<27:43,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15928\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 340/1243 [11:42<29:08,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15930\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  27%|       | 341/1243 [11:43<27:31,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16052\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 342/1243 [11:45<26:42,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16090\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 343/1243 [11:47<27:50,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16178\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class AnnotatedProtein {\\n    public static final String PEPTIDE_SELECTION_PROP = \"selectedAnnotation\";;\\n    private List<PeptideAnnotation> annotations;;\\n    private PeptideAnnotation selectedAnnotation;;\\n    private PropertyChangeHelper propertyChangeHelper;;\\n    /**\\n * these peptides can be duplicated\\n */\\nprivate int numOfValidPeptides = -1;;\\n    /**\\n * these peptides must be valid as well\\n */\\nprivate int numOfUniquePeptides = -1;;\\n    /**\\n * the number of amino acids covered\\n */\\nprivate int numOfAminoAcidCovered = -1;;\\n    public AnnotatedProtein(Protein protein);\\n    public AnnotatedProtein(String accession, String name, STATUS status, String sequence);\\n    public AnnotatedProtein(String accession);\\n    public boolean isValidPeptideAnnotation(PeptideAnnotation annotation)\\n    public boolean isStrictValidPeptideAnnotation(PeptideAnnotation annotation)\\n    public void addAnnotation(PeptideAnnotation annotation)\\n    public void removeAnnotation(PeptideAnnotation annotation)\\n    public List<PeptideAnnotation> getAnnotations()\\n    public PeptideAnnotation getSelectedAnnotation()\\n    public void setSelectedAnnotation(PeptideAnnotation selectedAnnotation)\\n    public int getNumOfPeptides()\\n    public int getNumOfValidPeptides()\\n    public void setNumOfValidPeptides(int numOfValidPeptides)\\n    public int getNumOfUniquePeptides()\\n    public void setNumOfUniquePeptides(int numOfUniquePeptides)\\n    public Set<Integer> searchStartingPosition(PeptideAnnotation annotation)\\n    public double getSequenceCoverage()\\n    public int getNumOfAminoAcidCovered()\\n    private void populateCoverage()\\n    public void setNumOfAminoAcidCovered(int numOfAminoAcidCovered)\\n    public void addPropertyChangeListener(PropertyChangeListener listener)\\n    public void removePropertyChangeListener(PropertyChangeListener listener)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/component/sequence/AnnotatedProtein.java', 'methodCount': 24, 'lineCount': 33}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class TaskEvent {\\n    private final V value;;\\n    public TaskEvent(Task source, V value);\\n    public V getValue()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/task/TaskEvent.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface TaskListener {\\n     void started(TaskEvent<Void> event)\\n     void process(TaskEvent<List<V>> event)\\n     void finished(TaskEvent<Void> event)\\n     void failed(TaskEvent<Throwable> event)\\n     void succeed(TaskEvent<T> event)\\n     void cancelled(TaskEvent<Void> event)\\n     void interrupted(TaskEvent<InterruptedException> iex)\\n     void progress(TaskEvent<Integer> progress)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/task/TaskListener.java', 'methodCount': 8, 'lineCount': 10}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ProteinAccession {\\n    private String accession;;\\n    private String mappedAccession;;\\n    public ProteinAccession(String accession, String mappedAccession);\\n    public String getAccession()\\n    public String getMappedAccession()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n    public int compareTo(Object o)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/utils/ProteinAccession.java', 'methodCount': 7, 'lineCount': 11}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Add protein related details\n",
      "     *\n",
      "     * @param newData protein detail map\n",
      "     */\n",
      "protected void addProteinDetails(Object newData)\n",
      "**************************************\n",
      "protected void addProteinDetails(Object newData) {\n",
      "    Map<String, Protein> proteins = (Map<String, Protein>) newData;\n",
      "    Collection<PeptideRow> peptideRows = spectrumIdToPeptideRow.values();\n",
      "    for (PeptideRow parentPeptideRow : peptideRows) {\n",
      "        addProteinDetailsForPeptideRow(proteins, (PeptideRow) getRoot(), parentPeptideRow);\n",
      "        for (PeptideRow childPeptideRow : parentPeptideRow.getChildren()) {\n",
      "            addProteinDetailsForPeptideRow(proteins, parentPeptideRow, childPeptideRow);\n",
      "        }\n",
      "    }\n",
      "}\n",
      "1140\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 344/1243 [11:49<26:55,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16181\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 345/1243 [11:51<30:31,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16236\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 346/1243 [11:53<28:42,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16379\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 347/1243 [11:55<30:12,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16473\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 348/1243 [11:57<30:30,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16480\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 15, 'classCount': 1, 'methodsOnly': 'public class PrideInspectorContext {\\n    private static final Logger logger = LoggerFactory.getLogger(PrideInspector.class);;\\n    /**\\n * triggered when the visibility property of data source browser is changed\\n */\\npublic static final String LEFT_CONTROL_PANE_VISIBILITY = \"leftControlPaneVisible\";;\\n    /**\\n * DataAccessMonitor manages a list of\\n */\\nprivate final DataAccessMonitor dataAccessMonitor;;\\n    /**\\n * This map maintains a reference between DataAccessController and its data content display pane\\n */\\nprivate final Map<DataAccessController, JComponent> dataContentPaneCache;;\\n    /**\\n * This map maintains a reference between DataAccessController and its Summary pane\\n */\\nprivate final Map<DataAccessController, JComponent> dataSummaryPaneCache;;\\n    /**\\n * This map maintains a reference between DataAccessController and its action\\n */\\nprivate final Map<DataAccessController, Map<Class<? extends PrideAction>, PrideAction>> sharedActionCache;;\\n    /**\\n * data source browser visibility\\n */\\nprivate boolean leftControlPaneVisible;;\\n    /**\\n * Tracking all the summary report in a list model for each database access controller\\n */\\nprivate final Map<DataAccessController, ListModel> summaryReportTracker;;\\n    /**\\n * The main help set for PRIDE Inspector\\n */\\nprivate HelpSet mainHelpSet;;\\n    /**\\n * The main help broker\\n */\\nprivate HelpBroker mainHelpBroker;;\\n    /**\\n * The path to open file\\n */\\nprivate String openFilePath;;\\n    /**\\n * List of project details after login to PRIDE-R\\n */\\nprivate LoginRecord loginRecord;;\\n    /**\\n * welcome pane\\n */\\nprivate WelcomePane welcomePane = null;;\\n    /**\\n * database search pane\\n */\\nprivate PrideArchiveWSSearchPane prideArchiveWSSearchPane = null;;\\n    /**\\n * selected data transfer protocol\\n */\\nprivate DataTransferProtocol dataTransferProtocol;;\\n    public PrideInspectorContext();\\n    public synchronized DataTransferProtocol getDataTransferProtocol()\\n    public synchronized void setDataTransferProtocol(DataTransferProtocol dataTransferProtocol)\\n    public final DataAccessMonitor getDataAccessMonitor()\\n    public WelcomePane getWelcomePane()\\n    public void setWelcomePane(WelcomePane welcomePane)\\n    public PrideArchiveWSSearchPane getPrideArchiveWSSearchPane()\\n    public void setPrideArchiveWSSearchPane(PrideArchiveWSSearchPane prideArchiveWSSearchPane)\\n    public final List<DataAccessController> getControllers()\\n    public final int getNumberOfControllers()\\n    public final boolean isForegroundDataAccessController(DataAccessController controller)\\n    public final DataAccessController getForegroundDataAccessController()\\n    public final void setForegroundDataAccessController(DataAccessController controller, String message)\\n    public final void removeDataAccessController(DataAccessController controller, boolean cancelTasks, String message)\\n    public final void replaceSummaryReport(DataAccessController controller, DataAccessController replacement)\\n    public final void replaceDataAccessController(DataAccessController original, DataAccessController replacement, boolean cancelTasks, String message)\\n    public final void addDataAccessController(DataAccessController controller, String message)\\n    public final void addDataAccessController(DataAccessController controller, boolean foreground, String message)\\n    public final JComponent getDataContentPane(DataAccessController controller)\\n    public final JComponent getSummaryPane(DataAccessController controller)\\n    public final void addDataContentPane(DataAccessController controller, JComponent component)\\n    public final void addSummaryContentPane(DataAccessController controller, JComponent component)\\n    public final void removeDataContentPane(DataAccessController controller)\\n    public final synchronized PrideAction getPrideAction(DataAccessController controller, Class<? extends PrideAction> actionClass)\\n    public final synchronized void addPrideAction(DataAccessController controller, PrideAction action)\\n    public final void removePrideAction(DataAccessController controller, Class<? extends PrideAction> actionClass)\\n    public final void removePrideAction(DataAccessController controller)\\n    public final synchronized boolean isLeftControlPaneVisible()\\n    public void setLeftControlPaneVisible(boolean leftControlPaneVisible)\\n    private void createHelp()\\n    public HelpSet getMainHelpSet()\\n    public HelpBroker getMainHelpBroker()\\n    public String getOpenFilePath()\\n    public void setOpenFilePath(String openFilePath)\\n    public LoginRecord getLoginRecord()\\n    public void setLoginRecord(LoginRecord loginRecord)\\n    public synchronized ListModel getSummaryReportModel(DataAccessController controller)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/PrideInspectorContext.java', 'methodCount': 37, 'lineCount': 54}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class PrideAction {\\n    private int keyCode;;\\n    private int keyMod;;\\n    private boolean hasAccelerator = false;;\\n    private boolean isCheckBoxItem = false;;\\n    private boolean isRadioItem = false;;\\n    public PrideAction();\\n    public PrideAction(String name);\\n    public PrideAction(String name, javax.swing.Icon icon);\\n    public abstract void actionPerformed(ActionEvent e)\\n    public void menuSelected(MenuEvent e)\\n    public void menuDeselected(MenuEvent e)\\n    public void menuCanceled(MenuEvent e)\\n    public void setAccelerator(int keyCode)\\n    public void setAccelerator(int keyCode, int keyMod)\\n    public boolean isAccelerated()\\n    public int getKeyCode()\\n    public void setKeyCode(int keyCode)\\n    public int getKeyMod()\\n    public void setKeyMod(int keyMod)\\n    public boolean isCheckBoxItem()\\n    public void setCheckBoxItem(boolean checkBoxItem)\\n    public boolean isRadioItem()\\n    public void setRadioItem(boolean radioItem)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/action/PrideAction.java', 'methodCount': 18, 'lineCount': 25}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class OpenHelpAction {\\n    private String helpIndex;;\\n    public OpenHelpAction(String name, Icon icon, String helpIndex);\\n    public OpenHelpAction(String name, Icon icon, String helpIndex, int acceleratorKey);\\n    public void actionPerformed(ActionEvent e)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/action/impl/OpenHelpAction.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class DataAccessControllerPane {\\n    /**\\n * DataAccessController behind this panel\\n */\\nprotected DataAccessController controller = null;;\\n    /**\\n * A reference to parent component\\n */\\nprotected JComponent parentComponent;;\\n    /**\\n * reference to application context\\n */\\nprotected PrideInspectorContext appContext;;\\n    /**\\n * Title for the panel\\n */\\nprivate String title = null;;\\n    /**\\n * Icon to use when there is no ongoing background tasks associated with this panel\\n */\\nprivate Icon finalIcon = null;;\\n    /**\\n * Icon to use when there is ongoing background tasks associated with this panel\\n */\\nprivate Icon loadingIcon = null;;\\n    public DataAccessControllerPane(DataAccessController controller);\\n    public DataAccessControllerPane(DataAccessController controller, JComponent parentComponent);\\n    public DataAccessControllerPane(DataAccessController controller, JComponent parentComponent, String title);\\n    public JComponent getParentComponent()\\n    public void setParentComponent(JComponent parentComponent)\\n    public PrideInspectorContext getAppContext()\\n    public String getTitle()\\n    public void setTitle(String title)\\n    public Icon getIcon()\\n    public void setIcon(Icon icon)\\n    public Icon getLoadingIcon()\\n    public void setLoadingIcon(Icon icon)\\n    protected void setupMainPane()\\n    protected void addComponents()\\n    public void populate()\\n    public DataAccessController getController()\\n    public void propertyChange(PropertyChangeEvent evt)\\n    public void started(TaskEvent<Void> event)\\n    public void process(TaskEvent<List<V>> listTaskEvent)\\n    public void finished(TaskEvent<Void> event)\\n    public void failed(TaskEvent<Throwable> event)\\n    public void succeed(TaskEvent<T> tTaskEvent)\\n    public void cancelled(TaskEvent<Void> event)\\n    public void interrupted(TaskEvent<InterruptedException> iex)\\n    public void progress(TaskEvent<Integer> progress)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/component/DataAccessControllerPane.java', 'methodCount': 25, 'lineCount': 33}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class PSMEvent {\\n    private final Comparable identificationId;;\\n    private final Comparable peptideId;;\\n    private final DataAccessController controller;;\\n    public PSMEvent(Object source, DataAccessController controller, Comparable identId, Comparable pepId);\\n    public Comparable getIdentificationId()\\n    public Comparable getPeptideId()\\n    public DataAccessController getController()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/event/container/PSMEvent.java', 'methodCount': 4, 'lineCount': 9}, {'fieldCount': 12, 'classCount': 1, 'methodsOnly': 'public class Task {\\n    private static final Logger logger = LoggerFactory.getLogger(Task.class);;\\n    public static final String NAME_PROP = \"name\";;\\n    public static final String DESCRIPTION_PROP = \"description\";;\\n    public static final String COMPLETED_PROP = \"completed\";;\\n    public static final String GUI_BLOCKER_PROP = \"blocker\";;\\n    // task name\\n@GuardedBy(\"this\")\\nprivate String name;;\\n    // task description\\n@GuardedBy(\"this\")\\nprivate String description;;\\n    // lock for owners collection\\nprivate final Object ownersLock = new Object();;\\n    @GuardedBy(\"ownersLock\")\\nprivate final Collection<Object> owners;;\\n    // lock for task listeners collection\\nprivate final Object taskListenersLock = new Object();;\\n    @GuardedBy(\"taskListenersLock\")\\nprivate final Collection<TaskListener<T, V>> taskListeners;;\\n    // GUI blocker for the task for blocking GUI components\\n@GuardedBy(\"this\")\\nprivate GUIBlocker blocker;;\\n    public Task();\\n    public synchronized String getName()\\n    public void setName(String n)\\n    public synchronized String getDescription()\\n    public void setDescription(String desc)\\n    public synchronized GUIBlocker getGUIBlocker()\\n    public void setGUIBlocker(GUIBlocker bl)\\n    public final void addOwner(Object owner)\\n    public final List<Object> getOwners()\\n    public void removeByOwner(Object owner)\\n    public final boolean isPending()\\n    public final boolean isStarted()\\n    public synchronized boolean hasPropertyChangeListener(PropertyChangeListener listener)\\n    public void addTaskListener(TaskListener<T, V> listener)\\n    public void removeTaskListener(TaskListener<T, V> listener)\\n    public boolean hasTaskListener(TaskListener<T, V> listener)\\n    public Collection<TaskListener<T, V>> getTaskListeners()\\n    protected void process(List<V> values)\\n    protected final void done()\\n    protected abstract void finished()\\n    protected void failed(Throwable error)\\n    protected abstract void succeed(T results)\\n    protected abstract void cancelled()\\n    protected abstract void interrupted(InterruptedException iex)\\n    private void fireStartedListeners()\\n    private void fireProcessListeners(List<V> values)\\n    private void fireCompletionListeners()\\n    private void fireCancelledListeners()\\n    private void fireInterruptedListeners(InterruptedException iex)\\n    private void fireSucceedListeners(T result)\\n    private void fireFailedListeners(Throwable error)\\n    private void fireFinishedListeners()\\n    private void fireProgressListeners(int progress)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/task/Task.java', 'methodCount': 33, 'lineCount': 47}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class TaskUtil {\\n    private static final DesktopContext context = Desktop.getInstance().getDesktopContext();;\\n    private TaskUtil();\\n    public static void startBackgroundTask(Task task)\\n    public static void startBackgroundTask(Task task, Object taskOwner)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/task/TaskUtil.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class RetrievePeptideTask {\\n    private static final Logger logger = LoggerFactory.getLogger(RetrievePeptideTask.class);;\\n    private final Comparable identId;;\\n    private final Comparable peptideId;;\\n    public RetrievePeptideTask(DataAccessController controller, Comparable identId, Comparable peptideId);\\n    protected Peptide retrieve() throws Exception\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/task/impl/RetrievePeptideTask.java', 'methodCount': 2, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Create chromatogram browser to display chromatogram\n",
      "     */\n",
      "private void createChromatogramBrowser()\n",
      "**************************************\n",
      "private void createChromatogramBrowser() {\n",
      "    chromaBrowser = new ChromatogramBrowser();\n",
      "    Icon helpIcon = GUIUtilities.loadIcon(context.getProperty(\"help.icon.small\"));\n",
      "    String helpTooltip = context.getProperty(\"help.tooltip\");\n",
      "    PrideAction helpAction = new OpenHelpAction(null, helpIcon, helpTooltip);\n",
      "    helpAction.putValue(Action.SHORT_DESCRIPTION, helpTooltip);\n",
      "    AbstractButton button = (AbstractButton) chromaBrowser.getSidePane().addAction(helpAction, false);\n",
      "    CSH.setHelpIDString(button, \"help.mzgraph.chroma\");\n",
      "    button.addActionListener(new CSH.DisplayHelpFromSource(context.getMainHelpBroker()));\n",
      "}\n",
      "3231\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 349/1243 [11:59<28:20,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16506\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 350/1243 [12:00<27:04,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16509\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 351/1243 [12:02<26:10,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16529\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 352/1243 [12:04<25:29,  1.72s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16530\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 353/1243 [12:06<27:16,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16531\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 15, 'classCount': 1, 'methodsOnly': 'public class PrideInspectorContext {\\n    private static final Logger logger = LoggerFactory.getLogger(PrideInspector.class);;\\n    /**\\n * triggered when the visibility property of data source browser is changed\\n */\\npublic static final String LEFT_CONTROL_PANE_VISIBILITY = \"leftControlPaneVisible\";;\\n    /**\\n * DataAccessMonitor manages a list of\\n */\\nprivate final DataAccessMonitor dataAccessMonitor;;\\n    /**\\n * This map maintains a reference between DataAccessController and its data content display pane\\n */\\nprivate final Map<DataAccessController, JComponent> dataContentPaneCache;;\\n    /**\\n * This map maintains a reference between DataAccessController and its Summary pane\\n */\\nprivate final Map<DataAccessController, JComponent> dataSummaryPaneCache;;\\n    /**\\n * This map maintains a reference between DataAccessController and its action\\n */\\nprivate final Map<DataAccessController, Map<Class<? extends PrideAction>, PrideAction>> sharedActionCache;;\\n    /**\\n * data source browser visibility\\n */\\nprivate boolean leftControlPaneVisible;;\\n    /**\\n * Tracking all the summary report in a list model for each database access controller\\n */\\nprivate final Map<DataAccessController, ListModel> summaryReportTracker;;\\n    /**\\n * The main help set for PRIDE Inspector\\n */\\nprivate HelpSet mainHelpSet;;\\n    /**\\n * The main help broker\\n */\\nprivate HelpBroker mainHelpBroker;;\\n    /**\\n * The path to open file\\n */\\nprivate String openFilePath;;\\n    /**\\n * List of project details after login to PRIDE-R\\n */\\nprivate LoginRecord loginRecord;;\\n    /**\\n * welcome pane\\n */\\nprivate WelcomePane welcomePane = null;;\\n    /**\\n * database search pane\\n */\\nprivate PrideArchiveWSSearchPane prideArchiveWSSearchPane = null;;\\n    /**\\n * selected data transfer protocol\\n */\\nprivate DataTransferProtocol dataTransferProtocol;;\\n    public PrideInspectorContext();\\n    public synchronized DataTransferProtocol getDataTransferProtocol()\\n    public synchronized void setDataTransferProtocol(DataTransferProtocol dataTransferProtocol)\\n    public final DataAccessMonitor getDataAccessMonitor()\\n    public WelcomePane getWelcomePane()\\n    public void setWelcomePane(WelcomePane welcomePane)\\n    public PrideArchiveWSSearchPane getPrideArchiveWSSearchPane()\\n    public void setPrideArchiveWSSearchPane(PrideArchiveWSSearchPane prideArchiveWSSearchPane)\\n    public final List<DataAccessController> getControllers()\\n    public final int getNumberOfControllers()\\n    public final boolean isForegroundDataAccessController(DataAccessController controller)\\n    public final DataAccessController getForegroundDataAccessController()\\n    public final void setForegroundDataAccessController(DataAccessController controller, String message)\\n    public final void removeDataAccessController(DataAccessController controller, boolean cancelTasks, String message)\\n    public final void replaceSummaryReport(DataAccessController controller, DataAccessController replacement)\\n    public final void replaceDataAccessController(DataAccessController original, DataAccessController replacement, boolean cancelTasks, String message)\\n    public final void addDataAccessController(DataAccessController controller, String message)\\n    public final void addDataAccessController(DataAccessController controller, boolean foreground, String message)\\n    public final JComponent getDataContentPane(DataAccessController controller)\\n    public final JComponent getSummaryPane(DataAccessController controller)\\n    public final void addDataContentPane(DataAccessController controller, JComponent component)\\n    public final void addSummaryContentPane(DataAccessController controller, JComponent component)\\n    public final void removeDataContentPane(DataAccessController controller)\\n    public final synchronized PrideAction getPrideAction(DataAccessController controller, Class<? extends PrideAction> actionClass)\\n    public final synchronized void addPrideAction(DataAccessController controller, PrideAction action)\\n    public final void removePrideAction(DataAccessController controller, Class<? extends PrideAction> actionClass)\\n    public final void removePrideAction(DataAccessController controller)\\n    public final synchronized boolean isLeftControlPaneVisible()\\n    public void setLeftControlPaneVisible(boolean leftControlPaneVisible)\\n    private void createHelp()\\n    public HelpSet getMainHelpSet()\\n    public HelpBroker getMainHelpBroker()\\n    public String getOpenFilePath()\\n    public void setOpenFilePath(String openFilePath)\\n    public LoginRecord getLoginRecord()\\n    public void setLoginRecord(LoginRecord loginRecord)\\n    public synchronized ListModel getSummaryReportModel(DataAccessController controller)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/PrideInspectorContext.java', 'methodCount': 37, 'lineCount': 54}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'SummaryReportMessage': 'public class SummaryReportMessage {\\n    private final Type type;;\\n    private final String message;;\\n    private final String description;;\\n    public SummaryReportMessage(Type type, String message, String description);\\n    public Type getType()\\n    public String getMessage()\\n    public String getDescription()\\n}\\n'}\n",
      "/**\n",
      "     * Get the paint for the message border\n",
      "     *\n",
      "     * @param type message type\n",
      "     * @return Paint   border color\n",
      "     */\n",
      "private Paint getBorderPaint(SummaryReportMessage.Type type)\n",
      "**************************************\n",
      "private Paint getBorderPaint(SummaryReportMessage.Type type) {\n",
      "    switch(type) {\n",
      "        case SUCCESS:\n",
      "            return new Color(40, 175, 99);\n",
      "        case ERROR:\n",
      "            return new Color(215, 39, 41);\n",
      "        case WARNING:\n",
      "            return new Color(251, 182, 1);\n",
      "        case INFO:\n",
      "            return new Color(27, 106, 165);\n",
      "        default:\n",
      "            return new Color(27, 106, 165);\n",
      "    }\n",
      "}\n",
      "1308\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  28%|       | 354/1243 [12:07<26:40,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16582\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 355/1243 [12:10<28:01,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16668\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 356/1243 [12:12<28:51,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16707\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 357/1243 [12:13<27:04,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 358/1243 [12:15<28:28,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16953\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 92, 'classCount': 1, 'methodsOnly': 'public interface Constants {\\n    /**\\n * Labels\\n */\\nString GENERAL = \"General\";;\\n    String SPECTRUM = \"Spectrum\";;\\n    String CHROMATOGRAM = \"Chromatogram\";;\\n    String DATA_PROCESSING = \"Data Processing\";;\\n    String SOURCE_FILE = \"Source File\";;\\n    String SCAN = \"Scan\";;\\n    String SCAN_WINDOW = \"Scan Window\";;\\n    String PRECURSOR = \"Precursor\";;\\n    String PRODUCTS = \"Products\";;\\n    String PRODUCT = \"Product\";;\\n    String PROCESSING_METHOD = \"Method\";;\\n    String INDEX = \"Index\";;\\n    String DEFAULT_ARR_LEN = \"Default Array Length\";;\\n    String ORDER = \"Order\";;\\n    String SOFTWARE_ID = \"Software ID\";;\\n    String SOFTWARE_VERSION = \"Software Version\";;\\n    String SOURCE_FILE_ID = \"Source File ID\";;\\n    String SOURCE_FILE_NAME = \"Source File Name\";;\\n    String SOURCE_FILE_PATH = \"Source File Path\";;\\n    String EXTERNAL_SPECTRUM_REF = \"External Spectrum ID\";;\\n    String SPECTRUM_ID = \"Spectrum ID\";;\\n    String ISOLATION_WINDOW = \"Isolation Window\";;\\n    String SELECTED_ION = \"Selected Ion\";;\\n    String ACTIVATION = \"Activation\";;\\n    String ID = \"ID\";;\\n    String ACCESSION = \"Accession\";;\\n    String VERSION = \"Version\";;\\n    String TITLE = \"Title\";;\\n    String SHORT_LABEL = \"Short Label\";;\\n    String PUBLIC_DATE = \"Public Date\";;\\n    String CREATION_DATE = \"Creation Date\";;\\n    String NAME = \"Name\";;\\n    String PATH = \"Path\";;\\n    String PARAMETER = \"Parameter\";;\\n    String SCAN_SETTING_ID = \"Scan Setting ID\";;\\n    String SOURCE = \"Source\";;\\n    String ANALYZER = \"Analyzer\";;\\n    String DETECTOR = \"Detector\";;\\n    String TARGET = \"Target\";;\\n    String EMPTY = \"\";;\\n    String PROTOCOL_STEP = \"Step\";;\\n    String FULL_REFERENCE = \"Full Reference\";;\\n    String PUBMED = \"PubMed\";;\\n    String DOI = \"DOI\";;\\n    String DOT = \".\";;\\n    String COMMA = \",\";;\\n    String TAB = \"\\\\t\";;\\n    String TAB_SEP_FILE = \".tsv\";;\\n    String MZTAB_SEP_FILE = \".mztab\";;\\n    String MZML_FILE = \".mzml\";;\\n    String XML_FILE = \".xml\";;\\n    String GZIPPED_FILE = \".gz\";;\\n    String MZIDENT_FILE = \".mzid\";;\\n    String MZIDENT_FILE_LARGE = \".mzidentml\";;\\n    String MZXML_FILE = \".mzxml\";;\\n    String MGF_FILE = \".mgf\";;\\n    String DTA_FILE = \".dta\";;\\n    String MS2_FILE = \".ms2\";;\\n    String PKL_FILE = \".pkl\";;\\n    String APL_FILE = \".apl\";;\\n    String MZTAB_FILE = \".mztab\";;\\n    String CDF_FILE = \".cdf\";;\\n    String LINE_SEPARATOR = System.getProperty(\"line.separator\");;\\n    String UNKNOWN = \"Unknown\";;\\n    String FIT = \"Fuzzy Fit\";;\\n    String NOT_FIT = \"No Fit\";;\\n    String STRICT_FIT = \"Fit\";;\\n    String OVERLAP = \"Overlap\";;\\n    String PTM = \"PTM\";;\\n    String SELECTED = \"Selected\";;\\n    String VIEW = \"View\";;\\n    String PUBMED_URL_PERFIX = \"http://www.ncbi.nlm.nih.gov/pubmed/\";;\\n    String DOI_URL_PREFIX = \"http://dx.doi.org/\";;\\n    String OLS_URL_PREFIX = \"http://www.ebi.ac.uk/ontology-lookup/?termId=\";;\\n    String NOT_AVAILABLE = \"N/A\";;\\n    /**\\n * Quantification\\n */\\nString QUANTIFICATION_RATIO_CHAR = \"/\";;\\n    /**\\n * Colours\\n */\\nColor PTM_BACKGROUND_COLOUR = new Color(215, 39, 41, 100);;\\n    Color PTM_HIGHLIGHT_COLOUR = Color.yellow.darker();;\\n    Color STRICT_FIT_PEPTIDE_BACKGROUND_COLOUR = new Color(40, 175, 99, 100);;\\n    Color FIT_PEPTIDE_BACKGROUND_COLOUR = new Color(251, 182, 1, 100);;\\n    Color PEPTIDE_OVERLAP_COLOUR = new Color(40, 175, 99);;\\n    Color PEPTIDE_HIGHLIGHT_COLOUR = Color.yellow;;\\n    Color NOT_FIT_PEPTIDE_BACKGROUND_COLOUR = new Color(215, 39, 41, 100);;\\n    /**\\n * color for tables\\n */\\nColor ALTER_ROW_COLOUR = new Color(214, 241, 249);;\\n    Color ROW_SELECTION_BACKGROUD = new Color(193, 210, 238);;\\n    Color ROW_SELECTION_FOREGROUND = Color.black;;\\n    Color DELTA_MZ_WARNING = new Color(215, 39, 41, 100);;\\n    Color DELTA_MZ_NORMAL = new Color(40, 175, 99, 100);;\\n    /**\\n * Number Formatter\\n */\\nDecimalFormat LARGE_DECIMAL_NUMBER_FORMATTER = new DecimalFormat(\"0.###E0\");;\\n    NumberFormat DECIMAL_FORMATTER = NumberFormat.getInstance();;\\n    /**\\n * Scientific number threshold\\n */\\ndouble MAX_NON_SCIENTIFIC_NUMBER = 999;;\\n    double MIN_MON_SCIENTIFIC_NUMBER = 0.001;;\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Toolsuite,pride-inspector/src/main/java/uk/ac/ebi/pride/toolsuite/gui/utils/Constants.java', 'methodCount': 0, 'lineCount': 94}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{'AnnotatedProtein': 'public class AnnotatedProtein {\\n    public static final String PEPTIDE_SELECTION_PROP = \"selectedAnnotation\";;\\n    private List<PeptideAnnotation> annotations;;\\n    private PeptideAnnotation selectedAnnotation;;\\n    private PropertyChangeHelper propertyChangeHelper;;\\n    /**\\n * these peptides can be duplicated\\n */\\nprivate int numOfValidPeptides = -1;;\\n    /**\\n * these peptides must be valid as well\\n */\\nprivate int numOfUniquePeptides = -1;;\\n    /**\\n * the number of amino acids covered\\n */\\nprivate int numOfAminoAcidCovered = -1;;\\n    public AnnotatedProtein(Protein protein);\\n    public AnnotatedProtein(String accession, String name, STATUS status, String sequence);\\n    public AnnotatedProtein(String accession);\\n    public boolean isValidPeptideAnnotation(PeptideAnnotation annotation)\\n    public boolean isStrictValidPeptideAnnotation(PeptideAnnotation annotation)\\n    public void addAnnotation(PeptideAnnotation annotation)\\n    public void removeAnnotation(PeptideAnnotation annotation)\\n    public List<PeptideAnnotation> getAnnotations()\\n    public PeptideAnnotation getSelectedAnnotation()\\n    public void setSelectedAnnotation(PeptideAnnotation selectedAnnotation)\\n    public int getNumOfPeptides()\\n    public int getNumOfValidPeptides()\\n    public void setNumOfValidPeptides(int numOfValidPeptides)\\n    public int getNumOfUniquePeptides()\\n    public void setNumOfUniquePeptides(int numOfUniquePeptides)\\n    public Set<Integer> searchStartingPosition(PeptideAnnotation annotation)\\n    public double getSequenceCoverage()\\n    public int getNumOfAminoAcidCovered()\\n    private void populateCoverage()\\n    public void setNumOfAminoAcidCovered(int numOfAminoAcidCovered)\\n    public void addPropertyChangeListener(PropertyChangeListener listener)\\n    public void removePropertyChangeListener(PropertyChangeListener listener)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n'}\n",
      "/**\n",
      "     * This method is responsible for create a formatted and styled protein sequence\n",
      "     * <p/>\n",
      "     * 1. insert white space between a segment of the sequence (PROTEIN_SEGMENT_LENGTH)\n",
      "     * 2. insert tabs between segments of the sequence (PROTEIN_SEGMENT_GAP)\n",
      "     * 3. highlight the peptide annotations\n",
      "     * 4. highlight the ptm annotations\n",
      "     * 5. highlight the selected peptide\n",
      "     *\n",
      "     * @param protein protein to be converted to formatted sequence string\n",
      "     * @return AttributedString    formatted sequence string\n",
      "     */\n",
      "public static AttributedString build(AnnotatedProtein protein)\n",
      "**************************************\n",
      "/**\n",
      " * This method is responsible for create a formatted and styled protein sequence\n",
      " * <p/>\n",
      " * 1. insert white space between a segment of the sequence (PROTEIN_SEGMENT_LENGTH)\n",
      " * 2. insert tabs between segments of the sequence (PROTEIN_SEGMENT_GAP)\n",
      " * 3. highlight the peptide annotations\n",
      " * 4. highlight the ptm annotations\n",
      " * 5. highlight the selected peptide\n",
      " *\n",
      " * @param protein protein to be converted to formatted sequence string\n",
      " * @return AttributedString    formatted sequence string\n",
      " */\n",
      "public static AttributedString build(AnnotatedProtein protein) {\n",
      "    AttributedString formattedSequence = null;\n",
      "    if (protein != null) {\n",
      "        String sequence = protein.getSequenceString();\n",
      "        if (sequence != null && !\"\".equals(sequence.trim())) {\n",
      "            // add sequence segment gap\n",
      "            String gappedSequence = insertSegmentGapToSequence(sequence);\n",
      "            // create attributed string\n",
      "            formattedSequence = new AttributedString(gappedSequence);\n",
      "            // set overall font\n",
      "            formattedSequence.addAttribute(TextAttribute.FONT, DEFAULT_FONT);\n",
      "            formattedSequence.addAttribute(TextAttribute.FOREGROUND, DEFAULT_FOREGROUND);\n",
      "            // color code peptides\n",
      "            addPeptideAnnotations(protein, formattedSequence);\n",
      "        }\n",
      "    }\n",
      "    return formattedSequence;\n",
      "}\n",
      "2234\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 359/1243 [12:17<29:20,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16956\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 360/1243 [12:20<29:32,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16987\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 361/1243 [12:22<29:55,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16988\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 362/1243 [12:23<28:21,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17099\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 363/1243 [12:26<29:38,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17218\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Store a collection of protein details\n",
      "     *\n",
      "     * @param proteins a collection of proteins\n",
      "     */\n",
      "public void addProteinDetails(Collection<Protein> proteins)\n",
      "**************************************\n",
      "public void addProteinDetails(Collection<Protein> proteins) {\n",
      "    if (proteins == null) {\n",
      "        String msg = \"Collection of proteins cannot be null\";\n",
      "        logger.error(msg);\n",
      "        throw new IllegalArgumentException(msg);\n",
      "    }\n",
      "    for (Protein protein : proteins) {\n",
      "        proteinNameCache.put(new Element(protein.getAccession(), protein));\n",
      "    }\n",
      "}\n",
      "151\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 364/1243 [12:28<29:54,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19171\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 365/1243 [12:29<27:58,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19172\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  29%|       | 366/1243 [12:31<29:04,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19176\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 367/1243 [12:33<29:11,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19515\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 368/1243 [12:36<29:48,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19519\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidInputFileException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 8737568813714627096L;;\\n    public InvalidInputFileException(String message);\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/datastructures/InvalidInputFileException.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 10, 'classCount': 1, 'methodsOnly': 'public class VariantPoolHeavy {\\n    //\\tprivate static ArrayList<String> usedPoolIDs = new ArrayList<String>();\\n//\\tprivate static ArrayList<String> generatedPoolIDs = new ArrayList<String>();\\nprivate static Logger logger = Logger.getLogger(VariantPoolHeavy.class);;\\n    private HashMap<String, VariantContext> hMap;;\\n    private HashMap<String, HashMap<String, VariantContext>> hMapChrPos;;\\n    //\\tprivate TreeMap<String, VariantContext> tMap;\\nprivate TreeSet<String> contigs;;\\n    private String currVarKey;;\\n    private Iterator<String> varKeyIter;;\\n    private Boolean hasGenotypeData;;\\n    private int potentialMatchingIndelAlleles;;\\n    private int potentialMatchingIndelRecords;;\\n    /* This boolean will track whether the VCF has been fully read into the\\n\\t * VariantPoolHeavy object. Once it has been, then getNextVar should use\\n\\t * the set of variants read into this object.\\n\\t */\\nprivate boolean vcfFileFullyParsed = false;;\\n    public VariantPoolHeavy(boolean addChr, String poolID) throws IOException;\\n    public VariantPoolHeavy(VariantPoolHeavy vp) throws IOException;\\n    public VariantPoolHeavy(String filePath, String poolID, boolean addChr) throws IOException;\\n    public VariantPoolHeavy(File file, String poolID, boolean addChr) throws IOException;\\n    public VariantPoolHeavy(File file, String poolID, boolean requireIndex, boolean addChr) throws IOException;\\n    public VariantPoolHeavy(String inputString, boolean requireIndex, boolean addChr) throws InvalidInputFileException, IOException;\\n    private void init() throws IOException\\n    public VariantContext getVariant(String chr, int pos, String ref)\\n    private HashMap<String, VariantContext> getVariantsByChrPos(String chr, int pos)\\n    public VariantContext getVariant(String key)\\n    public int getNumVarRecords()\\n    public TreeSet<String> getContigs()\\n    private Iterator<String> getVariantIterator()\\n    public VariantContext getNextVar() throws IOException\\n    public boolean hasGenotypeData()\\n    public int getPotentialMatchingIndelAlleles()\\n    public int getPotentialMatchingIndelRecords()\\n    public void setPotentialMatchingIndelAlleles(int count)\\n    public void setPotentialMatchingIndelRecords(int count)\\n    private void addContig(String contig)\\n    private void parseVCF(String filename, boolean requireIndex) throws IOException\\n    public void addVariant(VariantContext v, boolean union)\\n    public void updateVariant(String key, VariantContext newVar)\\n    private String generateChrString(String chr)\\n    private VariantContext buildNewVariantWithChr(String chr, VariantContext var)\\n    public void changeSampleNames(ArrayList<String> newSampleNames)\\n    public static Genotype renameGenotypeForSample(String sampleName, Genotype geno)\\n    public static Genotype changeAllelesForGenotype(Genotype geno, ArrayList<Allele> newAlleles)\\n    public static VariantContext buildVariant(VariantContext var, List<Allele> alleles, ArrayList<Genotype> genos)\\n    public void generateBasicHeader(SAMSequenceDictionary refDict, Set<String> sampleNames)\\n    public static void printVariantPool(String file, VariantPoolHeavy vp, File refDict, SupportedFileType fileType, boolean repairHeader) throws URISyntaxException, FileNotFoundException\\n    public static void printVariantPool(String fileName, String outputDirectory, VariantPoolHeavy vp, File refDict, SupportedFileType fileType, boolean repairHeader) throws URISyntaxException, FileNotFoundException\\n    private static void printVariantPoolToVCF(File file, VariantPoolHeavy vp, File refDict, boolean repairHeader) throws FileNotFoundException\\n    private static void addMissingHeaderLineAndWriteVariant(Exception e, VariantPoolHeavy vp, VariantContextWriter writer, VariantContext vc)\\n    public int getOverlappingIndelAlleleCount(VariantContext var)\\n    public VariantContext getOverlappingIndel(String chr, int pos, int indelLength, AltType type)\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/datastructures/VariantPoolHeavy.java', 'methodCount': 36, 'lineCount': 48}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class VariantPoolLight {\\n    public VariantPoolLight(boolean addChr, String poolID);\\n    public VariantPoolLight(File file, String poolID, boolean requireIndex, boolean addChr) throws IOException;\\n    public VariantPoolLight(String inputString, boolean requireIndex, boolean addChr) throws InvalidInputFileException, IOException;\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/datastructures/VariantPoolLight.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidOperationException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 492884518173784168L;;\\n    public InvalidOperationException(String message);\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/InvalidOperationException.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Operation {\\n    //\\tprivate String operationPatternString =\\n//\\t\\t\\t\"^((\\\\\\\\w+)=)?[iIuUcC]\\\\\\\\[(\\\\\\\\w+(\\\\\\\\[\\\\\\\\w+(,\\\\\\\\w+)*\\\\\\\\])?)(:\\\\\\\\w+(\\\\\\\\[\\\\\\\\w+(,\\\\\\\\w+)*\\\\\\\\])?)*\\\\\\\\]$\";\\n//\\tprivate Pattern operationPattern;\\nprivate String operationString;;\\n    private String operID;;\\n    private Operator operator;;\\n    private static ArrayList<String> usedOperationIDs = new ArrayList<String>();;\\n    private static ArrayList<String> generatedOperationIDs = new ArrayList<String>();;\\n    public Operation(String operationString, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public Operation(String operationString, String operID, Operator op);\\n    public String getOperationID()\\n    public abstract Collection<String> getAllPoolIDs()\\n    public Operator getOperator()\\n    private static String generateOperationID()\\n    private static void addOperationID(String operID)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/Operation.java', 'methodCount': 8, 'lineCount': 15}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class OperationFactory {\\n    public static Operation createOperation(String operation, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException\\n    private static Operation parseOperation(String operation, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException\\n    private static Operation createProperOperation(String operationString, String operID, Operator op, ArrayList<SamplePool> samplePools) throws InvalidOperationException\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/OperationFactory.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/varstats/AltType.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 12}\n",
      "{}\n",
      "/**\n",
      "\t * Count the differences between a ref and alt\n",
      "\t * \n",
      "\t * @param ref\n",
      "\t * @param alt\n",
      "\t * @return\n",
      "\t */\n",
      "public static int getDiffCount(Allele ref, Allele alt)\n",
      "**************************************\n",
      "public static int getDiffCount(Allele ref, Allele alt) {\n",
      "    String refNucs = ref.getBaseString();\n",
      "    String altNucs = alt.getBaseString();\n",
      "    if (refNucs.length() != altNucs.length()) {\n",
      "        System.err.println(\"Error: Running getDiffCount on two alleles of different lengths.\");\n",
      "        return -1;\n",
      "    }\n",
      "    int count = 0;\n",
      "    for (int i = 0; i < refNucs.length(); i++) {\n",
      "        if (refNucs.charAt(i) != altNucs.charAt(i)) {\n",
      "            count++;\n",
      "        }\n",
      "    }\n",
      "    return count;\n",
      "}\n",
      "1999\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 369/1243 [12:38<30:13,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19556\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 370/1243 [12:40<30:28,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19560\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 371/1243 [12:42<30:36,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19567\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 372/1243 [12:44<31:13,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19568\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 373/1243 [12:47<32:21,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19580\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class SamplePool {\\n    private static TreeSet<String> usedPoolIDs = new TreeSet<String>();;\\n    /* This poolID must match the poolID of its associated VariantPool object\\n\\t * since all samples come from the variants anyway.\\n\\t */\\nprivate String poolID;;\\n    private TreeSet<String> samples;;\\n    // Keep track of all variantPools defined on the command line\\nprivate static TreeSet<String> definedVariantPoolIDs;;\\n    // Match on the expected pattern for a sample pool (e.g. f1[s1,s3] )\\nprivate static Pattern samplePoolPattern = Pattern.compile(\"^(\\\\\\\\w+)(\\\\\\\\[(.+)\\\\\\\\])*$\");;\\n    public SamplePool(String pool, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public SamplePool();\\n    public TreeSet<String> getSamples()\\n    public String getPoolID()\\n    public static TreeSet<String> getAllPoolIDs()\\n    public void setPoolID(String poolID)\\n    public void addSamples(TreeSet<String> samples)\\n    public void addSample(String sample)\\n    private void parseSamplePool(String pool, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/datastructures/SamplePool.java', 'methodCount': 9, 'lineCount': 16}, {'fieldCount': 10, 'classCount': 1, 'methodsOnly': 'public class VariantPoolHeavy {\\n    //\\tprivate static ArrayList<String> usedPoolIDs = new ArrayList<String>();\\n//\\tprivate static ArrayList<String> generatedPoolIDs = new ArrayList<String>();\\nprivate static Logger logger = Logger.getLogger(VariantPoolHeavy.class);;\\n    private HashMap<String, VariantContext> hMap;;\\n    private HashMap<String, HashMap<String, VariantContext>> hMapChrPos;;\\n    //\\tprivate TreeMap<String, VariantContext> tMap;\\nprivate TreeSet<String> contigs;;\\n    private String currVarKey;;\\n    private Iterator<String> varKeyIter;;\\n    private Boolean hasGenotypeData;;\\n    private int potentialMatchingIndelAlleles;;\\n    private int potentialMatchingIndelRecords;;\\n    /* This boolean will track whether the VCF has been fully read into the\\n\\t * VariantPoolHeavy object. Once it has been, then getNextVar should use\\n\\t * the set of variants read into this object.\\n\\t */\\nprivate boolean vcfFileFullyParsed = false;;\\n    public VariantPoolHeavy(boolean addChr, String poolID) throws IOException;\\n    public VariantPoolHeavy(VariantPoolHeavy vp) throws IOException;\\n    public VariantPoolHeavy(String filePath, String poolID, boolean addChr) throws IOException;\\n    public VariantPoolHeavy(File file, String poolID, boolean addChr) throws IOException;\\n    public VariantPoolHeavy(File file, String poolID, boolean requireIndex, boolean addChr) throws IOException;\\n    public VariantPoolHeavy(String inputString, boolean requireIndex, boolean addChr) throws InvalidInputFileException, IOException;\\n    private void init() throws IOException\\n    public VariantContext getVariant(String chr, int pos, String ref)\\n    private HashMap<String, VariantContext> getVariantsByChrPos(String chr, int pos)\\n    public VariantContext getVariant(String key)\\n    public int getNumVarRecords()\\n    public TreeSet<String> getContigs()\\n    private Iterator<String> getVariantIterator()\\n    public VariantContext getNextVar() throws IOException\\n    public boolean hasGenotypeData()\\n    public int getPotentialMatchingIndelAlleles()\\n    public int getPotentialMatchingIndelRecords()\\n    public void setPotentialMatchingIndelAlleles(int count)\\n    public void setPotentialMatchingIndelRecords(int count)\\n    private void addContig(String contig)\\n    private void parseVCF(String filename, boolean requireIndex) throws IOException\\n    public void addVariant(VariantContext v, boolean union)\\n    public void updateVariant(String key, VariantContext newVar)\\n    private String generateChrString(String chr)\\n    private VariantContext buildNewVariantWithChr(String chr, VariantContext var)\\n    public void changeSampleNames(ArrayList<String> newSampleNames)\\n    public static Genotype renameGenotypeForSample(String sampleName, Genotype geno)\\n    public static Genotype changeAllelesForGenotype(Genotype geno, ArrayList<Allele> newAlleles)\\n    public static VariantContext buildVariant(VariantContext var, List<Allele> alleles, ArrayList<Genotype> genos)\\n    public void generateBasicHeader(SAMSequenceDictionary refDict, Set<String> sampleNames)\\n    public static void printVariantPool(String file, VariantPoolHeavy vp, File refDict, SupportedFileType fileType, boolean repairHeader) throws URISyntaxException, FileNotFoundException\\n    public static void printVariantPool(String fileName, String outputDirectory, VariantPoolHeavy vp, File refDict, SupportedFileType fileType, boolean repairHeader) throws URISyntaxException, FileNotFoundException\\n    private static void printVariantPoolToVCF(File file, VariantPoolHeavy vp, File refDict, boolean repairHeader) throws FileNotFoundException\\n    private static void addMissingHeaderLineAndWriteVariant(Exception e, VariantPoolHeavy vp, VariantContextWriter writer, VariantContext vc)\\n    public int getOverlappingIndelAlleleCount(VariantContext var)\\n    public VariantContext getOverlappingIndel(String chr, int pos, int indelLength, AltType type)\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/datastructures/VariantPoolHeavy.java', 'methodCount': 36, 'lineCount': 48}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ComplementOperation {\\n    private ArrayList<SamplePool> samplePools;;\\n    private ArrayList<String> allPoolIDs;;\\n    public ComplementOperation(String operationString, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public ComplementOperation(String operationString, String operID, Operator op, ArrayList<SamplePool> samplePools);\\n    public ArrayList<SamplePool> getSamplePools()\\n    public SamplePool getSamplePool(String poolID)\\n    public ArrayList<String> getAllPoolIDs()\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/ComplementOperation.java', 'methodCount': 5, 'lineCount': 9}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class IntersectOperation {\\n    private HashMap<String, SamplePool> samplePools;;\\n    public IntersectOperation(String operationString, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public IntersectOperation(String operationString, String operID, Operator op, HashMap<String, SamplePool> samplePools);\\n    public HashMap<String, SamplePool> getSamplePools()\\n    public SamplePool getSamplePool(String poolID)\\n    public Set<String> getAllPoolIDs()\\n    public void addSamplePool(SamplePool sp)\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/IntersectOperation.java', 'methodCount': 6, 'lineCount': 9}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class InvalidOperationException {\\n    /**\\n */\\nprivate static final long serialVersionUID = 492884518173784168L;;\\n    public InvalidOperationException(String message);\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/InvalidOperationException.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Operation {\\n    //\\tprivate String operationPatternString =\\n//\\t\\t\\t\"^((\\\\\\\\w+)=)?[iIuUcC]\\\\\\\\[(\\\\\\\\w+(\\\\\\\\[\\\\\\\\w+(,\\\\\\\\w+)*\\\\\\\\])?)(:\\\\\\\\w+(\\\\\\\\[\\\\\\\\w+(,\\\\\\\\w+)*\\\\\\\\])?)*\\\\\\\\]$\";\\n//\\tprivate Pattern operationPattern;\\nprivate String operationString;;\\n    private String operID;;\\n    private Operator operator;;\\n    private static ArrayList<String> usedOperationIDs = new ArrayList<String>();;\\n    private static ArrayList<String> generatedOperationIDs = new ArrayList<String>();;\\n    public Operation(String operationString, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public Operation(String operationString, String operID, Operator op);\\n    public String getOperationID()\\n    public abstract Collection<String> getAllPoolIDs()\\n    public Operator getOperator()\\n    private static String generateOperationID()\\n    private static void addOperationID(String operID)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/Operation.java', 'methodCount': 8, 'lineCount': 15}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class UnionOperation {\\n    private HashMap<String, SamplePool> samplePools;;\\n    public UnionOperation(String operationString, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public UnionOperation(String operationString, String operID, Operator op, HashMap<String, SamplePool> samplePools);\\n    public HashMap<String, SamplePool> getSamplePools()\\n    public SamplePool getSamplePool(String poolID)\\n    public Set<String> getAllPoolIDs()\\n    public void addSamplePool(SamplePool sp)\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/setoperator/operation/UnionOperation.java', 'methodCount': 6, 'lineCount': 9}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class UtilityBelt {\\n    private final static int MAX_ALLELE_SIZE_FOR_NON_SV = 150;;\\n    public UtilityBelt();\\n    public static String roundDoubleToString(double d)\\n    public static double round(double unrounded, int precision, int roundingMode)\\n    public static TreeMap<String, VariantPoolHeavy> createHeavyVariantPools(List<String> inputFiles, boolean addChr) throws InvalidInputFileException, InvalidOperationException, IOException\\n    public static TreeMap<String, VariantPoolLight> createLightVariantPools(List<String> inputFiles, boolean addChr) throws InvalidInputFileException, InvalidOperationException, IOException\\n    public static ArrayList<VariantPoolHeavy> getAssociatedVariantPoolsAsArrayList(Operation op, TreeMap<String, VariantPoolHeavy> vps)\\n    public static int getSmallestLength(TreeSet<String> alleles)\\n    public static int getLargestLength(TreeSet<String> alleles)\\n    public static double getAverageLength(TreeSet<String> alleles)\\n    public static AltType determineAltType(Allele ref, Allele alt)\\n    public static boolean altTypeIsIndel(AltType type)\\n    public static int getDiffCount(Allele ref, Allele alt)\\n    public static void printErrorUsageHelpAndExit(ArgumentParser parser, Logger logger, Exception e)\\n    public static void printErrorUsageAndExit(ArgumentParser parser, Logger logger, Exception e)\\n    public static void printUsageHelpAndExit(ArgumentParser parser)\\n    public static ArrayList<Operation> createOperations(ArrayList<Object> operations, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException\\n    public static Object getProperty(String prop)\\n    public static String getHGREF()\\n}\\n', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/utilitybelt/UtilityBelt.java', 'methodCount': 18, 'lineCount': 21}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/mebbert,VariantToolChest/src/main/java/vtc/tools/varstats/AltType.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 15}\n",
      "{'SamplePool': 'public class SamplePool {\\n    private static TreeSet<String> usedPoolIDs = new TreeSet<String>();;\\n    /* This poolID must match the poolID of its associated VariantPool object\\n\\t * since all samples come from the variants anyway.\\n\\t */\\nprivate String poolID;;\\n    private TreeSet<String> samples;;\\n    // Keep track of all variantPools defined on the command line\\nprivate static TreeSet<String> definedVariantPoolIDs;;\\n    // Match on the expected pattern for a sample pool (e.g. f1[s1,s3] )\\nprivate static Pattern samplePoolPattern = Pattern.compile(\"^(\\\\\\\\w+)(\\\\\\\\[(.+)\\\\\\\\])*$\");;\\n    public SamplePool(String pool, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException;\\n    public SamplePool();\\n    public TreeSet<String> getSamples()\\n    public String getPoolID()\\n    public static TreeSet<String> getAllPoolIDs()\\n    public void setPoolID(String poolID)\\n    public void addSamples(TreeSet<String> samples)\\n    public void addSample(String sample)\\n    private void parseSamplePool(String pool, TreeMap<String, VariantPoolHeavy> variantPools) throws InvalidOperationException\\n}\\n'}\n",
      "/**\n",
      "\t * Determine which samples are missing\n",
      "\t * @param gc\n",
      "\t * @param sp\n",
      "\t * @return\n",
      "\t */\n",
      "private ArrayList<String> getMissingSamples(GenotypesContext gc, SamplePool sp)\n",
      "**************************************\n",
      "private ArrayList<String> getMissingSamples(GenotypesContext gc, SamplePool sp) {\n",
      "    Iterator<String> sampIT = sp.getSamples().iterator();\n",
      "    String samp;\n",
      "    ArrayList<String> missingSamps = new ArrayList<String>();\n",
      "    while (sampIT.hasNext()) {\n",
      "        samp = sampIT.next();\n",
      "        if (!gc.containsSample(samp)) {\n",
      "            missingSamps.add(samp);\n",
      "        }\n",
      "    }\n",
      "    return missingSamps;\n",
      "}\n",
      "3130\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 374/1243 [12:48<29:47,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19622\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 375/1243 [12:50<30:14,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19623\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 376/1243 [12:53<30:00,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19626\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 377/1243 [12:54<28:19,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19640\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 378/1243 [12:57<30:11,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19644\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Get the number of Substitution PTMs for an specific Protein Identification\n",
      "     *\n",
      "     * @param ident ID of the Protein Identification\n",
      "     * @return Total number of Substitution PTMs\n",
      "     */\n",
      "public static int getNumberOfSubstitutionPTMs(Protein ident)\n",
      "**************************************\n",
      "public static int getNumberOfSubstitutionPTMs(Protein ident) {\n",
      "    int cnt = 0;\n",
      "    List<Peptide> peptides = ident.getPeptides();\n",
      "    for (Peptide peptide : peptides) {\n",
      "        PeptideSequence peptideSequence = peptide.getPeptideSequence();\n",
      "        if (peptideSequence != null) {\n",
      "            List<SubstitutionModification> mods = peptideSequence.getSubstitutionModifications();\n",
      "            cnt += mods.size();\n",
      "        }\n",
      "    }\n",
      "    return cnt;\n",
      "}\n",
      "195\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  30%|       | 379/1243 [12:59<32:48,  2.28s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19703\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 380/1243 [13:02<32:32,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19704\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 381/1243 [13:03<29:53,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19712\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 382/1243 [13:05<27:53,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19714\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 383/1243 [13:07<28:34,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19765\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Convert param group\n",
      "     *\n",
      "     * @param oldParam jmzML param group\n",
      "     * @param <T>      any jmzml objects which extends param group\n",
      "     * @return ParamGroup  param group\n",
      "     */\n",
      "public static ParamGroup transformParamGroup(T oldParam)\n",
      "**************************************\n",
      "public static <T extends uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.Param> ParamGroup transformParamGroup(T oldParam) {\n",
      "    ParamGroup newParamGroup = null;\n",
      "    if (oldParam != null) {\n",
      "        List<CvParam> cvParams = new ArrayList<CvParam>();\n",
      "        List<UserParam> userParams = new ArrayList<UserParam>();\n",
      "        transformCvParam(cvParams, oldParam.getCvParams());\n",
      "        transformUserParam(userParams, oldParam.getUserParams());\n",
      "        newParamGroup = new ParamGroup(cvParams, userParams);\n",
      "    }\n",
      "    return newParamGroup;\n",
      "}\n",
      "226\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 384/1243 [13:09<29:07,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19772\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 385/1243 [13:11<29:18,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19773\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 386/1243 [13:13<28:00,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19783\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 387/1243 [13:16<31:35,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19786\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|       | 388/1243 [13:17<29:11,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19787\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class DataAccessUtilities {\\n    public static final Logger logger = LoggerFactory.getLogger(DataAccessUtilities.class);;\\n    private DataAccessUtilities();\\n    public static List<String> getTaxonomy(ExperimentMetaData metaData)\\n    public static String getProjectName(ExperimentMetaData metaData)\\n    public static int getNumberOfPeaks(Spectrum spectrum)\\n    public static int getMsLevel(Spectrum spectrum)\\n    public static Integer getPrecursorChargeParamGroup(ParamGroup paramGroup)\\n    private static Double getSelectedIonCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static double getPrecursorMz(Spectrum spectrum)\\n    public static double getPrecursorMz(ParamGroup paramGroup)\\n    public static double getPrecursorIntensity(Spectrum spectrum)\\n    public static double getSumOfIntensity(Spectrum spectrum)\\n    public static Double getSelectedIonCharge(Precursor precursor, int index)\\n    public static Double getSelectedIonMz(Precursor precursor, int index)\\n    public static Double getSelectedIonIntensity(Precursor precursor, int index)\\n    private static Double getSelectedIonCvParamValue(Precursor precursor, int index, CvTermReference... refs)\\n    private static Double getSelectedCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static Peptide getPeptide(Protein ident, int index)\\n    public static QuantPeptide getQuantPeptide(Protein ident, int index)\\n    public static int getNumberOfPeptides(Protein ident)\\n    public static int getNumberOfUniquePeptides(Protein ident)\\n    public static boolean hasFragmentIon(Spectrum spectrum)\\n    public static boolean hasFragmentIon(Peptide peptide)\\n    public static int getNumberOfPTMs(Protein ident)\\n    public static int getNumberOfSubstitutionPTMs(Protein ident)\\n    public static int getNumberOfPTMs(Peptide peptide)\\n    public static int getNumberOfSubstitutionPTMs(Peptide peptide)\\n    public static List<SearchEngineCvTermReference> getSearchEngineTypes(ParamGroup paramGroup)\\n    public static Score getScore(ParamGroup params)\\n    public static List<SearchEngineCvTermReference> getSearchEngineCvTermReference(ParamGroup paramGroup)\\n    public static List<SearchEngineScoreCvTermReference> getSearchEngineScoresCvTermReference(ParamGroup paramGroup)\\n    public static List<CvParam> getCvParam(ParamGroup paramGroup, String cvLabel, String accession)\\n    public static List<CvParam> getCvParam(String name, String cvLabel, String accession, String value)\\n    public static List<Parameter> getParamByName(ParamGroup paramGroup, String name)\\n    public static Map<PeptideEvidence, List<Peptide>> getPeptideEvidence(List<Peptide> peptides)\\n    public static Integer getPrecursorCharge(List<Precursor> precursors)\\n    public static Integer getProteinCoverage(Protein protein)\\n    private static Set<Integer> searchStartingPosition(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isStrictValidPeptideAnnotation(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isValidPeptideAnnotation(Protein protein, Peptide peptide)\\n    public static String getRetentionTime(SpectrumIdentificationItem oldSpectrumIdentification)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessUtilities.java', 'methodCount': 40, 'lineCount': 43}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class BinaryDataUtils {\\n    private static final Logger logger = LoggerFactory.getLogger(BinaryDataUtils.class);;\\n    private BinaryDataUtils();\\n    public static Number[] toNumberArray(byte[] byteArray, CvTermReference dataType, ByteOrder order)\\n    public static double[] toDoubleArray(byte[] arr, CvTermReference dataType, ByteOrder order)\\n    private static int getNumOfByte(CvTermReference dataType)\\n    public static byte[] decompress(byte[] compressedData)\\n    public static byte[] compress(byte[] uncompressedData)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/BinaryDataUtils.java', 'methodCount': 6, 'lineCount': 9}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class CvUtilities {\\n    public static CvParam getCVTermFromCvReference(CvTermReference cvTerm, String value)\\n    public static CvParam getQuantTermFromQuantReference(QuantCvTermReference quantTerm, String value)\\n    public static String getMailFromCvParam(ParamGroup person)\\n    public static boolean isDecoyHit(ParamGroup paramGroup)\\n    public static String getValueFromParmGroup(List<CvParam> cvParams, String accession)\\n    public static boolean isAccessionDecoy(Protein identification)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/CvUtilities.java', 'methodCount': 6, 'lineCount': 8}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class PRIDEUtils {\\n    public static final String DUP_PROTEINS_SEARCH_ENGINES = \"duplicated_proteins_search_engines\";;\\n    public static final String DUP_PROTEINS_SEARCH_ENGINES_SCORES = \"duplicated_proteins_search_engines_scores\";;\\n    public static final String DUP_PROTEINS_BEST_SEARCH_ENGINES_SCORE = \"duplicated_proteins_best_search_engines_score\";;\\n    public static final String DUP_PROTEINS_HAD_QUANT = \"duplicated_proteins_had_quantification\";;\\n    public static final String NUM_MERGE_PROTEINS = \"num_merge_proteins\";;\\n    public static final List<String> PREFIX_PRIDE_DECOY_ENTRIES = Arrays.asList(\"DECOY_\", \"#C#\", \"#DECOY#\", \"###REV###\", \"REV_\", \"REVERSE_\", \"##REV\", \"DECOY_REV\", \"RANDOM_\", \"###RND###\", \"##RND\", \"REV\");;\\n    public static final List<String> POSTFIX_PRIDE_DECOY_ENTRIES = Arrays.asList(\"_REVERSED\", \"-DECOY\", \"RANDOM_\", \"|RND\");;\\n    public static final List<String> MIDDLE_PRIDE_DECOY_ENTRIES = Arrays.asList(\"RANDOM_\", \"REV_\", \"_REVERSED\");;\\n    public static CvParam convertSoftwareNameToCvPram(String name, String version)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/PRIDEUtils.java', 'methodCount': 1, 'lineCount': 11}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Convert precursor list\n",
      "     *\n",
      "     * @param rawPrecursors pride xml precursor list\n",
      "     * @return List<Precursor> a list of precursors\n",
      "     */\n",
      "public static List<Precursor> transformPrecursorList(uk.ac.ebi.pride.jaxb.model.PrecursorList rawPrecursors)\n",
      "**************************************\n",
      "public static List<Precursor> transformPrecursorList(uk.ac.ebi.pride.jaxb.model.PrecursorList rawPrecursors) {\n",
      "    List<Precursor> precursors = null;\n",
      "    if (rawPrecursors != null) {\n",
      "        precursors = new ArrayList<Precursor>();\n",
      "        for (uk.ac.ebi.pride.jaxb.model.Precursor rawPrecursor : rawPrecursors.getPrecursor()) {\n",
      "            precursors.add(transformPrecursor(rawPrecursor));\n",
      "        }\n",
      "    }\n",
      "    return precursors;\n",
      "}\n",
      "1769\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|      | 389/1243 [13:20<30:49,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19788\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|      | 390/1243 [13:22<31:21,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19803\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  31%|      | 391/1243 [13:24<30:39,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19883\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 392/1243 [13:26<31:04,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19890\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 393/1243 [13:29<31:34,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19891\n",
      "{'classCount': 6, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface DataAccessController {\\n    public String getUid()\\n    public String getName()\\n    public Type getType()\\n    public Collection<ContentCategory> getContentCategories()\\n    public Object getSource()\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessController.java', 'methodCount': 6, 'lineCount': 8}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessMode.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class MzTabCachingStrategy {\\n    private static final int INIT_BIG_HASH = 10000;;\\n    private static final Logger logger = LoggerFactory.getLogger(MzTabCachingStrategy.class);;\\n    public void cache()\\n    private boolean hasProteinGroup(MzTabUnmarshallerAdaptor unmarshaller)\\n    private void cacheProteinGroups(MzTabUnmarshallerAdaptor unmarshaller)\\n    private Map<String, String> cacheSpectrumIds(MzTabUnmarshallerAdaptor unmarshaller)\\n    private void cacheProteins(MzTabUnmarshallerAdaptor unmarshaller)\\n    private void cacheQuantPeptideIds(MzTabUnmarshallerAdaptor unmarshaller, Map<String, String> proteinAccession)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/strategy/MzTabCachingStrategy.java', 'methodCount': 6, 'lineCount': 10}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class MzTabTransformer {\\n    private static final Logger logger = LoggerFactory.getLogger(MzTabTransformer.class);;\\n    private final static String PROTOCOL_ID = \"protocol1\";;\\n    public static List<SourceFile> transformSourceFiles(Map<Integer, uk.ac.ebi.pride.jmztab.model.MsRun> msRunMap)\\n    public static SourceFile transformSourceFile(MsRun msRun)\\n    public static Collection<Person> transformContactToPersons(Map<Integer, uk.ac.ebi.pride.jmztab.model.Contact> contacts)\\n    public static Person transformContactToPerson(Contact contact)\\n    public static List<Organization> transformContactToOrganization(Map<Integer, uk.ac.ebi.pride.jmztab.model.Contact> contacts)\\n    public static Organization transformContactToOrganization(Contact contact)\\n    public static List<Sample> transformSamples(Map<Integer, uk.ac.ebi.pride.jmztab.model.Sample> oldSamples, Metadata metadata, boolean hasQuantitation)\\n    private static Sample transformSample(uk.ac.ebi.pride.jmztab.model.Sample oldSample)\\n    public static List<Software> transformSoftwares(Map<Integer, uk.ac.ebi.pride.jmztab.model.Software> dataSoftwares)\\n    public static Software transformSoftware(uk.ac.ebi.pride.jmztab.model.Software oldSoftware)\\n    public static Collection<Reference> transformReferences(Map<Integer, uk.ac.ebi.pride.jmztab.model.Publication> publications)\\n    public static ParamGroup transformAdditional(List<Param> additionalParams)\\n    public static Protein transformIdentification(uk.ac.ebi.pride.jmztab.model.Protein rawIdent, Integer rawIndex, Map<String, uk.ac.ebi.pride.jmztab.model.PSM> rawPsms, Map<String, uk.ac.ebi.pride.jmztab.model.Peptide> rawPeptides, uk.ac.ebi.pride.jmztab.model.Metadata metadata, boolean hasQuantitation)\\n    public static Protein transformIdent(uk.ac.ebi.pride.jmztab.model.Protein rawIdent, Integer rawIndex, Map<String, uk.ac.ebi.pride.jmztab.model.PSM> rawPsms, Map<String, uk.ac.ebi.pride.jmztab.model.Peptide> rawPeptides, uk.ac.ebi.pride.jmztab.model.Metadata metadata, boolean hasQuantitation)\\n    private static QuantScore transformQuantitationParams(uk.ac.ebi.pride.jmztab.model.Protein rawIdent, Metadata metadata)\\n    private static QuantScore transformQuantitationParams(uk.ac.ebi.pride.jmztab.model.Peptide rawIdent, Metadata metadata)\\n    public static Peptide transformPeptide(uk.ac.ebi.pride.jmztab.model.PSM rawPeptide, DBSequence dbSequence, Comparable index, Metadata metadata, int indexPeptide)\\n    private static CvParam transformPSMPrecursorMZ(PSM rawPeptide, Metadata metadata)\\n    public static QuantPeptide transformQuantPeptide(uk.ac.ebi.pride.jmztab.model.Peptide rawPeptide, DBSequence dbSequence, Comparable index, Metadata metadata, int indexPeptide)\\n    private static List<CvParam> transformPSMSearchEngineScoreCvTerm(PSM rawPeptide, Metadata metadata)\\n    private static List<CvParam> transformPeptideSearchEngineScoreCvTerm(uk.ac.ebi.pride.jmztab.model.Peptide rawPeptide, MsRun msRun, Metadata metadata)\\n    private static List<Modification> transformModification(List<uk.ac.ebi.pride.jmztab.model.Modification> rawMods, Metadata metadata)\\n    private static String getAccesion(uk.ac.ebi.pride.jmztab.model.Modification rawMod)\\n    private static String getModificationName(ParamGroup paramGroup, String accession)\\n    private static List<CvParam> transformSearchEngineProteinScores(uk.ac.ebi.pride.jmztab.model.Protein rawIdent, Metadata metadata)\\n    public static ExperimentProtocol transformProtocol(SortedMap<Integer, SplitList<Param>> sampleProcession)\\n    public static Map<Comparable, SpectraData> transformMsRunMap(Map<Integer, MsRun> mRunMap)\\n    private static SpectraData transformMsRunToSpectraData(MsRun msRun)\\n    public static Collection<InstrumentConfiguration> transformInstrument(Map<Integer, Instrument> oldInstruments)\\n    private static InstrumentComponent transformAnalyzer(int orderCnt, Param rawAnalyzer)\\n    private static InstrumentComponent transformDetector(int detectorOrder, Param detector)\\n    private static InstrumentComponent transformSource(int sourceOrder, Param source)\\n    public static List<SearchDataBase> transformDatabases(Set<Tuple<String, String>> oldDatabases)\\n    public static Map<Comparable, StudyVariable> transformStudyVariables(Metadata metadata, boolean quantitationExperiment)\\n    private static Assay transformAssay(uk.ac.ebi.pride.jmztab.model.Assay oldAssay)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/impl/Transformer/MzTabTransformer.java', 'methodCount': 35, 'lineCount': 39}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MzTabUnmarshallerAdaptor {\\n    //Todo: mzTab do not have a way to retrieve for each Protein the PSM of peptides\\nMap<Comparable, List<Comparable>> proteinPSMMap;;\\n    private int numIdentifiedPeptides;;\\n    private Map<Integer, Peptide> peptides;;\\n    public MzTabUnmarshallerAdaptor(File tabFile, OutputStream out) throws IOException;\\n    public Map<Integer, MsRun> getSourceFiles()\\n    public Map<Integer, Contact> getContacts()\\n    public Map<Integer, Sample> getSamples()\\n    public Map<Integer, Assay> getAssays()\\n    public Map<Integer, StudyVariable> getStudyVariables()\\n    public Map<Integer, Software> getDataSoftwares()\\n    public Map<Integer, Publication> getReferences()\\n    public List<Param> getAdditionalParams()\\n    public String getExpTitle()\\n    public String getExpAccession()\\n    public String getVersion()\\n    public int getNumberOfPeptides(int rank)\\n    public boolean hasProteinSequence()\\n    public Map<Integer, MsRun> getMRunMap()\\n    public Map<Integer, List<String>> getPSMtoMsRunMap()\\n    public Map<Integer, PSM> getPSMs()\\n    public boolean hasProteinGroup()\\n    public List<String> getProteinGroupIds()\\n    public Map<Integer, Protein> getAllProteins()\\n    public Tuple<Map<String, List<String>>, Map<String, List<String>>> getAllProteinAccessions()\\n    public Map<Integer, Instrument> getInstrument()\\n    public Set<Tuple<String, String>> getDatabases()\\n    public SortedMap<Integer, SplitList<Param>> getProtocol()\\n    public Tuple<Integer, Protein> getProteinById(Comparable proteinId)\\n    public Map<String, PSM> getSpectrumIdentificationsByIds(List<Comparable> spectrumIdentIds)\\n    public Metadata getMetadata()\\n    public int getNumIdentifiedPeptides()\\n    public boolean hasQuantitationData()\\n    public Map<Integer, Peptide> getPeptides()\\n    public Map<String, Peptide> getPeptideByIds(List<Comparable> peptideIds)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/io/file/MzTabUnmarshallerAdaptor.java', 'methodCount': 31, 'lineCount': 36}, {'fieldCount': 24, 'classCount': 1, 'methodsOnly': 'public class Constants {\\n    /**\\n * algebraic sign\\n */\\nprivate static final String SIGN = \"[+-]\";;\\n    public static final String TAB = \"\\\\t\";;\\n    public static final String LINE_SEPARATOR = \"\\\\n\";;\\n    /**\\n * integer expression\\n */\\npublic static final String INTEGER = SIGN + \"?\\\\\\\\d+\";;\\n    public static final String VERSION_NUMBER = \"^(\\\\\\\\d+\\\\\\\\.)(\\\\\\\\d+\\\\\\\\.)(\\\\\\\\d+)$\";;\\n    public static final String MZIDENTML_ID_PATTERN = \"^[a-zA-Z0-9.]*$\";;\\n    private Constants();\\n    public static final String NOT_AVAILABLE = \"N/A\";;\\n    public static final String MGF_EXT = \".mgf\";;\\n    public static final String DTA_EXT = \".dta\";;\\n    public static final String MS2_EXT = \".ms2\";;\\n    public static final String PKL_EXT = \".pkl\";;\\n    public static final String MZXML_EXT = \".mzxml\";;\\n    public static final String APL_EXT = \".apl\";;\\n    public static final String XML_EXT = \".xml\";;\\n    public static final String MZML_EXT = \".mzML\";;\\n    public static final String MZTAB_EXT = \".mztab\";;\\n    public static final String WIFF_EXT = \".wiff\";;\\n    public static final String NETCDF_EXT = \".cdf\";;\\n    public static final String MZIDENTML_NAMESPACE = \"http://psidev.info/psi/pi/mzIdentML/1.1\";;\\n    public static final Double DELTATHESHOLD = 4.0;;\\n    public static final String PSI_MOD = \"MOD\";;\\n    public static final String MS = \"MS\";;\\n    public static final String UNIMOD = \"UNIMOD\";;\\n    public static final String ANCHOR_PROTEIN = \"MS:1001591\";;\\n    public static SpecFileFormat getSpecFileFormat(String fileFormat)\\n    public static SpecFileFormat getSpecFileFormatFromLocation(String path)\\n    public static Constants.SpecFileFormat getSpectraDataFormat(SpectraData spectraData)\\n    public static Constants.SpecIdFormat getSpectraDataIdFormat(String accession)\\n    public static List<SpecFileFormat> getFileTypeSupported(SpectraData spectraData)\\n    public static Constants.SpecFileFormat getDataFormatFromFileExtension(SpectraData spectradata)\\n    public static Class getFileType(File file)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/Constants.java', 'methodCount': 8, 'lineCount': 34}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class CvUtilities {\\n    public static CvParam getCVTermFromCvReference(CvTermReference cvTerm, String value)\\n    public static CvParam getQuantTermFromQuantReference(QuantCvTermReference quantTerm, String value)\\n    public static String getMailFromCvParam(ParamGroup person)\\n    public static boolean isDecoyHit(ParamGroup paramGroup)\\n    public static String getValueFromParmGroup(List<CvParam> cvParams, String accession)\\n    public static boolean isAccessionDecoy(Protein identification)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/CvUtilities.java', 'methodCount': 6, 'lineCount': 8}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MD5Utils {\\n    private MD5Utils();\\n    public static String generateHash(String msg) throws NoSuchAlgorithmException\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/MD5Utils.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 8, 'javaImportCount': 2}\n",
      "{'DataAccessException': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n'}\n",
      "/**\n",
      "     * Return a List of Persons\n",
      "     * @return List of Persons\n",
      "     */\n",
      "public List<Person> getPersonContacts()\n",
      "**************************************\n",
      "public List<Person> getPersonContacts() {\n",
      "    logger.debug(\"Get person contacts\");\n",
      "    List<Person> personList = new ArrayList<Person>();\n",
      "    try {\n",
      "        personList.addAll(MzTabTransformer.transformContactToPersons(reader.getContacts()));\n",
      "    } catch (Exception ex) {\n",
      "        String msg = \"Error while getting person contacts\";\n",
      "        logger.error(msg, ex);\n",
      "        throw new DataAccessException(msg, ex);\n",
      "    }\n",
      "    return personList;\n",
      "}\n",
      "3379\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 394/1243 [13:30<29:27,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19896\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 395/1243 [13:33<29:47,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19905\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 396/1243 [13:35<29:28,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19912\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 397/1243 [13:37<30:40,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19917\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 398/1243 [13:40<32:03,  2.28s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19921\n",
      "{'classCount': 3, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface DataAccessController {\\n    public String getUid()\\n    public String getName()\\n    public Type getType()\\n    public Collection<ContentCategory> getContentCategories()\\n    public Object getSource()\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessController.java', 'methodCount': 6, 'lineCount': 8}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessMode.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MzXmlCachingStrategy {\\n    public void cache()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/strategy/MzXmlCachingStrategy.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MzXmlTransformer {\\n    private MzXmlTransformer();\\n    public static Spectrum transformSpectrum(uk.ac.ebi.pride.tools.jmzreader.model.Spectrum spectrum)\\n    private static List<BinaryDataArray> transformBinaryDataArrayList(Map<Double, Double> peakList)\\n    public static List<Person> transformPersonContacts(List<uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.Operator> operators)\\n    public static List<SourceFile> transformFileSources(List<uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.ParentFile> rawParentFiles)\\n    public static List<Software> transformSoftwares(List<uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.Software> rawSoftwares)\\n    public static Software transformSoftware(uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.Software rawSoftware)\\n    public static List<InstrumentConfiguration> transformMsInstrument(List<uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.MsInstrument> rawInstrumentList)\\n    public static List<DataProcessing> transformDataProcessingList(List<uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.DataProcessing> rawDataProcList)\\n    public static DataProcessing transformDataProcessing(int order, uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.DataProcessing oldDataProcessing)\\n    public static CvParam transformDurationToCvParam(Duration startDate, CvTermReference msScanDate)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/impl/Transformer/MzXmlTransformer.java', 'methodCount': 11, 'lineCount': 13}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MzXmlUnmarshallerAdaptor {\\n    private MzXMLFile unmarshaller = null;;\\n    public MzXmlUnmarshallerAdaptor(MzXMLFile um);\\n    public List<String> getSpectrumIds()\\n    public Spectrum getSpectrumById(String id) throws JMzReaderException\\n    public List<Operator> getPersonContacts() throws MzXMLParsingException\\n    public List<Software> getSoftwares() throws MzXMLParsingException\\n    public List<ParentFile> getParentFiles() throws MzXMLParsingException\\n    public List<MsInstrument> getMsInstruments() throws MzXMLParsingException\\n    public List<DataProcessing> getDataProcessing() throws MzXMLParsingException\\n    public Duration getStartDate() throws DatatypeConfigurationException\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/io/file/MzXmlUnmarshallerAdaptor.java', 'methodCount': 9, 'lineCount': 12}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MD5Utils {\\n    private MD5Utils();\\n    public static String generateHash(String msg) throws NoSuchAlgorithmException\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/MD5Utils.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 13}\n",
      "{}\n",
      "/**\n",
      "     * The softwares on mzXml are annotated with their names, the type, and the execution\n",
      "     * time processing the data. The function annotated also a CvParam related with the Type\n",
      "     * of the Software (Data Processing or Data Acquisition).\n",
      "     *\n",
      "     * @return\n",
      "     * @\n",
      "     */\n",
      "public List<Software> getSoftwares()\n",
      "**************************************\n",
      "public List<Software> getSoftwares() {\n",
      "    ExperimentMetaData metaData = super.getExperimentMetaData();\n",
      "    if (metaData == null) {\n",
      "        try {\n",
      "            List<uk.ac.ebi.pride.tools.mzxml_parser.mzxml.model.Software> rawSoftware = unmarshaller.getSoftwares();\n",
      "            return MzXmlTransformer.transformSoftwares(rawSoftware);\n",
      "        } catch (MzXMLParsingException e) {\n",
      "            String msg = \"Error while getting a list of software\";\n",
      "            logger.error(msg, e);\n",
      "            throw new DataAccessException(msg, e);\n",
      "        }\n",
      "    } else {\n",
      "        return metaData.getSoftwares();\n",
      "    }\n",
      "}\n",
      "1435\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 399/1243 [13:42<31:02,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19922\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 400/1243 [13:44<31:06,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19924\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 401/1243 [13:46<30:45,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19929\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 402/1243 [13:48<28:12,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20004\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  32%|      | 403/1243 [13:50<28:43,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20013\n",
      "{'classCount': 3, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface DataAccessController {\\n    public String getUid()\\n    public String getName()\\n    public Type getType()\\n    public Collection<ContentCategory> getContentCategories()\\n    public Object getSource()\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessController.java', 'methodCount': 6, 'lineCount': 8}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessMode.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MzDataCachingStrategy {\\n    public void cache()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/strategy/MzDataCachingStrategy.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MzDataTransformer {\\n    private MzDataTransformer();\\n    public static Spectrum transformSpectrum(uk.ac.ebi.pride.tools.jmzreader.model.Spectrum spectrum)\\n    public static ParamGroup transformParamGroup(T oldParam)\\n    public static List<ParamGroup> transformParamGroupList(List<T> oldParamGroupList)\\n    private static List<CvParam> transformCvParam(List<CvParam> newCvParams, List<uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.CvParam> oldCvParams)\\n    private static List<UserParam> transformUserParam(List<UserParam> newUserParams, List<uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.UserParam> oldUserParams)\\n    private static List<BinaryDataArray> transformBinaryDataArrayList(Map<Double, Double> peakList)\\n    public static List<CVLookup> transformCVList(List<uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.CvLookup> oldCvList)\\n    public static CVLookup transformCVLookup(uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.CvLookup oldCvLookup)\\n    public static List<Sample> transformSampleList(uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.Admin oldSampleDescription)\\n    public static List<Software> transformSoftware(uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.Software oldSoftware)\\n    public static List<InstrumentConfiguration> transformInstrumentConfiguration(uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.InstrumentDescription oldInstrument)\\n    private static InstrumentComponent transformInstrumentComponent(int order, uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.Param rawParamGroup)\\n    public static List<DataProcessing> transformDataProcessing(uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.DataProcessing oldDataProcessing)\\n    public static List<SourceFile> transformToFileSource(uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.SourceFile oldSourceFile)\\n    public static List<Person> transformToPerson(List<uk.ac.ebi.pride.tools.mzdata_parser.mzdata.model.Person> oldPersons)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/impl/Transformer/MzDataTransformer.java', 'methodCount': 16, 'lineCount': 18}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MzDataUnmarshallerAdaptor {\\n    private MzDataFile unmarshaller = null;;\\n    public MzDataUnmarshallerAdaptor(MzDataFile um);\\n    public List<String> getSpectrumIds()\\n    public Spectrum getSpectrumById(String id) throws JMzReaderException\\n    public List<Person> getPersonContacts() throws JMzReaderException\\n    public Software getSoftware() throws JMzReaderException\\n    public SourceFile getSourceFiles() throws JMzReaderException\\n    public InstrumentDescription getInstrument() throws JMzReaderException\\n    public List<CvLookup> getCvLookups() throws JMzReaderException\\n    public Admin getAdmin() throws JMzReaderException\\n    public DataProcessing getDataProcessing() throws JMzReaderException\\n    public String getIdMzData()\\n    public String getVersionMzData()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/io/file/MzDataUnmarshallerAdaptor.java', 'methodCount': 12, 'lineCount': 15}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MD5Utils {\\n    private MD5Utils();\\n    public static String generateHash(String msg) throws NoSuchAlgorithmException\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/MD5Utils.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 11}\n",
      "{}\n",
      "/**\n",
      "     * Get spectrum using a spectrum id, gives the option to choose whether to use cache.\n",
      "     * This implementation provides a way of by passing the cache.\n",
      "     *\n",
      "     * @param id       spectrum id\n",
      "     * @param useCache true means to use cache\n",
      "     * @return Spectrum spectrum object\n",
      "     * @ data access exception\n",
      "     */\n",
      " Spectrum getSpectrumById(Comparable id, boolean useCache)\n",
      "**************************************\n",
      "@Override\n",
      "Spectrum getSpectrumById(Comparable id, boolean useCache) {\n",
      "    Spectrum spectrum = super.getSpectrumById(id, useCache);\n",
      "    if (spectrum == null) {\n",
      "        try {\n",
      "            uk.ac.ebi.pride.tools.jmzreader.model.Spectrum rawSpec = unmarshaller.getSpectrumById(id.toString());\n",
      "            spectrum = MzDataTransformer.transformSpectrum(rawSpec);\n",
      "            if (useCache) {\n",
      "                getCache().store(CacheEntry.SPECTRUM, id, spectrum);\n",
      "            }\n",
      "        } catch (JMzReaderException ex) {\n",
      "            logger.error(\"Get spectrum by id\", ex);\n",
      "            throw new DataAccessException(\"Exception while trying to read Spectrum using Spectrum ID\", ex);\n",
      "        }\n",
      "    }\n",
      "    return spectrum;\n",
      "}\n",
      "1652\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 404/1243 [13:52<28:48,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20089\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 405/1243 [13:54<27:11,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20212\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 406/1243 [13:55<26:04,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20214\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 407/1243 [13:57<27:11,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20221\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 408/1243 [14:00<29:17,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20222\n",
      "{'classCount': 3, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessMode.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class DataAccessUtilities {\\n    public static final Logger logger = LoggerFactory.getLogger(DataAccessUtilities.class);;\\n    private DataAccessUtilities();\\n    public static List<String> getTaxonomy(ExperimentMetaData metaData)\\n    public static String getProjectName(ExperimentMetaData metaData)\\n    public static int getNumberOfPeaks(Spectrum spectrum)\\n    public static int getMsLevel(Spectrum spectrum)\\n    public static Integer getPrecursorChargeParamGroup(ParamGroup paramGroup)\\n    private static Double getSelectedIonCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static double getPrecursorMz(Spectrum spectrum)\\n    public static double getPrecursorMz(ParamGroup paramGroup)\\n    public static double getPrecursorIntensity(Spectrum spectrum)\\n    public static double getSumOfIntensity(Spectrum spectrum)\\n    public static Double getSelectedIonCharge(Precursor precursor, int index)\\n    public static Double getSelectedIonMz(Precursor precursor, int index)\\n    public static Double getSelectedIonIntensity(Precursor precursor, int index)\\n    private static Double getSelectedIonCvParamValue(Precursor precursor, int index, CvTermReference... refs)\\n    private static Double getSelectedCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static Peptide getPeptide(Protein ident, int index)\\n    public static QuantPeptide getQuantPeptide(Protein ident, int index)\\n    public static int getNumberOfPeptides(Protein ident)\\n    public static int getNumberOfUniquePeptides(Protein ident)\\n    public static boolean hasFragmentIon(Spectrum spectrum)\\n    public static boolean hasFragmentIon(Peptide peptide)\\n    public static int getNumberOfPTMs(Protein ident)\\n    public static int getNumberOfSubstitutionPTMs(Protein ident)\\n    public static int getNumberOfPTMs(Peptide peptide)\\n    public static int getNumberOfSubstitutionPTMs(Peptide peptide)\\n    public static List<SearchEngineCvTermReference> getSearchEngineTypes(ParamGroup paramGroup)\\n    public static Score getScore(ParamGroup params)\\n    public static List<SearchEngineCvTermReference> getSearchEngineCvTermReference(ParamGroup paramGroup)\\n    public static List<SearchEngineScoreCvTermReference> getSearchEngineScoresCvTermReference(ParamGroup paramGroup)\\n    public static List<CvParam> getCvParam(ParamGroup paramGroup, String cvLabel, String accession)\\n    public static List<CvParam> getCvParam(String name, String cvLabel, String accession, String value)\\n    public static List<Parameter> getParamByName(ParamGroup paramGroup, String name)\\n    public static Map<PeptideEvidence, List<Peptide>> getPeptideEvidence(List<Peptide> peptides)\\n    public static Integer getPrecursorCharge(List<Precursor> precursors)\\n    public static Integer getProteinCoverage(Protein protein)\\n    private static Set<Integer> searchStartingPosition(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isStrictValidPeptideAnnotation(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isValidPeptideAnnotation(Protein protein, Peptide peptide)\\n    public static String getRetentionTime(SpectrumIdentificationItem oldSpectrumIdentification)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessUtilities.java', 'methodCount': 40, 'lineCount': 43}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Cache {\\n     void store(CacheEntry type, Object key, Object value)\\n     void storeInBatch(CacheEntry type, Map values)\\n     void store(CacheEntry type, Object key)\\n     void storeInBatch(CacheEntry type, Collection values)\\n     Object get(CacheEntry type, Object key)\\n     Collection getInBatch(CacheEntry type, Collection keys)\\n     boolean hasCacheEntry(CacheEntry type)\\n     Object get(CacheEntry type)\\n     void clear(CacheEntry type)\\n     void clear()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/Cache.java', 'methodCount': 10, 'lineCount': 12}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class CacheAccessor {\\n    private static final Logger logger = LoggerFactory.getLogger(CacheAccessor.class);;\\n    /**\\n * All data are stored in here.\\n */\\nprivate final Map<CacheEntry, Object> contents;;\\n    public CacheAccessor();\\n    public void store(CacheEntry type, Object key)\\n    public void store(CacheEntry type, Object key, Object value)\\n    public void storeInBatch(CacheEntry type, Map values)\\n    public void storeInBatch(CacheEntry type, Collection values)\\n    public Object get(CacheEntry type, Object key)\\n    public Collection getInBatch(CacheEntry type, Collection keys)\\n    public Object get(CacheEntry type)\\n    public boolean hasCacheEntry(CacheEntry type)\\n    public void clear(CacheEntry type)\\n    public void clear()\\n    private Object createIfNotExist(CacheEntry type)\\n    private Object retrieveContent(CacheEntry type, Object key)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheAccessor.java', 'methodCount': 13, 'lineCount': 17}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface CachingStrategy {\\n     void cache()\\n     void setDataAccessController(DataAccessController dataAccessController)\\n     void setCache(Cache cache)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CachingStrategy.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class CollectionUtils {\\n    private CollectionUtils();\\n    public static int getIndex(Collection<T> collection, T element)\\n    public static T getElement(Collection<T> collection, int index)\\n    public static T getLastElement(Collection<T> collection)\\n    public static List<T> createListFromList(Collection<T> collection)\\n    public static void replaceValuesInCollection(Collection<T> from, Collection<T> to)\\n    public static List<T> createEmptyList()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/CollectionUtils.java', 'methodCount': 7, 'lineCount': 9}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Get search engine type\n",
      "     *\n",
      "     * @return SearchEngine    search engine\n",
      "     */\n",
      "public List<SearchEngineCvTermReference> getSearchEngineCvTermReferences()\n",
      "**************************************\n",
      "@Override\n",
      "public List<SearchEngineCvTermReference> getSearchEngineCvTermReferences() {\n",
      "    Collection<SearchEngineCvTermReference> searchEngineCvTermReferences = (Collection<SearchEngineCvTermReference>) cache.get(CacheEntry.SEARCH_ENGINE_TYPE);\n",
      "    if (searchEngineCvTermReferences != null && !searchEngineCvTermReferences.isEmpty()) {\n",
      "        return new ArrayList<SearchEngineCvTermReference>(searchEngineCvTermReferences);\n",
      "    } else if (!DataAccessMode.CACHE_ONLY.equals(mode)) {\n",
      "        return super.getSearchEngineCvTermReferences();\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "2001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 409/1243 [14:02<29:33,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20223\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 410/1243 [14:04<29:37,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20224\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 411/1243 [14:06<27:55,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20225\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 412/1243 [14:08<28:38,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20226\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 413/1243 [14:10<28:06,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20230\n",
      "{'classCount': 3, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessMode.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class DataAccessUtilities {\\n    public static final Logger logger = LoggerFactory.getLogger(DataAccessUtilities.class);;\\n    private DataAccessUtilities();\\n    public static List<String> getTaxonomy(ExperimentMetaData metaData)\\n    public static String getProjectName(ExperimentMetaData metaData)\\n    public static int getNumberOfPeaks(Spectrum spectrum)\\n    public static int getMsLevel(Spectrum spectrum)\\n    public static Integer getPrecursorChargeParamGroup(ParamGroup paramGroup)\\n    private static Double getSelectedIonCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static double getPrecursorMz(Spectrum spectrum)\\n    public static double getPrecursorMz(ParamGroup paramGroup)\\n    public static double getPrecursorIntensity(Spectrum spectrum)\\n    public static double getSumOfIntensity(Spectrum spectrum)\\n    public static Double getSelectedIonCharge(Precursor precursor, int index)\\n    public static Double getSelectedIonMz(Precursor precursor, int index)\\n    public static Double getSelectedIonIntensity(Precursor precursor, int index)\\n    private static Double getSelectedIonCvParamValue(Precursor precursor, int index, CvTermReference... refs)\\n    private static Double getSelectedCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static Peptide getPeptide(Protein ident, int index)\\n    public static QuantPeptide getQuantPeptide(Protein ident, int index)\\n    public static int getNumberOfPeptides(Protein ident)\\n    public static int getNumberOfUniquePeptides(Protein ident)\\n    public static boolean hasFragmentIon(Spectrum spectrum)\\n    public static boolean hasFragmentIon(Peptide peptide)\\n    public static int getNumberOfPTMs(Protein ident)\\n    public static int getNumberOfSubstitutionPTMs(Protein ident)\\n    public static int getNumberOfPTMs(Peptide peptide)\\n    public static int getNumberOfSubstitutionPTMs(Peptide peptide)\\n    public static List<SearchEngineCvTermReference> getSearchEngineTypes(ParamGroup paramGroup)\\n    public static Score getScore(ParamGroup params)\\n    public static List<SearchEngineCvTermReference> getSearchEngineCvTermReference(ParamGroup paramGroup)\\n    public static List<SearchEngineScoreCvTermReference> getSearchEngineScoresCvTermReference(ParamGroup paramGroup)\\n    public static List<CvParam> getCvParam(ParamGroup paramGroup, String cvLabel, String accession)\\n    public static List<CvParam> getCvParam(String name, String cvLabel, String accession, String value)\\n    public static List<Parameter> getParamByName(ParamGroup paramGroup, String name)\\n    public static Map<PeptideEvidence, List<Peptide>> getPeptideEvidence(List<Peptide> peptides)\\n    public static Integer getPrecursorCharge(List<Precursor> precursors)\\n    public static Integer getProteinCoverage(Protein protein)\\n    private static Set<Integer> searchStartingPosition(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isStrictValidPeptideAnnotation(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isValidPeptideAnnotation(Protein protein, Peptide peptide)\\n    public static String getRetentionTime(SpectrumIdentificationItem oldSpectrumIdentification)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessUtilities.java', 'methodCount': 40, 'lineCount': 43}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Cache {\\n     void store(CacheEntry type, Object key, Object value)\\n     void storeInBatch(CacheEntry type, Map values)\\n     void store(CacheEntry type, Object key)\\n     void storeInBatch(CacheEntry type, Collection values)\\n     Object get(CacheEntry type, Object key)\\n     Collection getInBatch(CacheEntry type, Collection keys)\\n     boolean hasCacheEntry(CacheEntry type)\\n     Object get(CacheEntry type)\\n     void clear(CacheEntry type)\\n     void clear()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/Cache.java', 'methodCount': 10, 'lineCount': 12}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class CacheAccessor {\\n    private static final Logger logger = LoggerFactory.getLogger(CacheAccessor.class);;\\n    /**\\n * All data are stored in here.\\n */\\nprivate final Map<CacheEntry, Object> contents;;\\n    public CacheAccessor();\\n    public void store(CacheEntry type, Object key)\\n    public void store(CacheEntry type, Object key, Object value)\\n    public void storeInBatch(CacheEntry type, Map values)\\n    public void storeInBatch(CacheEntry type, Collection values)\\n    public Object get(CacheEntry type, Object key)\\n    public Collection getInBatch(CacheEntry type, Collection keys)\\n    public Object get(CacheEntry type)\\n    public boolean hasCacheEntry(CacheEntry type)\\n    public void clear(CacheEntry type)\\n    public void clear()\\n    private Object createIfNotExist(CacheEntry type)\\n    private Object retrieveContent(CacheEntry type, Object key)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheAccessor.java', 'methodCount': 13, 'lineCount': 17}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface CachingStrategy {\\n     void cache()\\n     void setDataAccessController(DataAccessController dataAccessController)\\n     void setCache(Cache cache)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CachingStrategy.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class CollectionUtils {\\n    private CollectionUtils();\\n    public static int getIndex(Collection<T> collection, T element)\\n    public static T getElement(Collection<T> collection, int index)\\n    public static T getLastElement(Collection<T> collection)\\n    public static List<T> createListFromList(Collection<T> collection)\\n    public static void replaceValuesInCollection(Collection<T> from, Collection<T> to)\\n    public static List<T> createEmptyList()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/CollectionUtils.java', 'methodCount': 7, 'lineCount': 9}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Return a collection of the protein group identifiers\n",
      "     * @return java.lang.Collection<java.lang.Comparable> of Protein group identifiers\n",
      "     */\n",
      "public Collection<Comparable> getProteinAmbiguityGroupIds()\n",
      "**************************************\n",
      "@Override\n",
      "public Collection<Comparable> getProteinAmbiguityGroupIds() {\n",
      "    if (proteinsAreInferred()) {\n",
      "        return super.getProteinAmbiguityGroupIds();\n",
      "    }\n",
      "    Collection<Comparable> groupIds = (Collection<Comparable>) cache.get(CacheEntry.PROTEIN_GROUP_ID);\n",
      "    if (groupIds == null || groupIds.isEmpty()) {\n",
      "        groupIds = Collections.emptyList();\n",
      "    }\n",
      "    return groupIds;\n",
      "}\n",
      "1969\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 414/1243 [14:13<30:26,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20239\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 415/1243 [14:15<30:17,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20240\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  33%|      | 416/1243 [14:17<30:33,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20247\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 417/1243 [14:19<28:23,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20257\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 418/1243 [14:21<29:17,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20258\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DataAccessException {\\n    public DataAccessException(String message);\\n    public DataAccessException(String message, Throwable cause);\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessMode.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class DataAccessUtilities {\\n    public static final Logger logger = LoggerFactory.getLogger(DataAccessUtilities.class);;\\n    private DataAccessUtilities();\\n    public static List<String> getTaxonomy(ExperimentMetaData metaData)\\n    public static String getProjectName(ExperimentMetaData metaData)\\n    public static int getNumberOfPeaks(Spectrum spectrum)\\n    public static int getMsLevel(Spectrum spectrum)\\n    public static Integer getPrecursorChargeParamGroup(ParamGroup paramGroup)\\n    private static Double getSelectedIonCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static double getPrecursorMz(Spectrum spectrum)\\n    public static double getPrecursorMz(ParamGroup paramGroup)\\n    public static double getPrecursorIntensity(Spectrum spectrum)\\n    public static double getSumOfIntensity(Spectrum spectrum)\\n    public static Double getSelectedIonCharge(Precursor precursor, int index)\\n    public static Double getSelectedIonMz(Precursor precursor, int index)\\n    public static Double getSelectedIonIntensity(Precursor precursor, int index)\\n    private static Double getSelectedIonCvParamValue(Precursor precursor, int index, CvTermReference... refs)\\n    private static Double getSelectedCvParamValue(ParamGroup paramGroup, CvTermReference... refs)\\n    public static Peptide getPeptide(Protein ident, int index)\\n    public static QuantPeptide getQuantPeptide(Protein ident, int index)\\n    public static int getNumberOfPeptides(Protein ident)\\n    public static int getNumberOfUniquePeptides(Protein ident)\\n    public static boolean hasFragmentIon(Spectrum spectrum)\\n    public static boolean hasFragmentIon(Peptide peptide)\\n    public static int getNumberOfPTMs(Protein ident)\\n    public static int getNumberOfSubstitutionPTMs(Protein ident)\\n    public static int getNumberOfPTMs(Peptide peptide)\\n    public static int getNumberOfSubstitutionPTMs(Peptide peptide)\\n    public static List<SearchEngineCvTermReference> getSearchEngineTypes(ParamGroup paramGroup)\\n    public static Score getScore(ParamGroup params)\\n    public static List<SearchEngineCvTermReference> getSearchEngineCvTermReference(ParamGroup paramGroup)\\n    public static List<SearchEngineScoreCvTermReference> getSearchEngineScoresCvTermReference(ParamGroup paramGroup)\\n    public static List<CvParam> getCvParam(ParamGroup paramGroup, String cvLabel, String accession)\\n    public static List<CvParam> getCvParam(String name, String cvLabel, String accession, String value)\\n    public static List<Parameter> getParamByName(ParamGroup paramGroup, String name)\\n    public static Map<PeptideEvidence, List<Peptide>> getPeptideEvidence(List<Peptide> peptides)\\n    public static Integer getPrecursorCharge(List<Precursor> precursors)\\n    public static Integer getProteinCoverage(Protein protein)\\n    private static Set<Integer> searchStartingPosition(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isStrictValidPeptideAnnotation(Protein protein, PeptideEvidence uniquePeptide)\\n    private static boolean isValidPeptideAnnotation(Protein protein, Peptide peptide)\\n    public static String getRetentionTime(SpectrumIdentificationItem oldSpectrumIdentification)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/DataAccessUtilities.java', 'methodCount': 40, 'lineCount': 43}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PrideXmlCachingStrategy {\\n    public void cache()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/strategy/PrideXmlCachingStrategy.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 9, 'classCount': 1, 'methodsOnly': 'public class PrideXmlTransformer {\\n    /**\\n * These are a list of default names to be used for the conversion\\n */\\nprivate final static String SAMPLE_ID = \"sample1\";;\\n    private final static String COMMENTS = \"comments\";;\\n    private final static String COMPLETION_TIME = \"completion time\";;\\n    private final static String CONTACT_INFO = \"contact information\";;\\n    private final static String DATA_PROCESSING_ID = \"dataprocessing1\";;\\n    private final static String SOURCE_FILE_ID = \"sourcefile1\";;\\n    private final static String SOURCE_FILE_TYPE = \"source file type\";;\\n    private final static String PROTOCOL_ID = \"protocol1\";;\\n    private final static int PROCESSING_METHOD_ORDER = 1;;\\n    private PrideXmlTransformer();\\n    public static Spectrum transformSpectrum(uk.ac.ebi.pride.jaxb.model.Spectrum rawSpec)\\n    private static ParamGroup transformSpectrumParamGroup(uk.ac.ebi.pride.jaxb.model.SpectrumDesc rawSpecDesc)\\n    private static CvParam getSpectrumType(String value)\\n    private static ScanList transformScanList(uk.ac.ebi.pride.jaxb.model.SpectrumSettings rawSpecSettings)\\n    private static List<ParamGroup> getScanWindows(uk.ac.ebi.pride.jaxb.model.SpectrumInstrument rawSpecInstrument)\\n    private static CvParam getMethodOfCombination(String value)\\n    public static List<Precursor> transformPrecursorList(uk.ac.ebi.pride.jaxb.model.PrecursorList rawPrecursors)\\n    public static Precursor transformPrecursor(uk.ac.ebi.pride.jaxb.model.Precursor rawPrecursor)\\n    public static BinaryDataArray transformBinaryDataArray(uk.ac.ebi.pride.jaxb.model.PeakListBinary rawArr, CvTermReference binaryType)\\n    public static Protein transformIdentification(uk.ac.ebi.pride.jaxb.model.Identification identification)\\n    public static Protein transformTwoDimIdent(uk.ac.ebi.pride.jaxb.model.TwoDimensionalIdentification rawIdent)\\n    public static Protein transformGelFreeIdent(uk.ac.ebi.pride.jaxb.model.GelFreeIdentification rawIdent)\\n    private static Gel transformGel(uk.ac.ebi.pride.jaxb.model.SimpleGel rawGel, uk.ac.ebi.pride.jaxb.model.GelLocation gelLocation, Double mw, Double pI)\\n    public static Peptide transformPeptide(uk.ac.ebi.pride.jaxb.model.PeptideItem rawPeptide, DBSequence dbSequence, Comparable index)\\n    private static Modification transformModification(uk.ac.ebi.pride.jaxb.model.ModificationItem rawMod)\\n    private static String getModificationName(ParamGroup paramGroup, String accession)\\n    public static ExperimentProtocol transformProtocol(uk.ac.ebi.pride.jaxb.model.Protocol rawProt)\\n    public static ParamGroup transformParamGroup(uk.ac.ebi.pride.jaxb.model.Param rawParams)\\n    public static List<UserParam> transformUserParams(List<uk.ac.ebi.pride.jaxb.model.UserParam> rawUserParams)\\n    public static UserParam transformUserParam(uk.ac.ebi.pride.jaxb.model.UserParam rawUserParam)\\n    public static List<CvParam> transformCvParams(List<uk.ac.ebi.pride.jaxb.model.CvParam> rawCvParams)\\n    public static CvParam transformCvParam(uk.ac.ebi.pride.jaxb.model.CvParam rawCvParam)\\n    public static List<CVLookup> transformCvLookups(List<uk.ac.ebi.pride.jaxb.model.CvLookup> rawCvLookups)\\n    public static Sample transformSample(uk.ac.ebi.pride.jaxb.model.Admin rawAdmin)\\n    public static Software transformSoftware(uk.ac.ebi.pride.jaxb.model.DataProcessing rawDataProcessing)\\n    public static List<InstrumentConfiguration> transformInstrument(uk.ac.ebi.pride.jaxb.model.Instrument rawInstrument, uk.ac.ebi.pride.jaxb.model.DataProcessing rawDataProcessing)\\n    private static InstrumentComponent transformSource(int order, uk.ac.ebi.pride.jaxb.model.Param rawSource)\\n    private static InstrumentComponent transformAnalyzer(int order, uk.ac.ebi.pride.jaxb.model.Param rawAnalyzer)\\n    private static InstrumentComponent transformDetector(int order, uk.ac.ebi.pride.jaxb.model.Param rawDetector)\\n    public static DataProcessing transformDataProcessing(uk.ac.ebi.pride.jaxb.model.DataProcessing rawDataProcessing)\\n    public static List<ParamGroup> transformContacts(uk.ac.ebi.pride.jaxb.model.Admin rawAdmin)\\n    public static List<Person> transformContactToPerson(uk.ac.ebi.pride.jaxb.model.Admin rawAdmin)\\n    public static List<Organization> transformContactToOrganization(uk.ac.ebi.pride.jaxb.model.Admin rawAdmin)\\n    public static SourceFile transformSourceFile(uk.ac.ebi.pride.jaxb.model.Admin rawAdmin)\\n    public static List<Reference> transformReferences(List<uk.ac.ebi.pride.jaxb.model.Reference> rawReferences)\\n    public static ParamGroup transformAdditional(uk.ac.ebi.pride.jaxb.model.Param rawAdditionalParams)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/impl/Transformer/PrideXmlTransformer.java', 'methodCount': 37, 'lineCount': 48}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MD5Utils {\\n    private MD5Utils();\\n    public static String generateHash(String msg) throws NoSuchAlgorithmException\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/MD5Utils.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Get the number of peptides by Rank, in PRIDE XML all peptides are rank 1.\n",
      "     *\n",
      "     * @return int  the number of peptides.\n",
      "     */\n",
      "public int getNumberOfPeptidesByRank(int rank)\n",
      "**************************************\n",
      "@Override\n",
      "public int getNumberOfPeptidesByRank(int rank) {\n",
      "    int num;\n",
      "    try {\n",
      "        num = reader.getNumberOfPeptides();\n",
      "    } catch (Exception ex) {\n",
      "        throw new DataAccessException(\"Failed to retrieve number of peptides\", ex);\n",
      "    }\n",
      "    return num;\n",
      "}\n",
      "2603\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 419/1243 [14:23<27:44,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20263\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 420/1243 [14:25<29:58,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20280\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 421/1243 [14:27<27:51,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20289\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 422/1243 [14:30<29:35,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20294\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 423/1243 [14:32<29:18,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20437\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/cache/CacheEntry.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class MzXmlControllerImpl {\\n    private static final Logger logger = LoggerFactory.getLogger(MzXmlControllerImpl.class);;\\n    /**\\n * Pattern for validating mzXML format\\n */\\nprivate static final Pattern mzXmlHeaderPattern = Pattern.compile(\"^[^<]*(<\\\\\\\\?xml [^>]*>\\\\\\\\s*(<!--[^>]*-->\\\\\\\\s*)*)?<(mzXML) xmlns=.*\", Pattern.MULTILINE);;\\n    /**\\n * Reader for getting information from mzXML file\\n */\\nprivate MzXmlUnmarshallerAdaptor unmarshaller;;\\n    public MzXmlControllerImpl(File file);\\n    private void initialize()\\n    public MzXmlUnmarshallerAdaptor getUnmarshaller()\\n    public String getUid()\\n    public List<Person> getPersonContacts()\\n    public List<SourceFile> getSourceFiles()\\n    public ParamGroup getFileContent()\\n    public List<Software> getSoftwares()\\n    public List<InstrumentConfiguration> getInstrumentConfigurations()\\n    public List<DataProcessing> getDataProcessings()\\n    public ParamGroup getAdditional()\\n     Spectrum getSpectrumById(Comparable id, boolean useCache)\\n    public Chromatogram getChromatogramById(Comparable id, boolean useCache)\\n    public void close()\\n    public ExperimentMetaData getExperimentMetaData()\\n    public MzGraphMetaData getMzGraphMetaData()\\n    public IdentificationMetaData getIdentificationMetaData()\\n    public static boolean isValidFormat(File file)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/impl/ControllerImpl/MzXmlControllerImpl.java', 'methodCount': 18, 'lineCount': 23}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MzXmlUnmarshallerAdaptor {\\n    private MzXMLFile unmarshaller = null;;\\n    public MzXmlUnmarshallerAdaptor(MzXMLFile um);\\n    public List<String> getSpectrumIds()\\n    public Spectrum getSpectrumById(String id) throws JMzReaderException\\n    public List<Operator> getPersonContacts() throws MzXMLParsingException\\n    public List<Software> getSoftwares() throws MzXMLParsingException\\n    public List<ParentFile> getParentFiles() throws MzXMLParsingException\\n    public List<MsInstrument> getMsInstruments() throws MzXMLParsingException\\n    public List<DataProcessing> getDataProcessing() throws MzXMLParsingException\\n    public Duration getStartDate() throws DatatypeConfigurationException\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/io/file/MzXmlUnmarshallerAdaptor.java', 'methodCount': 9, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{'MzXmlControllerImpl': 'public class MzXmlControllerImpl {\\n    private static final Logger logger = LoggerFactory.getLogger(MzXmlControllerImpl.class);;\\n    /**\\n * Pattern for validating mzXML format\\n */\\nprivate static final Pattern mzXmlHeaderPattern = Pattern.compile(\"^[^<]*(<\\\\\\\\?xml [^>]*>\\\\\\\\s*(<!--[^>]*-->\\\\\\\\s*)*)?<(mzXML) xmlns=.*\", Pattern.MULTILINE);;\\n    /**\\n * Reader for getting information from mzXML file\\n */\\nprivate MzXmlUnmarshallerAdaptor unmarshaller;;\\n    public MzXmlControllerImpl(File file);\\n    private void initialize()\\n    public MzXmlUnmarshallerAdaptor getUnmarshaller()\\n    public String getUid()\\n    public List<Person> getPersonContacts()\\n    public List<SourceFile> getSourceFiles()\\n    public ParamGroup getFileContent()\\n    public List<Software> getSoftwares()\\n    public List<InstrumentConfiguration> getInstrumentConfigurations()\\n    public List<DataProcessing> getDataProcessings()\\n    public ParamGroup getAdditional()\\n     Spectrum getSpectrumById(Comparable id, boolean useCache)\\n    public Chromatogram getChromatogramById(Comparable id, boolean useCache)\\n    public void close()\\n    public ExperimentMetaData getExperimentMetaData()\\n    public MzGraphMetaData getMzGraphMetaData()\\n    public IdentificationMetaData getIdentificationMetaData()\\n    public static boolean isValidFormat(File file)\\n}\\n'}\n",
      "/**\n",
      "     * For the moment, MzXmlCacheBuilder only caches spectrum ids and chromatogram ids.\n",
      "     */\n",
      "public void cache()\n",
      "**************************************\n",
      "/**\n",
      " * For the moment, MzXmlCacheBuilder only caches spectrum ids and chromatogram ids.\n",
      " */\n",
      "@Override\n",
      "public void cache() {\n",
      "    // get a direct reference to unmarshaller\n",
      "    MzXmlUnmarshallerAdaptor unmarshaller = ((MzXmlControllerImpl) controller).getUnmarshaller();\n",
      "    // clear and add spectrum ids\n",
      "    cache.clear(CacheEntry.SPECTRUM_ID);\n",
      "    cache.storeInBatch(CacheEntry.SPECTRUM_ID, new ArrayList<Comparable>(unmarshaller.getSpectrumIds()));\n",
      "}\n",
      "1195\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 424/1243 [14:34<29:06,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20454\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 425/1243 [14:35<26:49,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20461\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 426/1243 [14:38<27:49,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20493\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 427/1243 [14:40<27:53,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20502\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  34%|      | 428/1243 [14:42<27:46,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20503\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class SpectraData {\\n    private CvParam spectrumIdFormat;;\\n    public SpectraData(String id, String name, String location, CvParam fileFormat, String externalFormatDocumentationURI, CvParam spectrumIdFormat);\\n    public CvParam getSpectrumIdFormat()\\n    public void setSpectrumIdFormat(CvParam spectrumIdFormat)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/core/SpectraData.java', 'methodCount': 5, 'lineCount': 8}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "   * Return the SpectrumFile format for an specific path such as: /myppath/spectrum_file.mgf\n",
      "   *\n",
      "   * @param path the specific path\n",
      "   * @return the SpectrumFile format such as MZXML or PKL\n",
      "   */\n",
      "public static SpecFileFormat getSpecFileFormatFromLocation(String path)\n",
      "**************************************\n",
      "public static SpecFileFormat getSpecFileFormatFromLocation(String path) {\n",
      "    if (path != null && path.length() > 0) {\n",
      "        if (path.toUpperCase().contains(MZXML_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.MZXML;\n",
      "        if (path.toUpperCase().contains(DTA_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.DTA;\n",
      "        if (path.toUpperCase().contains(MGF_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.MGF;\n",
      "        if (path.toUpperCase().contains(XML_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.MZDATA;\n",
      "        if (path.toUpperCase().contains(MZML_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.MZML;\n",
      "        if (path.toUpperCase().contains(APL_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.APL;\n",
      "        if (path.toUpperCase().contains(PKL_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.PKL;\n",
      "        if (path.toUpperCase().contains(MS2_EXT.toUpperCase()))\n",
      "            return SpecFileFormat.MS2;\n",
      "    }\n",
      "    return SpecFileFormat.NONE;\n",
      "}\n",
      "474\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 429/1243 [14:44<29:47,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 430/1243 [14:46<29:24,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20505\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 431/1243 [14:48<28:49,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20515\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 432/1243 [14:50<26:13,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20516\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 433/1243 [14:51<24:53,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20519\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Get the index of sublist using a given object\n",
      "     *\n",
      "     * @param o input object\n",
      "     * @return int index of the object\n",
      "     */\n",
      "public int indexOf(Object o)\n",
      "**************************************\n",
      "public int indexOf(Object o) {\n",
      "    checkArgument(o);\n",
      "    int cnt = 0;\n",
      "    for (Object anOriginalList : this.originalList) {\n",
      "        if (clazz.isInstance(anOriginalList)) {\n",
      "            if (o.equals(anOriginalList)) {\n",
      "                return cnt;\n",
      "            }\n",
      "            cnt++;\n",
      "        }\n",
      "    }\n",
      "    return -1;\n",
      "}\n",
      "150\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 434/1243 [14:54<27:09,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20548\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 435/1243 [14:56<26:12,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20756\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 436/1243 [14:58<26:44,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20760\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 437/1243 [15:00<26:52,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20761\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 438/1243 [15:02<26:34,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20764\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'CvParam': 'public class CvParam {\\n    /**\\n * Cv term accession\\n */\\nprivate String accession;;\\n    /**\\n * Id of the referenced CvLookup object\\n */\\nprivate String cvLookupID;;\\n    public CvParam(String accession, String name, String cvLookupID, String value, String unitAcc, String unitName, String unitCVLookupID);\\n    public String getCvLookupID()\\n    public void setCvLookupID(String cvLookupID)\\n    public void setAccession(String accession)\\n    public String getAccession()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Get cell line cv param of a sub sample with given index\n",
      "     *\n",
      "     * @param index sub sample index\n",
      "     * @return CvParam cell line cv param\n",
      "     */\n",
      "public CvParam getCellLine(int index)\n",
      "**************************************\n",
      "public CvParam getCellLine(int index) {\n",
      "    CvParam cellLine = null;\n",
      "    if (index >= 0 && index < MAX_SUB_SAMPLE_SIZE) {\n",
      "        SubSample sample = samples[index - 1];\n",
      "        if (sample != null) {\n",
      "            cellLine = sample.getCellLine();\n",
      "        }\n",
      "    }\n",
      "    return cellLine;\n",
      "}\n",
      "306\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 439/1243 [15:04<27:07,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20765\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 440/1243 [15:06<27:16,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20768\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  35%|      | 441/1243 [15:08<27:32,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20769\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 442/1243 [15:10<27:32,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20772\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 443/1243 [15:12<28:19,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20773\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'CvParam': 'public class CvParam {\\n    /**\\n * Cv term accession\\n */\\nprivate String accession;;\\n    /**\\n * Id of the referenced CvLookup object\\n */\\nprivate String cvLookupID;;\\n    public CvParam(String accession, String name, String cvLookupID, String value, String unitAcc, String unitName, String unitCVLookupID);\\n    public String getCvLookupID()\\n    public void setCvLookupID(String cvLookupID)\\n    public void setAccession(String accession)\\n    public String getAccession()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Add reagent information to a subsample\n",
      "     *\n",
      "     * @param cvParam reagent cv param\n",
      "     */\n",
      "public void setReagent(CvParam cvParam)\n",
      "**************************************\n",
      "public void setReagent(CvParam cvParam) {\n",
      "    int index = getSubSampleIndex(cvParam.getValue());\n",
      "    if (index > 0 && index <= MAX_SUB_SAMPLE_SIZE) {\n",
      "        SubSample sample = samples[index - 1];\n",
      "        if (sample == null) {\n",
      "            sample = new SubSample(index);\n",
      "            samples[index - 1] = sample;\n",
      "        }\n",
      "        sample.setReagent(cvParam);\n",
      "    }\n",
      "}\n",
      "314\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 444/1243 [15:15<29:32,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20776\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 445/1243 [15:16<27:11,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20777\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 446/1243 [15:19<27:34,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20780\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 447/1243 [15:21<27:16,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20781\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 448/1243 [15:23<27:47,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20784\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'CvParam': 'public class CvParam {\\n    /**\\n * Cv term accession\\n */\\nprivate String accession;;\\n    /**\\n * Id of the referenced CvLookup object\\n */\\nprivate String cvLookupID;;\\n    public CvParam(String accession, String name, String cvLookupID, String value, String unitAcc, String unitName, String unitCVLookupID);\\n    public String getCvLookupID()\\n    public void setCvLookupID(String cvLookupID)\\n    public void setAccession(String accession)\\n    public String getAccession()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Get description cv param of a sub sample with given index\n",
      "     *\n",
      "     * @param index sub sample index\n",
      "     * @return CvParam description cv param\n",
      "     */\n",
      "public CvParam getDescription(int index)\n",
      "**************************************\n",
      "public CvParam getDescription(int index) {\n",
      "    CvParam description = null;\n",
      "    if (index >= 0 && index < MAX_SUB_SAMPLE_SIZE) {\n",
      "        SubSample sample = samples[index - 1];\n",
      "        if (sample != null) {\n",
      "            description = sample.getDescription();\n",
      "        }\n",
      "    }\n",
      "    return description;\n",
      "}\n",
      "296\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 449/1243 [15:25<28:38,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20785\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 450/1243 [15:27<26:04,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20787\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 451/1243 [15:29<27:31,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20825\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 452/1243 [15:31<27:28,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20986\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  36%|      | 453/1243 [15:34<29:56,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21511\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 12, 'classCount': 1, 'methodsOnly': 'public class MzIdentML {\\n    private static final long serialVersionUID = 100L;;\\n    @XmlAttribute(required = true)\\nprotected String version;;\\n    @XmlAttribute\\nprotected String creationDate;;\\n    @XmlElement(required = true)\\nprotected CvList cvList;;\\n    @XmlElement(name = \"AnalysisSoftwareList\")\\nprotected AnalysisSoftwareList analysisSoftwareList;;\\n    @XmlElement(name = \"Provider\")\\nprotected Provider provider;;\\n    @XmlElement(name = \"AuditCollection\")\\nprotected AuditCollection auditCollection;;\\n    @XmlElement(name = \"AnalysisSampleCollection\")\\nprotected AnalysisSampleCollection analysisSampleCollection;;\\n    @XmlElement(name = \"SequenceCollection\")\\nprotected SequenceCollection sequenceCollection;;\\n    @XmlElement(name = \"AnalysisProtocolCollection\", required = true)\\nprotected AnalysisProtocolCollection analysisProtocolCollection;;\\n    @XmlElement(name = \"DataCollection\", required = true)\\nprotected DataCollection dataCollection;;\\n    @XmlElement(name = \"BibliographicReference\")\\nprotected List<BibliographicReference> bibliographicReference;;\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/lightModel/MzIdentML.java', 'methodCount': 0, 'lineCount': 14}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 8}\n",
      "{'FastMzIdentMLUnmarshaller': 'public class FastMzIdentMLUnmarshaller {\\n    private MzIdentML mzIdentML = null;;\\n    private JAXBContext jaxbContext;;\\n    private static volatile FastMzIdentMLUnmarshaller fastMzIdentMLUnmarshaller = null;;\\n    private FastMzIdentMLUnmarshaller(File mzIdentMLFile);\\n    public static FastMzIdentMLUnmarshaller getInstance(File mzIdentMLFile)\\n    public MzIdentML getMzIdentML()\\n    public void destroy()\\n}\\n'}\n",
      "/**\n",
      "   * This is the method should use to instantiate fastMzIdentMLUnmarshaller object. This method is\n",
      "   * thread safe and also do a lazy loading because of two reasons: 1. XML parsing from mzIdentML\n",
      "   * file is a memory intensive task 2. MzIdentML is an object with high memory consumption\n",
      "   *\n",
      "   * @param mzIdentMLFile Input MzIdentML file with .mzid extension\n",
      "   * @return FastMzIdentMLUnmarshaller\n",
      "   */\n",
      "public static FastMzIdentMLUnmarshaller getInstance(File mzIdentMLFile)\n",
      "**************************************\n",
      "/**\n",
      " * This is the method should use to instantiate fastMzIdentMLUnmarshaller object. This method is\n",
      " * thread safe and also do a lazy loading because of two reasons: 1. XML parsing from mzIdentML\n",
      " * file is a memory intensive task 2. MzIdentML is an object with high memory consumption\n",
      " *\n",
      " * @param mzIdentMLFile Input MzIdentML file with .mzid extension\n",
      " * @return FastMzIdentMLUnmarshaller\n",
      " */\n",
      "public static FastMzIdentMLUnmarshaller getInstance(File mzIdentMLFile) {\n",
      "    try {\n",
      "        if (fastMzIdentMLUnmarshaller == null) {\n",
      "            synchronized (FastMzIdentMLUnmarshaller.class) {\n",
      "                if (fastMzIdentMLUnmarshaller == null) {\n",
      "                    fastMzIdentMLUnmarshaller = new FastMzIdentMLUnmarshaller(mzIdentMLFile);\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    } catch (Exception e) {\n",
      "        log.error(\"mzIdentML file access error : \" + e.getMessage());\n",
      "    }\n",
      "    return fastMzIdentMLUnmarshaller;\n",
      "}\n",
      "805\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 454/1243 [15:36<31:00,  2.36s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21618\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 455/1243 [15:38<29:49,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21626\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 456/1243 [15:41<29:20,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 457/1243 [15:43<30:10,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21631\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 458/1243 [15:46<31:25,  2.40s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21633\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class LightModelsTransformer {\\n    private final static Map<String, CVLookup> cvLookupMap = new HashMap<>();;\\n    public static void setCvLookupMap(Map<String, CVLookup> cvLookupList)\\n    public static Software transformToSoftware(uk.ac.ebi.pride.utilities.data.lightModel.AnalysisSoftware analysisSoftware)\\n    public static List<SourceFile> transformToSourceFiles(List<uk.ac.ebi.pride.utilities.data.lightModel.SourceFile> sourceFilesLight)\\n    public static List<CvParam> transformToCvParam(List<uk.ac.ebi.pride.utilities.data.lightModel.CvParam> cvParamsLight)\\n    public static Set<CvParam> transformToCvParam(Set<uk.ac.ebi.pride.utilities.data.lightModel.CvParam> cvParamsLight)\\n    public static CvParam transformToCvParam(uk.ac.ebi.pride.utilities.data.lightModel.CvParam cvParamLight)\\n    public static List<UserParam> transformToUserParam(List<uk.ac.ebi.pride.utilities.data.lightModel.UserParam> userParamsLight)\\n    public static UserParam transformToUserParam(uk.ac.ebi.pride.utilities.data.lightModel.UserParam userParam)\\n    public static List<CVLookup> transformCVList(List<uk.ac.ebi.pride.utilities.data.lightModel.Cv> cvList)\\n    public static CVLookup transformToCVLookup(uk.ac.ebi.pride.utilities.data.lightModel.Cv CvLight)\\n    public static List<Person> transformToPerson(List<uk.ac.ebi.pride.utilities.data.lightModel.Person> personsLight)\\n    public static Person transformToPerson(uk.ac.ebi.pride.utilities.data.lightModel.Person lightPerson)\\n    public static List<Organization> transformToOrganization(List<uk.ac.ebi.pride.utilities.data.lightModel.Organization> lightOrganizations)\\n    public static Organization transformToOrganization(uk.ac.ebi.pride.utilities.data.lightModel.Organization lightOrganization)\\n    public static List<Organization> transformAffiliationToOrganization(List<uk.ac.ebi.pride.utilities.data.lightModel.Affiliation> lightAffiliations)\\n    public static Provider transformToProvider(uk.ac.ebi.pride.utilities.data.lightModel.Provider lightProvider)\\n    public static List<Reference> transformToReference(List<uk.ac.ebi.pride.utilities.data.lightModel.BibliographicReference> bibliographicReference)\\n    public static SpectraData transformToSpectraData(uk.ac.ebi.pride.utilities.data.lightModel.SpectraData lightSpectraData, boolean mgfTitle)\\n    public static List<SpectraData> transformToSpectraData(List<uk.ac.ebi.pride.utilities.data.lightModel.SpectraData> lightSpectraDataList, List<Comparable> usedTitle)\\n    public static CvParam transformDateToCvParam(Date creationDate)\\n    public static List<SearchDataBase> transformToSearchDataBase(List<uk.ac.ebi.pride.utilities.data.lightModel.SearchDatabase> lightSearchDatabases)\\n    public static SearchDataBase transformToSeachDatabase(uk.ac.ebi.pride.utilities.data.lightModel.SearchDatabase lightDatabase)\\n    public static String getValueFromCvTerm(Parameter cvTerm)\\n    public static List<Sample> transformToSample(List<uk.ac.ebi.pride.utilities.data.lightModel.Sample> lightSamples)\\n    public static List<Sample> transformSubSampleToSample(List<uk.ac.ebi.pride.utilities.data.lightModel.SubSample> lightSamples)\\n    public static Sample transformToSample(uk.ac.ebi.pride.utilities.data.lightModel.Sample lightSample)\\n    public static Map<Contact, CvParam> transformToRoleList(List<uk.ac.ebi.pride.utilities.data.lightModel.ContactRole> contactRoles)\\n    public static List<Enzyme> transformToEnzyme(List<uk.ac.ebi.pride.utilities.data.lightModel.Enzyme> oldEnzymes)\\n    private static Enzyme transformToEnzyme(uk.ac.ebi.pride.utilities.data.lightModel.Enzyme oldEnzyme)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/controller/impl/Transformer/LightModelsTransformer.java', 'methodCount': 29, 'lineCount': 32}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SourceFile {\\n    /**\\n * A URI to access documentation and tools to interpret the external format\\n * of the ExternalData instance. For example, XML Schema or static libraries\\n * (APIs) to access binary formats.\\n */\\nprivate String externalFormatDocumentationURI;;\\n    /**\\n * The format of the ExternalData file, for example \"tiff\" for image files.\\n */\\nprivate CvParam fileFormat;;\\n    /**\\n * location of the source file\\n */\\nprivate String path;;\\n    public SourceFile(String name, String path);\\n    public SourceFile(ParamGroup params, String id, String name, String path);\\n    public SourceFile(ParamGroup params, String id, String name, String path, CvParam fileFormat, String externalFormatDocumentationURI);\\n    public String getPath()\\n    public void setPath(String path)\\n    public CvParam getFileFormat()\\n    public void setFileFormat(CvParam fileFormat)\\n    public String getExternalFormatDocumentationURI()\\n    public void setExternalFormatDocumentationURI(String externalFormatDocumentationURI)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/core/SourceFile.java', 'methodCount': 11, 'lineCount': 16}, {'fieldCount': 24, 'classCount': 1, 'methodsOnly': 'public class Constants {\\n    /**\\n * algebraic sign\\n */\\nprivate static final String SIGN = \"[+-]\";;\\n    public static final String TAB = \"\\\\t\";;\\n    public static final String LINE_SEPARATOR = \"\\\\n\";;\\n    /**\\n * integer expression\\n */\\npublic static final String INTEGER = SIGN + \"?\\\\\\\\d+\";;\\n    public static final String VERSION_NUMBER = \"^(\\\\\\\\d+\\\\\\\\.)(\\\\\\\\d+\\\\\\\\.)(\\\\\\\\d+)$\";;\\n    public static final String MZIDENTML_ID_PATTERN = \"^[a-zA-Z0-9.]*$\";;\\n    private Constants();\\n    public static final String NOT_AVAILABLE = \"N/A\";;\\n    public static final String MGF_EXT = \".mgf\";;\\n    public static final String DTA_EXT = \".dta\";;\\n    public static final String MS2_EXT = \".ms2\";;\\n    public static final String PKL_EXT = \".pkl\";;\\n    public static final String MZXML_EXT = \".mzxml\";;\\n    public static final String APL_EXT = \".apl\";;\\n    public static final String XML_EXT = \".xml\";;\\n    public static final String MZML_EXT = \".mzML\";;\\n    public static final String MZTAB_EXT = \".mztab\";;\\n    public static final String WIFF_EXT = \".wiff\";;\\n    public static final String NETCDF_EXT = \".cdf\";;\\n    public static final String MZIDENTML_NAMESPACE = \"http://psidev.info/psi/pi/mzIdentML/1.1\";;\\n    public static final Double DELTATHESHOLD = 4.0;;\\n    public static final String PSI_MOD = \"MOD\";;\\n    public static final String MS = \"MS\";;\\n    public static final String UNIMOD = \"UNIMOD\";;\\n    public static final String ANCHOR_PROTEIN = \"MS:1001591\";;\\n    public static SpecFileFormat getSpecFileFormat(String fileFormat)\\n    public static SpecFileFormat getSpecFileFormatFromLocation(String path)\\n    public static Constants.SpecFileFormat getSpectraDataFormat(SpectraData spectraData)\\n    public static Constants.SpecIdFormat getSpectraDataIdFormat(String accession)\\n    public static List<SpecFileFormat> getFileTypeSupported(SpectraData spectraData)\\n    public static Constants.SpecFileFormat getDataFormatFromFileExtension(SpectraData spectradata)\\n    public static Class getFileType(File file)\\n}\\n', 'filePath': '/home/ubuntu/repos/PRIDE-Utilities,ms-data-core-api/src/main/java/uk/ac/ebi/pride/utilities/data/utils/Constants.java', 'methodCount': 8, 'lineCount': 34}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "   * Get creation date of the MzIdentML file. This is a optional attribute\n",
      "   *\n",
      "   * @return the date that the MzIdentML file created\n",
      "   */\n",
      "public Date getCreationDate() throws Exception\n",
      "**************************************\n",
      "public Date getCreationDate() throws Exception {\n",
      "    Date dateCreation = null;\n",
      "    String dateInString = fastMzIdentMLUnmarshaller.getMzIdentML().getCreationDate();\n",
      "    try {\n",
      "        if (dateInString != \"\") {\n",
      "            Calendar calendar = javax.xml.bind.DatatypeConverter.parseDateTime(dateInString);\n",
      "            dateCreation = calendar.getTime();\n",
      "        }\n",
      "    } catch (Exception ex) {\n",
      "        throw new Exception(\"Failed to convert creationDate attribute of the MzIdentML Element\", ex);\n",
      "    }\n",
      "    return dateCreation;\n",
      "}\n",
      "2084\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 459/1243 [15:47<28:19,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21693\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 460/1243 [15:49<26:19,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21700\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 461/1243 [15:51<26:27,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21711\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 462/1243 [15:53<25:00,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21782\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 463/1243 [15:54<23:58,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21796\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Acquires.\n",
      "     */\n",
      "public void acquire()\n",
      "**************************************\n",
      "/**\n",
      " * Acquires.\n",
      " */\n",
      "public void acquire() {\n",
      "    try {\n",
      "        mutex.acquire();\n",
      "    } catch (Exception e) {\n",
      "        throw new RuntimeException(e);\n",
      "    }\n",
      "}\n",
      "84\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 464/1243 [15:55<20:57,  1.61s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21797\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 465/1243 [15:57<20:38,  1.59s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21818\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  37%|      | 466/1243 [15:58<20:20,  1.57s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21824\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 467/1243 [16:00<20:20,  1.57s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21829\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 468/1243 [16:02<20:12,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21839\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * This method returns the coded column type for the specified column.\n",
      "     *\n",
      "     * @param   aColumnName String with the column name.\n",
      "     * @return  int with the coded column type for the specified column, or\n",
      "     *              '-1' if the specified column was not found.\n",
      "     */\n",
      "public int getCodedColumnType(String aColumnName)\n",
      "**************************************\n",
      "public int getCodedColumnType(String aColumnName) {\n",
      "    int result = -1;\n",
      "    for (int i = 0; i < iColumns.length; i++) {\n",
      "        String lColumn = iColumns[i].trim();\n",
      "        if (aColumnName.trim().equalsIgnoreCase(lColumn)) {\n",
      "            result = iColTypes[i];\n",
      "            break;\n",
      "        }\n",
      "    }\n",
      "    return result;\n",
      "}\n",
      "188\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 469/1243 [16:03<20:05,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21840\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 470/1243 [16:05<20:05,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21841\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 471/1243 [16:07<23:18,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21859\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 472/1243 [16:09<24:39,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21888\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 473/1243 [16:11<24:18,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21899\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public interface WaitingHandler {\\n    /**\\n * Convenience date format.\\n */\\npublic static final SimpleDateFormat SIMPLE_DATA_FORMAT = new SimpleDateFormat(\"dd MMM yyyy, HH:mm\");;\\n    /**\\n * The tab space to add when using HTML.\\n */\\npublic static final String TAB_HTML = \"&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\";;\\n    /**\\n * The tab space to add when not using HTML.\\n */\\npublic static final String TAB_NON_HTML = \"        \";;\\n    public void setPrimaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setMaxPrimaryProgressCounter(int maxProgressValue)\\n    public void increasePrimaryProgressCounter()\\n    public void increasePrimaryProgressCounter(int increment)\\n    public void setPrimaryProgressCounter(int value)\\n    public void setMaxSecondaryProgressCounter(int maxProgressValue)\\n    public void resetPrimaryProgressCounter()\\n    public void resetSecondaryProgressCounter()\\n    public void increaseSecondaryProgressCounter()\\n    public void setSecondaryProgressCounter(int value)\\n    public void increaseSecondaryProgressCounter(int value)\\n    public void setSecondaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setRunFinished()\\n    public void setRunCanceled()\\n    public void appendReport(String report, boolean includeDate, boolean addNewLine)\\n    public void appendReportNewLineNoDate()\\n    public void appendReportEndLine()\\n    public boolean isRunCanceled()\\n    public boolean isRunFinished()\\n    public void setSecondaryProgressText(String text)\\n    public boolean isReport()\\n    public void setWaitingText(String text)\\n    public int getPrimaryProgressCounter()\\n    public int getMaxPrimaryProgressCounter()\\n    public int getSecondaryProgressCounter()\\n    public int getMaxSecondaryProgressCounter()\\n    public void setDisplayProgress(boolean displayProgress)\\n    public boolean getDisplayProgress()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/waiting/WaitingHandler.java', 'methodCount': 28, 'lineCount': 33}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 8}\n",
      "{}\n",
      "/**\n",
      "     * Returns the object if present in the cache. Null if not.\n",
      "     *\n",
      "     * @param objectKey the key of the object\n",
      "     *\n",
      "     * @return the object of interest, null if not present in the cache\n",
      "     */\n",
      "public Object getObject(long objectKey)\n",
      "**************************************\n",
      "public Object getObject(long objectKey) {\n",
      "    Object object = null;\n",
      "    loadObjectMutex.acquire();\n",
      "    if (loadedObjects.containsKey(objectKey)) {\n",
      "        object = loadedObjects.get(objectKey).object;\n",
      "    }\n",
      "    loadObjectMutex.release();\n",
      "    return object;\n",
      "}\n",
      "620\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 474/1243 [16:13<23:22,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21908\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 475/1243 [16:15<23:13,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21910\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 476/1243 [16:17<24:25,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21928\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 477/1243 [16:19<24:14,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21958\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  38%|      | 478/1243 [16:20<23:01,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21978\n",
      "{'classCount': 5, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class IsotopicDistribution {\\n    // Class specific log4j logger for AASequenceImpl instances.\\nLogger logger = LogManager.getLogger(IsotopicDistribution.class);;\\n    /**\\n * The result of the isotopic distributions calculation. Percentage of the total contribution\\n */\\nprivate Vector<Double> iPercTot = null;;\\n    /**\\n * The result of the isotopic distributions calculation. Percentage of the contribution compared to the maximum\\n */\\nprivate Vector<Double> iPercMax = null;;\\n    /**\\n * The molecular formula\\n */\\nprivate MolecularFormula iMolecularFormula;;\\n    /**\\n * boolean that indicates if a LABEL isotopic distribution must be used\\n */\\nprivate boolean iLabel = false;;\\n    /**\\n * int with the dalton difference\\n */\\nprivate int iLabelDaltonDifference;;\\n    public IsotopicDistribution(MolecularFormula lFormula);\\n    public IsotopicDistribution(MolecularFormula lFormula, int lDaltonDifference);\\n    public void setLabelDifference(int lLabelDifference)\\n    public void calculate()\\n    public IsotopicDistribution(int lC, int lN, int lH, int lO, int lS);\\n    public Double[] getPercMax()\\n    public Double[] getPercTot()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/general/IsotopicDistribution.java', 'methodCount': 7, 'lineCount': 15}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Sequence {\\n    public void setSequence(String aSequence)\\n    public String getSequence()\\n    public double getMass()\\n    public int getLength()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/interfaces/Sequence.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public interface Modification {\\n    /**\\n * Constant to indicate the position of the monoisotopic mass in the mass array.\\n */\\npublic int MONOISOTOPIC = 0;;\\n    /**\\n * Constant to indicate the position of the average mass in the mass array.\\n */\\npublic int AVERAGE = 1;;\\n    /**\\n * The residue code for the N-terminus.\\n */\\npublic String NTERMINUS = \"0\";;\\n    /**\\n * The residue code for the C-terminus.\\n */\\npublic String CTERMINUS = \"1\";;\\n    public abstract int getLocation()\\n    public abstract boolean isArtifact()\\n    public abstract String getCode()\\n    public abstract String getTitle()\\n    public abstract double getMonoisotopicMassDelta(String aResidue)\\n    public abstract double getAverageMassDelta(String aResidue)\\n    public abstract void setLocation(int aLocation)\\n    public abstract Collection getResidues()\\n    public abstract String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/interfaces/Modification.java', 'methodCount': 9, 'lineCount': 15}, {'fieldCount': 10, 'classCount': 1, 'methodsOnly': 'public class MassCalc {\\n    // Class specific log4j logger for MassCalc instances.\\nstatic Logger logger = LogManager.getLogger(MassCalc.class);;\\n    /**\\n * This variable highlights which map was chosen for the element masses.\\n */\\nprivate int iChosen = -1;;\\n    /**\\n * This hash will contain all the masses for the currently selected\\n * element list.\\n */\\nprivate HashMap masses = null;;\\n    /**\\n * This Vector stores all the HashMaps that can be selected as element\\n * lists. <br>\\n * Please only access them via the proper index, as defined in static vars\\n * on this class.\\n */\\nprivate static Vector allMaps = new Vector(2);;\\n    /**\\n * An index into an array.\\n */\\nprivate static final int ELEMENT = 0;;\\n    /**\\n * An index into an array.\\n */\\nprivate static final int MULTIPLICITY = 1;;\\n    /**\\n * Index for the monoisotopic masses of the biochemically relevant\\n * elements.\\n */\\npublic static final int MONOELEMENTS = 0;;\\n    /**\\n * Index for the monoisotopic aminoacid masses.\\n */\\npublic static final int MONOAA = 1;;\\n    /**\\n * Index for the monoisotopic nucleotide masses.\\n */\\npublic static final int MONONUCLEOTIDES = 2;;\\n    /**\\n * Value for the self-defined masses.\\n */\\npublic static final int SELFDEFINED = -1;;\\n    public MassCalc();\\n    public MassCalc(int aMassListIdentifier);\\n    public MassCalc(String aFilename);\\n    public MassCalc(HashMap aElementList);\\n    public MassCalc(int aMassListIdentifier, HashMap aSupplElementList);\\n    public double calculateMass(String aFormula) throws UnknownElementMassException\\n    private void addResultToBrutoFormula(String aElement, int aMultiplicity, HashMap aBruto)\\n    private Object[] getElement(StringCharacterIterator aSCI)\\n    private int getMultiplicity(StringCharacterIterator aSCI)\\n    private double getInnerFormulaMass(StringCharacterIterator aSCI) throws UnknownElementMassException\\n    private String isolateInnerPartString(StringCharacterIterator aSCI, char aOpener, char aCloser, boolean aKeepTokens)\\n    private HashMap loadMassesFromPropFile(String aFilename)\\n    public static void main(String[] args)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/general/MassCalc.java', 'methodCount': 13, 'lineCount': 25}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UnknownElementMassException {\\n    // Class specific log4j logger for UnknownElementMassException instances.\\nLogger logger = LogManager.getLogger(UnknownElementMassException.class);;\\n    /**\\n * The element symbol that was not recognized.\\n */\\nprivate String element = null;;\\n    public UnknownElementMassException(String aElement);\\n    public String getElement()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/general/UnknownElementMassException.java', 'methodCount': 2, 'lineCount': 6}], 'asteriskCount': 0, 'importCount': 5, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * Simple setter for the sequence. It also clears the mass cache.\n",
      "     *\n",
      "     * @param\taSequence\tString with the sequence to be set.\n",
      "     */\n",
      "public void setSequence(String aSequence)\n",
      "**************************************\n",
      "/**\n",
      " * Simple setter for the sequence. It also clears the mass cache.\n",
      " *\n",
      " * @param\taSequence\tString with the sequence to be set.\n",
      " */\n",
      "public void setSequence(String aSequence) {\n",
      "    // Nullpointer check.\n",
      "    if (aSequence == null) {\n",
      "        throw new NullPointerException(\"Sequence cannot be 'null'!\\n\");\n",
      "    } else if (aSequence.trim().equals(\"\")) {\n",
      "        throw new IllegalArgumentException(\"Sequence cannot be empty String!\\n\");\n",
      "    }\n",
      "    this.iSequence = aSequence.trim();\n",
      "    this.iMass = -1.0;\n",
      "    this.iGravyCached = false;\n",
      "    this.iMeekCached = false;\n",
      "}\n",
      "1705\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 479/1243 [16:22<23:56,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22002\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 480/1243 [16:24<24:53,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22029\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 481/1243 [16:27<26:01,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22037\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 482/1243 [16:29<26:09,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22041\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 483/1243 [16:30<24:35,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22063\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Scales the given values using the maximum value.\n",
      "     *\n",
      "     * @param values The values to scale.\n",
      "     * @param percent If true, the values are returned in the [0-100] range.\n",
      "     *\n",
      "     * @return The given values scaled using the maximum value.\n",
      "     */\n",
      "public static double[] scaleToMax(double[] values, boolean percent)\n",
      "**************************************\n",
      "public static double[] scaleToMax(double[] values, boolean percent) {\n",
      "    double maxValue = Arrays.stream(values).max().orElse(1.0);\n",
      "    return scaleValues(values, maxValue, percent);\n",
      "}\n",
      "156\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 484/1243 [16:32<23:05,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22082\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 485/1243 [16:33<22:01,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22185\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 486/1243 [16:36<25:06,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22407\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 487/1243 [16:38<25:26,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22448\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 488/1243 [16:40<25:28,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22507\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ExperimentObject {\\n    /**\\n * Unique identifier.\\n */\\nprivate long id;;\\n    public ExperimentObject();\\n    public long getId()\\n    public void setId(long id)\\n    private final static long[] HASHVALUELIST = { -4785187655948605440L, 2351437082613919744L, -7620983146772158464L, -8747902906952306688L, -4496485731609829376L, -4047692981156677632L, 5720028397227347968L, -4748060763626276864L, 2063365725770047488L, -5756543758428897280L, -134698081630425088L, 867726525032218624L, -526450428666544128L, 4146020926348189696L, 4362296343029680128L, -672990070253072384L, -2559490283472277504L, 3187632952876974080L, -5716989432807307264L, -8332013824838645760L, 4253744629365506048L, 2097316067254513664L, 8951627463544416256L, -5600031980443258880L, 6380991404691560448L, 8903284868402118656L, -1115601857539225600L, 4631654322507227136L, 7771989044436795392L, 7773688932940122112L, -6019734220953055232L, 3392712990065328128L, -8921384047543447552L, -7767842613008707584L, -1186522791511611392L, -6926112736333537280L, 8736653739320072192L, 8867062073843642368L, 6298992568012455936L, -6831107491093487616L, -7084666134747267072L, -1674183307215181824L, 7180054879733344256L, -1774408477150697472L, -1102347028329271296L, 2837232313405440000L, 6789844965029836800L, -2021979153929187328L, -803643088872329216L, -6635474898121947136L, -1519775710292529152L, -7017431056149018624L, 8399941098113230848L, 6620078501932513280L, 8402686423795523584L, 7887825026517880832L, 6240511391300272128L, -2116326148598433792L, 3164957088731514880L, 6354445331039899648L, -2421944411545827328L, -6588274517877174272L, -5482092713179058176L, 1515440486213902336L, -3383185261582667776L, -2725557693718710272L, 2180993613288613888L, -4878984385226620928L, 4905597879284899840L, -8937278576235966464L, -4857623260077275136L, -6678664042745585664L, 6590419491356596224L, 3898378085667969024L, -8773012746479065088L, -4316629602317574144L, -578020323565103104L, 5815789437630859264L, 1330829571824838656L, 2058704620696928256L, 5775301559630338048L, -4128281782811285504L, -6189976155577464832L, -2204893487149668352L, -4107985148748068864L, -2803177563490273280L, 7139083951461890048L, -6547891565468342272L, 3512976861638146048L, 8446989268574042112L, -6262309160844883968L, -447362214463838208L, -4695191602764636160L, -8777129286526107648L, -2322220230279856128L, -3376371221541236736L, -352816524822126592L, -6489602716775188480L, -4340386299073419264L, -411238008103813120L, -7548606038504292352L, 3950672770391547904L, 1570846774247147520L, 2087897268844892160L, -6691005528687374336L, 1651506531346769920L, -9105826395118237696L, -920643688498837504L, 6741095098680469504L, -9196666188088858624L, 4628592761082419200L, 1805632260469598208L, -2595685425333377024L, -2001876750192766976L, 4498796613205751808L, -3322677024598908928L, 8658129466298726400L, 2854559136171276288L, 106897466552897536L, 5590481524594866176L, -4319460978758043648L, 1896969526688425984L, -2860223852340688896L, -2273634011107659776L, -6830438329227218944L, -1024496407927033856L, -1561168395559655424L, -1430574524350681088L };;\\n    public static final long NO_KEY = asLong(\"#!#_NO_KEY_#!#\");;\\n    private HashMap<Long, UrParameter> urParams = null;;\\n    public void addUrParam(UrParameter parameter)\\n    public void removeUrParam(long paramterKey)\\n    private synchronized void createParamsMap()\\n    public UrParameter getUrParam(UrParameter parameter)\\n    public void clearParametersMap()\\n    public void setUrParams(HashMap<Long, UrParameter> urParams)\\n    public HashMap<Long, UrParameter> getUrParams()\\n    public static long asLong(String key)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/personalization/ExperimentObject.java', 'methodCount': 11, 'lineCount': 17}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Returns a short description of the parameters.\n",
      "     *\n",
      "     * @return a short description of the parameters\n",
      "     */\n",
      "public String getShortDescription()\n",
      "**************************************\n",
      "public String getShortDescription() {\n",
      "    String newLine = System.getProperty(\"line.separator\");\n",
      "    StringBuilder output = new StringBuilder();\n",
      "    output.append(\"Protein FDR: \").append(defaultProteinFDR).append(\".\").append(newLine);\n",
      "    output.append(\"Peptide FDR: \").append(defaultPeptideFDR).append(\".\").append(newLine);\n",
      "    output.append(\"PSM FDR: \").append(defaultPsmFDR).append(\".\").append(newLine);\n",
      "    return output.toString();\n",
      "}\n",
      "1699\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 489/1243 [16:42<26:13,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23168\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  39%|      | 490/1243 [16:44<24:29,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23183\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 491/1243 [16:46<25:29,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23189\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 492/1243 [16:48<25:44,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23190\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 493/1243 [16:50<24:14,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23191\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 10, 'classCount': 1, 'methodsOnly': 'public class MassCalc {\\n    // Class specific log4j logger for MassCalc instances.\\nstatic Logger logger = LogManager.getLogger(MassCalc.class);;\\n    /**\\n * This variable highlights which map was chosen for the element masses.\\n */\\nprivate int iChosen = -1;;\\n    /**\\n * This hash will contain all the masses for the currently selected\\n * element list.\\n */\\nprivate HashMap masses = null;;\\n    /**\\n * This Vector stores all the HashMaps that can be selected as element\\n * lists. <br>\\n * Please only access them via the proper index, as defined in static vars\\n * on this class.\\n */\\nprivate static Vector allMaps = new Vector(2);;\\n    /**\\n * An index into an array.\\n */\\nprivate static final int ELEMENT = 0;;\\n    /**\\n * An index into an array.\\n */\\nprivate static final int MULTIPLICITY = 1;;\\n    /**\\n * Index for the monoisotopic masses of the biochemically relevant\\n * elements.\\n */\\npublic static final int MONOELEMENTS = 0;;\\n    /**\\n * Index for the monoisotopic aminoacid masses.\\n */\\npublic static final int MONOAA = 1;;\\n    /**\\n * Index for the monoisotopic nucleotide masses.\\n */\\npublic static final int MONONUCLEOTIDES = 2;;\\n    /**\\n * Value for the self-defined masses.\\n */\\npublic static final int SELFDEFINED = -1;;\\n    public MassCalc();\\n    public MassCalc(int aMassListIdentifier);\\n    public MassCalc(String aFilename);\\n    public MassCalc(HashMap aElementList);\\n    public MassCalc(int aMassListIdentifier, HashMap aSupplElementList);\\n    public double calculateMass(String aFormula) throws UnknownElementMassException\\n    private void addResultToBrutoFormula(String aElement, int aMultiplicity, HashMap aBruto)\\n    private Object[] getElement(StringCharacterIterator aSCI)\\n    private int getMultiplicity(StringCharacterIterator aSCI)\\n    private double getInnerFormulaMass(StringCharacterIterator aSCI) throws UnknownElementMassException\\n    private String isolateInnerPartString(StringCharacterIterator aSCI, char aOpener, char aCloser, boolean aKeepTokens)\\n    private HashMap loadMassesFromPropFile(String aFilename)\\n    public static void main(String[] args)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/general/MassCalc.java', 'methodCount': 13, 'lineCount': 25}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UnknownElementMassException {\\n    // Class specific log4j logger for UnknownElementMassException instances.\\nLogger logger = LogManager.getLogger(UnknownElementMassException.class);;\\n    /**\\n * The element symbol that was not recognized.\\n */\\nprivate String element = null;;\\n    public UnknownElementMassException(String aElement);\\n    public String getElement()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/general/UnknownElementMassException.java', 'methodCount': 2, 'lineCount': 6}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * This method attempts to parse a String into a HashMap.\n",
      "     *\n",
      "     * @param\taList\tString with the list to be parsed.\n",
      "     * @return\tHashMap\tif all goes well!\n",
      "     * @exception\tException\tif anything at all goes wrong!\n",
      "     */\n",
      "private HashMap parseSelfList(String aList) throws Exception\n",
      "**************************************\n",
      "private HashMap parseSelfList(String aList) throws Exception {\n",
      "    HashMap toReturn = new HashMap();\n",
      "    aList = aList.trim();\n",
      "    StringTokenizer st = new StringTokenizer(aList, \" =_:\\n\");\n",
      "    while (st.hasMoreTokens()) {\n",
      "        String key = st.nextToken();\n",
      "        Double value = Double.valueOf(st.nextToken());\n",
      "        toReturn.put(key, value);\n",
      "    }\n",
      "    return toReturn;\n",
      "}\n",
      "938\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 494/1243 [16:51<22:31,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23197\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 495/1243 [16:53<21:18,  1.71s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23198\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 496/1243 [16:55<22:29,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23216\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 497/1243 [16:57<21:44,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23227\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 498/1243 [16:58<20:55,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23265\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Prepares the renderer by querying the data model for the\n",
      "     * value and selection state\n",
      "     * of the cell at <code>row</code>, <code>column</code>.\n",
      "     * Returns the component (may be a <code>Component</code>\n",
      "     * or a <code>JComponent</code>) under the event location.\n",
      "     *\n",
      "     * @param renderer  the <code>TableCellRenderer</code> to prepare\n",
      "     * @param rowIndex  the row of the cell to render, where 0 is the first row\n",
      "     * @param vColIndex the column of the cell to render,\n",
      "     *\t\t\twhere 0 is the first column\n",
      "     * @return          the <code>Component</code> under the event location\n",
      "     */\n",
      "public Component prepareRenderer(TableCellRenderer renderer, int rowIndex, int vColIndex)\n",
      "**************************************\n",
      "/**\n",
      " *  Prepares the renderer by querying the data model for the\n",
      " *  value and selection state\n",
      " *  of the cell at <code>row</code>, <code>column</code>.\n",
      " *  Returns the component (may be a <code>Component</code>\n",
      " *  or a <code>JComponent</code>) under the event location.\n",
      " *\n",
      " *  @param renderer  the <code>TableCellRenderer</code> to prepare\n",
      " *  @param rowIndex  the row of the cell to render, where 0 is the first row\n",
      " *  @param vColIndex the column of the cell to render,\n",
      " * \t\t\twhere 0 is the first column\n",
      " *  @return          the <code>Component</code> under the event location\n",
      " */\n",
      "public Component prepareRenderer(TableCellRenderer renderer, int rowIndex, int vColIndex) {\n",
      "    Component c = super.prepareRenderer(renderer, rowIndex, vColIndex);\n",
      "    if (!isCellSelected(rowIndex, vColIndex)) {\n",
      "        if ((rowIndex + 1) % 2 == 0) {\n",
      "            if (iEvenRowColor == null) {\n",
      "                c.setBackground(getBackground());\n",
      "            } else {\n",
      "                c.setBackground(iEvenRowColor);\n",
      "            }\n",
      "        } else {\n",
      "            if (iUnevenRowColor == null) {\n",
      "                c.setBackground(getBackground());\n",
      "            } else {\n",
      "                c.setBackground(iUnevenRowColor);\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return c;\n",
      "}\n",
      "515\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 499/1243 [17:00<20:27,  1.65s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23295\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 500/1243 [17:02<22:36,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23302\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 501/1243 [17:04<22:15,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23382\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 502/1243 [17:05<21:40,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23774\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  40%|      | 503/1243 [17:08<23:21,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23785\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ExperimentObject {\\n    /**\\n * Unique identifier.\\n */\\nprivate long id;;\\n    public ExperimentObject();\\n    public long getId()\\n    public void setId(long id)\\n    private final static long[] HASHVALUELIST = { -4785187655948605440L, 2351437082613919744L, -7620983146772158464L, -8747902906952306688L, -4496485731609829376L, -4047692981156677632L, 5720028397227347968L, -4748060763626276864L, 2063365725770047488L, -5756543758428897280L, -134698081630425088L, 867726525032218624L, -526450428666544128L, 4146020926348189696L, 4362296343029680128L, -672990070253072384L, -2559490283472277504L, 3187632952876974080L, -5716989432807307264L, -8332013824838645760L, 4253744629365506048L, 2097316067254513664L, 8951627463544416256L, -5600031980443258880L, 6380991404691560448L, 8903284868402118656L, -1115601857539225600L, 4631654322507227136L, 7771989044436795392L, 7773688932940122112L, -6019734220953055232L, 3392712990065328128L, -8921384047543447552L, -7767842613008707584L, -1186522791511611392L, -6926112736333537280L, 8736653739320072192L, 8867062073843642368L, 6298992568012455936L, -6831107491093487616L, -7084666134747267072L, -1674183307215181824L, 7180054879733344256L, -1774408477150697472L, -1102347028329271296L, 2837232313405440000L, 6789844965029836800L, -2021979153929187328L, -803643088872329216L, -6635474898121947136L, -1519775710292529152L, -7017431056149018624L, 8399941098113230848L, 6620078501932513280L, 8402686423795523584L, 7887825026517880832L, 6240511391300272128L, -2116326148598433792L, 3164957088731514880L, 6354445331039899648L, -2421944411545827328L, -6588274517877174272L, -5482092713179058176L, 1515440486213902336L, -3383185261582667776L, -2725557693718710272L, 2180993613288613888L, -4878984385226620928L, 4905597879284899840L, -8937278576235966464L, -4857623260077275136L, -6678664042745585664L, 6590419491356596224L, 3898378085667969024L, -8773012746479065088L, -4316629602317574144L, -578020323565103104L, 5815789437630859264L, 1330829571824838656L, 2058704620696928256L, 5775301559630338048L, -4128281782811285504L, -6189976155577464832L, -2204893487149668352L, -4107985148748068864L, -2803177563490273280L, 7139083951461890048L, -6547891565468342272L, 3512976861638146048L, 8446989268574042112L, -6262309160844883968L, -447362214463838208L, -4695191602764636160L, -8777129286526107648L, -2322220230279856128L, -3376371221541236736L, -352816524822126592L, -6489602716775188480L, -4340386299073419264L, -411238008103813120L, -7548606038504292352L, 3950672770391547904L, 1570846774247147520L, 2087897268844892160L, -6691005528687374336L, 1651506531346769920L, -9105826395118237696L, -920643688498837504L, 6741095098680469504L, -9196666188088858624L, 4628592761082419200L, 1805632260469598208L, -2595685425333377024L, -2001876750192766976L, 4498796613205751808L, -3322677024598908928L, 8658129466298726400L, 2854559136171276288L, 106897466552897536L, 5590481524594866176L, -4319460978758043648L, 1896969526688425984L, -2860223852340688896L, -2273634011107659776L, -6830438329227218944L, -1024496407927033856L, -1561168395559655424L, -1430574524350681088L };;\\n    public static final long NO_KEY = asLong(\"#!#_NO_KEY_#!#\");;\\n    private HashMap<Long, UrParameter> urParams = null;;\\n    public void addUrParam(UrParameter parameter)\\n    public void removeUrParam(long paramterKey)\\n    private synchronized void createParamsMap()\\n    public UrParameter getUrParam(UrParameter parameter)\\n    public void clearParametersMap()\\n    public void setUrParams(HashMap<Long, UrParameter> urParams)\\n    public HashMap<Long, UrParameter> getUrParams()\\n    public static long asLong(String key)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/personalization/ExperimentObject.java', 'methodCount': 11, 'lineCount': 17}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'Atom': 'public class Atom {\\n    public Atom();\\n    /**\\n * The version UID for Serialization/Deserialization compatibility.\\n */\\nstatic final long serialVersionUID = 1059024301538472131L;;\\n    /**\\n * The hydrogen atom.\\n */\\npublic static final Atom H = new Hydrogen();;\\n    /**\\n * The nitrogen atom.\\n */\\npublic static final Atom N = new Nitrogen();;\\n    /**\\n * The oxygen atom.\\n */\\npublic static final Atom O = new Oxygen();;\\n    /**\\n * The carbon atom.\\n */\\npublic static final Atom C = new Carbon();;\\n    /**\\n * The sulfur atom.\\n */\\npublic static final Atom S = new Sulfur();;\\n    /**\\n * The phosphorus atom.\\n */\\npublic static final Atom P = new Phosphorus();;\\n    /**\\n * The helium atom.\\n */\\npublic static final Atom He = new Helium();;\\n    /**\\n * The phosphorus atom.\\n */\\npublic static final Atom Li = new Lithium();;\\n    /**\\n * The sodium atom.\\n */\\npublic static final Atom Na = new Sodium();;\\n    /**\\n * The selenium atom.\\n */\\npublic static final Atom Se = new Selenium();;\\n    /**\\n * The Iodine atom.\\n */\\npublic static final Atom I = new Iodine();;\\n    /**\\n * The Fluorine atom.\\n */\\npublic static final Atom F = new Fluorine();;\\n    /**\\n * The Iron atom.\\n */\\npublic static final Atom Fe = new Iron();;\\n    /**\\n * The Potassium atom.\\n */\\npublic static final Atom K = new Potassium();;\\n    /**\\n * The Calcium atom.\\n */\\npublic static final Atom Ca = new Calcium();;\\n    /**\\n * The Zinc atom.\\n */\\npublic static final Atom Zn = new Zinc();;\\n    /**\\n * The Magnesium atom.\\n */\\npublic static final Atom Mg = new Magnesium();;\\n    /**\\n * The Copper atom.\\n */\\npublic static final Atom Cu = new Copper();;\\n    /**\\n * The Chlorine atom.\\n */\\npublic static final Atom Cl = new Chlorine();;\\n    /**\\n * The Cobalt atom.\\n */\\npublic static final Atom Co = new Cobalt();;\\n    /**\\n * The Nickel atom.\\n */\\npublic static final Atom Ni = new Nickel();;\\n    /**\\n * The Manganese atom.\\n */\\npublic static final Atom Mn = new Manganese();;\\n    /**\\n * The monoisotopic mass. Access is faster then querying the isotope map.\\n */\\nprotected double monoisotopicMass;;\\n    /**\\n * Map of the isotope masses relative to the monoisotopic peak (+1 for\\n * carbon 13).\\n */\\nprotected HashMap<Integer, Double> isotopeMap;;\\n    /**\\n * Map of the isotope representative composition of the stable isotopes.\\n */\\nprotected HashMap<Integer, Double> representativeComposition;;\\n    /**\\n * The name of the atom.\\n */\\nprotected String name;;\\n    /**\\n * The symbol for the atom.\\n */\\nprotected String letter;;\\n    public static String[] getImplementedAtoms(boolean includeSelect)\\n    public static Atom getAtom(String shortName)\\n    public double getMonoisotopicMass()\\n    public String getName()\\n    public String getLetter()\\n    public ArrayList<Integer> getImplementedIsotopes()\\n    public Double getIsotopeMass(int isotopeNumber)\\n    public double getDifferenceToMonoisotopic(int isotopeNumber)\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Returns the atom as a string.\n",
      "     *\n",
      "     * @param isotopeCurlyBrackets if true, the isotopes are indicated as curly\n",
      "     * brackets after the atom, e.g. carbon 13 is written as C{13}, if false,\n",
      "     * the isotopes are indicated as a number before the atom, e.g. carbon 13 is\n",
      "     * written as 13C\n",
      "     *\n",
      "     * @return the atom as a string\n",
      "     */\n",
      "public String toString(boolean isotopeCurlyBrackets)\n",
      "**************************************\n",
      "public String toString(boolean isotopeCurlyBrackets) {\n",
      "    Atom tempAtom = Atom.getAtom(atomSymbol);\n",
      "    if (isotope == 0) {\n",
      "        return tempAtom.getLetter();\n",
      "    } else {\n",
      "        if (getMass() == null) {\n",
      "            throw new UnsupportedOperationException(\"Isotope \" + isotope + \" not implemented for atom \" + tempAtom + \".\");\n",
      "        }\n",
      "        if (isotopeCurlyBrackets) {\n",
      "            return tempAtom.getLetter() + \"{\" + Math.round(getMass()) + \"}\";\n",
      "        } else {\n",
      "            return Math.round(getMass()) + tempAtom.getLetter();\n",
      "        }\n",
      "    }\n",
      "}\n",
      "2664\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 504/1243 [17:10<23:56,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23890\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 505/1243 [17:12<24:05,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23911\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 506/1243 [17:14<24:38,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23920\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 507/1243 [17:15<23:06,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23922\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 508/1243 [17:18<24:45,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23931\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 38, 'classCount': 1, 'methodsOnly': 'public class AminoAcid {\\n    public AminoAcid();\\n    public static final AminoAcid A = new Alanine();;\\n    public static final AminoAcid C = new Cysteine();;\\n    public static final AminoAcid D = new AsparticAcid();;\\n    public static final AminoAcid E = new GlutamicAcid();;\\n    public static final AminoAcid F = new Phenylalanine();;\\n    public static final AminoAcid G = new Glycine();;\\n    public static final AminoAcid H = new Histidine();;\\n    public static final AminoAcid I = new Isoleucine();;\\n    public static final AminoAcid K = new Lysine();;\\n    public static final AminoAcid L = new Leucine();;\\n    public static final AminoAcid M = new Methionine();;\\n    public static final AminoAcid N = new Asparagine();;\\n    public static final AminoAcid P = new Proline();;\\n    public static final AminoAcid Q = new Glutamine();;\\n    public static final AminoAcid R = new Arginine();;\\n    public static final AminoAcid S = new Serine();;\\n    public static final AminoAcid T = new Threonine();;\\n    public static final AminoAcid V = new Valine();;\\n    public static final AminoAcid W = new Tryptophan();;\\n    public static final AminoAcid Y = new Tyrosine();;\\n    public static final AminoAcid U = new Selenocysteine();;\\n    public static final AminoAcid O = new Pyrrolysine();;\\n    public static final AminoAcid B = new B();;\\n    public static final AminoAcid J = new J();;\\n    public static final AminoAcid Z = new Z();;\\n    public static final AminoAcid X = new X();;\\n    /**\\n * Single letter code of the amino acid.\\n */\\npublic String singleLetterCode;;\\n    /**\\n * Three letter code of the amino acid.\\n */\\npublic String threeLetterCode;;\\n    /**\\n * Name of the amino acid.\\n */\\npublic String name;;\\n    /**\\n * The monoisotopic atom chain.\\n */\\nprotected AtomChain monoisotopicAtomChain;;\\n    /**\\n * The sub amino acids without combination.\\n */\\nprotected char[] subAminoAcidsWithoutCombination;;\\n    /**\\n * The sub amino acids with combination.\\n */\\nprotected char[] subAminoAcidsWithCombination;;\\n    /**\\n * The amino acid combinations.\\n */\\nprotected char[] aminoAcidCombinations;;\\n    /**\\n * The standard genetic code.\\n */\\nprotected String[] standardGeneticCode;;\\n    /**\\n * The amino acid one letter codes as char array.\\n */\\nprivate static final char[] AMINO_ACID_CHARS = { \\'A\\', \\'B\\', \\'C\\', \\'D\\', \\'E\\', \\'F\\', \\'G\\', \\'H\\', \\'I\\', \\'J\\', \\'K\\', \\'L\\', \\'M\\', \\'N\\', \\'O\\', \\'P\\', \\'Q\\', \\'R\\', \\'S\\', \\'T\\', \\'U\\', \\'V\\', \\'W\\', \\'X\\', \\'Y\\', \\'Z\\' };;\\n    /**\\n * A char array of the one letter code of amino acids without combinations\\n * of amino acids.\\n */\\nprivate static final char[] UNIQUE_AMINO_ACID_CHARS = { \\'A\\', \\'C\\', \\'D\\', \\'E\\', \\'F\\', \\'G\\', \\'H\\', \\'I\\', \\'K\\', \\'L\\', \\'M\\', \\'N\\', \\'O\\', \\'P\\', \\'Q\\', \\'R\\', \\'S\\', \\'T\\', \\'U\\', \\'V\\', \\'W\\', \\'Y\\' };;\\n    /**\\n * The amino acid one letter codes as string array.\\n */\\npublic static final String[] AMINO_ACID_STRINGS = new String[] { \"A\", \"B\", \"C\", \"D\", \"E\", \"F\", \"G\", \"H\", \"I\", \"J\", \"K\", \"L\", \"M\", \"N\", \"O\", \"P\", \"Q\", \"R\", \"S\", \"T\", \"U\", \"V\", \"W\", \"X\", \"Y\", \"Z\" };;\\n    public static char[] getAminoAcids()\\n    public char getSingleLetterCodeAsChar()\\n    public static List<String> getAminoAcidsList()\\n    public static char[] getUniqueAminoAcids()\\n    public static AminoAcid getAminoAcid(String aa) throws IllegalArgumentException\\n    public static AminoAcid getAminoAcid(char aa) throws IllegalArgumentException\\n    public abstract boolean iscombination()\\n    public char[] getSubAminoAcids()\\n    public char[] getSubAminoAcids(boolean includeCombinations)\\n    public char[] getCombinations()\\n    public static char getMatchingAminoAcid(char aa, SequenceMatchingParameters sequenceMatchingPreferences)\\n    public static String getMatchingSequence(String sequence, SequenceMatchingParameters sequenceMatchingPreferences)\\n    public String[] getStandardGeneticCode()\\n    public static AminoAcid getAminoAcidFromGeneticCode(String geneticCode)\\n    protected String[] getStandardGeneticCodeForCombination()\\n    public AtomChain getMonoisotopicAtomChain()\\n    public double getMonoisotopicMass()\\n    public abstract double getHydrophobicity()\\n    public abstract double getHelicity()\\n    public abstract double getBasicity()\\n    public abstract double getPI()\\n    public abstract double getPK1()\\n    public abstract double getPK2()\\n    public abstract double getPKa()\\n    public abstract int getVanDerWaalsVolume()\\n    public double getProperty(Property property)\\n    public static int getNUnique()\\n    private static final int[] AA_INDEXES = { 0, -1, 1, 2, 3, 4, 5, 6, 7, -1, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, -1, 21, -1 };;\\n    public static int getUniqueIndex(char aa)\\n    public static int getIndex(char aa)\\n    public static boolean isAa(char aa)\\n    public static boolean isAa(int aa)\\n    public static boolean isUniqueAa(char aa)\\n    public boolean equals(Object obj)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/aminoacids/AminoAcid.java', 'methodCount': 34, 'lineCount': 74}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class Ion {\\n    public Ion();\\n    /**\\n * Serial number for backward compatibility.\\n */\\nstatic final long serialVersionUID = -1505719074403886934L;;\\n    /**\\n * Cache for the neutral losses as string.\\n */\\nprivate String neutralLossesAsString = null;;\\n    /**\\n * Type of ion.\\n */\\nprotected IonType type = IonType.UNKNOWN;;\\n    /**\\n * The theoretic mass.\\n *\\n * @deprecated use the double value instead.\\n */\\nprotected Double theoreticMass;;\\n    /**\\n * The theoretic mass.\\n */\\nprotected double theoreticMass1;;\\n    /**\\n * The atomic composition of the ion.\\n */\\nprotected AtomChain atomChain;;\\n    public abstract String getName()\\n    public abstract CvTerm getPrideCvTerm()\\n    public abstract CvTerm getPsiMsCvTerm()\\n    public abstract int getSubType()\\n    public abstract String getSubTypeAsString()\\n    public static int[] getPossibleSubtypes(IonType ionType)\\n    public static HashSet<Integer> getPossibleSubtypesAsSet(IonType ionType)\\n    public abstract NeutralLoss[] getNeutralLosses()\\n    public boolean hasNeutralLosses()\\n    public abstract boolean isSameAs(Ion anotherIon)\\n    public String getNeutralLossesAsString()\\n    public static String getNeutralLossesAsString(NeutralLoss[] neutralLosses)\\n    public double getTheoreticMass()\\n    public double getTheoreticMz(Integer charge)\\n    public AtomChain getAtomicComposition()\\n    public void setAtomicComposition(AtomChain atomChain)\\n    public void setTheoreticMass(double theoreticMass)\\n    public IonType getType()\\n    public static ArrayList<IonType> getImplementedIonTypes()\\n    public String getTypeAsString()\\n    public static String getTypeAsString(IonType type)\\n    public static Ion getGenericIon(IonType ionType, int subType, NeutralLoss[] neutralLosses)\\n    public static Ion getGenericIon(IonType ionType, int subType)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/ions/Ion.java', 'methodCount': 24, 'lineCount': 32}, {'fieldCount': 14, 'classCount': 1, 'methodsOnly': 'public class NeutralLoss {\\n    /**\\n * Serial number for backward compatibility.\\n */\\nstatic final long serialVersionUID = 5540846193082177391L;;\\n    /**\\n * Map of available neutral losses.\\n */\\nprivate final static HashMap<String, NeutralLoss> neutralLosses = new HashMap<>(0);;\\n    /**\\n * H2O loss.\\n */\\npublic static final NeutralLoss H2O = new NeutralLoss(\"H2O\", AtomChain.getAtomChain(\"H(2)O\"), false, new char[] { \\'D\\', \\'E\\', \\'S\\', \\'T\\' });;\\n    /**\\n * NH3 loss.\\n */\\npublic static final NeutralLoss NH3 = new NeutralLoss(\"NH3\", AtomChain.getAtomChain(\"NH(3)\"), false, new char[] { \\'K\\', \\'N\\', \\'Q\\', \\'R\\' });;\\n    /**\\n * H3PO4 loss.\\n */\\npublic static final NeutralLoss H3PO4 = new NeutralLoss(\"H3PO4\", AtomChain.getAtomChain(\"H(3)PO(4)\"), false);;\\n    /**\\n * H3PO3 loss.\\n */\\npublic static final NeutralLoss HPO3 = new NeutralLoss(\"HPO3\", AtomChain.getAtomChain(\"HPO(3)\"), false);;\\n    /**\\n * CH4OS loss.\\n */\\npublic static final NeutralLoss CH4OS = new NeutralLoss(\"CH4OS\", AtomChain.getAtomChain(\"CH(4)OS\"), false);;\\n    /**\\n * C3H9N loss.\\n */\\npublic static final NeutralLoss C3H9N = new NeutralLoss(\"C3H9N\", AtomChain.getAtomChain(\"C(3)H(9)N\"), false);;\\n    /**\\n * HCNO loss.\\n */\\npublic static final NeutralLoss HCNO = new NeutralLoss(\"HCNO\", AtomChain.getAtomChain(\"HCNO\"), false);;\\n    /**\\n * The composition of the ion.\\n */\\nprivate AtomChain composition;;\\n    /**\\n * The name of the neutral loss.\\n */\\npublic final String name;;\\n    /**\\n * Boolean indicating whether the neutral loss will always be accounted for.\\n */\\nprivate boolean fixed = false;;\\n    /**\\n * The PSI MS CV term of the neutral loss, null if not set.\\n */\\nprivate CvTerm psiCvTerm = null;;\\n    /**\\n * Amino acids that are likely to induce this loss. Null if not a loss originating from amino acids.\\n */\\npublic final char[] aminoAcids;;\\n    public NeutralLoss(String name, AtomChain composition, boolean fixed, char[] aminoAcids);\\n    public NeutralLoss();\\n    public NeutralLoss(String name, AtomChain composition, boolean fixed);\\n    public NeutralLoss(String name, AtomChain composition, boolean fixed, char[] aminoAcids, boolean save);\\n    public static void addNeutralLoss(NeutralLoss neutralLoss)\\n    public static NeutralLoss getNeutralLoss(String name)\\n    public static void removeNeutralLoss(String name)\\n    public CvTerm getPsiMsCvTerm()\\n    public boolean isFixed()\\n    public void setFixed(boolean fixed)\\n    public AtomChain getComposition()\\n    public void setComposition(AtomChain composition)\\n    public double getMass()\\n    public boolean isSameAs(NeutralLoss anotherNeutralLoss)\\n    public NeutralLoss clone()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/ions/NeutralLoss.java', 'methodCount': 15, 'lineCount': 31}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/utils/StandardMasses.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class CvTerm {\\n    public CvTerm();\\n    /**\\n * The ontology.\\n */\\nprivate String ontology;;\\n    /**\\n * The accession number.\\n */\\nprivate String accession;;\\n    /**\\n * The name/term.\\n */\\nprivate String name;;\\n    /**\\n * The value for the given term.\\n */\\nprivate String value;;\\n    public CvTerm(String ontology, String accession, String name, String value);\\n    public String getOntology()\\n    public void setOntology(String ontology)\\n    public String getAccession()\\n    public void setAccession(String accession)\\n    public String getName()\\n    public void setName(String name)\\n    public String getValue()\\n    public void setValue(String value)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/pride/CvTerm.java', 'methodCount': 10, 'lineCount': 16}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Returns the possible subtypes.\n",
      "     *\n",
      "     * @return the possible subtypes\n",
      "     */\n",
      "public static int[] getPossibleSubtypes()\n",
      "**************************************\n",
      "public static int[] getPossibleSubtypes() {\n",
      "    int[] possibleTypes = new int[21];\n",
      "    for (int i = 0; i <= 20; i++) {\n",
      "        possibleTypes[i] = i;\n",
      "    }\n",
      "    return possibleTypes;\n",
      "}\n",
      "3380\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 509/1243 [17:19<23:27,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 510/1243 [17:22<24:29,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24036\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 511/1243 [17:24<24:56,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24438\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|      | 512/1243 [17:26<25:43,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24511\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|     | 513/1243 [17:28<26:10,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24575\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public interface WaitingHandler {\\n    /**\\n * Convenience date format.\\n */\\npublic static final SimpleDateFormat SIMPLE_DATA_FORMAT = new SimpleDateFormat(\"dd MMM yyyy, HH:mm\");;\\n    /**\\n * The tab space to add when using HTML.\\n */\\npublic static final String TAB_HTML = \"&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\";;\\n    /**\\n * The tab space to add when not using HTML.\\n */\\npublic static final String TAB_NON_HTML = \"        \";;\\n    public void setPrimaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setMaxPrimaryProgressCounter(int maxProgressValue)\\n    public void increasePrimaryProgressCounter()\\n    public void increasePrimaryProgressCounter(int increment)\\n    public void setPrimaryProgressCounter(int value)\\n    public void setMaxSecondaryProgressCounter(int maxProgressValue)\\n    public void resetPrimaryProgressCounter()\\n    public void resetSecondaryProgressCounter()\\n    public void increaseSecondaryProgressCounter()\\n    public void setSecondaryProgressCounter(int value)\\n    public void increaseSecondaryProgressCounter(int value)\\n    public void setSecondaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setRunFinished()\\n    public void setRunCanceled()\\n    public void appendReport(String report, boolean includeDate, boolean addNewLine)\\n    public void appendReportNewLineNoDate()\\n    public void appendReportEndLine()\\n    public boolean isRunCanceled()\\n    public boolean isRunFinished()\\n    public void setSecondaryProgressText(String text)\\n    public boolean isReport()\\n    public void setWaitingText(String text)\\n    public int getPrimaryProgressCounter()\\n    public int getMaxPrimaryProgressCounter()\\n    public int getSecondaryProgressCounter()\\n    public int getMaxSecondaryProgressCounter()\\n    public void setDisplayProgress(boolean displayProgress)\\n    public boolean getDisplayProgress()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/waiting/WaitingHandler.java', 'methodCount': 28, 'lineCount': 33}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * Returns the number of occurrences of a given character until position\n",
      "     * index.\n",
      "     *\n",
      "     * @param index the index\n",
      "     * @param character the character\n",
      "     * @return the rank\n",
      "     */\n",
      "public int getRankRecursive(int index, int character)\n",
      "**************************************\n",
      "public int getRankRecursive(int index, int character) {\n",
      "    if (index >= 0) {\n",
      "        int cell = character >>> BIT_SHIFT;\n",
      "        int pos = character & BIT_MASK;\n",
      "        boolean left = ((alphabetDirections[cell] >>> pos) & 1) == 1;\n",
      "        int result = rank.getRank(index, left);\n",
      "        if (left && leftChild != null) {\n",
      "            return leftChild.getRankRecursive(result - 1, character);\n",
      "        } else if (!left && rightChild != null) {\n",
      "            return rightChild.getRankRecursive(result - 1, character);\n",
      "        }\n",
      "        return result;\n",
      "    }\n",
      "    return 0;\n",
      "}\n",
      "701\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|     | 514/1243 [17:30<25:08,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24583\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  41%|     | 515/1243 [17:32<25:35,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 516/1243 [17:35<25:58,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24616\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 517/1243 [17:37<25:50,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24705\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 518/1243 [17:39<25:54,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24707\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ExperimentObject {\\n    /**\\n * Unique identifier.\\n */\\nprivate long id;;\\n    public ExperimentObject();\\n    public long getId()\\n    public void setId(long id)\\n    private final static long[] HASHVALUELIST = { -4785187655948605440L, 2351437082613919744L, -7620983146772158464L, -8747902906952306688L, -4496485731609829376L, -4047692981156677632L, 5720028397227347968L, -4748060763626276864L, 2063365725770047488L, -5756543758428897280L, -134698081630425088L, 867726525032218624L, -526450428666544128L, 4146020926348189696L, 4362296343029680128L, -672990070253072384L, -2559490283472277504L, 3187632952876974080L, -5716989432807307264L, -8332013824838645760L, 4253744629365506048L, 2097316067254513664L, 8951627463544416256L, -5600031980443258880L, 6380991404691560448L, 8903284868402118656L, -1115601857539225600L, 4631654322507227136L, 7771989044436795392L, 7773688932940122112L, -6019734220953055232L, 3392712990065328128L, -8921384047543447552L, -7767842613008707584L, -1186522791511611392L, -6926112736333537280L, 8736653739320072192L, 8867062073843642368L, 6298992568012455936L, -6831107491093487616L, -7084666134747267072L, -1674183307215181824L, 7180054879733344256L, -1774408477150697472L, -1102347028329271296L, 2837232313405440000L, 6789844965029836800L, -2021979153929187328L, -803643088872329216L, -6635474898121947136L, -1519775710292529152L, -7017431056149018624L, 8399941098113230848L, 6620078501932513280L, 8402686423795523584L, 7887825026517880832L, 6240511391300272128L, -2116326148598433792L, 3164957088731514880L, 6354445331039899648L, -2421944411545827328L, -6588274517877174272L, -5482092713179058176L, 1515440486213902336L, -3383185261582667776L, -2725557693718710272L, 2180993613288613888L, -4878984385226620928L, 4905597879284899840L, -8937278576235966464L, -4857623260077275136L, -6678664042745585664L, 6590419491356596224L, 3898378085667969024L, -8773012746479065088L, -4316629602317574144L, -578020323565103104L, 5815789437630859264L, 1330829571824838656L, 2058704620696928256L, 5775301559630338048L, -4128281782811285504L, -6189976155577464832L, -2204893487149668352L, -4107985148748068864L, -2803177563490273280L, 7139083951461890048L, -6547891565468342272L, 3512976861638146048L, 8446989268574042112L, -6262309160844883968L, -447362214463838208L, -4695191602764636160L, -8777129286526107648L, -2322220230279856128L, -3376371221541236736L, -352816524822126592L, -6489602716775188480L, -4340386299073419264L, -411238008103813120L, -7548606038504292352L, 3950672770391547904L, 1570846774247147520L, 2087897268844892160L, -6691005528687374336L, 1651506531346769920L, -9105826395118237696L, -920643688498837504L, 6741095098680469504L, -9196666188088858624L, 4628592761082419200L, 1805632260469598208L, -2595685425333377024L, -2001876750192766976L, 4498796613205751808L, -3322677024598908928L, 8658129466298726400L, 2854559136171276288L, 106897466552897536L, 5590481524594866176L, -4319460978758043648L, 1896969526688425984L, -2860223852340688896L, -2273634011107659776L, -6830438329227218944L, -1024496407927033856L, -1561168395559655424L, -1430574524350681088L };;\\n    public static final long NO_KEY = asLong(\"#!#_NO_KEY_#!#\");;\\n    private HashMap<Long, UrParameter> urParams = null;;\\n    public void addUrParam(UrParameter parameter)\\n    public void removeUrParam(long paramterKey)\\n    private synchronized void createParamsMap()\\n    public UrParameter getUrParam(UrParameter parameter)\\n    public void clearParametersMap()\\n    public void setUrParams(HashMap<Long, UrParameter> urParams)\\n    public HashMap<Long, UrParameter> getUrParams()\\n    public static long asLong(String key)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/personalization/ExperimentObject.java', 'methodCount': 11, 'lineCount': 17}, {'fieldCount': 14, 'classCount': 1, 'methodsOnly': 'public class NeutralLoss {\\n    /**\\n * Serial number for backward compatibility.\\n */\\nstatic final long serialVersionUID = 5540846193082177391L;;\\n    /**\\n * Map of available neutral losses.\\n */\\nprivate final static HashMap<String, NeutralLoss> neutralLosses = new HashMap<>(0);;\\n    /**\\n * H2O loss.\\n */\\npublic static final NeutralLoss H2O = new NeutralLoss(\"H2O\", AtomChain.getAtomChain(\"H(2)O\"), false, new char[] { \\'D\\', \\'E\\', \\'S\\', \\'T\\' });;\\n    /**\\n * NH3 loss.\\n */\\npublic static final NeutralLoss NH3 = new NeutralLoss(\"NH3\", AtomChain.getAtomChain(\"NH(3)\"), false, new char[] { \\'K\\', \\'N\\', \\'Q\\', \\'R\\' });;\\n    /**\\n * H3PO4 loss.\\n */\\npublic static final NeutralLoss H3PO4 = new NeutralLoss(\"H3PO4\", AtomChain.getAtomChain(\"H(3)PO(4)\"), false);;\\n    /**\\n * H3PO3 loss.\\n */\\npublic static final NeutralLoss HPO3 = new NeutralLoss(\"HPO3\", AtomChain.getAtomChain(\"HPO(3)\"), false);;\\n    /**\\n * CH4OS loss.\\n */\\npublic static final NeutralLoss CH4OS = new NeutralLoss(\"CH4OS\", AtomChain.getAtomChain(\"CH(4)OS\"), false);;\\n    /**\\n * C3H9N loss.\\n */\\npublic static final NeutralLoss C3H9N = new NeutralLoss(\"C3H9N\", AtomChain.getAtomChain(\"C(3)H(9)N\"), false);;\\n    /**\\n * HCNO loss.\\n */\\npublic static final NeutralLoss HCNO = new NeutralLoss(\"HCNO\", AtomChain.getAtomChain(\"HCNO\"), false);;\\n    /**\\n * The composition of the ion.\\n */\\nprivate AtomChain composition;;\\n    /**\\n * The name of the neutral loss.\\n */\\npublic final String name;;\\n    /**\\n * Boolean indicating whether the neutral loss will always be accounted for.\\n */\\nprivate boolean fixed = false;;\\n    /**\\n * The PSI MS CV term of the neutral loss, null if not set.\\n */\\nprivate CvTerm psiCvTerm = null;;\\n    /**\\n * Amino acids that are likely to induce this loss. Null if not a loss originating from amino acids.\\n */\\npublic final char[] aminoAcids;;\\n    public NeutralLoss(String name, AtomChain composition, boolean fixed, char[] aminoAcids);\\n    public NeutralLoss();\\n    public NeutralLoss(String name, AtomChain composition, boolean fixed);\\n    public NeutralLoss(String name, AtomChain composition, boolean fixed, char[] aminoAcids, boolean save);\\n    public static void addNeutralLoss(NeutralLoss neutralLoss)\\n    public static NeutralLoss getNeutralLoss(String name)\\n    public static void removeNeutralLoss(String name)\\n    public CvTerm getPsiMsCvTerm()\\n    public boolean isFixed()\\n    public void setFixed(boolean fixed)\\n    public AtomChain getComposition()\\n    public void setComposition(AtomChain composition)\\n    public double getMass()\\n    public boolean isSameAs(NeutralLoss anotherNeutralLoss)\\n    public NeutralLoss clone()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/ions/NeutralLoss.java', 'methodCount': 15, 'lineCount': 31}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Returns an arraylist of the names of the implemented neutral losses.\n",
      "     *\n",
      "     * @return an arraylist of the names of the implemented neutral losses\n",
      "     */\n",
      "public ArrayList<String> getAccountedNeutralLosses()\n",
      "**************************************\n",
      "public ArrayList<String> getAccountedNeutralLosses() {\n",
      "    if (accountedNeutralLosses == null) {\n",
      "        accountedNeutralLosses = new ArrayList<>(forwardBoundaries.keySet());\n",
      "    }\n",
      "    return accountedNeutralLosses;\n",
      "}\n",
      "2503\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 519/1243 [17:41<26:13,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24709\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 520/1243 [17:43<25:45,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24846\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 521/1243 [17:46<29:10,  2.42s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24848\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 522/1243 [17:48<28:39,  2.39s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25027\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 523/1243 [17:50<27:09,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25091\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ExperimentObject {\\n    /**\\n * Unique identifier.\\n */\\nprivate long id;;\\n    public ExperimentObject();\\n    public long getId()\\n    public void setId(long id)\\n    private final static long[] HASHVALUELIST = { -4785187655948605440L, 2351437082613919744L, -7620983146772158464L, -8747902906952306688L, -4496485731609829376L, -4047692981156677632L, 5720028397227347968L, -4748060763626276864L, 2063365725770047488L, -5756543758428897280L, -134698081630425088L, 867726525032218624L, -526450428666544128L, 4146020926348189696L, 4362296343029680128L, -672990070253072384L, -2559490283472277504L, 3187632952876974080L, -5716989432807307264L, -8332013824838645760L, 4253744629365506048L, 2097316067254513664L, 8951627463544416256L, -5600031980443258880L, 6380991404691560448L, 8903284868402118656L, -1115601857539225600L, 4631654322507227136L, 7771989044436795392L, 7773688932940122112L, -6019734220953055232L, 3392712990065328128L, -8921384047543447552L, -7767842613008707584L, -1186522791511611392L, -6926112736333537280L, 8736653739320072192L, 8867062073843642368L, 6298992568012455936L, -6831107491093487616L, -7084666134747267072L, -1674183307215181824L, 7180054879733344256L, -1774408477150697472L, -1102347028329271296L, 2837232313405440000L, 6789844965029836800L, -2021979153929187328L, -803643088872329216L, -6635474898121947136L, -1519775710292529152L, -7017431056149018624L, 8399941098113230848L, 6620078501932513280L, 8402686423795523584L, 7887825026517880832L, 6240511391300272128L, -2116326148598433792L, 3164957088731514880L, 6354445331039899648L, -2421944411545827328L, -6588274517877174272L, -5482092713179058176L, 1515440486213902336L, -3383185261582667776L, -2725557693718710272L, 2180993613288613888L, -4878984385226620928L, 4905597879284899840L, -8937278576235966464L, -4857623260077275136L, -6678664042745585664L, 6590419491356596224L, 3898378085667969024L, -8773012746479065088L, -4316629602317574144L, -578020323565103104L, 5815789437630859264L, 1330829571824838656L, 2058704620696928256L, 5775301559630338048L, -4128281782811285504L, -6189976155577464832L, -2204893487149668352L, -4107985148748068864L, -2803177563490273280L, 7139083951461890048L, -6547891565468342272L, 3512976861638146048L, 8446989268574042112L, -6262309160844883968L, -447362214463838208L, -4695191602764636160L, -8777129286526107648L, -2322220230279856128L, -3376371221541236736L, -352816524822126592L, -6489602716775188480L, -4340386299073419264L, -411238008103813120L, -7548606038504292352L, 3950672770391547904L, 1570846774247147520L, 2087897268844892160L, -6691005528687374336L, 1651506531346769920L, -9105826395118237696L, -920643688498837504L, 6741095098680469504L, -9196666188088858624L, 4628592761082419200L, 1805632260469598208L, -2595685425333377024L, -2001876750192766976L, 4498796613205751808L, -3322677024598908928L, 8658129466298726400L, 2854559136171276288L, 106897466552897536L, 5590481524594866176L, -4319460978758043648L, 1896969526688425984L, -2860223852340688896L, -2273634011107659776L, -6830438329227218944L, -1024496407927033856L, -1561168395559655424L, -1430574524350681088L };;\\n    public static final long NO_KEY = asLong(\"#!#_NO_KEY_#!#\");;\\n    private HashMap<Long, UrParameter> urParams = null;;\\n    public void addUrParam(UrParameter parameter)\\n    public void removeUrParam(long paramterKey)\\n    private synchronized void createParamsMap()\\n    public UrParameter getUrParam(UrParameter parameter)\\n    public void clearParametersMap()\\n    public void setUrParams(HashMap<Long, UrParameter> urParams)\\n    public HashMap<Long, UrParameter> getUrParams()\\n    public static long asLong(String key)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/personalization/ExperimentObject.java', 'methodCount': 11, 'lineCount': 17}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Returns the delta score at a given site. First amino acid is 1.\n",
      "     *\n",
      "     * @param site the site of interest\n",
      "     * @return the attached delta score. 0 if not found.\n",
      "     */\n",
      "public double getDeltaScore(int site)\n",
      "**************************************\n",
      "public double getDeltaScore(int site) {\n",
      "    if (deltaScoresAtAA == null) {\n",
      "        return 0.0;\n",
      "    }\n",
      "    Double score = deltaScoresAtAA.get(site);\n",
      "    if (score == null) {\n",
      "        return 0.0;\n",
      "    } else {\n",
      "        return score;\n",
      "    }\n",
      "}\n",
      "1686\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 524/1243 [17:53<26:45,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25093\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 525/1243 [17:54<24:49,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25103\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 526/1243 [17:56<23:15,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25108\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 527/1243 [17:58<21:47,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25134\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  42%|     | 528/1243 [18:00<23:10,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25161\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Peptide {\\n    /**\\n * The peptide sequence.\\n */\\nprivate String sequence;;\\n    /**\\n * The peptide key.\\n */\\nprivate long key = NO_KEY;;\\n    /**\\n * Boolean indicating whether the matching key is set.\\n */\\nprivate boolean keySet = false;;\\n    /**\\n * The peptide matching key.\\n */\\nprivate long matchingKey;;\\n    /**\\n * The peptide mass.\\n */\\nprivate double mass = -1.0;;\\n    /**\\n * The mapping of this peptide on proteins as a map, accession to position.\\n * Position on protein sequences is 0 based.\\n */\\nprivate TreeMap<String, int[]> proteinMapping = null;;\\n    /**\\n * The variable modifications carried by the peptide.\\n */\\nprivate ModificationMatch[] variableModifications = null;;\\n    /**\\n * Convenience array for no modifications.\\n */\\nprivate static final ModificationMatch[] NO_MOD = new ModificationMatch[0];;\\n    /**\\n * The variants observed when mapping this peptide to the database. Peptide\\n * variant matches are indexed by protein and by peptide start.\\n */\\nprivate HashMap<String, HashMap<Integer, PeptideVariantMatches>> variantMatches = null;;\\n    /**\\n * Separator preceding confident localization of the confident localization\\n * of a modification.\\n */\\npublic final static String MODIFICATION_LOCALIZATION_SEPARATOR = \"-ATAA-\";;\\n    /**\\n * Separator used to separate modifications in peptide keys as string.\\n */\\npublic final static String MODIFICATION_SEPARATOR = \"_\";;\\n    public Peptide();\\n    public Peptide(String aSequence, ModificationMatch[] variableModifications, HashMap<String, HashMap<Integer, PeptideVariantMatches>> variantMatches, boolean sanityCheck, double mass);\\n    public Peptide(String aSequence, ModificationMatch[] variableModifications, boolean sanityCheck, double mass);\\n    public Peptide(String aSequence, ModificationMatch[] variableModifications, boolean sanityCheck);\\n    public Peptide(String aSequence);\\n    public Peptide(String aSequence, ModificationMatch[] variableModifications);\\n    public Peptide(String aSequence, ModificationMatch[] variableModifications, HashMap<String, HashMap<Integer, PeptideVariantMatches>> variantMatches, boolean sanityCheck);\\n    private void sanityCheck()\\n    public void setMass(double mass)\\n    public void setKey(long key)\\n    public TreeMap<String, int[]> getProteinMapping()\\n    public void setProteinMapping(TreeMap<String, int[]> proteinMapping)\\n    public void setVariantMatches(HashMap<String, HashMap<Integer, PeptideVariantMatches>> variantMatches)\\n    public HashMap<String, HashMap<Integer, PeptideVariantMatches>> getVariantMatches()\\n    public double getMass()\\n    public double getMass(ModificationParameters modificationParameters, SequenceProvider sequenceProvider, SequenceMatchingParameters modificationSequenceMatchingParameters)\\n    public ModificationMatch[] getVariableModifications()\\n    public String[] getIndexedVariableModifications()\\n    public String[] getFixedModifications(ModificationParameters modificationParameters, SequenceProvider sequenceProvider, SequenceMatchingParameters modificationsSequenceMatchingParameters)\\n    public void setVariableModifications(ModificationMatch[] variableModifications)\\n    public void clearVariableModifications()\\n    public void addVariableModification(ModificationMatch modificationMatch)\\n    public void clearVariantMatches()\\n    public int getPeptideEnd(String proteinAccession, int peptideStart)\\n    public String getSequence()\\n    public void setSequence(String sequence)\\n    public int getNMissedCleavages(Enzyme enzyme)\\n    public int getNMissedCleavages(DigestionParameters digestionPreferences)\\n    public long getMatchingKey()\\n    public void setMatchingKey(long matchingKey)\\n    public long getMatchingKey(SequenceMatchingParameters sequenceMatchingPreferences)\\n    public long getKey()\\n    public static long getKey(String sequence, ModificationMatch[] variableModifications)\\n    public int getNVariableModifications(double modificationMass)\\n    public int getNVariableModifications()\\n    public ArrayList<Integer> getPotentialModificationSitesNoCombination(Modification modification, String proteinSequence, int peptideStart)\\n    public boolean isSameSequenceAndModificationStatus(Peptide anotherPeptide, SequenceMatchingParameters sequenceMatchingPreferences)\\n    public boolean isSameSequence(Peptide anotherPeptide, SequenceMatchingParameters sequenceMatchingPreferences)\\n    public boolean isSameModificationStatus(Peptide anotherPeptide)\\n    public boolean sameModificationsAs(Peptide anotherPeptide, ArrayList<String> modifications)\\n    public boolean sameModificationsAs(Peptide anotherPeptide)\\n    public String getTaggedModifiedSequence(ModificationParameters modificationProfile, SequenceProvider sequenceProvider, SequenceMatchingParameters modificationsSequenceMatchingParameters, boolean useHtmlColorCoding, boolean includeHtmlStartEndTags, boolean useShortName, HashSet<String> displayedModifications)\\n    public void estimateTheoreticMass(ModificationParameters modificationParameters, SequenceProvider sequenceProvider, SequenceMatchingParameters modificationSequenceMatchingParameters)\\n    public Peptide getNoModPeptide(HashSet<String> forbiddenModifications)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/proteins/Peptide.java', 'methodCount': 44, 'lineCount': 57}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class IdentificationMatch {\\n    public IdentificationMatch();\\n    /**\\n * Serial number for backward compatibility.\\n */\\nstatic final long serialVersionUID = -9132138792119651421L;;\\n    public abstract MatchType getType()\\n    public abstract long getKey()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/identification/IdentificationMatch.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ExperimentObject {\\n    /**\\n * Unique identifier.\\n */\\nprivate long id;;\\n    public ExperimentObject();\\n    public long getId()\\n    public void setId(long id)\\n    private final static long[] HASHVALUELIST = { -4785187655948605440L, 2351437082613919744L, -7620983146772158464L, -8747902906952306688L, -4496485731609829376L, -4047692981156677632L, 5720028397227347968L, -4748060763626276864L, 2063365725770047488L, -5756543758428897280L, -134698081630425088L, 867726525032218624L, -526450428666544128L, 4146020926348189696L, 4362296343029680128L, -672990070253072384L, -2559490283472277504L, 3187632952876974080L, -5716989432807307264L, -8332013824838645760L, 4253744629365506048L, 2097316067254513664L, 8951627463544416256L, -5600031980443258880L, 6380991404691560448L, 8903284868402118656L, -1115601857539225600L, 4631654322507227136L, 7771989044436795392L, 7773688932940122112L, -6019734220953055232L, 3392712990065328128L, -8921384047543447552L, -7767842613008707584L, -1186522791511611392L, -6926112736333537280L, 8736653739320072192L, 8867062073843642368L, 6298992568012455936L, -6831107491093487616L, -7084666134747267072L, -1674183307215181824L, 7180054879733344256L, -1774408477150697472L, -1102347028329271296L, 2837232313405440000L, 6789844965029836800L, -2021979153929187328L, -803643088872329216L, -6635474898121947136L, -1519775710292529152L, -7017431056149018624L, 8399941098113230848L, 6620078501932513280L, 8402686423795523584L, 7887825026517880832L, 6240511391300272128L, -2116326148598433792L, 3164957088731514880L, 6354445331039899648L, -2421944411545827328L, -6588274517877174272L, -5482092713179058176L, 1515440486213902336L, -3383185261582667776L, -2725557693718710272L, 2180993613288613888L, -4878984385226620928L, 4905597879284899840L, -8937278576235966464L, -4857623260077275136L, -6678664042745585664L, 6590419491356596224L, 3898378085667969024L, -8773012746479065088L, -4316629602317574144L, -578020323565103104L, 5815789437630859264L, 1330829571824838656L, 2058704620696928256L, 5775301559630338048L, -4128281782811285504L, -6189976155577464832L, -2204893487149668352L, -4107985148748068864L, -2803177563490273280L, 7139083951461890048L, -6547891565468342272L, 3512976861638146048L, 8446989268574042112L, -6262309160844883968L, -447362214463838208L, -4695191602764636160L, -8777129286526107648L, -2322220230279856128L, -3376371221541236736L, -352816524822126592L, -6489602716775188480L, -4340386299073419264L, -411238008103813120L, -7548606038504292352L, 3950672770391547904L, 1570846774247147520L, 2087897268844892160L, -6691005528687374336L, 1651506531346769920L, -9105826395118237696L, -920643688498837504L, 6741095098680469504L, -9196666188088858624L, 4628592761082419200L, 1805632260469598208L, -2595685425333377024L, -2001876750192766976L, 4498796613205751808L, -3322677024598908928L, 8658129466298726400L, 2854559136171276288L, 106897466552897536L, 5590481524594866176L, -4319460978758043648L, 1896969526688425984L, -2860223852340688896L, -2273634011107659776L, -6830438329227218944L, -1024496407927033856L, -1561168395559655424L, -1430574524350681088L };;\\n    public static final long NO_KEY = asLong(\"#!#_NO_KEY_#!#\");;\\n    private HashMap<Long, UrParameter> urParams = null;;\\n    public void addUrParam(UrParameter parameter)\\n    public void removeUrParam(long paramterKey)\\n    private synchronized void createParamsMap()\\n    public UrParameter getUrParam(UrParameter parameter)\\n    public void clearParametersMap()\\n    public void setUrParams(HashMap<Long, UrParameter> urParams)\\n    public HashMap<Long, UrParameter> getUrParams()\\n    public static long asLong(String key)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/personalization/ExperimentObject.java', 'methodCount': 11, 'lineCount': 17}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * Sets the matchKey field.\n",
      "     */\n",
      "private void setMatchKey()\n",
      "**************************************\n",
      "private void setMatchKey() {\n",
      "    matchKey = ExperimentObject.asLong(Arrays.stream(accessions).collect(Collectors.joining()));\n",
      "}\n",
      "2897\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 529/1243 [18:02<24:07,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25218\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 530/1243 [18:04<24:56,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25251\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 531/1243 [18:06<24:38,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25258\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 532/1243 [18:08<24:47,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25308\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 533/1243 [18:11<26:47,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25309\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Peak {\\n    public Peak();\\n    /**\\n * The mass over charge ratio of the peak.\\n */\\npublic final double mz;;\\n    /**\\n * The intensity of the peak.\\n */\\npublic final double intensity;;\\n    public Peak(double mz, double intensity);\\n    public boolean isSameAs(Peak aPeak)\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public double getMass(int chargeValue)\\n    public static final Comparator<Peak> AscendingIntensityComparator = new Comparator<Peak>() {\\n\\n    @Override\\n    public int compare(Peak o1, Peak o2) {\\n        return o1.intensity < o2.intensity ? -1 : o1.intensity == o2.intensity ? 0 : 1;\\n    }\\n};;\\n    public static final Comparator<Peak> DescendingIntensityComparator = new Comparator<Peak>() {\\n\\n    @Override\\n    public int compare(Peak o1, Peak o2) {\\n        return o1.intensity > o2.intensity ? -1 : o1.intensity == o2.intensity ? 0 : 1;\\n    }\\n};;\\n    public static final Comparator<Peak> AscendingMzComparator = new Comparator<Peak>() {\\n\\n    @Override\\n    public int compare(Peak o1, Peak o2) {\\n        return o1.mz < o2.mz ? -1 : o1.mz == o2.mz ? 0 : 1;\\n    }\\n};;\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/spectra/Peak.java', 'methodCount': 7, 'lineCount': 14}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class NonSymmetricalNormalDistribution {\\n    public NonSymmetricalNormalDistribution();\\n    /**\\n * Serial version number for backward compatibility.\\n */\\nstatic final long serialVersionUID = -5258258835569357886L;;\\n    /**\\n * The standard deviation on the right of the distribution.\\n */\\nprivate final double stdUp;;\\n    /**\\n * The standard deviation on the left of the distribution.\\n */\\nprivate final double stdDown;;\\n    /**\\n * The mean.\\n */\\nprivate final double mean;;\\n    /**\\n * Distribution equivalent to the right of the distribution.\\n */\\nprivate final NormalDistribution distributionUp;;\\n    /**\\n * Distribution equivalent to the left of the distribution.\\n */\\nprivate final NormalDistribution distributionDown;;\\n    public NonSymmetricalNormalDistribution(double mean, double stdDown, double stdUp);\\n    public double getStdUp()\\n    public double getStdDown()\\n    public double getMean()\\n    public static NonSymmetricalNormalDistribution getRobustNonSymmetricalNormalDistribution(ArrayList<Double> input)\\n    public static NonSymmetricalNormalDistribution getRobustNonSymmetricalNormalDistributionFromSortedList(ArrayList<Double> input)\\n    public static NonSymmetricalNormalDistribution getRobustNonSymmetricalNormalDistributionFromSortedArray(double[] input)\\n    public double getProbabilityAt(double x)\\n    public double getCumulativeProbabilityAt(double x)\\n    public double getDescendingCumulativeProbabilityAt(double x)\\n    public double getSmallestCumulativeProbabilityAt(double x)\\n    public double getMaxValueForProbability(double p)\\n    public double getMinValueForProbability(double p)\\n    public double getValueAtCumulativeProbability(double p)\\n    public double getValueAtDescendingCumulativeProbability(double p)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/math/statistics/distributions/NonSymmetricalNormalDistribution.java', 'methodCount': 16, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Get the binned logged cumulative probability.\n",
      "     * \n",
      "     * @param intensity the intensity\n",
      "     * \n",
      "     * @return the binned cumulative logged probability\n",
      "     */\n",
      "public double getBinnedCumulativeProbabilityLog(double intensity)\n",
      "**************************************\n",
      "public double getBinnedCumulativeProbabilityLog(double intensity) {\n",
      "    for (int i = 0; i < orderedBins.length; i++) {\n",
      "        int bin = orderedBins[i];\n",
      "        if (intensity > bin) {\n",
      "            return pLog[i];\n",
      "        }\n",
      "    }\n",
      "    return 0.0;\n",
      "}\n",
      "1038\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 534/1243 [18:13<26:19,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25373\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 535/1243 [18:16<27:36,  2.34s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25477\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 536/1243 [18:18<27:09,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25671\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 537/1243 [18:20<26:39,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25672\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 538/1243 [18:22<26:02,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25707\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class Charge {\\n    public Charge();\\n    public static String toString(int value)\\n    public static String getChargeAsFormattedString(int value)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/ions/Charge.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Precursor {\\n    /**\\n * The retention time when the precursor was isolated.\\n */\\npublic final double rt;;\\n    /**\\n * The measured m/z of the precursor.\\n */\\npublic final double mz;;\\n    /**\\n * The measured intensity of the precursor.\\n */\\npublic final double intensity;;\\n    /**\\n * The charge of the precursor.\\n */\\npublic final int[] possibleCharges;;\\n    public Precursor();\\n    public Precursor(double rt, double mz, int[] possibleCharges);\\n    public Precursor(double rt, double mz, double intensity, int[] possibleCharges);\\n    public Precursor(double mz, double intensity, int[] possibleCharges, double rtMin, double rtMax);\\n    public double getRtInMinutes()\\n    public String getPossibleChargesAsString()\\n    public double getMass(int chargeValue)\\n    public boolean isSameAs(Precursor otherPrecursor)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/spectra/Precursor.java', 'methodCount': 8, 'lineCount': 14}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Spectrum {\\n    /**\\n * The precursor if any.\\n */\\npublic Precursor precursor;;\\n    /**\\n * The array of the m/z of the peaks. Note: throughout the code, it is\\n * assumed that the m/z array is sorted by ascending m/z. Only minimal\\n * sanity check is conducted.\\n */\\npublic double[] mz;;\\n    /**\\n * The array of the intensities of the peaks.\\n */\\npublic double[] intensity;;\\n    /**\\n * The spectrum level, i.e. 1 for MS1, 2 for MS2, etc.\\n */\\nprivate Integer spectrumLevel = null;;\\n    public Spectrum();\\n    public Spectrum(Precursor precursor, double[] mz, double[] intensities, int spectrumLevel);\\n    public String getPeakListAsString()\\n    public double getTotalIntensity()\\n    public double getMaxIntensity()\\n    public double getMaxMz()\\n    public double getMinMz()\\n    public Precursor getPrecursor()\\n    public int getNPeaks()\\n    public boolean isSameAs(Spectrum otherSpectrum)\\n    public String toString()\\n    public int getSpectrumLevel()\\n    public void setSpectrumLevel(int spectrumLevel)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.java', 'methodCount': 13, 'lineCount': 19}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class SimpleFileWriter {\\n    /**\\n * The buffered writer used for writing.\\n */\\nprivate final BufferedWriter bw;;\\n    /**\\n * A simple mutex.\\n */\\nprivate final SimpleSemaphore mutex = new SimpleSemaphore(1);;\\n    public SimpleFileWriter(File file, boolean gz);\\n    public void writeLine(String... elements)\\n    public void writeLine(String line)\\n    public void write(char[] buffer, int offset, int length)\\n    public void newLine()\\n    public void write(String text)\\n    public void write(String text, boolean newLine)\\n    public void close()\\n    public BufferedWriter getWriter()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/flat/SimpleFileWriter.java', 'methodCount': 9, 'lineCount': 13}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     * Writes the given spectrum to the file.\n",
      "     *\n",
      "     * @param spectrumTitle The title of the spectrum to write.\n",
      "     * @param spectrum The spectrum to write.\n",
      "     */\n",
      "public void writeSpectrum(String spectrumTitle, Spectrum spectrum)\n",
      "**************************************\n",
      "/**\n",
      " * Writes the given spectrum to the file.\n",
      " *\n",
      " * @param spectrumTitle The title of the spectrum to write.\n",
      " * @param spectrum The spectrum to write.\n",
      " */\n",
      "public void writeSpectrum(String spectrumTitle, Spectrum spectrum) {\n",
      "    String toWrite = asMgf(spectrumTitle, spectrum);\n",
      "    writer.write(toWrite, true);\n",
      "}\n",
      "1109\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 539/1243 [18:24<25:43,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25743\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  43%|     | 540/1243 [18:26<23:55,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25746\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 541/1243 [18:28<22:19,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25752\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 542/1243 [18:30<23:13,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25757\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 543/1243 [18:32<23:58,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25758\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface SpectrumProvider {\\n    public Spectrum getSpectrum(String fileNameWithoutExtension, String spectrumTitle)\\n    public Precursor getPrecursor(String fileNameWithoutExtension, String spectrumTitle)\\n    public ArrayList<String> getPostcursorSpectrumTitles(String fileNameWithoutExtension, String spectrumTitle)\\n    public double getPrecursorMz(String fileNameWithoutExtension, String spectrumTitle)\\n    public double getPrecursorRt(String fileNameWithoutExtension, String spectrumTitle)\\n    public int getSpectrumLevel(String fileNameWithoutExtension, String spectrumTitle)\\n    public double[][] getPeaks(String fileNameWithoutExtension, String spectrumTitle)\\n    public double getMinPrecMz(String fileNameWithoutExtension)\\n    public double getMaxPrecMz(String fileNameWithoutExtension)\\n    public double getMaxPrecInt(String fileNameWithoutExtension)\\n    public double getMaxPrecRT(String fileNameWithoutExtension)\\n    public double getMinPrecMz()\\n    public double getMaxPrecMz()\\n    public double getMaxPrecInt()\\n    public double getMaxPrecRT()\\n    public String[] getOrderedFileNamesWithoutExtensions()\\n    public String[] getSpectrumTitles(String fileNameWithoutExtension)\\n    public HashMap<String, String> getFilePaths()\\n    public HashMap<String, String> getCmsFilePaths()\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/SpectrumProvider.java', 'methodCount': 20, 'lineCount': 22}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Precursor {\\n    /**\\n * The retention time when the precursor was isolated.\\n */\\npublic final double rt;;\\n    /**\\n * The measured m/z of the precursor.\\n */\\npublic final double mz;;\\n    /**\\n * The measured intensity of the precursor.\\n */\\npublic final double intensity;;\\n    /**\\n * The charge of the precursor.\\n */\\npublic final int[] possibleCharges;;\\n    public Precursor();\\n    public Precursor(double rt, double mz, int[] possibleCharges);\\n    public Precursor(double rt, double mz, double intensity, int[] possibleCharges);\\n    public Precursor(double mz, double intensity, int[] possibleCharges, double rtMin, double rtMax);\\n    public double getRtInMinutes()\\n    public String getPossibleChargesAsString()\\n    public double getMass(int chargeValue)\\n    public boolean isSameAs(Precursor otherPrecursor)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/spectra/Precursor.java', 'methodCount': 8, 'lineCount': 14}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Spectrum {\\n    /**\\n * The precursor if any.\\n */\\npublic Precursor precursor;;\\n    /**\\n * The array of the m/z of the peaks. Note: throughout the code, it is\\n * assumed that the m/z array is sorted by ascending m/z. Only minimal\\n * sanity check is conducted.\\n */\\npublic double[] mz;;\\n    /**\\n * The array of the intensities of the peaks.\\n */\\npublic double[] intensity;;\\n    /**\\n * The spectrum level, i.e. 1 for MS1, 2 for MS2, etc.\\n */\\nprivate Integer spectrumLevel = null;;\\n    public Spectrum();\\n    public Spectrum(Precursor precursor, double[] mz, double[] intensities, int spectrumLevel);\\n    public String getPeakListAsString()\\n    public double getTotalIntensity()\\n    public double getMaxIntensity()\\n    public double getMaxMz()\\n    public double getMinMz()\\n    public Precursor getPrecursor()\\n    public int getNPeaks()\\n    public boolean isSameAs(Spectrum otherSpectrum)\\n    public String toString()\\n    public int getSpectrumLevel()\\n    public void setSpectrumLevel(int spectrumLevel)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/mass_spectrometry/spectra/Spectrum.java', 'methodCount': 13, 'lineCount': 19}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class IoUtil {\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default separator for tabular files.\\n */\\npublic static final String DEFAULT_SEPARATOR = \"\\\\t\";;\\n    public static boolean emptyDir(File dir)\\n    public static boolean deleteDir(File dir)\\n    public static void copyFile(File in, File out) throws IOException\\n    public static void append(File in, File out) throws IOException\\n    public static void copyFile(File in, File out, boolean overwrite) throws IOException\\n    public static String getFileName(String filePath)\\n    public static String getFileName(File file)\\n    public static String getExtension(File file)\\n    public static String getExtension(String fileName)\\n    public static String getFilenameExtensionLowerCase(String fileName)\\n    public static File existsExtensionNotCaseSensitive(File file)\\n    public static String appendSuffix(String fileName, String suffix)\\n    public static String removeExtension(String fileName)\\n    public static File saveUrl(File saveFile, String targetUrlAsString, int fileSizeInBytes, String userName, String password, WaitingHandler waitingHandler) throws MalformedURLException, IOException, FileNotFoundException\\n    public static boolean checkIfURLExists(String targetUrlAsString, String userName, String password)\\n    public static int getFileSize(URL url)\\n    public static void closeBuffer(MappedByteBuffer buffer)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/IoUtil.java', 'methodCount': 17, 'lineCount': 21}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ZstdUtils {\\n    public static byte[] zstdDecompress(byte[] compressedByteArray, int uncompressedLength)\\n    public static byte[] zstdDecompress(ZstdDecompressor decompressor, byte[] compressedByteArray, int uncompressedLength)\\n    public static TempByteArray zstdCompress(byte[] uncompressedData)\\n    public static TempByteArray zstdCompress(ZstdCompressor compressor, byte[] uncompressedData)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/compression/ZstdUtils.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class SimpleSemaphore {\\n    /**\\n * The semaphore to use.\\n */\\nprivate final Semaphore mutex;;\\n    public SimpleSemaphore(int nPermits, boolean fair);\\n    public SimpleSemaphore(int nPermits);\\n    public void acquire()\\n    public void release()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/threading/SimpleSemaphore.java', 'methodCount': 4, 'lineCount': 7}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public interface WaitingHandler {\\n    /**\\n * Convenience date format.\\n */\\npublic static final SimpleDateFormat SIMPLE_DATA_FORMAT = new SimpleDateFormat(\"dd MMM yyyy, HH:mm\");;\\n    /**\\n * The tab space to add when using HTML.\\n */\\npublic static final String TAB_HTML = \"&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\";;\\n    /**\\n * The tab space to add when not using HTML.\\n */\\npublic static final String TAB_NON_HTML = \"        \";;\\n    public void setPrimaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setMaxPrimaryProgressCounter(int maxProgressValue)\\n    public void increasePrimaryProgressCounter()\\n    public void increasePrimaryProgressCounter(int increment)\\n    public void setPrimaryProgressCounter(int value)\\n    public void setMaxSecondaryProgressCounter(int maxProgressValue)\\n    public void resetPrimaryProgressCounter()\\n    public void resetSecondaryProgressCounter()\\n    public void increaseSecondaryProgressCounter()\\n    public void setSecondaryProgressCounter(int value)\\n    public void increaseSecondaryProgressCounter(int value)\\n    public void setSecondaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setRunFinished()\\n    public void setRunCanceled()\\n    public void appendReport(String report, boolean includeDate, boolean addNewLine)\\n    public void appendReportNewLineNoDate()\\n    public void appendReportEndLine()\\n    public boolean isRunCanceled()\\n    public boolean isRunFinished()\\n    public void setSecondaryProgressText(String text)\\n    public boolean isReport()\\n    public void setWaitingText(String text)\\n    public int getPrimaryProgressCounter()\\n    public int getMaxPrimaryProgressCounter()\\n    public int getSecondaryProgressCounter()\\n    public int getMaxSecondaryProgressCounter()\\n    public void setDisplayProgress(boolean displayProgress)\\n    public boolean getDisplayProgress()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/waiting/WaitingHandler.java', 'methodCount': 28, 'lineCount': 33}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 10}\n",
      "{}\n",
      "/**\n",
      "     * Returns the spectrum level of the spectrum with the given title.\n",
      "     *\n",
      "     * @param spectrumTitle the title of the spectrum\n",
      "     *\n",
      "     * @return the spectrum level of the spectrum\n",
      "     */\n",
      "public int getSpectrumLevel(String spectrumTitle)\n",
      "**************************************\n",
      "public int getSpectrumLevel(String spectrumTitle) {\n",
      "    int index = indexMap.get(spectrumTitle);\n",
      "    mutex.acquire();\n",
      "    mappedByteBuffer.position(index + 3 * Double.BYTES);\n",
      "    int spectrumLevel = mappedByteBuffer.getInt();\n",
      "    mutex.release();\n",
      "    return spectrumLevel;\n",
      "}\n",
      "2253\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 544/1243 [18:34<24:35,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25760\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 545/1243 [18:36<22:30,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25825\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 546/1243 [18:38<23:24,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25847\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 547/1243 [18:40<23:16,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25915\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 548/1243 [18:43<25:15,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25925\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PrideMarshaller {\\n    public PrideMarshaller();\\n    protected void init()\\n    public ProjectDetail getProjectDetail(String jsonURL) throws IOException\\n    public ProjectSummaryList getProjectSummaryList(String jsonURL) throws IOException\\n    public AssayDetailList getAssayDetailList(String jsonURL) throws IOException\\n    public FileDetailList getFileDetailList(String jsonURL) throws IOException\\n    public ProteinDetailList getProteinDetailList(String jsonURL) throws IOException\\n    public PsmDetailList getPeptideDetailList(String jsonURL) throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/json/marshallers/PrideMarshaller.java', 'methodCount': 8, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/pride/prideobjects/webservice/PrideQuery.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class PrideFilter {\\n    public PrideFilter();\\n    /**\\n * A list containing the project summaries.\\n */\\nprivate final String value;;\\n    /**\\n * A list containing the project summaries.\\n */\\nprivate final PrideFilterType type;;\\n    public PrideFilter(PrideFilterType type, String value);\\n    public String getValue()\\n    public PrideFilterType getType()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/pride/prideobjects/webservice/query/PrideFilter.java', 'methodCount': 4, 'lineCount': 8}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Returns a list of proteins, gotten from the paged query.\n",
      "     *\n",
      "     * @param queryURL the root query url\n",
      "     * @return a list of proteins, gotten from the paged query\n",
      "     * @throws IOException thrown if the webservice is down or the json response\n",
      "     * was invalid\n",
      "     */\n",
      "private static ProteinDetailList bufferProteinIdentifications(String queryURL) throws IOException\n",
      "**************************************\n",
      "private static ProteinDetailList bufferProteinIdentifications(String queryURL) throws IOException {\n",
      "    ProteinDetailList proteins = new ProteinDetailList();\n",
      "    boolean moreResults = true;\n",
      "    int page = 0;\n",
      "    while (moreResults) {\n",
      "        ProteinDetailList subList = (ProteinDetailList) MARSHALLER.getProteinDetailList(queryURL + \"?show=\" + PAGE_SIZE + \"&page=\" + page);\n",
      "        if (subList.getList().isEmpty()) {\n",
      "            moreResults = false;\n",
      "        } else {\n",
      "            proteins.getList().addAll(subList.getList());\n",
      "        }\n",
      "        page++;\n",
      "    }\n",
      "    return proteins;\n",
      "}\n",
      "669\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 549/1243 [18:44<23:18,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25932\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 550/1243 [18:46<21:43,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25956\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 551/1243 [18:47<20:55,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25958\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 552/1243 [18:49<21:20,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26010\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  44%|     | 553/1243 [18:51<20:08,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26012\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Update the file listing.\n",
      "     */\n",
      "private void updateFileListing()\n",
      "**************************************\n",
      "/**\n",
      " * Update the file listing.\n",
      " */\n",
      "private void updateFileListing() {\n",
      "    String fileListing = \"<html><ol>\";\n",
      "    for (int i = 0; i < files.size(); i++) {\n",
      "        fileListing += \"<li>\" + files.get(i).getName() + \" - <a href=\\\"\" + i + \"\\\">remove</a></li>\";\n",
      "    }\n",
      "    fileListing += \"</ol></html>\";\n",
      "    selectedFilesEditorPane.setText(fileListing);\n",
      "    selectedFilesEditorPane.setCaretPosition(0);\n",
      "}\n",
      "159\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 554/1243 [18:53<20:47,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26039\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 555/1243 [18:55<21:28,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26172\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 556/1243 [18:56<20:05,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 557/1243 [18:58<19:16,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26245\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 558/1243 [19:00<20:54,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26265\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class BareBonesBrowserLaunch {\\n    // list of browsers\\nprivate static final String[] browsers = { \"google-chrome\", \"firefox\", \"opera\", \"epiphany\", \"konqueror\", \"conkeror\", \"midori\", \"kazehakase\", \"mozilla\" };;\\n    public static void openURL(String url)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/examples/BareBonesBrowserLaunch.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class Enzyme {\\n    /*\\n     * The enzyme name.\\n     */\\nprivate final String name;;\\n    /*\\n     * The amino acids before cleavage.\\n     */\\nprivate final HashSet<Character> aminoAcidBefore = new HashSet<>(0);;\\n    /*\\n     * The amino acids after cleavage.\\n     */\\nprivate final HashSet<Character> aminoAcidAfter = new HashSet<>(0);;\\n    /*\\n     * The restriction amino acids before cleavage.\\n     */\\nprivate final HashSet<Character> restrictionBefore = new HashSet<>(0);;\\n    /*\\n     * The restriction amino acids after cleavage.\\n     */\\nprivate final HashSet<Character> restrictionAfter = new HashSet<>(0);;\\n    /**\\n * The CV term associated to this enzyme.\\n */\\nprivate CvTerm cvTerm;;\\n    public Enzyme();\\n    public Enzyme(String name);\\n    public String getName()\\n    public void addAminoAcidAfter(Character aminoAcid)\\n    public HashSet<Character> getAminoAcidAfter()\\n    public void addAminoAcidBefore(Character aminoAcid)\\n    public HashSet<Character> getAminoAcidBefore()\\n    public void addRestrictionAfter(Character aminoAcid)\\n    public HashSet<Character> getRestrictionAfter()\\n    public void addRestrictionBefore(Character aminoAcid)\\n    public HashSet<Character> getRestrictionBefore()\\n    public boolean isCleavageSite(String aaBefore, String aaAfter)\\n    public boolean isCleavageSite(char aaBefore, char aaAfter)\\n    public boolean isCleavageSiteNoCombination(Character aaBefore, Character aaAfter)\\n    public int getNmissedCleavages(String sequence)\\n    public HashSet<String> digest(String sequence, int nMissedCleavages, Integer nMin, Integer nMax)\\n    public HashSet<String> digest(String sequence, int nMissedCleavages, Double massMin, Double massMax)\\n    public boolean equals(Enzyme otherEnzyme)\\n    public String getDescription()\\n    public CvTerm getCvTerm()\\n    public void setCvTerm(CvTerm cvTerm)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/enzymes/Enzyme.java', 'methodCount': 21, 'lineCount': 29}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class EnzymeFactory {\\n    /**\\n * The imported enzymes.\\n */\\nprivate HashMap<String, Enzyme> enzymes = null;;\\n    /**\\n * The instance of the factory.\\n */\\nprivate static EnzymeFactory instance = null;;\\n    /**\\n * The alphabetically sorted enzyme names.\\n */\\nprivate static ArrayList<String> sortedEnzymeNames = null;;\\n    /**\\n * The folder containing the enzymes factory.\\n */\\nprivate static String SERIALIZATION_FILE_FOLDER = System.getProperty(\"user.home\") + \"/.compomics\";;\\n    /**\\n * The name of the enzymes factory back-up file. The version number follows\\n * the one of utilities.\\n */\\nprivate static final String SERIALIZATION_FILE_NAME = \"enzymeFactory-5.0.0.json\";;\\n    private EnzymeFactory();\\n    public static EnzymeFactory getInstance()\\n    public void saveFactory() throws IOException\\n    public static void setDefaultEnzymes()\\n    public static EnzymeFactory loadFromFile(File file) throws IOException\\n    public static void saveToFile(EnzymeFactory enzymeFactory, File file) throws IOException\\n    public static String getSerializationFolder()\\n    public static void setSerializationFolder(String serializationFilePath)\\n    public ArrayList<Enzyme> getEnzymes()\\n    public ArrayList<String> getSortedEnzymeNames()\\n    public Enzyme getEnzyme(String enzymeName)\\n    public void addEnzyme(Enzyme enzyme)\\n    public void removeEnzyme(String enzymeName)\\n    public boolean enzymeLoaded(String enzyme)\\n    public Enzyme getUtilitiesEnzyme(String cvTermAccession)\\n    private static ArrayList<Enzyme> getDefaultEnzymes()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/biology/enzymes/EnzymeFactory.java', 'methodCount': 16, 'lineCount': 23}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class HelpDialog {\\n    /**\\n * The parent of the parent is a Frame.\\n */\\nprivate Frame frameParent = null;;\\n    /**\\n * The parent of the parent is a JDialog.\\n */\\nprivate JDialog dialogParent = null;;\\n    public HelpDialog(Frame parent, URL fileName, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(Frame parent, URL fileName, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(Frame parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(Frame parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(JDialog parent, URL fileName, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(JDialog parent, URL fileName, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(JDialog parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(JDialog parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    private void initComponents()\\n    private void closeJButtonActionPerformed(java.awt.event.ActionEvent evt)\\n    private void textJEditorPaneHyperlinkUpdate(javax.swing.event.HyperlinkEvent evt)\\n    private javax.swing.JPanel backgroundPanel;;\\n    private javax.swing.JButton closeJButton;;\\n    private javax.swing.JScrollPane jScrollPane1;;\\n    private javax.swing.JEditorPane textJEditorPane;;\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/error_handlers/HelpDialog.java', 'methodCount': 11, 'lineCount': 19}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class FileChooserUtil {\\n    public static File getUserSelectedFile(Component parent, String aFileEnding, String aFileFormatDescription, String aDialogTitle, String lastSelectedFolder, String aSuggestedFileName, boolean openDialog)\\n    public static FileAndFileFilter getUserSelectedFile(Component parent, String[] fileEndings, String[] fileFormatDescriptions, String aDialogTitle, String lastSelectedFolder, String aSuggestedFileName, boolean openDialog, boolean formatSelectedByUser, boolean showAllFilesOption, int defaultFilterIndex)\\n    public static File getUserSelectedFolder(Component parent, String aDialogTitle, String lastSelectedFolder, String aFolderDescription, String approveButtonText, boolean openDialog)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/file_handling/FileChooserUtil.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class CvTerm {\\n    public CvTerm();\\n    /**\\n * The ontology.\\n */\\nprivate String ontology;;\\n    /**\\n * The accession number.\\n */\\nprivate String accession;;\\n    /**\\n * The name/term.\\n */\\nprivate String name;;\\n    /**\\n * The value for the given term.\\n */\\nprivate String value;;\\n    public CvTerm(String ontology, String accession, String name, String value);\\n    public String getOntology()\\n    public void setOntology(String ontology)\\n    public String getAccession()\\n    public void setAccession(String accession)\\n    public String getName()\\n    public void setName(String name)\\n    public String getValue()\\n    public void setValue(String value)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/pride/CvTerm.java', 'methodCount': 10, 'lineCount': 16}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 12}\n",
      "{}\n",
      "/**\n",
      "     * Save the enzyme details.\n",
      "     *\n",
      "     * @param evt\n",
      "     */\n",
      "private void formWindowClosing(java.awt.event.WindowEvent evt)\n",
      "**************************************\n",
      "private void formWindowClosing(java.awt.event.WindowEvent evt) {\n",
      "    try {\n",
      "        enzymeFactory.saveFactory();\n",
      "    } catch (IOException e) {\n",
      "        JOptionPane.showMessageDialog(this, \"An error occurred while saving the enzymes.\", \"Saving Error\", JOptionPane.WARNING_MESSAGE);\n",
      "    }\n",
      "    dispose();\n",
      "}\n",
      "2178\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 559/1243 [19:02<21:45,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26303\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 560/1243 [19:04<20:49,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26338\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 561/1243 [19:05<20:14,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26339\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 562/1243 [19:08<21:45,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26412\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 563/1243 [19:10<22:12,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26424\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Util {\\n    /**\\n * Forbidden characters in file names.\\n */\\npublic static final String[] FORBIDDEN_CHARACTERS = { \"!\", \":\", \";\", \"\\\\\\\\?\", \"/\", \"\\\\\\\\\\\\\\\\\", \"\\\\\\\\*\", \"<\", \">\", \"\\\\\"\", \"\\'\", \"\\\\\\\\|\" };;\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default column separator for text files.\\n */\\npublic static final String DEFAULT_COLUMN_SEPARATOR = \"\\\\t\";;\\n    /**\\n * The line separator.\\n */\\npublic static final String LINE_SEPARATOR = System.getProperty(\"line.separator\");;\\n    /**\\n * The mass added per amino acid as part of the reference mass when\\n * converting a tolerance in Dalton to ppm.\\n */\\npublic static final double MASS_PER_AA = 100.0;;\\n    public static String removeSubString(String string, String subString)\\n    public static boolean sendGAUpdate(String ua, String action, String label)\\n    public static String getVersion()\\n    public static String removeCharacters(String string, char forbiddenCharacter)\\n    public static String removeCharacters(String string, String[] forbiddenCharacters)\\n    public static String removeForbiddenCharacters(String string)\\n    public static boolean containsForbiddenCharacter(String string)\\n    public static String replaceAll(String string, char a, char b)\\n    public static double roundDouble(double d, int places)\\n    public static double floorDouble(double d, int places)\\n    public static double getPpmError(double theoreticalMzValue, double massError)\\n    public static String color2Hex(int colorRGB)\\n    public static String color2Hex(Color color)\\n    public static Color getColor(int colorRGB)\\n    public static String tableToText(JTable table, String separator, ProgressDialogX progressDialog, boolean removeHtml)\\n    public static void tableToFile(JTable table, String separator, ProgressDialogX progressDialog, boolean removeHtml, BufferedWriter writer) throws IOException\\n    public static boolean sameLists(ArrayList<?> list1, ArrayList<?> list2)\\n    public static int getOccurrence(String input, char character)\\n    public static ArrayList<Integer> getIndexes(String bigString, String smallString)\\n    public static double readDoubleAsString(String doubleAsString) throws NumberFormatException\\n    public static int convertBooleanToInteger(Boolean booleanToConvert)\\n    public static String keyValueToString(String key, String value)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/Util.java', 'methodCount': 22, 'lineCount': 29}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class RescalingEvent {\\n    // Class specific log4j logger for RescalingEvent instances.\\nLogger logger = LogManager.getLogger(RescalingEvent.class);;\\n    /**\\n * The minimal mass to display after rescaling.\\n */\\nprivate double iMinMass = 0;;\\n    /**\\n * The maximal mass to display after rescaling.\\n */\\nprivate double iMaxMass = 0;;\\n    /**\\n * This JPanel represents the source.\\n */\\nprivate JPanel iSource = null;;\\n    public RescalingEvent(JPanel aSource, double aMinMass, double aMaxMass);\\n    public double getMaxMass()\\n    public double getMinMass()\\n    public JPanel getSource()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/events/RescalingEvent.java', 'methodCount': 4, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface SpectrumAnnotation {\\n    public abstract double getMZ()\\n    public abstract double getErrorMargin()\\n    public abstract Color getColor()\\n    public abstract String getLabel()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/interfaces/SpectrumAnnotation.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface SpectrumPanelListener {\\n    public void rescaled(RescalingEvent aSe)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/interfaces/SpectrumPanelListener.java', 'methodCount': 1, 'lineCount': 3}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Annotate the annotated peaks.\n",
      "     *\n",
      "     * @param g the <code>Graphics</code> context in which to paint\n",
      "     * @param annotations the annotations\n",
      "     * @param mirrored if the annotations are for the mirrored spectra\n",
      "     */\n",
      "protected void annotatePeaks(Graphics g, Vector annotations, boolean mirrored)\n",
      "**************************************\n",
      "protected void annotatePeaks(Graphics g, Vector annotations, boolean mirrored) {\n",
      "    if (annotations != null && annotations.size() > 0 && !miniature) {\n",
      "        HashMap<String, Integer> annotatedPeaks = new HashMap<>();\n",
      "        for (Object o : annotations) {\n",
      "            if (o instanceof SpectrumAnnotation) {\n",
      "                SpectrumAnnotation sa = (SpectrumAnnotation) o;\n",
      "                this.annotate(sa, g, annotatedPeaks, mirrored);\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "}\n",
      "1206\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 564/1243 [19:12<22:50,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26510\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  45%|     | 565/1243 [19:14<23:18,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26511\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 566/1243 [19:16<23:27,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26682\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 567/1243 [19:18<23:38,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26697\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 568/1243 [19:21<23:49,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26754\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 45, 'classCount': 1, 'methodsOnly': 'public class Advocate {\\n    public Advocate();\\n    /**\\n * The Mascot search engine.\\n */\\npublic static final Advocate mascot = new Advocate(0, \"Mascot\", AdvocateType.search_engine, new java.awt.Color(255, 153, 255));;\\n    /**\\n * The OMSSA search engine.\\n */\\npublic static final Advocate omssa = new Advocate(1, \"OMSSA\", AdvocateType.search_engine, new java.awt.Color(153, 153, 255));;\\n    /**\\n * The X! Tandem search engine.\\n */\\npublic static final Advocate xtandem = new Advocate(2, \"X!Tandem\", AdvocateType.search_engine, new java.awt.Color(153, 255, 255));;\\n    /**\\n * The PepNovo+ de novo sequencing algorithm.\\n */\\npublic static final Advocate pepnovo = new Advocate(3, \"PepNovo+\", AdvocateType.sequencing_algorithm, new java.awt.Color(224, 130, 20));;\\n    /**\\n * The Andromeda search engine.\\n */\\npublic static final Advocate andromeda = new Advocate(4, \"Andromeda\", AdvocateType.search_engine, new Color(230, 230, 250));;\\n    /**\\n * The MS Amanda search engine.\\n */\\npublic static final Advocate msAmanda = new Advocate(5, \"MS Amanda\", AdvocateType.search_engine, new java.awt.Color(216, 191, 216));;\\n    /**\\n * The PeptideShaker multiple algorithm software.\\n */\\npublic static final Advocate peptideShaker = new Advocate(6, \"PeptideShaker\", AdvocateType.multiple_algorithm_software, new Color(110, 196, 97));;\\n    /**\\n * The MS-GF+ search engine.\\n */\\npublic static final Advocate msgf = new Advocate(7, \"MS-GF+\", AdvocateType.search_engine, new java.awt.Color(205, 92, 92));;\\n    /**\\n * The DirecTag sequencing algorithm.\\n */\\npublic static final Advocate direcTag = new Advocate(8, \"DirecTag\", AdvocateType.sequencing_algorithm, new java.awt.Color(189, 183, 107));;\\n    /**\\n * The Byonic search engine integrated in the Byonic protein metrics\\n * interface.\\n */\\n// @TODO: add color\\npublic static final Advocate byonic = new Advocate(9, \"Byonic\", AdvocateType.search_engine);;\\n    /**\\n * The Comet search engine, free version of Sequest.\\n */\\npublic static final Advocate comet = new Advocate(10, \"Comet\", AdvocateType.search_engine, new java.awt.Color(255, 160, 122));;\\n    /**\\n * The ProteinLynx search engine, waters instruments.\\n */\\npublic static final Advocate proteinLynx = new Advocate(11, \"ProteinLynx\", AdvocateType.search_engine);;\\n    /**\\n * The MS-Fit search engine, old school peptide mass fingerprinting.\\n */\\npublic static final Advocate msFit = new Advocate(12, \"MS-Fit\", AdvocateType.search_engine);;\\n    /**\\n * The MyriMatch search engine, old school peptide mass fingerprinting.\\n */\\npublic static final Advocate myriMatch = new Advocate(13, \"MyriMatch\", AdvocateType.search_engine, new Color(241, 226, 204));;\\n    /**\\n * The PEAKS sequencing algorithm.\\n */\\npublic static final Advocate peaks = new Advocate(14, \"PEAKS Studio\", AdvocateType.sequencing_algorithm, new Color(173, 255, 47));;\\n    /**\\n * The Phenyx search engine.\\n */\\npublic static final Advocate phenyx = new Advocate(15, \"Phenyx\", AdvocateType.search_engine);;\\n    /**\\n * The Profound search engine.\\n */\\npublic static final Advocate proFound = new Advocate(16, \"ProFound\", AdvocateType.search_engine);;\\n    /**\\n * The search engine results of Protein Prospector.\\n */\\npublic static final Advocate proteinProspector = new Advocate(17, \"ProteinProspector\", AdvocateType.search_engine);;\\n    /**\\n * The search engine integrated in Protein Scape, Bruker instruments.\\n */\\npublic static final Advocate proteinScape = new Advocate(18, \"ProteinScape\", AdvocateType.search_engine);;\\n    /**\\n * The Sequest search engine.\\n */\\npublic static final Advocate sequest = new Advocate(19, \"SEQUEST\", AdvocateType.search_engine);;\\n    /**\\n * The SeQuence IDentfication (SQID) search engine.\\n */\\npublic static final Advocate sqid = new Advocate(20, \"SQID\", AdvocateType.search_engine);;\\n    /**\\n * The Scaffold multiple search engine data interpretation software.\\n */\\npublic static final Advocate scaffold = new Advocate(21, \"Scaffold\", AdvocateType.multiple_algorithm_software);;\\n    /**\\n * The Sonar search engine, integrated in radars.\\n */\\npublic static final Advocate sonar = new Advocate(22, \"Sonar\", AdvocateType.search_engine);;\\n    /**\\n * The SpectraST spectral library search engine (TPP).\\n */\\npublic static final Advocate spectraST = new Advocate(23, \"SpectraST\", AdvocateType.spectral_library);;\\n    /**\\n * The Spectrum Mill search engine, Agilent.\\n */\\npublic static final Advocate spectrumMill = new Advocate(24, \"SpectrumMill\", AdvocateType.search_engine);;\\n    /**\\n * The ZCore search engine, ETD search engine.\\n */\\npublic static final Advocate zCore = new Advocate(25, \"ZCore\", AdvocateType.search_engine);;\\n    /**\\n * The Percolator rescoring algorithm.\\n */\\npublic static final Advocate percolator = new Advocate(26, \"Percolator\", AdvocateType.rescoring_algorithm);;\\n    /**\\n * The pNovo+ de novo sequencing algorithm.\\n */\\npublic static final Advocate pNovo = new Advocate(27, \"pNovo+\", AdvocateType.sequencing_algorithm, new Color(253, 192, 134));;\\n    /**\\n * The Tide search engine.\\n */\\npublic static final Advocate tide = new Advocate(28, \"Tide\", AdvocateType.search_engine, new Color(210, 105, 30));;\\n    /**\\n * The Novor de novo sequencing algorithm.\\n */\\npublic static final Advocate novor = new Advocate(29, \"Novor\", AdvocateType.sequencing_algorithm, new Color(135, 206, 235));;\\n    /**\\n * The Morpheus search engine.\\n */\\npublic static final Advocate morpheus = new Advocate(30, \"Morpheus\", AdvocateType.search_engine, new Color(255, 248, 220));;\\n    /**\\n * The ProteinPilot search engine.\\n */\\npublic static final Advocate proteinPilot = new Advocate(31, \"ProteinPilot Software\", AdvocateType.search_engine, new Color(75, 0, 130));;\\n    /**\\n * The Onyase engine script.\\n */\\npublic static final Advocate onyaseEngine = new Advocate(32, \"Onyase Engine\", AdvocateType.search_engine);;\\n    /**\\n * The MetaMorpheus search engine.\\n */\\npublic static final Advocate metaMorpheus = new Advocate(33, \"MetaMorpheus\", AdvocateType.search_engine, new Color(218, 165, 32));;\\n    /**\\n * The IdentiPy search engine.\\n */\\npublic static final Advocate identiPy = new Advocate(34, \"IdentiPy\", AdvocateType.search_engine, new Color(128, 128, 0));;\\n    /**\\n * The CompOmics Spectral Searching (COSS) engine.\\n */\\npublic static final Advocate coss = new Advocate(35, \"COSS\", AdvocateType.spectral_library, new Color(108, 108, 0));;\\n    /**\\n * The Sage search engine.\\n */\\npublic static final Advocate sage = new Advocate(36, \"Sage\", AdvocateType.search_engine, new Color(108, 108, 0));;\\n    /**\\n * Advocate type for mzId files where no software is annotated.\\n */\\npublic static final Advocate genericMzId = new Advocate(100, \"mzid\", AdvocateType.unknown);;\\n    /**\\n * Map of user-defined advocates indexed by index.\\n */\\nprivate static HashMap<Integer, Advocate> userAdvocates = new HashMap<>();;\\n    /**\\n * The index of the advocate.\\n */\\nprivate final int index;;\\n    /**\\n * The name of the advocate.\\n */\\nprivate final String name;;\\n    /**\\n * The type of advocate\\n */\\nprivate final AdvocateType type;;\\n    /**\\n * The color of the advocate. Defaults to light gray.\\n */\\nprivate Color color = Color.lightGray;;\\n    /**\\n * The search engine color map.\\n */\\nprivate static HashMap<Integer, java.awt.Color> advocateColorMap;;\\n    /**\\n * The search engine tool tip map.\\n */\\nprivate static HashMap<Integer, String> advocateToolTipMap;;\\n    private Advocate(int index, String name, AdvocateType type, Color color);\\n    private Advocate(int index, String name, AdvocateType type);\\n    private Advocate(int index, String name);\\n    public int getIndex()\\n    public String getName()\\n    public AdvocateType getType()\\n    public Color getColor()\\n    public String toString()\\n    public static Advocate[] values()\\n    public static Advocate getAdvocate(int index)\\n    public static Advocate getAdvocate(String advocateName)\\n    public static Advocate addUserAdvocate(String advocateName)\\n    public static HashMap<Integer, Advocate> getUserAdvocates()\\n    public static void setUserAdvocates(HashMap<Integer, Advocate> userAdvocates)\\n    public String getPmid()\\n    public static HashMap<Integer, java.awt.Color> getAdvocateColorMap()\\n    public static HashMap<Integer, String> getAdvocateToolTipMap()\\n    public static Advocate getAdvocateFromFile(String idFileName)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/identification/Advocate.java', 'methodCount': 19, 'lineCount': 66}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/experiment/identification/psm_scoring/PsmScore.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class HelpDialog {\\n    /**\\n * The parent of the parent is a Frame.\\n */\\nprivate Frame frameParent = null;;\\n    /**\\n * The parent of the parent is a JDialog.\\n */\\nprivate JDialog dialogParent = null;;\\n    public HelpDialog(Frame parent, URL fileName, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(Frame parent, URL fileName, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(Frame parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(Frame parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(JDialog parent, URL fileName, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(JDialog parent, URL fileName, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(JDialog parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(JDialog parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    private void initComponents()\\n    private void closeJButtonActionPerformed(java.awt.event.ActionEvent evt)\\n    private void textJEditorPaneHyperlinkUpdate(javax.swing.event.HyperlinkEvent evt)\\n    private javax.swing.JPanel backgroundPanel;;\\n    private javax.swing.JButton closeJButton;;\\n    private javax.swing.JScrollPane jScrollPane1;;\\n    private javax.swing.JEditorPane textJEditorPane;;\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/error_handlers/HelpDialog.java', 'methodCount': 11, 'lineCount': 19}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class PsmScoringParameters {\\n    /**\\n * The scores used to score the spectrum matches for every advocate in a\\n * map: advocate index &gt; list of score indexes.\\n */\\nprivate HashMap<Integer, HashSet<Integer>> spectrumMatchingScores = null;;\\n    /**\\n * The scores to use by default.\\n */\\nprivate HashSet<Integer> defaultScores;;\\n    /**\\n * The minimal number of decoys to include in a bin to set the bin size of\\n * the score histogram.\\n */\\nprivate int minDecoysInBin = 10;;\\n    public PsmScoringParameters();\\n    public void addScore(Integer advocateId, Integer scoreId)\\n    public void clearScores(Integer advocateId)\\n    public void clearAllScores()\\n    public HashSet<Integer> getScoreForAlgorithm(int advocateId)\\n    public boolean isScoringNeeded(int advocate)\\n    public Set<Integer> getAdvocates()\\n    public boolean isScoringNeeded(ArrayList<Integer> advocates)\\n    public boolean isTargetDecoyNeededForPsmScoring(ArrayList<Integer> advocates)\\n    public String getShortDescription()\\n    public boolean equals(PsmScoringParameters otherPsmScoringPreferences)\\n    public HashSet<Integer> getDefaultScores()\\n    private void setDefaultScores()\\n    public int getDecoysInFirstBin()\\n    public void setDecoysInFirstBin(Integer decoysInFirstBin)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/parameters/identification/advanced/PsmScoringParameters.java', 'methodCount': 15, 'lineCount': 20}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 13}\n",
      "{}\n",
      "/**\n",
      "     * Sets the table properties.\n",
      "     */\n",
      "private void setTableProperties()\n",
      "**************************************\n",
      "private void setTableProperties() {\n",
      "    TableColumnModel tableColumnModel = psmScoresJTable.getColumnModel();\n",
      "    tableColumnModel.getColumn(0).setMaxWidth(30);\n",
      "    psmScoresJScrollPane.getViewport().setOpaque(false);\n",
      "    for (int i = 0; i < PsmScore.values().length; i++) {\n",
      "        tableColumnModel.getColumn(i + 2).setCellRenderer(new TrueFalseIconRenderer(new ImageIcon(this.getClass().getResource(\"/icons/selected_green-new.png\")), null, \"On\", \"Off\"));\n",
      "    }\n",
      "}\n",
      "3425\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 569/1243 [19:23<24:00,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26757\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 570/1243 [19:24<22:06,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26840\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 571/1243 [19:26<20:40,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "26843\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 572/1243 [19:28<21:00,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27151\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 573/1243 [19:30<21:35,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27250\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class BareBonesBrowserLaunch {\\n    // list of browsers\\nprivate static final String[] browsers = { \"google-chrome\", \"firefox\", \"opera\", \"epiphany\", \"konqueror\", \"conkeror\", \"midori\", \"kazehakase\", \"mozilla\" };;\\n    public static void openURL(String url)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/examples/BareBonesBrowserLaunch.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface IdentificationAlgorithmParameter {\\n    public Advocate getAlgorithm()\\n    public boolean equals(IdentificationAlgorithmParameter identificationAlgorithmParameter)\\n    public String toString(boolean html)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/parameters/identification/IdentificationAlgorithmParameter.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class PNovoParameters {\\n    /**\\n * Version number for deserialization.\\n */\\nstatic final long serialVersionUID = 7525455518683797145L;;\\n    /**\\n * The number of peptides reported.\\n */\\nprivate int numberOfPeptides = 10;;\\n    /**\\n * The minimum precursor mass.\\n */\\nprivate int lowerPrecursorMass = 300;;\\n    /**\\n * The maximum precursor mass.\\n */\\nprivate int upperPrecursorMass = 5000;;\\n    /**\\n * The activation type (HCD, CID or ETD).\\n */\\nprivate String acticationType = \"HCD\";;\\n    /**\\n * A map from the pNovo PTM character to the utilities PTM names. pNovo PTM\\n * character &gt; utilities PTM name.\\n */\\nprivate HashMap<Character, String> pNovoPtmMap;;\\n    /**\\n * A map from the pNovo character to the original amino acids residue. pNovo\\n * PTM character &gt; original amino acids residue.\\n */\\nprivate HashMap<Character, Character> pNovoResidueMap;;\\n    public PNovoParameters();\\n    public Advocate getAlgorithm()\\n    public boolean equals(IdentificationAlgorithmParameter identificationAlgorithmParameter)\\n    public String toString(boolean html)\\n    public int getNumberOfPeptides()\\n    public void setNumberOfPeptides(int numberOfPeptides)\\n    public int getLowerPrecursorMass()\\n    public void setLowerPrecursorMass(int lowerPrecursorMass)\\n    public int getUpperPrecursorMass()\\n    public void setUpperPrecursorMass(int upperPrecursorMass)\\n    public String getActicationType()\\n    public void setActicationType(String acticationType)\\n    public HashMap<Character, String> getPNovoPtmMap()\\n    public String getUtilitiesPtmName(Character pNovoPtmCharacter)\\n    public void setPNovoPtmMap(HashMap<Character, String> pNovoPtmMap)\\n    public HashMap<Character, Character> getPNovoPtmResiduesMap()\\n    public Character getPtmResidue(Character pNovoPtmCharacter)\\n    public void setPNovoPtmResiduesMap(HashMap<Character, Character> pNovoResidueMap)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/parameters/identification/tool_specific/PNovoParameters.java', 'methodCount': 18, 'lineCount': 27}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class GuiUtilities {\\n    /**\\n * Escape key stroke.\\n */\\nprivate static final KeyStroke escapeStroke = KeyStroke.getKeyStroke(KeyEvent.VK_ESCAPE, 0);;\\n    /**\\n * Dispatch window closing action map key.\\n */\\nprivate static final String dispatchWindowClosingActionMapKey = \"com.compomics.util.gui:WINDOW_CLOSING\";;\\n    public static int getPreferredWidthOfCell(JTable table, int rowIndex, int colIndex)\\n    public static boolean validateIntegerInput(Component parentComponent, JLabel label, JTextField textField, String valueDescription, String errorTitle, boolean positiveValue, boolean showMessage, boolean valid)\\n    public static boolean validateDoubleInput(Component parentComponent, JLabel label, JTextField textField, String valueDescription, String errorTitle, boolean positiveValue, boolean showMessage, boolean valid)\\n    public static void installEscapeCloseOperation(final JDialog dialog)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/GuiUtilities.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface AlgorithmParametersDialog {\\n    public boolean isCancelled()\\n    public IdentificationAlgorithmParameter getParameters()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/parameters/identification/AlgorithmParametersDialog.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Sets up the GUI.\n",
      "     */\n",
      "private void setUpGUI()\n",
      "**************************************\n",
      "/**\n",
      " * Sets up the GUI.\n",
      " */\n",
      "private void setUpGUI() {\n",
      "    activationTypeCmb.setRenderer(new com.compomics.util.gui.renderers.AlignedListCellRenderer(SwingConstants.CENTER));\n",
      "    activationTypeCmb.setEnabled(editable);\n",
      "    minPrecursorMassTextField.setEditable(editable);\n",
      "    minPrecursorMassTextField.setEnabled(editable);\n",
      "    maxPrecursorMassTextField.setEditable(editable);\n",
      "    maxPrecursorMassTextField.setEnabled(editable);\n",
      "}\n",
      "1316\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 574/1243 [19:32<22:46,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27287\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 575/1243 [19:34<22:57,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27324\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 576/1243 [19:36<21:16,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27355\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  46%|     | 577/1243 [19:37<20:04,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27383\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 578/1243 [19:39<19:22,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27401\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 11}\n",
      "{'CheckableItem': 'public class CheckableItem {\\n    /**\\n * The text to display.\\n */\\npublic final String text;;\\n    /**\\n * Whether the item is selected or not.\\n */\\npublic boolean selected;;\\n    public CheckableItem(String text, boolean selected);\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Returns the checked items as an array of strings.\n",
      "     *\n",
      "     * @param model the list model\n",
      "     * @return the checked items as an array of strings\n",
      "     */\n",
      "public ArrayList<String> getCheckedItems(ListModel model)\n",
      "**************************************\n",
      "public ArrayList<String> getCheckedItems(ListModel model) {\n",
      "    ArrayList<String> sl = new ArrayList<>();\n",
      "    for (int i = 0; i < model.getSize(); i++) {\n",
      "        Object o = model.getElementAt(i);\n",
      "        if (o instanceof CheckableItem && ((CheckableItem) o).selected) {\n",
      "            sl.add(o.toString());\n",
      "        }\n",
      "    }\n",
      "    return sl;\n",
      "}\n",
      "245\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 579/1243 [19:41<18:49,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27402\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 580/1243 [19:42<18:27,  1.67s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27418\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 581/1243 [19:44<19:47,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27499\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 582/1243 [19:46<20:52,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27562\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 583/1243 [19:48<20:12,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27577\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class BareBonesBrowserLaunch {\\n    // list of browsers\\nprivate static final String[] browsers = { \"google-chrome\", \"firefox\", \"opera\", \"epiphany\", \"konqueror\", \"conkeror\", \"midori\", \"kazehakase\", \"mozilla\" };;\\n    public static void openURL(String url)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/examples/BareBonesBrowserLaunch.java', 'methodCount': 1, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DummyFrame {\\n    public DummyFrame(String title, String relativeImageIconPath);\\n    public DummyFrame(String title, Image imageIcon);\\n    public DummyFrame setNewTitle(String title)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/DummyFrame.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class SimpleSemaphore {\\n    /**\\n * The semaphore to use.\\n */\\nprivate final Semaphore mutex;;\\n    public SimpleSemaphore(int nPermits, boolean fair);\\n    public SimpleSemaphore(int nPermits);\\n    public void acquire()\\n    public void release()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/threading/SimpleSemaphore.java', 'methodCount': 4, 'lineCount': 7}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface WaitingActionListener {\\n    public void cancelPressed()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/waiting/WaitingActionListener.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public interface WaitingHandler {\\n    /**\\n * Convenience date format.\\n */\\npublic static final SimpleDateFormat SIMPLE_DATA_FORMAT = new SimpleDateFormat(\"dd MMM yyyy, HH:mm\");;\\n    /**\\n * The tab space to add when using HTML.\\n */\\npublic static final String TAB_HTML = \"&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\";;\\n    /**\\n * The tab space to add when not using HTML.\\n */\\npublic static final String TAB_NON_HTML = \"        \";;\\n    public void setPrimaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setMaxPrimaryProgressCounter(int maxProgressValue)\\n    public void increasePrimaryProgressCounter()\\n    public void increasePrimaryProgressCounter(int increment)\\n    public void setPrimaryProgressCounter(int value)\\n    public void setMaxSecondaryProgressCounter(int maxProgressValue)\\n    public void resetPrimaryProgressCounter()\\n    public void resetSecondaryProgressCounter()\\n    public void increaseSecondaryProgressCounter()\\n    public void setSecondaryProgressCounter(int value)\\n    public void increaseSecondaryProgressCounter(int value)\\n    public void setSecondaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setRunFinished()\\n    public void setRunCanceled()\\n    public void appendReport(String report, boolean includeDate, boolean addNewLine)\\n    public void appendReportNewLineNoDate()\\n    public void appendReportEndLine()\\n    public boolean isRunCanceled()\\n    public boolean isRunFinished()\\n    public void setSecondaryProgressText(String text)\\n    public boolean isReport()\\n    public void setWaitingText(String text)\\n    public int getPrimaryProgressCounter()\\n    public int getMaxPrimaryProgressCounter()\\n    public int getSecondaryProgressCounter()\\n    public int getMaxSecondaryProgressCounter()\\n    public void setDisplayProgress(boolean displayProgress)\\n    public boolean getDisplayProgress()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/waiting/WaitingHandler.java', 'methodCount': 28, 'lineCount': 33}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 15}\n",
      "{'DummyFrame': 'public class DummyFrame {\\n    public DummyFrame(String title, String relativeImageIconPath);\\n    public DummyFrame(String title, Image imageIcon);\\n    public DummyFrame setNewTitle(String title)\\n}\\n'}\n",
      "/**\n",
      "     * Cancels the process if ongoing or opens the results if finished.\n",
      "     *\n",
      "     * @param evt\n",
      "     */\n",
      "private void okButtonActionPerformed(java.awt.event.ActionEvent evt)\n",
      "**************************************\n",
      "private void okButtonActionPerformed(java.awt.event.ActionEvent evt) {\n",
      "    if (runFinished || runCanceled) {\n",
      "        if (waitingHandlerParent instanceof DummyFrame) {\n",
      "            ((DummyFrame) waitingHandlerParent).dispose();\n",
      "        }\n",
      "        this.dispose();\n",
      "    } else {\n",
      "        setRunCanceled();\n",
      "    }\n",
      "}\n",
      "1120\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 584/1243 [19:50<21:33,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27622\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 585/1243 [19:53<22:09,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27623\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 586/1243 [19:54<20:48,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27651\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 587/1243 [19:56<21:55,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27689\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 588/1243 [19:58<21:30,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27713\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class HelpDialog {\\n    /**\\n * The parent of the parent is a Frame.\\n */\\nprivate Frame frameParent = null;;\\n    /**\\n * The parent of the parent is a JDialog.\\n */\\nprivate JDialog dialogParent = null;;\\n    public HelpDialog(Frame parent, URL fileName, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(Frame parent, URL fileName, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(Frame parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(Frame parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(JDialog parent, URL fileName, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(JDialog parent, URL fileName, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    public HelpDialog(JDialog parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title);\\n    public HelpDialog(JDialog parent, URL fileName, String reference, Image helpIcon, Image aboutIcon, String title, int windowWidth, int windowHeightReduction);\\n    private void initComponents()\\n    private void closeJButtonActionPerformed(java.awt.event.ActionEvent evt)\\n    private void textJEditorPaneHyperlinkUpdate(javax.swing.event.HyperlinkEvent evt)\\n    private javax.swing.JPanel backgroundPanel;;\\n    private javax.swing.JButton closeJButton;;\\n    private javax.swing.JScrollPane jScrollPane1;;\\n    private javax.swing.JEditorPane textJEditorPane;;\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/error_handlers/HelpDialog.java', 'methodCount': 11, 'lineCount': 19}, {'fieldCount': 16, 'classCount': 1, 'methodsOnly': 'public class ExportGraphicsDialog {\\n    /**\\n * The chart panel to export.\\n */\\nprivate ChartPanel chartPanel;;\\n    /**\\n * The panel to export.\\n */\\nprivate Component graphicsPanel;;\\n    /**\\n * The progress dialog.\\n */\\nprivate ProgressDialogX progressDialog;;\\n    /**\\n * The parent frame.\\n */\\nprivate JFrame frame;;\\n    /**\\n * The normal icon.\\n */\\nprivate Image normalIcon;;\\n    /**\\n * The waiting icon.\\n */\\nprivate Image waitingIcon;;\\n    /**\\n * The last selected folder.\\n */\\nprivate LastSelectedFolder lastSelectedFolder = new LastSelectedFolder();;\\n    public ExportGraphicsDialog(JFrame frame, Image normalIcon, Image waitingIcon, boolean modal, Component graphicsPanel, LastSelectedFolder lastSelectedFolder);\\n    public ExportGraphicsDialog(JFrame frame, Image normalIcon, Image waitingIcon, boolean modal, ChartPanel chartPanel, LastSelectedFolder lastSelectedFolder);\\n    public ExportGraphicsDialog(JDialog dialog, Image normalIcon, Image waitingIcon, boolean modal, Component graphicsPanel, LastSelectedFolder lastSelectedFolder);\\n    public ExportGraphicsDialog(JDialog dialog, Image normalIcon, Image waitingIcon, boolean modal, ChartPanel chartPanel, LastSelectedFolder lastSelectedFolder);\\n    private void initComponents()\\n    private void exportJButtonActionPerformed(java.awt.event.ActionEvent evt)\\n    private void cancelJButtonActionPerformed(java.awt.event.ActionEvent evt)\\n    private javax.swing.JPanel backgroundPanel;;\\n    private javax.swing.ButtonGroup buttonGroup;;\\n    private javax.swing.JButton cancelJButton;;\\n    private javax.swing.JButton exportJButton;;\\n    private javax.swing.JRadioButton pdfJRadioButton;;\\n    private javax.swing.JPanel plottingTypeJPanel;;\\n    private javax.swing.JRadioButton pngJRadioButton;;\\n    private javax.swing.JRadioButton svgJRadioButton;;\\n    private javax.swing.JRadioButton tiffJRadioButton;;\\n    private void savePanel(String selectedFile, boolean aShowSavedMessage)\\n    private void addFileFilter(JFileChooser chooser)\\n    private String getLastSelectedFolder()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/gui/export/graphics/ExportGraphicsDialog.java', 'methodCount': 10, 'lineCount': 28}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class LastSelectedFolder {\\n    /**\\n * Serial version UID for backward compatibility.\\n */\\nstatic final long serialVersionUID = 7612497861284156966L;;\\n    /**\\n * Map of the last selected folders indexed by use cases.\\n */\\nprivate HashMap<String, String> lastSelectedFolder = null;;\\n    /**\\n * The default use case.\\n */\\nprivate static final String defaultUseCase = \"default\";;\\n    public LastSelectedFolder(String defaultFolder);\\n    public LastSelectedFolder();\\n    public void setLastSelectedFolder(String useCase, String folderPath)\\n    public void setLastSelectedFolder(String folderPath)\\n    public String getLastSelectedFolder(String useCase)\\n    public String getLastSelectedFolder()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/file/LastSelectedFolder.java', 'methodCount': 6, 'lineCount': 11}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Allow the user to change the value font size.\n",
      "     *\n",
      "     * @param evt\n",
      "     */\n",
      "private void valueFontSizeMenuItemActionPerformed(java.awt.event.ActionEvent evt)\n",
      "**************************************\n",
      "private void valueFontSizeMenuItemActionPerformed(java.awt.event.ActionEvent evt) {\n",
      "    String value = JOptionPane.showInputDialog(this, \"Values Font Size:\", vennDiagramPanel.getFontSize());\n",
      "    if (value != null) {\n",
      "        try {\n",
      "            vennDiagramPanel.setFontSize(Integer.parseInt(value));\n",
      "            vennDiagramPanel.updatePlot();\n",
      "        } catch (NumberFormatException e) {\n",
      "            JOptionPane.showMessageDialog(this, \"Font size has to be an integer!\", \"Font Error\", JOptionPane.INFORMATION_MESSAGE);\n",
      "        }\n",
      "    }\n",
      "}\n",
      "1375\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 589/1243 [20:00<20:19,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27714\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  47%|     | 590/1243 [20:02<21:16,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27815\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 591/1243 [20:04<20:21,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27824\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 592/1243 [20:05<19:30,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27829\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 593/1243 [20:07<19:03,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27851\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Exclude null values from an array of double.\n",
      "     *\n",
      "     * @param data the data\n",
      "     * @return another double array with no longer null values\n",
      "     */\n",
      "public double[] excludeNullValues(Double[] data)\n",
      "**************************************\n",
      "public double[] excludeNullValues(Double[] data) {\n",
      "    ArrayList<Double> list = new ArrayList<>();\n",
      "    for (Double value : data) {\n",
      "        if (value != null) {\n",
      "            list.add(value);\n",
      "        }\n",
      "    }\n",
      "    double[] newArray = new double[list.size()];\n",
      "    for (int i = 0; i < list.size(); i++) {\n",
      "        newArray[i] = list.get(i);\n",
      "    }\n",
      "    return newArray;\n",
      "}\n",
      "173\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 594/1243 [20:09<18:52,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27925\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 595/1243 [20:11<20:09,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27927\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 596/1243 [20:13<19:31,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27956\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 597/1243 [20:14<19:05,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27962\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 598/1243 [20:16<19:59,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27965\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * This method reads a line of text from the buffer.\n",
      "     * It returns a 'null' if the EOF is encountered.\n",
      "     *\n",
      "     * @return  String  with the current line in the buffer, or 'null' if the end\n",
      "     *                  of the buffer has been reached.\n",
      "     */\n",
      "public String readLine()\n",
      "**************************************\n",
      "/**\n",
      " * This method reads a line of text from the buffer.\n",
      " * It returns a 'null' if the EOF is encountered.\n",
      " *\n",
      " * @return  String  with the current line in the buffer, or 'null' if the end\n",
      " *                  of the buffer has been reached.\n",
      " */\n",
      "public String readLine() {\n",
      "    String result = null;\n",
      "    if (!iEOF) {\n",
      "        int returnIndex = iLocation;\n",
      "        iLocation++;\n",
      "        if (iLocation == iData.size()) {\n",
      "            iEOF = true;\n",
      "        } else if (iLocation > iData.size()) {\n",
      "            throw new ArrayIndexOutOfBoundsException(\"Vector elements count up to element \" + (iData.size() - 1) + \" and current index is \" + iLocation + \".\");\n",
      "        }\n",
      "        result = (String) iData.get(returnIndex);\n",
      "    }\n",
      "    return result;\n",
      "}\n",
      "284\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 599/1243 [20:18<19:09,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27968\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 600/1243 [20:20<20:03,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27988\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 601/1243 [20:22<21:01,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27990\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  48%|     | 602/1243 [20:24<21:29,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 603/1243 [20:26<20:11,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28046\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Reads an object from a serialized file.\n",
      "     *\n",
      "     * @param serializedFile the serialized file\n",
      "     * @return the object\n",
      "     * @throws IOException exception thrown whenever an error occurred while\n",
      "     * reading the file\n",
      "     * @throws ClassNotFoundException exception thrown whenever an unknown class\n",
      "     * is found\n",
      "     */\n",
      "public static Object readObject(File serializedFile) throws IOException, ClassNotFoundException\n",
      "**************************************\n",
      "public static Object readObject(File serializedFile) throws IOException, ClassNotFoundException {\n",
      "    FileInputStream fis = new FileInputStream(serializedFile);\n",
      "    try {\n",
      "        BufferedInputStream bis = new BufferedInputStream(fis);\n",
      "        try {\n",
      "            ObjectInputStream in = new ObjectInputStream(bis);\n",
      "            try {\n",
      "                Object object = in.readObject();\n",
      "                return object;\n",
      "            } finally {\n",
      "                in.close();\n",
      "            }\n",
      "        } finally {\n",
      "            bis.close();\n",
      "        }\n",
      "    } finally {\n",
      "        fis.close();\n",
      "    }\n",
      "}\n",
      "217\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 604/1243 [20:28<19:33,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28052\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 605/1243 [20:30<20:33,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28171\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 606/1243 [20:32<21:15,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28176\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 607/1243 [20:34<19:47,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28202\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 608/1243 [20:35<18:53,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28209\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{'FtpLoginException': 'public class FtpLoginException {\\n    public FtpLoginException();\\n    // Class specific log4j logger for FtpLoginException instances.\\nLogger logger = LogManager.getLogger(FtpLoginException.class);;\\n     FtpLoginException(String s);\\n}\\n'}\n",
      "/**\n",
      "     * Login user to a host with username <i>user</i> and password\n",
      "     * <i>password</i>.\n",
      "     *\n",
      "     * @param   user    String with the username to use.\n",
      "     * @param   password    String with the password to use (passwords are sent in plain text in FTP!)\n",
      "     * @throws IOException if an IOException occurs\n",
      "     */\n",
      "public void login(String user, String password) throws IOException\n",
      "**************************************\n",
      "public void login(String user, String password) throws IOException {\n",
      "    if (!serverIsOpen()) {\n",
      "        throw new FtpLoginException(\"Error: not connected to host.\\n\");\n",
      "    }\n",
      "    this.user = user;\n",
      "    this.password = password;\n",
      "    if (issueCommand(\"USER \" + user) == FTP_ERROR) {\n",
      "        throw new FtpLoginException(\"Error: User not found.\\n\");\n",
      "    }\n",
      "    if (password != null && issueCommand(\"PASS \" + password) == FTP_ERROR) {\n",
      "        throw new FtpLoginException(\"Error: Wrong Password.\\n\");\n",
      "    }\n",
      "}\n",
      "317\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 609/1243 [20:37<18:22,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28211\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 610/1243 [20:38<18:09,  1.72s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28212\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 611/1243 [20:40<17:50,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28239\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 612/1243 [20:42<18:58,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28240\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 613/1243 [20:44<19:55,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28255\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Util {\\n    /**\\n * Forbidden characters in file names.\\n */\\npublic static final String[] FORBIDDEN_CHARACTERS = { \"!\", \":\", \";\", \"\\\\\\\\?\", \"/\", \"\\\\\\\\\\\\\\\\\", \"\\\\\\\\*\", \"<\", \">\", \"\\\\\"\", \"\\'\", \"\\\\\\\\|\" };;\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default column separator for text files.\\n */\\npublic static final String DEFAULT_COLUMN_SEPARATOR = \"\\\\t\";;\\n    /**\\n * The line separator.\\n */\\npublic static final String LINE_SEPARATOR = System.getProperty(\"line.separator\");;\\n    /**\\n * The mass added per amino acid as part of the reference mass when\\n * converting a tolerance in Dalton to ppm.\\n */\\npublic static final double MASS_PER_AA = 100.0;;\\n    public static String removeSubString(String string, String subString)\\n    public static boolean sendGAUpdate(String ua, String action, String label)\\n    public static String getVersion()\\n    public static String removeCharacters(String string, char forbiddenCharacter)\\n    public static String removeCharacters(String string, String[] forbiddenCharacters)\\n    public static String removeForbiddenCharacters(String string)\\n    public static boolean containsForbiddenCharacter(String string)\\n    public static String replaceAll(String string, char a, char b)\\n    public static double roundDouble(double d, int places)\\n    public static double floorDouble(double d, int places)\\n    public static double getPpmError(double theoreticalMzValue, double massError)\\n    public static String color2Hex(int colorRGB)\\n    public static String color2Hex(Color color)\\n    public static Color getColor(int colorRGB)\\n    public static String tableToText(JTable table, String separator, ProgressDialogX progressDialog, boolean removeHtml)\\n    public static void tableToFile(JTable table, String separator, ProgressDialogX progressDialog, boolean removeHtml, BufferedWriter writer) throws IOException\\n    public static boolean sameLists(ArrayList<?> list1, ArrayList<?> list2)\\n    public static int getOccurrence(String input, char character)\\n    public static ArrayList<Integer> getIndexes(String bigString, String smallString)\\n    public static double readDoubleAsString(String doubleAsString) throws NumberFormatException\\n    public static int convertBooleanToInteger(Boolean booleanToConvert)\\n    public static String keyValueToString(String key, String value)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/Util.java', 'methodCount': 22, 'lineCount': 29}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class IoUtil {\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default separator for tabular files.\\n */\\npublic static final String DEFAULT_SEPARATOR = \"\\\\t\";;\\n    public static boolean emptyDir(File dir)\\n    public static boolean deleteDir(File dir)\\n    public static void copyFile(File in, File out) throws IOException\\n    public static void append(File in, File out) throws IOException\\n    public static void copyFile(File in, File out, boolean overwrite) throws IOException\\n    public static String getFileName(String filePath)\\n    public static String getFileName(File file)\\n    public static String getExtension(File file)\\n    public static String getExtension(String fileName)\\n    public static String getFilenameExtensionLowerCase(String fileName)\\n    public static File existsExtensionNotCaseSensitive(File file)\\n    public static String appendSuffix(String fileName, String suffix)\\n    public static String removeExtension(String fileName)\\n    public static File saveUrl(File saveFile, String targetUrlAsString, int fileSizeInBytes, String userName, String password, WaitingHandler waitingHandler) throws MalformedURLException, IOException, FileNotFoundException\\n    public static boolean checkIfURLExists(String targetUrlAsString, String userName, String password)\\n    public static int getFileSize(URL url)\\n    public static void closeBuffer(MappedByteBuffer buffer)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/IoUtil.java', 'methodCount': 17, 'lineCount': 21}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface SimpleFileReader {\\n    public static SimpleFileReader getFileReader(File file)\\n    public String readLine()\\n    public int read(char[] buffer)\\n    public int read(char[] buffer, int offset, int length)\\n    public Reader getReader()\\n    public double getProgressInPercent()\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/flat/SimpleFileReader.java', 'methodCount': 7, 'lineCount': 9}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class SimpleFileWriter {\\n    /**\\n * The buffered writer used for writing.\\n */\\nprivate final BufferedWriter bw;;\\n    /**\\n * A simple mutex.\\n */\\nprivate final SimpleSemaphore mutex = new SimpleSemaphore(1);;\\n    public SimpleFileWriter(File file, boolean gz);\\n    public void writeLine(String... elements)\\n    public void writeLine(String line)\\n    public void write(char[] buffer, int offset, int length)\\n    public void newLine()\\n    public void write(String text)\\n    public void write(String text, boolean newLine)\\n    public void close()\\n    public BufferedWriter getWriter()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/flat/SimpleFileWriter.java', 'methodCount': 9, 'lineCount': 13}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class SimpleSemaphore {\\n    /**\\n * The semaphore to use.\\n */\\nprivate final Semaphore mutex;;\\n    public SimpleSemaphore(int nPermits, boolean fair);\\n    public SimpleSemaphore(int nPermits);\\n    public void acquire()\\n    public void release()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/threading/SimpleSemaphore.java', 'methodCount': 4, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Writes content to the given section.\n",
      "     *\n",
      "     * @param sectionName The name of the section.\n",
      "     * @param content The content to write.\n",
      "     */\n",
      "public void write(String sectionName, String content)\n",
      "**************************************\n",
      "public void write(String sectionName, String content) {\n",
      "    SimpleSemaphore sectionSemaphore = semaphoreMap.get(sectionName);\n",
      "    sectionSemaphore.acquire();\n",
      "    SimpleFileWriter tempWriter = tempWriterMap.get(sectionName);\n",
      "    tempWriter.write(content);\n",
      "    sectionSemaphore.release();\n",
      "}\n",
      "1569\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 614/1243 [20:46<19:48,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28256\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  49%|     | 615/1243 [20:48<19:56,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28258\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 616/1243 [20:50<20:45,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28286\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 617/1243 [20:52<19:34,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28297\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 618/1243 [20:53<18:33,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28300\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public interface WaitingHandler {\\n    /**\\n * Convenience date format.\\n */\\npublic static final SimpleDateFormat SIMPLE_DATA_FORMAT = new SimpleDateFormat(\"dd MMM yyyy, HH:mm\");;\\n    /**\\n * The tab space to add when using HTML.\\n */\\npublic static final String TAB_HTML = \"&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;\";;\\n    /**\\n * The tab space to add when not using HTML.\\n */\\npublic static final String TAB_NON_HTML = \"        \";;\\n    public void setPrimaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setMaxPrimaryProgressCounter(int maxProgressValue)\\n    public void increasePrimaryProgressCounter()\\n    public void increasePrimaryProgressCounter(int increment)\\n    public void setPrimaryProgressCounter(int value)\\n    public void setMaxSecondaryProgressCounter(int maxProgressValue)\\n    public void resetPrimaryProgressCounter()\\n    public void resetSecondaryProgressCounter()\\n    public void increaseSecondaryProgressCounter()\\n    public void setSecondaryProgressCounter(int value)\\n    public void increaseSecondaryProgressCounter(int value)\\n    public void setSecondaryProgressCounterIndeterminate(boolean indeterminate)\\n    public void setRunFinished()\\n    public void setRunCanceled()\\n    public void appendReport(String report, boolean includeDate, boolean addNewLine)\\n    public void appendReportNewLineNoDate()\\n    public void appendReportEndLine()\\n    public boolean isRunCanceled()\\n    public boolean isRunFinished()\\n    public void setSecondaryProgressText(String text)\\n    public boolean isReport()\\n    public void setWaitingText(String text)\\n    public int getPrimaryProgressCounter()\\n    public int getMaxPrimaryProgressCounter()\\n    public int getSecondaryProgressCounter()\\n    public int getMaxSecondaryProgressCounter()\\n    public void setDisplayProgress(boolean displayProgress)\\n    public boolean getDisplayProgress()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/waiting/WaitingHandler.java', 'methodCount': 28, 'lineCount': 33}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 16}\n",
      "{}\n",
      "/**\n",
      "     * Returns the size of the file located at the given URL.\n",
      "     *\n",
      "     * @param url the url of the file\n",
      "     *\n",
      "     * @return the size of the file\n",
      "     */\n",
      "public static int getFileSize(URL url)\n",
      "**************************************\n",
      "public static int getFileSize(URL url) {\n",
      "    HttpURLConnection conn = null;\n",
      "    try {\n",
      "        conn = (HttpURLConnection) url.openConnection();\n",
      "        conn.setRequestMethod(\"HEAD\");\n",
      "        conn.getInputStream();\n",
      "        return conn.getContentLength();\n",
      "    } catch (IOException e) {\n",
      "        return -1;\n",
      "    } finally {\n",
      "        conn.disconnect();\n",
      "    }\n",
      "}\n",
      "627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 619/1243 [20:55<19:24,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28322\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 620/1243 [20:58<20:21,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28326\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 621/1243 [20:59<19:37,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28327\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 622/1243 [21:01<19:28,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28330\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 623/1243 [21:03<19:07,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28335\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class IoUtil {\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default separator for tabular files.\\n */\\npublic static final String DEFAULT_SEPARATOR = \"\\\\t\";;\\n    public static boolean emptyDir(File dir)\\n    public static boolean deleteDir(File dir)\\n    public static void copyFile(File in, File out) throws IOException\\n    public static void append(File in, File out) throws IOException\\n    public static void copyFile(File in, File out, boolean overwrite) throws IOException\\n    public static String getFileName(String filePath)\\n    public static String getFileName(File file)\\n    public static String getExtension(File file)\\n    public static String getExtension(String fileName)\\n    public static String getFilenameExtensionLowerCase(String fileName)\\n    public static File existsExtensionNotCaseSensitive(File file)\\n    public static String appendSuffix(String fileName, String suffix)\\n    public static String removeExtension(String fileName)\\n    public static File saveUrl(File saveFile, String targetUrlAsString, int fileSizeInBytes, String userName, String password, WaitingHandler waitingHandler) throws MalformedURLException, IOException, FileNotFoundException\\n    public static boolean checkIfURLExists(String targetUrlAsString, String userName, String password)\\n    public static int getFileSize(URL url)\\n    public static void closeBuffer(MappedByteBuffer buffer)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/IoUtil.java', 'methodCount': 17, 'lineCount': 21}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * Splits the input of comma separated command line input and returns the\n",
      "     * results as an arraylist.\n",
      "     *\n",
      "     * @param cliInput the CLI input\n",
      "     *\n",
      "     * @return an arraylist containing the results, empty list if empty string\n",
      "     */\n",
      "public static ArrayList<String> splitInput(String cliInput)\n",
      "**************************************\n",
      "public static ArrayList<String> splitInput(String cliInput) {\n",
      "    ArrayList<String> results = new ArrayList<>();\n",
      "    if (cliInput == null || cliInput.trim().length() == 0) {\n",
      "        return results;\n",
      "    }\n",
      "    for (String tempInput : cliInput.split(SEPARATOR)) {\n",
      "        results.add(tempInput.trim());\n",
      "    }\n",
      "    return results;\n",
      "}\n",
      "546\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 624/1243 [21:05<18:39,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28339\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 625/1243 [21:06<18:07,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28340\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 626/1243 [21:08<17:53,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28342\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  50%|     | 627/1243 [21:10<17:44,  1.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28344\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 628/1243 [21:11<17:28,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28346\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class IoUtil {\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default separator for tabular files.\\n */\\npublic static final String DEFAULT_SEPARATOR = \"\\\\t\";;\\n    public static boolean emptyDir(File dir)\\n    public static boolean deleteDir(File dir)\\n    public static void copyFile(File in, File out) throws IOException\\n    public static void append(File in, File out) throws IOException\\n    public static void copyFile(File in, File out, boolean overwrite) throws IOException\\n    public static String getFileName(String filePath)\\n    public static String getFileName(File file)\\n    public static String getExtension(File file)\\n    public static String getExtension(String fileName)\\n    public static String getFilenameExtensionLowerCase(String fileName)\\n    public static File existsExtensionNotCaseSensitive(File file)\\n    public static String appendSuffix(String fileName, String suffix)\\n    public static String removeExtension(String fileName)\\n    public static File saveUrl(File saveFile, String targetUrlAsString, int fileSizeInBytes, String userName, String password, WaitingHandler waitingHandler) throws MalformedURLException, IOException, FileNotFoundException\\n    public static boolean checkIfURLExists(String targetUrlAsString, String userName, String password)\\n    public static int getFileSize(URL url)\\n    public static void closeBuffer(MappedByteBuffer buffer)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/IoUtil.java', 'methodCount': 17, 'lineCount': 21}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': \"public class SequenceMatchingParameters {\\n    /**\\n * The amino acid matching type.\\n */\\nprivate MatchingType sequenceMatchingType;;\\n    /**\\n * Limit the share of X's a match can contain, range [0.0-1.0].\\n */\\nprivate double limitX = 0.25;;\\n    /**\\n * Boolean deciding whether tags should only be mapped to enzymatic\\n * peptides.\\n */\\nprivate Boolean enzymaticTagsOnly = false;;\\n    /**\\n * The maximum number of PTMs per peptide when mapping tags.\\n */\\nprivate int maxPtmsPerTagPeptide = 3;;\\n    /**\\n * The minimum amino acid score [0-100]. Used when converting Novor peptides\\n * into tags.\\n */\\nprivate Integer minAminoAcidScore = 30;;\\n    /**\\n * The minimum tag length. Used when converting Novor peptides into tags.\\n */\\nprivate Integer minTagLength = 3;;\\n    /**\\n * Default string matching.\\n */\\npublic static final SequenceMatchingParameters DEFAULT_STRING_MATCHING = getStringMatching();;\\n    public SequenceMatchingParameters();\\n    public static SequenceMatchingParameters getStringMatching()\\n    public static SequenceMatchingParameters getDefaultSequenceMatching()\\n    public MatchingType getSequenceMatchingType()\\n    public void setSequenceMatchingType(MatchingType sequenceMatchingType)\\n    public double getLimitX()\\n    public void setLimitX(double limitX)\\n    public boolean isSameAs(SequenceMatchingParameters sequenceMatchingPreferences)\\n    public String getShortDescription()\\n    public boolean isEnzymaticTagsOnly()\\n    public void setEnzymaticTagsOnly(boolean enzymaticTagsOnly)\\n    public int getMaxPtmsPerTagPeptide()\\n    public void setMaxPtmsPerTagPeptide(int numberOfPtmsPerTagPeptide)\\n    public int getMinAminoAcidScore()\\n    public void setMinAminoAcidScore(int minAminoAcidScore)\\n    public int getMinTagLength()\\n    public void setMinTagLength(int minTagLength)\\n}\\n\", 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/parameters/identification/advanced/SequenceMatchingParameters.java', 'methodCount': 17, 'lineCount': 26}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     * Returns true if the argument can be parsed as a double value.\n",
      "     *\n",
      "     * @param argType the name of the argument\n",
      "     * @param arg the content of the argument\n",
      "     * @return true if the argument can be parsed as a double value\n",
      "     */\n",
      "public static boolean isDouble(String argType, String arg)\n",
      "**************************************\n",
      "public static boolean isDouble(String argType, String arg) {\n",
      "    boolean valid = true;\n",
      "    try {\n",
      "        Double.valueOf(arg);\n",
      "    } catch (NumberFormatException e) {\n",
      "        System.out.println(System.getProperty(\"line.separator\") + \"Error parsing the \" + argType + \" option: Not a floating value!\" + System.getProperty(\"line.separator\"));\n",
      "        valid = false;\n",
      "    } catch (Exception e) {\n",
      "        System.out.println(System.getProperty(\"line.separator\") + \"Error parsing the \" + argType + \" option: \" + e.getLocalizedMessage() + System.getProperty(\"line.separator\"));\n",
      "        valid = false;\n",
      "    }\n",
      "    return valid;\n",
      "}\n",
      "1137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 629/1243 [21:14<18:45,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28355\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 630/1243 [21:15<16:34,  1.62s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28381\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 631/1243 [21:16<16:59,  1.67s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28484\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 632/1243 [21:18<18:07,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28485\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 633/1243 [21:21<19:03,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28487\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class Util {\\n    /**\\n * Forbidden characters in file names.\\n */\\npublic static final String[] FORBIDDEN_CHARACTERS = { \"!\", \":\", \";\", \"\\\\\\\\?\", \"/\", \"\\\\\\\\\\\\\\\\\", \"\\\\\\\\*\", \"<\", \">\", \"\\\\\"\", \"\\'\", \"\\\\\\\\|\" };;\\n    /**\\n * Default encoding, cf the second rule.\\n */\\npublic static final String ENCODING = \"UTF-8\";;\\n    /**\\n * Default column separator for text files.\\n */\\npublic static final String DEFAULT_COLUMN_SEPARATOR = \"\\\\t\";;\\n    /**\\n * The line separator.\\n */\\npublic static final String LINE_SEPARATOR = System.getProperty(\"line.separator\");;\\n    /**\\n * The mass added per amino acid as part of the reference mass when\\n * converting a tolerance in Dalton to ppm.\\n */\\npublic static final double MASS_PER_AA = 100.0;;\\n    public static String removeSubString(String string, String subString)\\n    public static boolean sendGAUpdate(String ua, String action, String label)\\n    public static String getVersion()\\n    public static String removeCharacters(String string, char forbiddenCharacter)\\n    public static String removeCharacters(String string, String[] forbiddenCharacters)\\n    public static String removeForbiddenCharacters(String string)\\n    public static boolean containsForbiddenCharacter(String string)\\n    public static String replaceAll(String string, char a, char b)\\n    public static double roundDouble(double d, int places)\\n    public static double floorDouble(double d, int places)\\n    public static double getPpmError(double theoreticalMzValue, double massError)\\n    public static String color2Hex(int colorRGB)\\n    public static String color2Hex(Color color)\\n    public static Color getColor(int colorRGB)\\n    public static String tableToText(JTable table, String separator, ProgressDialogX progressDialog, boolean removeHtml)\\n    public static void tableToFile(JTable table, String separator, ProgressDialogX progressDialog, boolean removeHtml, BufferedWriter writer) throws IOException\\n    public static boolean sameLists(ArrayList<?> list1, ArrayList<?> list2)\\n    public static int getOccurrence(String input, char character)\\n    public static ArrayList<Integer> getIndexes(String bigString, String smallString)\\n    public static double readDoubleAsString(String doubleAsString) throws NumberFormatException\\n    public static int convertBooleanToInteger(Boolean booleanToConvert)\\n    public static String keyValueToString(String key, String value)\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/Util.java', 'methodCount': 22, 'lineCount': 29}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class SimpleFileWriter {\\n    /**\\n * The buffered writer used for writing.\\n */\\nprivate final BufferedWriter bw;;\\n    /**\\n * A simple mutex.\\n */\\nprivate final SimpleSemaphore mutex = new SimpleSemaphore(1);;\\n    public SimpleFileWriter(File file, boolean gz);\\n    public void writeLine(String... elements)\\n    public void writeLine(String line)\\n    public void write(char[] buffer, int offset, int length)\\n    public void newLine()\\n    public void write(String text)\\n    public void write(String text, boolean newLine)\\n    public void close()\\n    public BufferedWriter getWriter()\\n}\\n', 'filePath': '/home/ubuntu/repos/compomics,compomics-utilities/src/main/java/com/compomics/util/io/flat/SimpleFileWriter.java', 'methodCount': 9, 'lineCount': 13}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Logs an error.\n",
      "     *\n",
      "     * @param message the error message\n",
      "     */\n",
      "public void logError(String message)\n",
      "**************************************\n",
      "public void logError(String message) {\n",
      "    String now = Instant.now().toString();\n",
      "    logWriter.writeLine(now, \"Error\", \"\\\"\" + message.replace(Util.LINE_SEPARATOR, \" \") + \"\\\"\");\n",
      "}\n",
      "902\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 634/1243 [21:23<20:43,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28582\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 635/1243 [21:25<20:59,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28583\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 636/1243 [21:27<21:00,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28606\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|     | 637/1243 [21:30<21:43,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28683\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|    | 638/1243 [21:32<21:59,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28684\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface IPatternFingerprinter {\\n    public abstract double[] getValuesAsArray()\\n    public abstract Collection<IFeature> getFeatures()\\n    public abstract Collection<Double> getValues()\\n    public abstract IFeature getFeature(int index) throws CDKException\\n    public abstract Double getWeight(String pattern)\\n    public abstract Double getWeight(int index)\\n    public abstract int getFingerprintSize()\\n    public abstract int hashCode()\\n    public abstract boolean equals(Object feature)\\n    public abstract String toString()\\n    public abstract BitSet getHashedFingerPrint()\\n    public abstract void addBinary(BitSet fingerprint) throws CDKException\\n    public abstract void add(IPatternFingerprinter fngp) throws CDKException\\n    public abstract String getFingerprintID()\\n    public abstract void setFingerprintID(String fingerprintID)\\n    public abstract int getFeatureCount()\\n    public abstract boolean hasFeature(IFeature key)\\n    public abstract double[] getWeightedHashedFingerPrint()\\n    public abstract void add(IFeature feature) throws CDKException\\n    public IPatternFingerprinter clone() throws CloneNotSupportedException\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/fingerprints/interfaces/IPatternFingerprinter.java', 'methodCount': 20, 'lineCount': 22}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class DirectReactionDrawer {\\n    private static final Logger LOG = getLogger(DirectReactionDrawer.class.getName());;\\n    private AbstractDirectReactionLayout reactionLayout;;\\n    private AbstractAWTReactionLayout exactReactionLayout;;\\n    private DirectMoleculeDrawer moleculeDrawer;;\\n    private DirectArrowDrawer arrowDrawer;;\\n    public DirectReactionDrawer(AbstractDirectReactionLayout layout);\\n    public DirectReactionDrawer(Params params);\\n    public DirectReactionDrawer(Params params, AbstractDirectReactionLayout layout);\\n    public DirectReactionDrawer(Params params, AbstractDirectReactionLayout layout, AbstractAWTReactionLayout exactReactionLayout);\\n    public AbstractDirectReactionLayout getLayout()\\n    private void setupLayout()\\n    public Image drawReaction(IReaction reaction)\\n    public Image drawReaction(IReaction reaction, int w, int h)\\n    public Map<String, String> makeLabelMap(IReaction reaction)\\n    public Image drawReaction(IReaction reaction, int w, int h, boolean invert)\\n    public Image drawReaction(IReaction reaction, boolean invert)\\n    public void drawReaction(IReaction reaction, BoundsTree boundsTree, int w, int h, double zoom, Graphics2D g)\\n    public void drawLabelPanel(Map<String, String> labelMap, BoundsTree labelBoundsTree, Graphics2D g)\\n    public double getMaxLabelHeight(BoundsTree tree, Map<String, String> labels, Graphics2D g)\\n    private void drawReaction(IReaction reaction, Graphics2D g, BoundsTree boundsTree, Vector2d reactionAxis)\\n    private void drawBoundsTree(BoundsTree tree, List<String> labels, Color color, Graphics2D g)\\n    public void printBoundsTree(BoundsTree tree, List<String> labels)\\n    public BoundsTree centerOnOrigin(IReaction reaction, BoundsTree boundsTree)\\n    public BoundsTree centerOn(IReaction reaction, double cx, double cy, BoundsTree boundsTree)\\n    public BoundsTree shift(IReaction reaction, BoundsTree unshiftedTree, double dx, double dy)\\n    private BoundsTree shift(IAtomContainerSet reactants, String rootLabel, BoundsTree unshiftedMolSetTree, String label, double dx, double dy)\\n    private Axis getAxisOfExpansion(int targetWidth, int targetHeight, double zoomedWidth, double zoomedHeight)\\n    public void align(IReaction reaction, int w, int h, double actualWidth, double actualHeight, double zoom)\\n    public void drawReactionID(IReaction reaction, int w, Graphics2D g)\\n    public BoundsTree getReactionBounds(IReaction reaction)\\n    public BoundsTree getExactReactionBounds(IReaction reaction, Graphics2D g)\\n    public double calculateZoom(int targetWidth, int targetHeight, Rectangle2D totalBounds)\\n    private double calculateZoom(int targetWidth, int targetHeight, double actualWidth, double actualHeight)\\n    public Rectangle2D getDrawnBounds(List<IAtom> atoms)\\n    public Vector2d getReactionAxis()\\n    public void drawMappings(IReaction reaction, Graphics2D g)\\n    private Rectangle2D molSetBounds(IAtomContainerSet molSet)\\n    public void drawMoleculeSet(IAtomContainerSet reactants, String reactionID, double yAxis, BoundsTree boundsTree, Graphics2D g)\\n    public void drawPlus(IAtomContainer ac, String acLabel, double yAxis, BoundsTree boundsTree, Graphics2D g)\\n    public Params getParams()\\n    public void highlightSubgraph(List<IAtom> atoms, Color color)\\n    public DirectMoleculeDrawer getMoleculeDrawer()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/graphics/direct/DirectReactionDrawer.java', 'methodCount': 37, 'lineCount': 44}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 6}\n",
      "{'PatternFingerprinter': 'public class PatternFingerprinter {\\n    private static final long serialVersionUID = 0156306561546552043757L;;\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(PatternFingerprinter.class);;\\n    public static IPatternFingerprinter makePatternFingerprint(Map<String, Double> map)\\n    public static IPatternFingerprinter makePatternFingerprint(Collection<String> keyCollection, Collection<Double> valueCollection)\\n    private final Set<IFeature> featureSet;;\\n    private String fingerprintID = \"?\";;\\n    private int fingerprintSize;;\\n    public PatternFingerprinter();\\n    public PatternFingerprinter(Collection<IFeature> features);\\n    public PatternFingerprinter(int fingerprintSize);\\n    public PatternFingerprinter(Collection<IFeature> features, int fingerprintSize);\\n    public synchronized void addBinary(BitSet fingerprint) throws CDKException\\n    public synchronized void add(IFeature feature) throws CDKException\\n    public synchronized void add(IPatternFingerprinter fngp) throws CDKException\\n    public double[] getValuesAsArray()\\n    public Collection<IFeature> getFeatures()\\n    public Collection<Double> getValues()\\n    public int getFeatureCount()\\n    public BitSet getHashedFingerPrint()\\n    public double[] getWeightedHashedFingerPrint()\\n    public String toString()\\n    public IFeature getFeature(int index) throws CDKException\\n    public Double getWeight(String pattern)\\n    public Double getWeight(int index)\\n    public String getFingerprintID()\\n    public void setFingerprintID(String fingerprintID)\\n    public synchronized int compare(IPatternFingerprinter o1, IPatternFingerprinter o2)\\n    public synchronized int compareTo(IPatternFingerprinter t)\\n    public boolean equals(Object object)\\n    public int hashCode()\\n    public int getFingerprintSize()\\n    public boolean hasFeature(IFeature key)\\n    public IPatternFingerprinter clone() throws CloneNotSupportedException\\n}\\n', 'FingerprintGenerator': 'public class FingerprintGenerator {\\n    private final static ILoggingTool LOGGER = createLoggingTool(FingerprintGenerator.class);;\\n    public static int getFingerprinterSize()\\n    final CircularFingerprinter fingerprinter;;\\n    public FingerprintGenerator();\\n    public synchronized BitSet getFingerprint(IAtomContainer mol) throws CDKException\\n}\\n', 'Feature': 'public class Feature {\\n    private static final long serialVersionUID = 0xe6c5aecf276L;;\\n    private final String pattern;;\\n    private double weight;;\\n    public Feature(String feature, double weight);\\n    public Feature(String feature);\\n    public boolean equals(Object obj)\\n    public int hashCode()\\n    public synchronized String toString()\\n    public String getPattern()\\n    public double getWeight()\\n    public int compareTo(IFeature feature)\\n    public int compare(IFeature o1, IFeature o2)\\n    public void setValue(double weight)\\n}\\n'}\n",
      "/**\n",
      "     *\n",
      "     * @param molSet\n",
      "     * @throws CDKException\n",
      "     */\n",
      "private static IPatternFingerprinter getSumOfFingerprints(IAtomContainerSet molSet) throws CDKException, Exception\n",
      "**************************************\n",
      "/**\n",
      " * @param molSet\n",
      " * @throws CDKException\n",
      " */\n",
      "private static IPatternFingerprinter getSumOfFingerprints(IAtomContainerSet molSet) throws CDKException, Exception {\n",
      "    FingerprintGenerator molFingerprint = new FingerprintGenerator();\n",
      "    IPatternFingerprinter fp = new PatternFingerprinter(getFingerprinterSize());\n",
      "    for (IAtomContainer mol : molSet.atomContainers()) {\n",
      "        BitSet booleanArray = molFingerprint.getFingerprint(mol);\n",
      "        for (int i = 0; i < booleanArray.size(); i++) {\n",
      "            if (booleanArray.get(i)) {\n",
      "                fp.add(new Feature(valueOf(i), 1.0));\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return fp;\n",
      "}\n",
      "2034\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|    | 639/1243 [21:34<23:04,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28685\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  51%|    | 640/1243 [21:37<23:32,  2.34s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28688\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 641/1243 [21:39<22:47,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28689\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 642/1243 [21:41<22:03,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28691\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 643/1243 [21:43<21:47,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28692\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param Molecule1\n",
      "     * @param Molecule2\n",
      "     * @return <B>Similarity <U>Cosine,Ochiai,Carbo</U></B>\n",
      "     * <B>c/sqrt(a*b)</B>\n",
      "     * @throws Exception\n",
      "     */\n",
      "public static synchronized double getCosineSimilarity(BitSet Molecule1, BitSet Molecule2) throws Exception\n",
      "**************************************\n",
      "public static synchronized double getCosineSimilarity(BitSet Molecule1, BitSet Molecule2) throws Exception {\n",
      "    BitSet bitset1 = (BitSet) Molecule1.clone();\n",
      "    BitSet bitset2 = (BitSet) Molecule2.clone();\n",
      "    float _bitset1_cardinality = bitset1.cardinality();\n",
      "    float _bitset2_cardinality = bitset2.cardinality();\n",
      "    if (bitset1.size() != bitset2.size()) {\n",
      "        throw new Exception(\"Bisets must have the same bit length\");\n",
      "    }\n",
      "    BitSet one_and_two = (BitSet) bitset1.clone();\n",
      "    one_and_two.and(bitset2);\n",
      "    float _common_bit_count = one_and_two.cardinality();\n",
      "    return _common_bit_count / (sqrt(_bitset1_cardinality * _bitset2_cardinality));\n",
      "}\n",
      "295\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 644/1243 [21:45<22:22,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28693\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 645/1243 [21:48<23:10,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28706\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 646/1243 [21:50<22:22,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28709\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 647/1243 [21:52<21:47,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28711\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 648/1243 [21:55<22:42,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28743\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/stereo/IStereoAndConformation.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface ChiralityTool {\\n    public Map<IAtom, IStereoAndConformation> getTetrahedralChiralities(IAtomContainer atomContainer)\\n    public Map<IAtom, IStereoAndConformation> getTetrahedralChiralities(IAtomContainer atomContainer, boolean getNoneAssignments)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/stereo/compare/ChiralityTool.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class WedgeStereoLifter {\\n    private final List<WedgeRule> rules;;\\n    public WedgeStereoLifter();\\n    private WedgeRule getRule(IBond.Stereo[] stereos)\\n    public IStereoElement lift(IAtom atom, IAtomContainer atomContainer)\\n    private double getFullAngle(IAtom centralAtom, IAtom partner1, IAtom partner2)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/stereo/wedge/WedgeStereoLifter.java', 'methodCount': 4, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param atomContainer\n",
      "     * @param getNoneAssesments\n",
      "     * @return\n",
      "     */\n",
      "public Map<IAtom, IStereoAndConformation> getTetrahedralChiralities(IAtomContainer atomContainer, boolean getNoneAssesments)\n",
      "**************************************\n",
      "@Override\n",
      "public Map<IAtom, IStereoAndConformation> getTetrahedralChiralities(IAtomContainer atomContainer, boolean getNoneAssesments) {\n",
      "    Map<IAtom, IStereoAndConformation> chiralities = new HashMap<>();\n",
      "    WedgeStereoLifter lifter = new WedgeStereoLifter();\n",
      "    for (IAtom atom : atomContainer.atoms()) {\n",
      "        IStereoAndConformation chirality = getChirality2D(lifter, atom, atomContainer);\n",
      "        if (getNoneAssesments || chirality != NONE) {\n",
      "            chiralities.put(atom, chirality);\n",
      "        }\n",
      "    }\n",
      "    return chiralities;\n",
      "}\n",
      "613\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 649/1243 [21:57<22:07,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28746\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 650/1243 [21:58<20:35,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28823\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 651/1243 [22:00<19:34,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28826\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  52%|    | 652/1243 [22:02<18:26,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28827\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 653/1243 [22:04<19:04,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28896\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{'SubgraphAtomSignature': 'public class SubgraphAtomSignature {\\n    private final IAtomContainer atomContainer;;\\n    private Map<Integer, int[]> subgraphAdjacencyLists;;\\n    // XXX - tmp fix to get around the architectural bugs of using adj lists.\\nprivate boolean useAdjLists = false;;\\n    public SubgraphAtomSignature(IAtomContainer atomContainer, int atomIndex);\\n    public SubgraphAtomSignature(IAtomContainer atomContainer, int atomIndex, Map<Integer, int[]> subgraphAdjacencyLists);\\n    protected int[] getConnected(int atomIndex)\\n    protected String getEdgeLabel(int atomIndexA, int atomIndexB)\\n    protected int getIntLabel(int arg0)\\n    protected String getVertexSymbol(int atomIndex)\\n    protected int convertEdgeLabelToColor(String edgeLabel)\\n}\\n'}\n",
      "/**\n",
      "     *\n",
      "     * @param subgraphAtomIndex\n",
      "     * @return\n",
      "     */\n",
      "public AbstractVertexSignature signatureForVertex(int subgraphAtomIndex)\n",
      "**************************************\n",
      "@Override\n",
      "public AbstractVertexSignature signatureForVertex(int subgraphAtomIndex) {\n",
      "    if (useAdjLists) {\n",
      "        int actualIndex = indexLookup.get(subgraphAtomIndex);\n",
      "        return new SubgraphAtomSignature(atomContainer, actualIndex, subgraphAdjacencyLists);\n",
      "    } else {\n",
      "        return new SubgraphAtomSignature(atomContainer, subgraphAtomIndex);\n",
      "    }\n",
      "}\n",
      "325\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 654/1243 [22:06<18:00,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28977\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 655/1243 [22:08<18:23,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28994\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 656/1243 [22:10<19:04,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29001\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 657/1243 [22:12<19:48,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29069\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 658/1243 [22:14<19:46,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29070\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ReactionFileData {\\n    private static final long serialVersionUID = 193790837047304701L;;\\n    private List<String> _data = null;;\\n    public ReactionFileData();\\n    public boolean addData(String data)\\n    public void addData(int index, String data)\\n    public boolean isPresent(String data)\\n    public String getValue(int index)\\n    public Collection<String> getValues()\\n    public boolean isEmpty()\\n    public int getSize()\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mapping/container/helper/ReactionFileData.java', 'methodCount': 9, 'lineCount': 13}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 6}\n",
      "{'ReactionFileData': 'public class ReactionFileData {\\n    private static final long serialVersionUID = 193790837047304701L;;\\n    private List<String> _data = null;;\\n    public ReactionFileData();\\n    public boolean addData(String data)\\n    public void addData(int index, String data)\\n    public boolean isPresent(String data)\\n    public String getValue(int index)\\n    public Collection<String> getValues()\\n    public boolean isEmpty()\\n    public int getSize()\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     *\n",
      "     * @param _recID\n",
      "     * @return entry ID of the reaction\n",
      "     */\n",
      "public synchronized String getEntry(String _recID)\n",
      "**************************************\n",
      "public synchronized String getEntry(String _recID) {\n",
      "    String _data = null;\n",
      "    if (reactionMap.containsKey(_recID)) {\n",
      "        Map<String, ReactionFileData> DataMap = reactionMap.get(_recID);\n",
      "        if (DataMap.containsKey(entry)) {\n",
      "            _data = DataMap.get(entry).getValue(0);\n",
      "        }\n",
      "    }\n",
      "    return _data.trim();\n",
      "}\n",
      "447\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 659/1243 [22:15<18:28,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29072\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 660/1243 [22:18<19:08,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29074\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 661/1243 [22:20<19:11,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29077\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 662/1243 [22:22<19:51,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29215\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 663/1243 [22:24<19:51,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29216\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 20, 'classCount': 1, 'methodsOnly': 'public class Holder {\\n    private static final boolean DEBUG = false;;\\n    private static final long serialVersionUID = 18989786786L;;\\n    private final static ILoggingTool LOGGER = createLoggingTool(Holder.class);;\\n    /*\\n     Final methods\\n     */\\nprivate final List<MolMapping> mappingMolPair;;\\n    private final EBIMatrix stereoMatrix;;\\n    private final EBIMatrix cliqueMatrix;;\\n    private final EBIMatrix graphSimilarityMatrix;;\\n    private final EBIMatrix fragmentMatrix;;\\n    private final EBIMatrix energyMatrix;;\\n    private final EBIMatrix carbonOverlapMatrix;;\\n    private final EBIMatrix fpSimMatrixWithoutHydrogen;;\\n    private final int row;;\\n    private final int coloumn;;\\n    /*\\n     Local methods\\n     */\\nprivate ReactionContainer structureInformation;;\\n    private BestMatch bestMatchContainer;;\\n    private List<String> eductCounter;;\\n    private List<String> productCounter;;\\n    private String reactionID;;\\n    private HydrogenFreeFingerPrintContainer hydFPFree;;\\n    private IMappingAlgorithm theory;;\\n    public Holder(IMappingAlgorithm theory, String reactionID, List<String> eductCounter, List<String> productCounter, ReactionContainer reactionContainer, BestMatch bestMatchContainer, HydrogenFreeFingerPrintContainer hydFPFree) throws IOException;\\n    public Holder(int row, int column);\\n    public synchronized EBIMatrix getStereoMatrix()\\n    public synchronized EBIMatrix getCliqueMatrix()\\n    public synchronized EBIMatrix getGraphSimilarityMatrix()\\n    public synchronized EBIMatrix getFragmentMatrix()\\n    public synchronized EBIMatrix getEnergyMatrix()\\n    private void initialize()\\n    private void setFingerprint()\\n    private void setMolMapping()\\n    public synchronized EBIMatrix getFPSimilarityMatrix()\\n    public synchronized ReactionContainer getReactionContainer()\\n    public synchronized List<MolMapping> getMappingMolPair()\\n    public synchronized Object clone() throws CloneNotSupportedException\\n    private void setData(double[][] sourceData, EBIMatrix sinkMatrix, int rows, int coloumns)\\n    public synchronized BestMatch getBestMatchContainer()\\n    public IMappingAlgorithm getTheory()\\n    public void setTheory(IMappingAlgorithm theory)\\n    public EBIMatrix getCarbonOverlapMatrix()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mapping/algorithm/Holder.java', 'methodCount': 19, 'lineCount': 41}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class ReactionContainer {\\n    static final long serialVersionUID = 17278639972837695L;;\\n    /*\\n     * Singleton Pattern Implementation\\n     */\\nprivate final Map<Integer, IAtomContainer> eAtomContainerMap;;\\n    private final Map<Integer, IAtomContainer> pAtomContainerMap;;\\n    private final Map<Integer, BitSet> eFingerPrintMap;;\\n    private final Map<Integer, BitSet> pFingerPrintMap;;\\n    private final Map<Integer, Boolean> eductContainerModificationMap;;\\n    private final Map<Integer, Boolean> productContainerModificationMap;;\\n    private final IFingerprintGenerator fpr;;\\n    public ReactionContainer() throws Exception;\\n    public String toString()\\n    public synchronized void Clear() throws IOException\\n    public synchronized void eraseEduct(int Key) throws IOException\\n    public synchronized void eraseProduct(int Key) throws IOException\\n    public synchronized Integer getEductCount()\\n    public synchronized Integer getProductCount()\\n    public synchronized IAtomContainer getEduct(int value) throws IOException, CDKException\\n    public synchronized IAtomContainer getProduct(int value) throws IOException, CDKException\\n    public synchronized Collection<IAtomContainer> getEducts() throws IOException, CDKException\\n    public synchronized Collection<IAtomContainer> getProducts() throws IOException, CDKException\\n    public synchronized boolean isEductModified(int value) throws IOException, CDKException\\n    public synchronized boolean isProductModified(int value) throws IOException, CDKException\\n    public synchronized void putEduct(int index, IAtomContainer educt) throws IOException, Exception\\n    public synchronized void putProduct(int index, IAtomContainer product) throws IOException, Exception\\n    public synchronized void setEductModified(int index, boolean flag) throws IOException, Exception\\n    public synchronized void setProductModified(int index, boolean flag) throws IOException, Exception\\n    public synchronized void putAllEduct(TreeMap<Integer, IAtomContainer> eudMap) throws IOException, Exception\\n    public synchronized void putAllProduct(TreeMap<Integer, IAtomContainer> prodMap) throws IOException, Exception\\n    public synchronized BitSet getFingerPrintofEduct(int value) throws IOException, CDKException\\n    public synchronized BitSet getFingerPrintofProduct(int value) throws IOException, CDKException\\n    private synchronized void setFingerPrintofEduct(int value, BitSet edu) throws IOException\\n    private synchronized void setFingerPrintofProduct(int value, BitSet prod) throws IOException\\n    public synchronized Object clone() throws CloneNotSupportedException\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mapping/container/ReactionContainer.java', 'methodCount': 24, 'lineCount': 34}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class EBIMatrix {\\n    static final String NEW_LINE = getProperty(\"line.separator\");;\\n    private static final long serialVersionUID = 19787786981017786L;;\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(EBIMatrix.class);;\\n    public static List<Double> elimination(EBIMatrix matrix, List<Double> vector)\\n    public static EBIMatrix random(int m, int n)\\n    public static EBIMatrix identity(int m, int n)\\n    public static EBIMatrix read(BufferedReader input) throws java.io.IOException\\n    public static EBIMatrix constructWithCopy(double[][] A)\\n    private double[][] matrix;;\\n    private int rows, columns;;\\n    public EBIMatrix(int rows, int columns);\\n    public EBIMatrix(double[][] A);\\n    public EBIMatrix(double[][] A, int m, int n);\\n    public EBIMatrix(int m, int n, double s);\\n    public EBIMatrix(double[] vals, int m);\\n    public synchronized void initMatrix(double v)\\n    public synchronized double getValue(int i, int j)\\n    public synchronized EBIMatrix duplicate()\\n    public synchronized Object clone() throws CloneNotSupportedException\\n    public synchronized double[][] getArrayCopy()\\n    public synchronized double[] getColumnPackedCopy()\\n    public synchronized double[] getRowPackedCopy()\\n    public synchronized double[][] getArray()\\n    public synchronized EBIMatrix getMatrix(int rowStart, int rowEnd, int colStart, int colEnd)\\n    public synchronized EBIMatrix transpose()\\n    public synchronized boolean setValue(int row, int col, double value)\\n    public synchronized int getRowDimension()\\n    public synchronized int getColumnDimension()\\n    public synchronized List<Double> getDiagonalElements()\\n    public synchronized EBIMatrix getMatrix(int[] r, int[] c)\\n    public synchronized EBIMatrix getMatrix(int rowStart, int rowEnd, int[] c)\\n    public synchronized EBIMatrix getMatrix(int[] r, int colStart, int colEnd)\\n    public synchronized void set(int i, int j, double s)\\n    public synchronized void setMatrix(int rowStart, int rowEnd, int colStart, int colEnd, EBIMatrix X)\\n    public synchronized void setMatrix(int[] r, int[] c, EBIMatrix X)\\n    public synchronized void setMatrix(int[] r, int colStart, int colEnd, EBIMatrix X)\\n    public synchronized void setMatrix(int rowStart, int rowEnd, int[] c, EBIMatrix X)\\n    public synchronized boolean is_element_max_in_column(int iPos, int jPos)\\n    public synchronized boolean is_element_min_in_column(int iPos, int jPos)\\n    public synchronized void reSizeMatrix(int RowSize, int ColSize)\\n    public synchronized boolean is_element_max_in_row(int iPos, int jPos)\\n    public synchronized boolean is_element_min_in_row(int iPos, int jPos)\\n    public synchronized void swapColumns(int coloumn1, int coloumn2)\\n    public synchronized void swapRows(int row1, int row2) throws CDKException\\n    public synchronized void pivot(int row, int col)\\n    public synchronized EBIMatrix normalize(EBIMatrix S)\\n    public synchronized EBIMatrix mul(double a)\\n    public synchronized List<Double> mul(List<Double> a)\\n    public synchronized EBIMatrix mul(EBIMatrix b)\\n    private synchronized void checkMatrixDimensions(EBIMatrix B)\\n    public synchronized EBIMatrix arrayTimes(EBIMatrix B)\\n    public synchronized EBIMatrix arrayTimesEquals(EBIMatrix B)\\n    public synchronized EBIMatrix arrayRightDivide(EBIMatrix B)\\n    public synchronized EBIMatrix arrayRightDivideEquals(EBIMatrix B)\\n    public synchronized EBIMatrix arrayLeftDivide(EBIMatrix B)\\n    public synchronized EBIMatrix arrayLeftDivideEquals(EBIMatrix B)\\n    public synchronized EBIMatrix times(double s)\\n    public synchronized EBIMatrix timesEquals(double s)\\n    public synchronized EBIMatrix times(EBIMatrix B)\\n    public synchronized LUDecomposition lu()\\n    public synchronized QRDecomposition qr()\\n    public synchronized CholeskyDecomposition chol()\\n    public synchronized SingularValueDecomposition svd()\\n    public synchronized EigenvalueDecomposition eig()\\n    public synchronized EBIMatrix solve(EBIMatrix B)\\n    public synchronized EBIMatrix solveTranspose(EBIMatrix B)\\n    public synchronized EBIMatrix inverse()\\n    public synchronized double det()\\n    public synchronized int rank()\\n    public synchronized double cond()\\n    public synchronized double trace()\\n    public synchronized EBIMatrix diagonalize(int nrot)\\n    public synchronized EBIMatrix orthonormalize(EBIMatrix S)\\n    public synchronized void print(NumberFormat format, int width)\\n    public synchronized void print(int w, int d)\\n    public synchronized void print(PrintWriter output, int w, int d)\\n    public synchronized void print(PrintWriter output, NumberFormat format, int width)\\n    public synchronized String toString()\\n    public double contraction()\\n    public EBIMatrix similar(EBIMatrix U)\\n    public synchronized double norm1()\\n    public synchronized double norm2()\\n    public synchronized double normInf()\\n    public synchronized double normF()\\n    public synchronized EBIMatrix uminus()\\n    public synchronized EBIMatrix plus(EBIMatrix B)\\n    public synchronized EBIMatrix plusEquals(EBIMatrix B)\\n    public synchronized EBIMatrix minus(EBIMatrix B)\\n    public synchronized EBIMatrix minusEquals(EBIMatrix B)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/EBIMatrix.java', 'methodCount': 84, 'lineCount': 91}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param mh\n",
      "     * @param IndexI\n",
      "     * @param IndexJ\n",
      "     * @return\n",
      "     * @throws IOException\n",
      "     * @throws CDKException\n",
      "     */\n",
      "protected synchronized boolean isMajorSubgraphRow(Holder mh, int IndexI, int IndexJ) throws IOException, CDKException\n",
      "**************************************\n",
      "protected synchronized boolean isMajorSubgraphRow(Holder mh, int IndexI, int IndexJ) throws IOException, CDKException {\n",
      "    double queryRowScore = mh.getCliqueMatrix().getValue(IndexI, IndexJ);\n",
      "    if (queryRowScore > 0) {\n",
      "        for (int row = 0; row < rowSize; row++) {\n",
      "            if (flagMatrix[row][IndexJ] && row != IndexI) {\n",
      "                int eSize = mh.getReactionContainer().getEduct(row).getAtomCount();\n",
      "                double rowRSize = mh.getCliqueMatrix().getValue(row, IndexJ);\n",
      "                if (queryRowScore < rowRSize) {\n",
      "                    return false;\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "2751\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 664/1243 [22:26<20:03,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29217\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  53%|    | 665/1243 [22:28<20:38,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29219\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 666/1243 [22:31<20:45,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29220\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 667/1243 [22:33<21:22,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29221\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 668/1243 [22:35<21:03,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29222\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 20, 'classCount': 1, 'methodsOnly': 'public class Holder {\\n    private static final boolean DEBUG = false;;\\n    private static final long serialVersionUID = 18989786786L;;\\n    private final static ILoggingTool LOGGER = createLoggingTool(Holder.class);;\\n    /*\\n     Final methods\\n     */\\nprivate final List<MolMapping> mappingMolPair;;\\n    private final EBIMatrix stereoMatrix;;\\n    private final EBIMatrix cliqueMatrix;;\\n    private final EBIMatrix graphSimilarityMatrix;;\\n    private final EBIMatrix fragmentMatrix;;\\n    private final EBIMatrix energyMatrix;;\\n    private final EBIMatrix carbonOverlapMatrix;;\\n    private final EBIMatrix fpSimMatrixWithoutHydrogen;;\\n    private final int row;;\\n    private final int coloumn;;\\n    /*\\n     Local methods\\n     */\\nprivate ReactionContainer structureInformation;;\\n    private BestMatch bestMatchContainer;;\\n    private List<String> eductCounter;;\\n    private List<String> productCounter;;\\n    private String reactionID;;\\n    private HydrogenFreeFingerPrintContainer hydFPFree;;\\n    private IMappingAlgorithm theory;;\\n    public Holder(IMappingAlgorithm theory, String reactionID, List<String> eductCounter, List<String> productCounter, ReactionContainer reactionContainer, BestMatch bestMatchContainer, HydrogenFreeFingerPrintContainer hydFPFree) throws IOException;\\n    public Holder(int row, int column);\\n    public synchronized EBIMatrix getStereoMatrix()\\n    public synchronized EBIMatrix getCliqueMatrix()\\n    public synchronized EBIMatrix getGraphSimilarityMatrix()\\n    public synchronized EBIMatrix getFragmentMatrix()\\n    public synchronized EBIMatrix getEnergyMatrix()\\n    private void initialize()\\n    private void setFingerprint()\\n    private void setMolMapping()\\n    public synchronized EBIMatrix getFPSimilarityMatrix()\\n    public synchronized ReactionContainer getReactionContainer()\\n    public synchronized List<MolMapping> getMappingMolPair()\\n    public synchronized Object clone() throws CloneNotSupportedException\\n    private void setData(double[][] sourceData, EBIMatrix sinkMatrix, int rows, int coloumns)\\n    public synchronized BestMatch getBestMatchContainer()\\n    public IMappingAlgorithm getTheory()\\n    public void setTheory(IMappingAlgorithm theory)\\n    public EBIMatrix getCarbonOverlapMatrix()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mapping/algorithm/Holder.java', 'methodCount': 19, 'lineCount': 41}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class ReactionContainer {\\n    static final long serialVersionUID = 17278639972837695L;;\\n    /*\\n     * Singleton Pattern Implementation\\n     */\\nprivate final Map<Integer, IAtomContainer> eAtomContainerMap;;\\n    private final Map<Integer, IAtomContainer> pAtomContainerMap;;\\n    private final Map<Integer, BitSet> eFingerPrintMap;;\\n    private final Map<Integer, BitSet> pFingerPrintMap;;\\n    private final Map<Integer, Boolean> eductContainerModificationMap;;\\n    private final Map<Integer, Boolean> productContainerModificationMap;;\\n    private final IFingerprintGenerator fpr;;\\n    public ReactionContainer() throws Exception;\\n    public String toString()\\n    public synchronized void Clear() throws IOException\\n    public synchronized void eraseEduct(int Key) throws IOException\\n    public synchronized void eraseProduct(int Key) throws IOException\\n    public synchronized Integer getEductCount()\\n    public synchronized Integer getProductCount()\\n    public synchronized IAtomContainer getEduct(int value) throws IOException, CDKException\\n    public synchronized IAtomContainer getProduct(int value) throws IOException, CDKException\\n    public synchronized Collection<IAtomContainer> getEducts() throws IOException, CDKException\\n    public synchronized Collection<IAtomContainer> getProducts() throws IOException, CDKException\\n    public synchronized boolean isEductModified(int value) throws IOException, CDKException\\n    public synchronized boolean isProductModified(int value) throws IOException, CDKException\\n    public synchronized void putEduct(int index, IAtomContainer educt) throws IOException, Exception\\n    public synchronized void putProduct(int index, IAtomContainer product) throws IOException, Exception\\n    public synchronized void setEductModified(int index, boolean flag) throws IOException, Exception\\n    public synchronized void setProductModified(int index, boolean flag) throws IOException, Exception\\n    public synchronized void putAllEduct(TreeMap<Integer, IAtomContainer> eudMap) throws IOException, Exception\\n    public synchronized void putAllProduct(TreeMap<Integer, IAtomContainer> prodMap) throws IOException, Exception\\n    public synchronized BitSet getFingerPrintofEduct(int value) throws IOException, CDKException\\n    public synchronized BitSet getFingerPrintofProduct(int value) throws IOException, CDKException\\n    private synchronized void setFingerPrintofEduct(int value, BitSet edu) throws IOException\\n    private synchronized void setFingerPrintofProduct(int value, BitSet prod) throws IOException\\n    public synchronized Object clone() throws CloneNotSupportedException\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mapping/container/ReactionContainer.java', 'methodCount': 24, 'lineCount': 34}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class EBIMatrix {\\n    static final String NEW_LINE = getProperty(\"line.separator\");;\\n    private static final long serialVersionUID = 19787786981017786L;;\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(EBIMatrix.class);;\\n    public static List<Double> elimination(EBIMatrix matrix, List<Double> vector)\\n    public static EBIMatrix random(int m, int n)\\n    public static EBIMatrix identity(int m, int n)\\n    public static EBIMatrix read(BufferedReader input) throws java.io.IOException\\n    public static EBIMatrix constructWithCopy(double[][] A)\\n    private double[][] matrix;;\\n    private int rows, columns;;\\n    public EBIMatrix(int rows, int columns);\\n    public EBIMatrix(double[][] A);\\n    public EBIMatrix(double[][] A, int m, int n);\\n    public EBIMatrix(int m, int n, double s);\\n    public EBIMatrix(double[] vals, int m);\\n    public synchronized void initMatrix(double v)\\n    public synchronized double getValue(int i, int j)\\n    public synchronized EBIMatrix duplicate()\\n    public synchronized Object clone() throws CloneNotSupportedException\\n    public synchronized double[][] getArrayCopy()\\n    public synchronized double[] getColumnPackedCopy()\\n    public synchronized double[] getRowPackedCopy()\\n    public synchronized double[][] getArray()\\n    public synchronized EBIMatrix getMatrix(int rowStart, int rowEnd, int colStart, int colEnd)\\n    public synchronized EBIMatrix transpose()\\n    public synchronized boolean setValue(int row, int col, double value)\\n    public synchronized int getRowDimension()\\n    public synchronized int getColumnDimension()\\n    public synchronized List<Double> getDiagonalElements()\\n    public synchronized EBIMatrix getMatrix(int[] r, int[] c)\\n    public synchronized EBIMatrix getMatrix(int rowStart, int rowEnd, int[] c)\\n    public synchronized EBIMatrix getMatrix(int[] r, int colStart, int colEnd)\\n    public synchronized void set(int i, int j, double s)\\n    public synchronized void setMatrix(int rowStart, int rowEnd, int colStart, int colEnd, EBIMatrix X)\\n    public synchronized void setMatrix(int[] r, int[] c, EBIMatrix X)\\n    public synchronized void setMatrix(int[] r, int colStart, int colEnd, EBIMatrix X)\\n    public synchronized void setMatrix(int rowStart, int rowEnd, int[] c, EBIMatrix X)\\n    public synchronized boolean is_element_max_in_column(int iPos, int jPos)\\n    public synchronized boolean is_element_min_in_column(int iPos, int jPos)\\n    public synchronized void reSizeMatrix(int RowSize, int ColSize)\\n    public synchronized boolean is_element_max_in_row(int iPos, int jPos)\\n    public synchronized boolean is_element_min_in_row(int iPos, int jPos)\\n    public synchronized void swapColumns(int coloumn1, int coloumn2)\\n    public synchronized void swapRows(int row1, int row2) throws CDKException\\n    public synchronized void pivot(int row, int col)\\n    public synchronized EBIMatrix normalize(EBIMatrix S)\\n    public synchronized EBIMatrix mul(double a)\\n    public synchronized List<Double> mul(List<Double> a)\\n    public synchronized EBIMatrix mul(EBIMatrix b)\\n    private synchronized void checkMatrixDimensions(EBIMatrix B)\\n    public synchronized EBIMatrix arrayTimes(EBIMatrix B)\\n    public synchronized EBIMatrix arrayTimesEquals(EBIMatrix B)\\n    public synchronized EBIMatrix arrayRightDivide(EBIMatrix B)\\n    public synchronized EBIMatrix arrayRightDivideEquals(EBIMatrix B)\\n    public synchronized EBIMatrix arrayLeftDivide(EBIMatrix B)\\n    public synchronized EBIMatrix arrayLeftDivideEquals(EBIMatrix B)\\n    public synchronized EBIMatrix times(double s)\\n    public synchronized EBIMatrix timesEquals(double s)\\n    public synchronized EBIMatrix times(EBIMatrix B)\\n    public synchronized LUDecomposition lu()\\n    public synchronized QRDecomposition qr()\\n    public synchronized CholeskyDecomposition chol()\\n    public synchronized SingularValueDecomposition svd()\\n    public synchronized EigenvalueDecomposition eig()\\n    public synchronized EBIMatrix solve(EBIMatrix B)\\n    public synchronized EBIMatrix solveTranspose(EBIMatrix B)\\n    public synchronized EBIMatrix inverse()\\n    public synchronized double det()\\n    public synchronized int rank()\\n    public synchronized double cond()\\n    public synchronized double trace()\\n    public synchronized EBIMatrix diagonalize(int nrot)\\n    public synchronized EBIMatrix orthonormalize(EBIMatrix S)\\n    public synchronized void print(NumberFormat format, int width)\\n    public synchronized void print(int w, int d)\\n    public synchronized void print(PrintWriter output, int w, int d)\\n    public synchronized void print(PrintWriter output, NumberFormat format, int width)\\n    public synchronized String toString()\\n    public double contraction()\\n    public EBIMatrix similar(EBIMatrix U)\\n    public synchronized double norm1()\\n    public synchronized double norm2()\\n    public synchronized double normInf()\\n    public synchronized double normF()\\n    public synchronized EBIMatrix uminus()\\n    public synchronized EBIMatrix plus(EBIMatrix B)\\n    public synchronized EBIMatrix plusEquals(EBIMatrix B)\\n    public synchronized EBIMatrix minus(EBIMatrix B)\\n    public synchronized EBIMatrix minusEquals(EBIMatrix B)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/EBIMatrix.java', 'methodCount': 84, 'lineCount': 91}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param similarityMatrix\n",
      "     * @param IndexI\n",
      "     * @param IndexJ\n",
      "     * @return\n",
      "     */\n",
      "protected synchronized boolean isMinorSubgraphRow(EBIMatrix similarityMatrix, int IndexI, int IndexJ)\n",
      "**************************************\n",
      "protected synchronized boolean isMinorSubgraphRow(EBIMatrix similarityMatrix, int IndexI, int IndexJ) {\n",
      "    boolean flag = true;\n",
      "    double queryRowScore = similarityMatrix.getValue(IndexI, IndexJ);\n",
      "    if (queryRowScore > 0.) {\n",
      "        for (int row = 0; row < rowSize; row++) {\n",
      "            if (flagMatrix[row][IndexJ] && row != IndexI) {\n",
      "                double targetRowScore = similarityMatrix.getValue(row, IndexJ);\n",
      "                if (queryRowScore > targetRowScore && targetRowScore > 0.) {\n",
      "                    return false;\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return flag;\n",
      "}\n",
      "2719\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 669/1243 [22:37<20:55,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29223\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 670/1243 [22:39<20:36,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29224\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 671/1243 [22:41<20:15,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29236\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 672/1243 [22:43<18:29,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29335\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 673/1243 [22:45<18:43,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29350\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface ICanonicalMoleculeLabeller {\\n    public IAtomContainer getCanonicalMolecule(IAtomContainer container)\\n    public int[] getCanonicalPermutation(IAtomContainer container)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/labelling/ICanonicalMoleculeLabeller.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'RBlastMoleculeSignature': 'public class RBlastMoleculeSignature {\\n    private boolean useAromatics = true;;\\n    private boolean useCharge = true;;\\n    private boolean isBondSensitive = true;;\\n    private RBlastMoleculeFromSignatureBuilder builder;;\\n    public RBlastMoleculeSignature(IAtomContainer atomContainer);\\n    public int getAtomCount()\\n    public boolean isUseAromatics()\\n    public void setUseAromatics(boolean useAromatics)\\n    public boolean isUseCharge()\\n    public void setUseCharge(boolean useCharge)\\n    public boolean isBondSensitive()\\n    public void setBondSensitive(boolean isBondSensitive)\\n    public String getMoleculeCanonicalSignatureString()\\n    public String getSignatureStringForAtom(int atomIndex)\\n    public String getSignatureStringForAtom(int atomIndex, int height)\\n    public RBlastAtomSignature getAtomSignature(int atomIndex)\\n    public RBlastAtomSignature getAtomSignature(int atomIndex, int height)\\n    private void setFlags(RBlastAtomSignature atomSignature)\\n    private void setFlags(RBlastMoleculeFromSignatureBuilder builder)\\n    public String getSmilesForAtomSignature(int atomIndex, int height)\\n    public String getSmilesForAtomSignature(int atomIndex, int height, ICanonicalMoleculeLabeller labeller)\\n    public IAtomContainer getFragment(int atomIndex, int height)\\n    public IAtomContainer makeMoleculeFromSignature(String signatureString)\\n}\\n'}\n",
      "/**\n",
      "     *\n",
      "     * @param container\n",
      "     * @return\n",
      "     */\n",
      "public int[] getCanonicalPermutation(IAtomContainer container)\n",
      "**************************************\n",
      "@Override\n",
      "public int[] getCanonicalPermutation(IAtomContainer container) {\n",
      "    RBlastMoleculeSignature molSig = new RBlastMoleculeSignature(container);\n",
      "    if (height == -1) {\n",
      "        return molSig.getAtomSignature(atomIndex).getCanonicalLabelling(container.getAtomCount());\n",
      "    } else {\n",
      "        return molSig.getAtomSignature(atomIndex, height).getCanonicalLabelling(container.getAtomCount());\n",
      "    }\n",
      "}\n",
      "589\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 674/1243 [22:46<17:34,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29362\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 675/1243 [22:48<16:52,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29373\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 676/1243 [22:51<19:12,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29418\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  54%|    | 677/1243 [22:53<18:57,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29444\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 678/1243 [22:55<18:57,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29530\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface IPatternFingerprinter {\\n    public abstract double[] getValuesAsArray()\\n    public abstract Collection<IFeature> getFeatures()\\n    public abstract Collection<Double> getValues()\\n    public abstract IFeature getFeature(int index) throws CDKException\\n    public abstract Double getWeight(String pattern)\\n    public abstract Double getWeight(int index)\\n    public abstract int getFingerprintSize()\\n    public abstract int hashCode()\\n    public abstract boolean equals(Object feature)\\n    public abstract String toString()\\n    public abstract BitSet getHashedFingerPrint()\\n    public abstract void addBinary(BitSet fingerprint) throws CDKException\\n    public abstract void add(IPatternFingerprinter fngp) throws CDKException\\n    public abstract String getFingerprintID()\\n    public abstract void setFingerprintID(String fingerprintID)\\n    public abstract int getFeatureCount()\\n    public abstract boolean hasFeature(IFeature key)\\n    public abstract double[] getWeightedHashedFingerPrint()\\n    public abstract void add(IFeature feature) throws CDKException\\n    public IPatternFingerprinter clone() throws CloneNotSupportedException\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/fingerprints/interfaces/IPatternFingerprinter.java', 'methodCount': 20, 'lineCount': 22}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param query\n",
      "     * @param target @Equation Unweighted comparison of patterns\n",
      "     * @return returns true if query fingerprint is a subset of target\n",
      "     * @throws CDKException\n",
      "     */\n",
      "public static boolean isSubset(double[] query, double[] target) throws CDKException\n",
      "**************************************\n",
      "public static boolean isSubset(double[] query, double[] target) throws CDKException {\n",
      "    if (query.length != target.length) {\n",
      "        throw new CDKException(\"Unequal Fingerprint size can't be processed\");\n",
      "    }\n",
      "    for (int i = 0; i < query.length; i++) {\n",
      "        if (query[i] > target[i]) {\n",
      "            return false;\n",
      "        }\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "502\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 679/1243 [22:57<20:07,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29531\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 680/1243 [22:59<19:58,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29532\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 681/1243 [23:01<19:46,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29541\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 682/1243 [23:05<23:41,  2.53s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 683/1243 [23:07<22:50,  2.45s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29547\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class EBIMatrix {\\n    static final String NEW_LINE = getProperty(\"line.separator\");;\\n    private static final long serialVersionUID = 19787786981017786L;;\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(EBIMatrix.class);;\\n    public static List<Double> elimination(EBIMatrix matrix, List<Double> vector)\\n    public static EBIMatrix random(int m, int n)\\n    public static EBIMatrix identity(int m, int n)\\n    public static EBIMatrix read(BufferedReader input) throws java.io.IOException\\n    public static EBIMatrix constructWithCopy(double[][] A)\\n    private double[][] matrix;;\\n    private int rows, columns;;\\n    public EBIMatrix(int rows, int columns);\\n    public EBIMatrix(double[][] A);\\n    public EBIMatrix(double[][] A, int m, int n);\\n    public EBIMatrix(int m, int n, double s);\\n    public EBIMatrix(double[] vals, int m);\\n    public synchronized void initMatrix(double v)\\n    public synchronized double getValue(int i, int j)\\n    public synchronized EBIMatrix duplicate()\\n    public synchronized Object clone() throws CloneNotSupportedException\\n    public synchronized double[][] getArrayCopy()\\n    public synchronized double[] getColumnPackedCopy()\\n    public synchronized double[] getRowPackedCopy()\\n    public synchronized double[][] getArray()\\n    public synchronized EBIMatrix getMatrix(int rowStart, int rowEnd, int colStart, int colEnd)\\n    public synchronized EBIMatrix transpose()\\n    public synchronized boolean setValue(int row, int col, double value)\\n    public synchronized int getRowDimension()\\n    public synchronized int getColumnDimension()\\n    public synchronized List<Double> getDiagonalElements()\\n    public synchronized EBIMatrix getMatrix(int[] r, int[] c)\\n    public synchronized EBIMatrix getMatrix(int rowStart, int rowEnd, int[] c)\\n    public synchronized EBIMatrix getMatrix(int[] r, int colStart, int colEnd)\\n    public synchronized void set(int i, int j, double s)\\n    public synchronized void setMatrix(int rowStart, int rowEnd, int colStart, int colEnd, EBIMatrix X)\\n    public synchronized void setMatrix(int[] r, int[] c, EBIMatrix X)\\n    public synchronized void setMatrix(int[] r, int colStart, int colEnd, EBIMatrix X)\\n    public synchronized void setMatrix(int rowStart, int rowEnd, int[] c, EBIMatrix X)\\n    public synchronized boolean is_element_max_in_column(int iPos, int jPos)\\n    public synchronized boolean is_element_min_in_column(int iPos, int jPos)\\n    public synchronized void reSizeMatrix(int RowSize, int ColSize)\\n    public synchronized boolean is_element_max_in_row(int iPos, int jPos)\\n    public synchronized boolean is_element_min_in_row(int iPos, int jPos)\\n    public synchronized void swapColumns(int coloumn1, int coloumn2)\\n    public synchronized void swapRows(int row1, int row2) throws CDKException\\n    public synchronized void pivot(int row, int col)\\n    public synchronized EBIMatrix normalize(EBIMatrix S)\\n    public synchronized EBIMatrix mul(double a)\\n    public synchronized List<Double> mul(List<Double> a)\\n    public synchronized EBIMatrix mul(EBIMatrix b)\\n    private synchronized void checkMatrixDimensions(EBIMatrix B)\\n    public synchronized EBIMatrix arrayTimes(EBIMatrix B)\\n    public synchronized EBIMatrix arrayTimesEquals(EBIMatrix B)\\n    public synchronized EBIMatrix arrayRightDivide(EBIMatrix B)\\n    public synchronized EBIMatrix arrayRightDivideEquals(EBIMatrix B)\\n    public synchronized EBIMatrix arrayLeftDivide(EBIMatrix B)\\n    public synchronized EBIMatrix arrayLeftDivideEquals(EBIMatrix B)\\n    public synchronized EBIMatrix times(double s)\\n    public synchronized EBIMatrix timesEquals(double s)\\n    public synchronized EBIMatrix times(EBIMatrix B)\\n    public synchronized LUDecomposition lu()\\n    public synchronized QRDecomposition qr()\\n    public synchronized CholeskyDecomposition chol()\\n    public synchronized SingularValueDecomposition svd()\\n    public synchronized EigenvalueDecomposition eig()\\n    public synchronized EBIMatrix solve(EBIMatrix B)\\n    public synchronized EBIMatrix solveTranspose(EBIMatrix B)\\n    public synchronized EBIMatrix inverse()\\n    public synchronized double det()\\n    public synchronized int rank()\\n    public synchronized double cond()\\n    public synchronized double trace()\\n    public synchronized EBIMatrix diagonalize(int nrot)\\n    public synchronized EBIMatrix orthonormalize(EBIMatrix S)\\n    public synchronized void print(NumberFormat format, int width)\\n    public synchronized void print(int w, int d)\\n    public synchronized void print(PrintWriter output, int w, int d)\\n    public synchronized void print(PrintWriter output, NumberFormat format, int width)\\n    public synchronized String toString()\\n    public double contraction()\\n    public EBIMatrix similar(EBIMatrix U)\\n    public synchronized double norm1()\\n    public synchronized double norm2()\\n    public synchronized double normInf()\\n    public synchronized double normF()\\n    public synchronized EBIMatrix uminus()\\n    public synchronized EBIMatrix plus(EBIMatrix B)\\n    public synchronized EBIMatrix plusEquals(EBIMatrix B)\\n    public synchronized EBIMatrix minus(EBIMatrix B)\\n    public synchronized EBIMatrix minusEquals(EBIMatrix B)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/EBIMatrix.java', 'methodCount': 84, 'lineCount': 91}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Return upper triangular factor\n",
      "     *\n",
      "     * @return U\n",
      "     */\n",
      "public EBIMatrix getU()\n",
      "**************************************\n",
      "public EBIMatrix getU() {\n",
      "    EBIMatrix X = new EBIMatrix(n, n);\n",
      "    double[][] U = X.getArray();\n",
      "    for (int i = 0; i < n; i++) {\n",
      "        for (int j = 0; j < n; j++) {\n",
      "            if (i <= j) {\n",
      "                U[i][j] = LU[i][j];\n",
      "            } else {\n",
      "                U[i][j] = 0.0;\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return X;\n",
      "}\n",
      "1468\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 684/1243 [23:09<21:32,  2.31s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29565\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 685/1243 [23:11<21:03,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29583\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 686/1243 [23:13<20:28,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29584\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 687/1243 [23:15<19:59,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29592\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 688/1243 [23:17<19:47,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29595\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 9}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param reaction\n",
      "     * @return\n",
      "     */\n",
      "protected Map<IChemObject, Integer> makeIndexMap(IReaction reaction)\n",
      "**************************************\n",
      "protected Map<IChemObject, Integer> makeIndexMap(IReaction reaction) {\n",
      "    Map<IChemObject, Integer> indexMap = new HashMap<>();\n",
      "    List<IAtomContainer> all = getAllAtomContainers(reaction);\n",
      "    int globalIndex = 0;\n",
      "    for (IAtomContainer ac : all) {\n",
      "        for (IAtom atom : ac.atoms()) {\n",
      "            indexMap.put(atom, globalIndex);\n",
      "            globalIndex++;\n",
      "        }\n",
      "    }\n",
      "    return indexMap;\n",
      "}\n",
      "166\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  55%|    | 689/1243 [23:19<19:29,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29601\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 690/1243 [23:21<19:01,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29614\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 691/1243 [23:24<20:39,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29619\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 692/1243 [23:26<20:50,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29620\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 693/1243 [23:29<20:20,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29621\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MoleculeInitializer {\\n    /**\\n * Prepare the molecule for analysis.\\n * <p/>\\n * We perform ring perception and aromaticity detection and set up the\\n * appropriate properties. Right now, this function is called each time we\\n * need to do a query and this is inefficient.\\n *\\n * @param atomContainer Atom container where rings are to be marked\\n * @throws CDKException if there is a problem in ring perception or\\n * aromaticity detection, which is usually related to a timeout in the ring\\n * finding code.\\n */\\nprivate static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(MoleculeInitializer.class);;\\n    public static synchronized void initializeMolecule(IAtomContainer atomContainer) throws CDKException\\n    public static boolean testIsSubgraphHeuristics(IAtomContainer q, IAtomContainer t)\\n    public static synchronized boolean testIsSubgraphHeuristics(IAtomContainer ac1, IAtomContainer ac2, AtomMatcher am, BondMatcher bm)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/helper/MoleculeInitializer.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ExtAtomContainerManipulator {\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(ExtAtomContainerManipulator.class);;\\n    static final long serialVersionUID = 1786786539472837495L;;\\n    public static IAtomContainer removeHydrogensExceptSingleAndPreserveAtomID(IAtomContainer org)\\n    public static IAtomContainer copyAndSuppressedHydrogens(IAtomContainer org)\\n    private static void printAtoms(IAtomContainer mol)\\n    public static synchronized IAtomContainer checkAndCleanMolecule(IAtomContainer molecule_orignal)\\n    public static void aromatizeMolecule(IAtomContainer mol)\\n    public static IAtomContainer cloneWithIDs(IAtomContainer container) throws CloneNotSupportedException\\n    public static IAtomContainer newInstanceWithIDs(IAtomContainer container) throws CloneNotSupportedException\\n    public static int getExplicitHydrogenCount(IAtomContainer atomContainer, IAtom atom)\\n    public static int getImplicitHydrogenCount(IAtom atom)\\n    public static int getHydrogenCount(IAtomContainer atomContainer, IAtom atom)\\n    public static IAtomContainer convertExplicitToImplicitHydrogens(IAtomContainer atomContainer)\\n    public static void percieveAtomTypesAndConfigureAtoms(IAtomContainer container) throws CDKException\\n    public static IAtomContainer addExplicitH(IAtomContainer gMol) throws CloneNotSupportedException\\n    public static void aromatizeDayLight(IAtomContainer molecule) throws CDKException\\n    public static void aromatizeCDK(IAtomContainer molecule) throws CDKException\\n    public static void fixDativeBonds(IAtomContainer mol)\\n    public static void setNullHCountToZero(IAtomContainer mol)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/tools/ExtAtomContainerManipulator.java', 'methodCount': 17, 'lineCount': 21}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface IMolDescriptors {\\n     double getAPolDescriptor()\\n     double getCPSADescriptor() throws CDKException\\n     int getCovalentCount()\\n     int getEccentricConnectivityIndexDescriptor()\\n     double getFragmentComplexityDescriptor()\\n     int getHBondAcceptors(boolean checkAromaticity) throws CDKException\\n     int getHBondDoners(boolean checkAromaticity) throws CDKException\\n     int getHeavyAtomCount()\\n     int getLargestPiSystemDescriptor(boolean checkAromaticity) throws CDKException\\n     double getMolecularWeight() throws CDKException\\n     double getPetitjeanNumberDescriptor()\\n     int getRotatableBondsCountDescriptor(boolean includeTerminals, boolean excludeAmides) throws CDKException\\n     double getTPSADescriptor(boolean checkAromaticity) throws CDKException\\n     double getTotalCharge()\\n     double getVAdjMaDescriptor()\\n     double getXlogP(boolean checkAromaticity) throws CDKException\\n     double getZagrebIndexDescriptor()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/interfaces/IMolDescriptors.java', 'methodCount': 17, 'lineCount': 19}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param checkAromaticity\n",
      "     * @return\n",
      "     */\n",
      "public double getXlogP(boolean checkAromaticity)\n",
      "**************************************\n",
      "@Override\n",
      "public double getXlogP(boolean checkAromaticity) {\n",
      "    IMolecularDescriptor xlogP = new XLogPDescriptor();\n",
      "    Object[] xlogPparams = { checkAromaticity, TRUE };\n",
      "    try {\n",
      "        xlogP.setParameters(xlogPparams);\n",
      "    } catch (CDKException ex) {\n",
      "        LOGGER.error(SEVERE, null, ex);\n",
      "    }\n",
      "    double xlogPvalue = ((DoubleResult) xlogP.calculate(molecule).getValue()).doubleValue();\n",
      "    return xlogPvalue;\n",
      "}\n",
      "1184\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 694/1243 [23:30<19:30,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29622\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 695/1243 [23:33<19:25,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29629\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 696/1243 [23:35<19:19,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29630\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 697/1243 [23:37<19:32,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29631\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 698/1243 [23:39<18:11,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29650\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param map\n",
      "     * @return\n",
      "     */\n",
      "public static synchronized Map<Object, Double> valueInAscendingOrder(Map<Object, Double> map)\n",
      "**************************************\n",
      "/**\n",
      " * @param map\n",
      " * @return\n",
      " */\n",
      "public static synchronized Map<Object, Double> valueInAscendingOrder(Map<Object, Double> map) {\n",
      "    List<Map.Entry<Object, Double>> list = new LinkedList<>(map.entrySet());\n",
      "    // Sort the list using an annonymous inner class implementing Comparator for the compare method\n",
      "    // Return 0 for bond match, -1 for less than and +1 for more then (Aceending Order Sort)\n",
      "    sort(// Return 0 for bond match, -1 for less than and +1 for more then (Aceending Order Sort)\n",
      "    list, // Return 0 for bond match, -1 for less than and +1 for more then (Aceending Order Sort)\n",
      "    (Map.Entry<Object, Double> entry, Map.Entry<Object, Double> entry1) -> (entry.getValue().equals(entry1.getValue()) ? 0 : (entry.getValue() > entry1.getValue() ? 1 : -1)));\n",
      "    // LOGGER.info(list);\n",
      "    Map<Object, Double> result = new LinkedHashMap<>();\n",
      "    list.forEach((entry) -> {\n",
      "        result.put(entry.getKey(), entry.getValue());\n",
      "    });\n",
      "    return result;\n",
      "}\n",
      "309\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 699/1243 [23:40<16:54,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29651\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 700/1243 [23:42<17:27,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29652\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 701/1243 [23:44<17:46,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29653\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  56%|    | 702/1243 [23:46<16:55,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29654\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 703/1243 [23:48<16:23,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29655\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param map\n",
      "     * @return\n",
      "     */\n",
      "public static synchronized Map<Double, Double> doubleValueInDescendingOrder(Map<Double, Double> map)\n",
      "**************************************\n",
      "public static synchronized Map<Double, Double> doubleValueInDescendingOrder(Map<Double, Double> map) {\n",
      "    List<Map.Entry<Double, Double>> list = new LinkedList<>(map.entrySet());\n",
      "    sort(list, (Map.Entry<Double, Double> entry, Map.Entry<Double, Double> entry1) -> (entry.getValue().equals(entry1.getValue()) ? 0 : (entry.getValue() < entry1.getValue() ? 1 : -1)));\n",
      "    Map<Double, Double> result = new LinkedHashMap<>();\n",
      "    list.forEach((entry) -> {\n",
      "        result.put(entry.getKey(), entry.getValue());\n",
      "    });\n",
      "    return result;\n",
      "}\n",
      "197\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 704/1243 [23:50<16:59,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29662\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 705/1243 [23:52<17:38,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29713\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 706/1243 [23:54<18:36,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29715\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 707/1243 [23:56<18:32,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29747\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 708/1243 [23:58<17:31,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29757\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 19}\n",
      "{}\n",
      "/**\n",
      "     * Create a new chem model for a single {@link IAtomContainer}.\n",
      "     *\n",
      "     * @param container the container to create the model for\n",
      "     * @return a new {@link IChemModel}\n",
      "     */\n",
      "private static IChemModel newModel(final IAtomContainer container)\n",
      "**************************************\n",
      "private static IChemModel newModel(final IAtomContainer container) {\n",
      "    if (container == null) {\n",
      "        throw new NullPointerException(\"cannot create chem model for a null container\");\n",
      "    }\n",
      "    final IChemObjectBuilder builder = container.getBuilder();\n",
      "    final IChemModel model = builder.newInstance(IChemModel.class);\n",
      "    final IAtomContainerSet containers = builder.newInstance(IAtomContainerSet.class);\n",
      "    containers.addAtomContainer(container);\n",
      "    model.setMoleculeSet(containers);\n",
      "    return model;\n",
      "}\n",
      "197\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 709/1243 [24:00<17:42,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29786\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 710/1243 [24:02<18:08,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29800\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 711/1243 [24:05<19:27,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29815\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 712/1243 [24:07<19:57,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29864\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 713/1243 [24:09<18:10,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29866\n",
      "{'classCount': 5, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class CholeskyDecomposition {\\n    private static final long serialVersionUID = 78619981017L;;\\n    /* ------------------------\\n     Class variables\\n     * ------------------------ */\\n/**\\n * Array for internal storage of decomposition.\\n *\\n * @serial internal array storage.\\n */\\nprivate final double[][] L;;\\n    /**\\n * Row and column dimension (square matrix).\\n *\\n * @serial matrix dimension.\\n */\\nprivate final int n;;\\n    /**\\n * Symmetric and positive definite flag.\\n *\\n * @serial is symmetric and positive definite flag.\\n */\\nprivate boolean isspd;;\\n    public CholeskyDecomposition(EBIMatrix Arg);\\n    public boolean isSPD()\\n    public EBIMatrix getL()\\n    public EBIMatrix solve(EBIMatrix B)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/matrix/CholeskyDecomposition.java', 'methodCount': 4, 'lineCount': 10}, {'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class EigenvalueDecomposition {\\n    private static final long serialVersionUID = 17869981017L;;\\n    /* ------------------------\\n     Class variables\\n     * ------------------------ */\\n/**\\n * Row and column dimension (square matrix).\\n *\\n * @serial matrix dimension.\\n */\\nprivate final int n;;\\n    /**\\n * Symmetry flag.\\n *\\n * @serial internal symmetry flag.\\n */\\nprivate boolean issymmetric;;\\n    /**\\n * Arrays for internal storage of eigenvalues.\\n *\\n * @serial internal storage of eigenvalues.\\n */\\nprivate final double[] d, e;;\\n    /**\\n * Array for internal storage of eigenvectors.\\n *\\n * @serial internal storage of eigenvectors.\\n */\\nprivate final double[][] V;;\\n    /**\\n * Array for internal storage of nonsymmetric Hessenberg form.\\n *\\n * @serial internal storage of nonsymmetric Hessenberg form.\\n */\\nprivate double[][] H;;\\n    /**\\n * Working storage for nonsymmetric algorithm.\\n *\\n * @serial working storage for nonsymmetric algorithm.\\n */\\nprivate double[] ort;;\\n    /* ------------------------\\n     Private Methods\\n     * ------------------------ */\\nprivate transient double cdivr, cdivi;;\\n    public EigenvalueDecomposition(EBIMatrix Arg);\\n    private void tred2()\\n    private void tql2()\\n    private void orthes()\\n    private void cdiv(double xr, double xi, double yr, double yi)\\n    private void hqr2()\\n    public EBIMatrix getV()\\n    public double[] getRealEigenvalues()\\n    public double[] getImagEigenvalues()\\n    public EBIMatrix getD()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/matrix/EigenvalueDecomposition.java', 'methodCount': 10, 'lineCount': 20}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class LUDecomposition {\\n    private static final long serialVersionUID = 19978681017L;;\\n    /* ------------------------\\n     Class variables\\n     * ------------------------ */\\n/**\\n * Array for internal storage of decomposition.\\n *\\n * @serial internal array storage.\\n */\\nprivate final double[][] LU;;\\n    /**\\n * Row and column dimensions, and pivot sign.\\n *\\n * @serial column dimension.\\n * @serial row dimension.\\n * @serial pivot sign.\\n */\\nprivate final int m, n;;\\n    private int pivsign;;\\n    /**\\n * Internal storage of pivot vector.\\n *\\n * @serial pivot vector.\\n */\\nprivate final int[] piv;;\\n    public LUDecomposition(EBIMatrix A);\\n    public boolean isNonsingular()\\n    public EBIMatrix getL()\\n    public EBIMatrix getU()\\n    public int[] getPivot()\\n    public double[] getDoublePivot()\\n    public double det()\\n    public EBIMatrix solve(EBIMatrix B)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/matrix/LUDecomposition.java', 'methodCount': 8, 'lineCount': 15}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class QRDecomposition {\\n    private static final long serialVersionUID = 199810878617L;;\\n    private final static ILoggingTool LOGGER = createLoggingTool(QRDecomposition.class);;\\n    /* ------------------------\\n     Class variables\\n     * ------------------------ */\\n/**\\n * Array for internal storage of decomposition.\\n *\\n * @serial internal array storage.\\n */\\nprivate final double[][] QR;;\\n    /**\\n * Row and column dimensions.\\n *\\n * @serial column dimension.\\n * @serial row dimension.\\n */\\nprivate final int m, n;;\\n    /**\\n * Array for internal storage of diagonal of R.\\n *\\n * @serial diagonal of R.\\n */\\nprivate final double[] Rdiag;;\\n    public QRDecomposition(EBIMatrix A);\\n    public boolean isFullRank()\\n    public EBIMatrix getH()\\n    public EBIMatrix getR()\\n    public EBIMatrix getQ()\\n    public EBIMatrix solve(EBIMatrix B)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/matrix/QRDecomposition.java', 'methodCount': 6, 'lineCount': 13}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class SingularValueDecomposition {\\n    private static final long serialVersionUID = 19981017786L;;\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(SingularValueDecomposition.class);;\\n    /* ------------------------\\n     Class variables\\n     * ------------------------ */\\n/**\\n * Arrays for internal storage of U and V.\\n *\\n * @serial internal storage of U.\\n * @serial internal storage of V.\\n */\\nprivate double[][] U, V;;\\n    /**\\n * Array for internal storage of singular values.\\n *\\n * @serial internal storage of singular values.\\n */\\nprivate double[] s;;\\n    /**\\n * Row and column dimensions.\\n *\\n * @serial row dimension.\\n * @serial column dimension.\\n */\\nprivate int m, n;;\\n    public SingularValueDecomposition(EBIMatrix Arg);\\n    public EBIMatrix getU()\\n    public EBIMatrix getV()\\n    public double[] getSingularValues()\\n    public EBIMatrix getS()\\n    public double norm2()\\n    public double cond()\\n    public int rank()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/matrix/SingularValueDecomposition.java', 'methodCount': 8, 'lineCount': 15}], 'asteriskCount': 0, 'importCount': 5, 'javaImportCount': 22}\n",
      "{}\n",
      "/**\n",
      "     * One norm\n",
      "     *\n",
      "     * @return maximum column sum.\n",
      "     */\n",
      "public synchronized double norm1()\n",
      "**************************************\n",
      "public synchronized double norm1() {\n",
      "    double f = 0;\n",
      "    for (int j = 0; j < columns; j++) {\n",
      "        double s = 0;\n",
      "        for (int i = 0; i < rows; i++) {\n",
      "            s += abs(matrix[i][j]);\n",
      "        }\n",
      "        f = max(f, s);\n",
      "    }\n",
      "    return f;\n",
      "}\n",
      "1804\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  57%|    | 714/1243 [24:11<18:17,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29868\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 715/1243 [24:13<18:23,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29875\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 716/1243 [24:15<19:41,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29878\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 717/1243 [24:17<18:30,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29879\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 718/1243 [24:19<18:32,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29884\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param Molecule1 BitSet\n",
      "     * @param Molecule2 BitSet\n",
      "     * <B>Note:</B> Can be Normalized to the range 1 to 0 if all attributes are\n",
      "     * normalized to this range and the result is divided in N\n",
      "     * @return <B>Distance (Dissimilarity) <U>Euclidean</U></B>\n",
      "     * <B>sqrt(a+b-2c)</B>\n",
      "     * <B>sqrt(OnA+OnB-2*c)</B>\n",
      "     * @throws java.lang.Exception\n",
      "     */\n",
      "public static double getEuclideanDistance(BitSet Molecule1, BitSet Molecule2) throws Exception\n",
      "**************************************\n",
      "/**\n",
      " * @param Molecule1 BitSet\n",
      " * @param Molecule2 BitSet\n",
      " * <B>Note:</B> Can be Normalized to the range 1 to 0 if all attributes are\n",
      " * normalized to this range and the result is divided in N\n",
      " * @return <B>Distance (Dissimilarity) <U>Euclidean</U></B>\n",
      " * <B>sqrt(a+b-2c)</B>\n",
      " * <B>sqrt(OnA+OnB-2*c)</B>\n",
      " * @throws java.lang.Exception\n",
      " */\n",
      "public static double getEuclideanDistance(BitSet Molecule1, BitSet Molecule2) throws Exception {\n",
      "    A = (BitSet) Molecule1.clone();\n",
      "    B = (BitSet) Molecule2.clone();\n",
      "    float _bitset1_cardinality = A.cardinality();\n",
      "    float _bitset2_cardinality = B.cardinality();\n",
      "    if (A.size() != B.size()) {\n",
      "        throw new Exception(\"Bisets must have the same bit length\");\n",
      "    }\n",
      "    BitSet one_and_two = (BitSet) A.clone();\n",
      "    one_and_two.and(B);\n",
      "    float _common_bit_count = one_and_two.cardinality();\n",
      "    double Val = sqrt((_bitset1_cardinality + _bitset2_cardinality - _common_bit_count) - _common_bit_count);\n",
      "    return Val;\n",
      "}\n",
      "462\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 719/1243 [24:22<18:38,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29891\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 720/1243 [24:24<18:22,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29901\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 721/1243 [24:26<18:18,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29907\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 722/1243 [24:28<18:09,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29908\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 723/1243 [24:30<18:06,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29943\n",
      "{'classCount': 5, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MoleculeInitializer {\\n    /**\\n * Prepare the molecule for analysis.\\n * <p/>\\n * We perform ring perception and aromaticity detection and set up the\\n * appropriate properties. Right now, this function is called each time we\\n * need to do a query and this is inefficient.\\n *\\n * @param atomContainer Atom container where rings are to be marked\\n * @throws CDKException if there is a problem in ring perception or\\n * aromaticity detection, which is usually related to a timeout in the ring\\n * finding code.\\n */\\nprivate static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(MoleculeInitializer.class);;\\n    public static synchronized void initializeMolecule(IAtomContainer atomContainer) throws CDKException\\n    public static boolean testIsSubgraphHeuristics(IAtomContainer q, IAtomContainer t)\\n    public static synchronized boolean testIsSubgraphHeuristics(IAtomContainer ac1, IAtomContainer ac2, AtomMatcher am, BondMatcher bm)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/helper/MoleculeInitializer.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class AtomAtomMappingContainer {\\n    private static final String NEW_LINE = System.getProperty(\"line.separator\");;\\n    private static final long serialVersionUID = 17879096958755L;;\\n    private List<IAtom> reactantAtomArray = new ArrayList<>();;\\n    private List<IAtom> productAtomArray = new ArrayList<>();;\\n    public AtomAtomMappingContainer(Reactor reactor, boolean withoutH) throws Exception;\\n    public AtomAtomMappingContainer(IAtomContainerSet reactants, IAtomContainerSet products, boolean withoutH);\\n    public AtomAtomMappingContainer(IReaction reaction, boolean withoutH);\\n    public synchronized String toString()\\n    public synchronized IAtom getMappedProductAtom(IAtom reactantAtom)\\n    public synchronized IAtom getReactantAtom(int idx)\\n    public synchronized IAtom getProductAtom(int idx)\\n    public synchronized int getSize()\\n    public synchronized int getSizeNoHydrogens()\\n    public synchronized boolean isReactantAtomPresent(IAtom atom)\\n    public synchronized boolean isProductAtomPresent(IAtom atom)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mechanism/helper/AtomAtomMappingContainer.java', 'methodCount': 11, 'lineCount': 17}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class AtomStereoChangeInformation {\\n    private static final long serialVersionUID = 1896986585959789L;;\\n    private final IAtom reactantAtom;;\\n    private final IAtom productAtom;;\\n    private boolean stereoChange = false;;\\n    private IStereoAndConformation atomStereoR;;\\n    private IStereoAndConformation atomStereoP;;\\n    public AtomStereoChangeInformation(IAtom rAtom, IAtom pAtom);\\n    public AtomStereoChangeInformation(IAtom atomE, IAtom atomP, IStereoAndConformation aStereoR, IStereoAndConformation aStereoP);\\n    public synchronized IAtom getReactantAtom()\\n    public synchronized IAtom getProductAtom()\\n    public synchronized boolean isStereoChange()\\n    private synchronized void setStereoChange(boolean stereoChange)\\n    public IStereoAndConformation getReactantAtomStereo()\\n    public IStereoAndConformation getProductAtomStereo()\\n    public void setAtomStereo(IStereoAndConformation atomStereoR, IStereoAndConformation atomStereoP)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mechanism/helper/AtomStereoChangeInformation.java', 'methodCount': 9, 'lineCount': 17}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class BondChange {\\n    private static final String NEW_LINE = System.getProperty(\"line.separator\");;\\n    private static final long serialVersionUID = 9890766688070991L;;\\n    public static synchronized int convertBondOrder(IBond bond)\\n    public static synchronized int convertBondStereo(IBond bond)\\n    private final IBond reactantBond;;\\n    private final IBond productBond;;\\n    private float bondChangeDelta;;\\n    public BondChange(IBond reactantBond, IBond productBond);\\n    public synchronized IBond getReactantBond()\\n    public synchronized IBond getProductBond()\\n    public synchronized float getBondChangeDelta()\\n    private synchronized void setBondChangeInformation(float bondChangeInformation)\\n    public synchronized String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mechanism/helper/BondChange.java', 'methodCount': 8, 'lineCount': 15}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface IChangeCalculator {\\n    public BEMatrix getEductBEMatrix()\\n    public BEMatrix getProductBEMatrix()\\n    public RMatrix getRMatrix()\\n    public void printBMatrix()\\n    public void printEMatrix()\\n    public void printRMatrix()\\n    public void writeBMatrix(File outputFile)\\n    public void writeEMatrix(File outputFile)\\n    public void writeRMatrix(File outputFile)\\n    public boolean hasRMatrix()\\n    public Map<IAtom, IAtom> getMappingMap()\\n    public List<BondChange> getBondChangeList()\\n    public Collection<IAtom> getReactionCenterSet()\\n    public List<AtomStereoChangeInformation> getStereoChangeList()\\n    public Iterable<AtomStereoChangeInformation> getConformationChangeList()\\n    public AtomAtomMappingContainer getMappingContainer()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mechanism/interfaces/IChangeCalculator.java', 'methodCount': 16, 'lineCount': 18}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/stereo/IStereoAndConformation.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class StereoCenteralityTool {\\n    private final static ILoggingTool LOGGER = createLoggingTool(StereoCenteralityTool.class);;\\n    private static final long serialVersionUID = 17867606807697859L;;\\n    private static IAtom getAtomByID(String id, IAtomContainer ac)\\n    public static Map<IAtom, IStereoAndConformation> getChirality2D(IReaction reaction) throws CDKException, CloneNotSupportedException\\n    public static Map<IAtom, IStereoAndConformation> getChirality2D(IAtomContainer ac, CDKPerceptor perceptor)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/stereo/ebi/StereoCenteralityTool.java', 'methodCount': 3, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 5, 'javaImportCount': 11}\n",
      "{}\n",
      "/**\n",
      "     * @param mappings to be set\n",
      "     */\n",
      "private void setMappingMap(Iterable<IMapping> mappings)\n",
      "**************************************\n",
      "/**\n",
      " * @param mappings to be set\n",
      " */\n",
      "private void setMappingMap(Iterable<IMapping> mappings) {\n",
      "    Iterator<IMapping> mappingIterator = mappings.iterator();\n",
      "    while (mappingIterator.hasNext()) {\n",
      "        IMapping mappingObject = mappingIterator.next();\n",
      "        IAtom atomEduct = (IAtom) mappingObject.getChemObject(0);\n",
      "        IAtom atomProduct = (IAtom) mappingObject.getChemObject(1);\n",
      "        mappingMap.put(atomEduct, atomProduct);\n",
      "    }\n",
      "}\n",
      "1819\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 724/1243 [24:32<18:30,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29968\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 725/1243 [24:34<19:01,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29978\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 726/1243 [24:37<18:36,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29989\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  58%|    | 727/1243 [24:38<17:21,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "29992\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 728/1243 [24:40<17:50,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30003\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class AtomAtomMapping {\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(AtomAtomMapping.class);;\\n    private static final long serialVersionUID = 1223637237262778L;;\\n    private final IAtomContainer query;;\\n    private final IAtomContainer target;;\\n    private final Map<IAtom, IAtom> mapping;;\\n    private final Map<Integer, Integer> mappingIndex;;\\n    public boolean equals(Object obj)\\n    public int hashCode()\\n    public AtomAtomMapping(IAtomContainer query, IAtomContainer target);\\n    public synchronized void put(IAtom atom1, IAtom atom2)\\n    public synchronized String toString()\\n    public synchronized boolean isEmpty()\\n    public synchronized void clear()\\n    public synchronized int getCount()\\n    public synchronized Map<IAtom, IAtom> getMappingsByAtoms()\\n    public synchronized Map<Integer, Integer> getMappingsByIndex()\\n    public synchronized int getQueryIndex(IAtom atom)\\n    public synchronized int getTargetIndex(IAtom atom)\\n    public synchronized IAtomContainer getQuery()\\n    public synchronized IAtomContainer getTarget()\\n    public synchronized IAtomContainer getMapCommonFragmentOnQuery() throws CloneNotSupportedException\\n    public synchronized IAtomContainer getMapCommonFragmentOnTarget() throws CloneNotSupportedException\\n    public synchronized IAtomContainer getCommonFragment() throws CloneNotSupportedException\\n    public synchronized String getCommonFragmentAsSMILES() throws CloneNotSupportedException, CDKException\\n    public Map<IAtom, IAtom> sortByValues(Map<IAtom, IAtom> map)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/AtomAtomMapping.java', 'methodCount': 19, 'lineCount': 27}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MoleculeInitializer {\\n    /**\\n * Prepare the molecule for analysis.\\n * <p/>\\n * We perform ring perception and aromaticity detection and set up the\\n * appropriate properties. Right now, this function is called each time we\\n * need to do a query and this is inefficient.\\n *\\n * @param atomContainer Atom container where rings are to be marked\\n * @throws CDKException if there is a problem in ring perception or\\n * aromaticity detection, which is usually related to a timeout in the ring\\n * finding code.\\n */\\nprivate static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(MoleculeInitializer.class);;\\n    public static synchronized void initializeMolecule(IAtomContainer atomContainer) throws CDKException\\n    public static boolean testIsSubgraphHeuristics(IAtomContainer q, IAtomContainer t)\\n    public static synchronized boolean testIsSubgraphHeuristics(IAtomContainer ac1, IAtomContainer ac2, AtomMatcher am, BondMatcher bm)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/helper/MoleculeInitializer.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class Feature {\\n    private static final long serialVersionUID = 0xe6c5aecf276L;;\\n    private final String pattern;;\\n    private double weight;;\\n    public Feature(String feature, double weight);\\n    public Feature(String feature);\\n    public boolean equals(Object obj)\\n    public int hashCode()\\n    public synchronized String toString()\\n    public String getPattern()\\n    public double getWeight()\\n    public int compareTo(IFeature feature)\\n    public int compare(IFeature o1, IFeature o2)\\n    public void setValue(double weight)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/fingerprints/Feature.java', 'methodCount': 10, 'lineCount': 15}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class PatternFingerprinter {\\n    private static final long serialVersionUID = 0156306561546552043757L;;\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(PatternFingerprinter.class);;\\n    public static IPatternFingerprinter makePatternFingerprint(Map<String, Double> map)\\n    public static IPatternFingerprinter makePatternFingerprint(Collection<String> keyCollection, Collection<Double> valueCollection)\\n    private final Set<IFeature> featureSet;;\\n    private String fingerprintID = \"?\";;\\n    private int fingerprintSize;;\\n    public PatternFingerprinter();\\n    public PatternFingerprinter(Collection<IFeature> features);\\n    public PatternFingerprinter(int fingerprintSize);\\n    public PatternFingerprinter(Collection<IFeature> features, int fingerprintSize);\\n    public synchronized void addBinary(BitSet fingerprint) throws CDKException\\n    public synchronized void add(IFeature feature) throws CDKException\\n    public synchronized void add(IPatternFingerprinter fngp) throws CDKException\\n    public double[] getValuesAsArray()\\n    public Collection<IFeature> getFeatures()\\n    public Collection<Double> getValues()\\n    public int getFeatureCount()\\n    public BitSet getHashedFingerPrint()\\n    public double[] getWeightedHashedFingerPrint()\\n    public String toString()\\n    public IFeature getFeature(int index) throws CDKException\\n    public Double getWeight(String pattern)\\n    public Double getWeight(int index)\\n    public String getFingerprintID()\\n    public void setFingerprintID(String fingerprintID)\\n    public synchronized int compare(IPatternFingerprinter o1, IPatternFingerprinter o2)\\n    public synchronized int compareTo(IPatternFingerprinter t)\\n    public boolean equals(Object object)\\n    public int hashCode()\\n    public int getFingerprintSize()\\n    public boolean hasFeature(IFeature key)\\n    public IPatternFingerprinter clone() throws CloneNotSupportedException\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/fingerprints/PatternFingerprinter.java', 'methodCount': 28, 'lineCount': 35}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface IPatternFingerprinter {\\n    public abstract double[] getValuesAsArray()\\n    public abstract Collection<IFeature> getFeatures()\\n    public abstract Collection<Double> getValues()\\n    public abstract IFeature getFeature(int index) throws CDKException\\n    public abstract Double getWeight(String pattern)\\n    public abstract Double getWeight(int index)\\n    public abstract int getFingerprintSize()\\n    public abstract int hashCode()\\n    public abstract boolean equals(Object feature)\\n    public abstract String toString()\\n    public abstract BitSet getHashedFingerPrint()\\n    public abstract void addBinary(BitSet fingerprint) throws CDKException\\n    public abstract void add(IPatternFingerprinter fngp) throws CDKException\\n    public abstract String getFingerprintID()\\n    public abstract void setFingerprintID(String fingerprintID)\\n    public abstract int getFeatureCount()\\n    public abstract boolean hasFeature(IFeature key)\\n    public abstract double[] getWeightedHashedFingerPrint()\\n    public abstract void add(IFeature feature) throws CDKException\\n    public IPatternFingerprinter clone() throws CloneNotSupportedException\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/fingerprints/interfaces/IPatternFingerprinter.java', 'methodCount': 20, 'lineCount': 22}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/mechanism/interfaces/EnumSubstrateProduct.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class RBlastMoleculeSignature {\\n    private boolean useAromatics = true;;\\n    private boolean useCharge = true;;\\n    private boolean isBondSensitive = true;;\\n    private RBlastMoleculeFromSignatureBuilder builder;;\\n    public RBlastMoleculeSignature(IAtomContainer atomContainer);\\n    public int getAtomCount()\\n    public boolean isUseAromatics()\\n    public void setUseAromatics(boolean useAromatics)\\n    public boolean isUseCharge()\\n    public void setUseCharge(boolean useCharge)\\n    public boolean isBondSensitive()\\n    public void setBondSensitive(boolean isBondSensitive)\\n    public String getMoleculeCanonicalSignatureString()\\n    public String getSignatureStringForAtom(int atomIndex)\\n    public String getSignatureStringForAtom(int atomIndex, int height)\\n    public RBlastAtomSignature getAtomSignature(int atomIndex)\\n    public RBlastAtomSignature getAtomSignature(int atomIndex, int height)\\n    private void setFlags(RBlastAtomSignature atomSignature)\\n    private void setFlags(RBlastMoleculeFromSignatureBuilder builder)\\n    public String getSmilesForAtomSignature(int atomIndex, int height)\\n    public String getSmilesForAtomSignature(int atomIndex, int height, ICanonicalMoleculeLabeller labeller)\\n    public IAtomContainer getFragment(int atomIndex, int height)\\n    public IAtomContainer makeMoleculeFromSignature(String signatureString)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/signature/RBlastMoleculeSignature.java', 'methodCount': 19, 'lineCount': 25}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class CDKSMILES {\\n    private final ILoggingTool LOGGER = createLoggingTool(CDKSMILES.class);;\\n    private final IAtomContainer molecule;;\\n    public CDKSMILES(IAtomContainer mol, boolean removeH, boolean remove_AAM) throws CloneNotSupportedException;\\n    public String getCanonicalSMILES()\\n    private boolean isPseudoAtoms()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/tools/CDKSMILES.java', 'methodCount': 3, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 16}\n",
      "{'ReactionCenterFragment': 'public class ReactionCenterFragment {\\n    private static final long serialVersionUID = 9879878799977781L;;\\n    private final String signature;;\\n    private final int level;;\\n    private final EnumSubstrateProduct rpf;;\\n    public ReactionCenterFragment(String signature, int level, EnumSubstrateProduct rpf);\\n    public String toString()\\n    public int getLevel()\\n    public EnumSubstrateProduct getReactantProductInfo()\\n    public String getSignature()\\n}\\n'}\n",
      "/**\n",
      "     *\n",
      "     * @param molOrignal\n",
      "     * @param type\n",
      "     * @param atom\n",
      "     * @return\n",
      "     * @throws Exception\n",
      "     * @throws CloneNotSupportedException\n",
      "     */\n",
      "protected static List<ReactionCenterFragment> getCircularReactionPatternFingerprints(IAtomContainer molOrignal, IAtom atom, EnumSubstrateProduct type) throws Exception, CloneNotSupportedException\n",
      "**************************************\n",
      "protected static List<ReactionCenterFragment> getCircularReactionPatternFingerprints(IAtomContainer molOrignal, IAtom atom, EnumSubstrateProduct type) throws Exception, CloneNotSupportedException {\n",
      "    List<ReactionCenterFragment> fragmentsRC = new ArrayList<>();\n",
      "    IAtomContainer clone = molOrignal.clone();\n",
      "    for (int i = 0; i < 3; i++) {\n",
      "        String smiles = getCircularSMILES(clone, atom, i, true);\n",
      "        ReactionCenterFragment reactionCenterFragment = new ReactionCenterFragment(smiles, i, type);\n",
      "        fragmentsRC.add(reactionCenterFragment);\n",
      "    }\n",
      "    String smiles = getCircularSMILES(clone, atom, -1, true);\n",
      "    ReactionCenterFragment reactionCenterFragment = new ReactionCenterFragment(smiles, -1, type);\n",
      "    fragmentsRC.add(reactionCenterFragment);\n",
      "    return fragmentsRC;\n",
      "}\n",
      "2689\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 729/1243 [24:42<15:38,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30209\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 730/1243 [24:44<16:17,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30215\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 731/1243 [24:46<16:42,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30217\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 732/1243 [24:48<15:54,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30223\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 733/1243 [24:50<16:53,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30231\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 14}\n",
      "{}\n",
      "/**\n",
      "     * Calculates the center of the given atoms and returns it as a Point2d. See\n",
      "     * comment for center(IAtomContainer atomCon, Dimension areaDim, HashMap\n",
      "     * renderingCoordinates) for details on coordinate sets\n",
      "     *\n",
      "     * @param atoms The Iterator of the given atoms\n",
      "     * @return The center of the given atoms as Point2d\n",
      "     */\n",
      "public static Point2d get2DCenter(Iterator<IAtom> atoms)\n",
      "**************************************\n",
      "public static Point2d get2DCenter(Iterator<IAtom> atoms) {\n",
      "    IAtom atom;\n",
      "    double xsum = 0;\n",
      "    double ysum = 0;\n",
      "    int length = 0;\n",
      "    while (atoms.hasNext()) {\n",
      "        atom = (IAtom) atoms.next();\n",
      "        if (atom.getPoint2d() != null) {\n",
      "            xsum += atom.getPoint2d().x;\n",
      "            ysum += atom.getPoint2d().y;\n",
      "        }\n",
      "        ++length;\n",
      "    }\n",
      "    return new Point2d(xsum / (double) length, ysum / (double) length);\n",
      "}\n",
      "263\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 734/1243 [24:52<17:08,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30234\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 735/1243 [24:54<17:33,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30237\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 736/1243 [24:56<18:09,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30238\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 737/1243 [24:58<16:43,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30239\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 738/1243 [25:00<15:55,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30242\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 14}\n",
      "{}\n",
      "/**\n",
      "     * Returns the atom of the given molecule that is closest to the given\n",
      "     * coordinates. See comment for center(IAtomContainer atomCon, Dimension\n",
      "     * areaDim, HashMap renderingCoordinates) for details on coordinate sets\n",
      "     *\n",
      "     * @param xPosition The x coordinate\n",
      "     * @param yPosition The y coordinate\n",
      "     * @param atomCon The molecule that is searched for the closest atom\n",
      "     * @return The atom that is closest to the given coordinates\n",
      "     */\n",
      "public static IAtom getClosestAtom(int xPosition, int yPosition, IAtomContainer atomCon)\n",
      "**************************************\n",
      "public static IAtom getClosestAtom(int xPosition, int yPosition, IAtomContainer atomCon) {\n",
      "    IAtom closestAtom = null;\n",
      "    IAtom currentAtom;\n",
      "    double smallestMouseDistance = -1;\n",
      "    double mouseDistance;\n",
      "    double atomX;\n",
      "    double atomY;\n",
      "    for (int i = 0; i < atomCon.getAtomCount(); i++) {\n",
      "        currentAtom = atomCon.getAtom(i);\n",
      "        atomX = currentAtom.getPoint2d().x;\n",
      "        atomY = currentAtom.getPoint2d().y;\n",
      "        mouseDistance = Math.sqrt(Math.pow(atomX - xPosition, 2) + Math.pow(atomY - yPosition, 2));\n",
      "        if (mouseDistance < smallestMouseDistance || smallestMouseDistance == -1) {\n",
      "            smallestMouseDistance = mouseDistance;\n",
      "            closestAtom = currentAtom;\n",
      "        }\n",
      "    }\n",
      "    return closestAtom;\n",
      "}\n",
      "350\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  59%|    | 739/1243 [25:02<16:29,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30243\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 740/1243 [25:03<15:30,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30245\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 741/1243 [25:06<16:14,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30246\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 742/1243 [25:08<17:02,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30247\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 743/1243 [25:10<17:18,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30255\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 14}\n",
      "{}\n",
      "/**\n",
      "     * Determines if this AtomContainer contains 2D coordinates for some or all\n",
      "     * molecules. See comment for center(IAtomContainer atomCon, Dimension\n",
      "     * areaDim, HashMap renderingCoordinates) for details on coordinate sets\n",
      "     *\n",
      "     *\n",
      "     * @param container the molecule to be considered\n",
      "     * @return 0 no 2d, 1=some, 2= for each atom\n",
      "     * @deprecated use\n",
      "     * {@link #get2DCoordinateCoverage(org.openscience.cdk.interfaces.IAtomContainer)}\n",
      "     * for determining partial coordinates\n",
      "     * @see\n",
      "     * #get2DCoordinateCoverage(org.openscience.cdk.interfaces.IAtomContainer)\n",
      "     */\n",
      "public static int has2DCoordinatesNew(IAtomContainer container)\n",
      "**************************************\n",
      "@Deprecated\n",
      "public static int has2DCoordinatesNew(IAtomContainer container) {\n",
      "    if (container == null) {\n",
      "        return 0;\n",
      "    }\n",
      "    boolean no2d = false;\n",
      "    boolean with2d = false;\n",
      "    for (IAtom atom : container.atoms()) {\n",
      "        if (atom.getPoint2d() == null) {\n",
      "            no2d = true;\n",
      "        } else {\n",
      "            with2d = true;\n",
      "        }\n",
      "    }\n",
      "    if (!no2d && with2d) {\n",
      "        return 2;\n",
      "    } else if (no2d && with2d) {\n",
      "        return 1;\n",
      "    } else {\n",
      "        return 0;\n",
      "    }\n",
      "}\n",
      "344\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 744/1243 [25:12<17:07,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30258\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 745/1243 [25:13<15:06,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30288\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 746/1243 [25:15<15:44,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30298\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 747/1243 [25:18<17:07,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30311\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 748/1243 [25:19<16:03,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30339\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 87, 'classCount': 1, 'methodsOnly': 'public class Params {\\n    private static final Logger LOG = getLogger(Params.class.getName());;\\n    /**\\n */\\npublic BondStrokeCap bondStrokeCap = BUTT;;\\n    /**\\n */\\npublic BondStrokeJoin bondStrokeJoin = MITRE;;\\n    /**\\n */\\npublic XAlign leftRightAlignment = XAlign.CENTER;;\\n    /**\\n */\\npublic YAlign topBottomAlignment = YAlign.CENTER;;\\n    /**\\n */\\npublic int bondLength = 30;;\\n    /**\\n */\\npublic int borderX = 20;;\\n    /**\\n */\\npublic int borderY = 20;;\\n    /**\\n */\\npublic int plusGap = 20;;\\n    /**\\n */\\npublic int arrowLength = 30;;\\n    /**\\n */\\npublic int arrowGap = 10;;\\n    /**\\n */\\npublic int arrowHeadLength = 10;;\\n    /**\\n */\\npublic boolean drawBounds = false;;\\n    /**\\n */\\npublic boolean drawCarbons = false;;\\n    /**\\n */\\npublic boolean drawExplicitHydrogens = true;;\\n    /**\\n */\\npublic boolean drawImplicitHydrogens = true;;\\n    /**\\n */\\npublic boolean drawTerminalCarbons = true;;\\n    /**\\n */\\npublic int atomSymbolFontSize = 10;;\\n    /**\\n */\\npublic int plusFontSize = 14;;\\n    /**\\n */\\npublic boolean drawMappings = true;;\\n    /**\\n */\\npublic int subgraphBoxXBorder = 1;;\\n    /**\\n */\\npublic int subgraphBoxYBorder = 2;;\\n    /**\\n */\\npublic double doubleBondGap = 2;;\\n    /**\\n */\\npublic int subscriptHeight = 2;;\\n    /**\\n */\\npublic int subscriptTextSize = 9;;\\n    /**\\n */\\npublic boolean drawAromaticCircles = true;;\\n    /**\\n */\\npublic double ringProportion = 0.75;;\\n    /**\\n */\\npublic float bondStrokeWidth = 1.1f;;\\n    /**\\n */\\npublic double offsetBondDistanceProportion = 0.75;;\\n    /**\\n */\\npublic int filledWedgeWidth = 6;;\\n    /**\\n */\\npublic double wiggleLineWidth = 4;;\\n    /**\\n */\\npublic boolean drawAtomID = false;;\\n    /**\\n */\\npublic int atomIDFontSize = 7;;\\n    /**\\n */\\npublic double labelYGap = 10;;\\n    /**\\n */\\npublic int moleculeLabelFontSize = 7;;\\n    /**\\n */\\npublic int leftToRightMoleculeLabelFontSize = 9;;\\n    /**\\n */\\npublic int topToBottomMoleculeLabelFontSize = 8;;\\n    /**\\n */\\npublic boolean drawMoleculeID = true;;\\n    /**\\n */\\npublic boolean drawLonePairs = true;;\\n    /**\\n */\\npublic double electronRadius = 1.0;;\\n    /**\\n */\\npublic double bondMarkLength = 6;;\\n    /**\\n */\\npublic boolean drawSubgraphBoxes = true;;\\n    /**\\n */\\npublic double doubleMarkGap = 1;;\\n    /**\\n */\\npublic int lonePairSeparation = 4;;\\n    /**\\n */\\npublic boolean drawHighlights = true;;\\n    /**\\n */\\npublic double highlightRadius = 8;;\\n    /**\\n */\\npublic Color highlightColor = BLUE;;\\n    /**\\n */\\npublic boolean highlightsAbove = true;;\\n    /**\\n */\\npublic boolean highlightsBelow = false;;\\n    /**\\n */\\npublic float highlightAlpha = 0.15f;;\\n    /**\\n */\\npublic float highlightBondStroke = 4.0f;;\\n    /**\\n */\\npublic boolean drawSubgraphMappingLines = false;;\\n    /**\\n */\\npublic boolean colorSubgraphBoxes = true;;\\n    /**\\n */\\npublic boolean drawReactionID = false;;\\n    /**\\n */\\npublic boolean layoutLeftToRight = true;;\\n    /**\\n */\\npublic boolean highlightSubgraphs = false;;\\n    /**\\n */\\npublic boolean drawBondStereoChanges = true;;\\n    /**\\n */\\npublic double arrowHeadAngle = 45;;\\n    /**\\n */\\npublic double circularHighlightBorder = 5;;\\n    /**\\n */\\npublic boolean useCircularHighlight = false;;\\n    /**\\n */\\npublic double circularHighlightMinRadius = 10;;\\n    /**\\n */\\npublic boolean circularHighlightIsConcentric = true;;\\n    /**\\n */\\npublic boolean circularHighlightTransparentFilled = false;;\\n    /**\\n */\\npublic boolean useAntialias = true;;\\n    /**\\n */\\npublic double tripleBondGap = 2.5;;\\n    /**\\n */\\npublic boolean drawRS = false;;\\n    /**\\n */\\npublic int chiralSymbolFontSize = 9;;\\n    /**\\n */\\npublic float dashedWedgeStroke = 1.0f;;\\n    /**\\n */\\npublic double dashedGapFactor = 0.1;;\\n    /**\\n */\\npublic double dashedWidthFactor = 0.2;;\\n    /**\\n */\\npublic double dashedWedgeWidth = 6;;\\n    /**\\n */\\npublic int arrowHeadIndent = 5;;\\n    /**\\n */\\npublic int arrowBodyWidth = 5;;\\n    /**\\n */\\npublic boolean drawFatArrow = false;;\\n    /**\\n */\\npublic boolean drawArrowFilled = false;;\\n    /**\\n */\\npublic ArrowType arrowType = FORWARD;;\\n    /**\\n */\\npublic boolean alignMolecules = false;;\\n    /**\\n */\\npublic MoleculeAlignMethod moleculeAlignMethod = MAX_AXIS;;\\n    /**\\n */\\npublic boolean circularHighlightShowAtoms = true;;\\n    /**\\n */\\npublic boolean drawBondFormedCleavedMarks = true;;\\n    /**\\n */\\npublic boolean drawBondOrderChangedMarks = true;;\\n    /**\\n */\\npublic boolean drawLabelPanel = false;;\\n    /**\\n */\\npublic String labelPanelFont = \"ROMAN\";;\\n    /**\\n */\\npublic int labelPanelFontSize = 14;;\\n    /**\\n */\\npublic boolean shouldCrop = true;;\\n    /**\\n */\\npublic double labelPanelHeight = 20;;\\n    /**\\n */\\npublic double labelGap = 10;;\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/graphics/direct/Params.java', 'methodCount': 0, 'lineCount': 89}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param ac\n",
      "     */\n",
      "public void invert(IAtomContainer ac)\n",
      "**************************************\n",
      "public void invert(IAtomContainer ac) {\n",
      "    if (shouldInvert && ac.getProperty(INVERTED) == null || !((Boolean) ac.getProperty(INVERTED))) {\n",
      "        for (IAtom atom : ac.atoms()) {\n",
      "            atom.getPoint2d().y *= -1;\n",
      "        }\n",
      "        ac.setProperty(INVERTED, TRUE);\n",
      "    }\n",
      "    shouldInvert = false;\n",
      "}\n",
      "1801\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 749/1243 [25:22<16:29,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30356\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 750/1243 [25:24<17:52,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30366\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 751/1243 [25:26<17:51,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30391\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  60%|    | 752/1243 [25:28<17:40,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30424\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 753/1243 [25:31<17:44,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30425\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class BoundsTree {\\n    private static final Logger LOG = getLogger(BoundsTree.class.getName());;\\n    private Rectangle2D root;;\\n    private String rootLabel;;\\n    private Map<String, Rectangle2D> childMap;;\\n    public BoundsTree(String rootLabel);\\n    public BoundsTree(String rootLabel, String firstLabel, Rectangle2D firstBox);\\n    public BoundsTree(String rootLabel, BoundsTree... boundsTrees);\\n    public BoundsTree getSubtree(String prefix)\\n    public Rectangle2D getRoot()\\n    public void add(String label, Rectangle2D bounds)\\n    public void add(String label, Point2D point)\\n    public Rectangle2D getBounds(List<String> labels)\\n    public void setRoot(Rectangle2D root)\\n    public void add(String prefix, BoundsTree tree)\\n    public List<String> getBoundLabels()\\n    public void shift(double dx, double dy)\\n    public Rectangle2D get(String label)\\n    public double getWidth()\\n    public double getHeight()\\n    public Iterator<Rectangle2D> iterator()\\n    public BoundsTree transform(AffineTransform transform)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/graphics/direct/layout/BoundsTree.java', 'methodCount': 18, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param labelMap\n",
      "     * @param labelBounds\n",
      "     * @param g\n",
      "     */\n",
      "public void draw(Map<String, String> labelMap, BoundsTree labelBounds, Graphics2D g)\n",
      "**************************************\n",
      "/**\n",
      " * @param labelMap\n",
      " * @param labelBounds\n",
      " * @param g\n",
      " */\n",
      "public void draw(Map<String, String> labelMap, BoundsTree labelBounds, Graphics2D g) {\n",
      "    // layout labels according to the positions in the moleculeBounds\n",
      "    for (String boundsLabel : labelMap.keySet()) {\n",
      "        String label = labelMap.get(boundsLabel);\n",
      "        Rectangle2D bounds = labelBounds.get(boundsLabel);\n",
      "        double x = bounds.getCenterX();\n",
      "        double y = bounds.getCenterY();\n",
      "        Point2f p = super.getTextPoint(g, label, x, y);\n",
      "        //            System.out.println(\"drawing string at \" + x + \" \" + y);\n",
      "        g.drawString(label, p.x, p.y);\n",
      "    }\n",
      "}\n",
      "558\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 754/1243 [25:33<17:22,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30429\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 755/1243 [25:35<17:27,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30448\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 756/1243 [25:37<17:36,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30450\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 757/1243 [25:39<17:23,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30456\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 758/1243 [25:41<17:06,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30474\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/graphics/direct/Axis.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class BoundsTree {\\n    private static final Logger LOG = getLogger(BoundsTree.class.getName());;\\n    private Rectangle2D root;;\\n    private String rootLabel;;\\n    private Map<String, Rectangle2D> childMap;;\\n    public BoundsTree(String rootLabel);\\n    public BoundsTree(String rootLabel, String firstLabel, Rectangle2D firstBox);\\n    public BoundsTree(String rootLabel, BoundsTree... boundsTrees);\\n    public BoundsTree getSubtree(String prefix)\\n    public Rectangle2D getRoot()\\n    public void add(String label, Rectangle2D bounds)\\n    public void add(String label, Point2D point)\\n    public Rectangle2D getBounds(List<String> labels)\\n    public void setRoot(Rectangle2D root)\\n    public void add(String prefix, BoundsTree tree)\\n    public List<String> getBoundLabels()\\n    public void shift(double dx, double dy)\\n    public Rectangle2D get(String label)\\n    public double getWidth()\\n    public double getHeight()\\n    public Iterator<Rectangle2D> iterator()\\n    public BoundsTree transform(AffineTransform transform)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/reactionblast/graphics/direct/layout/BoundsTree.java', 'methodCount': 18, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param molSet\n",
      "     * @param molSetBoundsTree\n",
      "     * @param dx\n",
      "     * @param dy\n",
      "     */\n",
      "public void shiftMoleculeSet(IAtomContainerSet molSet, BoundsTree molSetBoundsTree, double dx, double dy)\n",
      "**************************************\n",
      "/**\n",
      " * @param molSet\n",
      " * @param molSetBoundsTree\n",
      " * @param dx\n",
      " * @param dy\n",
      " */\n",
      "public void shiftMoleculeSet(IAtomContainerSet molSet, BoundsTree molSetBoundsTree, double dx, double dy) {\n",
      "    //        System.out.println(molSetBoundsTree);\n",
      "    int counter = 0;\n",
      "    for (IAtomContainer molecule : molSet.atomContainers()) {\n",
      "        String molLabel = molSet.getID() + \"_\" + molecule.getID() + \":\" + counter;\n",
      "        //            System.out.println(\"shifting \" + molLabel + \" from \" + BoundsPrinter.toString(GeometryTools.getRectangle2D(molecule)));\n",
      "        Rectangle2D bounds = molSetBoundsTree.get(molLabel);\n",
      "        bounds.setFrame(bounds.getMinX() + dx, bounds.getMinY() + dy, bounds.getWidth(), bounds.getHeight());\n",
      "        translate2D(molecule, dx, dy);\n",
      "        //            System.out.println(\"shifting \" + molecule.getID() + \" to \" + BoundsPrinter.toString(GeometryTools.getRectangle2D(molecule)));\n",
      "        counter++;\n",
      "    }\n",
      "}\n",
      "707\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 759/1243 [25:43<17:09,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30521\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 760/1243 [25:45<15:55,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|    | 761/1243 [25:47<14:58,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30547\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|   | 762/1243 [25:49<15:09,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30548\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|   | 763/1243 [25:51<15:28,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30549\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 10}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param position\n",
      "     * @return\n",
      "     */\n",
      "public Vector2d getRightPerpendicularFromPosition(AnnotationPosition position)\n",
      "**************************************\n",
      "public Vector2d getRightPerpendicularFromPosition(AnnotationPosition position) {\n",
      "    switch(position) {\n",
      "        case N:\n",
      "            return vE;\n",
      "        case NE:\n",
      "            return vSE;\n",
      "        case E:\n",
      "            return vS;\n",
      "        case SE:\n",
      "            return vSW;\n",
      "        case S:\n",
      "            return vW;\n",
      "        case SW:\n",
      "            return vNW;\n",
      "        case W:\n",
      "            return vN;\n",
      "        case NW:\n",
      "            return vNE;\n",
      "        default:\n",
      "            return vS;\n",
      "    }\n",
      "}\n",
      "172\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  61%|   | 764/1243 [25:53<16:37,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30551\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 765/1243 [25:55<15:46,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30577\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 766/1243 [25:57<15:57,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30638\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 767/1243 [25:59<16:26,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30641\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 768/1243 [26:01<15:28,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30686\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Descriptor {\\n    public Type getType()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/uk/ac/ebi/centres/Descriptor.java', 'methodCount': 1, 'lineCount': 3}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 8}\n",
      "{'DescriptorList': 'public class DescriptorList {\\n    private final List<Descriptor> descriptors = new ArrayList<>();;\\n    private final Set<Descriptor> ignore = new HashSet<>();;\\n    private int pairing = 0;;\\n    public DescriptorList(Descriptor... ignore);\\n    public DescriptorList(List<Descriptor> ignores);\\n    public DescriptorList(DescriptorList head, DescriptorList tail);\\n    public boolean ignore(Descriptor ignore)\\n    public boolean ignoreAll(Collection<? extends Descriptor> ignores)\\n    public boolean add(Descriptor descriptor)\\n    private boolean addAndPair(Descriptor descriptor)\\n    public boolean addAll(Collection<? extends Descriptor> descriptors)\\n    public int getPairing()\\n    public List<DescriptorList> append(Collection<? extends DescriptorList> lists)\\n    public String toString()\\n    public int compareTo(DescriptorList o)\\n    public void clear()\\n}\\n'}\n",
      "/**\n",
      "     * Appends multiple descriptor lists. If more then one list is provided the\n",
      "     * head (this list) is duplicate across the multiple tails (provided). If\n",
      "     * the contents of this list is 'RRSS' and we invoke append with two lists\n",
      "     * 'SRS' and 'RSR'. Two new lists will be returned with their contents\n",
      "     * 'RRSSSRS' and 'RRSSRSR' respectively.\n",
      "     * <p/>\n",
      "     * Empty descriptor lists are not appended, if all descriptor lists are\n",
      "     * empty then 'this' list is the single returned list\n",
      "     *\n",
      "     * @param lists multiple descriptor lists to be appended to this list.\n",
      "     *\n",
      "     * @return modified list of descriptors based on the provided input lists\n",
      "     */\n",
      "public List<DescriptorList> append(Collection<? extends DescriptorList> lists)\n",
      "**************************************\n",
      "public List<DescriptorList> append(Collection<? extends DescriptorList> lists) {\n",
      "    List<DescriptorList> created = new ArrayList<>(lists.size());\n",
      "    for (DescriptorList list : lists) {\n",
      "        if (!list.descriptors.isEmpty()) {\n",
      "            created.add(new DescriptorList(this, list));\n",
      "        }\n",
      "    }\n",
      "    if (created.isEmpty()) {\n",
      "        created.add(this);\n",
      "    }\n",
      "    return created;\n",
      "}\n",
      "562\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 769/1243 [26:03<16:27,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30710\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 770/1243 [26:05<15:10,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30842\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 771/1243 [26:07<15:41,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30860\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 772/1243 [26:09<14:49,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30884\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 773/1243 [26:10<14:37,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30912\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 10}\n",
      "{'Vertex': 'public class Vertex {\\n    private Integer query;;\\n    private Integer target;;\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public String toString()\\n    public String getLabel()\\n    public void setLabel(String label)\\n    private final int id;;\\n    private String label;;\\n    public Vertex(int node);\\n    public int compareTo(Vertex o)\\n    public int compare(Vertex o1, Vertex o2)\\n    public int getID()\\n    public void setCompatibilityBondPair(Integer a, Integer b)\\n    public Integer getTargetBondIndex()\\n    public Integer getQueryBondIndex()\\n}\\n', 'Edge': 'public class Edge {\\n    public int hashCode()\\n    public boolean equals(Object obj)\\n    public EdgeType getEdgeType()\\n    public void setEdgeType(EdgeType edgeType)\\n    public String toString()\\n    private static final long serialVersionUID = 52343464641L;;\\n    private final Integer source;;\\n    private final Integer sink;;\\n    private EdgeType edgeType;;\\n    public Edge(Integer i, Integer j);\\n    public Integer getSource()\\n    public Integer getSink()\\n    public boolean isC_Edge()\\n    public boolean isD_Edge()\\n}\\n'}\n",
      "/**\n",
      "     * Returns edges of the vertex\n",
      "     *\n",
      "     * @param currentVertex\n",
      "     * @return\n",
      "     */\n",
      "public Iterable<Edge> edgesOf(Vertex currentVertex)\n",
      "**************************************\n",
      "public Iterable<Edge> edgesOf(Vertex currentVertex) {\n",
      "    validateVertex(currentVertex);\n",
      "    Integer v = vertices.indexOf(currentVertex);\n",
      "    Set<Edge> edgesOfVertex = new LinkedHashSet<>();\n",
      "    edges().stream().map((e) -> {\n",
      "        if (e.getSource().equals(v)) {\n",
      "            edgesOfVertex.add(e);\n",
      "        }\n",
      "        return e;\n",
      "    }).filter((e) -> (e.getSink().equals(v))).forEachOrdered((e) -> {\n",
      "        edgesOfVertex.add(e);\n",
      "    });\n",
      "    return edgesOfVertex;\n",
      "}\n",
      "461\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 774/1243 [26:13<15:39,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31088\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 775/1243 [26:15<16:14,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31123\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  62%|   | 776/1243 [26:17<16:24,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31351\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 777/1243 [26:19<16:19,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31371\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 778/1243 [26:22<16:49,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31415\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Builds the initial extension set. This is the set of node that may be\n",
      "     * used as seed for the CDKRGraph parsing. This set depends on the\n",
      "     * constrains defined by the user.\n",
      "     *\n",
      "     * @param sourceBitSet constraint in the graph G1\n",
      "     * @param targetBitSet constraint in the graph G2\n",
      "     * @return\n",
      "     */\n",
      "private BitSet buildB(BitSet sourceBitSet, BitSet targetBitSet) throws CDKException\n",
      "**************************************\n",
      "private BitSet buildB(BitSet sourceBitSet, BitSet targetBitSet) throws CDKException {\n",
      "    this.setSourceBitSet(sourceBitSet);\n",
      "    this.setTargetBitSet(targetBitSet);\n",
      "    BitSet bistSet = new BitSet();\n",
      "    getGraph().stream().filter((rNode) -> ((sourceBitSet.get(rNode.getRMap().getId1()) || sourceBitSet.isEmpty()) && (targetBitSet.get(rNode.getRMap().getId2()) || targetBitSet.isEmpty()))).forEachOrdered((rNode) -> {\n",
      "        bistSet.set(getGraph().indexOf(rNode));\n",
      "    });\n",
      "    return bistSet;\n",
      "}\n",
      "269\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 779/1243 [26:24<17:25,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31421\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 780/1243 [26:26<17:23,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31424\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 781/1243 [26:28<15:44,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31455\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 782/1243 [26:30<15:44,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31461\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 783/1243 [26:32<16:12,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31468\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class AtomBondMatcher {\\n    private static final ILoggingTool LOGGER = LoggingToolFactory.createLoggingTool(AtomBondMatcher.class);;\\n    private final static boolean DEBUG = false;;\\n    public static boolean matchAtomAndBond(IBond b1, IBond b2, AtomMatcher atomMatcher, BondMatcher bondMatcher, boolean undirected)\\n    public static boolean matches(IBond bondA1, IBond bondA2, BondMatcher bm)\\n    public static boolean matches(IAtom a1, IAtom a2, AtomMatcher am)\\n    public static AtomMatcher atomMatcher(boolean matchAtomTypes, boolean shouldMatchRings)\\n    public static BondMatcher bondMatcher(boolean matchBond, boolean shouldMatchRings)\\n    public static AtomMatcher queryAtomMatcher()\\n    public static BondMatcher queryBondMatcher()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/algorithm/matchers/AtomBondMatcher.java', 'methodCount': 7, 'lineCount': 11}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class AtomMatcher {\\n    public abstract String toString()\\n    public abstract boolean matches(IAtom atom1, IAtom atom2)\\n    public static AtomMatcher forAny()\\n    public static AtomMatcher forElement()\\n    public static AtomMatcher forQuery()\\n    public static AtomMatcher forRingAtomTypeMatcher()\\n    public static AtomMatcher forAtomTypeMatcher()\\n    public static AtomMatcher forRingMatcher()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/algorithm/matchers/AtomMatcher.java', 'methodCount': 8, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class BondMatcher {\\n    public abstract String toString()\\n    public abstract boolean matches(IBond bond1, IBond bond2)\\n    public static BondMatcher forAny()\\n    public static BondMatcher forStrictOrder()\\n    public static BondMatcher forOrder()\\n    public static BondMatcher forRing()\\n    public static BondMatcher forQuery()\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/algorithm/matchers/BondMatcher.java', 'methodCount': 7, 'lineCount': 9}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class BinaryTree {\\n    public void setValue(int value)\\n    public BinaryTree(int value);\\n    public BinaryTree();\\n    private BinaryTree equal;;\\n    private BinaryTree notEqual;;\\n    private int value;;\\n    public synchronized int getValue()\\n    public synchronized BinaryTree getEqual()\\n    public synchronized void setEqual(BinaryTree tree)\\n    public synchronized BinaryTree getNotEqual()\\n    public synchronized void setNotEqual(BinaryTree tree)\\n    public static int remove_tree_structure(BinaryTree cur_struc)\\n}\\n', 'filePath': '/home/ubuntu/repos/asad,ReactionDecoder/src/main/java/org/openscience/smsd/helper/BinaryTree.java', 'methodCount': 9, 'lineCount': 14}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param TEMPMARCS\n",
      "     * @param neighborBondNumA\n",
      "     * @param neighborBondNumB\n",
      "     * @return count\n",
      "     */\n",
      "protected static int countArcsLeft(List<Integer> TEMPMARCS, int neighborBondNumA, int neighborBondNumB)\n",
      "**************************************\n",
      "protected static int countArcsLeft(List<Integer> TEMPMARCS, int neighborBondNumA, int neighborBondNumB) {\n",
      "    int arcsleft = 0;\n",
      "    for (int a = 0; a < neighborBondNumA; a++) {\n",
      "        for (int b = 0; b < neighborBondNumB; b++) {\n",
      "            if (TEMPMARCS.get(a * neighborBondNumB + b) == (1)) {\n",
      "                arcsleft++;\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return arcsleft;\n",
      "}\n",
      "931\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 784/1243 [26:34<15:03,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31564\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 785/1243 [26:35<14:07,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31580\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 786/1243 [26:37<14:42,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31611\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 787/1243 [26:40<15:18,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31613\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 788/1243 [26:42<15:20,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31632\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param molecule\n",
      "     * @throws CDKException\n",
      "     */\n",
      "public static void aromatizeDayLight(IAtomContainer molecule) throws CDKException\n",
      "**************************************\n",
      "public static void aromatizeDayLight(IAtomContainer molecule) throws CDKException {\n",
      "    ElectronDonation model = ElectronDonation.daylight();\n",
      "    CycleFinder cycles = Cycles.or(Cycles.all(), Cycles.relevant());\n",
      "    Aromaticity aromaticity = new Aromaticity(model, cycles);\n",
      "    try {\n",
      "        aromaticity.apply(molecule);\n",
      "    } catch (CDKException e) {\n",
      "        LOGGER.error(Level.WARNING, \"Aromaticity detection failed due to presence of unset \" + \"atom hybridisation\", molecule.getID());\n",
      "    }\n",
      "}\n",
      "186\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  63%|   | 789/1243 [26:44<15:20,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31633\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 790/1243 [26:46<15:44,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31664\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 791/1243 [26:49<17:17,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31686\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 792/1243 [26:51<17:38,  2.35s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31687\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 793/1243 [26:53<16:00,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31724\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class IlluminaUtil {\\n    public static final String BARCODE_DELIMITER = \"-\";;\\n    public static Integer getTileFromReadName(final String readName)\\n    public static byte[] makePhredBinaryFromSolexaQualityAscii_1_3(final String solexaQualities)\\n    public static byte[] makePhredBinaryFromSolexaQualityAscii_1_3(final String solexaQualities, final int offset, final int length)\\n    public static void convertSolexaQualityAscii_1_1_ToPhredBinary(final byte[] solexaQualities)\\n    public static byte getSolexaQualityCharFromFourQualities(final String[] qualities, final int cycleNumber, final FormatUtil formatter)\\n    public static String barcodeSeqsToString(final List<String> barcodes)\\n    public static String barcodeSeqsToString(final String[] barcodes)\\n    public static String barcodeSeqsToString(final byte[][] barcodes)\\n    public static String stringSeqsToString(final String[] barcodes, String delim)\\n    public static String byteArrayToString(final byte[][] barcodes, String delim)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/IlluminaUtil.java', 'methodCount': 10, 'lineCount': 13}, {'fieldCount': 28, 'classCount': 1, 'methodsOnly': 'public class HelpConstants {\\n    private HelpConstants();\\n    /**\\n * Definition of the group names / descriptions for documentation/help purposes.\\n */\\npublic final static String DOC_CAT_BASE_CALLING = \"Base Calling\";;\\n    public final static String DOC_CAT_BASE_CALLING_SUMMARY = \"Tools that process sequencing machine data, e.g. Illumina base calls, and detect sequencing level attributes, e.g. adapters\";;\\n    public final static String DOC_CAT_DIAGNOSTICS_AND_QC = \"Diagnostics and Quality Control\";;\\n    public final static String DOC_CAT_DIAGNOSTICS_AND_QC_SUMMARY = \"Tools that collect sequencing quality related and comparative metrics\";;\\n    public final static String DOC_CAT_INTERVALS_MANIPULATION = \"Intervals Manipulation\";;\\n    public final static String DOC_CAT_INTERVALS_MANIPULATION_SUMMARY = \"Tools that process genomic intervals in various formats\";;\\n    public final static String DOC_CAT_OTHER = \"Other\";;\\n    public final static String DOC_CAT_OTHER_SUMMARY = \"Miscellaneous tools, e.g. those that aid in data streaming\";;\\n    public final static String DOC_CAT_READ_DATA_MANIPULATION = \"Read Data Manipulation\";;\\n    public final static String DOC_CAT_READ_DATA_MANIPULATION_SUMMARY = \"Tools that manipulate read data in SAM, BAM or CRAM format\";;\\n    public final static String DOC_CAT_REFERENCE = \"Reference\";;\\n    public final static String DOC_CAT_REFERENCE_SUMMARY = \"Tools that analyze and manipulate FASTA format references\";;\\n    public final static String DOC_CAT_TEST = \"Test Tools\";;\\n    public final static String DOC_CAT_TEST_SUMMARY = \"Tools for internal test purposes\";;\\n    public final static String DOC_CAT_VARIANT_FILTERING = \"Variant Filtering\";;\\n    public final static String DOC_CAT_VARIANT_FILTERING_SUMMARY = \"Tools that filter variants by annotating the FILTER column\";;\\n    public final static String DOC_CAT_VARIANT_EVALUATION = \"Variant Evaluation and Refinement\";;\\n    public final static String DOC_CAT_VARIANT_EVALUATION_SUMMARY = \"Tools that evaluate and refine variant calls, e.g. with annotations not offered by the engine\";;\\n    public final static String DOC_CAT_VARIANT_MANIPULATION = \"Variant Manipulation\";;\\n    public final static String DOC_CAT_VARIANT_MANIPULATION_SUMMARY = \"Tools that manipulate variant call format (VCF) data\";;\\n    public final static String DOC_CAT_GENOTYPING_ARRAYS_MANIPULATION = \"Genotyping Arrays Manipulation\";;\\n    public final static String DOC_CAT_GENOTYPING_ARRAYS_MANIPULATION_SUMMARY = \"Tools that manipulate data generated by Genotyping arrays\";;\\n    public final static String DOC_CAT_METRICS = \"Metrics\";;\\n    public final static String DOC_CAT_METRICS_SUMMARY = \"Metrics\";;\\n    /**\\n * List of \"supercategory\" values used for doc purposes. Every doc group name can/should be put into\\n * one of the following supercategories.\\n */\\npublic final static String DOC_SUPERCAT_TOOLS = \"tools\";;\\n    public final static String DOC_SUPERCAT_UTILITIES = \"utilities\";;\\n    public final static String DOC_SUPERCAT_EXCLUDE = \"exclude\";;\\n    private static Map<String, String> groupToSuperCategory;;\\n    public static Map<String, String> getSuperCategoryMap()\\n    public static String getSuperCategoryProperty(final String groupName)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/help/HelpConstants.java', 'methodCount': 3, 'lineCount': 33}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'BarcodeMetric': 'public class BarcodeMetric {\\n    /**\\n * The barcode (from the set of expected barcodes) for which the following metrics apply.\\n * Note that the \"symbolic\" barcode of NNNNNN is used to report metrics for all reads that\\n * do not match a barcode.\\n */\\npublic String BARCODE;;\\n    public String BARCODE_WITHOUT_DELIMITER;;\\n    /**\\n * The barcode name.\\n */\\npublic String BARCODE_NAME = \"\";;\\n    /**\\n * The name of the library\\n */\\npublic String LIBRARY_NAME = \"\";;\\n    /**\\n * The total number of reads matching the barcode.\\n */\\npublic long READS = 0;;\\n    /**\\n * The number of PF reads matching this barcode (always less than or equal to READS).\\n */\\npublic long PF_READS = 0;;\\n    /**\\n * The number of all reads matching this barcode that matched with 0 errors or no-calls.\\n */\\npublic long PERFECT_MATCHES = 0;;\\n    /**\\n * The number of PF reads matching this barcode that matched with 0 errors or no-calls.\\n */\\npublic long PF_PERFECT_MATCHES = 0;;\\n    /**\\n * The number of all reads matching this barcode that matched with 1 error or no-call.\\n */\\npublic long ONE_MISMATCH_MATCHES = 0;;\\n    /**\\n * The number of PF reads matching this barcode that matched with 1 error or no-call.\\n */\\npublic long PF_ONE_MISMATCH_MATCHES = 0;;\\n    /**\\n * The fraction of all reads in the lane that matched to this barcode.\\n */\\npublic double PCT_MATCHES = 0d;;\\n    /**\\n * The rate of all reads matching this barcode to all reads matching the most prevalent barcode. For the\\n * most prevalent barcode this will be 1, for all others it will be less than 1 (except for the possible\\n * exception of when there are more orphan reads than for any other barcode, in which case the value\\n * may be arbitrarily large).  One over the lowest number in this column gives you the fold-difference\\n * in representation between barcodes.\\n */\\npublic double RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT = 0d;;\\n    /**\\n * The fraction of PF reads in the lane that matched to this barcode.\\n */\\npublic double PF_PCT_MATCHES = 0d;;\\n    /**\\n * The rate of PF reads matching this barcode to PF reads matching the most prevalent barcode. For the\\n * most prevalent barcode this will be 1, for all others it will be less than 1 (except for the possible\\n * exception of when there are more orphan reads than for any other barcode, in which case the value\\n * may be arbitrarily large).  One over the lowest number in this column gives you the fold-difference\\n * in representation of PF reads between barcodes.\\n */\\npublic double PF_RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT = 0d;;\\n    /**\\n * The \"normalized\" matches to each barcode. This is calculated as the number of pf reads matching\\n * this barcode over the sum of all pf reads matching any barcode (excluding orphans). If all barcodes\\n * are represented equally this will be 1.\\n */\\npublic double PF_NORMALIZED_MATCHES;;\\n    protected byte[][] barcodeBytes;;\\n    public BarcodeMetric(final String barcodeName, final String libraryName, final String barcodeDisplay, final String[] barcodeSeqs);\\n    public BarcodeMetric();\\n    public BarcodeMetric copy()\\n    public void merge(final BarcodeMetric metric)\\n}\\n'}\n",
      "/**\n",
      "     * Creates a copy of metric initialized with only non-accumulated and non-calculated values set\n",
      "     */\n",
      "public BarcodeMetric copy()\n",
      "**************************************\n",
      "/**\n",
      " * Creates a copy of metric initialized with only non-accumulated and non-calculated values set\n",
      " */\n",
      "public BarcodeMetric copy() {\n",
      "    final BarcodeMetric result = new BarcodeMetric();\n",
      "    result.BARCODE = this.BARCODE;\n",
      "    result.BARCODE_WITHOUT_DELIMITER = this.BARCODE_WITHOUT_DELIMITER;\n",
      "    result.BARCODE_NAME = this.BARCODE_NAME;\n",
      "    result.LIBRARY_NAME = this.LIBRARY_NAME;\n",
      "    result.barcodeBytes = this.barcodeBytes;\n",
      "    return result;\n",
      "}\n",
      "2105\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 794/1243 [26:56<17:46,  2.38s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31776\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 795/1243 [26:57<16:02,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31855\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 796/1243 [27:00<15:57,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31859\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 797/1243 [27:02<15:58,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31863\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 798/1243 [27:04<16:15,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31867\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Either set this value if not already set, or if already set, throw an exception if new value != current value.\n",
      "     */\n",
      "public void setOrCheckY(final int y)\n",
      "**************************************\n",
      "public void setOrCheckY(final int y) {\n",
      "    if (yIsSet()) {\n",
      "        if (this.y != y) {\n",
      "            throw new PicardException(\"Y value mismatch for \" + this + \" : \" + this.y + \" != \" + y);\n",
      "        }\n",
      "    } else {\n",
      "        this.y = y;\n",
      "    }\n",
      "}\n",
      "239\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 799/1243 [27:06<15:51,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31898\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 800/1243 [27:08<14:50,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31948\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  64%|   | 801/1243 [27:09<13:52,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32012\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 802/1243 [27:11<14:18,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32060\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 803/1243 [27:14<14:54,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32077\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     * Returns the tile of the next cluster that will be returned by PerTilePerCycleParser and therefore should be called before\n",
      "     * next() if you want to know the tile for the data returned by next()\n",
      "     *\n",
      "     * @return The tile number of the next ILLUMINA_DATA object to be returned\n",
      "     */\n",
      "public int getTileOfNextCluster()\n",
      "**************************************\n",
      "public int getTileOfNextCluster() {\n",
      "    if (cycleFilesParser.hasNext()) {\n",
      "        return currentTile;\n",
      "    }\n",
      "    if (currentTile < tileOrder.last()) {\n",
      "        return tileOrder.higher(currentTile);\n",
      "    }\n",
      "    throw new NoSuchElementException();\n",
      "}\n",
      "259\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 804/1243 [27:16<14:57,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32081\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 805/1243 [27:17<14:14,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32163\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 806/1243 [27:20<14:37,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32226\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 807/1243 [27:22<15:06,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32286\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 808/1243 [27:24<15:30,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32404\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{'InfiniumFileTOC': 'public class InfiniumFileTOC {\\n    private short tableOfContentsId;;\\n    private int offset;;\\n     short getTableOfContentsId()\\n     void setTableOfContentsId(short tableOfContentsId)\\n    public int getOffset()\\n    public void setOffset(int offset)\\n}\\n'}\n",
      "/**\n",
      "     * Parse the table of contents.\n",
      "     *\n",
      "     * @return The TOC\n",
      "     * @throws java.io.IOException Any errors reading in the TOC\n",
      "     */\n",
      " InfiniumFileTOC[] getTableOfContents() throws IOException\n",
      "**************************************\n",
      "InfiniumFileTOC[] getTableOfContents() throws IOException {\n",
      "    final InfiniumFileTOC[] tableOfContents = new InfiniumFileTOC[getNumberOfEntries()];\n",
      "    for (int i = 0; i < getNumberOfEntries(); i++) {\n",
      "        final InfiniumFileTOC toc = new InfiniumFileTOC();\n",
      "        toc.setTableOfContentsId(Short.reverseBytes(stream.readShort()));\n",
      "        toc.setOffset(Integer.reverseBytes(stream.readInt()));\n",
      "        tableOfContents[i] = toc;\n",
      "    }\n",
      "    return tableOfContents;\n",
      "}\n",
      "376\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 809/1243 [27:26<14:17,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32459\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 810/1243 [27:28<14:42,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32521\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 811/1243 [27:29<13:37,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32543\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 812/1243 [27:31<12:56,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32552\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 813/1243 [27:33<13:39,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32558\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Returns the index of the largest element in the array.  If there are multiple equal maxima then\n",
      "     * the earliest one in the array is returned.\n",
      "     */\n",
      "public static int indexOfMax(final double[] nums)\n",
      "**************************************\n",
      "public static int indexOfMax(final double[] nums) {\n",
      "    double max = nums[0];\n",
      "    int index = 0;\n",
      "    for (int i = 1; i < nums.length; ++i) {\n",
      "        if (nums[i] > max) {\n",
      "            max = nums[i];\n",
      "            index = i;\n",
      "        }\n",
      "    }\n",
      "    return index;\n",
      "}\n",
      "250\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  65%|   | 814/1243 [27:34<11:00,  1.54s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32560\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 815/1243 [27:36<11:33,  1.62s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32564\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 816/1243 [27:37<11:36,  1.63s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32565\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 817/1243 [27:39<11:19,  1.60s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32567\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 818/1243 [27:40<10:18,  1.46s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32568\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Returns the index of the smallest element in the array.  If there are multiple equal minima then\n",
      "     * the earliest one in the array is returned.\n",
      "     */\n",
      "public static int indexOfMin(final double[] nums)\n",
      "**************************************\n",
      "public static int indexOfMin(final double[] nums) {\n",
      "    double min = nums[0];\n",
      "    int index = 0;\n",
      "    for (int i = 1; i < nums.length; ++i) {\n",
      "        if (nums[i] < min) {\n",
      "            min = nums[i];\n",
      "            index = i;\n",
      "        }\n",
      "    }\n",
      "    return index;\n",
      "}\n",
      "250\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 819/1243 [27:43<13:00,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32573\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 820/1243 [27:44<11:33,  1.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32575\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 821/1243 [27:45<11:20,  1.61s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32577\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 822/1243 [27:47<11:28,  1.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32578\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 823/1243 [27:48<10:23,  1.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32579\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * calculates the sum of an array and a double.\n",
      "     */\n",
      "public static double[] sum(final double[] lhs, final double rhs)\n",
      "**************************************\n",
      "public static double[] sum(final double[] lhs, final double rhs) {\n",
      "    final int len = lhs.length;\n",
      "    final double[] result = new double[len];\n",
      "    for (int i = 0; i < len; ++i) {\n",
      "        result[i] = lhs[i] + rhs;\n",
      "    }\n",
      "    return result;\n",
      "}\n",
      "226\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 824/1243 [27:50<11:32,  1.65s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32612\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 825/1243 [27:52<11:45,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32637\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  66%|   | 826/1243 [27:54<11:42,  1.68s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32667\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 827/1243 [27:55<11:35,  1.67s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32673\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 828/1243 [27:57<12:23,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32702\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Parse the tile # from the read name.\n",
      "     * If we find that there are other elements needed from the read name, it might be a good idea to put\n",
      "     * makeReadName() and various get..() methods into a new class.\n",
      "     *\n",
      "     * @param readName As produced by IlluminaUtil.makeReadName()\n",
      "     * @return tile number, or null if read name is not in correct format.\n",
      "     */\n",
      "public static Integer getTileFromReadName(final String readName)\n",
      "**************************************\n",
      "/**\n",
      " * Parse the tile # from the read name.\n",
      " * If we find that there are other elements needed from the read name, it might be a good idea to put\n",
      " * makeReadName() and various get..() methods into a new class.\n",
      " *\n",
      " * @param readName As produced by IlluminaUtil.makeReadName()\n",
      " * @return tile number, or null if read name is not in correct format.\n",
      " */\n",
      "public static Integer getTileFromReadName(final String readName) {\n",
      "    final int first = readName.indexOf(':');\n",
      "    if (first > 0) {\n",
      "        final int second = readName.indexOf(':', first + 1);\n",
      "        if (second > 0) {\n",
      "            final int third = readName.indexOf(':', second + 1);\n",
      "            if (third > 0) {\n",
      "                return Integer.parseInt(readName.substring(second + 1, third));\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return null;\n",
      "}\n",
      "423\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 829/1243 [28:00<13:27,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32706\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 830/1243 [28:01<12:37,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32730\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 831/1243 [28:03<12:37,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32740\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 832/1243 [28:05<12:18,  1.80s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32789\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 833/1243 [28:07<12:07,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32805\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * Attempt to load a Properties object from an on-disk properties file.\n",
      "     * @param propertyFilePath name of the properties file to load. Must have a .properties extension\n",
      "     * @param clazz class used to obtain a class loader to use to locate the properties file\n",
      "     * @return null if the files doesn't exist or isn't readable, otherwise a Properties object\n",
      "     */\n",
      "public static Properties loadPropertiesFile(final String propertyFilePath, final Class<?> clazz)\n",
      "**************************************\n",
      "/**\n",
      " * Attempt to load a Properties object from an on-disk properties file.\n",
      " * @param propertyFilePath name of the properties file to load. Must have a .properties extension\n",
      " * @param clazz class used to obtain a class loader to use to locate the properties file\n",
      " * @return null if the files doesn't exist or isn't readable, otherwise a Properties object\n",
      " */\n",
      "public static Properties loadPropertiesFile(final String propertyFilePath, final Class<?> clazz) {\n",
      "    Utils.nonNull(propertyFilePath);\n",
      "    try (final InputStream inputStream = clazz.getClassLoader().getResourceAsStream(propertyFilePath)) {\n",
      "        if (inputStream != null) {\n",
      "            final Properties properties = new Properties();\n",
      "            properties.load(inputStream);\n",
      "            return properties;\n",
      "        } else {\n",
      "            return null;\n",
      "        }\n",
      "    } catch (IOException ex) {\n",
      "        throw new RuntimeException(String.format(\"IOException loading properties file %s\", propertyFilePath), ex);\n",
      "    }\n",
      "}\n",
      "317\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 834/1243 [28:09<12:42,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32840\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 835/1243 [28:10<12:30,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32865\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 836/1243 [28:13<13:05,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "32920\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 837/1243 [28:15<13:18,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33039\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 838/1243 [28:16<12:39,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33173\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class MathUtil {\\n    private MathUtil();\\n    /**\\n * The double value closest to 1 while still being less than 1.\\n */\\npublic static final double MAX_PROB_BELOW_ONE = 0.9999999999999999d;;\\n    /**\\n * Constant to convert between the natural base e and 4.0.  Useful for\\n * entropy calculations on DNA.\\n */\\npublic static final double LOG_4_BASE_E = Math.log(4.0);;\\n    public static double log10_1p(final double x)\\n    public static double mean(final double[] in, final int start, final int stop)\\n    public static double mean(final double[] in)\\n    public static double stddev(final double[] in, final int start, final int length)\\n    public static double stddev(final double[] in, final int start, final int stop, final double mean)\\n    public static double stddev(final double[] in, final double mean)\\n    public static int compare(final int v1, final int v2)\\n    public static double median(final double... in)\\n    public static Double percentageOrNull(final Long numerator, final Long denominator)\\n    public static double round(final double num, final int precision)\\n    public static double divide(final double numerator, final double denominator)\\n    public static double max(final double[] nums)\\n    public static double[] capFromBelow(final double[] nums, final double floor)\\n    public static int indexOfMax(final double[] nums)\\n    public static long max(final long[] nums)\\n    public static int indexOfMax(final long[] nums)\\n    public static double[] subtractMax(final double[] logLikelihoods)\\n    public static double min(final double[] nums)\\n    public static int min(final int[] nums)\\n    public static short min(final short[] nums)\\n    public static byte min(final byte[] nums)\\n    public static double[] capFromAbove(final double[] nums, final double top)\\n    public static int indexOfMin(final int[] nums)\\n    public static int indexOfMin(final double[] nums)\\n    public static double[] seq(final double from, final double to, final double by)\\n    public static double[] promote(final int[] is)\\n    public static double[] logLikelihoodsToProbs(final double[] likelihoods)\\n    public static double[] getProbabilityFromLog(final double[] lLikelihood)\\n    public static double[] getLogFromProbability(final double[] likelihood)\\n    public static double[] pNormalizeLogProbability(final double[] lPosterior)\\n    public static double[] divide(final double[] numerators, final double[] denominators)\\n    public static double[] pNormalizeVector(final double[] pPosterior)\\n    public static double[] multiply(final double[] lhs, final double[] rhs)\\n    public static double[] sum(final double[] lhs, final double[] rhs)\\n    public static double[] sum(final double[] lhs, final double rhs)\\n    public static double sum(final double[] arr)\\n    public static long sum(final long[] arr, final int start, final int stop)\\n    public static long[] round(final double... input)\\n    public static final LogMath LOG_2_MATH = new LogMath(2);;\\n    public static final LogMath NATURAL_LOG_MATH = new LogMath(Math.exp(1)) {\\n\\n    @Override\\n    public double getLogValue(final double nonLogValue) {\\n        return Math.log(nonLogValue);\\n    }\\n};;\\n    public static final LogMath LOG_10_MATH = new LogMath(10) {\\n\\n    @Override\\n    public double getLogValue(final double nonLogValue) {\\n        return Math.log10(nonLogValue);\\n    }\\n};;\\n    public static double klDivergance(double[] measured, double[] distribution)\\n    public static double[] permute(double[] array, RandomDataGenerator rdg)\\n    public static List<T> randomSublist(final List<T> list, final int n, final Random random)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/MathUtil.java', 'methodCount': 42, 'lineCount': 49}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Removes trailing zeros in a distribution.  The purpose of this function is to prevent other\n",
      "     * functions from evaluating in regions where the distribution has zero probability.\n",
      "     * If the array consists of only zeros or is empty, an empty array is returned.\n",
      "     *\n",
      "     * @param distribution Distribution of base qualities\n",
      "     * @return Distribution of base qualities removing any trailing zeros\n",
      "     */\n",
      " static double[] trimDistribution(final double[] distribution)\n",
      "**************************************\n",
      "static double[] trimDistribution(final double[] distribution) {\n",
      "    int endOfDistribution = distribution.length;\n",
      "    if (distribution.length == 0) {\n",
      "        return distribution;\n",
      "    }\n",
      "    while (distribution[endOfDistribution - 1] == 0) {\n",
      "        endOfDistribution--;\n",
      "        if (endOfDistribution == 0)\n",
      "            break;\n",
      "    }\n",
      "    return Arrays.copyOfRange(distribution, 0, endOfDistribution);\n",
      "}\n",
      "1235\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  67%|   | 839/1243 [28:19<13:29,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33213\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 840/1243 [28:21<14:00,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33278\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 841/1243 [28:23<14:09,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33323\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 842/1243 [28:25<14:37,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33353\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 843/1243 [28:28<14:40,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33354\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class AlleleSubsettingUtils {\\n    static final Allele NON_REF_ALLELE = Allele.create(\"<NON_REF>\");;\\n    static List<Allele> DIPLOID_NO_CALL = Arrays.asList(Allele.NO_CALL, Allele.NO_CALL);;\\n    public static VariantContext subsetVCToMatchSnp(final VariantContext ctx, final Snp snp)\\n    public static VariantContext subsetAlleles(final VariantContext originalVc, final List<Allele> allelesToKeep)\\n    public static VariantContext swapAlleles(final VariantContext originalVc, final Allele oldAllele, final Allele newAllele) throws IllegalArgumentException\\n    public static GenotypesContext subsetAlleles(final GenotypesContext originalGs, final List<Allele> originalAlleles, final List<Allele> allelesToKeep)\\n    public static int[] subsettedPLIndices(final List<Allele> originalAlleles, final List<Allele> newAlleles)\\n    private static void validateTrue(final boolean condition, final String msg)\\n    private static T nonNull(final T object, String message)\\n    private static T nonEmpty(T collection, String message)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/AlleleSubsettingUtils.java', 'methodCount': 8, 'lineCount': 12}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class MathUtil {\\n    private MathUtil();\\n    /**\\n * The double value closest to 1 while still being less than 1.\\n */\\npublic static final double MAX_PROB_BELOW_ONE = 0.9999999999999999d;;\\n    /**\\n * Constant to convert between the natural base e and 4.0.  Useful for\\n * entropy calculations on DNA.\\n */\\npublic static final double LOG_4_BASE_E = Math.log(4.0);;\\n    public static double log10_1p(final double x)\\n    public static double mean(final double[] in, final int start, final int stop)\\n    public static double mean(final double[] in)\\n    public static double stddev(final double[] in, final int start, final int length)\\n    public static double stddev(final double[] in, final int start, final int stop, final double mean)\\n    public static double stddev(final double[] in, final double mean)\\n    public static int compare(final int v1, final int v2)\\n    public static double median(final double... in)\\n    public static Double percentageOrNull(final Long numerator, final Long denominator)\\n    public static double round(final double num, final int precision)\\n    public static double divide(final double numerator, final double denominator)\\n    public static double max(final double[] nums)\\n    public static double[] capFromBelow(final double[] nums, final double floor)\\n    public static int indexOfMax(final double[] nums)\\n    public static long max(final long[] nums)\\n    public static int indexOfMax(final long[] nums)\\n    public static double[] subtractMax(final double[] logLikelihoods)\\n    public static double min(final double[] nums)\\n    public static int min(final int[] nums)\\n    public static short min(final short[] nums)\\n    public static byte min(final byte[] nums)\\n    public static double[] capFromAbove(final double[] nums, final double top)\\n    public static int indexOfMin(final int[] nums)\\n    public static int indexOfMin(final double[] nums)\\n    public static double[] seq(final double from, final double to, final double by)\\n    public static double[] promote(final int[] is)\\n    public static double[] logLikelihoodsToProbs(final double[] likelihoods)\\n    public static double[] getProbabilityFromLog(final double[] lLikelihood)\\n    public static double[] getLogFromProbability(final double[] likelihood)\\n    public static double[] pNormalizeLogProbability(final double[] lPosterior)\\n    public static double[] divide(final double[] numerators, final double[] denominators)\\n    public static double[] pNormalizeVector(final double[] pPosterior)\\n    public static double[] multiply(final double[] lhs, final double[] rhs)\\n    public static double[] sum(final double[] lhs, final double[] rhs)\\n    public static double[] sum(final double[] lhs, final double rhs)\\n    public static double sum(final double[] arr)\\n    public static long sum(final long[] arr, final int start, final int stop)\\n    public static long[] round(final double... input)\\n    public static final LogMath LOG_2_MATH = new LogMath(2);;\\n    public static final LogMath NATURAL_LOG_MATH = new LogMath(Math.exp(1)) {\\n\\n    @Override\\n    public double getLogValue(final double nonLogValue) {\\n        return Math.log(nonLogValue);\\n    }\\n};;\\n    public static final LogMath LOG_10_MATH = new LogMath(10) {\\n\\n    @Override\\n    public double getLogValue(final double nonLogValue) {\\n        return Math.log10(nonLogValue);\\n    }\\n};;\\n    public static double klDivergance(double[] measured, double[] distribution)\\n    public static double[] permute(double[] array, RandomDataGenerator rdg)\\n    public static List<T> randomSublist(final List<T> list, final int n, final Random random)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/MathUtil.java', 'methodCount': 42, 'lineCount': 49}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ThreadPoolExecutorWithExceptions {\\n    private static final Log log = Log.getInstance(ThreadPoolExecutorWithExceptions.class);;\\n    public Throwable exception = null;;\\n    public ThreadPoolExecutorWithExceptions(final int threads);\\n    protected void afterExecute(final Runnable r, Throwable t)\\n    protected void beforeExecute(Thread t, Runnable r)\\n    public boolean hasError()\\n    public void cleanUp()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/ThreadPoolExecutorWithExceptions.java', 'methodCount': 5, 'lineCount': 9}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 26}\n",
      "{'Fingerprint': 'public class Fingerprint {\\n    private final String sample;;\\n    private final Path source;;\\n    private final String info;;\\n    public Fingerprint(final String sample, final Path source, final String info);\\n    public String getSample()\\n    public Path getSource()\\n    public String getInfo()\\n    public String getPrintableId()\\n    public void add(final HaplotypeProbabilities h)\\n    public Fingerprint merge(final Fingerprint other)\\n    public static Function<FingerprintIdDetails, String> getFingerprintIdDetailsStringFunction(CrosscheckMetric.DataType CROSSCHECK_BY)\\n    public static Map<FingerprintIdDetails, Fingerprint> mergeFingerprintsBy(final Map<FingerprintIdDetails, Fingerprint> fingerprints, final Function<FingerprintIdDetails, String> by)\\n}\\n'}\n",
      "/**\n",
      "     * Loads genotypes from the supplied file into one or more Fingerprint objects and returns them in a\n",
      "     * Map of Sample->Fingerprint.\n",
      "     *\n",
      "     * @param fingerprintFile - VCF file containing genotypes for one or more samples\n",
      "     * @param specificSample  - null to load genotypes for all samples contained in the file or the name\n",
      "     *                        of an individual sample to load (and exclude all others).\n",
      "     * @return a Map of Sample name to Fingerprint\n",
      "     */\n",
      "public Map<String, Fingerprint> loadFingerprints(final Path fingerprintFile, final Path indexPath, final boolean forceIndex, final String specificSample)\n",
      "**************************************\n",
      "public Map<String, Fingerprint> loadFingerprints(final Path fingerprintFile, final Path indexPath, final boolean forceIndex, final String specificSample) {\n",
      "    final VCFFileReader reader = getVCFReader(fingerprintFile, indexPath, forceIndex);\n",
      "    checkDictionaryGoodForFingerprinting(reader.getFileHeader().getSequenceDictionary());\n",
      "    final Map<String, Fingerprint> fingerprints;\n",
      "    if (reader.isQueryable()) {\n",
      "        fingerprints = loadFingerprintsFromQueriableReader(reader, specificSample, fingerprintFile);\n",
      "    } else {\n",
      "        log.warn(\"Couldn't find index for file \" + fingerprintFile + \" going to read through it all.\");\n",
      "        fingerprints = loadFingerprintsFromVariantContexts(reader, specificSample, fingerprintFile);\n",
      "    }\n",
      "    reader.getFileHeader().getGenotypeSamples().forEach(sample -> fingerprints.computeIfAbsent(sample, s -> new Fingerprint(s, fingerprintFile, null)));\n",
      "    return fingerprints;\n",
      "}\n",
      "2022\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 844/1243 [28:30<14:42,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33360\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 845/1243 [28:32<15:10,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33362\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 846/1243 [28:35<16:12,  2.45s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33370\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 847/1243 [28:37<15:30,  2.35s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33416\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 848/1243 [28:40<15:22,  2.34s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33492\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 7}\n",
      "{'PedFile': 'public class PedFile {\\n    private static final Log log = Log.getInstance(PedFile.class);;\\n    static final Pattern WHITESPACE = Pattern.compile(\"\\\\\\\\s+\");;\\n    static final Pattern TAB = Pattern.compile(\"\\\\\\\\t\");;\\n    private final Pattern delimiterPattern;;\\n    // A textual representation of the delimiter, for output purposes\\nprivate final String delimiterString;;\\n    // These two are really for PedTrio, but they can\\'t be static in there and need to be accessed outside of PedFile\\npublic static final Number NO_PHENO = Integer.valueOf(-9);;\\n    public static final Sex UNKNOWN_SEX = Sex.Unknown;;\\n    public PedFile(final boolean isTabMode);\\n    public void add(final PedTrio trio)\\n    public void write(final File file)\\n    public static PedFile fromFile(final File file, final boolean isTabMode)\\n    public PedFile removeIncompleteTrios()\\n    public static PedFile fromSexMap(final Map<String, Sex> sampleSexes)\\n}\\n'}\n",
      "/**\n",
      "     * Scans through the pedigrees and removes all entries that do not have both paternal and maternal ids set.\n",
      "     */\n",
      "public PedFile removeIncompleteTrios()\n",
      "**************************************\n",
      "public PedFile removeIncompleteTrios() {\n",
      "    final Iterator<Map.Entry<String, PedTrio>> iterator = entrySet().iterator();\n",
      "    while (iterator.hasNext()) {\n",
      "        if (!iterator.next().getValue().hasBothParents())\n",
      "            iterator.remove();\n",
      "    }\n",
      "    return this;\n",
      "}\n",
      "355\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 849/1243 [28:41<13:56,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33538\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 850/1243 [28:43<12:45,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33561\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  68%|   | 851/1243 [28:45<12:21,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33587\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 852/1243 [28:46<12:11,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33593\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 853/1243 [28:53<20:58,  3.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33634\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 22, 'classCount': 1, 'methodsOnly': 'public class CommandLineProgram {\\n    private static String PROPERTY_USE_LEGACY_PARSER = \"picard.useLegacyParser\";;\\n    private static String PROPERTY_CONVERT_LEGACY_COMMAND_LINE = \"picard.convertCommandLine\";;\\n    private static Boolean useLegacyParser;;\\n    public static String SYNTAX_TRANSITION_URL = \"https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\";;\\n    /**\\n * CommandLineProgramProperties oneLineSummary attribute must be shorted than this in order to maintain\\n * reasonable help output formatting.\\n */\\npublic static int MAX_ALLOWABLE_ONE_LINE_SUMMARY_LENGTH = 120;;\\n    @Argument(doc = \"One or more directories with space available to be used by this program for temporary storage of working files\", common = true, optional = true)\\npublic List<File> TMP_DIR = new ArrayList<>();;\\n    @Argument(doc = \"Control verbosity of logging.\", common = true)\\npublic Log.LogLevel VERBOSITY = Log.LogLevel.INFO;;\\n    @Argument(doc = \"Whether to suppress job-summary info on System.err.\", common = true)\\npublic Boolean QUIET = false;;\\n    @Argument(doc = \"Validation stringency for all SAM files read by this program.  Setting stringency to SILENT \" + \"can improve performance when processing a BAM file in which variable-length data (read, qualities, tags) \" + \"do not otherwise need to be decoded.\", common = true)\\npublic ValidationStringency VALIDATION_STRINGENCY = ValidationStringency.DEFAULT_STRINGENCY;;\\n    @Argument(doc = \"Compression level for all compressed files created (e.g. BAM and VCF).\", common = true)\\npublic int COMPRESSION_LEVEL = Defaults.COMPRESSION_LEVEL;;\\n    @Argument(doc = \"When writing files that need to be sorted, this will specify the number of records stored in RAM before spilling to disk. \" + \"Increasing this number reduces the number of file handles needed to sort the file, and increases the amount of RAM needed.\", optional = true, common = true)\\npublic Integer MAX_RECORDS_IN_RAM = SAMFileWriterImpl.getDefaultMaxRecordsInRam();;\\n    @Argument(doc = \"Whether to create an index when writing VCF or coordinate sorted BAM output.\", common = true)\\npublic Boolean CREATE_INDEX = Defaults.CREATE_INDEX;;\\n    @Argument(doc = \"Whether to create an MD5 digest for any BAM or FASTQ files created.  \", common = true)\\npublic boolean CREATE_MD5_FILE = Defaults.CREATE_MD5;;\\n    @ArgumentCollection\\npublic ReferenceArgumentCollection referenceSequence = makeReferenceArgumentCollection();;\\n    // This is retained for compatibility with existing code that depends on accessing it, and is populated\\n// after argument parsing using the value established by the user in the referenceSequence argument collection.\\nprotected File REFERENCE_SEQUENCE = Defaults.REFERENCE_FASTA;;\\n    @ArgumentCollection(doc = \"Special Arguments that have meaning to the argument parsing system.  \" + \"It is unlikely these will ever need to be accessed by the command line program\")\\npublic Object specialArgumentsCollection = useLegacyParser() ? // legacy parser does not require these\\nnew Object() : new SpecialArgumentsCollection();;\\n    @Argument(shortName = \"use_jdk_deflater\", doc = \"Use the JDK Deflater instead of the Intel Deflater for writing compressed output\", common = true)\\npublic Boolean USE_JDK_DEFLATER = false;;\\n    @Argument(shortName = \"use_jdk_inflater\", doc = \"Use the JDK Inflater instead of the Intel Inflater for reading compressed input\", common = true)\\npublic Boolean USE_JDK_INFLATER = false;;\\n    private static final String[] PACKAGES_WITH_WEB_DOCUMENTATION = { \"picard\" };;\\n    /**\\n * Initialized in parseArgs.  Subclasses may want to access this to do their\\n * own validation, and then print usage using commandLineParser.\\n */\\nprivate CommandLineParser commandLineParser;;\\n    private final List<Header> defaultHeaders = new ArrayList<>();;\\n    /**\\n * The reconstructed commandline used to run this program. Used for logging\\n * and debugging.\\n */\\nprivate String commandLine;;\\n    protected abstract int doWork()\\n    protected boolean requiresReference()\\n    protected ReferenceArgumentCollection makeReferenceArgumentCollection()\\n    public void instanceMainWithExit(final String[] argv)\\n    public int instanceMain(final String[] argv)\\n    protected String[] customCommandLineValidation()\\n    protected boolean parseArgs(final String[] argv)\\n    protected MetricsFile<A, B> getMetricsFile()\\n    public String getStandardUsagePreamble()\\n    public CommandLineParser getCommandLineParserForArgs(String[] argv)\\n    public CommandLineParser getCommandLineParser()\\n    public static boolean useLegacyParser()\\n    public String getVersion()\\n    public String getCommandLine()\\n    public void setDefaultHeaders(final List<Header> headers)\\n    public SAMProgramRecord getPGRecord(final SAMFileHeader header)\\n    public List<Header> getDefaultHeaders()\\n    public static String getStandardUsagePreamble(final Class<?> mainClass)\\n    public static boolean hasWebDocumentation(final Class<?> clazz)\\n    public static String getFaqLink()\\n    public static boolean checkRInstallation(final boolean has_chart_output)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/cmdline/CommandLineProgram.java', 'methodCount': 21, 'lineCount': 45}, {'fieldCount': 18, 'classCount': 1, 'methodsOnly': 'public class StandardOptionDefinitions {\\n    public static final String INPUT_SHORT_NAME = \"I\";;\\n    public static final String OUTPUT_SHORT_NAME = \"O\";;\\n    public static final String REFERENCE_SHORT_NAME = \"R\";;\\n    public static final String SAMPLE_ALIAS_SHORT_NAME = \"ALIAS\";;\\n    public static final String LIBRARY_NAME_SHORT_NAME = \"LIB\";;\\n    public static final String EXPECTED_INSERT_SIZE_SHORT_NAME = \"INSERT\";;\\n    public static final String LANE_SHORT_NAME = \"L\";;\\n    public static final String SEQUENCE_DICTIONARY_SHORT_NAME = \"SD\";;\\n    public static final String METRICS_FILE_SHORT_NAME = \"M\";;\\n    public static final String ASSUME_SORTED_SHORT_NAME = \"AS\";;\\n    public static final String ASSUME_SORT_ORDER_SHORT_NAME = \"ASO\";;\\n    public static final String PF_READS_ONLY_SHORT_NAME = \"PF\";;\\n    public static final String MINIMUM_MAPPING_QUALITY_SHORT_NAME = \"MQ\";;\\n    public static final String READ_GROUP_ID_SHORT_NAME = \"RG\";;\\n    public static final String PROGRAM_RECORD_ID_SHORT_NAME = \"PG\";;\\n    public static final String MINIMUM_LOD_SHORT_NAME = \"LOD\";;\\n    public static final String SORT_ORDER_SHORT_NAME = \"SO\";;\\n    public static final String USE_ORIGINAL_QUALITIES_SHORT_NAME = \"OQ\";;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/cmdline/StandardOptionDefinitions.java', 'methodCount': 0, 'lineCount': 20}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ReadDataManipulationProgramGroup {\\n    public String getName()\\n    public String getDescription()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/cmdline/programgroups/ReadDataManipulationProgramGroup.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Write out a file of read names for debugging purposes.\n",
      "     *\n",
      "     * @param samOrBamFile The SAM/BAM/CRAM file for which we are going to write out a file of its\n",
      "     *                     containing read names\n",
      "     */\n",
      "private void writeReadsFile(final File samOrBamFile) throws IOException\n",
      "**************************************\n",
      "private void writeReadsFile(final File samOrBamFile) throws IOException {\n",
      "    final File readsFile = new File(OUTPUT.getParentFile(), IOUtil.basename(samOrBamFile) + \".reads\");\n",
      "    IOUtil.assertFileIsWritable(readsFile);\n",
      "    try (final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(samOrBamFile);\n",
      "        final BufferedWriter bw = IOUtil.openFileForBufferedWriting(readsFile, false)) {\n",
      "        for (final SAMRecord rec : reader) {\n",
      "            bw.write(rec.toString() + \"\\n\");\n",
      "        }\n",
      "    }\n",
      "    IOUtil.assertFileIsReadable(readsFile);\n",
      "}\n",
      "1907\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 854/1243 [28:55<19:06,  2.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33646\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 855/1243 [28:57<16:49,  2.60s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33648\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 856/1243 [28:59<15:58,  2.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33663\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 857/1243 [29:01<15:51,  2.46s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33668\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 858/1243 [29:03<14:32,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33691\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class PicardException {\\n    public PicardException(final String message);\\n    public PicardException(final String message, final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/PicardException.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Copies alignment info from aligned to unaligned read, if there is an alignment, and sets mate information.\n",
      "     *\n",
      "     * @param firstUnaligned  Original first of pair, into which alignment and pair info will be written.\n",
      "     * @param secondUnaligned Original second of pair, into which alignment and pair info will be written.\n",
      "     * @param firstAligned    Aligned first of pair, or null if no alignment.\n",
      "     * @param secondAligned   Aligned second of pair, or null if no alignment.\n",
      "     * @param isContaminant Should this pair be unmapped due to contamination?\n",
      "     */\n",
      "private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, final SAMRecord secondUnaligned, final SAMRecord firstAligned, final SAMRecord secondAligned, final boolean isContaminant, final boolean needsSafeReverseComplement)\n",
      "**************************************\n",
      "private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, final SAMRecord secondUnaligned, final SAMRecord firstAligned, final SAMRecord secondAligned, final boolean isContaminant, final boolean needsSafeReverseComplement) {\n",
      "    if (firstAligned != null) {\n",
      "        transferAlignmentInfoToFragment(firstUnaligned, firstAligned, isContaminant, needsSafeReverseComplement);\n",
      "    }\n",
      "    if (secondAligned != null) {\n",
      "        transferAlignmentInfoToFragment(secondUnaligned, secondAligned, isContaminant, needsSafeReverseComplement);\n",
      "    }\n",
      "    if (isClipOverlappingReads()) {\n",
      "        clipForOverlappingReads(firstUnaligned, secondUnaligned, hardClipOverlappingReads);\n",
      "    }\n",
      "    SamPairUtil.setMateInfo(secondUnaligned, firstUnaligned, addMateCigar);\n",
      "    if (!keepAlignerProperPairFlags) {\n",
      "        SamPairUtil.setProperPairFlags(secondUnaligned, firstUnaligned, expectedOrientations);\n",
      "    }\n",
      "}\n",
      "521\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 859/1243 [29:06<14:37,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33719\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 860/1243 [29:08<15:26,  2.42s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33720\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 861/1243 [29:13<20:04,  3.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33733\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 862/1243 [29:15<17:11,  2.71s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33743\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  69%|   | 863/1243 [29:16<15:03,  2.38s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33753\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class Pair {\\n    private X left;;\\n    private Y right;;\\n    public Pair(final X left, final Y right);\\n    public X getLeft()\\n    public Y getRight()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n    public int compareTo(final Pair<X, Y> o)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/sam/util/Pair.java', 'methodCount': 7, 'lineCount': 11}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public interface PhysicalLocation {\\n    public static int NO_VALUE = -1;;\\n    public short getReadGroup()\\n    public void setReadGroup(short rg)\\n    public short getTile()\\n    public void setTile(short tile)\\n    public int getX()\\n    public void setX(int x)\\n    public int getY()\\n    public void setY(int y)\\n    public short getLibraryId()\\n    public void setLibraryId(short libraryId)\\n    public boolean hasLocation()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/sam/util/PhysicalLocation.java', 'methodCount': 11, 'lineCount': 14}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class PhysicalLocationInt {\\n    public short tile = -1;;\\n    public int x = -1, y = -1;;\\n    public short getReadGroup()\\n    public void setReadGroup(final short readGroup)\\n    public short getTile()\\n    public void setTile(final short tile)\\n    public int getX()\\n    public void setX(final int x)\\n    public int getY()\\n    public void setY(final int y)\\n    public short getLibraryId()\\n    public void setLibraryId(final short libraryId)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/sam/util/PhysicalLocationInt.java', 'methodCount': 10, 'lineCount': 14}, {'fieldCount': 9, 'classCount': 1, 'methodsOnly': 'public class ReadNameParser {\\n    /**\\n * The read name regular expression (regex) is used to extract three pieces of information from the read name: tile, x location,\\n * and y location.  Any read name regex should parse the read name to produce these and only these values.  An example regex is:\\n *  (?:.*:)?([0-9]+)[^:]*:([0-9]+)[^:]*:([0-9]+)[^:]*$\\n * which assumes that fields in the read name are delimited by \\':\\' and the last three fields correspond to the tile, x and y locations,\\n * ignoring any trailing non-digit characters.\\n *\\n * The default regex is optimized for fast parsing (see {@link #getLastThreeFields(String, char, int[])}) by searching for the last\\n * three fields, ignoring any trailing non-digit characters, assuming the delimiter \\':\\'.  This should consider correctly read names\\n * where we have 5 or 7 field with the last three fields being tile/x/y, as is the case for the majority of read names produced by\\n * Illumina technology.\\n */\\npublic static final String DEFAULT_READ_NAME_REGEX = \"<optimized capture of last three \\':\\' separated fields as numeric values>\".intern();;\\n    private String readNameStored = null;;\\n    private PhysicalLocation physicalLocationStored = new PhysicalLocationInt();;\\n    // for optimization of addLocationInformation\\nprivate final int[] tmpLocationFields = new int[3];;\\n    // was the regex default?\\nprivate final boolean useOptimizedDefaultParsing;;\\n    protected final String readNameRegex;;\\n    private Pattern readNamePattern;;\\n    private boolean warnedAboutRegexNotMatching = false;;\\n    private transient final Log log;;\\n    public ReadNameParser();\\n    public ReadNameParser(final String readNameRegex);\\n    public ReadNameParser(final String readNameRegex, final Log log);\\n    private boolean readLocationInformation(final String readName, final PhysicalLocation loc)\\n    public boolean addLocationInformation(final String readName, final PhysicalLocation loc)\\n    public static int getLastThreeFields(final String readName, final char delim, final int[] tokens) throws NumberFormatException\\n    public static int rapidParseInt(final String input) throws NumberFormatException\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/sam/util/ReadNameParser.java', 'methodCount': 7, 'lineCount': 18}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Get the one-based cycle number of the base, taking the direction of the read into account\n",
      "     */\n",
      "private static int stratifyCycle(final RecordAndOffset recordAndOffset)\n",
      "**************************************\n",
      "private static int stratifyCycle(final RecordAndOffset recordAndOffset) {\n",
      "    final SAMRecord rec = recordAndOffset.getRecord();\n",
      "    final int offset = recordAndOffset.getOffset();\n",
      "    int retval = rec.getReadNegativeStrandFlag() ? (rec.getReadLength() - offset - 1) : offset;\n",
      "    retval += 1;\n",
      "    return retval;\n",
      "}\n",
      "1211\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 864/1243 [29:19<14:41,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33788\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 865/1243 [29:21<14:23,  2.28s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33800\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 866/1243 [29:23<15:00,  2.39s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33805\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 867/1243 [29:27<16:45,  2.67s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33808\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 868/1243 [29:29<15:47,  2.53s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33815\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MergeableMetricBase {\\n    public boolean canMerge(final MergeableMetricBase other)\\n    public boolean mergeIfCan(final MergeableMetricBase other)\\n    public MergeableMetricBase merge(final Collection<? extends MergeableMetricBase> many)\\n    public MergeableMetricBase merge(final MergeableMetricBase other)\\n    private static List<Field> getAllFields(Class clazz)\\n    public void calculateDerivedFields()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/analysis/MergeableMetricBase.java', 'methodCount': 6, 'lineCount': 8}, {'fieldCount': 28, 'classCount': 1, 'methodsOnly': 'public class HelpConstants {\\n    private HelpConstants();\\n    /**\\n * Definition of the group names / descriptions for documentation/help purposes.\\n */\\npublic final static String DOC_CAT_BASE_CALLING = \"Base Calling\";;\\n    public final static String DOC_CAT_BASE_CALLING_SUMMARY = \"Tools that process sequencing machine data, e.g. Illumina base calls, and detect sequencing level attributes, e.g. adapters\";;\\n    public final static String DOC_CAT_DIAGNOSTICS_AND_QC = \"Diagnostics and Quality Control\";;\\n    public final static String DOC_CAT_DIAGNOSTICS_AND_QC_SUMMARY = \"Tools that collect sequencing quality related and comparative metrics\";;\\n    public final static String DOC_CAT_INTERVALS_MANIPULATION = \"Intervals Manipulation\";;\\n    public final static String DOC_CAT_INTERVALS_MANIPULATION_SUMMARY = \"Tools that process genomic intervals in various formats\";;\\n    public final static String DOC_CAT_OTHER = \"Other\";;\\n    public final static String DOC_CAT_OTHER_SUMMARY = \"Miscellaneous tools, e.g. those that aid in data streaming\";;\\n    public final static String DOC_CAT_READ_DATA_MANIPULATION = \"Read Data Manipulation\";;\\n    public final static String DOC_CAT_READ_DATA_MANIPULATION_SUMMARY = \"Tools that manipulate read data in SAM, BAM or CRAM format\";;\\n    public final static String DOC_CAT_REFERENCE = \"Reference\";;\\n    public final static String DOC_CAT_REFERENCE_SUMMARY = \"Tools that analyze and manipulate FASTA format references\";;\\n    public final static String DOC_CAT_TEST = \"Test Tools\";;\\n    public final static String DOC_CAT_TEST_SUMMARY = \"Tools for internal test purposes\";;\\n    public final static String DOC_CAT_VARIANT_FILTERING = \"Variant Filtering\";;\\n    public final static String DOC_CAT_VARIANT_FILTERING_SUMMARY = \"Tools that filter variants by annotating the FILTER column\";;\\n    public final static String DOC_CAT_VARIANT_EVALUATION = \"Variant Evaluation and Refinement\";;\\n    public final static String DOC_CAT_VARIANT_EVALUATION_SUMMARY = \"Tools that evaluate and refine variant calls, e.g. with annotations not offered by the engine\";;\\n    public final static String DOC_CAT_VARIANT_MANIPULATION = \"Variant Manipulation\";;\\n    public final static String DOC_CAT_VARIANT_MANIPULATION_SUMMARY = \"Tools that manipulate variant call format (VCF) data\";;\\n    public final static String DOC_CAT_GENOTYPING_ARRAYS_MANIPULATION = \"Genotyping Arrays Manipulation\";;\\n    public final static String DOC_CAT_GENOTYPING_ARRAYS_MANIPULATION_SUMMARY = \"Tools that manipulate data generated by Genotyping arrays\";;\\n    public final static String DOC_CAT_METRICS = \"Metrics\";;\\n    public final static String DOC_CAT_METRICS_SUMMARY = \"Metrics\";;\\n    /**\\n * List of \"supercategory\" values used for doc purposes. Every doc group name can/should be put into\\n * one of the following supercategories.\\n */\\npublic final static String DOC_SUPERCAT_TOOLS = \"tools\";;\\n    public final static String DOC_SUPERCAT_UTILITIES = \"utilities\";;\\n    public final static String DOC_SUPERCAT_EXCLUDE = \"exclude\";;\\n    private static Map<String, String> groupToSuperCategory;;\\n    public static Map<String, String> getSuperCategoryMap()\\n    public static String getSuperCategoryProperty(final String groupName)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/help/HelpConstants.java', 'methodCount': 3, 'lineCount': 33}, {'fieldCount': 14, 'classCount': 1, 'methodsOnly': 'public class AbstractMarkDuplicatesCommandLineProgram {\\n    @ArgumentCollection\\nprotected final PGTagArgumentCollection pgTagArgumentCollection = new PGTagArgumentCollection();;\\n    @Argument(shortName = StandardOptionDefinitions.INPUT_SHORT_NAME, doc = \"One or more input SAM, BAM or CRAM files to analyze. Must be coordinate sorted.\")\\npublic List<String> INPUT;;\\n    @Argument(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = \"The output file to write marked records to\")\\npublic File OUTPUT;;\\n    @Argument(shortName = \"M\", doc = \"File to write duplication metrics to\")\\npublic File METRICS_FILE;;\\n    @Argument(doc = \"If true do not write duplicates to the output file instead of writing them with appropriate flags set.\")\\npublic boolean REMOVE_DUPLICATES = false;;\\n    @Deprecated\\n@Argument(shortName = StandardOptionDefinitions.ASSUME_SORTED_SHORT_NAME, doc = \"If true, assume that the input file is coordinate sorted even if the header says otherwise. \" + \"Deprecated, used ASSUME_SORT_ORDER=coordinate instead.\", mutex = { \"ASSUME_SORT_ORDER\" })\\npublic boolean ASSUME_SORTED = false;;\\n    @Argument(shortName = StandardOptionDefinitions.ASSUME_SORT_ORDER_SHORT_NAME, doc = \"If not null, assume that the input file has this order even if the header says otherwise.\", optional = true, mutex = { \"ASSUME_SORTED\" })\\npublic SAMFileHeader.SortOrder ASSUME_SORT_ORDER = null;;\\n    @Argument(shortName = \"DS\", doc = \"The scoring strategy for choosing the non-duplicate among candidates.\")\\npublic ScoringStrategy DUPLICATE_SCORING_STRATEGY = ScoringStrategy.TOTAL_MAPPED_REFERENCE_LENGTH;;\\n    @Argument(shortName = StandardOptionDefinitions.PROGRAM_RECORD_ID_SHORT_NAME, doc = \"The program record ID for the @PG record(s) created by this program. Set to null to disable \" + \"PG record creation.  This string may have a suffix appended to avoid collision with other \" + \"program record IDs.\", optional = true)\\npublic String PROGRAM_RECORD_ID = \"MarkDuplicates\";;\\n    @Argument(shortName = \"PG_VERSION\", doc = \"Value of VN tag of PG record to be created. If not specified, the version will be detected automatically.\", optional = true)\\npublic String PROGRAM_GROUP_VERSION;;\\n    @Argument(shortName = \"PG_COMMAND\", doc = \"Value of CL tag of PG record to be created. If not supplied the command line will be detected automatically.\", optional = true)\\npublic String PROGRAM_GROUP_COMMAND_LINE;;\\n    @Argument(shortName = \"PG_NAME\", doc = \"Value of PN tag of PG record to be created.\")\\npublic String PROGRAM_GROUP_NAME = getClass().getSimpleName();;\\n    @Argument(shortName = \"CO\", doc = \"Comment(s) to include in the output file\\'s header.\", optional = true)\\npublic List<String> COMMENT = new ArrayList<>();;\\n    /**\\n * The program groups that have been seen during the course of examining the input records.\\n */\\nprotected final Set<String> pgIdsSeen = new HashSet<>();;\\n    protected Map<String, String> getChainedPgIds(final SAMFileHeader outputHeader)\\n    public static void finalizeAndWriteMetrics(final LibraryIdGenerator libraryIdGenerator, final MetricsFile<DuplicationMetrics, Double> metricsFile, final File outputFile)\\n    public static DuplicationMetrics addReadToLibraryMetrics(final SAMRecord rec, final SAMFileHeader header, final LibraryIdGenerator libraryIdGenerator, final boolean flowMetrics)\\n    protected SamHeaderAndIterator openInputs(boolean eagerlyDecode)\\n    public static void trackOpticalDuplicates(List<? extends ReadEnds> ends, final ReadEnds keeper, final OpticalDuplicateFinder opticalDuplicateFinder, final LibraryIdGenerator libraryIdGenerator)\\n    public static void addSingletonToCount(final LibraryIdGenerator libraryIdGenerator)\\n    private static int trackOpticalDuplicates(final List<? extends ReadEnds> list, final ReadEnds keeper, final OpticalDuplicateFinder opticalDuplicateFinder, final Histogram<Short> opticalDuplicatesByLibraryId)\\n    private static void trackDuplicateCounts(final int listSize, final int optDupCnt, final LibraryIdGenerator libraryIdGenerator)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/sam/markduplicates/util/AbstractMarkDuplicatesCommandLineProgram.java', 'methodCount': 8, 'lineCount': 24}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class MathUtil {\\n    private MathUtil();\\n    /**\\n * The double value closest to 1 while still being less than 1.\\n */\\npublic static final double MAX_PROB_BELOW_ONE = 0.9999999999999999d;;\\n    /**\\n * Constant to convert between the natural base e and 4.0.  Useful for\\n * entropy calculations on DNA.\\n */\\npublic static final double LOG_4_BASE_E = Math.log(4.0);;\\n    public static double log10_1p(final double x)\\n    public static double mean(final double[] in, final int start, final int stop)\\n    public static double mean(final double[] in)\\n    public static double stddev(final double[] in, final int start, final int length)\\n    public static double stddev(final double[] in, final int start, final int stop, final double mean)\\n    public static double stddev(final double[] in, final double mean)\\n    public static int compare(final int v1, final int v2)\\n    public static double median(final double... in)\\n    public static Double percentageOrNull(final Long numerator, final Long denominator)\\n    public static double round(final double num, final int precision)\\n    public static double divide(final double numerator, final double denominator)\\n    public static double max(final double[] nums)\\n    public static double[] capFromBelow(final double[] nums, final double floor)\\n    public static int indexOfMax(final double[] nums)\\n    public static long max(final long[] nums)\\n    public static int indexOfMax(final long[] nums)\\n    public static double[] subtractMax(final double[] logLikelihoods)\\n    public static double min(final double[] nums)\\n    public static int min(final int[] nums)\\n    public static short min(final short[] nums)\\n    public static byte min(final byte[] nums)\\n    public static double[] capFromAbove(final double[] nums, final double top)\\n    public static int indexOfMin(final int[] nums)\\n    public static int indexOfMin(final double[] nums)\\n    public static double[] seq(final double from, final double to, final double by)\\n    public static double[] promote(final int[] is)\\n    public static double[] logLikelihoodsToProbs(final double[] likelihoods)\\n    public static double[] getProbabilityFromLog(final double[] lLikelihood)\\n    public static double[] getLogFromProbability(final double[] likelihood)\\n    public static double[] pNormalizeLogProbability(final double[] lPosterior)\\n    public static double[] divide(final double[] numerators, final double[] denominators)\\n    public static double[] pNormalizeVector(final double[] pPosterior)\\n    public static double[] multiply(final double[] lhs, final double[] rhs)\\n    public static double[] sum(final double[] lhs, final double[] rhs)\\n    public static double[] sum(final double[] lhs, final double rhs)\\n    public static double sum(final double[] arr)\\n    public static long sum(final long[] arr, final int start, final int stop)\\n    public static long[] round(final double... input)\\n    public static final LogMath LOG_2_MATH = new LogMath(2);;\\n    public static final LogMath NATURAL_LOG_MATH = new LogMath(Math.exp(1)) {\\n\\n    @Override\\n    public double getLogValue(final double nonLogValue) {\\n        return Math.log(nonLogValue);\\n    }\\n};;\\n    public static final LogMath LOG_10_MATH = new LogMath(10) {\\n\\n    @Override\\n    public double getLogValue(final double nonLogValue) {\\n        return Math.log10(nonLogValue);\\n    }\\n};;\\n    public static double klDivergance(double[] measured, double[] distribution)\\n    public static double[] permute(double[] array, RandomDataGenerator rdg)\\n    public static List<T> randomSublist(final List<T> list, final int n, final Random random)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/util/MathUtil.java', 'methodCount': 42, 'lineCount': 49}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Adds duplicated read to the metrics\n",
      "     */\n",
      "public void addDuplicateReadToMetrics(final SAMRecord rec)\n",
      "**************************************\n",
      "public void addDuplicateReadToMetrics(final SAMRecord rec) {\n",
      "    if (!rec.isSecondaryOrSupplementary() && !rec.getReadUnmappedFlag()) {\n",
      "        if (!rec.getReadPairedFlag() || rec.getMateUnmappedFlag()) {\n",
      "            ++UNPAIRED_READ_DUPLICATES;\n",
      "        } else {\n",
      "            ++READ_PAIR_DUPLICATES;\n",
      "        }\n",
      "    }\n",
      "}\n",
      "3003\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 869/1243 [29:31<15:11,  2.44s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33816\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 870/1243 [29:33<13:51,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33840\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 871/1243 [29:36<14:28,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33879\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 872/1243 [29:37<13:03,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33889\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 873/1243 [29:39<12:25,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33969\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class OpticalDuplicateFinder {\\n    public int opticalDuplicatePixelDistance;;\\n    public static final int DEFAULT_OPTICAL_DUPLICATE_DISTANCE = 100;;\\n    public static final int DEFAULT_BIG_DUPLICATE_SET_SIZE = 1000;;\\n    // larger than this number will generate over 100 billion comparisons in the n^2 algorithm below\\npublic static final int DEFAULT_MAX_DUPLICATE_SET_SIZE = 300000;;\\n    private int bigDuplicateSetSize = DEFAULT_BIG_DUPLICATE_SET_SIZE;;\\n    private long maxDuplicateSetSize = DEFAULT_MAX_DUPLICATE_SET_SIZE;;\\n    public void setBigDuplicateSetSize(final int bigDuplicateSetSize)\\n    public void setMaxDuplicateSetSize(final long maxDuplicateSetSize)\\n    public OpticalDuplicateFinder();\\n    public OpticalDuplicateFinder(final String readNameRegex, final int opticalDuplicatePixelDistance, final Log log);\\n    public OpticalDuplicateFinder(final String readNameRegex, final int opticalDuplicatePixelDistance, final long maxDuplicateSetSize, final Log log);\\n    public boolean[] findOpticalDuplicates(final List<? extends PhysicalLocation> list, final PhysicalLocation keeper)\\n    private boolean[] getOpticalDuplicatesFlagFast(List<? extends PhysicalLocation> list, PhysicalLocation actualKeeper, boolean[] opticalDuplicateFlags, Log log, ProgressLogger progressLoggerForKeeper, ProgressLogger progressLoggerForRest, boolean logProgress)\\n    private boolean[] getOpticalDuplicatesFlagWithGraph(List<? extends PhysicalLocation> list, PhysicalLocation keeper, boolean[] opticalDuplicateFlags, Log log, ProgressLogger progressLoggerForKeeper, ProgressLogger progressLoggerForRest, boolean logProgress)\\n    private void fillGraphFromAGroup(final List<? extends PhysicalLocation> wholeList, final List<Integer> groupList, final boolean logProgress, final ProgressLogger progressLoggerForKeeper, final int distance, final GraphUtils.Graph<Integer> opticalDistanceRelationGraph)\\n    private PhysicalLocation keeperOrNull(final List<? extends PhysicalLocation> list, final PhysicalLocation keeper)\\n    private boolean closeEnough(final PhysicalLocation lhs, final PhysicalLocation rhs, final int distance)\\n    private boolean closeEnoughShort(final PhysicalLocation lhs, final PhysicalLocation rhs, final int distance)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,picard/src/main/java/picard/sam/markduplicates/util/OpticalDuplicateFinder.java', 'methodCount': 12, 'lineCount': 20}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Compare bases corresponding to the hashes.\n",
      "     *\n",
      "     * @return errors number for current part of the read\n",
      "     */\n",
      "private int compareHashes(byte[] read1, byte[] read2, int hashNumber)\n",
      "**************************************\n",
      "private int compareHashes(byte[] read1, byte[] read2, int hashNumber) {\n",
      "    int errors = 0;\n",
      "    int position = minIdenticalBases + hashNumber;\n",
      "    while (position < minReadLenInGroup) {\n",
      "        if (read1[position] != read2[position]) {\n",
      "            errors++;\n",
      "        }\n",
      "        position += numberOfHashesInGroup;\n",
      "    }\n",
      "    return errors;\n",
      "}\n",
      "695\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 874/1243 [29:41<12:59,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33979\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 875/1243 [29:44<13:25,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34005\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  70%|   | 876/1243 [29:46<13:05,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34006\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 877/1243 [29:48<12:49,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34078\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 878/1243 [29:50<13:06,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34124\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 13}\n",
      "{'CommandLineProgram': 'public class CommandLineProgram {\\n    private static String PROPERTY_USE_LEGACY_PARSER = \"picard.useLegacyParser\";;\\n    private static String PROPERTY_CONVERT_LEGACY_COMMAND_LINE = \"picard.convertCommandLine\";;\\n    private static Boolean useLegacyParser;;\\n    public static String SYNTAX_TRANSITION_URL = \"https://github.com/broadinstitute/picard/wiki/Command-Line-Syntax-Transition-For-Users-(Pre-Transition)\";;\\n    /**\\n * CommandLineProgramProperties oneLineSummary attribute must be shorted than this in order to maintain\\n * reasonable help output formatting.\\n */\\npublic static int MAX_ALLOWABLE_ONE_LINE_SUMMARY_LENGTH = 120;;\\n    @Argument(doc = \"One or more directories with space available to be used by this program for temporary storage of working files\", common = true, optional = true)\\npublic List<File> TMP_DIR = new ArrayList<>();;\\n    @Argument(doc = \"Control verbosity of logging.\", common = true)\\npublic Log.LogLevel VERBOSITY = Log.LogLevel.INFO;;\\n    @Argument(doc = \"Whether to suppress job-summary info on System.err.\", common = true)\\npublic Boolean QUIET = false;;\\n    @Argument(doc = \"Validation stringency for all SAM files read by this program.  Setting stringency to SILENT \" + \"can improve performance when processing a BAM file in which variable-length data (read, qualities, tags) \" + \"do not otherwise need to be decoded.\", common = true)\\npublic ValidationStringency VALIDATION_STRINGENCY = ValidationStringency.DEFAULT_STRINGENCY;;\\n    @Argument(doc = \"Compression level for all compressed files created (e.g. BAM and VCF).\", common = true)\\npublic int COMPRESSION_LEVEL = Defaults.COMPRESSION_LEVEL;;\\n    @Argument(doc = \"When writing files that need to be sorted, this will specify the number of records stored in RAM before spilling to disk. \" + \"Increasing this number reduces the number of file handles needed to sort the file, and increases the amount of RAM needed.\", optional = true, common = true)\\npublic Integer MAX_RECORDS_IN_RAM = SAMFileWriterImpl.getDefaultMaxRecordsInRam();;\\n    @Argument(doc = \"Whether to create an index when writing VCF or coordinate sorted BAM output.\", common = true)\\npublic Boolean CREATE_INDEX = Defaults.CREATE_INDEX;;\\n    @Argument(doc = \"Whether to create an MD5 digest for any BAM or FASTQ files created.  \", common = true)\\npublic boolean CREATE_MD5_FILE = Defaults.CREATE_MD5;;\\n    @ArgumentCollection\\npublic ReferenceArgumentCollection referenceSequence = makeReferenceArgumentCollection();;\\n    // This is retained for compatibility with existing code that depends on accessing it, and is populated\\n// after argument parsing using the value established by the user in the referenceSequence argument collection.\\nprotected File REFERENCE_SEQUENCE = Defaults.REFERENCE_FASTA;;\\n    @ArgumentCollection(doc = \"Special Arguments that have meaning to the argument parsing system.  \" + \"It is unlikely these will ever need to be accessed by the command line program\")\\npublic Object specialArgumentsCollection = useLegacyParser() ? // legacy parser does not require these\\nnew Object() : new SpecialArgumentsCollection();;\\n    @Argument(shortName = \"use_jdk_deflater\", doc = \"Use the JDK Deflater instead of the Intel Deflater for writing compressed output\", common = true)\\npublic Boolean USE_JDK_DEFLATER = false;;\\n    @Argument(shortName = \"use_jdk_inflater\", doc = \"Use the JDK Inflater instead of the Intel Inflater for reading compressed input\", common = true)\\npublic Boolean USE_JDK_INFLATER = false;;\\n    private static final String[] PACKAGES_WITH_WEB_DOCUMENTATION = { \"picard\" };;\\n    /**\\n * Initialized in parseArgs.  Subclasses may want to access this to do their\\n * own validation, and then print usage using commandLineParser.\\n */\\nprivate CommandLineParser commandLineParser;;\\n    private final List<Header> defaultHeaders = new ArrayList<>();;\\n    /**\\n * The reconstructed commandline used to run this program. Used for logging\\n * and debugging.\\n */\\nprivate String commandLine;;\\n    protected abstract int doWork()\\n    protected boolean requiresReference()\\n    protected ReferenceArgumentCollection makeReferenceArgumentCollection()\\n    public void instanceMainWithExit(final String[] argv)\\n    public int instanceMain(final String[] argv)\\n    protected String[] customCommandLineValidation()\\n    protected boolean parseArgs(final String[] argv)\\n    protected MetricsFile<A, B> getMetricsFile()\\n    public String getStandardUsagePreamble()\\n    public CommandLineParser getCommandLineParserForArgs(String[] argv)\\n    public CommandLineParser getCommandLineParser()\\n    public static boolean useLegacyParser()\\n    public String getVersion()\\n    public String getCommandLine()\\n    public void setDefaultHeaders(final List<Header> headers)\\n    public SAMProgramRecord getPGRecord(final SAMFileHeader header)\\n    public List<Header> getDefaultHeaders()\\n    public static String getStandardUsagePreamble(final Class<?> mainClass)\\n    public static boolean hasWebDocumentation(final Class<?> clazz)\\n    public static String getFaqLink()\\n    public static boolean checkRInstallation(final boolean has_chart_output)\\n}\\n'}\n",
      "/**\n",
      "     * Process each {@code CommandLineProgram}-derived class given a list of packages.\n",
      "     * @param packageList list of packages to search\n",
      "     * @param clpClassProcessor function to process each CommandLineProgram class found in {@code packageList} (note\n",
      "     *                          that the {@code CommandLineProgramProperties} argument may be null)\n",
      "     */\n",
      "public static void processAllCommandLinePrograms(final List<String> packageList, final BiConsumer<Class<CommandLineProgram>, CommandLineProgramProperties> clpClassProcessor)\n",
      "**************************************\n",
      "public static void processAllCommandLinePrograms(final List<String> packageList, final BiConsumer<Class<CommandLineProgram>, CommandLineProgramProperties> clpClassProcessor) {\n",
      "    final ClassFinder classFinder = new ClassFinder();\n",
      "    packageList.forEach(pkg -> classFinder.find(pkg, CommandLineProgram.class));\n",
      "    for (final Class clazz : classFinder.getClasses()) {\n",
      "        if (!clazz.isInterface() && !clazz.isSynthetic() && !clazz.isPrimitive() && !clazz.isLocalClass() && !Modifier.isAbstract(clazz.getModifiers())) {\n",
      "            clpClassProcessor.accept(clazz, PicardCommandLine.getProgramProperty(clazz));\n",
      "        }\n",
      "    }\n",
      "}\n",
      "1417\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 879/1243 [29:52<13:10,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34156\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 880/1243 [29:54<12:01,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34409\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 881/1243 [29:55<10:42,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34411\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 882/1243 [29:56<09:52,  1.64s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34412\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 883/1243 [29:59<12:09,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34430\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 23, 'classCount': 1, 'methodsOnly': 'public class IGV {\\n    private static Logger log = LogManager.getLogger(IGV.class);;\\n    private static IGV theInstance;;\\n    public static final String DATA_PANEL_NAME = \"DataPanel\";;\\n    public static final String FEATURE_PANEL_NAME = \"FeaturePanel\";;\\n    // Window components\\nprivate Frame mainFrame;;\\n    private JRootPane rootPane;;\\n    private IGVContentPane contentPane;;\\n    private IGVMenuBar menuBar;;\\n    private StatusWindow statusWindow;;\\n    // Glass panes\\nComponent glassPane;;\\n    GhostGlassPane dNdGlassPane;;\\n    // Cursors\\npublic static Cursor fistCursor;;\\n    public static Cursor zoomInCursor;;\\n    public static Cursor zoomOutCursor;;\\n    public static Cursor dragNDropCursor;;\\n    /**\\n * Object to hold state that defines a user session.  There is always a user session, even if not initialized\\n * from a \"session\" file.\\n */\\nprivate Session session;;\\n    // Misc state\\nprivate Map<String, List<Track>> overlayTracksMap = new HashMap();;\\n    private Set<Track> overlaidTracks = new HashSet();;\\n    private LinkedList<String> recentSessionList = new LinkedList<String>();;\\n    // Vertical line that follows the mouse\\nprivate boolean rulerEnabled;;\\n    public static IGV createInstance(Frame frame, Main.IGVArgs igvArgs)\\n    public static IGV getInstance()\\n    public static boolean hasInstance()\\n     static void destroyInstance()\\n    private IGV(Frame frame, Main.IGVArgs igvArgs);\\n    public JRootPane getRootPane()\\n    public Frame getMainFrame()\\n    public GhostGlassPane getDnDGlassPane()\\n    public void startDnD()\\n    public void endDnD()\\n    public Dimension getPreferredSize()\\n    public void addRegionOfInterest(RegionOfInterest roi)\\n    public void beginROI(JButton button)\\n    public void endROI()\\n    public void focusSearchBox()\\n    public void selectGenomeFromList(String genomeId)\\n    public void defineGenome(javax.swing.ProgressMonitor monitor)\\n    public void enableExtrasMenu()\\n    public Future loadTracks(final Collection<ResourceLocator> locators)\\n    public List<Map<TrackPanelScrollPane, Integer>> getTrackPanelAttrs()\\n    public void resetPanelHeights(Map<TrackPanelScrollPane, Integer> trackCountMap, Map<TrackPanelScrollPane, Integer> panelSizeMap)\\n    public void setGeneList(GeneList geneList)\\n    public void setGeneList(final GeneList geneList, final boolean recordHistory)\\n    public void setDefaultFrame(String searchString)\\n    public final void doViewPreferences()\\n    public final void saveStateForExit()\\n    public final void doShowAttributeDisplay(boolean enableAttributeView)\\n    public final void doSelectDisplayableAttribute()\\n    public final void saveImage(Component target, String extension)\\n    public final void saveImage(Component target, String title, String extension)\\n    public final void createSnapshot(final Component target, final File defaultFile)\\n    public String createSnapshotNonInteractive(Component target, File file, boolean batch) throws Exception\\n    public File selectSnapshotFile(File defaultFile)\\n    private void createZoomCursors() throws HeadlessException, IndexOutOfBoundsException\\n    private void createHandCursor() throws HeadlessException, IndexOutOfBoundsException\\n    private void createDragAndDropCursor() throws HeadlessException, IndexOutOfBoundsException\\n    private Cursor createCustomCursor(Image image, Point hotspot, String name, int defaultCursor)\\n    private void subscribeToEvents()\\n    public void setStatusBarMessage(String message)\\n    public void setStatusBarMessag2(String message)\\n    public void setStatusBarMessage3(String message)\\n    public void enableStopButton(boolean enable)\\n    public void resetToFactorySettings()\\n    public void setFilterMatchAll(boolean value)\\n    public boolean isFilterMatchAll()\\n    public void setFilterShowAllTracks(boolean value)\\n    public boolean isFilterShowAllTracks()\\n    public TrackPanelScrollPane addDataPanel(String name)\\n    public TrackPanel getTrackPanel(String name)\\n    public List<TrackPanel> getTrackPanels()\\n    public boolean scrollToTrack(String trackName)\\n    public Session getSession()\\n    public void resetSession(String sessionPath)\\n    public void newSession()\\n    public boolean loadSession(String sessionPath, String locus)\\n    public boolean loadSessionFromStream(String sessionPath, InputStream inputStream) throws IOException\\n    public void saveSession(File targetFile) throws IOException\\n    public void resetStatusMessage()\\n    public void showLoadedTrackCount()\\n    private void closeWindow(final ProgressBar.ProgressDialog progressDialog)\\n    public void goToLocus(String locus)\\n    public void tweakPanelDivider()\\n    public void removeDataPanel(String name)\\n    public void layoutMainPanel()\\n    public MainPanel getMainPanel()\\n    public LinkedList<String> getRecentSessionList()\\n    public void setRecentSessionList(LinkedList<String> recentSessionList)\\n    public IGVContentPane getContentPane()\\n    public boolean isShowDetailsOnClick()\\n    public boolean isShowDetailsOnHover()\\n    public void openStatusWindow()\\n    public void setStatusWindowText(String text)\\n    public void loadResources(Collection<ResourceLocator> locators)\\n    public void addTracks(List<Track> tracks, PanelName panelName)\\n    public void addTracks(List<Track> tracks, ResourceLocator locator)\\n    public List<Track> load(ResourceLocator locator) throws DataLoadException\\n    public void load(final ResourceLocator locator, final TrackPanel panel) throws DataLoadException\\n    public TrackPanel getPanelFor(Track track)\\n    public TrackPanel getPanelFor(ResourceLocator locator)\\n    public Set<TrackType> getLoadedTypes()\\n    public TrackPanel getVcfBamPanel()\\n    private boolean isAnnotationFile(String format)\\n    public void sortAlignmentTracks(SortOption option, String tag, final boolean invertSort)\\n    public void sortAlignmentTracks(SortOption option, Double location, String tag, boolean invertSort)\\n    public void groupAlignmentTracks(AlignmentTrack.GroupOption option, String tag, Range pos)\\n    public void colorAlignmentTracks(AlignmentTrack.ColorOption option, String tag)\\n    public void packAlignmentTracks()\\n    public void resetOverlayTracks()\\n    public List<Track> getOverlayTracks(Track track)\\n    public int getVisibleTrackCount()\\n    public List<Track> getAllTracks()\\n    public List<FeatureTrack> getFeatureTracks()\\n    public List<DataTrack> getDataTracks()\\n    public List<AlignmentTrack> getAlignmentTracks()\\n    public void clearSelections()\\n    public void setTrackSelections(Iterable<Track> selectedTracks)\\n    public void shiftSelectTracks(Track track)\\n    public void toggleTrackSelections(Iterable<Track> selectedTracks)\\n    public List<Track> getSelectedTracks()\\n    public Set<ResourceLocator> getDataResourceLocators()\\n    public void setAllTrackHeights(int newHeight)\\n    public void deleteTracks(Collection<? extends Track> tracksToRemove)\\n    public void setGenomeTracks(Track newGeneTrack)\\n    public boolean hasSequenceTrack()\\n    public SequenceTrack getSequenceTrack()\\n    public void sortAllTracksByAttributes(final String[] attributeNames, final boolean[] ascending)\\n    public void sortByRegionScore(RegionOfInterest region, final RegionScoreType type, final ReferenceFrame frame)\\n    private List<String> sortSamplesByRegionScore(final RegionOfInterest region, final RegionScoreType type, final ReferenceFrame frame)\\n     static void sortByRegionScore(List<Track> tracks, final RegionOfInterest region, final RegionScoreType type, ReferenceFrame frame)\\n    public String getGroupByAttribute()\\n    public void setGroupByAttribute(String attributeName)\\n    private void resetGroups()\\n    public Future startUp(Main.IGVArgs igvArgs)\\n    public void setRulerEnabled(boolean rulerEnabled)\\n    public boolean isRulerEnabled()\\n    private static void startCommandsServer(Main.IGVArgs igvArgs, IGVPreferences prefMgr)\\n    public static void copySequenceToClipboard(Genome genome, String chr, int start, int end, Strand strand)\\n    public boolean waitForNotify(long timeout)\\n    public void postEvent(Object event)\\n    public void receiveEvent(Object event)\\n    public void resetFrames()\\n    public void revalidateTrackPanels()\\n    public void repaintNamePanels()\\n    public void fitTracksToPanel()\\n    public void repaint()\\n    public void repaint(Track track)\\n    public void repaint(Collection<? extends Track> tracks)\\n    private void repaint(final JComponent component)\\n    private boolean isLoading = false;;\\n    private Collection<? extends Track> pending = null;;\\n    private void repaint(final JComponent component, Collection<? extends Track> trackList)\\n    private void checkPanelLayouts()\\n    public List<Track> visibleTracks(DataPanelContainer dataPanelContainer)\\n    private static final ExecutorService threadExecutor = Executors.newFixedThreadPool(5);;\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/ui/IGV.java', 'methodCount': 131, 'lineCount': 156}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * A wrapper around invokeOnEventThread.  If the runnable is already in the event dispatching\n",
      "     * queue it is just run.  Otherwise it is placed in the queue via invokeOnEventThread.\n",
      "     * <p/>\n",
      "     * I'm not sure this is strictly necessary,  but is safe.\n",
      "     *\n",
      "     * @param runnable\n",
      "     */\n",
      "public static void invokeAndWaitOnEventThread(Runnable runnable)\n",
      "**************************************\n",
      "public static void invokeAndWaitOnEventThread(Runnable runnable) {\n",
      "    if (SwingUtilities.isEventDispatchThread()) {\n",
      "        runnable.run();\n",
      "    } else {\n",
      "        try {\n",
      "            SwingUtilities.invokeAndWait(runnable);\n",
      "        } catch (InterruptedException e) {\n",
      "            log.error(\"Error invoking runnable\", e);\n",
      "            UIUtilities.invokeOnEventThread(runnable);\n",
      "        } catch (InvocationTargetException e) {\n",
      "            log.error(\"Error invoking runnable\", e);\n",
      "            UIUtilities.invokeOnEventThread(runnable);\n",
      "        }\n",
      "    }\n",
      "}\n",
      "2252\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 884/1243 [30:02<12:38,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34499\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|   | 885/1243 [30:03<12:12,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34663\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|  | 886/1243 [30:06<12:37,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34706\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|  | 887/1243 [30:07<11:43,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35046\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  71%|  | 888/1243 [30:10<12:23,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35124\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class BatchRunner {\\n    private static Logger log = LogManager.getLogger(BatchRunner.class);;\\n    private String inputFile;;\\n    private String rootPath;;\\n    private IGV igv;;\\n    public BatchRunner(final String inputFile, IGV igv);\\n    public String getName()\\n    public static void setIsBatchMode(boolean isBatchMode)\\n    public void run()\\n    public void runWithDefaultGenome(String genomeId)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/batch/BatchRunner.java', 'methodCount': 5, 'lineCount': 11}, {'fieldCount': 9, 'classCount': 1, 'methodsOnly': 'public class PreferencesManager {\\n    public static final String SEPARATOR_KEY = \"---\";;\\n    public static final String INFO_KEY = \"info\";;\\n    private static List<PreferenceGroup> preferenceGroupList;;\\n    private static Logger log = LogManager.getLogger(PreferencesManager.class);;\\n    private static Map<String, IGVPreferences> preferencesMap = Collections.synchronizedMap(new HashMap<>());;\\n    private static Map<String, String> genericDefaults;;\\n    // User preferences file\\nprivate static String prefFile;;\\n    static Hashtable<String, String> aliasTable = new Hashtable<String, String>();;\\n    private PreferencesManager();\\n    private static PreferencesManager theInstance = new PreferencesManager();;\\n    public static IGVPreferences getPreferences(String category)\\n    private static void init()\\n    private static void extractMutationColors(IGVPreferences prefs)\\n    public static String getDefault(String key)\\n    public static IGVPreferences getPreferences()\\n    public static Collection<IGVPreferences> getAllPreferences()\\n    public static void setPrefsFile(String prefsFile)\\n    private static Map<String, Map<String, String>> loadUserPreferences()\\n    public static void updateAll(Map<String, Map<String, String>> preferenceMap)\\n    public static void loadOverrides(String overridePropertyFilePath)\\n    private static Map<String, Map<String, String>> load(String prefFileName)\\n    private static KeyValue translate(String key, String value)\\n    public static List<PreferenceGroup> loadPreferenceList() throws IOException\\n    public static Map<String, String> loadDefaults23() throws IOException\\n    public static void setOverride(String prefKey, String prefVal)\\n    private synchronized void storePreferences()\\n    public void receiveEvent(Object event)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/prefs/PreferencesManager.java', 'methodCount': 18, 'lineCount': 29}, {'fieldCount': 23, 'classCount': 1, 'methodsOnly': 'public class IGV {\\n    private static Logger log = LogManager.getLogger(IGV.class);;\\n    private static IGV theInstance;;\\n    public static final String DATA_PANEL_NAME = \"DataPanel\";;\\n    public static final String FEATURE_PANEL_NAME = \"FeaturePanel\";;\\n    // Window components\\nprivate Frame mainFrame;;\\n    private JRootPane rootPane;;\\n    private IGVContentPane contentPane;;\\n    private IGVMenuBar menuBar;;\\n    private StatusWindow statusWindow;;\\n    // Glass panes\\nComponent glassPane;;\\n    GhostGlassPane dNdGlassPane;;\\n    // Cursors\\npublic static Cursor fistCursor;;\\n    public static Cursor zoomInCursor;;\\n    public static Cursor zoomOutCursor;;\\n    public static Cursor dragNDropCursor;;\\n    /**\\n * Object to hold state that defines a user session.  There is always a user session, even if not initialized\\n * from a \"session\" file.\\n */\\nprivate Session session;;\\n    // Misc state\\nprivate Map<String, List<Track>> overlayTracksMap = new HashMap();;\\n    private Set<Track> overlaidTracks = new HashSet();;\\n    private LinkedList<String> recentSessionList = new LinkedList<String>();;\\n    // Vertical line that follows the mouse\\nprivate boolean rulerEnabled;;\\n    public static IGV createInstance(Frame frame, Main.IGVArgs igvArgs)\\n    public static IGV getInstance()\\n    public static boolean hasInstance()\\n     static void destroyInstance()\\n    private IGV(Frame frame, Main.IGVArgs igvArgs);\\n    public JRootPane getRootPane()\\n    public Frame getMainFrame()\\n    public GhostGlassPane getDnDGlassPane()\\n    public void startDnD()\\n    public void endDnD()\\n    public Dimension getPreferredSize()\\n    public void addRegionOfInterest(RegionOfInterest roi)\\n    public void beginROI(JButton button)\\n    public void endROI()\\n    public void focusSearchBox()\\n    public void selectGenomeFromList(String genomeId)\\n    public void defineGenome(javax.swing.ProgressMonitor monitor)\\n    public void enableExtrasMenu()\\n    public Future loadTracks(final Collection<ResourceLocator> locators)\\n    public List<Map<TrackPanelScrollPane, Integer>> getTrackPanelAttrs()\\n    public void resetPanelHeights(Map<TrackPanelScrollPane, Integer> trackCountMap, Map<TrackPanelScrollPane, Integer> panelSizeMap)\\n    public void setGeneList(GeneList geneList)\\n    public void setGeneList(final GeneList geneList, final boolean recordHistory)\\n    public void setDefaultFrame(String searchString)\\n    public final void doViewPreferences()\\n    public final void saveStateForExit()\\n    public final void doShowAttributeDisplay(boolean enableAttributeView)\\n    public final void doSelectDisplayableAttribute()\\n    public final void saveImage(Component target, String extension)\\n    public final void saveImage(Component target, String title, String extension)\\n    public final void createSnapshot(final Component target, final File defaultFile)\\n    public String createSnapshotNonInteractive(Component target, File file, boolean batch) throws Exception\\n    public File selectSnapshotFile(File defaultFile)\\n    private void createZoomCursors() throws HeadlessException, IndexOutOfBoundsException\\n    private void createHandCursor() throws HeadlessException, IndexOutOfBoundsException\\n    private void createDragAndDropCursor() throws HeadlessException, IndexOutOfBoundsException\\n    private Cursor createCustomCursor(Image image, Point hotspot, String name, int defaultCursor)\\n    private void subscribeToEvents()\\n    public void setStatusBarMessage(String message)\\n    public void setStatusBarMessag2(String message)\\n    public void setStatusBarMessage3(String message)\\n    public void enableStopButton(boolean enable)\\n    public void resetToFactorySettings()\\n    public void setFilterMatchAll(boolean value)\\n    public boolean isFilterMatchAll()\\n    public void setFilterShowAllTracks(boolean value)\\n    public boolean isFilterShowAllTracks()\\n    public TrackPanelScrollPane addDataPanel(String name)\\n    public TrackPanel getTrackPanel(String name)\\n    public List<TrackPanel> getTrackPanels()\\n    public boolean scrollToTrack(String trackName)\\n    public Session getSession()\\n    public void resetSession(String sessionPath)\\n    public void newSession()\\n    public boolean loadSession(String sessionPath, String locus)\\n    public boolean loadSessionFromStream(String sessionPath, InputStream inputStream) throws IOException\\n    public void saveSession(File targetFile) throws IOException\\n    public void resetStatusMessage()\\n    public void showLoadedTrackCount()\\n    private void closeWindow(final ProgressBar.ProgressDialog progressDialog)\\n    public void goToLocus(String locus)\\n    public void tweakPanelDivider()\\n    public void removeDataPanel(String name)\\n    public void layoutMainPanel()\\n    public MainPanel getMainPanel()\\n    public LinkedList<String> getRecentSessionList()\\n    public void setRecentSessionList(LinkedList<String> recentSessionList)\\n    public IGVContentPane getContentPane()\\n    public boolean isShowDetailsOnClick()\\n    public boolean isShowDetailsOnHover()\\n    public void openStatusWindow()\\n    public void setStatusWindowText(String text)\\n    public void loadResources(Collection<ResourceLocator> locators)\\n    public void addTracks(List<Track> tracks, PanelName panelName)\\n    public void addTracks(List<Track> tracks, ResourceLocator locator)\\n    public List<Track> load(ResourceLocator locator) throws DataLoadException\\n    public void load(final ResourceLocator locator, final TrackPanel panel) throws DataLoadException\\n    public TrackPanel getPanelFor(Track track)\\n    public TrackPanel getPanelFor(ResourceLocator locator)\\n    public Set<TrackType> getLoadedTypes()\\n    public TrackPanel getVcfBamPanel()\\n    private boolean isAnnotationFile(String format)\\n    public void sortAlignmentTracks(SortOption option, String tag, final boolean invertSort)\\n    public void sortAlignmentTracks(SortOption option, Double location, String tag, boolean invertSort)\\n    public void groupAlignmentTracks(AlignmentTrack.GroupOption option, String tag, Range pos)\\n    public void colorAlignmentTracks(AlignmentTrack.ColorOption option, String tag)\\n    public void packAlignmentTracks()\\n    public void resetOverlayTracks()\\n    public List<Track> getOverlayTracks(Track track)\\n    public int getVisibleTrackCount()\\n    public List<Track> getAllTracks()\\n    public List<FeatureTrack> getFeatureTracks()\\n    public List<DataTrack> getDataTracks()\\n    public List<AlignmentTrack> getAlignmentTracks()\\n    public void clearSelections()\\n    public void setTrackSelections(Iterable<Track> selectedTracks)\\n    public void shiftSelectTracks(Track track)\\n    public void toggleTrackSelections(Iterable<Track> selectedTracks)\\n    public List<Track> getSelectedTracks()\\n    public Set<ResourceLocator> getDataResourceLocators()\\n    public void setAllTrackHeights(int newHeight)\\n    public void deleteTracks(Collection<? extends Track> tracksToRemove)\\n    public void setGenomeTracks(Track newGeneTrack)\\n    public boolean hasSequenceTrack()\\n    public SequenceTrack getSequenceTrack()\\n    public void sortAllTracksByAttributes(final String[] attributeNames, final boolean[] ascending)\\n    public void sortByRegionScore(RegionOfInterest region, final RegionScoreType type, final ReferenceFrame frame)\\n    private List<String> sortSamplesByRegionScore(final RegionOfInterest region, final RegionScoreType type, final ReferenceFrame frame)\\n     static void sortByRegionScore(List<Track> tracks, final RegionOfInterest region, final RegionScoreType type, ReferenceFrame frame)\\n    public String getGroupByAttribute()\\n    public void setGroupByAttribute(String attributeName)\\n    private void resetGroups()\\n    public Future startUp(Main.IGVArgs igvArgs)\\n    public void setRulerEnabled(boolean rulerEnabled)\\n    public boolean isRulerEnabled()\\n    private static void startCommandsServer(Main.IGVArgs igvArgs, IGVPreferences prefMgr)\\n    public static void copySequenceToClipboard(Genome genome, String chr, int start, int end, Strand strand)\\n    public boolean waitForNotify(long timeout)\\n    public void postEvent(Object event)\\n    public void receiveEvent(Object event)\\n    public void resetFrames()\\n    public void revalidateTrackPanels()\\n    public void repaintNamePanels()\\n    public void fitTracksToPanel()\\n    public void repaint()\\n    public void repaint(Track track)\\n    public void repaint(Collection<? extends Track> tracks)\\n    private void repaint(final JComponent component)\\n    private boolean isLoading = false;;\\n    private Collection<? extends Track> pending = null;;\\n    private void repaint(final JComponent component, Collection<? extends Track> trackList)\\n    private void checkPanelLayouts()\\n    public List<Track> visibleTracks(DataPanelContainer dataPanelContainer)\\n    private static final ExecutorService threadExecutor = Executors.newFixedThreadPool(5);;\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/ui/IGV.java', 'methodCount': 131, 'lineCount': 156}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class FileDialogUtils {\\n    public static int LOAD = FileDialog.LOAD;;\\n    public static int SAVE = FileDialog.SAVE;;\\n    private static Logger log = LogManager.getLogger(FileDialogUtils.class);;\\n    public static File chooseFile(String title, File initialDirectory, int mode)\\n    public static File chooseFile(String title)\\n    public static File chooseFile(String title, File initialDirectory, File initialFile, int mode)\\n    private static File chooseFile(String title, File initialDirectory, File initialFile, FilenameFilter filter, int directoriesMode, int mode)\\n    public static File chooseDirectory(String title, File initialDirectory)\\n    public static File[] chooseMultiple(String title, File initialDirectory, final FilenameFilter filter)\\n    private static FileDialog getNativeChooser(String title, File initialDirectory, File initialFile, FilenameFilter filter, int directoryMode, int mode)\\n    private static File chooseNative(String title, File initialDirectory, File initialFile, FilenameFilter filter, int directoryMode, int mode)\\n    private static File chooseSwing(String title, File initialDirectory, File initialFile, final FilenameFilter filter, int directoryMode, int mode)\\n    private static JFileChooser getJFileChooser(String title, File initialDirectory, File initialFile, final FilenameFilter filter, int directoryMode)\\n    private static boolean setMultipleMode(FileDialog fd, boolean b)\\n    private static String fixMacExtension(File initialFile, String fname)\\n    private static Frame getParentFrame()\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/ui/util/FileDialogUtils.java', 'methodCount': 13, 'lineCount': 18}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 2}\n",
      "{'BatchRunner': 'public class BatchRunner {\\n    private static Logger log = LogManager.getLogger(BatchRunner.class);;\\n    private String inputFile;;\\n    private String rootPath;;\\n    private IGV igv;;\\n    public BatchRunner(final String inputFile, IGV igv);\\n    public String getName()\\n    public static void setIsBatchMode(boolean isBatchMode)\\n    public void run()\\n    public void runWithDefaultGenome(String genomeId)\\n}\\n'}\n",
      "/**\n",
      "     * Run the batch script.  This is PURPOSELY run on the event dispatch thread to maintain absolute synchronization\n",
      "     * @param e\n",
      "     */\n",
      "public void actionPerformed(ActionEvent e)\n",
      "**************************************\n",
      "/**\n",
      " * Run the batch script.  This is PURPOSELY run on the event dispatch thread to maintain absolute synchronization\n",
      " * @param e\n",
      " */\n",
      "public void actionPerformed(ActionEvent e) {\n",
      "    if (e.getActionCommand().equalsIgnoreCase(\"run batch script...\")) {\n",
      "        File[] script = chooseScriptFile();\n",
      "        if (script != null && script.length > 0) {\n",
      "            for (File f : script) {\n",
      "                final BatchRunner bRun = new BatchRunner(f.getPath(), igv);\n",
      "                bRun.run();\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "}\n",
      "3287\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 889/1243 [30:12<12:26,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35755\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 890/1243 [30:14<12:36,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35962\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 891/1243 [30:16<12:03,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36006\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 892/1243 [30:18<12:23,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36073\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 893/1243 [30:21<12:39,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36194\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class DirectoryManager {\\n    private static Logger log = LogManager.getLogger(DirectoryManager.class);;\\n    private static File USER_HOME;;\\n    // FileSystemView.getFileSystemView().getDefaultDirectory();\\nprivate static File USER_DIRECTORY;;\\n    // The IGV application directory\\nprivate static File IGV_DIRECTORY;;\\n    private static File GENOME_CACHE_DIRECTORY;;\\n    private static File GENE_LIST_DIRECTORY;;\\n    private static File BAM_CACHE_DIRECTORY;;\\n    final public static String IGV_DIR_USERPREF = \"igvDir\";;\\n    private static File getUserHome()\\n    public static synchronized File getUserDirectory()\\n    public static void setIgvDirectory(File dir)\\n    public static File getIgvDirectory()\\n    private static File getIgvDirectoryOverride()\\n    public static File getGenomeCacheDirectory()\\n    public static File getFastaCacheDirectory()\\n    public static File getGeneListDirectory()\\n    public static synchronized File getCacheDirectory()\\n    public static synchronized File getSamDirectory()\\n    public static synchronized File getLogFile() throws IOException\\n    public static synchronized File getPreferencesFile()\\n    public static void moveIGVDirectory(final File newIGVDirectoryParent)\\n    public static void moveDirectoryContents(File oldDirectory, File newDirectory)\\n    public static boolean isChildOf(File base, File child) throws IOException\\n    private static synchronized File getLegacyPreferencesFile()\\n    private static File getLegacyIGVDirectory()\\n    private static boolean canWrite(File directory)\\n    public static void initializeLog()\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/DirectoryManager.java', 'methodCount': 19, 'lineCount': 29}, {'fieldCount': 43, 'classCount': 1, 'methodsOnly': 'public class Globals {\\n    private static Logger log = LogManager.getLogger(Globals.class);;\\n    public static final int DESIGN_DPI = 96;;\\n    public static final DecimalFormat DECIMAL_FORMAT = new DecimalFormat();;\\n    final static public String HISTORY_DELIMITER = \";\";;\\n    public static final String DEFAULT_GENOME = \"hg19\";;\\n    public static final String CHR_ALL = \"All\";;\\n    public static final String TRACK_NAME_ATTRIBUTE = \"NAME\";;\\n    public static final String TRACK_DATA_FILE_ATTRIBUTE = \"DATA FILE\";;\\n    public static final String TRACK_DATA_TYPE_ATTRIBUTE = \"DATA TYPE\";;\\n    private static boolean headless = false;;\\n    private static boolean suppressMessages = false;;\\n    private static boolean batch = false;;\\n    private static boolean testing = false;;\\n    // 20 seconds\\npublic static int CONNECT_TIMEOUT = 20000;;\\n    // 3 minutes\\npublic static int READ_TIMEOUT = 1000 * 3 * 60;;\\n    // 1 minute\\npublic static int TOKEN_EXPIRE_GRACE_TIME = 1000 * 1 * 60;;\\n    /**\\n * Field description\\n */\\nfinal public static String SESSION_FILE_EXTENSION = \".xml\";;\\n    /**\\n * GENOME ARCHIVE CONSTANTS\\n */\\nfinal public static String GENOME_FILE_EXTENSION = \".genome\";;\\n    final public static String ZIP_EXTENSION = \".zip\";;\\n    final public static String GZIP_FILE_EXTENSION = \".gz\";;\\n    // Default user folder\\nfinal static public Pattern commaPattern = Pattern.compile(\",\");;\\n    final static public Pattern tabPattern = Pattern.compile(\"\\\\t\");;\\n    final static public Pattern multiTabPattern = Pattern.compile(\"\\\\t+\");;\\n    final static public Pattern colonPattern = Pattern.compile(\":\");;\\n    final static public Pattern dashPattern = Pattern.compile(\"-\");;\\n    final static public Pattern equalPattern = Pattern.compile(\"=\");;\\n    final static public Pattern semicolonPattern = Pattern.compile(\";\");;\\n    final static public Pattern singleTabMultiSpacePattern = Pattern.compile(\"\\\\t|( +)\");;\\n    final static public Pattern forwardSlashPattern = Pattern.compile(\"/\");;\\n    final static public Pattern whitespacePattern = Pattern.compile(\"\\\\\\\\s+\");;\\n    public static List emptyList = new ArrayList();;\\n    public static String VERSION;;\\n    public static String BUILD;;\\n    public static String TIMESTAMP;;\\n    final public static boolean IS_WINDOWS = System.getProperty(\"os.name\").toLowerCase().startsWith(\"windows\");;\\n    final public static boolean IS_MAC = System.getProperty(\"os.name\").toLowerCase().startsWith(\"mac\");;\\n    final public static boolean IS_LINUX = System.getProperty(\"os.name\").toLowerCase().contains(\"linux\");;\\n    final public static boolean IS_JWS = System.getProperty(\"webstart.version\", null) != null || System.getProperty(\"javawebstart.version\", null) != null;;\\n    public static final String JAVA_VERSION_STRING = \"java.version\";;\\n    //Location of bedtools executable\\n//Note: It is recommended you use an absolute path here.\\n//System paths can be finnicky and vary depending on how IGV is launched\\n//However, the path environment variable will be checked if the executable\\n//is named rather than the full path given\\n//\"bedtools\"\\npublic static String BEDtoolsPath = \"/usr/local/bin/bedtools\";;\\n    public static boolean toolsMenuEnabled = false;;\\n    public static boolean development = false;;\\n    public static String downloadURL = \"https://software.broadinstitute.org/software/igv/download\";;\\n    public static void setHeadless(boolean bool)\\n    public static boolean isHeadless()\\n    public static void setTesting(boolean testing)\\n    public static boolean isTesting()\\n    public static void setSuppressMessages(boolean bool)\\n    public static boolean isSuppressMessages()\\n    public static String applicationString()\\n    public static String versionString()\\n    public static boolean isDevelopment()\\n    public static boolean isBatch()\\n    public static void setBatch(boolean batch)\\n    public static boolean checkJavaVersion(String minVersion)\\n    public static double log2(final double value)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/Globals.java', 'methodCount': 13, 'lineCount': 58}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class FeatureCollectionSource {\\n    private TrackType type;;\\n    private Map<String, List<htsjdk.tribble.Feature>> featureMap;;\\n    CoverageDataSource coverageData;;\\n    Genome genome;;\\n    // usually null\\nObject header;;\\n    public FeatureCollectionSource(Iterable<? extends Feature> allFeatures, Genome genome);\\n    public List<LocusScore> getCoverageScores(String chr, int startLocation, int endLocation, int zoom)\\n    public Iterator<Feature> getFeatures(String chr, int start, int end)\\n    public List<Feature> getFeatureList(String chr, int start, int end)\\n    public boolean isLoaded(ReferenceFrame frame)\\n    public List<Feature> getFeatures(String chr)\\n    private void initFeatures(Iterable<? extends Feature> allFeatures)\\n    private void setFeatures(String chr, List<Feature> features)\\n    public TrackType getType()\\n    public void setType(TrackType type)\\n    public Object getHeader()\\n    public void setHeader(Object header)\\n    protected void sampleGenomeFeatures()\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/track/FeatureCollectionSource.java', 'methodCount': 13, 'lineCount': 20}, {'fieldCount': 24, 'classCount': 1, 'methodsOnly': 'public class FeatureTrack {\\n    private static Logger log = LogManager.getLogger(FeatureTrack.class);;\\n    public static final int MINIMUM_FEATURE_SPACING = 5;;\\n    public static final int DEFAULT_MARGIN = 5;;\\n    public static final int NO_FEATURE_ROW_SELECTED = -1;;\\n    protected static final Color SELECTED_FEATURE_ROW_COLOR = new Color(100, 100, 100, 30);;\\n    private static final int DEFAULT_EXPANDED_HEIGHT = 35;;\\n    private static final int DEFAULT_SQUISHED_HEIGHT = 12;;\\n    private int expandedRowHeight = DEFAULT_EXPANDED_HEIGHT;;\\n    private int squishedRowHeight = DEFAULT_SQUISHED_HEIGHT;;\\n    boolean fatalLoadError = false;;\\n    // Keeps track of the feature display mode before an auto-switch to COLLAPSE\\nTrack.DisplayMode lastFeatureMode = null;;\\n    // TODO -- this is a memory leak, this cache needs cleared when the reference frame collection (gene list) changes\\n/**\\n * Map of reference frame name -> packed features\\n */\\nprotected Map<String, PackedFeatures<Feature>> packedFeaturesMap = Collections.synchronizedMap(new HashMap<>());;\\n    protected Renderer renderer;;\\n    private DataRenderer coverageRenderer;;\\n    // true == features,  false =  coverage\\nprivate boolean showFeatures = true;;\\n    public FeatureSource source;;\\n    //track which row of the expanded track is selected by the user. Selection goes away if tracks are collpased\\nprotected int selectedFeatureRowIndex = NO_FEATURE_ROW_SELECTED;;\\n    //Feature selected by the user.  This is repopulated on each handleDataClick() call.\\nprotected IGVFeature selectedFeature = null;;\\n    int margin = DEFAULT_MARGIN;;\\n    private static boolean drawBorder = true;;\\n    private boolean alternateExonColor = false;;\\n    private String trackLine = null;;\\n    private boolean groupByStrand = false;;\\n    private String labelField;;\\n    public FeatureTrack();\\n    public FeatureTrack(ResourceLocator locator, String id, String name);\\n    public FeatureTrack(String id, String name, FeatureSource source);\\n    public FeatureTrack(ResourceLocator locator, String id, String name, FeatureSource source);\\n    public FeatureTrack(ResourceLocator locator, FeatureSource source);\\n    public FeatureTrack(ResourceLocator locator, String id, FeatureSource source);\\n    public FeatureTrack(FeatureTrack featureTrack);\\n    protected void init(ResourceLocator locator, FeatureSource source)\\n    public void unload()\\n    public void receiveEvent(Object e)\\n    public int getHeight()\\n    public int getExpandedRowHeight()\\n    public void setExpandedRowHeight(int expandedRowHeight)\\n    public int getSquishedRowHeight()\\n    public void setSquishedRowHeight(int squishedRowHeight)\\n    public void setRendererClass(Class rc)\\n    public void setMargin(int margin)\\n    public void setProperties(TrackProperties trackProperties)\\n    public void setWindowFunction(WindowFunction type)\\n    public int getNumberOfFeatureLevels()\\n    public float getRegionScore(String chr, int start, int end, int zoom, RegionScoreType scoreType, String frameName)\\n    public Renderer getRenderer()\\n    public String getValueStringAt(String chr, double position, int mouseX, int mouseY, ReferenceFrame frame)\\n    private String getFeatureURL(IGVFeature igvFeature)\\n    public String getLabelField()\\n    public void setLabelField(String labelField)\\n    public List<Feature> getFeatures(String chr, int start, int end)\\n    protected List<Feature> getAllFeatureAt(double position, int y, ReferenceFrame frame)\\n    private int getFeatureRow(int y)\\n    public List<Feature> getFeaturesAtPositionInFeatureRow(double position, int featureRow, ReferenceFrame frame)\\n    public WindowFunction getWindowFunction()\\n    public boolean handleDataClick(TrackClickEvent te)\\n    public Feature getFeatureAtMousePosition(TrackClickEvent te)\\n    public boolean isLogNormalized()\\n    public void overlay(RenderContext context, Rectangle rect)\\n    public void setDisplayMode(DisplayMode mode)\\n    public boolean isReadyToPaint(ReferenceFrame frame)\\n    public void load(ReferenceFrame frame)\\n    protected void loadFeatures(final String chr, final int start, final int end, final String frame)\\n    public void render(RenderContext context, Rectangle rect)\\n    protected boolean isShowFeatures(ReferenceFrame frame)\\n    protected void renderCoverage(RenderContext context, Rectangle inputRect)\\n    protected String getZoomInMessage(String chr)\\n    private float getMaxEstimate(List<LocusScore> scores)\\n    protected void renderFeatures(RenderContext context, Rectangle inputRect)\\n    protected void renderFeatureImpl(RenderContext context, Rectangle inputRect, PackedFeatures packedFeatures)\\n    public Feature nextFeature(String chr, double center, boolean forward, ReferenceFrame frame) throws IOException\\n    public void setVisibilityWindow(int windowSize)\\n    public int getSelectedFeatureRowIndex()\\n    public void setSelectedFeatureRowIndex(int selectedFeatureRowIndex)\\n    public IGVFeature getSelectedFeature()\\n    public static boolean isDrawBorder()\\n    public static void setDrawBorder(boolean drawBorder)\\n    public boolean isAlternateExonColor()\\n    public void updateTrackReferences(List<Track> allTracks)\\n    public void clearPackedFeatures()\\n    public List<Feature> getVisibleFeatures(ReferenceFrame frame)\\n    public void setTrackLine(String trackLine)\\n    public String getExportTrackLine()\\n    public void setGroupByStrand(boolean selected)\\n    public boolean isGroupByStrand()\\n    public void marshalXML(Document document, Element element)\\n    public void unmarshalXML(Element element, Integer version)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/track/FeatureTrack.java', 'methodCount': 63, 'lineCount': 89}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MessageUtils {\\n    private static Logger log = LogManager.getLogger(MessageUtils.class);;\\n    public static void showErrorMessage(String message, Exception e)\\n    public static void showMessage(String message)\\n    public static void showMessage(Level level, String message)\\n    public static void setStatusBarMessage(final String message)\\n    public static synchronized boolean confirm(final String message)\\n    public static synchronized boolean confirm(final Component component, final String message)\\n    public static String showInputDialog(String message, String defaultValue)\\n    public static String showInputDialog(final String message)\\n    public static void main(String[] args) throws Exception\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/ui/util/MessageUtils.java', 'methodCount': 9, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Create an annotation track for the genome from a supplied list of features\n",
      "     *\n",
      "     * @param genome\n",
      "     * @param features\n",
      "     */\n",
      "public static FeatureTrack createGeneTrack(Genome genome, List<htsjdk.tribble.Feature> features)\n",
      "**************************************\n",
      "public static FeatureTrack createGeneTrack(Genome genome, List<htsjdk.tribble.Feature> features) {\n",
      "    FeatureDB.clearFeatures();\n",
      "    FeatureTrack geneFeatureTrack = null;\n",
      "    String name = \"Annotations\";\n",
      "    String id = genome.getId() + \"_genes\";\n",
      "    geneFeatureTrack = new FeatureTrack(id, name, new FeatureCollectionSource(features, genome));\n",
      "    geneFeatureTrack.setMinimumHeight(5);\n",
      "    geneFeatureTrack.setHeight(35);\n",
      "    geneFeatureTrack.setColor(Color.BLUE.darker());\n",
      "    return geneFeatureTrack;\n",
      "}\n",
      "3362\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 894/1243 [30:22<11:48,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36440\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 895/1243 [30:24<11:10,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36457\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 896/1243 [30:26<11:38,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36459\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 897/1243 [30:27<10:18,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36718\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 898/1243 [30:29<10:02,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36945\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class FTPUtils {\\n    static Map<String, String> userCredentials = new HashMap<String, String>();;\\n    static int TIMEOUT = 10000;;\\n    public static boolean resourceAvailable(URL url)\\n    public static long getContentLength(URL url) throws IOException\\n    public static synchronized FTPClient connect(String host, String userInfo, UserPasswordInput userPasswordInput) throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/util/ftp/FTPUtils.java', 'methodCount': 3, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Delete a directory and all subdirectories\n",
      "     *\n",
      "     * @param dir\n",
      "     * @return\n",
      "     */\n",
      "public static boolean deleteDir(File dir)\n",
      "**************************************\n",
      "public static boolean deleteDir(File dir) {\n",
      "    if (dir.isDirectory()) {\n",
      "        String[] children = dir.list();\n",
      "        for (int i = 0; i < children.length; i++) {\n",
      "            boolean success = deleteDir(new File(dir, children[i]));\n",
      "            if (!success) {\n",
      "                return false;\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return dir.delete();\n",
      "}\n",
      "293\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 899/1243 [30:31<10:01,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36960\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 900/1243 [30:32<09:53,  1.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "36971\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  72%|  | 901/1243 [30:34<09:40,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37029\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 902/1243 [30:36<09:28,  1.67s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37030\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 903/1243 [30:37<09:32,  1.68s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37222\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * URL encodes the given string.  This allows us to pass special characters\n",
      "     * that would otherwise be rejected when building a URI instance.  Because we\n",
      "     * need to retain the URI's path structure we subsequently need to replace\n",
      "     * percent encoded path delimiters back to their decoded counterparts.\n",
      "     *\n",
      "     * @param str the string to encode\n",
      "     * @return the encoded string\n",
      "     */\n",
      "private static String preprocessUrlStr(final String str, final boolean encode)\n",
      "**************************************\n",
      "private static String preprocessUrlStr(final String str, final boolean encode) {\n",
      "    if (encode) {\n",
      "        try {\n",
      "            return (URLEncoder.encode(str, \"UTF-8\").replace(\"%3A\", \":\").replace(\"%2F\", \"/\").replace(\"+\", \"%20\"));\n",
      "        } catch (UnsupportedEncodingException e) {\n",
      "            throw new RuntimeException(e);\n",
      "        }\n",
      "    }\n",
      "    return str;\n",
      "}\n",
      "219\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 904/1243 [30:40<10:32,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37223\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 905/1243 [30:42<11:21,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37225\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 906/1243 [30:44<11:31,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37290\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 907/1243 [30:47<12:07,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37296\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 908/1243 [30:49<11:37,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37378\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 8, 'classCount': 1, 'methodsOnly': 'public class DirectoryManager {\\n    private static Logger log = LogManager.getLogger(DirectoryManager.class);;\\n    private static File USER_HOME;;\\n    // FileSystemView.getFileSystemView().getDefaultDirectory();\\nprivate static File USER_DIRECTORY;;\\n    // The IGV application directory\\nprivate static File IGV_DIRECTORY;;\\n    private static File GENOME_CACHE_DIRECTORY;;\\n    private static File GENE_LIST_DIRECTORY;;\\n    private static File BAM_CACHE_DIRECTORY;;\\n    final public static String IGV_DIR_USERPREF = \"igvDir\";;\\n    private static File getUserHome()\\n    public static synchronized File getUserDirectory()\\n    public static void setIgvDirectory(File dir)\\n    public static File getIgvDirectory()\\n    private static File getIgvDirectoryOverride()\\n    public static File getGenomeCacheDirectory()\\n    public static File getFastaCacheDirectory()\\n    public static File getGeneListDirectory()\\n    public static synchronized File getCacheDirectory()\\n    public static synchronized File getSamDirectory()\\n    public static synchronized File getLogFile() throws IOException\\n    public static synchronized File getPreferencesFile()\\n    public static void moveIGVDirectory(final File newIGVDirectoryParent)\\n    public static void moveDirectoryContents(File oldDirectory, File newDirectory)\\n    public static boolean isChildOf(File base, File child) throws IOException\\n    private static synchronized File getLegacyPreferencesFile()\\n    private static File getLegacyIGVDirectory()\\n    private static boolean canWrite(File directory)\\n    public static void initializeLog()\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/DirectoryManager.java', 'methodCount': 19, 'lineCount': 29}, {'fieldCount': 9, 'classCount': 1, 'methodsOnly': 'public class PreferencesManager {\\n    public static final String SEPARATOR_KEY = \"---\";;\\n    public static final String INFO_KEY = \"info\";;\\n    private static List<PreferenceGroup> preferenceGroupList;;\\n    private static Logger log = LogManager.getLogger(PreferencesManager.class);;\\n    private static Map<String, IGVPreferences> preferencesMap = Collections.synchronizedMap(new HashMap<>());;\\n    private static Map<String, String> genericDefaults;;\\n    // User preferences file\\nprivate static String prefFile;;\\n    static Hashtable<String, String> aliasTable = new Hashtable<String, String>();;\\n    private PreferencesManager();\\n    private static PreferencesManager theInstance = new PreferencesManager();;\\n    public static IGVPreferences getPreferences(String category)\\n    private static void init()\\n    private static void extractMutationColors(IGVPreferences prefs)\\n    public static String getDefault(String key)\\n    public static IGVPreferences getPreferences()\\n    public static Collection<IGVPreferences> getAllPreferences()\\n    public static void setPrefsFile(String prefsFile)\\n    private static Map<String, Map<String, String>> loadUserPreferences()\\n    public static void updateAll(Map<String, Map<String, String>> preferenceMap)\\n    public static void loadOverrides(String overridePropertyFilePath)\\n    private static Map<String, Map<String, String>> load(String prefFileName)\\n    private static KeyValue translate(String key, String value)\\n    public static List<PreferenceGroup> loadPreferenceList() throws IOException\\n    public static Map<String, String> loadDefaults23() throws IOException\\n    public static void setOverride(String prefKey, String prefVal)\\n    private synchronized void storePreferences()\\n    public void receiveEvent(Object event)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/prefs/PreferencesManager.java', 'methodCount': 18, 'lineCount': 29}, {'fieldCount': 9, 'classCount': 1, 'methodsOnly': 'public class AmazonUtils {\\n    private static Logger log = LogManager.getLogger(AmazonUtils.class);;\\n    // AWS specific objects\\nprivate static S3Client s3Client;;\\n    private static List<String> bucketsFinalList = new ArrayList<>();;\\n    private static CognitoIdentityClient cognitoIdentityClient;;\\n    private static Region AWSREGION;;\\n    private static Boolean awsCredentialsPresent = null;;\\n    /**\\n * Maps s3:// URLs to presigned URLs\\n */\\nprivate static Map<String, String> s3ToPresignedMap = new HashMap<>();;\\n    /**\\n * Maps aws presigned URLs to s3://.  This is needed in some cases (e.g. Tribble) to regenerate an expired URL\\n */\\nprivate static Map<String, String> presignedToS3Map = new HashMap<>();;\\n    private static JsonObject CognitoConfig;;\\n    public static void setCognitoConfig(JsonObject json)\\n    public static JsonObject GetCognitoConfig()\\n    public static boolean isAwsProviderPresent()\\n    private static Region getAWSREGION()\\n    public static Credentials GetCognitoAWSCredentials()\\n    public static void updateS3Client(Credentials credentials)\\n    public static List<String> ListBucketsForUser()\\n    public static HeadObjectResponse getObjectMetadata(String bucket, String key)\\n    public static s3ObjectAccessResult isObjectAccessible(String bucket, String key)\\n    private static List<String> getReadableBuckets(List<String> buckets)\\n    public static ArrayList<IGVS3Object> ListBucketObjects(String bucketName, String prefix)\\n    public static String getBucketFromS3URL(String s3URL)\\n    public static String getKeyFromS3URL(String s3URL)\\n    private static String createPresignedURL(String s3Path) throws IOException\\n    public static String translateAmazonCloudURL(String s3UrlString) throws IOException\\n    public static Boolean isAwsS3Path(String path)\\n    public static boolean isPresignedURL(String urlString)\\n    public static String updatePresignedURL(String urlString) throws IOException\\n    public static void checkLogin()\\n    private static boolean isPresignedURLValid(URL url)\\n    private static long signedURLValidity(URL url) throws ParseException, UnsupportedEncodingException\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/util/AmazonUtils.java', 'methodCount': 21, 'lineCount': 32}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class GoogleUtils {\\n    private static Logger log = LogManager.getLogger(GoogleUtils.class);;\\n    /**\\n * Google project ID for \"user pays\" requests.\\n */\\nprivate static String ProjectID;;\\n    public static final String GOOGLE_API_HOST = \"www.googleapis.com\";;\\n    public static final String GOOGLE_DRIVE_HOST = \"drive.google.com\";;\\n    public static boolean isGoogleURL(String url)\\n    public static boolean isGoogleDrive(String url)\\n    public static boolean isGoogleStorageURL(String url)\\n    public static String translateGoogleCloudURL(String gsUrl)\\n    public static void enterGoogleProjectID()\\n    public static String getProjectID()\\n    public static void setProjectID(String projectID)\\n    public static String driveDownloadURL(String link)\\n    public static String getGoogleDriveFileID(String link)\\n    public static JsonObject getDriveFileInfo(String googleDriveURL)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/util/GoogleUtils.java', 'methodCount': 10, 'lineCount': 16}, {'fieldCount': 14, 'classCount': 1, 'methodsOnly': 'public class HttpUtils {\\n    private static Logger log = LogManager.getLogger(HttpUtils.class);;\\n    private static HttpUtils instance;;\\n    private Map<String, Boolean> byteRangeTestMap;;\\n    private ProxySettings proxySettings = null;;\\n    private final int MAX_REDIRECTS = 5;;\\n    private String defaultUserName = null;;\\n    private char[] defaultPassword = null;;\\n    private Map<String, Collection<String>> headerMap = new HashMap<>();;\\n    // static provided to support unit testing\\nprivate static boolean BYTE_RANGE_DISABLED = false;;\\n    private Map<URL, Boolean> headURLCache = new HashMap<URL, Boolean>();;\\n    // remember HTTP redirects\\nprivate final int DEFAULT_REDIRECT_EXPIRATION_MIN = 15;;\\n    private Map<URL, CachedRedirect> redirectCache = new HashMap<URL, CachedRedirect>();;\\n    // oauth tokens set from command line script\\nprivate Map<Pattern, String> accessTokens = new HashMap<>();;\\n    public static HttpUtils getInstance()\\n    private HttpUtils();\\n    public void setAccessToken(String token, String host)\\n     String getAccessTokenFor(URL url)\\n    public void clearAccessTokens()\\n    public static URL createURL(String urlString) throws MalformedURLException\\n    public static String mapURL(String urlString) throws MalformedURLException\\n    public static boolean isRemoteURL(String string)\\n    public static void disableByteRange(boolean b)\\n    public String getContentsAsString(URL url) throws IOException\\n    public String getContentsAsString(URL url, Map<String, String> headers) throws IOException\\n    public byte[] getContentsAsBytes(URL url, Map<String, String> headers) throws IOException\\n    public String getContentsAsJSON(URL url) throws IOException\\n    public String doPost(URL url, Map<String, String> params) throws IOException\\n    public InputStream openConnectionStream(URL url) throws IOException\\n    public InputStream openConnectionStream(URL url, Map<String, String> requestProperties) throws IOException\\n    public boolean resourceAvailable(String urlString)\\n    private HttpURLConnection openConnectionHeadOrGet(URL url) throws IOException\\n    public String getHeaderField(URL url, String key) throws IOException\\n    public long getLastModified(URL url) throws IOException\\n    public long getContentLength(URL url) throws IOException\\n    public void updateProxySettings()\\n    private Proxy getSystemProxy(String uri)\\n    public RunnableResult downloadFile(String url, File outputFile) throws IOException\\n    public URLDownloader downloadFile(String url, File outputFile, Frame dialogsParent, String dialogTitle) throws IOException\\n    private void disableCertificateValidation()\\n    private String readContents(InputStream is) throws IOException\\n    public String readErrorStream(HttpURLConnection connection) throws IOException\\n    public HttpURLConnection delete(URL url) throws IOException\\n    public HttpURLConnection openConnection(URL url, Map<String, String> requestProperties) throws IOException\\n    private HttpURLConnection openConnection(URL url, Map<String, String> requestProperties, String method) throws IOException\\n    private HttpURLConnection openConnection(URL url, Map<String, String> requestProperties, String method, int redirectCount, int retries) throws IOException\\n    public HttpURLConnection openProxiedConnection(URL url) throws IOException\\n    private boolean isDropboxHost(String host)\\n    private URL addQueryParameter(URL url, String param, String value)\\n    private boolean hasQueryParameter(URL url, String parameter)\\n    private void logHeaders(HttpURLConnection conn)\\n    public void setDefaultPassword(String defaultPassword)\\n    public void setDefaultUserName(String defaultUserName)\\n    public void clearDefaultCredentials()\\n    public boolean useByteRange(URL url) throws IOException\\n    public boolean testByteRange(URL url) throws IOException\\n    public void addHeaders(Collection<String> headers, List<String> urls)\\n    private String stripParameters(String url)\\n    public void shutdown()\\n    public static void readFully(InputStream is, byte[] b) throws IOException\\n    private static Map<String, String> urlMappings;;\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/util/HttpUtils.java', 'methodCount': 46, 'lineCount': 62}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 8}\n",
      "{'OAuthProvider': 'public class OAuthProvider {\\n    private static Logger log = LogManager.getLogger(OAuthProvider.class);;\\n    private String authProvider = \"\";;\\n    private String appIdURI = null;;\\n    public static String findString = null;;\\n    public static String replaceString = null;;\\n    private static final String REFRESH_TOKEN_KEY = \"oauth_refresh_token\";;\\n    // \"RFC6749: An opaque value used by the client to maintain state\"\\nprivate String state;;\\n    private String portNumber = PreferencesManager.getPreferences().getPortNumber();;\\n    private String redirectURI = \"http%3A%2F%2Flocalhost%3A\" + portNumber + \"%2FoauthCallback\";;\\n    private String clientId;;\\n    private String clientSecret;;\\n    private String authEndpoint;;\\n    private String tokenEndpoint;;\\n    private String accessToken;;\\n    private String refreshToken;;\\n    // in milliseconds\\nprivate long expirationTime;;\\n    private String scope;;\\n    private String[] hosts;;\\n    private String currentUserName;;\\n    private String currentUserID;;\\n    private String currentUserEmail;;\\n    private JsonObject response;;\\n    private JsonObject config;;\\n    public OAuthProvider(JsonObject obj) throws IOException;\\n    public String getState()\\n    public String[] getHosts()\\n    public void setHosts(String[] hosts)\\n    public void openAuthorizationPage() throws IOException, URISyntaxException\\n    public void setAuthorizationCode(String authorizationCode) throws IOException\\n    public void setAccessToken(String accessToken)\\n    private void refreshAccessToken() throws IOException\\n    private void reauthorize() throws IOException\\n    public JsonObject fetchUserProfile(JsonObject jwt_payload)\\n    public String getAccessToken()\\n    public Duration getExpirationTime()\\n    public boolean isLoggedIn()\\n    public String getCurrentUserName()\\n    public void logout()\\n    private void removeRefreshToken()\\n    public synchronized void checkLogin()\\n    public boolean appliesToUrl(URL url)\\n    public boolean isGoogle()\\n    public JsonObject getResponse()\\n    public void setResponse(JsonObject res)\\n    public String getAuthProvider()\\n    public void setAuthProvider(String authProvider)\\n}\\n'}\n",
      "/**\n",
      "     * Parse Oauth provider configuration and update state\n",
      "     * <p>\n",
      "     * TODO -- refactor to remove side effects.\n",
      "     *\n",
      "     * @param json\n",
      "     * @throws IOException\n",
      "     */\n",
      "private OAuthProvider parseProviderJson(String json) throws IOException\n",
      "**************************************\n",
      "private OAuthProvider parseProviderJson(String json) throws IOException {\n",
      "    JsonParser parser = new JsonParser();\n",
      "    JsonObject obj = parser.parse(json).getAsJsonObject();\n",
      "    OAuthProvider p = new OAuthProvider(obj);\n",
      "    providerCache.put(p.getState(), p);\n",
      "    if (obj.has(\"auth_provider\") && obj.get(\"auth_provider\").getAsString().equals(\"Amazon\")) {\n",
      "        awsProvider = p;\n",
      "        AmazonUtils.setCognitoConfig(obj);\n",
      "    }\n",
      "    if (p.isGoogle()) {\n",
      "        defaultProvider = p;\n",
      "    }\n",
      "    return p;\n",
      "}\n",
      "3205\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 909/1243 [30:50<11:02,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38166\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 910/1243 [30:53<11:27,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38168\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 911/1243 [30:55<11:30,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "38515\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 912/1243 [30:57<11:37,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "39824\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  73%|  | 913/1243 [30:59<11:11,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40129\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 43, 'classCount': 1, 'methodsOnly': 'public class Globals {\\n    private static Logger log = LogManager.getLogger(Globals.class);;\\n    public static final int DESIGN_DPI = 96;;\\n    public static final DecimalFormat DECIMAL_FORMAT = new DecimalFormat();;\\n    final static public String HISTORY_DELIMITER = \";\";;\\n    public static final String DEFAULT_GENOME = \"hg19\";;\\n    public static final String CHR_ALL = \"All\";;\\n    public static final String TRACK_NAME_ATTRIBUTE = \"NAME\";;\\n    public static final String TRACK_DATA_FILE_ATTRIBUTE = \"DATA FILE\";;\\n    public static final String TRACK_DATA_TYPE_ATTRIBUTE = \"DATA TYPE\";;\\n    private static boolean headless = false;;\\n    private static boolean suppressMessages = false;;\\n    private static boolean batch = false;;\\n    private static boolean testing = false;;\\n    // 20 seconds\\npublic static int CONNECT_TIMEOUT = 20000;;\\n    // 3 minutes\\npublic static int READ_TIMEOUT = 1000 * 3 * 60;;\\n    // 1 minute\\npublic static int TOKEN_EXPIRE_GRACE_TIME = 1000 * 1 * 60;;\\n    /**\\n * Field description\\n */\\nfinal public static String SESSION_FILE_EXTENSION = \".xml\";;\\n    /**\\n * GENOME ARCHIVE CONSTANTS\\n */\\nfinal public static String GENOME_FILE_EXTENSION = \".genome\";;\\n    final public static String ZIP_EXTENSION = \".zip\";;\\n    final public static String GZIP_FILE_EXTENSION = \".gz\";;\\n    // Default user folder\\nfinal static public Pattern commaPattern = Pattern.compile(\",\");;\\n    final static public Pattern tabPattern = Pattern.compile(\"\\\\t\");;\\n    final static public Pattern multiTabPattern = Pattern.compile(\"\\\\t+\");;\\n    final static public Pattern colonPattern = Pattern.compile(\":\");;\\n    final static public Pattern dashPattern = Pattern.compile(\"-\");;\\n    final static public Pattern equalPattern = Pattern.compile(\"=\");;\\n    final static public Pattern semicolonPattern = Pattern.compile(\";\");;\\n    final static public Pattern singleTabMultiSpacePattern = Pattern.compile(\"\\\\t|( +)\");;\\n    final static public Pattern forwardSlashPattern = Pattern.compile(\"/\");;\\n    final static public Pattern whitespacePattern = Pattern.compile(\"\\\\\\\\s+\");;\\n    public static List emptyList = new ArrayList();;\\n    public static String VERSION;;\\n    public static String BUILD;;\\n    public static String TIMESTAMP;;\\n    final public static boolean IS_WINDOWS = System.getProperty(\"os.name\").toLowerCase().startsWith(\"windows\");;\\n    final public static boolean IS_MAC = System.getProperty(\"os.name\").toLowerCase().startsWith(\"mac\");;\\n    final public static boolean IS_LINUX = System.getProperty(\"os.name\").toLowerCase().contains(\"linux\");;\\n    final public static boolean IS_JWS = System.getProperty(\"webstart.version\", null) != null || System.getProperty(\"javawebstart.version\", null) != null;;\\n    public static final String JAVA_VERSION_STRING = \"java.version\";;\\n    //Location of bedtools executable\\n//Note: It is recommended you use an absolute path here.\\n//System paths can be finnicky and vary depending on how IGV is launched\\n//However, the path environment variable will be checked if the executable\\n//is named rather than the full path given\\n//\"bedtools\"\\npublic static String BEDtoolsPath = \"/usr/local/bin/bedtools\";;\\n    public static boolean toolsMenuEnabled = false;;\\n    public static boolean development = false;;\\n    public static String downloadURL = \"https://software.broadinstitute.org/software/igv/download\";;\\n    public static void setHeadless(boolean bool)\\n    public static boolean isHeadless()\\n    public static void setTesting(boolean testing)\\n    public static boolean isTesting()\\n    public static void setSuppressMessages(boolean bool)\\n    public static boolean isSuppressMessages()\\n    public static String applicationString()\\n    public static String versionString()\\n    public static boolean isDevelopment()\\n    public static boolean isBatch()\\n    public static void setBatch(boolean batch)\\n    public static boolean checkJavaVersion(String minVersion)\\n    public static double log2(final double value)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/Globals.java', 'methodCount': 13, 'lineCount': 58}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/feature/Strand.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/track/WindowFunction.java', 'methodCount': 0, 'lineCount': 0}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{'Alignment': 'public interface Alignment {\\n     String getReadName()\\n     String getReadSequence()\\n     String getChr()\\n     int getAlignmentStart()\\n     int getAlignmentEnd()\\n     boolean contains(double location)\\n     AlignmentBlock[] getAlignmentBlocks()\\n     AlignmentBlock[] getInsertions()\\n     String getCigarString()\\n     java.util.List<Gap> getGaps()\\n     int getInferredInsertSize()\\n     int getLeadingHardClipLength()\\n     int getMappingQuality()\\n     ReadMate getMate()\\n     Strand getReadStrand()\\n     boolean isProperPair()\\n     boolean isMapped()\\n     boolean isPaired()\\n     boolean isFirstOfPair()\\n     boolean isSecondOfPair()\\n     boolean isNegativeStrand()\\n     boolean isDuplicate()\\n     boolean isPrimary()\\n     boolean isSupplementary()\\n     byte getBase(double position)\\n     byte getPhred(double position)\\n     Object getAttribute(String key)\\n     void setMateSequence(String sequence)\\n     String getPairOrientation()\\n     Strand getFirstOfPairStrand()\\n     Strand getSecondOfPairStrand()\\n     boolean isVendorFailedRead()\\n     Color getYcColor()\\n     String getSample()\\n     String getReadGroup()\\n     String getLibrary()\\n     String getClipboardString(double location, int mouseX)\\n     void finish()\\n     AlignmentBlock getInsertionAt(int position)\\n     Gap getDeletionAt(int position)\\n     void setHaplotypeName(String hap)\\n     String getHaplotypeName()\\n     void setHapDistance(int dist)\\n     int getHapDistance()\\n     Map<Integer, BaseModificationUtils> getBaseModificationMap()\\n     List<BaseModificationSet> getBaseModificationSets()\\n     SMRTKinetics getSmrtKinetics()\\n     String getAlignmentValueString(double position, int mouseX, AlignmentTrack.RenderOptions renderOptions)\\n     Alignment getSpecificAlignment(double location)\\n}\\n'}\n",
      "/**\n",
      "     * This chooses the smallest overlapping alignment (i.e. if a group of linked alignments overlap take the smallest one).\n",
      "     */\n",
      "public Alignment getSpecificAlignment(final double location)\n",
      "**************************************\n",
      "@Override\n",
      "public Alignment getSpecificAlignment(final double location) {\n",
      "    Alignment sa = null;\n",
      "    for (Alignment a : alignments) {\n",
      "        if (a.contains(location)) {\n",
      "            if (sa == null || (a.getLengthOnReference() < sa.getLengthOnReference())) {\n",
      "                sa = a;\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return sa;\n",
      "}\n",
      "1658\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 914/1243 [31:01<11:28,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40184\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 915/1243 [31:03<11:39,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40215\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 916/1243 [31:05<11:49,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40285\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 917/1243 [31:07<11:11,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40286\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 918/1243 [31:10<11:38,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40288\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 43, 'classCount': 1, 'methodsOnly': 'public class Globals {\\n    private static Logger log = LogManager.getLogger(Globals.class);;\\n    public static final int DESIGN_DPI = 96;;\\n    public static final DecimalFormat DECIMAL_FORMAT = new DecimalFormat();;\\n    final static public String HISTORY_DELIMITER = \";\";;\\n    public static final String DEFAULT_GENOME = \"hg19\";;\\n    public static final String CHR_ALL = \"All\";;\\n    public static final String TRACK_NAME_ATTRIBUTE = \"NAME\";;\\n    public static final String TRACK_DATA_FILE_ATTRIBUTE = \"DATA FILE\";;\\n    public static final String TRACK_DATA_TYPE_ATTRIBUTE = \"DATA TYPE\";;\\n    private static boolean headless = false;;\\n    private static boolean suppressMessages = false;;\\n    private static boolean batch = false;;\\n    private static boolean testing = false;;\\n    // 20 seconds\\npublic static int CONNECT_TIMEOUT = 20000;;\\n    // 3 minutes\\npublic static int READ_TIMEOUT = 1000 * 3 * 60;;\\n    // 1 minute\\npublic static int TOKEN_EXPIRE_GRACE_TIME = 1000 * 1 * 60;;\\n    /**\\n * Field description\\n */\\nfinal public static String SESSION_FILE_EXTENSION = \".xml\";;\\n    /**\\n * GENOME ARCHIVE CONSTANTS\\n */\\nfinal public static String GENOME_FILE_EXTENSION = \".genome\";;\\n    final public static String ZIP_EXTENSION = \".zip\";;\\n    final public static String GZIP_FILE_EXTENSION = \".gz\";;\\n    // Default user folder\\nfinal static public Pattern commaPattern = Pattern.compile(\",\");;\\n    final static public Pattern tabPattern = Pattern.compile(\"\\\\t\");;\\n    final static public Pattern multiTabPattern = Pattern.compile(\"\\\\t+\");;\\n    final static public Pattern colonPattern = Pattern.compile(\":\");;\\n    final static public Pattern dashPattern = Pattern.compile(\"-\");;\\n    final static public Pattern equalPattern = Pattern.compile(\"=\");;\\n    final static public Pattern semicolonPattern = Pattern.compile(\";\");;\\n    final static public Pattern singleTabMultiSpacePattern = Pattern.compile(\"\\\\t|( +)\");;\\n    final static public Pattern forwardSlashPattern = Pattern.compile(\"/\");;\\n    final static public Pattern whitespacePattern = Pattern.compile(\"\\\\\\\\s+\");;\\n    public static List emptyList = new ArrayList();;\\n    public static String VERSION;;\\n    public static String BUILD;;\\n    public static String TIMESTAMP;;\\n    final public static boolean IS_WINDOWS = System.getProperty(\"os.name\").toLowerCase().startsWith(\"windows\");;\\n    final public static boolean IS_MAC = System.getProperty(\"os.name\").toLowerCase().startsWith(\"mac\");;\\n    final public static boolean IS_LINUX = System.getProperty(\"os.name\").toLowerCase().contains(\"linux\");;\\n    final public static boolean IS_JWS = System.getProperty(\"webstart.version\", null) != null || System.getProperty(\"javawebstart.version\", null) != null;;\\n    public static final String JAVA_VERSION_STRING = \"java.version\";;\\n    //Location of bedtools executable\\n//Note: It is recommended you use an absolute path here.\\n//System paths can be finnicky and vary depending on how IGV is launched\\n//However, the path environment variable will be checked if the executable\\n//is named rather than the full path given\\n//\"bedtools\"\\npublic static String BEDtoolsPath = \"/usr/local/bin/bedtools\";;\\n    public static boolean toolsMenuEnabled = false;;\\n    public static boolean development = false;;\\n    public static String downloadURL = \"https://software.broadinstitute.org/software/igv/download\";;\\n    public static void setHeadless(boolean bool)\\n    public static boolean isHeadless()\\n    public static void setTesting(boolean testing)\\n    public static boolean isTesting()\\n    public static void setSuppressMessages(boolean bool)\\n    public static boolean isSuppressMessages()\\n    public static String applicationString()\\n    public static String versionString()\\n    public static boolean isDevelopment()\\n    public static boolean isBatch()\\n    public static void setBatch(boolean batch)\\n    public static boolean checkJavaVersion(String minVersion)\\n    public static double log2(final double value)\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/Globals.java', 'methodCount': 13, 'lineCount': 58}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface AlignmentReader {\\n     void close() throws IOException\\n     List<String> getSequenceNames() throws IOException\\n     Map<String, Long> getSequenceDictionary()\\n     SAMFileHeader getFileHeader()\\n     Set<String> getPlatforms()\\n     CloseableIterator<T> iterator() throws IOException\\n     CloseableIterator<T> query(final String sequence, final int start, final int end, final boolean contained) throws IOException\\n     void cancelQuery()\\n     boolean hasIndex()\\n}\\n', 'filePath': '/home/ubuntu/repos/igvteam,igv/src/main/java/org/broad/igv/sam/reader/AlignmentReader.java', 'methodCount': 9, 'lineCount': 11}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'ByteSubarray': 'public class ByteSubarray {\\n    public byte[] backingArray;;\\n    public int startOffset;;\\n    public int length;;\\n    byte fillByte;;\\n    public ByteSubarray(byte[] backingArray, int startOffset, int length, byte fillByte);\\n    public byte getByte(int idx)\\n    public byte[] copyOfRange(int start, int end)\\n    public String getString()\\n    public byte[] getBytes()\\n}\\n'}\n",
      "/**\n",
      "     * Reverses and complements a copy of the original array\n",
      "     */\n",
      "public static ByteSubarray reverseComplementCopy(ByteSubarray bases)\n",
      "**************************************\n",
      "public static ByteSubarray reverseComplementCopy(ByteSubarray bases) {\n",
      "    final int lastIndex = bases.length - 1;\n",
      "    byte[] out = new byte[bases.length];\n",
      "    int i;\n",
      "    for (i = 0; i <= lastIndex; i++) {\n",
      "        out[lastIndex - i] = complement(bases.getByte(i));\n",
      "    }\n",
      "    return new ByteSubarray(out, 0, out.length, bases.fillByte);\n",
      "}\n",
      "1392\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 919/1243 [31:12<11:44,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40406\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 920/1243 [31:14<11:42,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40447\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 921/1243 [31:16<11:24,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40546\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 922/1243 [31:18<11:32,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "40962\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 923/1243 [31:20<11:37,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41106\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 13}\n",
      "{'RangeSliderDemo': 'public class RangeSliderDemo {\\n    private JLabel rangeSliderLabel1 = new JLabel();;\\n    private JLabel rangeSliderValue1 = new JLabel();;\\n    private JLabel rangeSliderLabel2 = new JLabel();;\\n    private JLabel rangeSliderValue2 = new JLabel();;\\n    private RangeSlider rangeSlider = new RangeSlider();;\\n    public RangeSliderDemo();\\n    public void display()\\n    public static void main(String[] args)\\n}\\n'}\n",
      "/**\n",
      "     * Main application method.\n",
      "     * @param args String[]\n",
      "     */\n",
      "public static void main(String[] args)\n",
      "**************************************\n",
      "public static void main(String[] args) {\n",
      "    try {\n",
      "        UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());\n",
      "    } catch (Exception ex) {\n",
      "        ex.printStackTrace();\n",
      "    }\n",
      "    SwingUtilities.invokeLater(new Runnable() {\n",
      "\n",
      "        public void run() {\n",
      "            new RangeSliderDemo().display();\n",
      "        }\n",
      "    });\n",
      "}\n",
      "229\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 924/1243 [31:22<10:45,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41120\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 925/1243 [31:24<10:32,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41134\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  74%|  | 926/1243 [31:26<10:37,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41156\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 927/1243 [31:28<10:36,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41587\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 928/1243 [31:30<10:05,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41655\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Log {\\n    private static LogLevel globalLogLevel = LogLevel.INFO;;\\n    private static PrintStream out = System.err;;\\n    private final Class<?> clazz;;\\n    private final String className;;\\n    private Log(final Class<?> clazz);\\n    public static Log getInstance(final Class<?> clazz)\\n    public static void setGlobalLogLevel(final LogLevel logLevel)\\n    public static LogLevel getGlobalLogLevel()\\n    public static void setGlobalPrintStream(final PrintStream stream)\\n    public static PrintStream getGlobalPrintStream()\\n    public static final boolean isEnabled(final LogLevel level)\\n    private void emit(final LogLevel level, final Throwable throwable, final Object... parts)\\n    protected String getTimestamp()\\n    public final void error(final Throwable throwable, final Object... messageParts)\\n    public final void warn(final Throwable throwable, final Object... messageParts)\\n    public final void info(final Throwable throwable, final Object... messageParts)\\n    public final void debug(final Throwable throwable, final Object... messageParts)\\n    public final void error(final Object... messageParts)\\n    public final void warn(final Object... messageParts)\\n    public final void info(final Object... messageParts)\\n    public final void debug(final Object... messageParts)\\n    private final void getAndEmitIfEnabled(LogLevel logLevel, final Supplier<Object> messageParts)\\n    public final void error(final Supplier<Object> messageParts)\\n    public final void warn(final Supplier<Object> messageParts)\\n    public final void info(final Supplier<Object> messageParts)\\n    public final void debug(final Supplier<Object> messageParts)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/Log.java', 'methodCount': 22, 'lineCount': 28}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 14}\n",
      "{}\n",
      "/**\n",
      "     * Scans the entries in a ZIP/JAR file for classes under the parent package.\n",
      "     * @param file the jar file to be scanned\n",
      "     * @param packagePath the top level package to start from\n",
      "     */\n",
      "protected void scanJar(final File file, final String packagePath) throws IOException\n",
      "**************************************\n",
      "protected void scanJar(final File file, final String packagePath) throws IOException {\n",
      "    final ZipFile zip = new ZipFile(file);\n",
      "    final Enumeration<? extends ZipEntry> entries = zip.entries();\n",
      "    while (entries.hasMoreElements()) {\n",
      "        final ZipEntry entry = entries.nextElement();\n",
      "        final String name = entry.getName();\n",
      "        if (name.startsWith(packagePath)) {\n",
      "            handleItem(name);\n",
      "        }\n",
      "    }\n",
      "}\n",
      "584\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 929/1243 [31:31<09:40,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41656\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 930/1243 [31:34<10:07,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41674\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 931/1243 [31:35<09:44,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41678\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 932/1243 [31:37<09:34,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41680\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 933/1243 [31:39<09:58,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41687\n",
      "{'classCount': 6, 'imports': [{'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class AbstractFeatureReader {\\n    // the logging destination for this source\\n//private final static Logger log = Logger.getLogger(\"BasicFeatureSource\");\\n/**\\n * The path to underlying data file, this must be the input path converted with {@link FeatureCodec#getPathToDataFile(String)}\\n */\\nString path;;\\n    // a wrapper to apply to the raw stream of the Feature file to allow features like prefetching and caching to be injected\\nfinal Function<SeekableByteChannel, SeekableByteChannel> wrapper;;\\n    // a wrapper to apply to the raw stream of the index file\\nfinal Function<SeekableByteChannel, SeekableByteChannel> indexWrapper;;\\n    // the query source, codec, and header\\n// protected final QuerySource querySource;\\nprotected final FeatureCodec<T, SOURCE> codec;;\\n    protected FeatureCodecHeader header;;\\n    private static ComponentMethods methods = new ComponentMethods();;\\n    /**\\n * @deprecated since June 2019 use {@link FileExtensions#BLOCK_COMPRESSED} instead.\\n */\\n@Deprecated\\npublic static final Set<String> BLOCK_COMPRESSED_EXTENSIONS = FileExtensions.BLOCK_COMPRESSED;;\\n    public static AbstractFeatureReader<FEATURE, SOURCE> getFeatureReader(final String featureFile, final FeatureCodec<FEATURE, SOURCE> codec) throws TribbleException\\n    public static AbstractFeatureReader<FEATURE, SOURCE> getFeatureReader(final String featureResource, final FeatureCodec<FEATURE, SOURCE> codec, final boolean requireIndex) throws TribbleException\\n    public static AbstractFeatureReader<FEATURE, SOURCE> getFeatureReader(final String featureResource, String indexResource, final FeatureCodec<FEATURE, SOURCE> codec, final boolean requireIndex) throws TribbleException\\n    public static AbstractFeatureReader<FEATURE, SOURCE> getFeatureReader(final String featureResource, String indexResource, final FeatureCodec<FEATURE, SOURCE> codec, final boolean requireIndex, Function<SeekableByteChannel, SeekableByteChannel> wrapper, Function<SeekableByteChannel, SeekableByteChannel> indexWrapper) throws TribbleException\\n    public static AbstractFeatureReader<FEATURE, SOURCE> getFeatureReader(final String featureResource, final FeatureCodec<FEATURE, SOURCE> codec, final Index index) throws TribbleException\\n    protected AbstractFeatureReader(final String path, final FeatureCodec<T, SOURCE> codec);\\n    protected AbstractFeatureReader(final String path, final FeatureCodec<T, SOURCE> codec, final Function<SeekableByteChannel, SeekableByteChannel> wrapper, final Function<SeekableByteChannel, SeekableByteChannel> indexWrapper);\\n    public boolean hasIndex()\\n    public boolean isQueryable()\\n    public static void setComponentMethods(ComponentMethods methods)\\n    public static boolean hasBlockCompressedExtension(final String fileName)\\n    public static boolean hasBlockCompressedExtension(final File file)\\n    public static boolean hasBlockCompressedExtension(final URI uri)\\n    public Object getHeader()\\n    public static boolean isTabix(String resourcePath, String indexPath) throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/AbstractFeatureReader.java', 'methodCount': 15, 'lineCount': 24}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SimpleFeature {\\n    private final String contig;;\\n    private final int start;;\\n    private final int end;;\\n    public SimpleFeature(final String contig, final int start, final int end);\\n    public String getContig()\\n    public int getStart()\\n    public int getEnd()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/SimpleFeature.java', 'methodCount': 4, 'lineCount': 9}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class BinaryFeatureCodec {\\n    public PositionalBufferedStream makeSourceFromStream(final InputStream bufferedInputStream)\\n    public LocationAware makeIndexableSourceFromStream(final InputStream bufferedInputStream)\\n    public void close(final PositionalBufferedStream source)\\n    public boolean isDone(final PositionalBufferedStream source)\\n    public final TabixFormat getTabixFormat()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/BinaryFeatureCodec.java', 'methodCount': 5, 'lineCount': 7}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Feature {\\n    public String getChr()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/Feature.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface FeatureCodec {\\n    public Feature decodeLoc(final SOURCE source) throws IOException\\n    public FEATURE_TYPE decode(final SOURCE source) throws IOException\\n    public FeatureCodecHeader readHeader(final SOURCE source) throws IOException\\n    public Class<FEATURE_TYPE> getFeatureType()\\n    public SOURCE makeSourceFromStream(final InputStream bufferedInputStream)\\n    public LocationAware makeIndexableSourceFromStream(final InputStream inputStream)\\n    public boolean isDone(final SOURCE source)\\n    public void close(final SOURCE source)\\n    public boolean canDecode(final String path)\\n    public TabixFormat getTabixFormat()\\n     String getPathToDataFile(String path)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/FeatureCodec.java', 'methodCount': 11, 'lineCount': 13}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': \"public class FeatureCodecHeader {\\n    /**\\n * The value of the headerEnd field when there's no header\\n */\\npublic final static long NO_HEADER_END = 0;;\\n    /**\\n * A public instance representing no header\\n */\\npublic final static FeatureCodecHeader EMPTY_HEADER = new FeatureCodecHeader(null, NO_HEADER_END);;\\n    private final Object headerValue;;\\n    private final long headerEnd;;\\n    public FeatureCodecHeader(final Object headerValue, final long headerEnd);\\n    public Object getHeaderValue()\\n    public long getHeaderEnd()\\n    public boolean skipHeaderBytes()\\n}\\n\", 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/FeatureCodecHeader.java', 'methodCount': 4, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface FeatureReader {\\n     CloseableTribbleIterator<T> query(final String chr, final int start, final int end) throws IOException\\n     CloseableTribbleIterator<T> query(Locatable locus) throws IOException\\n     CloseableTribbleIterator<T> iterator() throws IOException\\n     void close() throws IOException\\n     List<String> getSequenceNames()\\n     Object getHeader()\\n     boolean isQueryable()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/FeatureReader.java', 'methodCount': 7, 'lineCount': 9}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': \"public class AsciiLineReader {\\n    private static final Log log = Log.getInstance(AsciiLineReader.class);;\\n    private static final int BUFFER_OVERFLOW_INCREASE_FACTOR = 2;;\\n    private static final byte LINEFEED = (byte) ('\\\\n' & 0xff);;\\n    private static final byte CARRIAGE_RETURN = (byte) ('\\\\r' & 0xff);;\\n    private PositionalBufferedStream is;;\\n    private char[] lineBuffer;;\\n    private int lineTerminatorLength = -1;;\\n    protected AsciiLineReader();\\n    public AsciiLineReader(final InputStream is);\\n    public AsciiLineReader(final PositionalBufferedStream is);\\n    public static AsciiLineReader from(final InputStream inputStream)\\n    public long getPosition()\\n    public int getLineTerminatorLength()\\n    public String readLine(final PositionalBufferedStream stream) throws IOException\\n    public String readLine() throws IOException\\n    public void close()\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/readers/AsciiLineReader.java', 'methodCount': 10, 'lineCount': 19}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface LineIterator {\\n    public String peek()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/readers/LineIterator.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class PositionalBufferedStream {\\n    final InputStream is;;\\n    final byte[] buffer;;\\n    int nextChar;;\\n    int nChars;;\\n    long position;;\\n    public PositionalBufferedStream(final InputStream is);\\n    public PositionalBufferedStream(final InputStream is, final int bufferSize);\\n    public final long getPosition()\\n    public final int read() throws IOException\\n    public final int read(final byte[] bytes, final int start, final int len) throws IOException\\n    public final int read(final byte[] bytes) throws IOException\\n    public final boolean isDone() throws IOException\\n    public final int peek() throws IOException\\n    private final int fill() throws IOException\\n    public final long skip(final long nBytes) throws IOException\\n    public final void close()\\n    private static final int byteToInt(byte b)\\n    public static void main(String[] args) throws Exception\\n    private static void readFileByLine(final String name, final InputStream is) throws IOException\\n    private static void readFileInChunks(final String name, final InputStream is) throws IOException\\n    private static final void printStatus(final String name, long lineCount, double rate, long dt)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/readers/PositionalBufferedStream.java', 'methodCount': 16, 'lineCount': 23}], 'asteriskCount': 0, 'importCount': 6, 'javaImportCount': 9}\n",
      "{}\n",
      "/**\n",
      "     * Convenience method that creates an ExampleBinaryCodec file from another feature file.\n",
      "     *\n",
      "     * For testing purposes really\n",
      "     *\n",
      "     * @throws IOException\n",
      "     */\n",
      "public static void convertToBinaryTest(final FeatureReader<FEATURE_TYPE> reader, final OutputStream out) throws IOException\n",
      "**************************************\n",
      "public static <FEATURE_TYPE extends Feature> void convertToBinaryTest(final FeatureReader<FEATURE_TYPE> reader, final OutputStream out) throws IOException {\n",
      "    DataOutputStream dos = new DataOutputStream(out);\n",
      "    dos.writeBytes(HEADER_LINE + \"\\n\");\n",
      "    Iterator<FEATURE_TYPE> it = reader.iterator();\n",
      "    while (it.hasNext()) {\n",
      "        final Feature f = it.next();\n",
      "        dos.writeUTF(f.getContig());\n",
      "        dos.writeInt(f.getStart());\n",
      "        dos.writeInt(f.getEnd());\n",
      "    }\n",
      "    dos.close();\n",
      "    reader.close();\n",
      "}\n",
      "2514\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 934/1243 [31:42<10:32,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41727\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 935/1243 [31:43<10:02,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41736\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 936/1243 [31:45<09:46,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41831\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 937/1243 [31:47<09:44,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42022\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  75%|  | 938/1243 [31:48<08:34,  1.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42078\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Resets the stream to the most recent mark.\n",
      "     *\n",
      "     * @throws IOException If the stream has never been marked,\n",
      "     *                     or if the mark has been invalidated\n",
      "     */\n",
      "public void reset() throws IOException\n",
      "**************************************\n",
      "@Override\n",
      "public void reset() throws IOException {\n",
      "    synchronized (lock) {\n",
      "        ensureOpen();\n",
      "        if (markedChar < 0)\n",
      "            throw new IOException((markedChar == INVALIDATED) ? \"Mark invalid\" : \"Stream not marked\");\n",
      "        nextChar = markedChar;\n",
      "        skipLF = markedSkipLF;\n",
      "    }\n",
      "}\n",
      "152\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 939/1243 [31:50<09:15,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42243\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 940/1243 [31:52<09:34,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42253\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 941/1243 [31:55<10:07,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42384\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 942/1243 [31:57<09:47,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42387\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 943/1243 [31:59<10:17,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42388\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface HtsHeader {\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/beta/plugin/HtsHeader.java', 'methodCount': 0, 'lineCount': 2}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SAMSequenceDictionary {\\n    public static final long serialVersionUID = 1L;;\\n    private List<SAMSequenceRecord> mSequences = new ArrayList<>();;\\n    private final Map<String, SAMSequenceRecord> mSequenceMap = new HashMap<>();;\\n    public SAMSequenceDictionary();\\n    public SAMSequenceDictionary(final List<SAMSequenceRecord> list);\\n    public List<SAMSequenceRecord> getSequences()\\n    private static Log log = Log.getInstance(SAMSequenceDictionary.class);;\\n    public SAMSequenceRecord getSequence(final String name)\\n    public void setSequences(final List<SAMSequenceRecord> list)\\n    public void addSequence(final SAMSequenceRecord sequenceRecord)\\n    public SAMSequenceRecord getSequence(final int sequenceIndex)\\n    public int getSequenceIndex(final String sequenceName)\\n    public int size()\\n    public long getReferenceLength()\\n    public boolean isEmpty()\\n    private static String DICT_MISMATCH_TEMPLATE = \"SAM dictionaries are not the same: %s.\";;\\n    public void assertSameDictionary(final SAMSequenceDictionary that)\\n    public boolean isSameDictionary(final SAMSequenceDictionary that)\\n    public boolean equals(Object o)\\n    public SAMSequenceRecord addSequenceAlias(final String originalName, final String altName)\\n    public SAMSequenceRecord addAlternativeSequenceName(final String originalName, final String altName)\\n    public String md5()\\n    public int hashCode()\\n    public String toString()\\n    public static final List<String> DEFAULT_DICTIONARY_EQUAL_TAG = Arrays.asList(SAMSequenceRecord.MD5_TAG, SAMSequenceRecord.SEQUENCE_LENGTH_TAG);;\\n    public static SAMSequenceDictionary mergeDictionaries(final SAMSequenceDictionary dict1, final SAMSequenceDictionary dict2, final List<String> tagsToMatch)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMSequenceDictionary.java', 'methodCount': 20, 'lineCount': 28}, {'fieldCount': 21, 'classCount': 1, 'methodsOnly': 'public class SAMSequenceRecord {\\n    // AbstractSAMHeaderRecord implements Serializable\\npublic static final long serialVersionUID = 1L;;\\n    public static final int UNAVAILABLE_SEQUENCE_INDEX = -1;;\\n    // Value must be interned() if it\\'s ever set/modified\\nprivate final String mSequenceName;;\\n    private Set<String> mAlternativeSequenceName = new LinkedHashSet<>();;\\n    private int mSequenceIndex = UNAVAILABLE_SEQUENCE_INDEX;;\\n    private int mSequenceLength = 0;;\\n    public static final String SEQUENCE_NAME_TAG = \"SN\";;\\n    public static final String ALTERNATIVE_SEQUENCE_NAME_TAG = \"AN\";;\\n    public static final String SEQUENCE_LENGTH_TAG = \"LN\";;\\n    public static final String MD5_TAG = \"M5\";;\\n    public static final String ASSEMBLY_TAG = \"AS\";;\\n    public static final String URI_TAG = \"UR\";;\\n    public static final String SPECIES_TAG = \"SP\";;\\n    public static final String DESCRIPTION_TAG = \"DS\";;\\n    /**\\n * If one sequence has this length, and another sequence had a different length, isSameSequence will\\n * not complain that they are different sequences.\\n */\\npublic static final int UNKNOWN_SEQUENCE_LENGTH = 0;;\\n    /**\\n * This is not a valid sequence name, because it is reserved in the RNEXT field of SAM text format\\n * to mean \"same reference as RNAME field.\"\\n */\\npublic static final String RESERVED_RNEXT_SEQUENCE_NAME = \"=\";;\\n    /* use RESERVED_RNEXT_SEQUENCE_NAME instead. */\\n@Deprecated\\npublic static final String RESERVED_MRNM_SEQUENCE_NAME = RESERVED_RNEXT_SEQUENCE_NAME;;\\n    /**\\n * The standard tags are stored in text header without type information, because the type of these tags is known.\\n */\\npublic static final Set<String> STANDARD_TAGS = new HashSet<>(Arrays.asList(SEQUENCE_NAME_TAG, SEQUENCE_LENGTH_TAG, ASSEMBLY_TAG, ALTERNATIVE_SEQUENCE_NAME_TAG, MD5_TAG, URI_TAG, SPECIES_TAG));;\\n    // These are the chars matched by \\\\\\\\s.\\n// \\\\013 is vertical tab\\nprivate static final char[] WHITESPACE_CHARS = { \\' \\', \\'\\\\t\\', \\'\\\\n\\', \\'\\\\013\\', \\'\\\\f\\', \\'\\\\r\\' };;\\n    // alternative sequence name separator\\nprivate static final String ALTERNATIVE_SEQUENCE_NAME_SEPARATOR = \",\";;\\n    private static final Pattern LEGAL_RNAME_PATTERN = Pattern.compile(\"[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*\");;\\n    public SAMSequenceRecord(final String name);\\n    public SAMSequenceRecord(final String name, final int sequenceLength);\\n    public String getSequenceName()\\n    public int getSequenceLength()\\n    public SAMSequenceRecord setSequenceLength(final int value)\\n    public String getAssembly()\\n    public SAMSequenceRecord setAssembly(final String value)\\n    public String getSpecies()\\n    public SAMSequenceRecord setSpecies(final String value)\\n    public String getMd5()\\n    public SAMSequenceRecord setMd5(final String value)\\n    public String getDescription()\\n    public SAMSequenceRecord setDescription(final String value)\\n    public int getSequenceIndex()\\n    public SAMSequenceRecord setSequenceIndex(final int value)\\n    public Set<String> getAlternativeSequenceNames()\\n    public void addAlternativeSequenceName(final String name)\\n    public SAMSequenceRecord setAlternativeSequenceName(final Collection<String> alternativeSequences)\\n    private static void validateAltRegExp(final String name)\\n    private void encodeAltSequences(final Collection<String> alternativeSequences)\\n    public boolean hasAlternativeSequenceNames()\\n    public boolean isSameSequence(final SAMSequenceRecord that)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n     Set<String> getStandardTags()\\n    public final SAMSequenceRecord clone()\\n    public static String truncateSequenceName(final String sequenceName)\\n    public static void validateSequenceName(final String name)\\n    public String toString()\\n    public String getSAMString()\\n    public final String getContig()\\n    public final int getStart()\\n    public final int getEnd()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMSequenceRecord.java', 'methodCount': 33, 'lineCount': 56}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class TribbleException {\\n    // what file or input source we are working from\\nString source;;\\n    public TribbleException(String msg);\\n    public TribbleException(String message, Throwable throwable);\\n    public void setSource(String source)\\n    public String getMessage()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/TribbleException.java', 'methodCount': 4, 'lineCount': 7}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ParsingUtils {\\n    public static final Map<Object, Color> colorCache = new WeakHashMap<>(100);;\\n    private static URLHelperFactory urlHelperFactory = RemoteURLHelper::new;;\\n    // HTML 4.1 color table,  + orange and magenta\\nprivate static Map<String, String> colorSymbols = new HashMap();;\\n    public static InputStream openInputStream(String path) throws IOException\\n    static private final Set<String> URL_SCHEMES = new HashSet<>(Arrays.asList(\"http\", \"ftp\", \"https\"));;\\n    public static InputStream openInputStream(final String uri, final Function<SeekableByteChannel, SeekableByteChannel> wrapper) throws IOException\\n    public static String join(String separator, Collection<T> objects)\\n    public static List<T> sortList(Collection<T> list)\\n    public static String sortedString(Map<T, V> c)\\n    public static String join(String separator, String[] strings)\\n    public static String join(String separator, String[] strings, int start, int end)\\n    public static List<String> split(String input, char delim)\\n    public static int split(String aString, String[] tokens, char delim)\\n    public static int split(String aString, String[] tokens, char delim, boolean condenseTrailingTokens)\\n    public static String trim(String str, char ch)\\n    public static int splitWhitespace(String aString, String[] tokens)\\n    public static boolean isSorted(Iterable<T> iterable)\\n    public static Color parseColor(String string)\\n    private static Color hexToColor(String string)\\n    public static boolean resourceExists(String resource) throws IOException\\n    public static URLHelper getURLHelper(URL url)\\n    public static void setURLHelperFactory(URLHelperFactory factory)\\n    public static URLHelperFactory getURLHelperFactory()\\n    public static String appendToPath(String filepath, String indexExtension)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/util/ParsingUtils.java', 'methodCount': 20, 'lineCount': 26}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ValidationUtils {\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String nameOfObject)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(final T collection, final String nameOfObject)\\n    public static boolean isNonEmpty(final Collection<?> collection)\\n    public static String nonEmpty(final String string, final String nameOfObject)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(final T collection)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/utils/ValidationUtils.java', 'methodCount': 10, 'lineCount': 12}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': \"public class GeneralUtils {\\n    /**\\n * Setting this to true causes the VCF/BCF/VariantContext classes to emit debugging information\\n * to standard error\\n */\\npublic static final boolean DEBUG_MODE_ENABLED = false;;\\n    /**\\n * The smallest log10 value we'll emit from normalizeFromLog10 and other functions\\n * where the real-space value is 0.0.\\n */\\npublic final static double LOG10_P_OF_ZERO = -1000000.0;;\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static double[] normalizeFromLog10(double[] array)\\n    public static double[] normalizeFromLog10(double[] array, boolean takeLog10OfOutput)\\n    public static double[] normalizeFromLog10(double[] array, boolean takeLog10OfOutput, boolean keepInLogSpace)\\n    public static double arrayMax(final double[] array)\\n    public static int maxElementIndex(final double[] array)\\n    public static int maxElementIndex(final double[] array, final int endIndex)\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static byte compareDoubles(double a, double b)\\n    public static byte compareDoubles(double a, double b, double epsilon)\\n    public static final List<T> reverse(final List<T> l)\\n}\\n\", 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/variant/utils/GeneralUtils.java', 'methodCount': 12, 'lineCount': 16}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class VariantContextComparator {\\n    private static final long serialVersionUID = 1L;;\\n    public static List<String> getSequenceNameList(final SAMSequenceDictionary dictionary)\\n    private final Map<String, Integer> contigIndexLookup;;\\n    public VariantContextComparator(final List<String> contigs);\\n    public VariantContextComparator(final Collection<VCFContigHeaderLine> headerLines);\\n    public VariantContextComparator(final SAMSequenceDictionary dictionary);\\n    public int compare(final VariantContext firstVariantContext, final VariantContext secondVariantContext)\\n    public boolean isCompatible(final Collection<VCFContigHeaderLine> headerLines)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/variant/variantcontext/VariantContextComparator.java', 'methodCount': 6, 'lineCount': 10}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 13}\n",
      "{'VCFHeaderLine': 'public class VCFHeaderLine {\\n    public static final long serialVersionUID = 1L;;\\n    protected static final boolean ALLOW_UNBOUND_DESCRIPTIONS = true;;\\n    protected static final String UNBOUND_DESCRIPTION = \"Not provided in original VCF header\";;\\n    private String mKey = null;;\\n    private String mValue = null;;\\n    public VCFHeaderLine(String key, String value);\\n    public String getKey()\\n    public String getValue()\\n    public boolean shouldBeAddedToDictionary()\\n    public String toString()\\n    protected String toStringEncoding()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public int compareTo(Object other)\\n    public static boolean isHeaderLine(String line)\\n    public static String toStringEncoding(Map<String, ? extends Object> keyValues)\\n    private static String escapeQuotes(final String value)\\n}\\n', 'VCFIDHeaderLine': 'public interface VCFIDHeaderLine {\\n     String getID()\\n}\\n'}\n",
      "/**\n",
      "     * @return all of the VCF ID-based header lines in their original file order, or an empty list if none were present\n",
      "     */\n",
      "public List<VCFIDHeaderLine> getIDHeaderLines()\n",
      "**************************************\n",
      "public List<VCFIDHeaderLine> getIDHeaderLines() {\n",
      "    final List<VCFIDHeaderLine> lines = new ArrayList<VCFIDHeaderLine>();\n",
      "    for (final VCFHeaderLine line : mMetaData) {\n",
      "        if (line instanceof VCFIDHeaderLine) {\n",
      "            lines.add((VCFIDHeaderLine) line);\n",
      "        }\n",
      "    }\n",
      "    return lines;\n",
      "}\n",
      "3280\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 944/1243 [32:01<10:37,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42389\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 945/1243 [32:03<10:00,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42494\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 946/1243 [32:05<10:03,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42502\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 947/1243 [32:08<10:54,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42539\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 948/1243 [32:10<10:54,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42581\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * @param bases  bases representing an allele\n",
      "     * @return true if the bases represent a symbolic allele, including breakpoints and breakends\n",
      "     */\n",
      " static boolean wouldBeSymbolicAllele(byte[] bases)\n",
      "**************************************\n",
      "@Deprecated\n",
      "static boolean wouldBeSymbolicAllele(byte[] bases) {\n",
      "    if (bases.length <= 1)\n",
      "        return false;\n",
      "    else {\n",
      "        return bases[0] == Allele.SYMBOLIC_ALLELE_START || bases[bases.length - 1] == Allele.SYMBOLIC_ALLELE_END || wouldBeBreakpoint(bases) || wouldBeSingleBreakend(bases);\n",
      "    }\n",
      "}\n",
      "171\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 949/1243 [32:12<10:52,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42582\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  76%|  | 950/1243 [32:14<11:01,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42668\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 951/1243 [32:17<11:06,  2.28s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42671\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 952/1243 [32:19<11:09,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42683\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 953/1243 [32:21<11:05,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42694\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ParsingUtils {\\n    public static final Map<Object, Color> colorCache = new WeakHashMap<>(100);;\\n    private static URLHelperFactory urlHelperFactory = RemoteURLHelper::new;;\\n    // HTML 4.1 color table,  + orange and magenta\\nprivate static Map<String, String> colorSymbols = new HashMap();;\\n    public static InputStream openInputStream(String path) throws IOException\\n    static private final Set<String> URL_SCHEMES = new HashSet<>(Arrays.asList(\"http\", \"ftp\", \"https\"));;\\n    public static InputStream openInputStream(final String uri, final Function<SeekableByteChannel, SeekableByteChannel> wrapper) throws IOException\\n    public static String join(String separator, Collection<T> objects)\\n    public static List<T> sortList(Collection<T> list)\\n    public static String sortedString(Map<T, V> c)\\n    public static String join(String separator, String[] strings)\\n    public static String join(String separator, String[] strings, int start, int end)\\n    public static List<String> split(String input, char delim)\\n    public static int split(String aString, String[] tokens, char delim)\\n    public static int split(String aString, String[] tokens, char delim, boolean condenseTrailingTokens)\\n    public static String trim(String str, char ch)\\n    public static int splitWhitespace(String aString, String[] tokens)\\n    public static boolean isSorted(Iterable<T> iterable)\\n    public static Color parseColor(String string)\\n    private static Color hexToColor(String string)\\n    public static boolean resourceExists(String resource) throws IOException\\n    public static URLHelper getURLHelper(URL url)\\n    public static void setURLHelperFactory(URLHelperFactory factory)\\n    public static URLHelperFactory getURLHelperFactory()\\n    public static String appendToPath(String filepath, String indexExtension)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/util/ParsingUtils.java', 'methodCount': 20, 'lineCount': 26}, {'fieldCount': 90, 'classCount': 1, 'methodsOnly': 'public class VCFConstants {\\n    public static final Locale VCF_LOCALE = Locale.US;;\\n    // reserved INFO/FORMAT field keys\\npublic static final String ANCESTRAL_ALLELE_KEY = \"AA\";;\\n    public static final String ALLELE_COUNT_KEY = \"AC\";;\\n    public static final String ALLELE_FREQUENCY_KEY = \"AF\";;\\n    public static final String ALLELE_NUMBER_KEY = \"AN\";;\\n    public static final String RMS_BASE_QUALITY_KEY = \"BQ\";;\\n    public static final String CIGAR_KEY = \"CIGAR\";;\\n    public static final String DBSNP_KEY = \"DB\";;\\n    public static final String DEPTH_KEY = \"DP\";;\\n    public static final String END_KEY = \"END\";;\\n    public static final String GENOTYPE_FILTER_KEY = \"FT\";;\\n    public static final String GENOTYPE_KEY = \"GT\";;\\n    public static final String GENOTYPE_POSTERIORS_KEY = \"GP\";;\\n    public static final String GENOTYPE_QUALITY_KEY = \"GQ\";;\\n    //AD isn\\'t reserved, but is specifically handled by VariantContext\\npublic static final String GENOTYPE_ALLELE_DEPTHS = \"AD\";;\\n    // phred-scaled genotype likelihoods\\npublic static final String GENOTYPE_PL_KEY = \"PL\";;\\n    public static final String EXPECTED_ALLELE_COUNT_KEY = \"EC\";;\\n    // log10 scaled genotype likelihoods\\n@Deprecated\\npublic static final String GENOTYPE_LIKELIHOODS_KEY = \"GL\";;\\n    public static final String HAPMAP2_KEY = \"H2\";;\\n    public static final String HAPMAP3_KEY = \"H3\";;\\n    public static final String HAPLOTYPE_QUALITY_KEY = \"HQ\";;\\n    public static final String RMS_MAPPING_QUALITY_KEY = \"MQ\";;\\n    public static final String MAPPING_QUALITY_ZERO_KEY = \"MQ0\";;\\n    public static final String SAMPLE_NUMBER_KEY = \"NS\";;\\n    public static final String PHASE_QUALITY_KEY = \"PQ\";;\\n    public static final String PHASE_SET_KEY = \"PS\";;\\n    public static final String OLD_DEPTH_KEY = \"RD\";;\\n    public static final String STRAND_BIAS_KEY = \"SB\";;\\n    public static final String SOMATIC_KEY = \"SOMATIC\";;\\n    public static final String VALIDATED_KEY = \"VALIDATED\";;\\n    public static final String THOUSAND_GENOMES_KEY = \"1000G\";;\\n    // reserved INFO for structural variants\\n/**\\n * INFO Type of structural variant\\n */\\npublic static final String SVTYPE = \"SVTYPE\";;\\n    // separators\\npublic static final String FORMAT_FIELD_SEPARATOR = \":\";;\\n    public static final String GENOTYPE_FIELD_SEPARATOR = \":\";;\\n    public static final char GENOTYPE_FIELD_SEPARATOR_CHAR = \\':\\';;\\n    public static final String FIELD_SEPARATOR = \"\\\\t\";;\\n    public static final char FIELD_SEPARATOR_CHAR = \\'\\\\t\\';;\\n    public static final String FILTER_CODE_SEPARATOR = \";\";;\\n    public static final String INFO_FIELD_ARRAY_SEPARATOR = \",\";;\\n    public static final char INFO_FIELD_ARRAY_SEPARATOR_CHAR = \\',\\';;\\n    public static final String ID_FIELD_SEPARATOR = \";\";;\\n    public static final String INFO_FIELD_SEPARATOR = \";\";;\\n    public static final char INFO_FIELD_SEPARATOR_CHAR = \\';\\';;\\n    public static final String UNPHASED = \"/\";;\\n    public static final String PHASED = \"|\";;\\n    public static final String PHASED_SWITCH_PROB_v3 = \"\\\\\\\\\";;\\n    public static final String PHASING_TOKENS = \"/|\\\\\\\\\";;\\n    // header lines\\npublic static final String FILTER_HEADER_START = \"##FILTER\";;\\n    public static final String FORMAT_HEADER_START = \"##FORMAT\";;\\n    public static final String INFO_HEADER_START = \"##INFO\";;\\n    public static final String ALT_HEADER_KEY = \"ALT\";;\\n    public static final String ALT_HEADER_START = VCFHeader.METADATA_INDICATOR + ALT_HEADER_KEY;;\\n    public static final String CONTIG_HEADER_KEY = \"contig\";;\\n    public static final String CONTIG_HEADER_START = \"##\" + CONTIG_HEADER_KEY;;\\n    public static final int ALT_HEADER_OFFSET = ALT_HEADER_START.length() + 1;;\\n    public static final String PEDIGREE_HEADER_KEY = \"PEDIGREE\";;\\n    public static final String PEDIGREE_HEADER_START = VCFHeader.METADATA_INDICATOR + PEDIGREE_HEADER_KEY;;\\n    public static final int PEDIGREE_HEADER_OFFSET = PEDIGREE_HEADER_START.length() + 1;;\\n    public static final String SAMPLE_HEADER_KEY = \"SAMPLE\";;\\n    public static final String SAMPLE_HEADER_START = VCFHeader.METADATA_INDICATOR + SAMPLE_HEADER_KEY;;\\n    public static final int SAMPLE_HEADER_OFFSET = SAMPLE_HEADER_START.length() + 1;;\\n    public static final String META_HEADER_KEY = \"META\";;\\n    public static final String META_HEADER_START = VCFHeader.METADATA_INDICATOR + META_HEADER_KEY;;\\n    public static final int META_HEADER_OFFSET = META_HEADER_START.length() + 1;;\\n    // old indel alleles\\npublic static final char DELETION_ALLELE_v3 = \\'D\\';;\\n    public static final char INSERTION_ALLELE_v3 = \\'I\\';;\\n    // special alleles\\npublic static final char SPANNING_DELETION_ALLELE = \\'*\\';;\\n    public static final char NO_CALL_ALLELE = \\'.\\';;\\n    public static final char NULL_ALLELE = \\'-\\';;\\n    // missing/default values\\npublic static final String UNFILTERED = \".\";;\\n    public static final String PASSES_FILTERS_v3 = \"0\";;\\n    public static final String PASSES_FILTERS_v4 = \"PASS\";;\\n    public static final String EMPTY_ID_FIELD = \".\";;\\n    public static final String EMPTY_INFO_FIELD = \".\";;\\n    public static final String EMPTY_ALTERNATE_ALLELE_FIELD = \".\";;\\n    public static final String MISSING_VALUE_v4 = \".\";;\\n    public static final String MISSING_QUALITY_v3 = \"-1\";;\\n    public static final Double MISSING_QUALITY_v3_DOUBLE = Double.valueOf(MISSING_QUALITY_v3);;\\n    public static final String MISSING_GENOTYPE_QUALITY_v3 = \"-1\";;\\n    public static final String MISSING_HAPLOTYPE_QUALITY_v3 = \"-1\";;\\n    public static final String MISSING_DEPTH_v3 = \"-1\";;\\n    public static final String UNBOUNDED_ENCODING_v4 = \".\";;\\n    public static final String UNBOUNDED_ENCODING_v3 = \"-1\";;\\n    public static final String PER_ALTERNATE_COUNT = \"A\";;\\n    public static final String PER_ALLELE_COUNT = \"R\";;\\n    public static final String PER_GENOTYPE_COUNT = \"G\";;\\n    public static final String EMPTY_ALLELE = \".\";;\\n    public static final String EMPTY_GENOTYPE = \"./.\";;\\n    public static final int MAX_GENOTYPE_QUAL = 99;;\\n    // when we consider fields equal(), used in the Qual compare\\npublic static final Double VCF_ENCODING_EPSILON = 0.00005;;\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/variant/vcf/VCFConstants.java', 'methodCount': 0, 'lineCount': 92}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     * Reset all of the builder attributes to their defaults.  After this\n",
      "     * function you must provide sampleName and alleles before trying to\n",
      "     * make more Genotypes.\n",
      "     */\n",
      "public final void reset(final boolean keepSampleName)\n",
      "**************************************\n",
      "public final void reset(final boolean keepSampleName) {\n",
      "    if (!keepSampleName)\n",
      "        sampleName = null;\n",
      "    alleles = Collections.emptyList();\n",
      "    isPhased = false;\n",
      "    GQ = -1;\n",
      "    DP = -1;\n",
      "    AD = null;\n",
      "    PL = null;\n",
      "    filters = null;\n",
      "    extendedAttributes = null;\n",
      "}\n",
      "2318\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 954/1243 [32:24<11:02,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42870\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 955/1243 [32:26<10:58,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42876\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 956/1243 [32:28<11:01,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42878\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 957/1243 [32:30<10:04,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42880\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 958/1243 [32:32<10:20,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42902\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 13}\n",
      "{'Genotype': 'public class Genotype {\\n    public static final long serialVersionUID = 1L;;\\n    /**\\n * A list of genotype field keys corresponding to values we\\n * manage inline in the Genotype object.  They must not appear in the\\n * extended attributes map\\n */\\npublic final static Collection<String> PRIMARY_KEYS = Arrays.asList(VCFConstants.GENOTYPE_FILTER_KEY, VCFConstants.GENOTYPE_KEY, VCFConstants.GENOTYPE_QUALITY_KEY, VCFConstants.DEPTH_KEY, VCFConstants.GENOTYPE_ALLELE_DEPTHS, VCFConstants.GENOTYPE_PL_KEY);;\\n    public final static String PHASED_ALLELE_SEPARATOR = \"|\";;\\n    public final static String UNPHASED_ALLELE_SEPARATOR = \"/\";;\\n    private final String sampleName;;\\n    private GenotypeType type = null;;\\n    private final String filters;;\\n    protected Genotype(final String sampleName, final String filters);\\n    public abstract List<Allele> getAlleles()\\n    public int countAllele(final Allele allele)\\n    public abstract Allele getAllele(int i)\\n    public abstract boolean isPhased()\\n    public int getPloidy()\\n    public abstract int getDP()\\n    public abstract int[] getAD()\\n    public String getSampleName()\\n    public abstract int getGQ()\\n    public boolean hasPL()\\n    public boolean hasAD()\\n    public boolean hasGQ()\\n    public boolean hasDP()\\n    public GenotypeType getType()\\n    protected GenotypeType determineType()\\n    public boolean isHom()\\n    public boolean isHomRef()\\n    public boolean isHomVar()\\n    public boolean isHet()\\n    public boolean isHetNonRef()\\n    public boolean isNoCall()\\n    public boolean isCalled()\\n    public boolean isMixed()\\n    public boolean isAvailable()\\n    public boolean hasLikelihoods()\\n    public String getLikelihoodsString()\\n    public GenotypeLikelihoods getLikelihoods()\\n    public boolean isNonInformative()\\n    public abstract int[] getPL()\\n    public String getGenotypeString()\\n    public String getGenotypeString(boolean ignoreRefState)\\n    protected List<String> getAlleleStrings()\\n    public String toString()\\n    public String toBriefString()\\n    public int compareTo(final Genotype genotype)\\n    public boolean sameGenotype(final Genotype other)\\n    public boolean sameGenotype(final Genotype other, boolean ignorePhase)\\n    public abstract Map<String, Object> getExtendedAttributes()\\n    public boolean hasExtendedAttribute(final String key)\\n    public Object getExtendedAttribute(final String key, final Object defaultValue)\\n    public Object getExtendedAttribute(final String key)\\n    public final String getFilters()\\n    public final boolean isFiltered()\\n    public boolean hasLog10PError()\\n    public double getLog10PError()\\n    public int getPhredScaledQual()\\n    public String getAttributeAsString(String key, String defaultValue)\\n    public int getAttributeAsInt(String key, int defaultValue)\\n    public double getAttributeAsDouble(String key, double defaultValue)\\n    public Object getAnyAttribute(final String key)\\n    public boolean hasAnyAttribute(final String key)\\n    protected static String sortedString(Map<T, V> c)\\n    protected static final String toStringIfExists(final String name, final int v)\\n    protected static final String toStringIfExists(final String name, final String v)\\n    protected static final String toStringIfExists(final String name, final int[] vs)\\n    protected static final boolean hasForbiddenKey(final Map<String, Object> attributes)\\n    protected static final boolean isForbiddenKey(final String key)\\n}\\n'}\n",
      "/**\n",
      "     * Adds all of the genotypes to this context\n",
      "     *\n",
      "     * See {@link #add(Genotype)} for important information about this functions\n",
      "     * constraints and performance costs\n",
      "     *\n",
      "     * @param genotypes\n",
      "     * @return\n",
      "     */\n",
      "public boolean addAll(final Collection<? extends Genotype> genotypes)\n",
      "**************************************\n",
      "@Override\n",
      "public boolean addAll(final Collection<? extends Genotype> genotypes) {\n",
      "    checkImmutability();\n",
      "    invalidateSampleOrdering();\n",
      "    if (sampleNameToOffset != null) {\n",
      "        int pos = size();\n",
      "        for (final Genotype g : genotypes) {\n",
      "            sampleNameToOffset.put(g.getSampleName(), pos++);\n",
      "        }\n",
      "    }\n",
      "    return getGenotypes().addAll(genotypes);\n",
      "}\n",
      "1000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 959/1243 [32:34<10:23,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42908\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 960/1243 [32:37<10:23,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42931\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 961/1243 [32:39<10:15,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42976\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 962/1243 [32:41<09:40,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43028\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  77%|  | 963/1243 [32:43<09:59,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43058\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface HtsRecord {\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/beta/plugin/HtsRecord.java', 'methodCount': 0, 'lineCount': 2}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface Feature {\\n    public String getChr()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/Feature.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class TribbleException {\\n    // what file or input source we are working from\\nString source;;\\n    public TribbleException(String msg);\\n    public TribbleException(String message, Throwable throwable);\\n    public void setSource(String source)\\n    public String getMessage()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/TribbleException.java', 'methodCount': 4, 'lineCount': 7}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class ParsingUtils {\\n    public static final Map<Object, Color> colorCache = new WeakHashMap<>(100);;\\n    private static URLHelperFactory urlHelperFactory = RemoteURLHelper::new;;\\n    // HTML 4.1 color table,  + orange and magenta\\nprivate static Map<String, String> colorSymbols = new HashMap();;\\n    public static InputStream openInputStream(String path) throws IOException\\n    static private final Set<String> URL_SCHEMES = new HashSet<>(Arrays.asList(\"http\", \"ftp\", \"https\"));;\\n    public static InputStream openInputStream(final String uri, final Function<SeekableByteChannel, SeekableByteChannel> wrapper) throws IOException\\n    public static String join(String separator, Collection<T> objects)\\n    public static List<T> sortList(Collection<T> list)\\n    public static String sortedString(Map<T, V> c)\\n    public static String join(String separator, String[] strings)\\n    public static String join(String separator, String[] strings, int start, int end)\\n    public static List<String> split(String input, char delim)\\n    public static int split(String aString, String[] tokens, char delim)\\n    public static int split(String aString, String[] tokens, char delim, boolean condenseTrailingTokens)\\n    public static String trim(String str, char ch)\\n    public static int splitWhitespace(String aString, String[] tokens)\\n    public static boolean isSorted(Iterable<T> iterable)\\n    public static Color parseColor(String string)\\n    private static Color hexToColor(String string)\\n    public static boolean resourceExists(String resource) throws IOException\\n    public static URLHelper getURLHelper(URL url)\\n    public static void setURLHelperFactory(URLHelperFactory factory)\\n    public static URLHelperFactory getURLHelperFactory()\\n    public static String appendToPath(String filepath, String indexExtension)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/util/ParsingUtils.java', 'methodCount': 20, 'lineCount': 26}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': \"public class GeneralUtils {\\n    /**\\n * Setting this to true causes the VCF/BCF/VariantContext classes to emit debugging information\\n * to standard error\\n */\\npublic static final boolean DEBUG_MODE_ENABLED = false;;\\n    /**\\n * The smallest log10 value we'll emit from normalizeFromLog10 and other functions\\n * where the real-space value is 0.0.\\n */\\npublic final static double LOG10_P_OF_ZERO = -1000000.0;;\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static double[] normalizeFromLog10(double[] array)\\n    public static double[] normalizeFromLog10(double[] array, boolean takeLog10OfOutput)\\n    public static double[] normalizeFromLog10(double[] array, boolean takeLog10OfOutput, boolean keepInLogSpace)\\n    public static double arrayMax(final double[] array)\\n    public static int maxElementIndex(final double[] array)\\n    public static int maxElementIndex(final double[] array, final int endIndex)\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static byte compareDoubles(double a, double b)\\n    public static byte compareDoubles(double a, double b, double epsilon)\\n    public static final List<T> reverse(final List<T> l)\\n}\\n\", 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/variant/utils/GeneralUtils.java', 'methodCount': 12, 'lineCount': 16}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 13}\n",
      "{'Allele': 'public interface Allele {\\n    /**\\n * A generic static NO_CALL allele for use\\n */\\nString NO_CALL_STRING = \".\";;\\n    /**\\n * A generic static SPAN_DEL allele for use\\n */\\nString SPAN_DEL_STRING = \"*\";;\\n    /**\\n * Non ref allele representations\\n */\\nchar SINGLE_BREAKEND_INDICATOR = \\'.\\';;\\n    char BREAKEND_EXTENDING_RIGHT = \\'[\\';;\\n    char BREAKEND_EXTENDING_LEFT = \\']\\';;\\n    char SYMBOLIC_ALLELE_START = \\'<\\';;\\n    char SYMBOLIC_ALLELE_END = \\'>\\';;\\n    String NON_REF_STRING = \"<NON_REF>\";;\\n    String UNSPECIFIED_ALTERNATE_ALLELE_STRING = \"<*>\";;\\n    Allele REF_A = new SimpleAllele(\"A\", true);;\\n    Allele ALT_A = new SimpleAllele(\"A\", false);;\\n    Allele REF_C = new SimpleAllele(\"C\", true);;\\n    Allele ALT_C = new SimpleAllele(\"C\", false);;\\n    Allele REF_G = new SimpleAllele(\"G\", true);;\\n    Allele ALT_G = new SimpleAllele(\"G\", false);;\\n    Allele REF_T = new SimpleAllele(\"T\", true);;\\n    Allele ALT_T = new SimpleAllele(\"T\", false);;\\n    Allele REF_N = new SimpleAllele(\"N\", true);;\\n    Allele ALT_N = new SimpleAllele(\"N\", false);;\\n    Allele SPAN_DEL = new SimpleAllele(SPAN_DEL_STRING, false);;\\n    Allele NO_CALL = new SimpleAllele(NO_CALL_STRING, false);;\\n    Allele NON_REF_ALLELE = new SimpleAllele(NON_REF_STRING, false);;\\n    Allele UNSPECIFIED_ALTERNATE_ALLELE = new SimpleAllele(UNSPECIFIED_ALTERNATE_ALLELE_STRING, false);;\\n    // for simple deletion, e.g. \"ALT==<DEL>\" (note that the spec allows, for now at least, alt alleles like <DEL:ME>)\\n@SuppressWarnings(\"unused\")\\nAllele SV_SIMPLE_DEL = StructuralVariantType.DEL.toSymbolicAltAllele();;\\n    // for simple insertion, e.g. \"ALT==<INS>\"\\n@SuppressWarnings(\"unused\")\\nAllele SV_SIMPLE_INS = StructuralVariantType.INS.toSymbolicAltAllele();;\\n    // for simple inversion, e.g. \"ALT==<INV>\"\\n@SuppressWarnings(\"unused\")\\nAllele SV_SIMPLE_INV = StructuralVariantType.INV.toSymbolicAltAllele();;\\n    // for simple generic cnv, e.g. \"ALT==<CNV>\"\\n@SuppressWarnings(\"unused\")\\nAllele SV_SIMPLE_CNV = StructuralVariantType.CNV.toSymbolicAltAllele();;\\n    // for simple duplication, e.g. \"ALT==<DUP>\"\\n@SuppressWarnings(\"unused\")\\nAllele SV_SIMPLE_DUP = StructuralVariantType.DUP.toSymbolicAltAllele();;\\n     static Allele create(byte[] bases, boolean isRef)\\n     static Allele create(byte base, boolean isRef)\\n     static Allele create(byte base)\\n     static Allele extend(Allele left, byte[] right)\\n     static boolean wouldBeNullAllele(byte[] bases)\\n     static boolean wouldBeStarAllele(byte[] bases)\\n     static boolean wouldBeNoCallAllele(byte[] bases)\\n     static boolean wouldBeSymbolicAllele(byte[] bases)\\n     static boolean wouldBeBreakpoint(byte[] bases)\\n     static boolean wouldBeSingleBreakend(byte[] bases)\\n     static boolean acceptableAlleleBases(String bases)\\n     static boolean acceptableAlleleBases(String bases, boolean isReferenceAllele)\\n     static boolean acceptableAlleleBases(byte[] bases)\\n     static boolean acceptableAlleleBases(byte[] bases, boolean isReferenceAllele)\\n     static Allele create(String bases, boolean isRef)\\n     static Allele create(String bases)\\n     static Allele create(byte[] bases)\\n     static Allele create(Allele allele, boolean ignoreRefState)\\n     static boolean oneIsPrefixOfOther(final Allele a1, final Allele a2)\\n     boolean isPrefixOf(final Allele other)\\n     boolean isNoCall()\\n     boolean isCalled()\\n     boolean isReference()\\n     boolean isNonReference()\\n     boolean isSymbolic()\\n     boolean isBreakpoint()\\n     boolean isSingleBreakend()\\n     String toString()\\n     byte[] getBases()\\n     String getBaseString()\\n     String getDisplayString()\\n     byte[] getDisplayBases()\\n     boolean equals(Object other)\\n     int hashCode()\\n     boolean equals(Allele other, boolean ignoreRefState)\\n     boolean basesMatch(byte[] test)\\n     boolean basesMatch(String test)\\n     boolean basesMatch(Allele test)\\n     int length()\\n     boolean isNonRefAllele()\\n}\\n'}\n",
      "/**\n",
      "     * Run all extra-strict validation tests on a Variant Context object\n",
      "     *\n",
      "     * @param reportedReference   the reported reference allele\n",
      "     * @param observedReference     the observed reference allele\n",
      "     * @param rsIDs               the true dbSNP IDs\n",
      "     */\n",
      "public void extraStrictValidation(final Allele reportedReference, final Allele observedReference, final Set<String> rsIDs)\n",
      "**************************************\n",
      "public void extraStrictValidation(final Allele reportedReference, final Allele observedReference, final Set<String> rsIDs) {\n",
      "    validateReferenceBases(reportedReference, observedReference);\n",
      "    validateRSIDs(rsIDs);\n",
      "    validateAlternateAlleles();\n",
      "    validateChromosomeCounts();\n",
      "}\n",
      "2406\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 964/1243 [32:45<09:51,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43112\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 965/1243 [32:47<09:52,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43113\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 966/1243 [32:49<10:03,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43163\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 967/1243 [32:51<09:16,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43201\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 968/1243 [32:53<09:12,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43208\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 24, 'classCount': 1, 'methodsOnly': 'public class FileExtensions {\\n    /**\\n * extensions for read files and related formats.\\n */\\npublic static final Set<String> FASTA = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(\".fasta\", \".fasta.gz\", \".fas\", \".fas.gz\", \".fa\", \".fa.gz\", \".fna\", \".fna.gz\", \".txt\", \".txt.gz\")));;\\n    public static final String FASTA_INDEX = \".fai\";;\\n    /**\\n * extensions for alignment files SAM, BAM, CRAM.\\n */\\npublic static final String SAM = \".sam\";;\\n    public static final String BAM = \".bam\";;\\n    public static final String BAI_INDEX = \".bai\";;\\n    public static final String CRAM = \".cram\";;\\n    public static final String CRAM_INDEX = \".crai\";;\\n    public static final String BED = \".bed\";;\\n    public static final String TABIX_INDEX = \".tbi\";;\\n    public static final String TRIBBLE_INDEX = \".idx\";;\\n    /**\\n * extensions for VCF files and related formats.\\n */\\npublic static final String VCF = \".vcf\";;\\n    public static final String VCF_INDEX = TRIBBLE_INDEX;;\\n    public static final String BCF = \".bcf\";;\\n    public static final String COMPRESSED_VCF = \".vcf.gz\";;\\n    public static final String COMPRESSED_VCF_INDEX = \".tbi\";;\\n    public static final List<String> VCF_LIST = Collections.unmodifiableList(Arrays.asList(VCF, COMPRESSED_VCF, BCF));;\\n    public static final String INTERVAL_LIST = \".interval_list\";;\\n    public static final String COMPRESSED_INTERVAL_LIST = \".interval_list.gz\";;\\n    public static final String DICT = \".dict\";;\\n    public static final String GZI = \".gzi\";;\\n    public static final String SBI = \".sbi\";;\\n    public static final String CSI = \".csi\";;\\n    public static final Set<String> BLOCK_COMPRESSED = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(\".gz\", \".gzip\", \".bgz\", \".bgzf\")));;\\n    public static final Set<String> GFF3 = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(\".gff3\", \".gff\", \".gff3.gz\", \".gff.gz\")));;\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/FileExtensions.java', 'methodCount': 0, 'lineCount': 26}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class TribbleException {\\n    // what file or input source we are working from\\nString source;;\\n    public TribbleException(String msg);\\n    public TribbleException(String message, Throwable throwable);\\n    public void setSource(String source)\\n    public String getMessage()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/tribble/TribbleException.java', 'methodCount': 4, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 13}\n",
      "{}\n",
      "/**\n",
      "     * Collapse multiple strings into a comma separated list\n",
      "     *\n",
      "     * [\"s1\", \"s2\", \"s3\"] =&gt; \",s1,s2,s3\"\n",
      "     *\n",
      "     * @param strings size &gt; 1 list of strings\n",
      "     * @return\n",
      "     */\n",
      "public static String collapseStringList(final List<String> strings)\n",
      "**************************************\n",
      "public static String collapseStringList(final List<String> strings) {\n",
      "    if (strings.isEmpty())\n",
      "        return \"\";\n",
      "    else if (strings.size() == 1)\n",
      "        return strings.get(0);\n",
      "    else {\n",
      "        final StringBuilder b = new StringBuilder();\n",
      "        for (final String s : strings) {\n",
      "            if (s != null) {\n",
      "                assert s.indexOf(\",\") == -1;\n",
      "                b.append(',').append(s);\n",
      "            }\n",
      "        }\n",
      "        return b.toString();\n",
      "    }\n",
      "}\n",
      "912\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 969/1243 [32:55<08:49,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43249\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 970/1243 [32:57<09:03,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43264\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 971/1243 [32:59<09:31,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43274\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 972/1243 [33:02<09:45,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43283\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 973/1243 [33:04<09:24,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43338\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class SAMSequenceDictionaryCodec {\\n    private static final SAMFileHeader EMPTY_HEADER = new SAMFileHeader();;\\n    private final SAMTextHeaderCodec codec;;\\n    public SAMSequenceDictionaryCodec(final Writer writer);\\n    public void encodeSequenceRecord(final SAMSequenceRecord sequenceRecord)\\n    public void encodeHeaderLine(final boolean keepExistingVersionNumber)\\n    public SAMSequenceDictionary decode(final LineReader reader, final String source)\\n    public void encode(final SAMSequenceDictionary dictionary)\\n    public void setValidationStringency(final ValidationStringency validationStringency)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMSequenceDictionaryCodec.java', 'methodCount': 6, 'lineCount': 10}, {'fieldCount': 21, 'classCount': 1, 'methodsOnly': 'public class SAMSequenceRecord {\\n    // AbstractSAMHeaderRecord implements Serializable\\npublic static final long serialVersionUID = 1L;;\\n    public static final int UNAVAILABLE_SEQUENCE_INDEX = -1;;\\n    // Value must be interned() if it\\'s ever set/modified\\nprivate final String mSequenceName;;\\n    private Set<String> mAlternativeSequenceName = new LinkedHashSet<>();;\\n    private int mSequenceIndex = UNAVAILABLE_SEQUENCE_INDEX;;\\n    private int mSequenceLength = 0;;\\n    public static final String SEQUENCE_NAME_TAG = \"SN\";;\\n    public static final String ALTERNATIVE_SEQUENCE_NAME_TAG = \"AN\";;\\n    public static final String SEQUENCE_LENGTH_TAG = \"LN\";;\\n    public static final String MD5_TAG = \"M5\";;\\n    public static final String ASSEMBLY_TAG = \"AS\";;\\n    public static final String URI_TAG = \"UR\";;\\n    public static final String SPECIES_TAG = \"SP\";;\\n    public static final String DESCRIPTION_TAG = \"DS\";;\\n    /**\\n * If one sequence has this length, and another sequence had a different length, isSameSequence will\\n * not complain that they are different sequences.\\n */\\npublic static final int UNKNOWN_SEQUENCE_LENGTH = 0;;\\n    /**\\n * This is not a valid sequence name, because it is reserved in the RNEXT field of SAM text format\\n * to mean \"same reference as RNAME field.\"\\n */\\npublic static final String RESERVED_RNEXT_SEQUENCE_NAME = \"=\";;\\n    /* use RESERVED_RNEXT_SEQUENCE_NAME instead. */\\n@Deprecated\\npublic static final String RESERVED_MRNM_SEQUENCE_NAME = RESERVED_RNEXT_SEQUENCE_NAME;;\\n    /**\\n * The standard tags are stored in text header without type information, because the type of these tags is known.\\n */\\npublic static final Set<String> STANDARD_TAGS = new HashSet<>(Arrays.asList(SEQUENCE_NAME_TAG, SEQUENCE_LENGTH_TAG, ASSEMBLY_TAG, ALTERNATIVE_SEQUENCE_NAME_TAG, MD5_TAG, URI_TAG, SPECIES_TAG));;\\n    // These are the chars matched by \\\\\\\\s.\\n// \\\\013 is vertical tab\\nprivate static final char[] WHITESPACE_CHARS = { \\' \\', \\'\\\\t\\', \\'\\\\n\\', \\'\\\\013\\', \\'\\\\f\\', \\'\\\\r\\' };;\\n    // alternative sequence name separator\\nprivate static final String ALTERNATIVE_SEQUENCE_NAME_SEPARATOR = \",\";;\\n    private static final Pattern LEGAL_RNAME_PATTERN = Pattern.compile(\"[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*\");;\\n    public SAMSequenceRecord(final String name);\\n    public SAMSequenceRecord(final String name, final int sequenceLength);\\n    public String getSequenceName()\\n    public int getSequenceLength()\\n    public SAMSequenceRecord setSequenceLength(final int value)\\n    public String getAssembly()\\n    public SAMSequenceRecord setAssembly(final String value)\\n    public String getSpecies()\\n    public SAMSequenceRecord setSpecies(final String value)\\n    public String getMd5()\\n    public SAMSequenceRecord setMd5(final String value)\\n    public String getDescription()\\n    public SAMSequenceRecord setDescription(final String value)\\n    public int getSequenceIndex()\\n    public SAMSequenceRecord setSequenceIndex(final int value)\\n    public Set<String> getAlternativeSequenceNames()\\n    public void addAlternativeSequenceName(final String name)\\n    public SAMSequenceRecord setAlternativeSequenceName(final Collection<String> alternativeSequences)\\n    private static void validateAltRegExp(final String name)\\n    private void encodeAltSequences(final Collection<String> alternativeSequences)\\n    public boolean hasAlternativeSequenceNames()\\n    public boolean isSameSequence(final SAMSequenceRecord that)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n     Set<String> getStandardTags()\\n    public final SAMSequenceRecord clone()\\n    public static String truncateSequenceName(final String sequenceName)\\n    public static void validateSequenceName(final String name)\\n    public String toString()\\n    public String getSAMString()\\n    public final String getContig()\\n    public final int getStart()\\n    public final int getEnd()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMSequenceRecord.java', 'methodCount': 33, 'lineCount': 56}, {'fieldCount': 16, 'classCount': 1, 'methodsOnly': 'public class SequenceUtil {\\n    /**\\n * Byte typed variables for all normal bases.\\n */\\npublic static final byte a = \\'a\\', c = \\'c\\', g = \\'g\\', t = \\'t\\', n = \\'n\\', A = \\'A\\', C = \\'C\\', G = \\'G\\', T = \\'T\\', N = \\'N\\';;\\n    public static final byte[] VALID_BASES_UPPER = new byte[] { A, C, G, T };;\\n    public static final byte[] VALID_BASES_LOWER = new byte[] { a, c, g, t };;\\n    private static final byte[] ACGTN_BASES = new byte[] { A, C, G, T, N };;\\n    private static final String IUPAC_CODES_STRING = \".aAbBcCdDgGhHkKmMnNrRsStTvVwWyY\";;\\n    /**\\n * A set of bases supported by BAM in reads, see http://samtools.github.io/hts-specs/SAMv1.pdf chapter 4.2 on \\'seq\\' field.\\n * Effectively these are upper cased IUPAC codes with equals sign (\\'=\\') and without dot (\\'.\\').\\n */\\nprivate static final byte[] BAM_READ_BASE_SET = \"=ABCDGHKMNRSTVWY\".getBytes();;\\n    private static final int BASES_ARRAY_LENGTH = 127;;\\n    private static final int SHIFT_TO_LOWER_CASE = a - A;;\\n    /**\\n * A lookup table to find a corresponding BAM read base.\\n */\\nprivate static final byte[] bamReadBaseLookup = new byte[BASES_ARRAY_LENGTH];;\\n    private static final byte A_MASK = 1;;\\n    private static final byte C_MASK = 2;;\\n    private static final byte G_MASK = 4;;\\n    private static final byte T_MASK = 8;;\\n    private static final byte[] bases = new byte[BASES_ARRAY_LENGTH];;\\n    private static final byte NON_IUPAC_CODE = 0;;\\n    public static String reverseComplement(final String sequenceData)\\n    public static boolean basesEqual(final byte lhs, final byte rhs)\\n    public static boolean readBaseMatchesRefBaseWithAmbiguity(final byte readBase, final byte refBase)\\n    public static boolean isNoCall(final byte base)\\n    public static boolean isValidBase(final byte b)\\n    private static boolean isValidBase(final byte b, final byte[] validBases)\\n    public static boolean isUpperACGTN(final byte base)\\n    public static String getIUPACCodesString()\\n    public static boolean isIUPAC(final byte base)\\n    public static double calculateGc(final byte[] bases)\\n    public static boolean isBamReadBase(final byte base)\\n    public static byte[] toBamReadBasesInPlace(final byte[] bases)\\n    public static void assertSequenceListsEqual(final List<SAMSequenceRecord> s1, final List<SAMSequenceRecord> s2)\\n    public static void assertSequenceListsEqual(final List<SAMSequenceRecord> s1, final List<SAMSequenceRecord> s2, final boolean checkPrefixOnly)\\n    public static boolean areSequenceDictionariesEqual(final SAMSequenceDictionary s1, final SAMSequenceDictionary s2)\\n    public static void assertSequenceDictionariesEqual(final SAMSequenceDictionary s1, final SAMSequenceDictionary s2)\\n    public static void assertSequenceDictionariesEqual(final SAMSequenceDictionary s1, final SAMSequenceDictionary s2, final boolean checkPrefixOnly)\\n    public static void assertSequenceDictionariesEqual(final SAMSequenceDictionary s1, final SAMSequenceDictionary s2, final File f1, final File f2)\\n    public static String makeCigarStringWithPossibleClipping(final int alignmentStart, final int readLength, final int referenceSequenceLength)\\n    public static String makeCigarStringWithIndelPossibleClipping(final int alignmentStart, final int readLength, final int referenceSequenceLength, final int indelPosition, final int indelLength)\\n    public static String makeSoftClipCigar(final int clipLength)\\n    private static boolean basesMatch(final byte readBase, final byte refBase, final boolean negativeStrand, final boolean bisulfiteSequence, final boolean matchAmbiguousRef)\\n    public static int countMismatches(final SAMRecord read, final byte[] referenceBases)\\n    public static int countMismatches(final SAMRecord read, final byte[] referenceBases, final int referenceOffset)\\n    public static int countMismatches(final SAMRecord read, final byte[] referenceBases, final int referenceOffset, final boolean bisulfiteSequence)\\n    public static int countMismatches(final SAMRecord read, final byte[] referenceBases, final int referenceOffset, final boolean bisulfiteSequence, final boolean matchAmbiguousRef)\\n    public static int countMismatches(final SAMRecord read, final byte[] referenceBases, final boolean bisulfiteSequence)\\n    public static int sumQualitiesOfMismatches(final SAMRecord read, final byte[] referenceBases)\\n    public static int sumQualitiesOfMismatches(final SAMRecord read, final byte[] referenceBases, final int referenceOffset)\\n    public static int sumQualitiesOfMismatches(final SAMRecord read, final byte[] referenceBases, final int referenceOffset, final boolean bisulfiteSequence)\\n    public static int countInsertedBases(final Cigar cigar)\\n    public static int countDeletedBases(final Cigar cigar)\\n    public static int countInsertedBases(final SAMRecord read)\\n    public static int countDeletedBases(final SAMRecord read)\\n    public static int calculateSamNmTag(final SAMRecord read, final byte[] referenceBases)\\n    public static int calculateSamNmTag(final SAMRecord read, final byte[] referenceBases, final int referenceOffset)\\n    public static int calculateSamNmTag(final SAMRecord read, final byte[] referenceBases, final int referenceOffset, final boolean bisulfiteSequence)\\n    public static int calculateSamNmTagFromCigar(final SAMRecord record)\\n    public static byte complement(final byte b)\\n    public static boolean bisulfiteBasesEqual(final boolean negativeStrand, final byte read, final byte reference)\\n    public static boolean bisulfiteBasesEqual(final byte read, final byte reference)\\n    public static boolean bisulfiteBasesMatchWithAmbiguity(final boolean negativeStrand, final byte read, final byte reference)\\n    public static boolean isBisulfiteConverted(final byte read, final byte reference, final boolean negativeStrand)\\n    public static boolean isBisulfiteConverted(final byte read, final byte reference)\\n    static final Pattern mdPat = Pattern.compile(\"\\\\\\\\G(?:([0-9]+)|([ACTGNactgn])|(\\\\\\\\^[ACTGNactgn]+))\");;\\n    public static byte[] makeReferenceFromAlignment(final SAMRecord rec, final boolean includeReferenceBasesForDeletions)\\n    public static void reverseComplement(final byte[] bases)\\n    public static void reverseQualities(final byte[] quals)\\n    public static void reverse(final byte[] array, final int offset, final int len)\\n    public static void reverseComplement(final byte[] bases, final int offset, final int len)\\n    public static String calculateMD5String(final byte[] data)\\n    public static String calculateMD5String(final byte[] data, final int offset, final int len)\\n    public static String md5DigestToString(final byte[] digest)\\n    public static byte[] calculateMD5(final byte[] data, final int offset, final int len)\\n    public static void calculateMdAndNmTags(final SAMRecord record, final byte[] ref, final boolean calcMD, final boolean calcNM)\\n    public static byte upperCase(final byte base)\\n    public static byte[] upperCase(final byte[] bases)\\n    public static List<byte[]> generateAllKmers(final int length)\\n    public static String getSamReadNameFromFastqHeader(final String fastqHeader)\\n    public static byte[] getRandomBases(Random random, final int length)\\n    public static void getRandomBases(Random random, final int length, final byte[] bases)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/SequenceUtil.java', 'methodCount': 60, 'lineCount': 78}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ValidationUtils {\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String nameOfObject)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(final T collection, final String nameOfObject)\\n    public static boolean isNonEmpty(final Collection<?> collection)\\n    public static String nonEmpty(final String string, final String nameOfObject)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(final T collection)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/utils/ValidationUtils.java', 'methodCount': 10, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 11}\n",
      "{}\n",
      "/**\n",
      "     * Closes this writer flushing all remaining writing operation input the output resources.\n",
      "     * <p>\n",
      "     * Further calls to {@link #appendBases} or {@link #startSequence} will result in an exception.\n",
      "     * </p>\n",
      "     *\n",
      "     * @throws IOException if such exception is thrown when closing output writers and output streams.\n",
      "     * @throws IllegalStateException if closing without writing any sequences or closing when writing a sequence is in progress\n",
      "     */\n",
      "public void close() throws IOException\n",
      "**************************************\n",
      "@Override\n",
      "public void close() throws IOException {\n",
      "    if (!closed) {\n",
      "        try {\n",
      "            closeSequence();\n",
      "            if (sequenceNames.isEmpty()) {\n",
      "                throw new IllegalStateException(\"no sequences were added to the reference\");\n",
      "            }\n",
      "        } finally {\n",
      "            closed = true;\n",
      "            fastaStream.close();\n",
      "            faiIndexWriter.close();\n",
      "            dictWriter.close();\n",
      "        }\n",
      "    }\n",
      "}\n",
      "3450\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 974/1243 [33:06<10:21,  2.31s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43615\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  78%|  | 975/1243 [33:09<10:24,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43630\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 976/1243 [33:11<10:15,  2.31s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43640\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 977/1243 [33:13<09:21,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43677\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 978/1243 [33:15<09:26,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43713\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface HtsHeader {\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/beta/plugin/HtsHeader.java', 'methodCount': 0, 'lineCount': 2}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class Log {\\n    private static LogLevel globalLogLevel = LogLevel.INFO;;\\n    private static PrintStream out = System.err;;\\n    private final Class<?> clazz;;\\n    private final String className;;\\n    private Log(final Class<?> clazz);\\n    public static Log getInstance(final Class<?> clazz)\\n    public static void setGlobalLogLevel(final LogLevel logLevel)\\n    public static LogLevel getGlobalLogLevel()\\n    public static void setGlobalPrintStream(final PrintStream stream)\\n    public static PrintStream getGlobalPrintStream()\\n    public static final boolean isEnabled(final LogLevel level)\\n    private void emit(final LogLevel level, final Throwable throwable, final Object... parts)\\n    protected String getTimestamp()\\n    public final void error(final Throwable throwable, final Object... messageParts)\\n    public final void warn(final Throwable throwable, final Object... messageParts)\\n    public final void info(final Throwable throwable, final Object... messageParts)\\n    public final void debug(final Throwable throwable, final Object... messageParts)\\n    public final void error(final Object... messageParts)\\n    public final void warn(final Object... messageParts)\\n    public final void info(final Object... messageParts)\\n    public final void debug(final Object... messageParts)\\n    private final void getAndEmitIfEnabled(LogLevel logLevel, final Supplier<Object> messageParts)\\n    public final void error(final Supplier<Object> messageParts)\\n    public final void warn(final Supplier<Object> messageParts)\\n    public final void info(final Supplier<Object> messageParts)\\n    public final void debug(final Supplier<Object> messageParts)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/Log.java', 'methodCount': 22, 'lineCount': 28}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'SAMSequenceRecord': 'public class SAMSequenceRecord {\\n    // AbstractSAMHeaderRecord implements Serializable\\npublic static final long serialVersionUID = 1L;;\\n    public static final int UNAVAILABLE_SEQUENCE_INDEX = -1;;\\n    // Value must be interned() if it\\'s ever set/modified\\nprivate final String mSequenceName;;\\n    private Set<String> mAlternativeSequenceName = new LinkedHashSet<>();;\\n    private int mSequenceIndex = UNAVAILABLE_SEQUENCE_INDEX;;\\n    private int mSequenceLength = 0;;\\n    public static final String SEQUENCE_NAME_TAG = \"SN\";;\\n    public static final String ALTERNATIVE_SEQUENCE_NAME_TAG = \"AN\";;\\n    public static final String SEQUENCE_LENGTH_TAG = \"LN\";;\\n    public static final String MD5_TAG = \"M5\";;\\n    public static final String ASSEMBLY_TAG = \"AS\";;\\n    public static final String URI_TAG = \"UR\";;\\n    public static final String SPECIES_TAG = \"SP\";;\\n    public static final String DESCRIPTION_TAG = \"DS\";;\\n    /**\\n * If one sequence has this length, and another sequence had a different length, isSameSequence will\\n * not complain that they are different sequences.\\n */\\npublic static final int UNKNOWN_SEQUENCE_LENGTH = 0;;\\n    /**\\n * This is not a valid sequence name, because it is reserved in the RNEXT field of SAM text format\\n * to mean \"same reference as RNAME field.\"\\n */\\npublic static final String RESERVED_RNEXT_SEQUENCE_NAME = \"=\";;\\n    /* use RESERVED_RNEXT_SEQUENCE_NAME instead. */\\n@Deprecated\\npublic static final String RESERVED_MRNM_SEQUENCE_NAME = RESERVED_RNEXT_SEQUENCE_NAME;;\\n    /**\\n * The standard tags are stored in text header without type information, because the type of these tags is known.\\n */\\npublic static final Set<String> STANDARD_TAGS = new HashSet<>(Arrays.asList(SEQUENCE_NAME_TAG, SEQUENCE_LENGTH_TAG, ASSEMBLY_TAG, ALTERNATIVE_SEQUENCE_NAME_TAG, MD5_TAG, URI_TAG, SPECIES_TAG));;\\n    // These are the chars matched by \\\\\\\\s.\\n// \\\\013 is vertical tab\\nprivate static final char[] WHITESPACE_CHARS = { \\' \\', \\'\\\\t\\', \\'\\\\n\\', \\'\\\\013\\', \\'\\\\f\\', \\'\\\\r\\' };;\\n    // alternative sequence name separator\\nprivate static final String ALTERNATIVE_SEQUENCE_NAME_SEPARATOR = \",\";;\\n    private static final Pattern LEGAL_RNAME_PATTERN = Pattern.compile(\"[0-9A-Za-z!#$%&+./:;?@^_|~-][0-9A-Za-z!#$%&*+./:;=?@^_|~-]*\");;\\n    public SAMSequenceRecord(final String name);\\n    public SAMSequenceRecord(final String name, final int sequenceLength);\\n    public String getSequenceName()\\n    public int getSequenceLength()\\n    public SAMSequenceRecord setSequenceLength(final int value)\\n    public String getAssembly()\\n    public SAMSequenceRecord setAssembly(final String value)\\n    public String getSpecies()\\n    public SAMSequenceRecord setSpecies(final String value)\\n    public String getMd5()\\n    public SAMSequenceRecord setMd5(final String value)\\n    public String getDescription()\\n    public SAMSequenceRecord setDescription(final String value)\\n    public int getSequenceIndex()\\n    public SAMSequenceRecord setSequenceIndex(final int value)\\n    public Set<String> getAlternativeSequenceNames()\\n    public void addAlternativeSequenceName(final String name)\\n    public SAMSequenceRecord setAlternativeSequenceName(final Collection<String> alternativeSequences)\\n    private static void validateAltRegExp(final String name)\\n    private void encodeAltSequences(final Collection<String> alternativeSequences)\\n    public boolean hasAlternativeSequenceNames()\\n    public boolean isSameSequence(final SAMSequenceRecord that)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n     Set<String> getStandardTags()\\n    public final SAMSequenceRecord clone()\\n    public static String truncateSequenceName(final String sequenceName)\\n    public static void validateSequenceName(final String name)\\n    public String toString()\\n    public String getSAMString()\\n    public final String getContig()\\n    public final int getStart()\\n    public final int getEnd()\\n}\\n', 'SAMSequenceDictionary': 'public class SAMSequenceDictionary {\\n    public static final long serialVersionUID = 1L;;\\n    private List<SAMSequenceRecord> mSequences = new ArrayList<>();;\\n    private final Map<String, SAMSequenceRecord> mSequenceMap = new HashMap<>();;\\n    public SAMSequenceDictionary();\\n    public SAMSequenceDictionary(final List<SAMSequenceRecord> list);\\n    public List<SAMSequenceRecord> getSequences()\\n    private static Log log = Log.getInstance(SAMSequenceDictionary.class);;\\n    public SAMSequenceRecord getSequence(final String name)\\n    public void setSequences(final List<SAMSequenceRecord> list)\\n    public void addSequence(final SAMSequenceRecord sequenceRecord)\\n    public SAMSequenceRecord getSequence(final int sequenceIndex)\\n    public int getSequenceIndex(final String sequenceName)\\n    public int size()\\n    public long getReferenceLength()\\n    public boolean isEmpty()\\n    private static String DICT_MISMATCH_TEMPLATE = \"SAM dictionaries are not the same: %s.\";;\\n    public void assertSameDictionary(final SAMSequenceDictionary that)\\n    public boolean isSameDictionary(final SAMSequenceDictionary that)\\n    public boolean equals(Object o)\\n    public SAMSequenceRecord addSequenceAlias(final String originalName, final String altName)\\n    public SAMSequenceRecord addAlternativeSequenceName(final String originalName, final String altName)\\n    public String md5()\\n    public int hashCode()\\n    public String toString()\\n    public static final List<String> DEFAULT_DICTIONARY_EQUAL_TAG = Arrays.asList(SAMSequenceRecord.MD5_TAG, SAMSequenceRecord.SEQUENCE_LENGTH_TAG);;\\n    public static SAMSequenceDictionary mergeDictionaries(final SAMSequenceDictionary dict1, final SAMSequenceDictionary dict2, final List<String> tagsToMatch)\\n}\\n'}\n",
      "/**\n",
      "     * Non-comprehensive {@link #equals(Object)}-validation: instead of calling {@link SAMSequenceRecord#equals(Object)} on constituent\n",
      "     * {@link SAMSequenceRecord}s in this dictionary against its pair in the target dictionary, in order, call\n",
      "     * {@link SAMSequenceRecord#isSameSequence(SAMSequenceRecord)}.\n",
      "     *\n",
      "     * @param that {@link SAMSequenceDictionary} to compare against\n",
      "     * @return true if the dictionaries are the same, false otherwise\n",
      "     *\n",
      "     */\n",
      "public boolean isSameDictionary(final SAMSequenceDictionary that)\n",
      "**************************************\n",
      "public boolean isSameDictionary(final SAMSequenceDictionary that) {\n",
      "    if (that == null || that.mSequences == null)\n",
      "        return false;\n",
      "    if (this == that)\n",
      "        return true;\n",
      "    final Iterator<SAMSequenceRecord> thatSequences = that.mSequences.iterator();\n",
      "    for (final SAMSequenceRecord thisSequence : mSequences) {\n",
      "        if (!thatSequences.hasNext()) {\n",
      "            return false;\n",
      "        } else {\n",
      "            final SAMSequenceRecord thatSequence = thatSequences.next();\n",
      "            if (!thatSequence.isSameSequence(thisSequence)) {\n",
      "                return false;\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return !thatSequences.hasNext();\n",
      "}\n",
      "2084\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 979/1243 [33:17<09:19,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43715\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 980/1243 [33:19<09:30,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43725\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 981/1243 [33:21<09:25,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43767\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 982/1243 [33:24<09:24,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43771\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 983/1243 [33:25<08:43,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43774\n",
      "{'classCount': 3, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class FileTruncatedException {\\n    public FileTruncatedException();\\n    public FileTruncatedException(final String s);\\n    public FileTruncatedException(final String s, final Throwable throwable);\\n    public FileTruncatedException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/FileTruncatedException.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class SAMException {\\n    public SAMException();\\n    public SAMException(final String s);\\n    public SAMException(final String s, final Throwable throwable);\\n    public SAMException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMException.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class SeekableBufferedStream {\\n    public static final int DEFAULT_BUFFER_SIZE = 512000;;\\n    final private int bufferSize;;\\n    final SeekableStream wrappedStream;;\\n    ExtBufferedInputStream bufferedStream;;\\n    long position;;\\n    public SeekableBufferedStream(final SeekableStream stream, final int bufferSize);\\n    public SeekableBufferedStream(final SeekableStream stream);\\n    public long length()\\n    public long skip(final long skipLength) throws IOException\\n    public void seek(final long position) throws IOException\\n    private void seekInternal(final long position) throws IOException\\n    public int read() throws IOException\\n    public int read(final byte[] buffer, final int offset, final int length) throws IOException\\n    public void close() throws IOException\\n    public boolean eof() throws IOException\\n    public String getSource()\\n    public long position() throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableBufferedStream.java', 'methodCount': 12, 'lineCount': 19}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SeekableFileStream {\\n    /**\\n * Collection of all open instances.  SeekableFileStream objects are usually open and kept open for the\\n * duration of a session.  This collection supports a method to close them all.\\n */\\nstatic Collection<SeekableFileStream> allInstances = Collections.synchronizedCollection(new HashSet<SeekableFileStream>());;\\n    File file;;\\n    RandomAccessFile fis;;\\n    public SeekableFileStream(final File file) throws FileNotFoundException;\\n    public long length()\\n    public boolean eof() throws IOException\\n    public void seek(final long position) throws IOException\\n    public long position() throws IOException\\n    public long skip(long n) throws IOException\\n    public int read(final byte[] buffer, final int offset, final int length) throws IOException\\n    public int read() throws IOException\\n    public int read(byte[] b) throws IOException\\n    public String getSource()\\n    public void close() throws IOException\\n    public static synchronized void closeAllInstances()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableFileStream.java', 'methodCount': 12, 'lineCount': 17}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class SeekableHTTPStream {\\n    private long position = 0;;\\n    private long contentLength = -1;;\\n    private final URL url;;\\n    private final Proxy proxy;;\\n    public SeekableHTTPStream(final URL url);\\n    public SeekableHTTPStream(final URL url, Proxy proxy);\\n    public long position()\\n    public long length()\\n    public long skip(long n) throws IOException\\n    public boolean eof() throws IOException\\n    public void seek(final long position)\\n    public int read(byte[] buffer, int offset, int len) throws IOException\\n    public void close() throws IOException\\n    public int read() throws IOException\\n    public String getSource()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableHTTPStream.java', 'methodCount': 11, 'lineCount': 17}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class SeekableStream {\\n    /**\\n * If the stream is marked with {@link #mark(int)} this represents the {@link #position()}\\n * where the stream was; otherwise, this is empty.\\n */\\nprotected OptionalLong mark = OptionalLong.empty();;\\n    public abstract long length()\\n    public abstract long position() throws IOException\\n    public abstract void seek(long position) throws IOException\\n    public abstract int read(byte[] buffer, int offset, int length) throws IOException\\n    public abstract void close() throws IOException\\n    public abstract boolean eof() throws IOException\\n    public abstract String getSource()\\n    public void readFully(byte[] b) throws IOException\\n    public int available() throws IOException\\n    public final synchronized void mark(int readlimit)\\n    public synchronized void reset() throws IOException\\n    public final boolean markSupported()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableStream.java', 'methodCount': 12, 'lineCount': 15}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class InflaterFactory {\\n    public InflaterFactory();\\n    public Inflater makeInflater(final boolean gzipCompatible)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/zip/InflaterFactory.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 8}\n",
      "{}\n",
      "/**\n",
      "     * @param stream Must be at start of file.  Throws RuntimeException if !stream.markSupported().\n",
      "     * @return true if the given file looks like a valid BGZF file.\n",
      "     */\n",
      "public static boolean isValidFile(final InputStream stream) throws IOException\n",
      "**************************************\n",
      "public static boolean isValidFile(final InputStream stream) throws IOException {\n",
      "    if (!stream.markSupported()) {\n",
      "        throw new RuntimeException(\"Cannot test non-buffered stream\");\n",
      "    }\n",
      "    stream.mark(BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH);\n",
      "    final byte[] buffer = new byte[BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH];\n",
      "    final int count = readBytes(stream, buffer, 0, BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH);\n",
      "    stream.reset();\n",
      "    return count == BlockCompressedStreamConstants.BLOCK_HEADER_LENGTH && isValidBlockHeader(buffer);\n",
      "}\n",
      "1536\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 984/1243 [33:27<08:16,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43778\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 985/1243 [33:29<08:59,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43791\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 986/1243 [33:32<09:01,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43828\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 987/1243 [33:34<09:00,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43835\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  79%|  | 988/1243 [33:36<09:01,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43836\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 20, 'classCount': 1, 'methodsOnly': 'public class Defaults {\\n    private static final Log log = Log.getInstance(Defaults.class);;\\n    /**\\n * Should BAM index files be created when writing out coordinate sorted BAM files?  Default = false.\\n */\\npublic static final boolean CREATE_INDEX;;\\n    /**\\n * Should MD5 files be created when writing out SAM and BAM files?  Default = false.\\n */\\npublic static final boolean CREATE_MD5;;\\n    /**\\n * Should asynchronous read I/O be used where supported by the samtools package (one thread per file).\\n *  Default = false.\\n */\\npublic static final boolean USE_ASYNC_IO_READ_FOR_SAMTOOLS;;\\n    /**\\n * Should asynchronous write I/O be used where supported by the samtools package (one thread per file).\\n *  Default = false.\\n */\\npublic static final boolean USE_ASYNC_IO_WRITE_FOR_SAMTOOLS;;\\n    /**\\n * Should asynchronous write I/O be used where supported by the tribble package (one thread per file).\\n *  Default = false.\\n */\\npublic static final boolean USE_ASYNC_IO_WRITE_FOR_TRIBBLE;;\\n    /**\\n * Compression level to be used for writing BAM and other block-compressed outputs.  Default = 5.\\n */\\npublic static final int COMPRESSION_LEVEL;;\\n    /**\\n * Buffer size, in bytes, used whenever reading/writing files or streams.  Default = 128k.\\n */\\npublic static final int BUFFER_SIZE;;\\n    /**\\n * The output format of the flag field when writing SAM text.  Ignored for reading SAM text.  Default = DECIMAL\\n */\\npublic static final SamFlagField SAM_FLAG_FIELD_FORMAT;;\\n    /**\\n * The extension to assume a sam file has when the actual file doesn\\'t have an extension, useful\\n * for when outputing to /dev/stdout, for example.\\n */\\npublic static final String DEFAULT_SAM_EXTENSION;;\\n    /**\\n * The extension to assume a vcf has when the actual file doesn\\'t have an extension, useful\\n * for when outputing to /dev/stdout, for example.\\n */\\npublic static final String DEFAULT_VCF_EXTENSION;;\\n    /**\\n * Even if BUFFER_SIZE is 0, this is guaranteed to be non-zero.  If BUFFER_SIZE is non-zero,\\n * this == BUFFER_SIZE (Default = 128k).\\n */\\npublic static final int NON_ZERO_BUFFER_SIZE;;\\n    /**\\n * The reference FASTA file.  If this is not set, the file is null.  This file may be required for reading\\n * writing SAM files (ex. CRAM).  Default = null.\\n */\\npublic static final File REFERENCE_FASTA;;\\n    /**\\n * Custom reader factory able to handle URL based resources like ga4gh.\\n *  Expected format: <url prefix>,<fully qualified factory class name>[,<jar file name>]\\n *  E.g. https://www.googleapis.com/genomics/v1beta/reads/,com.google.genomics.ReaderFactory\\n *  OR https://www.googleapis.com/genomics/v1beta/reads/,com.google.genomics.ReaderFactory,/tmp/genomics.jar\\n *  Default = \"\".\\n */\\npublic static final String CUSTOM_READER_FACTORY;;\\n    /**\\n * Boolean describing whether downloading a reference file is allowed (for CRAM files),\\n * in case the reference file is not specified by the user\\n * Enabling this is not necessarily a good idea, since this process often fails.  Default = false.\\n */\\npublic static final boolean USE_CRAM_REF_DOWNLOAD;;\\n    /**\\n * A mask (pattern) to use when building EBI reference service URL for a\\n * given MD5 checksum. Must contain one and only one string placeholder.\\n * Default = \"https://www.ebi.ac.uk/ena/cram/md5/%s\".\\n */\\npublic static final String EBI_REFERENCE_SERVICE_URL_MASK;;\\n    /**\\n * Boolean describing whether downloading of SRA native libraries is allowed,\\n * in case such native libraries are not found locally.  Default = false.\\n */\\npublic static final boolean SRA_LIBRARIES_DOWNLOAD;;\\n    /**\\n * The name of the system property that disables snappy.  Default = \"snappy.disable\".\\n */\\npublic static final String DISABLE_SNAPPY_PROPERTY_NAME = \"snappy.disable\";;\\n    /**\\n * Disable use of the Snappy compressor.  Default = false.\\n */\\npublic static final boolean DISABLE_SNAPPY_COMPRESSOR;;\\n    public static final String SAMJDK_PREFIX = \"samjdk.\";;\\n    public static SortedMap<String, Object> allDefaults()\\n    private static String getStringProperty(final String name, final String def)\\n    private static boolean hasProperty(final String name)\\n    private static boolean getBooleanProperty(final String name, final boolean def)\\n    private static int getIntProperty(final String name, final int def)\\n    private static File getFileProperty(final String name, final String def)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/Defaults.java', 'methodCount': 6, 'lineCount': 28}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class SAMException {\\n    public SAMException();\\n    public SAMException(final String s);\\n    public SAMException(final String s, final Throwable throwable);\\n    public SAMException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMException.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class SeekableBufferedStream {\\n    public static final int DEFAULT_BUFFER_SIZE = 512000;;\\n    final private int bufferSize;;\\n    final SeekableStream wrappedStream;;\\n    ExtBufferedInputStream bufferedStream;;\\n    long position;;\\n    public SeekableBufferedStream(final SeekableStream stream, final int bufferSize);\\n    public SeekableBufferedStream(final SeekableStream stream);\\n    public long length()\\n    public long skip(final long skipLength) throws IOException\\n    public void seek(final long position) throws IOException\\n    private void seekInternal(final long position) throws IOException\\n    public int read() throws IOException\\n    public int read(final byte[] buffer, final int offset, final int length) throws IOException\\n    public void close() throws IOException\\n    public boolean eof() throws IOException\\n    public String getSource()\\n    public long position() throws IOException\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableBufferedStream.java', 'methodCount': 12, 'lineCount': 19}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class SeekableFileStream {\\n    /**\\n * Collection of all open instances.  SeekableFileStream objects are usually open and kept open for the\\n * duration of a session.  This collection supports a method to close them all.\\n */\\nstatic Collection<SeekableFileStream> allInstances = Collections.synchronizedCollection(new HashSet<SeekableFileStream>());;\\n    File file;;\\n    RandomAccessFile fis;;\\n    public SeekableFileStream(final File file) throws FileNotFoundException;\\n    public long length()\\n    public boolean eof() throws IOException\\n    public void seek(final long position) throws IOException\\n    public long position() throws IOException\\n    public long skip(long n) throws IOException\\n    public int read(final byte[] buffer, final int offset, final int length) throws IOException\\n    public int read() throws IOException\\n    public int read(byte[] b) throws IOException\\n    public String getSource()\\n    public void close() throws IOException\\n    public static synchronized void closeAllInstances()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableFileStream.java', 'methodCount': 12, 'lineCount': 17}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class SeekableHTTPStream {\\n    private long position = 0;;\\n    private long contentLength = -1;;\\n    private final URL url;;\\n    private final Proxy proxy;;\\n    public SeekableHTTPStream(final URL url);\\n    public SeekableHTTPStream(final URL url, Proxy proxy);\\n    public long position()\\n    public long length()\\n    public long skip(long n) throws IOException\\n    public boolean eof() throws IOException\\n    public void seek(final long position)\\n    public int read(byte[] buffer, int offset, int len) throws IOException\\n    public void close() throws IOException\\n    public int read() throws IOException\\n    public String getSource()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableHTTPStream.java', 'methodCount': 11, 'lineCount': 17}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class SeekableStream {\\n    /**\\n * If the stream is marked with {@link #mark(int)} this represents the {@link #position()}\\n * where the stream was; otherwise, this is empty.\\n */\\nprotected OptionalLong mark = OptionalLong.empty();;\\n    public abstract long length()\\n    public abstract long position() throws IOException\\n    public abstract void seek(long position) throws IOException\\n    public abstract int read(byte[] buffer, int offset, int length) throws IOException\\n    public abstract void close() throws IOException\\n    public abstract boolean eof() throws IOException\\n    public abstract String getSource()\\n    public void readFully(byte[] b) throws IOException\\n    public int available() throws IOException\\n    public final synchronized void mark(int readlimit)\\n    public synchronized void reset() throws IOException\\n    public final boolean markSupported()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/seekablestream/SeekableStream.java', 'methodCount': 12, 'lineCount': 15}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class DeleteOnExitPathHook {\\n    private static final Log LOG = Log.getInstance(DeleteOnExitPathHook.class);;\\n    private static LinkedHashSet<Path> paths = new LinkedHashSet<>();;\\n    private DeleteOnExitPathHook();\\n    public static synchronized void add(Path path)\\n     static void runHooks()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/nio/DeleteOnExitPathHook.java', 'methodCount': 3, 'lineCount': 7}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 45}\n",
      "{}\n",
      "/**\n",
      "     * Checks that a directory is non-null, extent, readable and a directory\n",
      "     * otherwise a runtime exception is thrown.\n",
      "     *\n",
      "     * @param dir the dir to check for writability\n",
      "     */\n",
      "public static void assertDirectoryIsReadable(final File dir)\n",
      "**************************************\n",
      "public static void assertDirectoryIsReadable(final File dir) {\n",
      "    if (dir == null) {\n",
      "        throw new IllegalArgumentException(\"Cannot check readability of null file.\");\n",
      "    } else if (!dir.exists()) {\n",
      "        throw new SAMException(\"Directory does not exist: \" + dir.getAbsolutePath());\n",
      "    } else if (!dir.isDirectory()) {\n",
      "        throw new SAMException(\"Cannot read from directory because it is not a directory: \" + dir.getAbsolutePath());\n",
      "    } else if (!dir.canRead()) {\n",
      "        throw new SAMException(\"Directory exists but is not readable: \" + dir.getAbsolutePath());\n",
      "    }\n",
      "}\n",
      "2639\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 989/1243 [33:38<08:59,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43840\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 990/1243 [33:40<08:21,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43859\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 991/1243 [33:41<08:09,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43862\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 992/1243 [33:43<07:50,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43863\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 993/1243 [33:45<08:07,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43927\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * merge two Sets, assumes sets of size 1 are immutale\n",
      "     */\n",
      "private static BiFunction<Set<T>, Set<T>, Set<T>> mergeSetsAccountingForSingletons()\n",
      "**************************************\n",
      "private static <T> BiFunction<Set<T>, Set<T>, Set<T>> mergeSetsAccountingForSingletons() {\n",
      "    return (newValue, oldValue) -> {\n",
      "        final Set<T> mutableSet = oldValue.size() == 1 ? new HashSet<>() : oldValue;\n",
      "        mutableSet.addAll(oldValue);\n",
      "        mutableSet.addAll(newValue);\n",
      "        return mutableSet;\n",
      "    };\n",
      "}\n",
      "157\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 994/1243 [33:47<08:15,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43928\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 995/1243 [33:50<08:29,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43929\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 996/1243 [33:52<08:34,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "43948\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 997/1243 [33:54<08:29,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44092\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 998/1243 [33:55<07:50,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44108\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ValidationUtils {\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String nameOfObject)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(final T collection, final String nameOfObject)\\n    public static boolean isNonEmpty(final Collection<?> collection)\\n    public static String nonEmpty(final String string, final String nameOfObject)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(final T collection)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/utils/ValidationUtils.java', 'methodCount': 10, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Find the least interval in the tree.\n",
      "     * @return The earliest interval, or null if the tree is empty.\n",
      "     */\n",
      "public Node<V> min()\n",
      "**************************************\n",
      "public Node<V> min() {\n",
      "    Node<V> result = null;\n",
      "    Node<V> node = mRoot;\n",
      "    while (node != null) {\n",
      "        result = node;\n",
      "        node = node.getLeft();\n",
      "    }\n",
      "    return result;\n",
      "}\n",
      "336\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 999/1243 [33:57<06:54,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44111\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  80%|  | 1000/1243 [33:58<06:55,  1.71s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44162\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1001/1243 [33:59<06:16,  1.56s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1002/1243 [34:01<06:49,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44212\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1003/1243 [34:03<06:58,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44231\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class RuntimeIOException {\\n    public RuntimeIOException();\\n    public RuntimeIOException(final String s);\\n    public RuntimeIOException(final String s, final Throwable throwable);\\n    public RuntimeIOException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/RuntimeIOException.java', 'methodCount': 4, 'lineCount': 6}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Lazily generates an InputStream over this response's data from the concatenation of the InputStreams from\n",
      "     * each of the response's data blocks\n",
      "     *\n",
      "     * @return InputStream over this response's data\n",
      "     */\n",
      "public InputStream getDataStream()\n",
      "**************************************\n",
      "public InputStream getDataStream() {\n",
      "    return new SequenceInputStream(new Enumeration<InputStream>() {\n",
      "\n",
      "        private final Iterator<Block> iterator = HtsgetResponse.this.blocks.iterator();\n",
      "\n",
      "        @Override\n",
      "        public boolean hasMoreElements() {\n",
      "            return this.iterator.hasNext();\n",
      "        }\n",
      "\n",
      "        @Override\n",
      "        public InputStream nextElement() {\n",
      "            return this.iterator.next().getData();\n",
      "        }\n",
      "    });\n",
      "}\n",
      "274\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1004/1243 [34:05<07:18,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44285\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1005/1243 [34:07<07:00,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44311\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1006/1243 [34:09<07:35,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44324\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1007/1243 [34:12<07:56,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44350\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1008/1243 [34:14<08:01,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44388\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class SAMException {\\n    public SAMException();\\n    public SAMException(final String s);\\n    public SAMException(final String s, final Throwable throwable);\\n    public SAMException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMException.java', 'methodCount': 4, 'lineCount': 6}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{}\n",
      "/**\n",
      "     * Wrap the given InputStream in a SnappyInputStream if available.\n",
      "     * @return If Snappy is available, a SnappyInputStream wrapping inputStream.\n",
      "     * If not, and bufferSize > 0, a BufferedInputStream.\n",
      "     * Otherwise inputStream is returned.\n",
      "     */\n",
      "public InputStream wrapTempInputStream(final InputStream inputStream, final int bufferSize)\n",
      "**************************************\n",
      "public InputStream wrapTempInputStream(final InputStream inputStream, final int bufferSize) {\n",
      "    InputStream is = IOUtil.maybeBufferInputStream(inputStream, bufferSize);\n",
      "    if (getSnappyLoader().isSnappyAvailable()) {\n",
      "        try {\n",
      "            return getSnappyLoader().wrapInputStream(is);\n",
      "        } catch (Exception e) {\n",
      "            throw new SAMException(\"Error creating SnappyInputStream\", e);\n",
      "        }\n",
      "    } else {\n",
      "        return is;\n",
      "    }\n",
      "}\n",
      "305\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%|  | 1009/1243 [34:16<08:02,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44389\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%| | 1010/1243 [34:17<07:32,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44428\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%| | 1011/1243 [34:20<07:43,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44462\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%| | 1012/1243 [34:22<08:00,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44484\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  81%| | 1013/1243 [34:24<08:03,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44503\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class DeflaterFactory {\\n    public DeflaterFactory();\\n    public Deflater makeDeflater(final int compressionLevel, final boolean gzipCompatible)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/zip/DeflaterFactory.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Writes len bytes from the specified byte array starting at offset off to this output stream. The general\n",
      "     * contract for write(b, off, len) is that some of the bytes in the array b are written to the output stream in order;\n",
      "     * element b[off] is the first byte written and b[off+len-1] is the last byte written by this operation.\n",
      "     *\n",
      "     * @param bytes the data\n",
      "     * @param startIndex the start offset in the data\n",
      "     * @param numBytes the number of bytes to write\n",
      "     */\n",
      "public void write(final byte[] bytes, int startIndex, int numBytes) throws IOException\n",
      "**************************************\n",
      "@Override\n",
      "public void write(final byte[] bytes, int startIndex, int numBytes) throws IOException {\n",
      "    assert (numUncompressedBytes < uncompressedBuffer.length);\n",
      "    while (numBytes > 0) {\n",
      "        final int bytesToWrite = Math.min(uncompressedBuffer.length - numUncompressedBytes, numBytes);\n",
      "        System.arraycopy(bytes, startIndex, uncompressedBuffer, numUncompressedBytes, bytesToWrite);\n",
      "        numUncompressedBytes += bytesToWrite;\n",
      "        startIndex += bytesToWrite;\n",
      "        numBytes -= bytesToWrite;\n",
      "        assert (numBytes >= 0);\n",
      "        if (numUncompressedBytes == uncompressedBuffer.length) {\n",
      "            deflateBlock();\n",
      "        }\n",
      "    }\n",
      "}\n",
      "414\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1014/1243 [34:26<08:03,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44512\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1015/1243 [34:28<07:49,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44540\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1016/1243 [34:30<07:17,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44542\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1017/1243 [34:32<07:28,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44543\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1018/1243 [34:34<07:11,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44544\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 14}\n",
      "{}\n",
      "/**\n",
      "     * Write a 32-bit unsigned int.\n",
      "     * NOTE: This method will break if we change to big-endian.\n",
      "     */\n",
      "public void writeUInt(final long val)\n",
      "**************************************\n",
      "public void writeUInt(final long val) {\n",
      "    if (val < 0) {\n",
      "        throw new IllegalArgumentException(\"Negative value (\" + val + \") passed to unsigned writing method.\");\n",
      "    }\n",
      "    if (val > MAX_UINT) {\n",
      "        throw new IllegalArgumentException(\"Value (\" + val + \") to large to be written as uint.\");\n",
      "    }\n",
      "    byteBuffer.clear();\n",
      "    byteBuffer.putLong(val);\n",
      "    writeByteBuffer(4);\n",
      "}\n",
      "161\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1019/1243 [34:35<06:57,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44550\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1020/1243 [34:37<07:08,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44552\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1021/1243 [34:39<06:53,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44611\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1022/1243 [34:41<07:19,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44633\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1023/1243 [34:44<07:31,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44726\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class SAMSequenceDictionary {\\n    public static final long serialVersionUID = 1L;;\\n    private List<SAMSequenceRecord> mSequences = new ArrayList<>();;\\n    private final Map<String, SAMSequenceRecord> mSequenceMap = new HashMap<>();;\\n    public SAMSequenceDictionary();\\n    public SAMSequenceDictionary(final List<SAMSequenceRecord> list);\\n    public List<SAMSequenceRecord> getSequences()\\n    private static Log log = Log.getInstance(SAMSequenceDictionary.class);;\\n    public SAMSequenceRecord getSequence(final String name)\\n    public void setSequences(final List<SAMSequenceRecord> list)\\n    public void addSequence(final SAMSequenceRecord sequenceRecord)\\n    public SAMSequenceRecord getSequence(final int sequenceIndex)\\n    public int getSequenceIndex(final String sequenceName)\\n    public int size()\\n    public long getReferenceLength()\\n    public boolean isEmpty()\\n    private static String DICT_MISMATCH_TEMPLATE = \"SAM dictionaries are not the same: %s.\";;\\n    public void assertSameDictionary(final SAMSequenceDictionary that)\\n    public boolean isSameDictionary(final SAMSequenceDictionary that)\\n    public boolean equals(Object o)\\n    public SAMSequenceRecord addSequenceAlias(final String originalName, final String altName)\\n    public SAMSequenceRecord addAlternativeSequenceName(final String originalName, final String altName)\\n    public String md5()\\n    public int hashCode()\\n    public String toString()\\n    public static final List<String> DEFAULT_DICTIONARY_EQUAL_TAG = Arrays.asList(SAMSequenceRecord.MD5_TAG, SAMSequenceRecord.SEQUENCE_LENGTH_TAG);;\\n    public static SAMSequenceDictionary mergeDictionaries(final SAMSequenceDictionary dict1, final SAMSequenceDictionary dict2, final List<String> tagsToMatch)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMSequenceDictionary.java', 'methodCount': 20, 'lineCount': 28}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{'Interval': 'public class Interval {\\n    private final boolean negativeStrand;;\\n    private final String name;;\\n    private final String contig;;\\n    private final int start;;\\n    private final int end;;\\n    public Interval(final String sequence, final int start, final int end);\\n    public Interval(final Locatable locatable);\\n    public Interval(final String sequence, final int start, final int end, final boolean negative, final String name);\\n    public String getSequence()\\n    public boolean isNegativeStrand()\\n    public boolean isPositiveStrand()\\n    public Strand getStrand()\\n    public String getName()\\n    public boolean intersects(final Interval other)\\n    public int getIntersectionLength(final Interval other)\\n    public Interval intersect(final Interval that)\\n    public boolean abuts(final Interval other)\\n    public int length()\\n    public Interval pad(final int left, final int right)\\n    public static long countBases(final Collection<Interval> intervals)\\n    public int compareTo(final Interval that)\\n    public boolean equals(final Object other)\\n    public boolean equalsWithStrandAndName(final Object other)\\n    public int hashCode()\\n    public String toString()\\n    public Interval clone()\\n    public String getContig()\\n    public int getStart()\\n    public int getEnd()\\n}\\n'}\n",
      "/**\n",
      "     * Writes the interval to the output stream.\n",
      "     * @param interval the interval to write.\n",
      "     */\n",
      "public void encode(final Interval interval)\n",
      "**************************************\n",
      "@Override\n",
      "public void encode(final Interval interval) {\n",
      "    final String name = interval.getName();\n",
      "    binaryCodec.writeInt(dict.getSequenceIndex(interval.getContig()));\n",
      "    binaryCodec.writeInt(interval.getStart());\n",
      "    binaryCodec.writeInt(interval.getEnd());\n",
      "    binaryCodec.writeBoolean(interval.isNegativeStrand());\n",
      "    binaryCodec.writeBoolean(name != null);\n",
      "    if (name != null) {\n",
      "        binaryCodec.writeString(name, false, true);\n",
      "    }\n",
      "}\n",
      "885\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1024/1243 [35:18<42:42, 11.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44765\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  82%| | 1025/1243 [35:20<31:56,  8.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44861\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1026/1243 [35:22<24:37,  6.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45096\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1027/1243 [35:24<19:28,  5.41s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45111\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1028/1243 [35:26<15:49,  4.42s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45113\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface BitInputStream {\\n     boolean readBit()\\n     int readBits(int length)\\n     long readLongBits(int length)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/io/BitInputStream.java', 'methodCount': 3, 'lineCount': 5}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface BitOutputStream {\\n     void write(int bitContainer, int nofBits)\\n     void write(long bitContainer, int nofBits)\\n     void write(byte bitContainer, int nofBits)\\n     void write(boolean bit)\\n     void write(boolean bit, long repeat)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/io/BitOutputStream.java', 'methodCount': 5, 'lineCount': 7}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ValidationUtils {\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String nameOfObject)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(final T collection, final String nameOfObject)\\n    public static boolean isNonEmpty(final Collection<?> collection)\\n    public static String nonEmpty(final String string, final String nameOfObject)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(final T collection)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/utils/ValidationUtils.java', 'methodCount': 10, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 0}\n",
      "{'BitOutputStream': 'public interface BitOutputStream {\\n     void write(int bitContainer, int nofBits)\\n     void write(long bitContainer, int nofBits)\\n     void write(byte bitContainer, int nofBits)\\n     void write(boolean bit)\\n     void write(boolean bit, long repeat)\\n}\\n', 'HuffmanBitCode': 'public class HuffmanBitCode {\\n    private final T symbol;;\\n    private final int codeWord;;\\n    private final int codeWordBitLength;;\\n    public HuffmanBitCode(final T symbol, final int codeWord, final int codeWordBitLength);\\n    public T getSymbol()\\n    public int getCodeWord()\\n    public int getCodeWordBitLength()\\n    public String toString()\\n    public String getBitCodeWithPrefix()\\n}\\n'}\n",
      "/**\n",
      "     *\n",
      "     * @param bitOutputStream stream to which the symbol should be written\n",
      "     * @param symbol symbol from the alphabet to be written to the stream\n",
      "     * @return the length of the codeword that was written\n",
      "     */\n",
      "public final long write(final BitOutputStream bitOutputStream, final T symbol)\n",
      "**************************************\n",
      "public final long write(final BitOutputStream bitOutputStream, final T symbol) {\n",
      "    final HuffmanBitCode<T> code = huffmanBitCodesBySymbol.get(symbol);\n",
      "    if (code == null) {\n",
      "        throw new RuntimeException(String.format(\"Attempt to write a symbol (%d) that is not in the symbol alphabet for this huffman encoder (found code word %s).\", symbol, code == null ? \"null\" : code.toString()));\n",
      "    }\n",
      "    bitOutputStream.write(code.getCodeWord(), code.getCodeWordBitLength());\n",
      "    return code.getCodeWordBitLength();\n",
      "}\n",
      "832\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1029/1243 [35:28<13:19,  3.73s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45117\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1030/1243 [35:30<11:36,  3.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45209\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1031/1243 [35:32<10:00,  2.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45291\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1032/1243 [35:34<09:09,  2.60s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45339\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1033/1243 [35:36<08:16,  2.36s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45369\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class CRAMException {\\n    public CRAMException();\\n    public CRAMException(final String s);\\n    public CRAMException(final String s, final Throwable throwable);\\n    public CRAMException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/CRAMException.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class CRAMVersion {\\n    private final int major;;\\n    private final int minor;;\\n    public CRAMVersion(final int major, final int minor);\\n    public CRAMVersion(final String version);\\n    public int getMajor()\\n    public int getMinor()\\n    public String toString()\\n    public int compareTo(final CRAMVersion o)\\n    public boolean compatibleWith(final CRAMVersion cramVersion)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/common/CRAMVersion.java', 'methodCount': 9, 'lineCount': 13}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': \"public class Block {\\n    /**\\n * Only external blocks have meaningful Content IDs\\n * Other blocks are required to have a Content ID of 0\\n */\\npublic static final int NO_CONTENT_ID = 0;;\\n    /**\\n * Compression method that applied to this block's content.\\n */\\nprivate final BlockCompressionMethod compressionMethod;;\\n    /**\\n * Identifies CRAM content type of the block.\\n */\\nprivate final BlockContentType contentType;;\\n    /**\\n * The External Block Content ID, or NO_CONTENT_ID for non-External block\\n */\\nprivate final int contentId;;\\n    /**\\n * The content stored in this block, in compressed form (if applicable). For blocks with\\n * BlockCompressionMethod.RAW, this is the raw content, which is the same in compressed\\n * and uncompressed form.\\n */\\nprivate final byte[] compressedContent;;\\n    /**\\n * The length of the content stored in this block when uncompressed\\n */\\nprivate final int uncompressedLength;;\\n    protected Block(final BlockCompressionMethod compressionMethod, final BlockContentType contentType, final int contentId, final byte[] compressedContent, final int uncompressedLength);\\n    private static Block createRawNonExternalBlock(final BlockContentType contentType, final byte[] rawContent)\\n    public static Block createGZIPFileHeaderBlock(final byte[] rawContent)\\n    public static Block createRawCompressionHeaderBlock(final byte[] rawContent)\\n    public static Block createRawSliceHeaderBlock(final byte[] rawContent)\\n    public static Block createRawCoreDataBlock(final byte[] rawContent)\\n    public static Block createExternalBlock(final BlockCompressionMethod compressionMethod, final int contentId, final byte[] compressedContent, final int uncompressedLength)\\n    public final BlockCompressionMethod getCompressionMethod()\\n    public final BlockContentType getContentType()\\n    public int getContentId()\\n    public final byte[] getRawContent()\\n    public final byte[] getUncompressedContent(final CompressorCache compressorCache)\\n    public int getUncompressedContentSize()\\n    public final byte[] getCompressedContent()\\n    public final int getCompressedContentSize()\\n    public static Block read(final CRAMVersion cramVersion, InputStream inputStream)\\n    public final void write(final CRAMVersion cramVersion, final OutputStream outputStream)\\n    private void doWrite(final OutputStream outputStream) throws IOException\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/structure/block/Block.java', 'methodCount': 19, 'lineCount': 27}, {'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/structure/block/BlockContentType.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ValidationUtils {\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String nameOfObject)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(final T collection, final String nameOfObject)\\n    public static boolean isNonEmpty(final Collection<?> collection)\\n    public static String nonEmpty(final String string, final String nameOfObject)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(final T collection)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/utils/ValidationUtils.java', 'methodCount': 10, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Write the coreBlock and each external block out to a CRAM stream. There is no predefined\n",
      "     * order that must be honored, since each block is identified by a unique contentID.\n",
      "     * @param cramVersion CRAM major version being written\n",
      "     * @param outputStream stream to write blocks to\n",
      "     */\n",
      "public void writeBlocks(final CRAMVersion cramVersion, final OutputStream outputStream)\n",
      "**************************************\n",
      "public void writeBlocks(final CRAMVersion cramVersion, final OutputStream outputStream) {\n",
      "    if (coreBlock == null) {\n",
      "        throw new IllegalArgumentException(\"A core block must be provided before slice blocks can be written to a CRAM stream\");\n",
      "    }\n",
      "    coreBlock.write(cramVersion, outputStream);\n",
      "    for (final Block block : externalBlocks.values()) {\n",
      "        block.write(cramVersion, outputStream);\n",
      "    }\n",
      "}\n",
      "1354\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1034/1243 [35:38<07:28,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45371\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1035/1243 [35:40<07:22,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45407\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1036/1243 [35:42<07:21,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45567\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  83%| | 1037/1243 [35:44<06:48,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45570\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1038/1243 [35:46<07:36,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45576\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class MutableInt {\\n    public int value;;\\n    public MutableInt();\\n    public MutableInt(final int initialValue);\\n    public MutableInt incrementValue(final int increment)\\n    public int hashCode()\\n    public boolean equals(Object o)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/common/MutableInt.java', 'methodCount': 5, 'lineCount': 8}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class ExternalCompressor {\\n    final public static int NO_COMPRESSION_ARG = -1;;\\n    final private static String argErrorMessage = \"Invalid compression arg (%d) requested for CRAM %s compressor\";;\\n    private BlockCompressionMethod method;;\\n    protected ExternalCompressor(final BlockCompressionMethod method);\\n    public abstract byte[] compress(byte[] data)\\n    public abstract byte[] uncompress(byte[] data)\\n    public BlockCompressionMethod getMethod()\\n    public String toString()\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public static ExternalCompressor getCompressorForMethod(final BlockCompressionMethod compressionMethod, final int compressorSpecificArg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/compression/ExternalCompressor.java', 'methodCount': 8, 'lineCount': 13}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class CanonicalHuffmanIntegerEncoding {\\n    private final HuffmanParams<Integer> huffmanParams;;\\n    public CanonicalHuffmanIntegerEncoding(final int[] symbols, final int[] bitLengths);\\n    public static CanonicalHuffmanIntegerEncoding fromSerializedEncodingParams(final byte[] serializedParams)\\n    public byte[] toSerializedEncodingParams()\\n    public CRAMCodec<Integer> buildCodec(final SliceBlocksReadStreams sliceBlocksReadStreams, final SliceBlocksWriteStreams sliceBlocksWriteStreams)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/encoding/core/CanonicalHuffmanIntegerEncoding.java', 'methodCount': 7, 'lineCount': 10}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface ReadFeature {\\n     int getPosition()\\n     byte getOperator()\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/encoding/readfeatures/ReadFeature.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': \"public class Substitution {\\n    public static final int NO_CODE = -1;;\\n    public static final byte operator = 'X';;\\n    /**\\n * zero-based position in read\\n */\\nprivate int position;;\\n    /**\\n * The read base, allowed values are ACGTN.\\n */\\nprivate byte base = -1;;\\n    /**\\n * The reference sequence base matching the position of this substitution, allowed values are ACGTN.\\n */\\nprivate byte referenceBase = -1;;\\n    /**\\n * A byte value denoting combination of the read base and the reference base.\\n */\\nprivate byte code = NO_CODE;;\\n    public Substitution(final int position, final byte code);\\n    public Substitution(int position, byte base, byte referenceBase);\\n    public byte getCode()\\n    public void setCode(final byte code)\\n    public byte getOperator()\\n    public int getPosition()\\n    public byte getBase()\\n    public byte getReferenceBase()\\n    public boolean equals(final Object obj)\\n    public int hashCode()\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/cram/encoding/readfeatures/Substitution.java', 'methodCount': 11, 'lineCount': 19}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class RuntimeIOException {\\n    public RuntimeIOException();\\n    public RuntimeIOException(final String s);\\n    public RuntimeIOException(final String s, final Throwable throwable);\\n    public RuntimeIOException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/RuntimeIOException.java', 'methodCount': 4, 'lineCount': 6}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ValidationUtils {\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String nameOfObject)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(final T collection, final String nameOfObject)\\n    public static boolean isNonEmpty(final Collection<?> collection)\\n    public static String nonEmpty(final String string, final String nameOfObject)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(final T collection)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/utils/ValidationUtils.java', 'methodCount': 10, 'lineCount': 12}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 9}\n",
      "{}\n",
      "/**\n",
      "     * Find a byte value never mentioned in the array\n",
      "     * @param array bytes\n",
      "     * @return byte value or -1 if the array contains all possible byte values.\n",
      "     */\n",
      " static int getUnusedByte(final byte[] array)\n",
      "**************************************\n",
      "static int getUnusedByte(final byte[] array) {\n",
      "    final byte[] usage = new byte[BYTE_SPACE_SIZE];\n",
      "    for (final byte b : array) {\n",
      "        usage[0xFF & b] = 1;\n",
      "    }\n",
      "    for (int i = 0; i < usage.length; i++) {\n",
      "        if (usage[i] == 0) {\n",
      "            return i;\n",
      "        }\n",
      "    }\n",
      "    return ALL_BYTES_USED;\n",
      "}\n",
      "1434\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1039/1243 [35:48<06:57,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45600\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1040/1243 [35:50<06:34,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45601\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1041/1243 [35:52<06:22,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45608\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1042/1243 [35:54<06:33,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45893\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1043/1243 [35:55<06:07,  1.84s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46026\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class RuntimeIOException {\\n    public RuntimeIOException();\\n    public RuntimeIOException(final String s);\\n    public RuntimeIOException(final String s, final Throwable throwable);\\n    public RuntimeIOException(final Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/RuntimeIOException.java', 'methodCount': 4, 'lineCount': 6}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Read enough bytes to fill the input buffer.\n",
      "     * @param b\n",
      "     * @throws EOFException If EOF is reached before buffer is filled\n",
      "     */\n",
      "public void readFully(byte[] b) throws IOException\n",
      "**************************************\n",
      "public void readFully(byte[] b) throws IOException {\n",
      "    int len = b.length;\n",
      "    if (len < 0) {\n",
      "        throw new IndexOutOfBoundsException();\n",
      "    }\n",
      "    int n = 0;\n",
      "    while (n < len) {\n",
      "        int count = read(b, n, len - n);\n",
      "        if (count < 0) {\n",
      "            throw new EOFException();\n",
      "        }\n",
      "        n += count;\n",
      "    }\n",
      "}\n",
      "279\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1044/1243 [35:57<05:53,  1.78s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46113\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1045/1243 [35:59<06:13,  1.88s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46164\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1046/1243 [36:01<06:28,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46265\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1047/1243 [36:03<06:42,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46266\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1048/1243 [36:06<06:46,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46346\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{'SAMRecord': 'public class SAMRecord {\\n    private final static Log LOG = Log.getInstance(SAMRecord.class);;\\n    public static final long serialVersionUID = 1L;;\\n    /**\\n * Alignment score for a good alignment, but where computing a Phred-score is not feasible.\\n */\\npublic static final int UNKNOWN_MAPPING_QUALITY = 255;;\\n    /**\\n * Alignment score for an unaligned read.\\n */\\npublic static final int NO_MAPPING_QUALITY = 0;;\\n    /**\\n * If a read has this reference name, it is unaligned, but not all unaligned reads have\\n * this reference name (see above).\\n */\\npublic static final String NO_ALIGNMENT_REFERENCE_NAME = \"*\";;\\n    /**\\n * If a read has this reference index, it is unaligned, but not all unaligned reads have\\n * this reference index (see above).\\n */\\npublic static final int NO_ALIGNMENT_REFERENCE_INDEX = -1;;\\n    /**\\n * Cigar string for an unaligned read.\\n */\\npublic static final String NO_ALIGNMENT_CIGAR = \"*\";;\\n    /**\\n * If a read has reference name \"*\", it will have this value for position.\\n */\\npublic static final int NO_ALIGNMENT_START = GenomicIndexUtil.UNSET_GENOMIC_LOCATION;;\\n    /**\\n * This should rarely be used, since a read with no sequence doesn\\'t make much sense.\\n */\\npublic static final byte[] NULL_SEQUENCE = new byte[0];;\\n    public static final String NULL_SEQUENCE_STRING = \"*\";;\\n    /**\\n * This should rarely be used, since all reads should have quality scores.\\n */\\npublic static final byte[] NULL_QUALS = new byte[0];;\\n    public static final String NULL_QUALS_STRING = \"*\";;\\n    /**\\n * abs(insertSize) must be <= this\\n */\\npublic static final int MAX_INSERT_SIZE = Integer.MAX_VALUE;;\\n    /**\\n * Tags that are known to need the reverse complement if the read is reverse complemented.\\n */\\n@SuppressWarnings(\"deprecated\")\\npublic static final List<String> TAGS_TO_REVERSE_COMPLEMENT = Arrays.asList(SAMTag.E2.name(), SAMTag.SQ.name());;\\n    /**\\n * Tags that are known to need the reverse if the read is reverse complemented.\\n */\\npublic static final List<String> TAGS_TO_REVERSE = Arrays.asList(SAMTag.OQ.name(), SAMTag.U2.name());;\\n    private String mReadName = null;;\\n    private byte[] mReadBases = NULL_SEQUENCE;;\\n    private byte[] mBaseQualities = NULL_QUALS;;\\n    private String mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;;\\n    private int mAlignmentStart = NO_ALIGNMENT_START;;\\n    private transient int mAlignmentEnd = NO_ALIGNMENT_START;;\\n    private int mMappingQuality = NO_MAPPING_QUALITY;;\\n    private String mCigarString = NO_ALIGNMENT_CIGAR;;\\n    private Cigar mCigar = null;;\\n    private List<AlignmentBlock> mAlignmentBlocks = null;;\\n    private int mFlags = 0;;\\n    private String mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;;\\n    private int mMateAlignmentStart = 0;;\\n    private int mInferredInsertSize = 0;;\\n    private SAMBinaryTagAndValue mAttributes = null;;\\n    protected Integer mReferenceIndex = null;;\\n    protected Integer mMateReferenceIndex = null;;\\n    /**\\n * Some attributes (e.g. CIGAR) are not decoded immediately.  Use this to decide how to validate when decoded.\\n */\\nprivate ValidationStringency mValidationStringency = ValidationStringency.SILENT;;\\n    /**\\n * File source of this record. May be null. Note that this field is not serializable (and therefore marked\\n * as transient) due to encapsulated stream objects within it -- so serializing a SAMRecord will cause its\\n * file source to be lost (if it had one).\\n */\\nprivate transient SAMFileSource mFileSource;;\\n    private SAMFileHeader mHeader = null;;\\n    /**\\n * Transient Map of attributes for use by anyone.\\n */\\nprivate transient Map<Object, Object> transientAttributes;;\\n    public SAMRecord(final SAMFileHeader header);\\n    public String getReadName()\\n    public int getReadNameLength()\\n    public void setReadName(final String value)\\n    public String getReadString()\\n    public void setReadString(final String value)\\n    public byte[] getReadBases()\\n    public void setReadBases(final byte[] value)\\n    public int getReadLength()\\n    public String getBaseQualityString()\\n    public void setBaseQualityString(final String value)\\n    public byte[] getBaseQualities()\\n    public void setBaseQualities(final byte[] value)\\n    public byte[] getOriginalBaseQualities()\\n    public void setOriginalBaseQualities(final byte[] oq)\\n    private static boolean hasReferenceName(final Integer referenceIndex, final String referenceName)\\n    private boolean hasReferenceName()\\n    private boolean hasMateReferenceName()\\n    public String getReferenceName()\\n    public void setReferenceName(final String referenceName)\\n    public Integer getReferenceIndex()\\n    public void setReferenceIndex(final int referenceIndex)\\n    public String getMateReferenceName()\\n    public void setMateReferenceName(final String mateReferenceName)\\n    public Integer getMateReferenceIndex()\\n    public void setMateReferenceIndex(final int mateReferenceIndex)\\n    protected static Integer resolveIndexFromName(final String referenceName, final SAMFileHeader header, final boolean strict)\\n    protected static String resolveNameFromIndex(final int referenceIndex, final SAMFileHeader header)\\n    public int getAlignmentStart()\\n    public void setAlignmentStart(final int value)\\n    public int getAlignmentEnd()\\n    public int getUnclippedStart()\\n    public int getUnclippedEnd()\\n    public int getReferencePositionAtReadPosition(final int position)\\n    public static int getReferencePositionAtReadPosition(final SAMRecord rec, final int position)\\n    public int getReadPositionAtReferencePosition(final int pos)\\n    public int getReadPositionAtReferencePosition(final int pos, final boolean returnLastBaseIfDeleted)\\n    public static int getReadPositionAtReferencePosition(final SAMRecord rec, final int pos, final boolean returnLastBaseIfDeleted)\\n    public int getMateAlignmentStart()\\n    public void setMateAlignmentStart(final int mateAlignmentStart)\\n    public int getInferredInsertSize()\\n    public void setInferredInsertSize(final int inferredInsertSize)\\n    public int getMappingQuality()\\n    public void setMappingQuality(final int value)\\n    public String getCigarString()\\n    public void setCigarString(final String value)\\n    public Cigar getCigar()\\n    public int getCigarLength()\\n    public void setCigar(final Cigar cigar)\\n    protected void initializeCigar(final Cigar cigar)\\n    public SAMReadGroupRecord getReadGroup()\\n    public int getFlags()\\n    public void setFlags(final int value)\\n    public boolean getReadPairedFlag()\\n    private void requireReadPaired()\\n    public boolean getProperPairFlag()\\n    private boolean getProperPairFlagUnchecked()\\n    public boolean getReadUnmappedFlag()\\n    public boolean getMateUnmappedFlag()\\n    private boolean getMateUnmappedFlagUnchecked()\\n    public boolean getReadNegativeStrandFlag()\\n    public boolean getMateNegativeStrandFlag()\\n    private boolean getMateNegativeStrandFlagUnchecked()\\n    public boolean getFirstOfPairFlag()\\n    private boolean getFirstOfPairFlagUnchecked()\\n    public boolean getSecondOfPairFlag()\\n    private boolean getSecondOfPairFlagUnchecked()\\n    public boolean getNotPrimaryAlignmentFlag()\\n    public boolean isSecondaryAlignment()\\n    public boolean getSupplementaryAlignmentFlag()\\n    public boolean getReadFailsVendorQualityCheckFlag()\\n    public boolean getDuplicateReadFlag()\\n    public void setReadPairedFlag(final boolean flag)\\n    public void setProperPairFlag(final boolean flag)\\n    public void setReadUmappedFlag(final boolean flag)\\n    public void setReadUnmappedFlag(final boolean flag)\\n    public void setMateUnmappedFlag(final boolean flag)\\n    public void setReadNegativeStrandFlag(final boolean flag)\\n    public void setMateNegativeStrandFlag(final boolean flag)\\n    public void setFirstOfPairFlag(final boolean flag)\\n    public void setSecondOfPairFlag(final boolean flag)\\n    public void setNotPrimaryAlignmentFlag(final boolean flag)\\n    public void setSecondaryAlignment(final boolean flag)\\n    public void setSupplementaryAlignmentFlag(final boolean flag)\\n    public void setReadFailsVendorQualityCheckFlag(final boolean flag)\\n    public void setDuplicateReadFlag(final boolean flag)\\n    public boolean isSecondaryOrSupplementary()\\n    private void setFlag(final boolean flag, final int bit)\\n    public ValidationStringency getValidationStringency()\\n    public void setValidationStringency(final ValidationStringency validationStringency)\\n    public boolean hasAttribute(final String tag)\\n    public boolean hasAttribute(final SAMTag tag)\\n    public Object getAttribute(final String tag)\\n    public Object getAttribute(final SAMTag tag)\\n    public Integer getIntegerAttribute(final SAMTag tag)\\n    public Integer getIntegerAttribute(final String tag)\\n    public Long getUnsignedIntegerAttribute(final String tag) throws SAMException\\n    public Long getUnsignedIntegerAttribute(final SAMTag tag) throws SAMException\\n    public Long getUnsignedIntegerAttribute(final short tag) throws SAMException\\n    public Short getShortAttribute(final SAMTag tag)\\n    public Short getShortAttribute(final String tag)\\n    public Byte getByteAttribute(final SAMTag tag)\\n    public Byte getByteAttribute(final String tag)\\n    public String getStringAttribute(final SAMTag tag)\\n    public String getStringAttribute(final String tag)\\n    public Character getCharacterAttribute(final SAMTag tag)\\n    public Character getCharacterAttribute(final String tag)\\n    public Float getFloatAttribute(final SAMTag tag)\\n    public Float getFloatAttribute(final String tag)\\n    public byte[] getByteArrayAttribute(final SAMTag tag)\\n    public byte[] getByteArrayAttribute(final String tag)\\n    public byte[] getUnsignedByteArrayAttribute(final SAMTag tag)\\n    public byte[] getUnsignedByteArrayAttribute(final String tag)\\n    public byte[] getSignedByteArrayAttribute(final SAMTag tag)\\n    public byte[] getSignedByteArrayAttribute(final String tag)\\n    public short[] getUnsignedShortArrayAttribute(final SAMTag tag)\\n    public short[] getUnsignedShortArrayAttribute(final String tag)\\n    public short[] getSignedShortArrayAttribute(final SAMTag tag)\\n    public short[] getSignedShortArrayAttribute(final String tag)\\n    public int[] getUnsignedIntArrayAttribute(final SAMTag tag)\\n    public int[] getUnsignedIntArrayAttribute(final String tag)\\n    public int[] getSignedIntArrayAttribute(final SAMTag tag)\\n    public int[] getSignedIntArrayAttribute(final String tag)\\n    public float[] getFloatArrayAttribute(final SAMTag tag)\\n    public float[] getFloatArrayAttribute(final String tag)\\n    public boolean isUnsignedArrayAttribute(final String tag)\\n    private void requireSigned(final String tag)\\n    private void requireUnsigned(final String tag)\\n    public Object getAttribute(final short tag)\\n    public void setAttribute(final String tag, final Object value)\\n    public void setUnsignedArrayAttribute(final String tag, final Object value)\\n    protected void setAttribute(final short tag, final Object value)\\n    public void setAttribute(final SAMTag tag, final Object value)\\n    protected static boolean isAllowedAttributeValue(final Object value)\\n    protected void setAttribute(final SAMTag tag, final Object value, final boolean isUnsignedArray)\\n    protected void setAttribute(final short tag, final Object value, final boolean isUnsignedArray)\\n    public void clearAttributes()\\n    protected void setAttributes(final SAMBinaryTagAndValue attributes)\\n    protected SAMBinaryTagAndValue getBinaryAttributes()\\n    public String getContig()\\n    public int getStart()\\n    public int getEnd()\\n    public List<SAMTagAndValue> getAttributes()\\n    public int computeIndexingBinIfAbsent(final SAMRecord alignment)\\n     int computeIndexingBin()\\n    public SAMFileHeader getHeader()\\n    public void setHeader(final SAMFileHeader header)\\n    public void setHeaderStrict(final SAMFileHeader header)\\n    public byte[] getVariableBinaryRepresentation()\\n    public int getAttributesBinarySize()\\n    public String format()\\n    private void addField(final StringBuilder buffer, final Object value, final Object defaultValue, final String defaultString)\\n    private void addField(final StringBuilder buffer, final String field)\\n    private static String formatTagValue(final short tag, final Object value)\\n    private boolean safeEquals(final Object o1, final Object o2)\\n    protected void eagerDecode()\\n    public List<AlignmentBlock> getAlignmentBlocks()\\n    public List<SAMValidationError> validateCigar(final long recordNumber)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public List<SAMValidationError> isValid()\\n    public List<SAMValidationError> isValid(final boolean firstOnly)\\n    public SAMFileSource getFileSource()\\n    public void setFileSource(final SAMFileSource fileSource)\\n    private List<SAMValidationError> isValidReferenceIndexAndPosition(final Integer referenceIndex, final String referenceName, final int alignmentStart, final boolean isMate)\\n    private List<SAMValidationError> isValidReferenceIndexAndPosition(final Integer referenceIndex, final String referenceName, final int alignmentStart, final boolean isMate, final boolean firstOnly)\\n    private String buildMessage(final String baseMessage, final boolean isMate)\\n    public Object clone() throws CloneNotSupportedException\\n    public SAMRecord deepCopy()\\n    public String toString()\\n    public String getSAMString()\\n    public String getPairedReadName()\\n    public final Set<SAMFlag> getSAMFlags()\\n    public final Object getTransientAttribute(final Object key)\\n    public final Object setTransientAttribute(final Object key, final Object value)\\n    public final Object removeTransientAttribute(final Object key)\\n    public void reverseComplement()\\n    public void reverseComplement(boolean inplace)\\n    public void reverseComplement(final Collection<String> tagsToRevcomp, final Collection<String> tagsToReverse, boolean inplace)\\n    private static void reverseArray(final byte[] array)\\n    private static void reverseArray(final short[] array)\\n    private static void reverseArray(final int[] array)\\n    private static void reverseArray(final float[] array)\\n}\\n'}\n",
      "/**\n",
      "     * Basic sanity check for a pair of SAMRecords.\n",
      "     * @throws SanityCheckFailedException if the sanity check failed\n",
      "     */\n",
      "public void assertPairValid(final SAMRecord firstEnd, final SAMRecord secondEnd) throws SanityCheckFailedException\n",
      "**************************************\n",
      "/**\n",
      " * Basic sanity check for a pair of SAMRecords.\n",
      " * @throws SanityCheckFailedException if the sanity check failed\n",
      " */\n",
      "public void assertPairValid(final SAMRecord firstEnd, final SAMRecord secondEnd) throws SanityCheckFailedException {\n",
      "    assertEquals(firstEnd.getReadName(), secondEnd.getReadName());\n",
      "    assertTrue(firstEnd.getFirstOfPairFlag());\n",
      "    assertTrue(secondEnd.getSecondOfPairFlag());\n",
      "    assertFalse(secondEnd.getFirstOfPairFlag());\n",
      "    assertFalse(firstEnd.getSecondOfPairFlag());\n",
      "    if (!firstEnd.getReadUnmappedFlag() && !secondEnd.getReadUnmappedFlag()) {\n",
      "        assertNotSame(firstEnd.getReadNegativeStrandFlag(), secondEnd.getReadNegativeStrandFlag());\n",
      "    }\n",
      "}\n",
      "3429\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1049/1243 [36:08<06:48,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46369\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  84%| | 1050/1243 [36:09<06:21,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46375\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1051/1243 [36:11<06:21,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46412\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1052/1243 [36:14<06:42,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46414\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1053/1243 [36:16<06:57,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46422\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 36, 'classCount': 1, 'methodsOnly': 'public class SAMRecord {\\n    private final static Log LOG = Log.getInstance(SAMRecord.class);;\\n    public static final long serialVersionUID = 1L;;\\n    /**\\n * Alignment score for a good alignment, but where computing a Phred-score is not feasible.\\n */\\npublic static final int UNKNOWN_MAPPING_QUALITY = 255;;\\n    /**\\n * Alignment score for an unaligned read.\\n */\\npublic static final int NO_MAPPING_QUALITY = 0;;\\n    /**\\n * If a read has this reference name, it is unaligned, but not all unaligned reads have\\n * this reference name (see above).\\n */\\npublic static final String NO_ALIGNMENT_REFERENCE_NAME = \"*\";;\\n    /**\\n * If a read has this reference index, it is unaligned, but not all unaligned reads have\\n * this reference index (see above).\\n */\\npublic static final int NO_ALIGNMENT_REFERENCE_INDEX = -1;;\\n    /**\\n * Cigar string for an unaligned read.\\n */\\npublic static final String NO_ALIGNMENT_CIGAR = \"*\";;\\n    /**\\n * If a read has reference name \"*\", it will have this value for position.\\n */\\npublic static final int NO_ALIGNMENT_START = GenomicIndexUtil.UNSET_GENOMIC_LOCATION;;\\n    /**\\n * This should rarely be used, since a read with no sequence doesn\\'t make much sense.\\n */\\npublic static final byte[] NULL_SEQUENCE = new byte[0];;\\n    public static final String NULL_SEQUENCE_STRING = \"*\";;\\n    /**\\n * This should rarely be used, since all reads should have quality scores.\\n */\\npublic static final byte[] NULL_QUALS = new byte[0];;\\n    public static final String NULL_QUALS_STRING = \"*\";;\\n    /**\\n * abs(insertSize) must be <= this\\n */\\npublic static final int MAX_INSERT_SIZE = Integer.MAX_VALUE;;\\n    /**\\n * Tags that are known to need the reverse complement if the read is reverse complemented.\\n */\\n@SuppressWarnings(\"deprecated\")\\npublic static final List<String> TAGS_TO_REVERSE_COMPLEMENT = Arrays.asList(SAMTag.E2.name(), SAMTag.SQ.name());;\\n    /**\\n * Tags that are known to need the reverse if the read is reverse complemented.\\n */\\npublic static final List<String> TAGS_TO_REVERSE = Arrays.asList(SAMTag.OQ.name(), SAMTag.U2.name());;\\n    private String mReadName = null;;\\n    private byte[] mReadBases = NULL_SEQUENCE;;\\n    private byte[] mBaseQualities = NULL_QUALS;;\\n    private String mReferenceName = NO_ALIGNMENT_REFERENCE_NAME;;\\n    private int mAlignmentStart = NO_ALIGNMENT_START;;\\n    private transient int mAlignmentEnd = NO_ALIGNMENT_START;;\\n    private int mMappingQuality = NO_MAPPING_QUALITY;;\\n    private String mCigarString = NO_ALIGNMENT_CIGAR;;\\n    private Cigar mCigar = null;;\\n    private List<AlignmentBlock> mAlignmentBlocks = null;;\\n    private int mFlags = 0;;\\n    private String mMateReferenceName = NO_ALIGNMENT_REFERENCE_NAME;;\\n    private int mMateAlignmentStart = 0;;\\n    private int mInferredInsertSize = 0;;\\n    private SAMBinaryTagAndValue mAttributes = null;;\\n    protected Integer mReferenceIndex = null;;\\n    protected Integer mMateReferenceIndex = null;;\\n    /**\\n * Some attributes (e.g. CIGAR) are not decoded immediately.  Use this to decide how to validate when decoded.\\n */\\nprivate ValidationStringency mValidationStringency = ValidationStringency.SILENT;;\\n    /**\\n * File source of this record. May be null. Note that this field is not serializable (and therefore marked\\n * as transient) due to encapsulated stream objects within it -- so serializing a SAMRecord will cause its\\n * file source to be lost (if it had one).\\n */\\nprivate transient SAMFileSource mFileSource;;\\n    private SAMFileHeader mHeader = null;;\\n    /**\\n * Transient Map of attributes for use by anyone.\\n */\\nprivate transient Map<Object, Object> transientAttributes;;\\n    public SAMRecord(final SAMFileHeader header);\\n    public String getReadName()\\n    public int getReadNameLength()\\n    public void setReadName(final String value)\\n    public String getReadString()\\n    public void setReadString(final String value)\\n    public byte[] getReadBases()\\n    public void setReadBases(final byte[] value)\\n    public int getReadLength()\\n    public String getBaseQualityString()\\n    public void setBaseQualityString(final String value)\\n    public byte[] getBaseQualities()\\n    public void setBaseQualities(final byte[] value)\\n    public byte[] getOriginalBaseQualities()\\n    public void setOriginalBaseQualities(final byte[] oq)\\n    private static boolean hasReferenceName(final Integer referenceIndex, final String referenceName)\\n    private boolean hasReferenceName()\\n    private boolean hasMateReferenceName()\\n    public String getReferenceName()\\n    public void setReferenceName(final String referenceName)\\n    public Integer getReferenceIndex()\\n    public void setReferenceIndex(final int referenceIndex)\\n    public String getMateReferenceName()\\n    public void setMateReferenceName(final String mateReferenceName)\\n    public Integer getMateReferenceIndex()\\n    public void setMateReferenceIndex(final int mateReferenceIndex)\\n    protected static Integer resolveIndexFromName(final String referenceName, final SAMFileHeader header, final boolean strict)\\n    protected static String resolveNameFromIndex(final int referenceIndex, final SAMFileHeader header)\\n    public int getAlignmentStart()\\n    public void setAlignmentStart(final int value)\\n    public int getAlignmentEnd()\\n    public int getUnclippedStart()\\n    public int getUnclippedEnd()\\n    public int getReferencePositionAtReadPosition(final int position)\\n    public static int getReferencePositionAtReadPosition(final SAMRecord rec, final int position)\\n    public int getReadPositionAtReferencePosition(final int pos)\\n    public int getReadPositionAtReferencePosition(final int pos, final boolean returnLastBaseIfDeleted)\\n    public static int getReadPositionAtReferencePosition(final SAMRecord rec, final int pos, final boolean returnLastBaseIfDeleted)\\n    public int getMateAlignmentStart()\\n    public void setMateAlignmentStart(final int mateAlignmentStart)\\n    public int getInferredInsertSize()\\n    public void setInferredInsertSize(final int inferredInsertSize)\\n    public int getMappingQuality()\\n    public void setMappingQuality(final int value)\\n    public String getCigarString()\\n    public void setCigarString(final String value)\\n    public Cigar getCigar()\\n    public int getCigarLength()\\n    public void setCigar(final Cigar cigar)\\n    protected void initializeCigar(final Cigar cigar)\\n    public SAMReadGroupRecord getReadGroup()\\n    public int getFlags()\\n    public void setFlags(final int value)\\n    public boolean getReadPairedFlag()\\n    private void requireReadPaired()\\n    public boolean getProperPairFlag()\\n    private boolean getProperPairFlagUnchecked()\\n    public boolean getReadUnmappedFlag()\\n    public boolean getMateUnmappedFlag()\\n    private boolean getMateUnmappedFlagUnchecked()\\n    public boolean getReadNegativeStrandFlag()\\n    public boolean getMateNegativeStrandFlag()\\n    private boolean getMateNegativeStrandFlagUnchecked()\\n    public boolean getFirstOfPairFlag()\\n    private boolean getFirstOfPairFlagUnchecked()\\n    public boolean getSecondOfPairFlag()\\n    private boolean getSecondOfPairFlagUnchecked()\\n    public boolean getNotPrimaryAlignmentFlag()\\n    public boolean isSecondaryAlignment()\\n    public boolean getSupplementaryAlignmentFlag()\\n    public boolean getReadFailsVendorQualityCheckFlag()\\n    public boolean getDuplicateReadFlag()\\n    public void setReadPairedFlag(final boolean flag)\\n    public void setProperPairFlag(final boolean flag)\\n    public void setReadUmappedFlag(final boolean flag)\\n    public void setReadUnmappedFlag(final boolean flag)\\n    public void setMateUnmappedFlag(final boolean flag)\\n    public void setReadNegativeStrandFlag(final boolean flag)\\n    public void setMateNegativeStrandFlag(final boolean flag)\\n    public void setFirstOfPairFlag(final boolean flag)\\n    public void setSecondOfPairFlag(final boolean flag)\\n    public void setNotPrimaryAlignmentFlag(final boolean flag)\\n    public void setSecondaryAlignment(final boolean flag)\\n    public void setSupplementaryAlignmentFlag(final boolean flag)\\n    public void setReadFailsVendorQualityCheckFlag(final boolean flag)\\n    public void setDuplicateReadFlag(final boolean flag)\\n    public boolean isSecondaryOrSupplementary()\\n    private void setFlag(final boolean flag, final int bit)\\n    public ValidationStringency getValidationStringency()\\n    public void setValidationStringency(final ValidationStringency validationStringency)\\n    public boolean hasAttribute(final String tag)\\n    public boolean hasAttribute(final SAMTag tag)\\n    public Object getAttribute(final String tag)\\n    public Object getAttribute(final SAMTag tag)\\n    public Integer getIntegerAttribute(final SAMTag tag)\\n    public Integer getIntegerAttribute(final String tag)\\n    public Long getUnsignedIntegerAttribute(final String tag) throws SAMException\\n    public Long getUnsignedIntegerAttribute(final SAMTag tag) throws SAMException\\n    public Long getUnsignedIntegerAttribute(final short tag) throws SAMException\\n    public Short getShortAttribute(final SAMTag tag)\\n    public Short getShortAttribute(final String tag)\\n    public Byte getByteAttribute(final SAMTag tag)\\n    public Byte getByteAttribute(final String tag)\\n    public String getStringAttribute(final SAMTag tag)\\n    public String getStringAttribute(final String tag)\\n    public Character getCharacterAttribute(final SAMTag tag)\\n    public Character getCharacterAttribute(final String tag)\\n    public Float getFloatAttribute(final SAMTag tag)\\n    public Float getFloatAttribute(final String tag)\\n    public byte[] getByteArrayAttribute(final SAMTag tag)\\n    public byte[] getByteArrayAttribute(final String tag)\\n    public byte[] getUnsignedByteArrayAttribute(final SAMTag tag)\\n    public byte[] getUnsignedByteArrayAttribute(final String tag)\\n    public byte[] getSignedByteArrayAttribute(final SAMTag tag)\\n    public byte[] getSignedByteArrayAttribute(final String tag)\\n    public short[] getUnsignedShortArrayAttribute(final SAMTag tag)\\n    public short[] getUnsignedShortArrayAttribute(final String tag)\\n    public short[] getSignedShortArrayAttribute(final SAMTag tag)\\n    public short[] getSignedShortArrayAttribute(final String tag)\\n    public int[] getUnsignedIntArrayAttribute(final SAMTag tag)\\n    public int[] getUnsignedIntArrayAttribute(final String tag)\\n    public int[] getSignedIntArrayAttribute(final SAMTag tag)\\n    public int[] getSignedIntArrayAttribute(final String tag)\\n    public float[] getFloatArrayAttribute(final SAMTag tag)\\n    public float[] getFloatArrayAttribute(final String tag)\\n    public boolean isUnsignedArrayAttribute(final String tag)\\n    private void requireSigned(final String tag)\\n    private void requireUnsigned(final String tag)\\n    public Object getAttribute(final short tag)\\n    public void setAttribute(final String tag, final Object value)\\n    public void setUnsignedArrayAttribute(final String tag, final Object value)\\n    protected void setAttribute(final short tag, final Object value)\\n    public void setAttribute(final SAMTag tag, final Object value)\\n    protected static boolean isAllowedAttributeValue(final Object value)\\n    protected void setAttribute(final SAMTag tag, final Object value, final boolean isUnsignedArray)\\n    protected void setAttribute(final short tag, final Object value, final boolean isUnsignedArray)\\n    public void clearAttributes()\\n    protected void setAttributes(final SAMBinaryTagAndValue attributes)\\n    protected SAMBinaryTagAndValue getBinaryAttributes()\\n    public String getContig()\\n    public int getStart()\\n    public int getEnd()\\n    public List<SAMTagAndValue> getAttributes()\\n    public int computeIndexingBinIfAbsent(final SAMRecord alignment)\\n     int computeIndexingBin()\\n    public SAMFileHeader getHeader()\\n    public void setHeader(final SAMFileHeader header)\\n    public void setHeaderStrict(final SAMFileHeader header)\\n    public byte[] getVariableBinaryRepresentation()\\n    public int getAttributesBinarySize()\\n    public String format()\\n    private void addField(final StringBuilder buffer, final Object value, final Object defaultValue, final String defaultString)\\n    private void addField(final StringBuilder buffer, final String field)\\n    private static String formatTagValue(final short tag, final Object value)\\n    private boolean safeEquals(final Object o1, final Object o2)\\n    protected void eagerDecode()\\n    public List<AlignmentBlock> getAlignmentBlocks()\\n    public List<SAMValidationError> validateCigar(final long recordNumber)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public List<SAMValidationError> isValid()\\n    public List<SAMValidationError> isValid(final boolean firstOnly)\\n    public SAMFileSource getFileSource()\\n    public void setFileSource(final SAMFileSource fileSource)\\n    private List<SAMValidationError> isValidReferenceIndexAndPosition(final Integer referenceIndex, final String referenceName, final int alignmentStart, final boolean isMate)\\n    private List<SAMValidationError> isValidReferenceIndexAndPosition(final Integer referenceIndex, final String referenceName, final int alignmentStart, final boolean isMate, final boolean firstOnly)\\n    private String buildMessage(final String baseMessage, final boolean isMate)\\n    public Object clone() throws CloneNotSupportedException\\n    public SAMRecord deepCopy()\\n    public String toString()\\n    public String getSAMString()\\n    public String getPairedReadName()\\n    public final Set<SAMFlag> getSAMFlags()\\n    public final Object getTransientAttribute(final Object key)\\n    public final Object setTransientAttribute(final Object key, final Object value)\\n    public final Object removeTransientAttribute(final Object key)\\n    public void reverseComplement()\\n    public void reverseComplement(boolean inplace)\\n    public void reverseComplement(final Collection<String> tagsToRevcomp, final Collection<String> tagsToReverse, boolean inplace)\\n    private static void reverseArray(final byte[] array)\\n    private static void reverseArray(final short[] array)\\n    private static void reverseArray(final int[] array)\\n    private static void reverseArray(final float[] array)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/SAMRecord.java', 'methodCount': 183, 'lineCount': 221}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Determines whether a SAMRecord matches this filter\n",
      "     *\n",
      "     * @param record the SAMRecord to evaluate\n",
      "     *\n",
      "     * @return true if the SAMRecord matches the filter, otherwise false\n",
      "     */\n",
      "public boolean filterOut(final SAMRecord record)\n",
      "**************************************\n",
      "/**\n",
      " * Determines whether a SAMRecord matches this filter\n",
      " *\n",
      " * @param record the SAMRecord to evaluate\n",
      " *\n",
      " * @return true if the SAMRecord matches the filter, otherwise false\n",
      " */\n",
      "@Override\n",
      "public boolean filterOut(final SAMRecord record) {\n",
      "    if (includeAligned) {\n",
      "        if (!record.getReadUnmappedFlag()) {\n",
      "            return false;\n",
      "        }\n",
      "    } else {\n",
      "        // exclude aligned\n",
      "        if (record.getReadUnmappedFlag()) {\n",
      "            return false;\n",
      "        }\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "3448\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1054/1243 [36:19<07:04,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46423\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1055/1243 [36:21<06:59,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46525\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1056/1243 [36:23<07:01,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46528\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1057/1243 [36:25<06:56,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46554\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1058/1243 [36:27<06:25,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46569\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class StringUtil {\\n    private static final byte UPPER_CASE_OFFSET = \\'A\\' - \\'a\\';;\\n    public static String join(final String separator, final Collection<T> objs)\\n    public static String join(final String separator, final T... objs)\\n    public static int split(final String aString, final String[] tokens, final char delim)\\n    public static int splitConcatenateExcessTokens(final String aString, final String[] tokens, final char delim)\\n    public static byte toLowerCase(final byte b)\\n    public static byte toUpperCase(final byte b)\\n    public static void toUpperCase(final byte[] bytes)\\n    public static String assertCharactersNotInString(final String illegalChars, final char... chars)\\n    public static String wordWrap(final String s, final int maxLineLength)\\n    public static String wordWrapSingleLine(final String s, final int maxLineLength)\\n    public static String intValuesToString(final int[] intVals)\\n    public static String intValuesToString(final short[] shortVals)\\n    public static String bytesToString(final byte[] data)\\n    public static String bytesToString(final byte[] buffer, final int offset, final int length)\\n    public static byte[] stringToBytes(final String s)\\n    public static byte[] stringToBytes(final String s, final int offset, final int length)\\n    public static String readNullTerminatedString(final BinaryCodec binaryCodec)\\n    public static void charsToBytes(final char[] chars, final int charOffset, final int length, final byte[] bytes, final int byteOffset)\\n    public static byte charToByte(final char c)\\n    public static char byteToChar(final byte b)\\n    public static String bytesToHexString(final byte[] data)\\n    public static byte[] hexStringToBytes(final String s) throws NumberFormatException\\n    public static char toHexDigit(final int value)\\n    public static int fromHexDigit(final char c) throws NumberFormatException\\n    public static String reverseString(final String s)\\n    public static boolean isBlank(String str)\\n    public static String repeatCharNTimes(char c, int repeatNumber)\\n    public static final String EMPTY_STRING = \"\";;\\n    public static String asEmptyIfNull(final Object string)\\n    public static int levenshteinDistance(final String string1, final String string2, int swap, int substitution, int insertion, int deletion)\\n    public static int hammingDistance(final String s1, final String s2)\\n    public static boolean isWithinHammingDistance(final String s1, final String s2, final int maxHammingDistance)\\n    public static String humanReadableByteCount(long bytes)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/StringUtil.java', 'methodCount': 32, 'lineCount': 36}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'Cigar': 'public class Cigar {\\n    public static final long serialVersionUID = 1L;;\\n    private final List<CigarElement> cigarElements = new ArrayList<CigarElement>();;\\n    public Cigar();\\n    public Cigar(final List<CigarElement> cigarElements);\\n    public List<CigarElement> getCigarElements()\\n    public CigarElement getCigarElement(final int i)\\n    public void add(final CigarElement cigarElement)\\n    public int numCigarElements()\\n    public boolean isEmpty()\\n    public int getReferenceLength()\\n    public int getPaddedReferenceLength()\\n    public int getReadLength()\\n    public static int getReadLength(final List<CigarElement> cigarElements)\\n    public List<SAMValidationError> isValid(final String readName, final long recordNumber)\\n    private static boolean isRealOperator(final CigarOperator op)\\n    private static boolean isInDelOperator(final CigarOperator op)\\n    private static boolean isClippingOperator(final CigarOperator op)\\n    private static boolean isPaddingOperator(final CigarOperator op)\\n    public boolean equals(final Object o)\\n    public static Cigar fromCigarOperators(final List<CigarOperator> cigarOperators)\\n    public static Cigar fromCigarString(String cigarString)\\n    public Iterator<CigarElement> iterator()\\n    public boolean containsOperator(final CigarOperator operator)\\n    public CigarElement getFirstCigarElement()\\n    public CigarElement getLastCigarElement()\\n    public boolean isLeftClipped()\\n    public boolean isRightClipped()\\n    public boolean isClipped()\\n    public int hashCode()\\n    public String toString()\\n}\\n', 'CigarElement': 'public class CigarElement {\\n    public static final long serialVersionUID = 1L;;\\n    private final int length;;\\n    private final CigarOperator operator;;\\n    public CigarElement(final int length, final CigarOperator operator);\\n    public int getLength()\\n    public CigarOperator getOperator()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Convert from Cigar class representation to String.\n",
      "     * @param cigar in Cigar class format\n",
      "     * @return CIGAR in String form ala SAM text file.  \"*\" means empty CIGAR.\n",
      "     */\n",
      "public static String encode(final Cigar cigar)\n",
      "**************************************\n",
      "/**\n",
      " * Convert from Cigar class representation to String.\n",
      " * @param cigar in Cigar class format\n",
      " * @return CIGAR in String form ala SAM text file.  \"*\" means empty CIGAR.\n",
      " */\n",
      "public static String encode(final Cigar cigar) {\n",
      "    if (cigar.isEmpty()) {\n",
      "        return SAMRecord.NO_ALIGNMENT_CIGAR;\n",
      "    }\n",
      "    final StringBuilder ret = new StringBuilder();\n",
      "    for (final CigarElement cigarElement : cigar.getCigarElements()) {\n",
      "        ret.append(cigarElement.getLength());\n",
      "        ret.append(cigarElement.getOperator());\n",
      "    }\n",
      "    return ret.toString();\n",
      "}\n",
      "1308\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1059/1243 [36:29<06:12,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46602\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1060/1243 [36:31<06:13,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46682\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1061/1243 [36:33<06:24,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46737\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  85%| | 1062/1243 [36:35<06:02,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46749\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1063/1243 [36:37<05:43,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46758\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class StringUtil {\\n    private static final byte UPPER_CASE_OFFSET = \\'A\\' - \\'a\\';;\\n    public static String join(final String separator, final Collection<T> objs)\\n    public static String join(final String separator, final T... objs)\\n    public static int split(final String aString, final String[] tokens, final char delim)\\n    public static int splitConcatenateExcessTokens(final String aString, final String[] tokens, final char delim)\\n    public static byte toLowerCase(final byte b)\\n    public static byte toUpperCase(final byte b)\\n    public static void toUpperCase(final byte[] bytes)\\n    public static String assertCharactersNotInString(final String illegalChars, final char... chars)\\n    public static String wordWrap(final String s, final int maxLineLength)\\n    public static String wordWrapSingleLine(final String s, final int maxLineLength)\\n    public static String intValuesToString(final int[] intVals)\\n    public static String intValuesToString(final short[] shortVals)\\n    public static String bytesToString(final byte[] data)\\n    public static String bytesToString(final byte[] buffer, final int offset, final int length)\\n    public static byte[] stringToBytes(final String s)\\n    public static byte[] stringToBytes(final String s, final int offset, final int length)\\n    public static String readNullTerminatedString(final BinaryCodec binaryCodec)\\n    public static void charsToBytes(final char[] chars, final int charOffset, final int length, final byte[] bytes, final int byteOffset)\\n    public static byte charToByte(final char c)\\n    public static char byteToChar(final byte b)\\n    public static String bytesToHexString(final byte[] data)\\n    public static byte[] hexStringToBytes(final String s) throws NumberFormatException\\n    public static char toHexDigit(final int value)\\n    public static int fromHexDigit(final char c) throws NumberFormatException\\n    public static String reverseString(final String s)\\n    public static boolean isBlank(String str)\\n    public static String repeatCharNTimes(char c, int repeatNumber)\\n    public static final String EMPTY_STRING = \"\";;\\n    public static String asEmptyIfNull(final Object string)\\n    public static int levenshteinDistance(final String string1, final String string2, int swap, int substitution, int insertion, int deletion)\\n    public static int hammingDistance(final String s1, final String s2)\\n    public static boolean isWithinHammingDistance(final String s1, final String s2, final int maxHammingDistance)\\n    public static String humanReadableByteCount(long bytes)\\n}\\n', 'filePath': '/home/ubuntu/repos/samtools,htsjdk/src/main/java/htsjdk/samtools/util/StringUtil.java', 'methodCount': 32, 'lineCount': 36}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Removes all attributes.\n",
      "     */\n",
      "public void clearAttributes()\n",
      "**************************************\n",
      "@Override\n",
      "public void clearAttributes() {\n",
      "    if (!mAttributesDecoded) {\n",
      "        getCigar();\n",
      "    }\n",
      "    mAttributesDecoded = true;\n",
      "    mBinaryDataStale = true;\n",
      "    super.clearAttributes();\n",
      "}\n",
      "725\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1064/1243 [36:40<06:35,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46852\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1065/1243 [36:42<06:31,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46902\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1066/1243 [36:44<06:37,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46932\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1067/1243 [36:46<06:09,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "46933\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1068/1243 [36:48<06:12,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47156\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface GATKRead {\\n     String getName()\\n     int getFlags()\\n     void setName(final String name)\\n     int getLength()\\n     boolean isEmpty()\\n     void setPosition(final String contig, final int start)\\n     void setPosition(final Locatable locatable)\\n     String getAssignedContig()\\n     int getAssignedStart()\\n     int getUnclippedStart()\\n     int getUnclippedEnd()\\n     int getSoftStart()\\n     int getSoftEnd()\\n     int getAdaptorBoundary()\\n     String getMateContig()\\n     int getMateStart()\\n     void setMatePosition(final String contig, final int start)\\n     void setMatePosition(final Locatable locatable)\\n     int getFragmentLength()\\n     void setFragmentLength(final int fragmentLength)\\n     int getMappingQuality()\\n     void setMappingQuality(final int mappingQuality)\\n     byte[] getBases()\\n     byte[] getBasesNoCopy()\\n     int copyBases(final int offset, final byte[] destination, final int destinationOffset, final int maxLength)\\n     int copyBaseQualities(final int offset, final byte[] destination, final int destinationOffset, final int maxLength)\\n     boolean hasBases()\\n     boolean hasBaseQualities()\\n     byte getBase(final int i)\\n     String getBasesString()\\n     void setBases(final byte[] bases)\\n     byte[] getBaseQualities()\\n     byte[] getBaseQualitiesNoCopy()\\n     int getBaseQualityCount()\\n     byte getBaseQuality(final int i)\\n     void setBaseQualities(final byte[] baseQualities)\\n     Cigar getCigar()\\n     List<CigarElement> getCigarElements()\\n     CigarElement getCigarElement(final int i)\\n     int numCigarElements()\\n     void setCigar(final Cigar cigar)\\n     void setCigar(final String cigarString)\\n     String getReadGroup()\\n     void setReadGroup(final String readGroupID)\\n     boolean isPaired()\\n     void setIsPaired(final boolean isPaired)\\n     boolean isProperlyPaired()\\n     void setIsProperlyPaired(final boolean isProperlyPaired)\\n     boolean isUnmapped()\\n     void setIsUnmapped()\\n     boolean isUnplaced()\\n     void setIsUnplaced()\\n     boolean mateIsUnmapped()\\n     void setMateIsUnmapped()\\n     boolean mateIsUnplaced()\\n     void setMateIsUnplaced()\\n     boolean isReverseStrand()\\n     void setIsReverseStrand(final boolean isReverseStrand)\\n     boolean mateIsReverseStrand()\\n     void setMateIsReverseStrand(final boolean mateIsReverseStrand)\\n     boolean isFirstOfPair()\\n     void setIsFirstOfPair()\\n     boolean isSecondOfPair()\\n     void setIsSecondOfPair()\\n     boolean isSecondaryAlignment()\\n     void setIsSecondaryAlignment(final boolean isSecondaryAlignment)\\n     boolean isSupplementaryAlignment()\\n     void setIsSupplementaryAlignment(final boolean isSupplementaryAlignment)\\n     Optional<SamPairUtil.PairOrientation> getPairOrientation()\\n     boolean failsVendorQualityCheck()\\n     void setFailsVendorQualityCheck(final boolean failsVendorQualityCheck)\\n     boolean isDuplicate()\\n     void setIsDuplicate(final boolean isDuplicate)\\n     boolean hasAttribute(final String attributeName)\\n     Integer getAttributeAsInteger(final String attributeName)\\n     Float getAttributeAsFloat(final String attributeName)\\n     String getAttributeAsString(final String attributeName)\\n     byte[] getAttributeAsByteArray(final String attributeName)\\n     Object getTransientAttribute(final Object key)\\n     Optional<T> getOptionalTransientAttribute(final Object key, final Class<T> clazz)\\n     void setAttribute(final String attributeName, final Integer attributeValue)\\n     void setAttribute(final String attributeName, final String attributeValue)\\n     void setAttribute(final String attributeName, final byte[] attributeValue)\\n    public void setTransientAttribute(final Object key, final Object value)\\n     void clearAttribute(final String attributeName)\\n     void clearAttributes()\\n     void clearTransientAttribute(final String attributeName)\\n     GATKRead copy()\\n     GATKRead deepCopy()\\n     SAMRecord convertToSAMRecord(final SAMFileHeader header)\\n     String getSAMString()\\n    public void reverseComplement()\\n     String commonToString()\\n     void hardClipAttributes(final int newStart, final int newLength, int originalLength)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java', 'methodCount': 94, 'lineCount': 96}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 0}\n",
      "{'GATKException': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n'}\n",
      "/**\n",
      "     * @param val mapping quality value (ranged 0-250)\n",
      "     * @return a score (as is performed by DRAGEN) corresponding to the phred likelihood of mipmapping event given that mapping quality.\n",
      "     */\n",
      "public static double mapMappingQualityToPhredLikelihoodScore(final int val)\n",
      "**************************************\n",
      "/**\n",
      " * @param val mapping quality value (ranged 0-250)\n",
      " * @return a score (as is performed by DRAGEN) corresponding to the phred likelihood of mipmapping event given that mapping quality.\n",
      " */\n",
      "@VisibleForTesting\n",
      "public static double mapMappingQualityToPhredLikelihoodScore(final int val) {\n",
      "    for (int i = 1; i < mQTableX.length; i++) {\n",
      "        if (val <= mQTableX[i]) {\n",
      "            final double xfactor = 1.0 * (val - mQTableX[i - 1]) / (mQTableX[i] - mQTableX[i - 1]);\n",
      "            return mQTableY[i - 1] + (xfactor * (mQTableY[i] - mQTableY[i - 1]));\n",
      "        }\n",
      "    }\n",
      "    //This is technically a failure state because the MQ is invalid if somehow it exceeds 256 by the SAMSpec....\n",
      "    throw new GATKException(\"Something went wrong trying to map an an invalid XQ tag '\" + val + \"' val must fall between 0 and 250\");\n",
      "}\n",
      "1528\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1069/1243 [36:50<06:22,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47177\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1070/1243 [36:53<06:22,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47178\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1071/1243 [36:55<06:17,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47188\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1072/1243 [36:57<06:24,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47219\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1073/1243 [36:59<06:04,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47239\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ParamUtils {\\n    public static final double INV_LOG_2 = 1 / Math.log(2.0);;\\n    private ParamUtils();\\n    public static double inRange(final double val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final long min, final long max, final String message)\\n    public static int inRange(final int val, final int min, final int max, final String message)\\n    public static double isPositiveOrZero(final double val, final String message)\\n    public static double isNegativeOrZero(final double val, final String message)\\n    public static long isPositiveOrZero(final long val, final String message)\\n    public static int isPositiveOrZero(final int val, final String message)\\n    public static long isPositive(final long val, final String message)\\n    public static int isPositive(final int val, final String message)\\n    public static double isPositive(final double val, final String message)\\n    public static double isFinite(final double val, final String message)\\n    public static double logb(final double val, final double base)\\n    public static double log2(final double val)\\n    public static void writeStringListToFile(final Iterable<String> stringIterator, final File outputFile)\\n    public static void writeValuesToFile(final double[] data, final File outputFile)\\n    public static double[] readValuesFromFile(final File inputFile)\\n    public static double inRange(final DoubleRange validRange, final double value, final String definition)\\n    public static int inRange(final IntRange validRange, final int value, final String definition)\\n    public static void isValidArrayOffsetAndRangeLength(final int from, final int length, final int arrayLength, final String definition)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/param/ParamUtils.java', 'methodCount': 21, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Given an int array returns the difference between the second smallest element\n",
      "     * and the smallest element.\n",
      "     * @param values the input values.\n",
      "     * @param defaultValue value to return in case the input array has zero or only one element.\n",
      "     * @return 0 if the input array has less than 2 elements, otherwise the second smallest\n",
      "     * - the smallest.\n",
      "     * @throws IllegalArgumentException if {@code values} is {@code null}.\n",
      "     */\n",
      "public static int secondSmallestMinusSmallest(final int[] values, final int defaultValue)\n",
      "**************************************\n",
      "public static int secondSmallestMinusSmallest(final int[] values, final int defaultValue) {\n",
      "    Utils.nonNull(values);\n",
      "    if (values.length <= 1) {\n",
      "        return defaultValue;\n",
      "    } else {\n",
      "        int smallest = values[0];\n",
      "        int secondSmallest = Integer.MAX_VALUE;\n",
      "        for (int i = 1; i < values.length; i++) {\n",
      "            if (values[i] < smallest) {\n",
      "                secondSmallest = smallest;\n",
      "                smallest = values[i];\n",
      "            } else if (values[i] < secondSmallest) {\n",
      "                secondSmallest = values[i];\n",
      "            }\n",
      "        }\n",
      "        return secondSmallest - smallest;\n",
      "    }\n",
      "}\n",
      "759\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1074/1243 [37:01<05:39,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47269\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  86%| | 1075/1243 [37:03<05:51,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47281\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1076/1243 [37:05<05:20,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47320\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1077/1243 [37:06<05:09,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47321\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1078/1243 [37:08<05:04,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47322\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ParamUtils {\\n    public static final double INV_LOG_2 = 1 / Math.log(2.0);;\\n    private ParamUtils();\\n    public static double inRange(final double val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final long min, final long max, final String message)\\n    public static int inRange(final int val, final int min, final int max, final String message)\\n    public static double isPositiveOrZero(final double val, final String message)\\n    public static double isNegativeOrZero(final double val, final String message)\\n    public static long isPositiveOrZero(final long val, final String message)\\n    public static int isPositiveOrZero(final int val, final String message)\\n    public static long isPositive(final long val, final String message)\\n    public static int isPositive(final int val, final String message)\\n    public static double isPositive(final double val, final String message)\\n    public static double isFinite(final double val, final String message)\\n    public static double logb(final double val, final double base)\\n    public static double log2(final double val)\\n    public static void writeStringListToFile(final Iterable<String> stringIterator, final File outputFile)\\n    public static void writeValuesToFile(final double[] data, final File outputFile)\\n    public static double[] readValuesFromFile(final File inputFile)\\n    public static double inRange(final DoubleRange validRange, final double value, final String definition)\\n    public static int inRange(final IntRange validRange, final int value, final String definition)\\n    public static void isValidArrayOffsetAndRangeLength(final int from, final int length, final int arrayLength, final String definition)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/param/ParamUtils.java', 'methodCount': 21, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     *\n",
      "     * @param array array of integers\n",
      "     * @return index of the max. In case of a tie, return the smallest index\n",
      "     */\n",
      "public static int maxElementIndex(final int[] array)\n",
      "**************************************\n",
      "public static int maxElementIndex(final int[] array) {\n",
      "    int maxIndex = 0;\n",
      "    int currentMax = Integer.MIN_VALUE;\n",
      "    for (int i = 0; i < array.length; i++) {\n",
      "        if (array[i] > currentMax) {\n",
      "            maxIndex = i;\n",
      "            currentMax = array[i];\n",
      "        }\n",
      "    }\n",
      "    return maxIndex;\n",
      "}\n",
      "631\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1079/1243 [37:10<05:21,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47408\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1080/1243 [37:12<05:09,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47415\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1081/1243 [37:14<05:02,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47434\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1082/1243 [37:16<04:53,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47437\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1083/1243 [37:18<05:12,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47460\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class GATKPath {\\n    private static final long serialVersionUID = 1L;;\\n    public static final String HDFS_SCHEME = \"hdfs\";;\\n    private String tagName;;\\n    private Map<String, String> tagAttributes;;\\n    public GATKPath(final String uriString);\\n    public GATKPath(final GATKPath sourceGATKPath);\\n    public Path toPath()\\n    public InputStream getInputStream()\\n    public OutputStream getOutputStream()\\n    public boolean isHadoopURL()\\n    public void setTag(String tagName)\\n    public String getTag()\\n    public void setTagAttributes(final Map<String, String> attributes)\\n    public Map<String, String> getTagAttributes()\\n    public String toString()\\n     static boolean tagNameAndAttributesAreEqual(final TaggedArgument first, final TaggedArgument second)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/GATKPath.java', 'methodCount': 14, 'lineCount': 20}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface ReferenceDataSource {\\n    public static ReferenceDataSource of(final Path fastaPath)\\n    public static ReferenceDataSource of(final Path fastaPath, final boolean preserveAmbiguityCodesAndCapitalization)\\n    public static ReferenceDataSource of(final ReferenceBases bases, final SAMSequenceDictionary referenceSequenceDictionary)\\n    public ReferenceSequence queryAndPrefetch(final SimpleInterval interval)\\n    public ReferenceSequence queryAndPrefetch(final String contig, final long start, final long stop)\\n    public Iterator<Byte> query(final SimpleInterval interval)\\n    public SAMSequenceDictionary getSequenceDictionary()\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/ReferenceDataSource.java', 'methodCount': 8, 'lineCount': 10}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class SimpleInterval {\\n    private static final long serialVersionUID = 1L;;\\n    public static final char CONTIG_SEPARATOR = \\':\\';;\\n    public static final char START_END_SEPARATOR = \\'-\\';;\\n    //note: needs to be a String because it\\'s used in an endsWith call.\\npublic static final String END_OF_CONTIG = \"+\";;\\n    private final int start;;\\n    private final int end;;\\n    private final String contig;;\\n    public SimpleInterval(final String contig, final int start, final int end);\\n    public SimpleInterval(final Locatable locatable);\\n     static void validatePositions(final String contig, final int start, final int end)\\n    public static boolean isValid(final String contig, final int start, final int end)\\n    public SimpleInterval(final String str);\\n     static int parsePositionThrowOnFailure(final String pos) throws NumberFormatException\\n     static int parsePosition(final String pos)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String getContig()\\n    public int getStart()\\n    public long getGA4GHStart()\\n    public int getEnd()\\n    public long getGA4GHEnd()\\n    public int size()\\n    public boolean overlaps(final Locatable other)\\n    public boolean overlapsWithMargin(final Locatable other, final int margin)\\n    public boolean contains(final Locatable other)\\n    public SimpleInterval intersect(final Locatable that)\\n    public SimpleInterval mergeWithContiguous(final Locatable that)\\n    public SimpleInterval spanWith(final Locatable other)\\n    private boolean contiguous(final Locatable that)\\n    public SimpleInterval expandWithinContig(final int padding, final int contigLength)\\n    public SimpleInterval expandWithinContig(final int padding, final SAMSequenceDictionary sequenceDictionary)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java', 'methodCount': 25, 'lineCount': 34}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': 'public class CachingIndexedFastaSequenceFile {\\n    private static final long serialVersionUID = 1L;;\\n    protected static final Logger logger = LogManager.getLogger(CachingIndexedFastaSequenceFile.class);;\\n    // TODO: this gets around numerous serialization issues with non-serializable lambdas\\nprivate transient final ReferenceSequenceFile sequenceFile;;\\n    /**\\n * do we want to print debugging information about cache efficiency?\\n */\\nprivate static final boolean PRINT_EFFICIENCY = false;;\\n    /**\\n * If we are printing efficiency info, what frequency should we do it at?\\n */\\nprivate static final int PRINT_FREQUENCY = 10000;;\\n    /**\\n * The default cache size in bp\\n */\\npublic static final long DEFAULT_CACHE_SIZE = 1000000;;\\n    /**\\n * The cache size of this CachingIndexedFastaSequenceFile\\n */\\nprivate final long cacheSize;;\\n    /**\\n * When we have a cache miss at position X, we load sequence from X - cacheMissBackup\\n */\\nprivate final long cacheMissBackup;;\\n    /**\\n * If true, we will preserve the case of the original base in the genome\\n */\\nprivate final boolean preserveCase;;\\n    /**\\n * If true, we will preserve the IUPAC bases in the genome\\n */\\nprivate final boolean preserveIUPAC;;\\n    // information about checking efficiency\\nlong cacheHits = 0;;\\n    long cacheMisses = 0;;\\n    private final Cache cache = new Cache();;\\n    private static ReferenceSequenceFile requireIndex(Path fasta, ReferenceSequenceFile referenceSequenceFile)\\n    public CachingIndexedFastaSequenceFile(final GATKPath fasta);\\n    public CachingIndexedFastaSequenceFile(final Path fasta);\\n    public CachingIndexedFastaSequenceFile(final Path fasta, boolean preserveAmbiguityCodesAndCapitalization);\\n    public CachingIndexedFastaSequenceFile(final Path fasta, final long cacheSize, final boolean preserveCase, final boolean preserveIUPAC);\\n    private static void checkFastaPath(final Path fastaPath)\\n    public void printEfficiency(final Level priority)\\n    public double calcEfficiency()\\n    public long getCacheHits()\\n    public long getCacheMisses()\\n    public long getCacheSize()\\n    public boolean isPreservingCase()\\n    public SAMSequenceDictionary getSequenceDictionary()\\n    public ReferenceSequence nextSequence()\\n    public void reset()\\n    public boolean isIndexed()\\n    public ReferenceSequence getSequence(String contig)\\n    public ReferenceSequence getSubsequenceAt(final String contig, long start, final long stop)\\n    public void close()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/fasta/CachingIndexedFastaSequenceFile.java', 'methodCount': 19, 'lineCount': 34}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 7}\n",
      "{}\n",
      "/**\n",
      "     * Given a fasta dictionary file, returns its sequence dictionary\n",
      "     *\n",
      "     * @param fastaDictionaryFile fasta dictionary file\n",
      "     * @return the SAMSequenceDictionary from fastaDictionaryFile\n",
      "     */\n",
      "public static SAMSequenceDictionary loadFastaDictionary(final GATKPath fastaDictionaryFile)\n",
      "**************************************\n",
      "public static SAMSequenceDictionary loadFastaDictionary(final GATKPath fastaDictionaryFile) {\n",
      "    try (final InputStream fastaDictionaryStream = fastaDictionaryFile.getInputStream()) {\n",
      "        return loadFastaDictionary(fastaDictionaryStream);\n",
      "    } catch (IOException e) {\n",
      "        throw new UserException.CouldNotReadInputFile(\"Error loading fasta dictionary file \" + fastaDictionaryFile, e);\n",
      "    } catch (UserException.MalformedFile e) {\n",
      "        throw new UserException.MalformedFile(\"Could not read sequence dictionary from given fasta file \" + fastaDictionaryFile);\n",
      "    }\n",
      "}\n",
      "2072\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1084/1243 [37:20<05:23,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47462\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1085/1243 [37:22<05:02,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47538\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1086/1243 [37:24<05:18,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47548\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  87%| | 1087/1243 [37:26<04:55,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47571\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1088/1243 [37:28<05:06,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47733\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class BucketUtils {\\n    public static final String GCS_PREFIX = GoogleCloudStorageFileSystem.SCHEME + \"://\";;\\n    public static final String HTTP_PREFIX = HttpFileSystemProvider.SCHEME + \"://\";;\\n    public static final String HTTPS_PREFIX = HttpsFileSystemProvider.SCHEME + \"://\";;\\n    public static final String HDFS_SCHEME = \"hdfs\";;\\n    public static final String HDFS_PREFIX = HDFS_SCHEME + \"://\";;\\n    // slashes omitted since hdfs paths seem to only have 1 slash which would be weirder to include than no slashes\\npublic static final String FILE_PREFIX = \"file:\";;\\n    private BucketUtils();\\n    public static boolean isGcsUrl(final String path)\\n    public static boolean isGcsUrl(final GATKPath pathSpec)\\n    public static boolean isEligibleForPrefetching(final GATKPath pathSpec)\\n    public static boolean isEligibleForPrefetching(final java.nio.file.Path path)\\n    private static boolean isEligibleForPrefetching(final String scheme)\\n    public static boolean isHttpUrl(String path)\\n    public static boolean isHadoopUrl(String path)\\n    public static boolean isRemoteStorageUrl(String path)\\n    public static String makeFilePathAbsolute(String path)\\n    public static InputStream openFile(String path)\\n    public static OutputStream createFile(String path)\\n    public static void copyFile(String sourcePath, String destPath) throws IOException\\n    public static void deleteFile(String pathToDelete) throws IOException\\n    public static String getTempFilePath(String prefix, String extension)\\n    public static String randomRemotePath(String stagingLocation, String prefix, String suffix)\\n    public static boolean fileExists(String path)\\n    public static long fileSize(String path) throws IOException\\n    public static long dirSize(final GATKPath pathSpecifier)\\n    public static boolean isFileUrl(String path)\\n    public static String getBucket(String path)\\n    public static String getPathWithoutBucket(String path)\\n    public static void setGlobalNIODefaultOptions(int maxReopens, String requesterProject)\\n    public static java.nio.file.Path getPathOnGcs(String gcsUrl)\\n    public static CloudStorageConfiguration getCloudStorageConfiguration(int maxReopens, String requesterProject)\\n    private static StorageOptions.Builder setGenerousTimeouts(StorageOptions.Builder builder)\\n    public static java.nio.file.FileSystem getAuthenticatedGcs(String projectId, String bucket, byte[] credentials) throws IOException\\n    public static SeekableByteChannel addPrefetcher(final int bufferSizeMB, final SeekableByteChannel channel)\\n    public static Function<SeekableByteChannel, SeekableByteChannel> getPrefetchingWrapper(final int cloudPrefetchBuffer)\\n    public static String createSignedUrlToGcsObject(String path, final long hoursToLive)\\n    public static String bucketPathToPublicHttpUrl(String path)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java', 'methodCount': 31, 'lineCount': 39}, {'fieldCount': 30, 'classCount': 1, 'methodsOnly': 'public class RecalUtils {\\n    public static final String ARGUMENT_REPORT_TABLE_TITLE = \"Arguments\";;\\n    public static final String QUANTIZED_REPORT_TABLE_TITLE = \"Quantized\";;\\n    public static final String READGROUP_REPORT_TABLE_TITLE = \"RecalTable0\";;\\n    public static final String QUALITY_SCORE_REPORT_TABLE_TITLE = \"RecalTable1\";;\\n    public static final String ALL_COVARIATES_REPORT_TABLE_TITLE = \"RecalTable2\";;\\n    public static final String ARGUMENT_COLUMN_NAME = \"Argument\";;\\n    public static final String ARGUMENT_VALUE_COLUMN_NAME = \"Value\";;\\n    public static final String QUANTIZED_VALUE_COLUMN_NAME = \"QuantizedScore\";;\\n    public static final String QUANTIZED_COUNT_COLUMN_NAME = \"Count\";;\\n    public static final String READGROUP_COLUMN_NAME = \"ReadGroup\";;\\n    public static final String EVENT_TYPE_COLUMN_NAME = \"EventType\";;\\n    public static final String EMPIRICAL_QUALITY_COLUMN_NAME = \"EmpiricalQuality\";;\\n    public static final String ESTIMATED_Q_REPORTED_COLUMN_NAME = \"EstimatedQReported\";;\\n    public static final String QUALITY_SCORE_COLUMN_NAME = \"QualityScore\";;\\n    public static final String COVARIATE_VALUE_COLUMN_NAME = \"CovariateValue\";;\\n    public static final String COVARIATE_NAME_COLUMN_NAME = \"CovariateName\";;\\n    public static final String NUMBER_OBSERVATIONS_COLUMN_NAME = \"Observations\";;\\n    public static final String NUMBER_ERRORS_COLUMN_NAME = \"Errors\";;\\n    private static boolean warnUserNullPlatform = false;;\\n    private static final String SCRIPT_FILE = \"BQSR.R\";;\\n    public static final int EMPIRICAL_QUAL_DECIMAL_PLACES = 4;;\\n    public static final int EMPIRICAL_Q_REPORTED_DECIMAL_PLACES = 4;;\\n    public static final int NUMBER_ERRORS_DECIMAL_PLACES = 2;;\\n    private static final Pair<String, String> covariateValue = new MutablePair<>(RecalUtils.COVARIATE_VALUE_COLUMN_NAME, \"%s\");;\\n    private static final Pair<String, String> covariateName = new MutablePair<>(RecalUtils.COVARIATE_NAME_COLUMN_NAME, \"%s\");;\\n    private static final Pair<String, String> eventType = new MutablePair<>(RecalUtils.EVENT_TYPE_COLUMN_NAME, \"%s\");;\\n    private static final Pair<String, String> empiricalQuality = new MutablePair<>(RecalUtils.EMPIRICAL_QUALITY_COLUMN_NAME, \"%.\" + EMPIRICAL_QUAL_DECIMAL_PLACES + \\'f\\');;\\n    private static final Pair<String, String> estimatedQReported = new MutablePair<>(RecalUtils.ESTIMATED_Q_REPORTED_COLUMN_NAME, \"%.\" + EMPIRICAL_Q_REPORTED_DECIMAL_PLACES + \\'f\\');;\\n    private static final Pair<String, String> nObservations = new MutablePair<>(RecalUtils.NUMBER_OBSERVATIONS_COLUMN_NAME, \"%d\");;\\n    private static final Pair<String, String> nErrors = new MutablePair<>(RecalUtils.NUMBER_ERRORS_COLUMN_NAME, \"%.\" + NUMBER_ERRORS_DECIMAL_PLACES + \\'f\\');;\\n    protected static CsvPrinter csvPrinter(final File out, final StandardCovariateList covs) throws FileNotFoundException\\n    public static void generateCsv(final File out, final Map<String, RecalibrationReport> reports) throws FileNotFoundException\\n    private static void writeCsv(final File out, final Map<String, RecalibrationReport> reports, final StandardCovariateList covs) throws FileNotFoundException\\n    public static List<GATKReportTable> generateReportTables(final RecalibrationTables recalibrationTables, final StandardCovariateList covariates)\\n    private static String getReportTableName(RecalibrationTables recalibrationTables, NestedIntegerArray<RecalDatum> table)\\n    private static GATKReportTable makeNewTableWithColumns(ArrayList<Pair<String, String>> columnNames, String reportTableName, GATKReportTable.Sorting sort)\\n    public static void outputRecalibrationReport(final PrintStream recalTableStream, final RecalibrationArgumentCollection RAC, final QuantizationInfo quantizationInfo, final RecalibrationTables recalibrationTables, final StandardCovariateList covariates)\\n    public static RecalibrationReport createRecalibrationReport(final GATKReportTable argumentTable, final GATKReportTable quantizationTable, final List<GATKReportTable> recalTables)\\n    public static GATKReport createRecalibrationGATKReport(final GATKReportTable argumentTable, final QuantizationInfo quantizationInfo, final RecalibrationTables recalibrationTables, final StandardCovariateList covariates)\\n    public static GATKReport createRecalibrationGATKReport(final GATKReportTable argumentTable, final GATKReportTable quantizationTable, final List<GATKReportTable> recalTables)\\n    public static void generatePlots(final File csvFile, final File maybeGzipedExampleReportFile, final File output)\\n    private static void writeCsv(final PrintStream deltaTableFile, final RecalibrationTables recalibrationTables, final String recalibrationMode, final StandardCovariateList covariates, final boolean printHeader)\\n    private static void printHeader(PrintStream out)\\n    private static NestedIntegerArray<RecalDatum> createDeltaTable(final RecalibrationTables recalibrationTables, final int numCovariates)\\n     static List<Object> generateValuesFromKeys(final int[] keys, final StandardCovariateList covariates)\\n    private static void addToDeltaTable(final NestedIntegerArray<RecalDatum> deltaTable, final int[] deltaKey, final RecalDatum recalDatum)\\n    public static void parsePlatformForRead(final GATKRead read, final SAMFileHeader header, final RecalibrationArgumentCollection RAC)\\n    public static ReadCovariates computeCovariates(final GATKRead read, final SAMFileHeader header, final StandardCovariateList covariates, final boolean recordIndelValues, final CovariateKeyCache keyCache)\\n    public static void computeCovariates(final GATKRead read, final SAMFileHeader header, final StandardCovariateList covariates, final ReadCovariates resultsStorage, final boolean recordIndelValues)\\n    public static void combineTables(final NestedIntegerArray<RecalDatum> table1, final NestedIntegerArray<RecalDatum> table2)\\n    public static void incrementDatumOrPutIfNecessary2keys(final NestedIntegerArray<RecalDatum> table, final byte qual, final double isError, final int key0, final int key1)\\n    public static void incrementDatumOrPutIfNecessary3keys(final NestedIntegerArray<RecalDatum> table, final byte qual, final double isError, final int key0, final int key1, final int key2)\\n    public static void incrementDatumOrPutIfNecessary4keys(final NestedIntegerArray<RecalDatum> table, final byte qual, final double isError, final int key0, final int key1, final int key2, final int key3)\\n    private static RecalDatum createDatumObject(final byte reportedQual, final double isError)\\n    protected static Resource loadBQSRScriptResource()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/recalibration/RecalUtils.java', 'methodCount': 25, 'lineCount': 57}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 14}\n",
      "{'GATKReport': 'public class GATKReport {\\n    public static final String RECAL_FILE = \"input covariates table file for base quality score recalibration\";;\\n    public static final String GATKREPORT_HEADER_PREFIX = \"#:GATKReport.\";;\\n    public static final GATKReportVersion LATEST_REPORT_VERSION = GATKReportVersion.V1_1;;\\n    private static final String SEPARATOR = \":\";;\\n    private GATKReportVersion version = LATEST_REPORT_VERSION;;\\n    private final NavigableMap<String, GATKReportTable> tables = new TreeMap<>();;\\n    public GATKReport();\\n    public GATKReport(String filename);\\n    public GATKReport(File file);\\n    public GATKReport(InputStream in);\\n    public GATKReport(GATKReportTable... tables);\\n    public SortedSet<String> getReadGroups()\\n    private void loadReport(Reader in)\\n    public void addTable(final String tableName, final String tableDescription, final int numColumns)\\n    public void addTable(final String tableName, final String tableDescription, final int numColumns, final GATKReportTable.Sorting sortingWay)\\n    public void addTable(GATKReportTable table)\\n    public void addTables(List<GATKReportTable> gatkReportTableV2s)\\n    public GATKReportTable getTable(String tableName)\\n    public void print(PrintStream out)\\n    public void print(PrintStream out, GATKReportTable.Sorting sortingWay)\\n    public Collection<GATKReportTable> getTables()\\n    public void concat(GATKReport input)\\n    public GATKReportVersion getVersion()\\n    public boolean isSameFormat(GATKReport report)\\n    public boolean equals(GATKReport report)\\n    public static GATKReport newSimpleReport(final String tableName, GATKReportTable.Sorting sorting, final String... columns)\\n    public static GATKReport newSimpleReportWithDescription(final String tableName, final String desc, GATKReportTable.Sorting sorting, final String... columns)\\n    public void addRow(final Object... values)\\n}\\n'}\n",
      "/**\n",
      "     * Checks that the reports are exactly the same.\n",
      "     *\n",
      "     * @param report another GATK report\n",
      "     * @return true if all field in the reports, tables, and columns are equal.\n",
      "     */\n",
      "public boolean equals(GATKReport report)\n",
      "**************************************\n",
      "public boolean equals(GATKReport report) {\n",
      "    if (!version.equals(report.version)) {\n",
      "        return false;\n",
      "    }\n",
      "    if (!tables.keySet().equals(report.tables.keySet())) {\n",
      "        return false;\n",
      "    }\n",
      "    for (String tableName : tables.keySet()) {\n",
      "        if (!getTable(tableName).equals(report.getTable(tableName)))\n",
      "            return false;\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "3238\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1089/1243 [37:30<04:53,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47748\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1090/1243 [37:31<04:42,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47751\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1091/1243 [37:33<04:54,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47765\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1092/1243 [37:36<05:12,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47766\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1093/1243 [37:37<04:53,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47883\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 10, 'classCount': 1, 'methodsOnly': 'public class Haplotype {\\n    private static final long serialVersionUID = 1L;;\\n    /**\\n * Compares two haplotypes first by their lengths and then by lexicographic order of their bases.\\n */\\npublic static final Comparator<Haplotype> SIZE_AND_BASE_ORDER = Comparator.comparingInt((Haplotype hap) -> hap.getBases().length).thenComparing(Allele::getBaseString);;\\n    private Locatable genomeLocation = null;;\\n    private EventMap eventMap = null;;\\n    private Cigar cigar;;\\n    //NOTE: this is the offset to a supposed array of reference bases held in memory and has nothing to do with start positions\\nprivate int alignmentStartHapwrtRef;;\\n    private double score = Double.NaN;;\\n    /**\\n * see {@link org.broadinstitute.hellbender.tools.walkers.haplotypecaller.LongHomopolymerHaplotypeCollapsingEngine} for a description of the semantics of collapsing\\n */\\nprivate boolean isCollapsed;;\\n    // uniquely diffrentiates the haplotype from others with same ref/bases.\\nprivate int uniquenessValue;;\\n    // debug information for tracking kmer sizes used in graph construction for debug output\\nprivate int kmerSize = 0;;\\n    public Haplotype(final byte[] bases, final boolean isRef);\\n    public Haplotype(final byte[] bases);\\n    public Haplotype(final byte[] bases, final boolean isRef, final int alignmentStartHapwrtRef, final Cigar cigar);\\n    public Haplotype(final byte[] bases, final boolean isRef, final Locatable loc, final Cigar cigar);\\n    public Haplotype(final byte[] bases, final Locatable loc);\\n    public Haplotype trim(final Locatable loc)\\n    public Haplotype trim(final Locatable loc, boolean ignoreRefState)\\n    public boolean equals(final Object h)\\n    public int hashCode()\\n    public EventMap getEventMap()\\n    public void setEventMap(final EventMap eventMap)\\n    public String toString()\\n    public Locatable getLocation()\\n    public void setGenomeLocation(final Locatable genomeLocation)\\n    public long getStartPosition()\\n    public long getStopPosition()\\n    public int getAlignmentStartHapwrtRef()\\n    public void setAlignmentStartHapwrtRef(final int alignmentStartHapwrtRef)\\n    public Cigar getCigar()\\n    public Cigar getConsolidatedPaddedCigar(final int padSize)\\n    public void setCigar(final Cigar cigar)\\n    public Haplotype insertAllele(final Allele refAllele, final Allele altAllele, final int insertLocation)\\n    public double getScore()\\n    public void setScore(final double score)\\n    public Locatable getGenomeLocation()\\n    public boolean isCollapsed()\\n    public void setCollapsed(boolean collapsed)\\n    public int getUniquenessValue()\\n    public int getKmerSize()\\n    public void setUniquenessValue(int uniquenessValue)\\n    public void setKmerSize(int kmerSize)\\n    public boolean isPartiallyDetermined()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/haplotype/Haplotype.java', 'methodCount': 32, 'lineCount': 44}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Converts a base including extended bases (ATCG + D and N bases) to a base index\n",
      "     *\n",
      "     * @param base [AaCcGgTtDdNn]\n",
      "     * @return 0, 1, 2, 3, 4, or -1 if the base can't be understood\n",
      "     */\n",
      "public static int extendedBaseToBaseIndex(byte base)\n",
      "**************************************\n",
      "static public int extendedBaseToBaseIndex(byte base) {\n",
      "    switch(base) {\n",
      "        case 'd':\n",
      "        case 'D':\n",
      "            return Base.D.ordinal();\n",
      "        case 'n':\n",
      "        case 'N':\n",
      "            return Base.N.ordinal();\n",
      "        default:\n",
      "            return simpleBaseToBaseIndex(base);\n",
      "    }\n",
      "}\n",
      "1099\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1094/1243 [37:39<04:36,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47887\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1095/1243 [37:41<04:55,  2.00s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47978\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1096/1243 [37:44<04:59,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "47997\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1097/1243 [37:46<05:00,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48020\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1098/1243 [37:48<04:58,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48033\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class GATKPath {\\n    private static final long serialVersionUID = 1L;;\\n    public static final String HDFS_SCHEME = \"hdfs\";;\\n    private String tagName;;\\n    private Map<String, String> tagAttributes;;\\n    public GATKPath(final String uriString);\\n    public GATKPath(final GATKPath sourceGATKPath);\\n    public Path toPath()\\n    public InputStream getInputStream()\\n    public OutputStream getOutputStream()\\n    public boolean isHadoopURL()\\n    public void setTag(String tagName)\\n    public String getTag()\\n    public void setTagAttributes(final Map<String, String> attributes)\\n    public Map<String, String> getTagAttributes()\\n    public String toString()\\n     static boolean tagNameAndAttributesAreEqual(final TaggedArgument first, final TaggedArgument second)\\n    public boolean equals(Object o)\\n    public int hashCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/GATKPath.java', 'methodCount': 14, 'lineCount': 20}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'Sample': 'public class Sample {\\n    private final String familyID;;\\n    private final String paternalID;;\\n    private final String maternalID;;\\n    private final Sex gender;;\\n    private final String ID;;\\n    private final Affection affection;;\\n    public Sample(final String ID, final String familyID, final String paternalID, final String maternalID, final Sex gender, final Affection affection);\\n    public Sample(final String ID, final String familyID, final String paternalID, final String maternalID, final Sex gender);\\n    public String getID()\\n    public String getFamilyID()\\n    public String getPaternalID()\\n    public String getMaternalID()\\n    public Sex getSex()\\n    public Affection getAffection()\\n    public String toString()\\n    public int compareTo(final Sample other)\\n    public int hashCode()\\n    public boolean equals(final Object o)\\n     final Sample mergeSamples(final Sample newSample)\\n    private static T mergeValues(final String name, final String field, final T o1, final T o2, final T emptyValue)\\n}\\n', 'SampleDBBuilder': 'public class SampleDBBuilder {\\n    private final PedigreeValidationType validationStrictness;;\\n    private final SampleDB sampleDB = new SampleDB();;\\n    private final Set<Sample> samplesFromDataSources = new LinkedHashSet<>();;\\n    private final Set<Sample> samplesFromPedigrees = new LinkedHashSet<>();;\\n    public SampleDBBuilder(PedigreeValidationType validationStrictness);\\n    public SampleDBBuilder addSamplesFromPedigreeFiles(final List<GATKPath> pedigreeFiles)\\n    public SampleDBBuilder addSamplesFromPedigreeStrings(final List<String> pedigreeStrings)\\n    private Collection<Sample> addSamplesFromPedigreeArgument(GATKPath sampleFile)\\n    public SampleDBBuilder addSamplesFromSampleNames(final Collection<String> sampleNames)\\n    private Collection<Sample> addSamplesFromPedigreeArgument(final String string)\\n    public SampleDB getFinalSampleDB()\\n    private EnumSet<PedReader.MissingPedField> getMissingFields(final Object engineArg)\\n    private void validate()\\n    private void validatePedigreeIDUniqueness()\\n}\\n'}\n",
      "/**\n",
      "     * Integrates the collection of sample names with the samples already present\n",
      "     */\n",
      "public SampleDBBuilder addSamplesFromSampleNames(final Collection<String> sampleNames)\n",
      "**************************************\n",
      "public SampleDBBuilder addSamplesFromSampleNames(final Collection<String> sampleNames) {\n",
      "    Utils.nonNull(sampleNames);\n",
      "    for (final String sampleName : sampleNames) {\n",
      "        if (sampleDB.getSample(sampleName) == null) {\n",
      "            final Sample newSample = new Sample(sampleName, null, null, null, Sex.UNKNOWN);\n",
      "            sampleDB.addSample(newSample);\n",
      "            samplesFromDataSources.add(newSample);\n",
      "        }\n",
      "    }\n",
      "    return this;\n",
      "}\n",
      "2794\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1099/1243 [37:50<04:59,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48088\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  88%| | 1100/1243 [37:52<04:57,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48089\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1101/1243 [37:54<04:34,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48097\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1102/1243 [37:55<04:14,  1.81s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48102\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1103/1243 [37:57<04:07,  1.77s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48214\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{'SampleList': 'public interface SampleList {\\n    static final SampleList EMPTY_LIST = new SampleList() {\\n\\n    @Override\\n    public int numberOfSamples() {\\n        return 0;\\n    }\\n\\n    @Override\\n    public int indexOfSample(final String sample) {\\n        Utils.nonNull(sample);\\n        return -1;\\n    }\\n\\n    @Override\\n    public String getSample(final int sampleIndex) {\\n        //we know this without checking because it\\'s an empty list\\n        throw new IllegalArgumentException(\"index is out of valid range\");\\n    }\\n};;\\n    public static SampleList emptySampleList()\\n    public int numberOfSamples()\\n    public int indexOfSample(final String sample)\\n    public String getSample(final int sampleIndex)\\n    public static boolean equals(final SampleList first, final SampleList second)\\n    public List<String> asListOfSamples()\\n    public Set<String> asSetOfSamples()\\n    public static SampleList singletonSampleList(final String sampleName)\\n}\\n'}\n",
      "/**\n",
      "     * Checks whether two sample lists are in fact the same.\n",
      "     * @param first one list to compare.\n",
      "     * @param second another list to compare.\n",
      "     *\n",
      "     * @throws IllegalArgumentException if if either list is {@code null}.\n",
      "     *\n",
      "     * @return {@code true} iff both list are equal.\n",
      "     */\n",
      "public static boolean equals(final SampleList first, final SampleList second)\n",
      "**************************************\n",
      "public static boolean equals(final SampleList first, final SampleList second) {\n",
      "    Utils.nonNull(first, \"first list is null\");\n",
      "    Utils.nonNull(second, \"second list is null\");\n",
      "    final int sampleCount = first.numberOfSamples();\n",
      "    if (sampleCount != second.numberOfSamples()) {\n",
      "        return false;\n",
      "    }\n",
      "    for (int i = 0; i < sampleCount; i++) {\n",
      "        final String firstSample = first.getSample(i);\n",
      "        final String secondSample = second.getSample(i);\n",
      "        if (!firstSample.equals(secondSample)) {\n",
      "            return false;\n",
      "        }\n",
      "    }\n",
      "    return true;\n",
      "}\n",
      "2183\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1104/1243 [37:58<03:56,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48215\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1105/1243 [38:00<04:10,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48236\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1106/1243 [38:02<04:19,  1.89s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48351\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1107/1243 [38:05<04:29,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48352\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1108/1243 [38:06<04:23,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48355\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Find the least interval in the tree.\n",
      "     *\n",
      "     * @return The earliest interval, or null if the tree is empty.\n",
      "     */\n",
      "public Entry<V> min()\n",
      "**************************************\n",
      "public Entry<V> min() {\n",
      "    Node<V> result = null;\n",
      "    Node<V> node = root;\n",
      "    while (node != null) {\n",
      "        result = node;\n",
      "        node = node.getLeft();\n",
      "    }\n",
      "    return result;\n",
      "}\n",
      "235\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1109/1243 [38:08<03:48,  1.70s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48359\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1110/1243 [38:10<04:02,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48374\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1111/1243 [38:11<03:51,  1.76s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48393\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  89%| | 1112/1243 [38:13<03:47,  1.74s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48395\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1113/1243 [38:15<03:41,  1.71s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48396\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ParamUtils {\\n    public static final double INV_LOG_2 = 1 / Math.log(2.0);;\\n    private ParamUtils();\\n    public static double inRange(final double val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final long min, final long max, final String message)\\n    public static int inRange(final int val, final int min, final int max, final String message)\\n    public static double isPositiveOrZero(final double val, final String message)\\n    public static double isNegativeOrZero(final double val, final String message)\\n    public static long isPositiveOrZero(final long val, final String message)\\n    public static int isPositiveOrZero(final int val, final String message)\\n    public static long isPositive(final long val, final String message)\\n    public static int isPositive(final int val, final String message)\\n    public static double isPositive(final double val, final String message)\\n    public static double isFinite(final double val, final String message)\\n    public static double logb(final double val, final double base)\\n    public static double log2(final double val)\\n    public static void writeStringListToFile(final Iterable<String> stringIterator, final File outputFile)\\n    public static void writeValuesToFile(final double[] data, final File outputFile)\\n    public static double[] readValuesFromFile(final File inputFile)\\n    public static double inRange(final DoubleRange validRange, final double value, final String definition)\\n    public static int inRange(final IntRange validRange, final int value, final String definition)\\n    public static void isValidArrayOffsetAndRangeLength(final int from, final int length, final int arrayLength, final String definition)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/param/ParamUtils.java', 'methodCount': 21, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 17}\n",
      "{}\n",
      "/**\n",
      "     * Returns a string of the values in ints joined by separator, such as A,B,C\n",
      "     *\n",
      "     * @param separator separator character\n",
      "     * @param ints   the array with values\n",
      "     * @return a string with the values separated by the separator\n",
      "     */\n",
      "public static String join(final String separator, final int[] ints)\n",
      "**************************************\n",
      "public static String join(final String separator, final int[] ints) {\n",
      "    Utils.nonNull(separator, \"the separator cannot be null\");\n",
      "    Utils.nonNull(ints, \"the ints cannot be null\");\n",
      "    if (ints.length == 0) {\n",
      "        return \"\";\n",
      "    } else {\n",
      "        final StringBuilder ret = new StringBuilder();\n",
      "        ret.append(ints[0]);\n",
      "        for (int i = 1; i < ints.length; ++i) {\n",
      "            ret.append(separator);\n",
      "            ret.append(ints[i]);\n",
      "        }\n",
      "        return ret.toString();\n",
      "    }\n",
      "}\n",
      "804\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1114/1243 [38:16<03:33,  1.65s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48397\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1115/1243 [38:18<03:32,  1.66s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48402\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1116/1243 [38:19<03:29,  1.65s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48403\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1117/1243 [38:21<03:30,  1.67s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48409\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1118/1243 [38:23<03:43,  1.79s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48413\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ParamUtils {\\n    public static final double INV_LOG_2 = 1 / Math.log(2.0);;\\n    private ParamUtils();\\n    public static double inRange(final double val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final long min, final long max, final String message)\\n    public static int inRange(final int val, final int min, final int max, final String message)\\n    public static double isPositiveOrZero(final double val, final String message)\\n    public static double isNegativeOrZero(final double val, final String message)\\n    public static long isPositiveOrZero(final long val, final String message)\\n    public static int isPositiveOrZero(final int val, final String message)\\n    public static long isPositive(final long val, final String message)\\n    public static int isPositive(final int val, final String message)\\n    public static double isPositive(final double val, final String message)\\n    public static double isFinite(final double val, final String message)\\n    public static double logb(final double val, final double base)\\n    public static double log2(final double val)\\n    public static void writeStringListToFile(final Iterable<String> stringIterator, final File outputFile)\\n    public static void writeValuesToFile(final double[] data, final File outputFile)\\n    public static double[] readValuesFromFile(final File inputFile)\\n    public static double inRange(final DoubleRange validRange, final double value, final String definition)\\n    public static int inRange(final IntRange validRange, final int value, final String definition)\\n    public static void isValidArrayOffsetAndRangeLength(final int from, final int length, final int arrayLength, final String definition)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/param/ParamUtils.java', 'methodCount': 21, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 17}\n",
      "{}\n",
      "/**\n",
      "     * Make all combinations of N size of objects\n",
      "     *\n",
      "     * if objects = [A, B, C]\n",
      "     * if N = 1 => [[A], [B], [C]]\n",
      "     * if N = 2 => [[A, A], [B, A], [C, A], [A, B], [B, B], [C, B], [A, C], [B, C], [C, C]]\n",
      "     *\n",
      "     * @param objects         list of objects\n",
      "     * @param n               size of each combination\n",
      "     * @param withReplacement if false, the resulting permutations will only contain unique objects from objects\n",
      "     * @return a list with all combinations with size n of objects.\n",
      "     */\n",
      "public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\n",
      "**************************************\n",
      "public static <T> List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement) {\n",
      "    final List<List<T>> combinations = new ArrayList<>();\n",
      "    if (n == 1) {\n",
      "        for (final T o : objects) {\n",
      "            combinations.add(Collections.singletonList(o));\n",
      "        }\n",
      "    } else if (n > 1) {\n",
      "        final List<List<T>> sub = makePermutations(objects, n - 1, withReplacement);\n",
      "        for (final List<T> subI : sub) {\n",
      "            for (final T a : objects) {\n",
      "                if (withReplacement || !subI.contains(a)) {\n",
      "                    combinations.add(Utils.cons(a, subI));\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return combinations;\n",
      "}\n",
      "957\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1119/1243 [38:25<03:36,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48445\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1120/1243 [38:26<03:27,  1.68s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48457\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1121/1243 [38:29<03:48,  1.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48466\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1122/1243 [38:31<04:00,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48475\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1123/1243 [38:33<04:07,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48534\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': \"public class IndexRange {\\n    /**\\n * First index in the range.\\n * <p>\\n *     It won't ever be negative nor greater than {@link #to}.\\n * </p>\\n */\\nprivate int from;;\\n    /**\\n * Index following the last index included in the range.\\n *\\n * <p>\\n *     It won't ever be negative nor less than {@link #from}.\\n * </p>\\n */\\nprivate int to;;\\n    public IndexRange(final int fromIndex, final int toIndex);\\n    public void shift(final int shift)\\n    public void shiftLeft(final int shift)\\n    public void shiftStart(final int shift)\\n    public void shiftStartLeft(final int shift)\\n    public void shiftEnd(final int shift)\\n    public void shiftEndLeft(final int shift)\\n    public int getStart()\\n    public int getEnd()\\n    private void validate()\\n    public boolean isValidLength(final int length)\\n    public int size()\\n    public void forEach(final IntConsumer lambda)\\n    public double[] mapToDouble(final IntToDoubleFunction lambda)\\n    public double sum(final IntToDoubleFunction lambda)\\n    public int[] mapToInteger(final IntUnaryOperator lambda)\\n    public List<Integer> filter(final IntPredicate predicate)\\n    public boolean equals(final Object other)\\n    public int hashCode()\\n    public String toString()\\n}\\n\", 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/IndexRange.java', 'methodCount': 20, 'lineCount': 24}, {'fieldCount': 13, 'classCount': 1, 'methodsOnly': \"public class MathUtils {\\n    /**\\n * The smallest log10 value we'll emit from normalizeFromLog10 and other functions\\n * where the real-space value is 0.0.\\n */\\npublic static final double LOG10_P_OF_ZERO = -1000000.0;;\\n    public static final double LOG10_ONE_HALF = Math.log10(0.5);;\\n    public static final double LOG10_ONE_THIRD = -Math.log10(3.0);;\\n    public static final double LOG_ONE_THIRD = -Math.log(3.0);;\\n    public static final double LOG_2 = Math.log(2.0);;\\n    public static final double INV_LOG_2 = 1.0 / Math.log(2.0);;\\n    public static final double LOG_10 = Math.log(10);;\\n    public static final double INV_LOG_10 = 1.0 / LOG_10;;\\n    public static final double LOG10_E = Math.log10(Math.E);;\\n    private static final double ROOT_TWO_PI = Math.sqrt(2.0 * Math.PI);;\\n    private static final Log10Cache LOG_10_CACHE = new Log10Cache();;\\n    private static final Log10FactorialCache LOG_10_FACTORIAL_CACHE = new Log10FactorialCache();;\\n    private static final DigammaCache DIGAMMA_CACHE = new DigammaCache();;\\n    private MathUtils();\\n    public static E randomSelect(final List<E> choices, final Function<E, Double> probabilityFunction, final RandomGenerator rng)\\n    public static int secondSmallestMinusSmallest(final int[] values, final int defaultValue)\\n    public static int[] normalizePLs(int[] PLs)\\n    public static int[] ebeAdd(final int[] a, final int[] b) throws DimensionMismatchException\\n    public static double[] ebeAdd(final double[] a, final double[] b) throws DimensionMismatchException\\n    public static double[] sumArrayFunction(final int min, final int max, final IntToDoubleArrayFunction function)\\n    public static void addToArrayInPlace(final double[] array, final double[] summand)\\n    public static void addToArrayInPlace(final int[] array, final int[] summand)\\n    public static int median(final int[] values)\\n    public static int median(final int[] values, final Percentile.EstimationType type)\\n    public static double dotProduct(double[] a, double[] b)\\n    public static double[] doubles(final double start, final double limit, final double step)\\n    public static double[] doubles(final int repeats, final double val)\\n    public static double sumOfSquares(final Collection<Integer> collection)\\n    public static int[] sampleIndicesWithoutReplacement(final int n, final int k)\\n    public static double log10OneMinusPow10(final double a)\\n    public static boolean isValidLog10ProbabilityVector(final double[] vector, final int expectedSize, final boolean shouldSumToOne)\\n    public static double sumLog10(final double[] log10values)\\n    public static int[] vectorDiff(final int[] x, final int[] y)\\n    public static double log10MultinomialCoefficient(final int n, final int[] k)\\n    public static double log10(int i)\\n    public static double digamma(int i)\\n    public static double log10sumLog10(final double[] log10values)\\n    public static double log10sumLog10(final double[] log10p, final int start)\\n    public static double log10sumLog10(final double[] log10p, final int start, final int finish)\\n    public static int fastRound(final double d)\\n    public static double logToLog10(final double ln)\\n    public static double approximateLog10SumLog10(final double[] vals)\\n    public static double approximateLog10SumLog10(final double[] vals, final int endIndex)\\n    public static double approximateLog10SumLog10(final double a, final double b, final double c)\\n    public static double approximateLog10SumLog10(final double a, final double b)\\n    public static double approximateLog10SumLog10(final double[] vals, final int fromIndex, final int toIndex)\\n    public static double sum(final double[] values)\\n    public static long sum(final int[] x)\\n    public static long sum(final long[] x)\\n    public static double sum(final double[] arr, final int start, final int stop)\\n    public static double sumDoubleFunction(final Collection<E> collection, final ToDoubleFunction<E> function)\\n    public static int sumIntFunction(final Collection<E> collection, final ToIntFunction<E> function)\\n    public static byte compareDoubles(final double a, final double b)\\n    public static byte compareDoubles(final double a, final double b, final double epsilon)\\n    public static double binomialCoefficient(final int n, final int k)\\n    public static double log10BinomialCoefficient(final int n, final int k)\\n    public static double binomialProbability(final int n, final int k, final double p)\\n    public static double log10BinomialProbability(final int n, final int k, final double log10p)\\n    public static double log10BinomialProbability(final int n, final int k)\\n    public static double log10SumLog10(final double[] log10Values, final int start)\\n    public static double log10SumLog10(final double[] log10Values)\\n    public static double log10SumLog10(final double[] log10Values, final int start, final int finish)\\n    public static double log10SumLog10(final double a, final double b)\\n    public static double log10SumLog10(final double a, final double b, final double c)\\n    public static double normalDistributionLog10(final double mean, final double sd, final double x)\\n    public static double square(final double x)\\n    public static double distanceSquared(final double[] x, final double[] y)\\n    public static double[] normalizeFromLog10ToLinearSpace(final double[] array)\\n    public static double[] normalizeLog10(final double[] array)\\n    public static double[] normalizeLog10(final double[] array, final boolean takeLog10OfOutput, final boolean inPlace)\\n    public static double[] normalizeLog10DeleteMePlease(final double[] array, final boolean takeLog10OfOutput)\\n    public static double[] scaleLogSpaceArrayForNumericalStability(final double[] array)\\n    public static double[] normalizeSumToOne(final double[] array)\\n    public static int maxElementIndex(final double[] array)\\n    public static int maxElementIndex(final double[] array, final int start, final int endIndex)\\n    public static int maxElementIndex(final double[] array, final int endIndex)\\n    public static int arrayMax(final int[] array)\\n    public static int arrayMax(final int[] array, final int from, final int to, final int defaultValue)\\n    public static double arrayMax(final double[] array)\\n    public static int arrayMin(final int[] array)\\n    public static int minElementIndex(final int[] array)\\n    public static boolean isValidLog10Probability(final double result)\\n    public static boolean isValidProbability(final double result)\\n    public static double log10ToLog(final double log10)\\n    public static double log10Gamma(final double x)\\n    public static double log10Factorial(final int n)\\n    public static double[] toLog10(final double[] prRealSpace)\\n    public static double log10OneMinusX(final double x)\\n    public static double median(final Collection<T> values)\\n    public static double roundToNDecimalPlaces(final double in, final int n)\\n    public static boolean wellFormedDouble(final double val)\\n    public static double normalDistribution(final double mean, final double sd, final double x)\\n    public static double dirichletMultinomial(double[] params, int[] counts)\\n    public static double[] applyToArray(final double[] array, final DoubleUnaryOperator func)\\n    public static double[] applyToArray(final int[] array, final IntToDoubleFunction func)\\n    public static double[] applyToArrayInPlace(final double[] array, final DoubleUnaryOperator func)\\n    public static boolean allMatch(final double[] array, final DoublePredicate pred)\\n    public static boolean allMatch(final int[] array, final IntPredicate pred)\\n    public static int maxElementIndex(final int[] array)\\n    public static int toIntExactOrThrow(long toConvert, Supplier<RuntimeException> exceptionSupplier)\\n    public static final double fastBernoulliEntropy(final double p)\\n    public static OptionalDouble binarySearchFindZero(final DoubleUnaryOperator func, final double lower, final double upper, final double precision)\\n}\\n\", 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/MathUtils.java', 'methodCount': 89, 'lineCount': 104}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class QualityUtils {\\n    /**\\n * Maximum quality score that can be encoded in a SAM/BAM file\\n */\\npublic static final byte MAX_SAM_QUAL_SCORE = SAMUtils.MAX_PHRED_SCORE;;\\n    /**\\n * conversion factor from phred scaled quality to log error probability and vice versa\\n */\\npublic static final double PHRED_TO_LOG_PROB_MULTIPLIER = -Math.log(10) / 10.0;;\\n    public static final double LOG_PROB_TO_PHRED_MULTIPLIER = 1 / PHRED_TO_LOG_PROB_MULTIPLIER;;\\n    private static final double MIN_LOG10_SCALED_QUAL = Math.log10(Double.MIN_VALUE);;\\n    protected static final double MIN_PHRED_SCALED_QUAL = -10.0 * MIN_LOG10_SCALED_QUAL;;\\n    /**\\n * bams containing quals above this value are extremely suspicious and we should warn the user\\n */\\npublic static final byte MAX_REASONABLE_Q_SCORE = 60;;\\n    /**\\n * The lowest quality score for a base that is considered reasonable for statistical analysis.  This is\\n * because Q 6 => you stand a 25% of being right, which means all bases are equally likely\\n */\\npublic static final byte MIN_USABLE_Q_SCORE = 6;;\\n    public static final int MAPPING_QUALITY_UNAVAILABLE = 255;;\\n    /**\\n * Maximum sense quality value.\\n */\\npublic static final int MAX_QUAL = 254;;\\n    /**\\n * Cached values for qual as byte calculations so they are very fast\\n */\\nprivate static final double[] qualToErrorProbCache = new double[MAX_QUAL + 1];;\\n    private static final double[] qualToProbLog10Cache = new double[MAX_QUAL + 1];;\\n    private QualityUtils();\\n    public static double qualToProb(final byte qual)\\n    public static double qualToProb(final double qual)\\n    public static double qualToProbLog10(final byte qual)\\n    public static double logProbToPhred(final double prob)\\n    public static double qualToErrorProb(final double qual)\\n    public static double qualToErrorProb(final byte qual)\\n    public static double qualToErrorProbLog10(final byte qual)\\n    public static double qualToErrorProbLog10(final double qual)\\n    public static byte errorProbToQual(final double errorRate)\\n    public static byte errorProbToQual(final double errorRate, final byte maxQual)\\n    public static byte errorProbToQual(final double prob, final int maxQual)\\n    public static byte trueProbToQual(final double prob)\\n    public static byte trueProbToQual(final double trueProb, final byte maxQual)\\n    public static byte trueProbToQual(final double prob, final int maxQual)\\n    public static double phredScaleCorrectRate(final double trueRate)\\n    public static double phredScaleErrorRate(final double errorRate)\\n    public static double phredScaleLog10ErrorRate(final double errorRateLog10)\\n    public static double phredScaleLog10CorrectRate(final double trueRateLog10)\\n    public static byte boundQual(int qual)\\n    public static byte boundQual(final int qual, final byte maxQual)\\n    public static double phredSum(final double[] phreds)\\n    public static double phredSum(final double a, final double b)\\n    public static double phredSum(final double a, final double b, final double c)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/QualityUtils.java', 'methodCount': 24, 'lineCount': 37}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Returns the error rate (in real space) of this interval, or 0 if there are no observations\n",
      "     * @return the empirical error rate ~= N errors / N obs\n",
      "     */\n",
      "public double getEmpiricalErrorRate()\n",
      "**************************************\n",
      "public double getEmpiricalErrorRate() {\n",
      "    if (numObservations == 0)\n",
      "        return 0.0;\n",
      "    else {\n",
      "        final double doubleMismatches = (numMismatches / MULTIPLIER) + SMOOTHING_CONSTANT;\n",
      "        final double doubleObservations = numObservations + SMOOTHING_CONSTANT + SMOOTHING_CONSTANT;\n",
      "        return doubleMismatches / doubleObservations;\n",
      "    }\n",
      "}\n",
      "3295\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  90%| | 1124/1243 [38:35<04:08,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48550\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1125/1243 [38:38<04:13,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48657\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1126/1243 [38:40<04:12,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48702\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1127/1243 [38:42<04:10,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48712\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1128/1243 [38:44<04:04,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48723\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{'GATKWDLWorkUnitHandler': 'public class GATKWDLWorkUnitHandler {\\n    private final static String GATK_FREEMARKER_TEMPLATE_NAME = \"wdlToolTemplate.wdl.ftl\";;\\n    // This must be kept in sync with the value used in build.gradle, where the file is created\\nprivate final static String dummyWDLTestFileName = \"dummyWDLTestFile\";;\\n    private final static String WDL_TEST_VALUE_PROPERTY = \"testValue\";;\\n    // Map of Java argument types that the WDL generator knows how to convert to a WDL type, along with the\\n// corresponding string substitution that needs to be run on the (Barclay-generated) string that describes\\n// the type. From a purely string perspective, some of these transforms are no-ops in that no actual\\n// conversion is required because the type names are identical in Java and WDL (i.e, File->File or\\n// String->String), but they\\'re included here for completeness, and to document the allowed type transitions.\\nprivate final static Map<Class<?>, ImmutablePair<String, String>> javaToWDLTypeMap = new HashMap<Class<?>, ImmutablePair<String, String>>() {\\n\\n    private static final long serialVersionUID = 1L;\\n\\n    {\\n        // GATK-specific File Types\\n        put(GATKPath.class, new ImmutablePair<>(GATKPath.class.getSimpleName(), \"File\"));\\n        // FeatureInputs require special handling to account for the generic type param(s)\\n        put(FeatureInput.class, new ImmutablePair<>(FeatureInput.class.getSimpleName(), \"File\"));\\n        put(Iso8601Date.class, new ImmutablePair<>(Iso8601Date.class.getSimpleName(), \"String\"));\\n        put(Date.class, new ImmutablePair<>(Date.class.getSimpleName(), \"String\"));\\n        put(ReadStructure.class, new ImmutablePair<>(ReadStructure.class.getSimpleName(), \"String\"));\\n    }\\n};;\\n    public GATKWDLWorkUnitHandler(final HelpDoclet doclet);\\n    public String getTemplateName(final DocWorkUnit workUnit)\\n    public String getDestinationFilename(final DocWorkUnit workUnit)\\n    public String getJSONFilename(final DocWorkUnit workUnit)\\n    protected void processNamedArgument(final DocWorkUnit currentWorkUnit, final Map<String, List<Map<String, Object>>> args, final NamedArgumentDefinition argDef)\\n    protected String processNamedArgument(final Map<String, Object> argBindings, final NamedArgumentDefinition argDef, final String fieldCommentText)\\n    protected void processPositionalArguments(final CommandLineArgumentParser clp, final Map<String, List<Map<String, Object>>> args)\\n    protected String testValueAsJSON(final String longName, final ArgumentDefinition argDef, final String wdlType, final String defaultWDLValue)\\n    protected String getDefaultValueForType(final ArgumentDefinition argDef)\\n    protected String convertJavaTypeToWDLType(final WorkflowOutput workflowOutput, final Class<?> argumentClass, final String docType, final String sourceContext)\\n    protected Pair<String, String> transformToWDLType(final Class<?> argumentClass)\\n    protected Pair<String, String> transformToWDLCollectionType(final Class<?> argumentCollectionClass)\\n    protected void addCustomBindings(final DocWorkUnit currentWorkUnit)\\n}\\n'}\n",
      "/**\n",
      "     * @return Create and return a DocWorkUnit-derived object to handle documentation\n",
      "     * for the target feature(s) represented by documentedFeature.\n",
      "     *\n",
      "     * @param classElement the Element for the class for this workunit\n",
      "     * @param clazz class of the target feature\n",
      "     * @param documentedFeature DocumentedFeature annotation for the target feature\n",
      "     * @return\n",
      "     */\n",
      "public DocWorkUnit createWorkUnit(final Element classElement, final Class<?> clazz, final DocumentedFeature documentedFeature)\n",
      "**************************************\n",
      "@Override\n",
      "public DocWorkUnit createWorkUnit(final Element classElement, final Class<?> clazz, final DocumentedFeature documentedFeature) {\n",
      "    return includeInDocs(documentedFeature, clazz) ? new DocWorkUnit(new GATKWDLWorkUnitHandler(this), classElement, clazz, documentedFeature) : null;\n",
      "}\n",
      "907\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1129/1243 [38:46<04:00,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48725\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1130/1243 [38:48<04:04,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48738\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1131/1243 [38:50<03:41,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48750\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1132/1243 [38:52<03:44,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48763\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1133/1243 [38:54<03:48,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48772\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Returns a compact String representation of the sequence dictionary it's passed\n",
      "     *\n",
      "     * The format of the returned String is:\n",
      "     * [ contig1Name(length: contig1Length) contig2Name(length: contig2Length) ... ]\n",
      "     *\n",
      "     * @param dict a non-null SAMSequenceDictionary\n",
      "     * @return A String containing all of the contig names and lengths from the sequence dictionary it's passed\n",
      "     */\n",
      "public static String getDictionaryAsString(final SAMSequenceDictionary dict)\n",
      "**************************************\n",
      "public static String getDictionaryAsString(final SAMSequenceDictionary dict) {\n",
      "    Utils.nonNull(dict, \"Sequence dictionary must be non-null\");\n",
      "    StringBuilder s = new StringBuilder(\"[ \");\n",
      "    for (SAMSequenceRecord dictionaryEntry : dict.getSequences()) {\n",
      "        s.append(dictionaryEntry.getSequenceName());\n",
      "        s.append(\"(length:\");\n",
      "        s.append(dictionaryEntry.getSequenceLength());\n",
      "        s.append(\") \");\n",
      "    }\n",
      "    s.append(\"]\");\n",
      "    return s.toString();\n",
      "}\n",
      "504\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%| | 1134/1243 [38:56<03:33,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48797\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%|| 1135/1243 [38:58<03:48,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48837\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%|| 1136/1243 [39:01<03:47,  2.13s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "48864\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  91%|| 1137/1243 [39:02<03:32,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49033\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1138/1243 [39:05<03:39,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49225\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class SimpleInterval {\\n    private static final long serialVersionUID = 1L;;\\n    public static final char CONTIG_SEPARATOR = \\':\\';;\\n    public static final char START_END_SEPARATOR = \\'-\\';;\\n    //note: needs to be a String because it\\'s used in an endsWith call.\\npublic static final String END_OF_CONTIG = \"+\";;\\n    private final int start;;\\n    private final int end;;\\n    private final String contig;;\\n    public SimpleInterval(final String contig, final int start, final int end);\\n    public SimpleInterval(final Locatable locatable);\\n     static void validatePositions(final String contig, final int start, final int end)\\n    public static boolean isValid(final String contig, final int start, final int end)\\n    public SimpleInterval(final String str);\\n     static int parsePositionThrowOnFailure(final String pos) throws NumberFormatException\\n     static int parsePosition(final String pos)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String getContig()\\n    public int getStart()\\n    public long getGA4GHStart()\\n    public int getEnd()\\n    public long getGA4GHEnd()\\n    public int size()\\n    public boolean overlaps(final Locatable other)\\n    public boolean overlapsWithMargin(final Locatable other, final int margin)\\n    public boolean contains(final Locatable other)\\n    public SimpleInterval intersect(final Locatable that)\\n    public SimpleInterval mergeWithContiguous(final Locatable that)\\n    public SimpleInterval spanWith(final Locatable other)\\n    private boolean contiguous(final Locatable that)\\n    public SimpleInterval expandWithinContig(final int padding, final int contigLength)\\n    public SimpleInterval expandWithinContig(final int padding, final SAMSequenceDictionary sequenceDictionary)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java', 'methodCount': 25, 'lineCount': 34}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Converts the given {@link String} into a {@link Strand}.\n",
      "     * @param s {@link String} to convert into a {@link Strand}.\n",
      "     * @return The {@link Strand} corresponding to {@code s}.\n",
      "     */\n",
      "private static Strand convertStringToStrand(final String s)\n",
      "**************************************\n",
      "/**\n",
      " * Converts the given {@link String} into a {@link Strand}.\n",
      " * @param s {@link String} to convert into a {@link Strand}.\n",
      " * @return The {@link Strand} corresponding to {@code s}.\n",
      " */\n",
      "private static Strand convertStringToStrand(final String s) {\n",
      "    if (s.equals(\"+\")) {\n",
      "        return Strand.POSITIVE;\n",
      "    } else if (s.equals(\"-\")) {\n",
      "        return Strand.NEGATIVE;\n",
      "    } else {\n",
      "        throw new IllegalArgumentException(\"Unexpected value: \" + s);\n",
      "    }\n",
      "}\n",
      "2544\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1139/1243 [39:07<03:42,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49234\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1140/1243 [39:09<03:27,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49286\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1141/1243 [39:11<03:27,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49309\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1142/1243 [39:12<03:13,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49312\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1143/1243 [39:14<03:18,  1.99s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49395\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class IntervalUtils {\\n    private static final Logger log = LogManager.getLogger(IntervalUtils.class);;\\n    /**\\n * Recognized extensions for interval files\\n */\\npublic static final List<String> GATK_INTERVAL_FILE_EXTENSIONS = Collections.unmodifiableList(Arrays.asList(\".list\", \".intervals\"));;\\n    /**\\n * Lexicographical (contig) order comparator.\\n * <p>\\n * Intervals from different contigs order is according their enclosing\\n * contigs name ascending lexicographical order.\\n * </p>\\n * <p>\\n * Intervals from the same contigs order is according to their start\\n * position ascending numerical order, and, in case of a tie, the stop position\\'s.\\n * </p>\\n * <p>\\n * The {@code null} contig is supported and comes last.\\n * </p>\\n */\\npublic static final Comparator<Locatable> LEXICOGRAPHICAL_ORDER_COMPARATOR = Comparator.comparing(Locatable::getContig, Comparator.nullsLast(String::compareTo)).thenComparingInt(Locatable::getStart).thenComparingInt(Locatable::getEnd);;\\n    private static final Logger logger = LogManager.getLogger(IntervalUtils.class);;\\n    public static final int compareLocatables(Locatable first, Locatable second, SAMSequenceDictionary dictionary)\\n    public static boolean isBefore(final Locatable first, final Locatable second, final SAMSequenceDictionary dictionary)\\n    public static boolean isAfter(final Locatable first, final Locatable second, final SAMSequenceDictionary dictionary)\\n    public static int compareContigs(final Locatable first, final Locatable second, final SAMSequenceDictionary dictionary)\\n    public static SimpleInterval getSpanningInterval(final List<? extends Locatable> locations)\\n    public static List<SimpleInterval> convertGenomeLocsToSimpleIntervals(final List<GenomeLoc> genomeLocIntervals)\\n    public static QueryInterval convertSimpleIntervalToQueryInterval(final SimpleInterval interval, final SAMSequenceDictionary sequenceDictionary)\\n    public static GenomeLocSortedSet loadIntervals(final List<String> intervalStrings, final IntervalSetRule intervalSetRule, final IntervalMergingRule intervalMergingRule, final int padding, final GenomeLocParser genomeLocParser)\\n    public static List<GenomeLoc> loadIntervalsNonMerging(final List<String> intervalStrings, final int padding, final GenomeLocParser genomeLocParser)\\n    public static List<GenomeLoc> parseIntervalArguments(final GenomeLocParser parser, final List<String> argList)\\n    public static List<GenomeLoc> parseIntervalArguments(final GenomeLocParser parser, final String arg)\\n    public static List<GenomeLoc> featureFileToIntervals(final GenomeLocParser parser, final String featureFile)\\n    private static List<GenomeLoc> gatkIntervalFileToList(final GenomeLocParser glParser, final String fileName)\\n    public static List<GenomeLoc> mergeListsBySetOperator(final List<GenomeLoc> setOne, final List<GenomeLoc> setTwo, final IntervalSetRule rule)\\n    public static GenomeLocSortedSet sortAndMergeIntervals(final GenomeLocParser parser, List<GenomeLoc> intervals, final IntervalMergingRule mergingRule)\\n    public static String equateIntervals(final List<GenomeLoc> masterArg, final List<GenomeLoc> testArg)\\n    private static List<T> reverse(final List<T> l)\\n    public static boolean isGatkIntervalFile(final String str)\\n    public static boolean isGatkIntervalFile(final String str, final boolean checkExists)\\n    public static Map<String, Integer> getContigSizes(final Path reference)\\n    public static void scatterContigIntervals(final SAMFileHeader fileHeader, final List<GenomeLoc> locs, final List<File> scatterParts)\\n    public static List<List<GenomeLoc>> splitIntervalsToSubLists(final List<GenomeLoc> locs, final List<Integer> splits)\\n    public static void scatterFixedIntervals(final SAMFileHeader fileHeader, final List<List<GenomeLoc>> splits, final List<File> scatterParts)\\n    public static List<List<GenomeLoc>> splitFixedIntervals(final List<GenomeLoc> locs, final int numParts)\\n    public static List<List<GenomeLoc>> splitLocusIntervals(final List<GenomeLoc> locs, final int numParts)\\n    private static SplitLocusRecursive splitLocusIntervals1(final LinkedList<GenomeLoc> remaining, final long idealSplitSize)\\n    public static boolean overlaps(final Locatable left, final Locatable right)\\n    public static String locatableToString(Locatable interval)\\n    public static List<GenomeLoc> flattenSplitIntervals(final List<List<GenomeLoc>> splits)\\n    private static void addFixedSplit(final List<Integer> splitPoints, final List<GenomeLoc> locs, final long locsSize, final int startIndex, final int stopIndex, final int numParts)\\n    private static Pair<Integer, Long> getFixedSplit(final List<GenomeLoc> locs, final long locsSize, final int startIndex, final int stopIndex, final int minLocs, final int maxLocs)\\n    private static Interval toInterval(final GenomeLoc loc, final int locIndex)\\n    public static List<GenomeLoc> mergeIntervalLocations(final List<GenomeLoc> raw, final IntervalMergingRule rule)\\n    public static long intervalSize(final List<GenomeLoc> locs)\\n    public static List<GenomeLoc> getIntervalsWithFlanks(final GenomeLocParser parser, final List<GenomeLoc> locs, final int basePairs)\\n    public static List<SimpleInterval> getIntervalsWithFlanks(final List<SimpleInterval> intervals, final int basePairs, final SAMSequenceDictionary dictionary)\\n    public static List<List<SimpleInterval>> groupIntervalsByContig(final List<SimpleInterval> sortedIntervals)\\n    private static LinkedHashMap<String, List<GenomeLoc>> splitByContig(final List<GenomeLoc> sorted)\\n    public static List<GenomeLoc> genomeLocsFromLocatables(final GenomeLocParser parser, final Collection<? extends Locatable> locatables)\\n    public static List<SimpleInterval> getAllIntervalsForReference(final SAMSequenceDictionary sequenceDictionary)\\n    public static List<SimpleInterval> getResolvedIntervals(final String intervalQueryString, final SAMSequenceDictionary sequenceDictionary)\\n    public static SimpleInterval trimIntervalToContig(final String contig, final int start, final int stop, final int contigLength)\\n    public static boolean intervalIsOnDictionaryContig(final SimpleInterval interval, final SAMSequenceDictionary dictionary)\\n    public static List<SimpleInterval> cutToShards(Iterable<SimpleInterval> intervals, int shardSize)\\n    public static int shardIndex(int oneBasedOffset, int shardSize)\\n    public static int beginOfShard(int shardIndex, int shardSize)\\n    public static int endOfShard(int shardIndex, int shardSize)\\n    public static List<SimpleInterval> getSpanningIntervals(final List<? extends Locatable> locations, final SAMSequenceDictionary sequenceDictionary)\\n    public static List<Locatable> combineAndSortBreakpoints(final List<T> unsortedLocatables1, final List<T> unsortedLocatables2, final SAMSequenceDictionary dictionary)\\n    public static void validateNoOverlappingIntervals(List<T> locatables)\\n    public static List<T> sortLocatablesBySequenceDictionary(final Collection<T> locatables, final SAMSequenceDictionary dictionary)\\n    public static Map<T, List<U>> createOverlapMap(final List<T> keys, final List<U> vals, final SAMSequenceDictionary dictionary)\\n    public static Comparator<T> getDictionaryOrderComparator(final SAMSequenceDictionary dictionary)\\n    public static boolean isReciprocalOverlap(final SimpleInterval interval1, final SimpleInterval interval2, final double reciprocalOverlapThreshold)\\n    public static Map<String, List<SimpleInterval>> sortAndMergeIntervals(final Collection<SimpleInterval> input, final SAMSequenceDictionary dictionary, final IntervalMergingRule rule)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java', 'methodCount': 55, 'lineCount': 61}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class SimpleInterval {\\n    private static final long serialVersionUID = 1L;;\\n    public static final char CONTIG_SEPARATOR = \\':\\';;\\n    public static final char START_END_SEPARATOR = \\'-\\';;\\n    //note: needs to be a String because it\\'s used in an endsWith call.\\npublic static final String END_OF_CONTIG = \"+\";;\\n    private final int start;;\\n    private final int end;;\\n    private final String contig;;\\n    public SimpleInterval(final String contig, final int start, final int end);\\n    public SimpleInterval(final Locatable locatable);\\n     static void validatePositions(final String contig, final int start, final int end)\\n    public static boolean isValid(final String contig, final int start, final int end)\\n    public SimpleInterval(final String str);\\n     static int parsePositionThrowOnFailure(final String pos) throws NumberFormatException\\n     static int parsePosition(final String pos)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String getContig()\\n    public int getStart()\\n    public long getGA4GHStart()\\n    public int getEnd()\\n    public long getGA4GHEnd()\\n    public int size()\\n    public boolean overlaps(final Locatable other)\\n    public boolean overlapsWithMargin(final Locatable other, final int margin)\\n    public boolean contains(final Locatable other)\\n    public SimpleInterval intersect(final Locatable that)\\n    public SimpleInterval mergeWithContiguous(final Locatable that)\\n    public SimpleInterval spanWith(final Locatable other)\\n    private boolean contiguous(final Locatable that)\\n    public SimpleInterval expandWithinContig(final int padding, final int contigLength)\\n    public SimpleInterval expandWithinContig(final int padding, final SAMSequenceDictionary sequenceDictionary)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java', 'methodCount': 25, 'lineCount': 34}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 2}\n",
      "{}\n",
      "/**\n",
      "     * Returns true if the specified interval 'that' overlaps with any of the exons actually spliced into this transcript.\n",
      "     *\n",
      "     * NOTE: this is is checking that the locatable is entirely contained within at least one exon.\n",
      "     * */\n",
      "public boolean contains(Locatable that)\n",
      "**************************************\n",
      "@Override\n",
      "public boolean contains(Locatable that) {\n",
      "    if (exons.isEmpty()) {\n",
      "        return getLocation().contains(that);\n",
      "    }\n",
      "    for (SimpleInterval exon : exons) {\n",
      "        if (IntervalUtils.overlaps(exon, that)) {\n",
      "            return true;\n",
      "        }\n",
      "    }\n",
      "    return false;\n",
      "}\n",
      "2548\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1144/1243 [39:17<03:25,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49537\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1145/1243 [39:18<03:13,  1.97s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49636\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1146/1243 [39:21<03:14,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49655\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1147/1243 [39:23<03:16,  2.05s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49659\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1148/1243 [39:25<03:21,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49669\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 0, 'methodsOnly': '', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Tail.java', 'methodCount': 0, 'lineCount': 0}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public interface SmithWatermanAligner {\\n    Logger logger = LogManager.getLogger(SmithWatermanAligner.class);;\\n     SmithWatermanAlignment align(final byte[] ref, final byte[] alt, SWParameters parameters, SWOverhangStrategy overhangStrategy)\\n     void close()\\n     static SmithWatermanAligner getAligner(final Implementation type)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanAligner.java', 'methodCount': 3, 'lineCount': 6}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface SmithWatermanAlignment {\\n     Cigar getCigar()\\n     int getAlignmentOffset()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/smithwaterman/SmithWatermanAlignment.java', 'methodCount': 2, 'lineCount': 4}], 'asteriskCount': 0, 'importCount': 3, 'javaImportCount': 4}\n",
      "{'CigarBuilder': 'public class CigarBuilder {\\n    private final List<CigarElement> cigarElements = new ArrayList<>();;\\n    // track the last operator so we can merge consecutive elements with the same operator\\n// for example, adding 3M and 4M is equivalent to adding 7M\\n// also we ignore leading deletions so for example 10S + 5D = 10S\\nprivate CigarOperator lastOperator = null;;\\n    private Section section = Section.LEFT_HARD_CLIP;;\\n    private final boolean removeDeletionsAtEnds;;\\n    private int leadingDeletionBasesRemoved = 0;;\\n    private int trailingDeletionBasesRemoved = 0;;\\n    private int trailingDeletionBasesRemovedInMake = 0;;\\n    public CigarBuilder(final boolean removeDeletionsAtEnds);\\n    public CigarBuilder();\\n    public CigarBuilder add(final CigarElement element)\\n    private boolean lastTwoElementsWereDeletionAndInsertion()\\n    public CigarBuilder addAll(final Iterable<CigarElement> elements)\\n    public Cigar make(final boolean allowEmpty)\\n    public Cigar make()\\n    private void advanceSectionAndValidateCigarOrder(CigarOperator operator)\\n    public int getLeadingDeletionBasesRemoved()\\n    public int getTrailingDeletionBasesRemoved()\\n    public Result makeAndRecordDeletionsRemovedResult()\\n}\\n'}\n",
      "/**\n",
      "     * replace soft clips (S) with match (M) operators, normalizing the result by all the transformations of the {@link CigarBuilder} class:\n",
      "     * merging consecutive identical operators and removing zero-length elements.  For example 10S10M -> 20M and 10S10M10I10I -> 20M20I.\n",
      "     */\n",
      "public static Cigar revertSoftClips(final Cigar originalCigar)\n",
      "**************************************\n",
      "public static Cigar revertSoftClips(final Cigar originalCigar) {\n",
      "    final CigarBuilder builder = new CigarBuilder();\n",
      "    for (final CigarElement element : originalCigar.getCigarElements()) {\n",
      "        if (element.getOperator() == CigarOperator.SOFT_CLIP) {\n",
      "            builder.add(new CigarElement(element.getLength(), CigarOperator.MATCH_OR_MISMATCH));\n",
      "        } else {\n",
      "            builder.add(element);\n",
      "        }\n",
      "    }\n",
      "    return builder.make();\n",
      "}\n",
      "2560\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  92%|| 1149/1243 [39:27<03:21,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49820\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1150/1243 [39:30<03:30,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49825\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1151/1243 [39:31<03:11,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49840\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1152/1243 [39:33<03:09,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49874\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1153/1243 [39:36<03:08,  2.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49877\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Randomly picks one of the best alignments and puts it into the 0th slot of the list.\n",
      "     * @param recs List of alignments sorted in descending order of alignment quality.\n",
      "     */\n",
      "private void randomlySelectPrimaryFromBest(List<SAMRecord> recs)\n",
      "**************************************\n",
      "private void randomlySelectPrimaryFromBest(List<SAMRecord> recs) {\n",
      "    if (recs.isEmpty())\n",
      "        return;\n",
      "    final int bestMapq = recs.get(0).getMappingQuality();\n",
      "    int i;\n",
      "    for (i = 1; i < recs.size() && recs.get(i).getMappingQuality() == bestMapq; ++i) {\n",
      "    }\n",
      "    final int bestIndex = random.nextInt(i);\n",
      "    if (bestIndex == 0)\n",
      "        return;\n",
      "    final SAMRecord tmp = recs.get(0);\n",
      "    recs.set(0, recs.get(bestIndex));\n",
      "    recs.set(bestIndex, tmp);\n",
      "}\n",
      "234\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1154/1243 [39:37<02:59,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49887\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1155/1243 [39:39<02:49,  1.93s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "49888\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1156/1243 [39:41<02:50,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50064\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1157/1243 [39:43<02:39,  1.86s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50245\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1158/1243 [39:45<02:46,  1.95s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50339\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Returns true if there is a dbSnp entry at pos in sequenceName, otherwise false\n",
      "     */\n",
      "public boolean isDbSnpSite(final String sequenceName, final int pos)\n",
      "**************************************\n",
      "public boolean isDbSnpSite(final String sequenceName, final int pos) {\n",
      "    if (sequenceToBitSet.get(sequenceName) == null) {\n",
      "        return false;\n",
      "    }\n",
      "    if (pos > sequenceToBitSet.get(sequenceName).length()) {\n",
      "        return false;\n",
      "    }\n",
      "    return sequenceToBitSet.get(sequenceName).get(pos);\n",
      "}\n",
      "1864\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1159/1243 [39:47<02:51,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50369\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1160/1243 [39:49<02:38,  1.91s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50410\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1161/1243 [39:50<02:29,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50461\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  93%|| 1162/1243 [39:52<02:27,  1.83s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50466\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1163/1243 [39:54<02:20,  1.75s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50517\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Print out in a human readable form the matrix for debugging\n",
      "     * @param name the name of this matrix\n",
      "     * @param matrix the matrix of values\n",
      "     */\n",
      "private void dumpMatrix(final String name, final double[][] matrix)\n",
      "**************************************\n",
      "private void dumpMatrix(final String name, final double[][] matrix) {\n",
      "    System.out.printf(\"%s%n\", name);\n",
      "    for (int i = 0; i < matrix.length; i++) {\n",
      "        System.out.printf(\"\\t%s[%d]\", name, i);\n",
      "        for (int j = 0; j < matrix[i].length; j++) {\n",
      "            if (Double.isInfinite(matrix[i][j]))\n",
      "                System.out.printf(\" %15s\", String.format(\"%f\", matrix[i][j]));\n",
      "            else\n",
      "                System.out.printf(\" % 15.5e\", matrix[i][j]);\n",
      "        }\n",
      "        System.out.println();\n",
      "    }\n",
      "}\n",
      "223\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1164/1243 [39:56<02:24,  1.82s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50520\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1165/1243 [39:58<02:29,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50530\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1166/1243 [40:00<02:32,  1.98s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50534\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1167/1243 [40:02<02:34,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50670\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1168/1243 [40:04<02:30,  2.01s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50676\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{'ProcessOutput': 'public class ProcessOutput {\\n    private final int exitValue;;\\n    private final StreamOutput stdout;;\\n    private final StreamOutput stderr;;\\n    public ProcessOutput(int exitValue, StreamOutput stdout, StreamOutput stderr);\\n    public int getExitValue()\\n    public StreamOutput getStdout()\\n    public StreamOutput getStderr()\\n    public String getStatusSummary(boolean includeStdout)\\n}\\n'}\n",
      "/**\n",
      "     * Inspect process output exit code and construct a corresponding exception message.\n",
      "     *\n",
      "     * @param po script process output\n",
      "     * @return script exception message\n",
      "     */\n",
      "public String getExceptionMessageFromScriptError(final ProcessOutput po)\n",
      "**************************************\n",
      "public String getExceptionMessageFromScriptError(final ProcessOutput po) {\n",
      "    Utils.nonNull(po, \"process output cannot be null\");\n",
      "    final int exitValue = po.getExitValue();\n",
      "    final String commandLineMessage = String.format(\"\\n%s exited with %d\\nCommand Line: %s\", externalScriptExecutableName, exitValue, String.join(\" \", Utils.nonNull(commandLineArgs, \"command line args have not been set yet\")));\n",
      "    final boolean outputStdout = !logger.isDebugEnabled();\n",
      "    return commandLineMessage.concat(po.getStatusSummary(outputStdout));\n",
      "}\n",
      "2278\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1169/1243 [40:06<02:21,  1.92s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50693\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1170/1243 [40:08<02:15,  1.85s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50756\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1171/1243 [40:10<02:19,  1.94s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50770\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1172/1243 [40:12<02:26,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50775\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1173/1243 [40:15<02:32,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50791\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ReadsSparkSink {\\n    private final static Logger logger = LogManager.getLogger(ReadsSparkSink.class);;\\n    public static void writeReads(final JavaSparkContext ctx, final String outputFile, final GATKPath referencePathSpecifier, final JavaRDD<GATKRead> reads, final SAMFileHeader header, ReadsWriteFormat format) throws IOException\\n    public static void writeReads(final JavaSparkContext ctx, final String outputFile, final GATKPath referencePathSpecifier, final JavaRDD<GATKRead> reads, final SAMFileHeader header, ReadsWriteFormat format, final int numReducers, final String outputPartsDir, final boolean sortReadsToHeader, final long splittingIndexGranularity) throws IOException\\n    public static void writeReads(final JavaSparkContext ctx, final String outputFile, final GATKPath referencePathSpecifier, final JavaRDD<GATKRead> reads, final SAMFileHeader header, ReadsWriteFormat format, final int numReducers, final String outputPartsDir, final boolean writeBai, final boolean writeSbi, final boolean sortReadsToHeader, final long splittingIndexGranularity) throws IOException\\n    private static void writeReads(final JavaSparkContext ctx, final String outputFile, final GATKPath referencePathSpecifier, final JavaRDD<SAMRecord> reads, final SAMFileHeader header, final long sbiIndexGranularity, final WriteOption... writeOptions) throws IOException\\n    private static void writeReadsADAM(final JavaSparkContext ctx, final String outputFile, final JavaRDD<SAMRecord> reads, final SAMFileHeader header) throws IOException\\n    private static void deleteHadoopFile(String fileToObliterate, Configuration conf) throws IOException\\n    public static String getDefaultPartsDirectory(String file)\\n    private static JavaRDD<SAMRecord> sortSamRecordsToMatchHeader(final JavaRDD<SAMRecord> reads, final SAMFileHeader header, final int numReducers)\\n    private static Comparator<SAMRecord> getSAMRecordComparator(final SAMFileHeader header)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSink.java', 'methodCount': 9, 'lineCount': 12}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Sort reads into queryname order if they are not already sorted\n",
      "     */\n",
      "public static JavaRDD<GATKRead> querynameSortReadsIfNecessary(JavaRDD<GATKRead> reads, int numReducers, SAMFileHeader header)\n",
      "**************************************\n",
      "public static JavaRDD<GATKRead> querynameSortReadsIfNecessary(JavaRDD<GATKRead> reads, int numReducers, SAMFileHeader header) {\n",
      "    JavaRDD<GATKRead> sortedReadsForMarking;\n",
      "    if (ReadUtils.isReadNameGroupedBam(header)) {\n",
      "        sortedReadsForMarking = reads;\n",
      "    } else {\n",
      "        header.setSortOrder(SAMFileHeader.SortOrder.queryname);\n",
      "        sortedReadsForMarking = sortReadsAccordingToHeader(reads, header, numReducers);\n",
      "    }\n",
      "    return sortedReadsForMarking;\n",
      "}\n",
      "2698\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  94%|| 1174/1243 [40:17<02:34,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50866\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1175/1243 [40:19<02:30,  2.22s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50867\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1176/1243 [40:21<02:28,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50981\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1177/1243 [40:24<02:25,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51032\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1178/1243 [40:26<02:21,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51034\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface HasGenomeLocation {\\n    public Locatable getLocation()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/HasGenomeLocation.java', 'methodCount': 1, 'lineCount': 3}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class ReadPileup {\\n    private final Locatable loc;;\\n    private final List<PileupElement> pileupElements;;\\n    /**\\n * Constant used by samtools to downgrade a quality for overlapping reads that disagrees in their base.\\n */\\npublic static final double SAMTOOLS_OVERLAP_LOW_CONFIDENCE = 0.8;;\\n    public ReadPileup(final Locatable loc, final List<PileupElement> pileup);\\n    public ReadPileup(final Locatable loc, final Map<String, ReadPileup> stratifiedPileup);\\n    public ReadPileup(final Locatable loc);\\n    public ReadPileup(final Locatable loc, final List<GATKRead> reads, final int offset);\\n    public ReadPileup(final Locatable loc, final List<GATKRead> reads, final List<Integer> offsets);\\n    public ReadPileup(final Locatable loc, final Iterable<GATKRead> reads);\\n     PileupElement getElementForRead(final GATKRead read)\\n    private static List<PileupElement> readsOffsetsToPileup(final List<GATKRead> reads, final List<Integer> offsets)\\n    private static List<PileupElement> readsOffsetsToPileup(final List<GATKRead> reads, final int offset)\\n    public ReadPileup makeFilteredPileup(final Predicate<PileupElement> filter)\\n    public ReadPileup getPileupForLane(final String laneID)\\n    public ReadPileup getPileupForSample(final String sample, final SAMFileHeader header)\\n    public Set<String> getReadGroupIDs()\\n    public Set<String> getSamples(final SAMFileHeader header)\\n    public Map<String, ReadPileup> splitBySample(final SAMFileHeader header, final String unknownSampleName)\\n    public Iterator<PileupElement> iterator()\\n    public Iterator<PileupElement> sortedIterator()\\n    public int size()\\n    public boolean isEmpty()\\n    public Locatable getLocation()\\n    public int[] getBaseCounts()\\n    public void fixOverlaps()\\n     static void fixPairOverlappingQualities(final PileupElement firstElement, final PileupElement secondElement)\\n    public final static Comparator<PileupElement> baseQualTieBreaker = new Comparator<PileupElement>() {\\n\\n    @Override\\n    public int compare(PileupElement o1, PileupElement o2) {\\n        return Byte.compare(o1.getQual(), o2.getQual());\\n    }\\n};;\\n    public final static Comparator<PileupElement> mapQualTieBreaker = new Comparator<PileupElement>() {\\n\\n    @Override\\n    public int compare(PileupElement o1, PileupElement o2) {\\n        return Integer.compare(o1.getMappingQual(), o2.getMappingQual());\\n    }\\n};;\\n    public ReadPileup getOverlappingFragmentFilteredPileup(SAMFileHeader header)\\n    public ReadPileup getOverlappingFragmentFilteredPileup(boolean discardDiscordant, Comparator<PileupElement> tieBreaker, SAMFileHeader header)\\n    private Collection<PileupElement> filterSingleSampleForOverlaps(ReadPileup pileup, Comparator<PileupElement> tieBreaker, boolean discardDiscordant)\\n    public String toString()\\n    public String getPileupString(final char ref)\\n    public List<GATKRead> getReads()\\n    private Stream<PileupElement> getElementStream()\\n    public int getNumberOfElements(final Predicate<PileupElement> peFilter)\\n    public List<Integer> getOffsets()\\n    private int[] extractIntArray(final ToIntFunction<PileupElement> map)\\n    public byte[] getBases()\\n    public byte[] getBaseQuals()\\n    private byte[] toByteArray(final int[] ints)\\n    public int[] getMappingQuals()\\n    private String getQualsString()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/pileup/ReadPileup.java', 'methodCount': 38, 'lineCount': 45}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'AlignmentContext': 'public class AlignmentContext {\\n    private final Locatable loc;;\\n    private final ReadPileup basePileup;;\\n    private boolean hasPileupBeenDownsampled;;\\n    public AlignmentContext(final Locatable loc, final ReadPileup basePileup);\\n    public AlignmentContext(final Locatable loc, final ReadPileup basePileup, final boolean hasPileupBeenDownsampled);\\n    public int size()\\n    public String getContig()\\n    public int getStart()\\n    public int getEnd()\\n    public long getPosition()\\n    public Locatable getLocation()\\n    public boolean hasPileupBeenDownsampled()\\n    public ReadPileup getBasePileup()\\n    public AlignmentContext stratify(final ReadOrientation type)\\n    public Map<String, AlignmentContext> splitContextBySampleName(final SAMFileHeader header)\\n    public Map<String, AlignmentContext> splitContextBySampleName(final String assumedSingleSample, final SAMFileHeader header)\\n    public static Map<String, AlignmentContext> splitContextBySampleName(final ReadPileup pileup, final SAMFileHeader header)\\n    public String toString()\\n}\\n'}\n",
      "/**\n",
      "     * Splits the given AlignmentContext into a StratifiedAlignmentContext per sample, but referenced by sample name instead\n",
      "     * of sample object.\n",
      "     *\n",
      "     * @param assumedSingleSample If non-null, assume this is the only sample in our pileup and return immediately.\n",
      "     *                            If null, get the list of samples from the provided header and do the work of splitting by sample.\n",
      "     * @return a Map of sample name to StratifiedAlignmentContext; samples without coverage are not included\n",
      "     **/\n",
      "public Map<String, AlignmentContext> splitContextBySampleName(final String assumedSingleSample, final SAMFileHeader header)\n",
      "**************************************\n",
      "public Map<String, AlignmentContext> splitContextBySampleName(final String assumedSingleSample, final SAMFileHeader header) {\n",
      "    if (assumedSingleSample != null) {\n",
      "        return Collections.singletonMap(assumedSingleSample, this);\n",
      "    }\n",
      "    final Locatable loc = this.getLocation();\n",
      "    final Map<String, ReadPileup> pileups = this.getBasePileup().splitBySample(header, assumedSingleSample);\n",
      "    final Map<String, AlignmentContext> contexts = new LinkedHashMap<>(pileups.size());\n",
      "    for (final Map.Entry<String, ReadPileup> entry : pileups.entrySet()) {\n",
      "        if (entry.getValue().isEmpty()) {\n",
      "            continue;\n",
      "        }\n",
      "        contexts.put(entry.getKey(), new AlignmentContext(loc, entry.getValue()));\n",
      "    }\n",
      "    return contexts;\n",
      "}\n",
      "3327\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1179/1243 [40:28<02:17,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51039\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1180/1243 [40:30<02:15,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51040\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1181/1243 [40:32<02:16,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51065\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1182/1243 [40:35<02:17,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51069\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1183/1243 [40:37<02:15,  2.27s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51075\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Stop the progress meter and output summary statistics to the logger\n",
      "     * @throws IllegalStateException if the meter has not been started yet or has been stopped already\n",
      "     */\n",
      "public void stop()\n",
      "**************************************\n",
      "public void stop() {\n",
      "    if (disabled) {\n",
      "        return;\n",
      "    }\n",
      "    Utils.validate(started, \"the progress meter has not been started yet\");\n",
      "    Utils.validate(!stopped, \"the progress meter has been stopped already\");\n",
      "    this.stopped = true;\n",
      "    currentTimeMs = timeFunction.getAsLong();\n",
      "    printProgress();\n",
      "    logger.info(String.format(\"Traversal complete. Processed %d total %s in %.1f minutes.\", numRecordsProcessed, recordLabel, elapsedTimeInMinutes()));\n",
      "}\n",
      "1891\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1184/1243 [40:39<02:16,  2.32s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51107\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1185/1243 [40:42<02:13,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51229\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1186/1243 [40:43<02:02,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51235\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  95%|| 1187/1243 [40:46<02:13,  2.39s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51251\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1188/1243 [40:48<02:03,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51296\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class SimpleInterval {\\n    private static final long serialVersionUID = 1L;;\\n    public static final char CONTIG_SEPARATOR = \\':\\';;\\n    public static final char START_END_SEPARATOR = \\'-\\';;\\n    //note: needs to be a String because it\\'s used in an endsWith call.\\npublic static final String END_OF_CONTIG = \"+\";;\\n    private final int start;;\\n    private final int end;;\\n    private final String contig;;\\n    public SimpleInterval(final String contig, final int start, final int end);\\n    public SimpleInterval(final Locatable locatable);\\n     static void validatePositions(final String contig, final int start, final int end)\\n    public static boolean isValid(final String contig, final int start, final int end)\\n    public SimpleInterval(final String str);\\n     static int parsePositionThrowOnFailure(final String pos) throws NumberFormatException\\n     static int parsePosition(final String pos)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String getContig()\\n    public int getStart()\\n    public long getGA4GHStart()\\n    public int getEnd()\\n    public long getGA4GHEnd()\\n    public int size()\\n    public boolean overlaps(final Locatable other)\\n    public boolean overlapsWithMargin(final Locatable other, final int margin)\\n    public boolean contains(final Locatable other)\\n    public SimpleInterval intersect(final Locatable that)\\n    public SimpleInterval mergeWithContiguous(final Locatable that)\\n    public SimpleInterval spanWith(final Locatable other)\\n    private boolean contiguous(final Locatable that)\\n    public SimpleInterval expandWithinContig(final int padding, final int contigLength)\\n    public SimpleInterval expandWithinContig(final int padding, final SAMSequenceDictionary sequenceDictionary)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java', 'methodCount': 25, 'lineCount': 34}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Determines whether all records overlapping the provided interval are already contained in our cache.\n",
      "     *\n",
      "     * @param interval the interval to check against the contents of our cache\n",
      "     * @return true if all records overlapping the provided interval are already contained in our cache, otherwise false\n",
      "     */\n",
      "public boolean cacheHit(final Locatable interval)\n",
      "**************************************\n",
      "public boolean cacheHit(final Locatable interval) {\n",
      "    final boolean cacheHit = cachedInterval != null && cachedInterval.contains(interval);\n",
      "    if (cacheHit) {\n",
      "        ++numCacheHits;\n",
      "    } else {\n",
      "        ++numCacheMisses;\n",
      "    }\n",
      "    return cacheHit;\n",
      "}\n",
      "745\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1189/1243 [40:50<01:51,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51300\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1190/1243 [40:52<01:53,  2.15s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51321\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1191/1243 [40:54<01:45,  2.02s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51322\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1192/1243 [40:56<01:47,  2.11s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51326\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1193/1243 [40:59<01:48,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51335\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 7, 'classCount': 1, 'methodsOnly': 'public class SimpleInterval {\\n    private static final long serialVersionUID = 1L;;\\n    public static final char CONTIG_SEPARATOR = \\':\\';;\\n    public static final char START_END_SEPARATOR = \\'-\\';;\\n    //note: needs to be a String because it\\'s used in an endsWith call.\\npublic static final String END_OF_CONTIG = \"+\";;\\n    private final int start;;\\n    private final int end;;\\n    private final String contig;;\\n    public SimpleInterval(final String contig, final int start, final int end);\\n    public SimpleInterval(final Locatable locatable);\\n     static void validatePositions(final String contig, final int start, final int end)\\n    public static boolean isValid(final String contig, final int start, final int end)\\n    public SimpleInterval(final String str);\\n     static int parsePositionThrowOnFailure(final String pos) throws NumberFormatException\\n     static int parsePosition(final String pos)\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public String toString()\\n    public String getContig()\\n    public int getStart()\\n    public long getGA4GHStart()\\n    public int getEnd()\\n    public long getGA4GHEnd()\\n    public int size()\\n    public boolean overlaps(final Locatable other)\\n    public boolean overlapsWithMargin(final Locatable other, final int margin)\\n    public boolean contains(final Locatable other)\\n    public SimpleInterval intersect(final Locatable that)\\n    public SimpleInterval mergeWithContiguous(final Locatable that)\\n    public SimpleInterval spanWith(final Locatable other)\\n    private boolean contiguous(final Locatable that)\\n    public SimpleInterval expandWithinContig(final int padding, final int contigLength)\\n    public SimpleInterval expandWithinContig(final int padding, final SAMSequenceDictionary sequenceDictionary)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/SimpleInterval.java', 'methodCount': 25, 'lineCount': 34}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}, {'fieldCount': 3, 'classCount': 1, 'methodsOnly': 'public class ByteArrayIterator {\\n    private final byte[] byteArray;;\\n    private int currentPosition;;\\n    private int stopIndex;;\\n    public ByteArrayIterator(final byte[] byteArray);\\n    public ByteArrayIterator(final byte[] byteArray, int firstIndex, int stopIndex);\\n    public boolean hasNext()\\n    public Byte next()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/iterators/ByteArrayIterator.java', 'methodCount': 4, 'lineCount': 9}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 1}\n",
      "{}\n",
      "/**\n",
      "     * Get a kmer around a position in reference without altering the internal state of the object\n",
      "     * The position must lie within the window\n",
      "     *\n",
      "     * Returns null when, at the ends of a contig, we cannot expand the window to the requested size\n",
      "     */\n",
      "public String getKmerAround(final int center, final int numBasesOnEachSide)\n",
      "**************************************\n",
      "public String getKmerAround(final int center, final int numBasesOnEachSide) {\n",
      "    Utils.validateArg(center >= 1, () -> \"start position must be positive\");\n",
      "    Utils.validateArg(window.getStart() <= center && center <= window.getEnd(), \"position must be smaller than end position\");\n",
      "    final SimpleInterval newWindow = new SimpleInterval(window.getContig(), center, center).expandWithinContig(numBasesOnEachSide, getContigLength(window.getContig()));\n",
      "    if (newWindow.getEnd() - newWindow.getStart() < 2 * numBasesOnEachSide) {\n",
      "        return null;\n",
      "    }\n",
      "    return new String(getBases(newWindow));\n",
      "}\n",
      "2848\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1194/1243 [41:01<01:50,  2.25s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51441\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1195/1243 [41:03<01:50,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51459\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1196/1243 [41:06<01:47,  2.29s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51611\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1197/1243 [41:08<01:39,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51789\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1198/1243 [41:09<01:32,  2.06s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51803\n",
      "{'classCount': 1, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class PicardNonZeroExitException {\\n    private static final long serialVersionUID = 0L;;\\n    private final int picardToolReturnCode;;\\n    public PicardNonZeroExitException(final int picardToolReturnCode);\\n    public int getToolReturnCode()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/PicardNonZeroExitException.java', 'methodCount': 2, 'lineCount': 6}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class ClassUtils {\\n    private ClassUtils();\\n    public static boolean canMakeInstances(final Class<?> clazz)\\n    public static List<T> makeInstancesOfSubclasses(final Class<? extends T> clazz, final Package pack)\\n    public static T makeInstanceOf(final Class<T> clazz)\\n    public static List<String> knownSubInterfaceSimpleNames(final Class<?> iface)\\n    public static Set<Class<?>> knownSubInterfaces(final Class<?> iface)\\n    public static List<Class<?>> getClassesOfType(final Class<?> targetClass, final List<Class<?>> classesToSearch)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/ClassUtils.java', 'methodCount': 7, 'lineCount': 9}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}, {'fieldCount': 5, 'classCount': 1, 'methodsOnly': 'public class ConfigFactory {\\n    private static final Logger logger = LogManager.getLogger(ConfigFactory.class);;\\n    //=======================================\\n// Singleton members / methods:\\nprivate static final ConfigFactory instance;;\\n    public static ConfigFactory getInstance()\\n    private ConfigFactory();\\n    private static final Pattern sourcesAnnotationPathVariablePattern = Pattern.compile(\"\\\\\\\\$\\\\\\\\{(.*)}\");;\\n    @VisibleForTesting\\nstatic final String NO_PATH_VARIABLE_VALUE = \"/dev/null\";;\\n    private final Set<Class<? extends Config>> alreadyResolvedPathVariables = new HashSet<>();;\\n     void checkFileNamePropertyExistenceAndSetConfigFactoryProperties(final List<String> filenameProperties)\\n     List<String> getConfigPathVariableNamesFromConfigClasses(final List<Class<?>> configurationClasses)\\n     List<String> getSourcesAnnotationPathVariables(final Class<? extends T> configClass)\\n     void injectToSystemProperties(final Map<String, String> properties)\\n    public GATKConfig getGATKConfig()\\n    public static void dumpConfigSettings(final T config, final Path outFilePath)\\n    public T create(final Class<? extends T> clazz, final Map<?, ?>... imports)\\n    public T getOrCreate(final Class<? extends T> clazz, final Map<?, ?>... imports)\\n    public T getOrCreate(final Factory factory, final Class<? extends T> clazz, final Map<?, ?>... imports)\\n    public T getOrCreate(final Object key, final Class<? extends T> clazz, final Map<?, ?>... imports)\\n    public T getOrCreate(final Factory factory, final Object key, final Class<? extends T> clazz, final Map<?, ?>... imports)\\n    private synchronized void resolvePathVariables(final Class<? extends T> clazz)\\n    public T get(final Object key)\\n    public static String getConfigFilenameFromArgs(final String[] args, final String configFileOption)\\n    public synchronized void initializeConfigurationsFromCommandLineArgs(final String[] argList, final String configFileOption)\\n    public synchronized void initializeConfigurationsFromCommandLineArgs(final String[] argList, final String configFileOption, final Class<? extends T> configClass)\\n     synchronized T getOrCreateConfigFromFile(final String configFileName, final Class<? extends T> configClass)\\n     synchronized T createConfigFromFile(final String configFileName, final Class<? extends T> configClass)\\n    public synchronized void injectSystemPropertiesFromConfig(final T config)\\n    public static void logConfigFields(final T config)\\n     static LinkedHashMap<String, String> getSystemPropertiesFromConfig(final T config)\\n    public static void logConfigFields(final T config, final Log.LogLevel logLevel)\\n     static LinkedHashMap<String, Object> getConfigMap(final T config, final boolean onlySystemProperties)\\n    private static LinkedHashMap<String, String> convertConfigMapToStringStringMap(final LinkedHashMap<String, Object> configMap)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/config/ConfigFactory.java', 'methodCount': 26, 'lineCount': 33}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class RuntimeUtils {\\n    public static final String[] PATHS;;\\n    public static File which(String executable)\\n    public static String toolDisplayName(final Class<?> toolClass)\\n    public static String getToolkitName(Class<?> clazz)\\n    public static Manifest getManifest(Class<?> clazz)\\n    public static String getVersion(Class<?> clazz)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/runtime/RuntimeUtils.java', 'methodCount': 5, 'lineCount': 8}], 'asteriskCount': 0, 'importCount': 1, 'javaImportCount': 5}\n",
      "{}\n",
      "/**\n",
      "     * Handle any exception that does not come from the user. Default implementation prints the stack trace.\n",
      "     * @param exception the exception to handle (never an {@link UserException}).\n",
      "     */\n",
      "protected void handleStorageException(final StorageException exception)\n",
      "**************************************\n",
      "protected void handleStorageException(final StorageException exception) {\n",
      "    System.err.println(\"code:      \" + exception.getCode());\n",
      "    System.err.println(\"message:   \" + exception.getMessage());\n",
      "    System.err.println(\"reason:    \" + exception.getReason());\n",
      "    System.err.println(\"location:  \" + exception.getLocation());\n",
      "    System.err.println(\"retryable: \" + exception.isRetryable());\n",
      "    exception.printStackTrace();\n",
      "}\n",
      "3156\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  96%|| 1199/1243 [41:12<01:33,  2.12s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51914\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1200/1243 [41:14<01:32,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51934\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1201/1243 [41:16<01:29,  2.14s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51947\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1202/1243 [41:18<01:29,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "51948\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1203/1243 [41:20<01:27,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52333\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ParamUtils {\\n    public static final double INV_LOG_2 = 1 / Math.log(2.0);;\\n    private ParamUtils();\\n    public static double inRange(final double val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final long min, final long max, final String message)\\n    public static int inRange(final int val, final int min, final int max, final String message)\\n    public static double isPositiveOrZero(final double val, final String message)\\n    public static double isNegativeOrZero(final double val, final String message)\\n    public static long isPositiveOrZero(final long val, final String message)\\n    public static int isPositiveOrZero(final int val, final String message)\\n    public static long isPositive(final long val, final String message)\\n    public static int isPositive(final int val, final String message)\\n    public static double isPositive(final double val, final String message)\\n    public static double isFinite(final double val, final String message)\\n    public static double logb(final double val, final double base)\\n    public static double log2(final double val)\\n    public static void writeStringListToFile(final Iterable<String> stringIterator, final File outputFile)\\n    public static void writeValuesToFile(final double[] data, final File outputFile)\\n    public static double[] readValuesFromFile(final File inputFile)\\n    public static double inRange(final DoubleRange validRange, final double value, final String definition)\\n    public static int inRange(final IntRange validRange, final int value, final String definition)\\n    public static void isValidArrayOffsetAndRangeLength(final int from, final int length, final int arrayLength, final String definition)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/param/ParamUtils.java', 'methodCount': 21, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 3}\n",
      "{'FuncotationFiltrationRule': 'public interface FuncotationFiltrationRule {\\n     boolean checkRule(final Set<Map.Entry<String, String>> funcotations, VariantContext variant)\\n}\\n'}\n",
      "/**\n",
      "     * Build a {@link FuncotationFiltrationRule} matching Funcotations from variants with a\n",
      "     * maximum MAF less than some threshold.\n",
      "     *\n",
      "     * @param maxMaf the MAF threshold to check in the rule. Must be in the range [0, 1]\n",
      "     * @param afDataSource the allele frequency data source (ExAC or gnomAD) with which the original VCF was Funcotated.\n",
      "     * @return a {@link FuncotationFiltrationRule} matching Funcotations with a MAF\n",
      "     *         less than {@code maxMaf} across all sub-populations.\n",
      "     */\n",
      "public static FuncotationFiltrationRule buildMaxMafRule(final double maxMaf, final AlleleFrequencyDataSource afDataSource)\n",
      "**************************************\n",
      "/**\n",
      " * Build a {@link FuncotationFiltrationRule} matching Funcotations from variants with a\n",
      " * maximum MAF less than some threshold.\n",
      " *\n",
      " * @param maxMaf the MAF threshold to check in the rule. Must be in the range [0, 1]\n",
      " * @param afDataSource the allele frequency data source (ExAC or gnomAD) with which the original VCF was Funcotated.\n",
      " * @return a {@link FuncotationFiltrationRule} matching Funcotations with a MAF\n",
      " *         less than {@code maxMaf} across all sub-populations.\n",
      " */\n",
      "public static FuncotationFiltrationRule buildMaxMafRule(final double maxMaf, final AlleleFrequencyDataSource afDataSource) {\n",
      "    ParamUtils.inRange(maxMaf, 0, 1, \"MAF must be between 0 and 1\");\n",
      "    if (afDataSource.equals(AlleleFrequencyDataSource.exac)) {\n",
      "        return (funcotations, variant) -> AlleleFrequencyExacUtils.getMaxMinorAlleleFreq(funcotations) <= maxMaf;\n",
      "    } else {\n",
      "        return (funcotations, variant) -> {\n",
      "            final Map<String, List<String>> condensedFuncotations = funcotations.stream().collect(Collectors.groupingBy(Map.Entry::getKey, Collectors.mapping(Map.Entry::getValue, Collectors.toList())));\n",
      "            return (!AlleleFrequencyGnomadUtils.allFrequenciesFiltered(condensedFuncotations) && AlleleFrequencyGnomadUtils.getMaxMinorAlleleFreq(condensedFuncotations) <= maxMaf);\n",
      "        };\n",
      "    }\n",
      "}\n",
      "1041\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1204/1243 [41:23<01:26,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52807\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1205/1243 [41:25<01:23,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "52820\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1206/1243 [41:27<01:16,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53018\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1207/1243 [41:29<01:15,  2.08s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53019\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1208/1243 [41:31<01:15,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53231\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}, {'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class ParamUtils {\\n    public static final double INV_LOG_2 = 1 / Math.log(2.0);;\\n    private ParamUtils();\\n    public static double inRange(final double val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final double min, final double max, final String message)\\n    public static long inRange(final long val, final long min, final long max, final String message)\\n    public static int inRange(final int val, final int min, final int max, final String message)\\n    public static double isPositiveOrZero(final double val, final String message)\\n    public static double isNegativeOrZero(final double val, final String message)\\n    public static long isPositiveOrZero(final long val, final String message)\\n    public static int isPositiveOrZero(final int val, final String message)\\n    public static long isPositive(final long val, final String message)\\n    public static int isPositive(final int val, final String message)\\n    public static double isPositive(final double val, final String message)\\n    public static double isFinite(final double val, final String message)\\n    public static double logb(final double val, final double base)\\n    public static double log2(final double val)\\n    public static void writeStringListToFile(final Iterable<String> stringIterator, final File outputFile)\\n    public static void writeValuesToFile(final double[] data, final File outputFile)\\n    public static double[] readValuesFromFile(final File inputFile)\\n    public static double inRange(final DoubleRange validRange, final double value, final String definition)\\n    public static int inRange(final IntRange validRange, final int value, final String definition)\\n    public static void isValidArrayOffsetAndRangeLength(final int from, final int length, final int arrayLength, final String definition)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/param/ParamUtils.java', 'methodCount': 21, 'lineCount': 24}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'GCBiasCorrector': \"public class GCBiasCorrector {\\n    private static final Logger logger = LogManager.getLogger(GCBiasCorrector.class);;\\n    // GC bins are 0%, 1% . . . 100%\\nprivate static final int NUMBER_OF_GC_BINS = 101;;\\n    // scale (in units of GC content from 0 to 1) over which GC bias correlation decays\\n// i.e. the bias at GC content = 0.3 and at 0.2 are correlated ~exp(-0.1 / correlationLength)\\n// this is effectively a smoothness parameter used for regularizing the GC-bias estimate for\\n// GC content bins that contain few genomic intervals)\\nprivate static final double correlationLength = 0.02;;\\n    private static final double correlationDecayRatePerBin = 1.0 / (correlationLength * NUMBER_OF_GC_BINS);;\\n    // multiply by these to get a GC-bias correction as a function of GC\\nprivate final double[] gcCorrectionFactors;;\\n    // Apache Commons median doesn't work on empty arrays; this value is a placeholder to avoid exceptions\\nprivate static final double DUMMY_VALUE_NEVER_USED = 1.0;;\\n    private GCBiasCorrector(final RealVector readCounts, final double[] intervalGCContent);\\n    public static void correctGCBias(final RealMatrix readCounts, final double[] intervalGCContent)\\n    private double[] calculateCorrectionFactors(final List<List<Double>> readCountsByGC)\\n    private double correctedCoverage(final double readCount, final double gcContent)\\n    private static double medianOrDefault(final List<Double> list)\\n    private static int gcContentToBinIndex(final double gcContent)\\n}\\n\"}\n",
      "/**\n",
      "     * As described above, calculate medians of each GC bin and convolve with an exponential kernel.\n",
      "     *\n",
      "     * @param readCountsByGC integer read counts for each GC bin\n",
      "     * @return multiplicative correction factors for each GC bin\n",
      "     */\n",
      "private double[] calculateCorrectionFactors(final List<List<Double>> readCountsByGC)\n",
      "**************************************\n",
      "private double[] calculateCorrectionFactors(final List<List<Double>> readCountsByGC) {\n",
      "    final RealVector medians = new ArrayRealVector(readCountsByGC.stream().mapToDouble(GCBiasCorrector::medianOrDefault).toArray());\n",
      "    return IntStream.range(0, NUMBER_OF_GC_BINS).mapToDouble(bin -> {\n",
      "        final RealVector weights = new ArrayRealVector(IntStream.range(0, NUMBER_OF_GC_BINS).mapToDouble(n -> readCountsByGC.get(n).size() * Math.exp(-Math.abs(bin - n) * correlationDecayRatePerBin)).toArray());\n",
      "        return weights.dotProduct(medians) / weights.getL1Norm();\n",
      "    }).map(x -> 1 / x).toArray();\n",
      "}\n",
      "2817\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1209/1243 [41:33<01:14,  2.18s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "53489\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1210/1243 [42:07<06:24, 11.66s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54361\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  97%|| 1211/1243 [42:09<04:43,  8.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54820\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1212/1243 [42:12<03:33,  6.87s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54939\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1213/1243 [42:14<02:45,  5.52s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54943\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 1}\n",
      "{'Kmer': 'public class Kmer {\\n    // this values may be updated in the course of interacting with this kmer\\nprivate byte[] bases;;\\n    private int start;;\\n    // two constants\\nprivate final int length;;\\n    private final int hash;;\\n    public Kmer(final byte[] kmer);\\n    public Kmer(final String kmer);\\n    public Kmer(final byte[] bases, final int start, final int length);\\n    private static int hashCode(final byte[] bases, final int start, final int length)\\n    public Kmer subKmer(final int newStart, final int newLength)\\n    public byte[] bases()\\n    public int length()\\n    public int getDifferingPositions(final Kmer other, final int maxDistance, final int[] differingIndeces, final byte[] differingBases)\\n    public String toString()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n     int start()\\n}\\n'}\n",
      "/**\n",
      "     * Gets a set of differing positions and bases from another k-mer, limiting up to a max distance.\n",
      "     * For example, if this = \"ACATT\" and other = \"ACGGT\":\n",
      "     * - if maxDistance < 2 then -1 will be returned, since distance between kmers is 2.\n",
      "     * - If maxDistance >=2, then 2 will be returned, and arrays will be filled as follows:\n",
      "     * differingIndeces = {2,3}\n",
      "     * differingBases = {'G','G'}\n",
      "     * @param other                 Other k-mer to test\n",
      "     * @param maxDistance           Maximum distance to search. If this and other k-mers are beyond this Hamming distance,\n",
      "     *                              search is aborted and -1 is returned\n",
      "     * @param differingIndeces      Array with indices of differing bytes in array\n",
      "     * @param differingBases        Actual differing bases\n",
      "     * @return                      Set of mappings of form (int->byte), where each elements represents index\n",
      "     *                              of k-mer array where bases mismatch, and the byte is the base from other kmer.\n",
      "     *                              If both k-mers differ by more than maxDistance, returns null\n",
      "     */\n",
      "public int getDifferingPositions(final Kmer other, final int maxDistance, final int[] differingIndeces, final byte[] differingBases)\n",
      "**************************************\n",
      "public int getDifferingPositions(final Kmer other, final int maxDistance, final int[] differingIndeces, final byte[] differingBases) {\n",
      "    Utils.nonNull(other);\n",
      "    Utils.nonNull(differingIndeces);\n",
      "    Utils.nonNull(differingBases);\n",
      "    Utils.validateArg(maxDistance > 0, \"maxDistance must be positive but was \" + maxDistance);\n",
      "    int dist = 0;\n",
      "    if (length == other.length()) {\n",
      "        final byte[] f2 = other.bases;\n",
      "        for (int i = 0; i < length; i++) {\n",
      "            if (bases[start + i] != f2[i]) {\n",
      "                differingIndeces[dist] = i;\n",
      "                differingBases[dist++] = f2[i];\n",
      "                if (dist > maxDistance) {\n",
      "                    return -1;\n",
      "                }\n",
      "            }\n",
      "        }\n",
      "    }\n",
      "    return dist;\n",
      "}\n",
      "2446\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1214/1243 [42:16<02:10,  4.49s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55169\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1215/1243 [42:18<01:43,  3.69s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55191\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1216/1243 [42:20<01:23,  3.10s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55209\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1217/1243 [42:21<01:10,  2.71s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55210\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1218/1243 [42:23<01:01,  2.45s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55220\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class UserException {\\n    private static final long serialVersionUID = 0L;;\\n    public UserException();\\n    public UserException(final String msg);\\n    public UserException(final String message, final Throwable throwable);\\n    protected static String getMessage(final Throwable t)\\n    public static final MissingReference MISSING_REFERENCE_FOR_CRAM = new MissingReference(\"A reference file is required when using CRAM files.\", true);;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/UserException.java', 'methodCount': 4, 'lineCount': 8}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 6}\n",
      "{}\n",
      "/**\n",
      "     * Remove all vertices on the graph that cannot be accessed by following any edge,\n",
      "     * regardless of its direction, from the reference source vertex\n",
      "     */\n",
      "public final void removeVerticesNotConnectedToRefRegardlessOfEdgeDirection()\n",
      "**************************************\n",
      "public final void removeVerticesNotConnectedToRefRegardlessOfEdgeDirection() {\n",
      "    final Collection<V> toRemove = new HashSet<>(vertexSet());\n",
      "    final V refV = getReferenceSourceVertex();\n",
      "    if (refV != null) {\n",
      "        for (final V v : new BaseGraphIterator<>(this, refV, true, true)) {\n",
      "            toRemove.remove(v);\n",
      "        }\n",
      "    }\n",
      "    removeAllVertices(toRemove);\n",
      "}\n",
      "2041\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1219/1243 [42:26<00:57,  2.38s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55229\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1220/1243 [42:27<00:49,  2.17s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55230\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1221/1243 [42:29<00:44,  2.04s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55273\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1222/1243 [42:31<00:41,  1.96s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55644\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1223/1243 [42:32<00:38,  1.90s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "55649\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class FeatureInput {\\n    final protected static Logger logger = LogManager.getLogger(FeatureInput.class);;\\n    private static final long serialVersionUID = 1L;;\\n    /**\\n * File containing Features as specified by the user on the command line\\n */\\n/**\\n * Cache the codec for this feature input the first time we discover it, so we only do it once\\n */\\nprivate transient Class<FeatureCodec<T, ?>> featureCodecClass;;\\n    /**\\n * Delimiter between the logical name and the file name in the --argument_name logical_name:feature_file syntax\\n */\\npublic static final String FEATURE_ARGUMENT_TAG_DELIMITER = \":\";;\\n     FeatureInput(final String rawArgumentValue);\\n    public FeatureInput(final GATKPath gatkPath);\\n    public FeatureInput(final GATKPath gatkPath, final String name);\\n     FeatureInput(final String featurePath, final String name);\\n    public FeatureInput(final String rawInputSpecifier, final String name, final Map<String, String> keyValueMap);\\n    public void setFeatureCodecClass(final Class<FeatureCodec<T, ?>> featureCodecClass)\\n    public Class<FeatureCodec<T, ?>> getFeatureCodecClass()\\n    private String makeIntoAbsolutePath()\\n    public String getAttribute(final String key)\\n    public String getName()\\n    public boolean hasUserSuppliedName()\\n    public String getFeaturePath()\\n    public int hashCode()\\n    public boolean equals(final Object other)\\n    public String toString()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/FeatureInput.java', 'methodCount': 15, 'lineCount': 21}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Given a map of sample to contamination fraction, determines whether any samples have\n",
      "     * a non-zero contamination fraction set.\n",
      "     *\n",
      "     * @param contaminationMap sample -> contamination fraction map to test\n",
      "     * @return true if at least one sample has a non-zero contamination fraction, otherwise false\n",
      "     */\n",
      "private boolean contaminationIsPresentInMap(final Map<String, Double> contaminationMap)\n",
      "**************************************\n",
      "private boolean contaminationIsPresentInMap(final Map<String, Double> contaminationMap) {\n",
      "    if (contaminationMap == null) {\n",
      "        return false;\n",
      "    }\n",
      "    for (final Map.Entry<String, Double> mapEntry : contaminationMap.entrySet()) {\n",
      "        if (contaminationFractionIsSet(mapEntry.getValue())) {\n",
      "            return true;\n",
      "        }\n",
      "    }\n",
      "    return false;\n",
      "}\n",
      "2295\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  98%|| 1224/1243 [42:35<00:38,  2.03s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56106\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1225/1243 [42:37<00:38,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56436\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1226/1243 [42:40<00:37,  2.23s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56476\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1227/1243 [42:42<00:36,  2.26s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56477\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1228/1243 [42:44<00:34,  2.28s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56483\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class ReadFilter {\\n    protected SAMFileHeader samHeader = null;;\\n    public void setHeader(SAMFileHeader samHeader)\\n    private static final long serialVersionUID = 1L;;\\n    public static ReadFilter fromList(final List<ReadFilter> readFilters, final SAMFileHeader samHeader)\\n    public ReadFilter and(ReadFilter other)\\n    public ReadFilter or(ReadFilter other)\\n    public ReadFilter negate()\\n    public abstract boolean test(GATKRead read)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilter.java', 'methodCount': 6, 'lineCount': 10}, {'fieldCount': 33, 'classCount': 1, 'methodsOnly': 'public class ReadFilterLibrary {\\n    private ReadFilterLibrary();\\n    /**\\n * Static, stateless read filter instances\\n */\\npublic static final AllowAllReadsReadFilter ALLOW_ALL_READS = new AllowAllReadsReadFilter();;\\n    public static final CigarContainsNoNOperator CIGAR_CONTAINS_NO_N_OPERATOR = new CigarContainsNoNOperator();;\\n    public static final FirstOfPairReadFilter FIRST_OF_PAIR = new FirstOfPairReadFilter();;\\n    public static final GoodCigarReadFilter GOOD_CIGAR = new GoodCigarReadFilter();;\\n    public static final HasReadGroupReadFilter HAS_READ_GROUP = new HasReadGroupReadFilter();;\\n    public static final MappedReadFilter MAPPED = new MappedReadFilter();;\\n    public static final MappingQualityAvailableReadFilter MAPPING_QUALITY_AVAILABLE = new MappingQualityAvailableReadFilter();;\\n    public static final MappingQualityNotZeroReadFilter MAPPING_QUALITY_NOT_ZERO = new MappingQualityNotZeroReadFilter();;\\n    public static final MatchingBasesAndQualsReadFilter HAS_MATCHING_BASES_AND_QUALS = new MatchingBasesAndQualsReadFilter();;\\n    public static final MateOnSameContigOrNoMappedMateReadFilter MATE_ON_SAME_CONTIG_OR_NO_MAPPED_MATE = new MateOnSameContigOrNoMappedMateReadFilter();;\\n    public static final MateDifferentStrandReadFilter MATE_DIFFERENT_STRAND = new MateDifferentStrandReadFilter();;\\n    public static final MateDistantReadFilter MATE_DISTANT = new MateDistantReadFilter();;\\n    public static final NonZeroReferenceLengthAlignmentReadFilter NON_ZERO_REFERENCE_LENGTH_ALIGNMENT = new NonZeroReferenceLengthAlignmentReadFilter();;\\n    public static final NonZeroFragmentLengthReadFilter NONZERO_FRAGMENT_LENGTH_READ_FILTER = new NonZeroFragmentLengthReadFilter();;\\n    public static final NotDuplicateReadFilter NOT_DUPLICATE = new NotDuplicateReadFilter();;\\n    public static final NotProperlyPairedReadFilter NOT_PROPERLY_PAIRED = new NotProperlyPairedReadFilter();;\\n    public static final NotSecondaryAlignmentReadFilter NOT_SECONDARY_ALIGNMENT = new NotSecondaryAlignmentReadFilter();;\\n    public static final NotSupplementaryAlignmentReadFilter NOT_SUPPLEMENTARY_ALIGNMENT = new NotSupplementaryAlignmentReadFilter();;\\n    public static final PairedReadFilter PAIRED = new PairedReadFilter();;\\n    public static final ProperlyPairedReadFilter PROPERLY_PAIRED = new ProperlyPairedReadFilter();;\\n    public static final PassesVendorQualityCheckReadFilter PASSES_VENDOR_QUALITY_CHECK = new PassesVendorQualityCheckReadFilter();;\\n    public static final PrimaryLineReadFilter PRIMARY_LINE = new PrimaryLineReadFilter();;\\n    public static final ReadLengthEqualsCigarLengthReadFilter READLENGTH_EQUALS_CIGARLENGTH = new ReadLengthEqualsCigarLengthReadFilter();;\\n    public static final SecondOfPairReadFilter SECOND_OF_PAIR = new SecondOfPairReadFilter();;\\n    public static final SeqIsStoredReadFilter SEQ_IS_STORED = new SeqIsStoredReadFilter();;\\n    public static final ValidAlignmentStartReadFilter VALID_ALIGNMENT_START = new ValidAlignmentStartReadFilter();;\\n    public static final ValidAlignmentEndReadFilter VALID_ALIGNMENT_END = new ValidAlignmentEndReadFilter();;\\n    public static final NonChimericOriginalAlignmentReadFilter NON_CHIMERIC_ORIGINAL_ALIGNMENT_READ_FILTER = new NonChimericOriginalAlignmentReadFilter();;\\n    public static final MateUnmappedAndUnmappedReadFilter MATE_UNMAPPED_AND_UNMAPPED_READ_FILTER = new MateUnmappedAndUnmappedReadFilter();;\\n    public static final ReadGroupHasFlowOrderReadFilter READ_GROUP_HAS_FLOW_ORDER_READ_FILTER = new ReadGroupHasFlowOrderReadFilter();;\\n    public static final HmerQualitySymetricReadFilter HMER_QUALITY_SYMETRIC_READ_FILTER = new HmerQualitySymetricReadFilter();;\\n    public static final FlowBasedTPAttributeValidReadFilter FLOW_BASED_TP_ATTRIBUTE_VALID_READ_FILTER = new FlowBasedTPAttributeValidReadFilter();;\\n    public static final FlowBasedTPAttributeSymetricReadFilter FLOW_BASED_TP_ATTRIBUTE_SYMETRIC_READ_FILTER = new FlowBasedTPAttributeSymetricReadFilter();;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/engine/filters/ReadFilterLibrary.java', 'methodCount': 1, 'lineCount': 36}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class Utils {\\n    /**\\n * Comparator for strings that sorts null first;\\n */\\npublic static final Comparator<? super String> COMPARE_STRINGS_NULLS_FIRST = Comparator.nullsFirst(Comparator.naturalOrder());;\\n    private Utils();\\n    private final static DateTimeFormatter longDateTimeFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.LONG);;\\n    /**\\n *  Static random number generator and seed.\\n */\\nprivate static final long GATK_RANDOM_SEED = 47382911L;;\\n    private static final Random randomGenerator = new Random(GATK_RANDOM_SEED);;\\n    private static final RandomDataGenerator randomDataGenerator = new RandomDataGenerator(new Well19937c(GATK_RANDOM_SEED));;\\n    public static Random getRandomGenerator()\\n    public static RandomDataGenerator getRandomDataGenerator()\\n    public static void resetRandomGenerator()\\n    private static final int TEXT_WARNING_WIDTH = 68;;\\n    private static final String TEXT_WARNING_PREFIX = \"* \";;\\n    private static final String TEXT_WARNING_BORDER = StringUtils.repeat(\\'*\\', TEXT_WARNING_PREFIX.length() + TEXT_WARNING_WIDTH);;\\n    private static final char ESCAPE_CHAR = \\'\\\\u001B\\';;\\n    public static final float JAVA_DEFAULT_HASH_LOAD_FACTOR = 0.75f;;\\n    private static final Logger logger = LogManager.getLogger(Utils.class);;\\n    public static List<T> cons(final T elt, final List<T> l)\\n    public static void warnUser(final String msg)\\n    public static void warnUser(final Logger logger, final String msg)\\n    public static List<String> warnUserLines(final String msg)\\n    private static void prettyPrintWarningMessage(final List<String> results, final String message)\\n    private static int getLastSpace(final CharSequence message, final int width)\\n    public static String join(final CharSequence separator, final Object... objects)\\n    public static String join(final String separator, final int[] ints)\\n    public static String join(final String separator, final double[] doubles)\\n    public static String join(final String separator, final Collection<T> objects)\\n    public static byte[] concat(final byte[]... allBytes)\\n    public static T[] concat(final T[] a, final T[] b, final IntFunction<T[]> constructor)\\n    public static byte[] concat(final byte[] a, final byte[] b)\\n    public static List<Integer> asList(final int... values)\\n    public static List<Double> asList(final double... values)\\n    public static List<T> append(final List<T> left, final T... elts)\\n    public static String dupChar(final char c, final int nCopies)\\n    public static String dupString(final String s, int nCopies)\\n    public static byte[] dupBytes(final byte b, final int nCopies)\\n    public static int countBooleanOccurrences(final boolean element, final boolean[] array)\\n    public static String[] escapeExpressions(final String args)\\n    private static String[] escapeExpressions(final String args, final String delimiter)\\n    public static byte[] repeatChars(final char c, final int n)\\n    public static byte[] repeatBytes(final byte b, final int n)\\n    public static List<List<T>> makePermutations(final List<T> objects, final int n, final boolean withReplacement)\\n    public static String calcMD5(final String s)\\n    public static String calcMD5(final byte[] bytes)\\n    public static String calculateFileMD5(final File file) throws IOException\\n    public static String calculatePathMD5(final Path path) throws IOException\\n    private static String MD5ToString(final byte[] bytes)\\n    public static T nonNull(final T object)\\n    public static T nonNull(final T object, final String message)\\n    public static T nonNull(final T object, final Supplier<String> message)\\n    public static T nonEmpty(T collection, String message)\\n    public static boolean isNonEmpty(Collection<?> collection)\\n    public static String nonEmpty(String string, String message)\\n    public static String nonEmpty(final String string)\\n    public static T nonEmpty(T collection)\\n    public static void containsNoNull(final Collection<?> collection, final String message)\\n    public static Set<E> checkForDuplicatesAndReturnSet(final Collection<E> c, final String message)\\n    public static int validIndex(final int index, final int length)\\n    public static int validIndex(final int index, final int length, final String errorMessage)\\n    public static void validateArg(final boolean condition, final String msg)\\n    public static void validateArg(final boolean condition, final Supplier<String> msg)\\n    public static void validate(final boolean condition, final String msg)\\n    public static void validate(final boolean condition, final Supplier<String> msg)\\n    public static int optimumHashSize(final int maxElements)\\n    public static boolean equalRange(final byte[] left, final int leftOffset, final byte[] right, final int rightOffset, final int length)\\n    private static void validRange(final int length, final int offset, final int size, final String msg)\\n    public static T skimArray(final T original, final boolean[] remove)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset)\\n    public static T skimArray(final T source, final int sourceOffset, final T dest, final int destOffset, final boolean[] remove, final int removeOffset, final int length)\\n    private static T skimArrayBuildResultArray(final T dest, final int destOffset, final Class<T> destClazz, final int newLength)\\n    private static Class<T> skimArrayDetermineDestArrayClass(final T dest, final Class<T> sourceClazz)\\n    public static void warnOnNonIlluminaReadGroups(final SAMFileHeader readsHeader, final Logger logger)\\n    public static boolean xor(final boolean x, final boolean y)\\n    public static int lastIndexOf(final byte[] reference, final byte[] query)\\n    public static List<Integer> listFromPrimitives(final int[] ar)\\n    public static Iterator<T> concatIterators(final Iterator<? extends Iterable<T>> iterator)\\n    public static Stream<T> stream(final Enumeration<T> enumeration)\\n    public static Stream<T> stream(final Iterable<T> iterable)\\n    public static Stream<T> stream(final Iterator<T> iterator)\\n    public static Function<T, T> identityFunction()\\n    public static Iterator<T> transformParallel(final Iterator<F> fromIterator, final Function<F, T> function, final int numThreads)\\n    public static Set<T> getDuplicatedItems(final Collection<T> objects)\\n    public static String getDateTimeForDisplay(final ZonedDateTime dateTime)\\n    public static void forceJVMLocaleToUSEnglish()\\n    public static T getMedianValue(List<T> values)\\n    public static List<String> split(final String str, final char delimiter)\\n    public static List<String> split(final String str, final String delimiter)\\n    private static List<String> split(final String str, final String delimiter, final int expectedNumTokens)\\n    private static void removeTrailingEmptyStringsFromEnd(final List<String> result)\\n    public static Map<U, Set<T>> getReverseValueToListMap(final Map<T, List<U>> somethingToListMap)\\n    public static String formattedPercent(final long x, final long total)\\n    public static String formattedRatio(final long num, final long denom)\\n    public static Set<String> filterCollectionByExpressions(final Collection<String> sourceValues, final Collection<String> filterExpressions, final boolean exactMatch)\\n    private static Collection<Pattern> compilePatterns(final Collection<String> filters)\\n    public static T runInParallel(final int threads, final Supplier<T> supplier)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/Utils.java', 'methodCount': 82, 'lineCount': 95}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{'ExampleMultiMetricsArgumentCollection': 'public class ExampleMultiMetricsArgumentCollection {\\n    public static final long serialVersionUID = 1L;;\\n    //This example collector only uses the standard argument collection for selecting a collection level.\\n@ArgumentCollection\\nMetricAccumulationLevelArgumentCollection metricAccumulationLevel = new MetricAccumulationLevelArgumentCollection();;\\n}\\n'}\n",
      "/**\n",
      "     * @param inputArgs ExampleMultiMetricsArgumentCollection populated with argument values. May not be null.\n",
      "     * @param samHeader samHeader for the input to be processed. May not be null.\n",
      "     */\n",
      "public void initialize(final ExampleMultiMetricsArgumentCollection inputArgs, final SAMFileHeader samHeader)\n",
      "**************************************\n",
      "/**\n",
      " * @param inputArgs ExampleMultiMetricsArgumentCollection populated with argument values. May not be null.\n",
      " * @param samHeader samHeader for the input to be processed. May not be null.\n",
      " */\n",
      "public void initialize(final ExampleMultiMetricsArgumentCollection inputArgs, final SAMFileHeader samHeader) {\n",
      "    Utils.nonNull(inputArgs);\n",
      "    Utils.nonNull(samHeader);\n",
      "    this.inputArgs = inputArgs;\n",
      "    // call setup to create the level distributors\n",
      "    setup(inputArgs.metricAccumulationLevel.accumulationLevels, samHeader.getReadGroups());\n",
      "}\n",
      "3175\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1229/1243 [42:47<00:34,  2.44s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56486\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1230/1243 [42:50<00:32,  2.49s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56489\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1231/1243 [42:52<00:29,  2.43s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56499\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1232/1243 [42:54<00:24,  2.21s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56604\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1233/1243 [42:55<00:20,  2.07s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57263\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 0}\n",
      "{}\n",
      "/**\n",
      "     * Computes number of reads filtered at each step and total final reads.\n",
      "     */\n",
      "public void computeDerivedMetrics()\n",
      "**************************************\n",
      "/**\n",
      " * Computes number of reads filtered at each step and total final reads.\n",
      " */\n",
      "public void computeDerivedMetrics() {\n",
      "    if (PRIMARY_READS == null || READS_AFTER_PREALIGNED_HOST_FILTER == null || READS_AFTER_QUALITY_AND_COMPLEXITY_FILTER == null || READS_AFTER_HOST_FILTER == null || READS_AFTER_DEDUPLICATION == null || FINAL_PAIRED_READS == null) {\n",
      "        throw new IllegalStateException(\"Cannot compute metrics if primary, pre-aligned host, quality, host, duplicate, or final paired read counts are not initialized\");\n",
      "    }\n",
      "    FINAL_TOTAL_READS = READS_AFTER_DEDUPLICATION;\n",
      "    FINAL_UNPAIRED_READS = FINAL_TOTAL_READS - FINAL_PAIRED_READS;\n",
      "    LOW_QUALITY_OR_LOW_COMPLEXITY_READS_FILTERED = READS_AFTER_PREALIGNED_HOST_FILTER - READS_AFTER_QUALITY_AND_COMPLEXITY_FILTER;\n",
      "    HOST_READS_FILTERED = PRIMARY_READS - READS_AFTER_PREALIGNED_HOST_FILTER + READS_AFTER_QUALITY_AND_COMPLEXITY_FILTER - READS_AFTER_HOST_FILTER;\n",
      "    DUPLICATE_READS_FILTERED = READS_AFTER_HOST_FILTER - READS_AFTER_DEDUPLICATION;\n",
      "}\n",
      "299\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1234/1243 [42:58<00:19,  2.16s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57280\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1235/1243 [43:00<00:17,  2.20s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57281\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows:  99%|| 1236/1243 [43:02<00:15,  2.24s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57304\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1237/1243 [43:05<00:13,  2.19s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57319\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1238/1243 [43:07<00:11,  2.39s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57461\n",
      "{'classCount': 0, 'imports': [{'fieldCount': 1, 'classCount': 1, 'methodsOnly': 'public class GATKException {\\n    private static final long serialVersionUID = 0L;;\\n    public GATKException(String msg);\\n    public GATKException(String message, Throwable throwable);\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/exceptions/GATKException.java', 'methodCount': 2, 'lineCount': 5}, {'fieldCount': 11, 'classCount': 1, 'methodsOnly': 'public class SvType {\\n    public static final int NO_APPLICABLE_END = -1;;\\n    public static final int NO_APPLICABLE_LEN = -1;;\\n    // fields whose name starts with \"variant\" are following the VCF spec, i.e. they can be used directly to construct VariantContextBuilder\\nprotected final String variantChromosome;;\\n    protected final int variantStart;;\\n    protected final int variantStop;;\\n    protected final String variantId;;\\n    protected final Allele refAllele;;\\n    protected final Allele altAllele;;\\n    protected final int svLen;;\\n    protected final Map<String, Object> typeSpecificAttributes;;\\n    protected static final Map<String, Object> noExtraAttributes = Collections.emptyMap();;\\n    public SvType(final String variantChromosome, final int variantStart, final int variantStop, final String variantId, final Allele refAllele, final Allele altAllele, final int svLen, final Map<String, Object> typeSpecificAttributes);\\n    public final VariantContextBuilder getBasicInformation()\\n    public String getVariantChromosome()\\n    public int getVariantStart()\\n    public int getVariantStop()\\n    public final String getInternalVariantId()\\n     Allele getRefAllele()\\n    public final Allele getAltAllele()\\n    public final int getSVLength()\\n    public final Map<String, Object> getTypeSpecificAttributes()\\n    public abstract boolean hasApplicableEnd()\\n    public abstract boolean hasApplicableLength()\\n    public boolean equals(final Object o)\\n    public int hashCode()\\n    public static SortedSet<String> getKnownTypes()\\n    public static String makeLocationString(final String chr1, final int pos1, final String chr2, final int pos2)\\n    public static String makeLocationString(final NovelAdjacencyAndAltHaplotype novelAdjacencyAndAltHaplotype)\\n     static byte[] extractRefBases(final SimpleInterval interval, final BasicReference reference)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvType.java', 'methodCount': 18, 'lineCount': 31}, {'fieldCount': 4, 'classCount': 1, 'methodsOnly': 'public class IntervalUtils {\\n    private static final Logger log = LogManager.getLogger(IntervalUtils.class);;\\n    /**\\n * Recognized extensions for interval files\\n */\\npublic static final List<String> GATK_INTERVAL_FILE_EXTENSIONS = Collections.unmodifiableList(Arrays.asList(\".list\", \".intervals\"));;\\n    /**\\n * Lexicographical (contig) order comparator.\\n * <p>\\n * Intervals from different contigs order is according their enclosing\\n * contigs name ascending lexicographical order.\\n * </p>\\n * <p>\\n * Intervals from the same contigs order is according to their start\\n * position ascending numerical order, and, in case of a tie, the stop position\\'s.\\n * </p>\\n * <p>\\n * The {@code null} contig is supported and comes last.\\n * </p>\\n */\\npublic static final Comparator<Locatable> LEXICOGRAPHICAL_ORDER_COMPARATOR = Comparator.comparing(Locatable::getContig, Comparator.nullsLast(String::compareTo)).thenComparingInt(Locatable::getStart).thenComparingInt(Locatable::getEnd);;\\n    private static final Logger logger = LogManager.getLogger(IntervalUtils.class);;\\n    public static final int compareLocatables(Locatable first, Locatable second, SAMSequenceDictionary dictionary)\\n    public static boolean isBefore(final Locatable first, final Locatable second, final SAMSequenceDictionary dictionary)\\n    public static boolean isAfter(final Locatable first, final Locatable second, final SAMSequenceDictionary dictionary)\\n    public static int compareContigs(final Locatable first, final Locatable second, final SAMSequenceDictionary dictionary)\\n    public static SimpleInterval getSpanningInterval(final List<? extends Locatable> locations)\\n    public static List<SimpleInterval> convertGenomeLocsToSimpleIntervals(final List<GenomeLoc> genomeLocIntervals)\\n    public static QueryInterval convertSimpleIntervalToQueryInterval(final SimpleInterval interval, final SAMSequenceDictionary sequenceDictionary)\\n    public static GenomeLocSortedSet loadIntervals(final List<String> intervalStrings, final IntervalSetRule intervalSetRule, final IntervalMergingRule intervalMergingRule, final int padding, final GenomeLocParser genomeLocParser)\\n    public static List<GenomeLoc> loadIntervalsNonMerging(final List<String> intervalStrings, final int padding, final GenomeLocParser genomeLocParser)\\n    public static List<GenomeLoc> parseIntervalArguments(final GenomeLocParser parser, final List<String> argList)\\n    public static List<GenomeLoc> parseIntervalArguments(final GenomeLocParser parser, final String arg)\\n    public static List<GenomeLoc> featureFileToIntervals(final GenomeLocParser parser, final String featureFile)\\n    private static List<GenomeLoc> gatkIntervalFileToList(final GenomeLocParser glParser, final String fileName)\\n    public static List<GenomeLoc> mergeListsBySetOperator(final List<GenomeLoc> setOne, final List<GenomeLoc> setTwo, final IntervalSetRule rule)\\n    public static GenomeLocSortedSet sortAndMergeIntervals(final GenomeLocParser parser, List<GenomeLoc> intervals, final IntervalMergingRule mergingRule)\\n    public static String equateIntervals(final List<GenomeLoc> masterArg, final List<GenomeLoc> testArg)\\n    private static List<T> reverse(final List<T> l)\\n    public static boolean isGatkIntervalFile(final String str)\\n    public static boolean isGatkIntervalFile(final String str, final boolean checkExists)\\n    public static Map<String, Integer> getContigSizes(final Path reference)\\n    public static void scatterContigIntervals(final SAMFileHeader fileHeader, final List<GenomeLoc> locs, final List<File> scatterParts)\\n    public static List<List<GenomeLoc>> splitIntervalsToSubLists(final List<GenomeLoc> locs, final List<Integer> splits)\\n    public static void scatterFixedIntervals(final SAMFileHeader fileHeader, final List<List<GenomeLoc>> splits, final List<File> scatterParts)\\n    public static List<List<GenomeLoc>> splitFixedIntervals(final List<GenomeLoc> locs, final int numParts)\\n    public static List<List<GenomeLoc>> splitLocusIntervals(final List<GenomeLoc> locs, final int numParts)\\n    private static SplitLocusRecursive splitLocusIntervals1(final LinkedList<GenomeLoc> remaining, final long idealSplitSize)\\n    public static boolean overlaps(final Locatable left, final Locatable right)\\n    public static String locatableToString(Locatable interval)\\n    public static List<GenomeLoc> flattenSplitIntervals(final List<List<GenomeLoc>> splits)\\n    private static void addFixedSplit(final List<Integer> splitPoints, final List<GenomeLoc> locs, final long locsSize, final int startIndex, final int stopIndex, final int numParts)\\n    private static Pair<Integer, Long> getFixedSplit(final List<GenomeLoc> locs, final long locsSize, final int startIndex, final int stopIndex, final int minLocs, final int maxLocs)\\n    private static Interval toInterval(final GenomeLoc loc, final int locIndex)\\n    public static List<GenomeLoc> mergeIntervalLocations(final List<GenomeLoc> raw, final IntervalMergingRule rule)\\n    public static long intervalSize(final List<GenomeLoc> locs)\\n    public static List<GenomeLoc> getIntervalsWithFlanks(final GenomeLocParser parser, final List<GenomeLoc> locs, final int basePairs)\\n    public static List<SimpleInterval> getIntervalsWithFlanks(final List<SimpleInterval> intervals, final int basePairs, final SAMSequenceDictionary dictionary)\\n    public static List<List<SimpleInterval>> groupIntervalsByContig(final List<SimpleInterval> sortedIntervals)\\n    private static LinkedHashMap<String, List<GenomeLoc>> splitByContig(final List<GenomeLoc> sorted)\\n    public static List<GenomeLoc> genomeLocsFromLocatables(final GenomeLocParser parser, final Collection<? extends Locatable> locatables)\\n    public static List<SimpleInterval> getAllIntervalsForReference(final SAMSequenceDictionary sequenceDictionary)\\n    public static List<SimpleInterval> getResolvedIntervals(final String intervalQueryString, final SAMSequenceDictionary sequenceDictionary)\\n    public static SimpleInterval trimIntervalToContig(final String contig, final int start, final int stop, final int contigLength)\\n    public static boolean intervalIsOnDictionaryContig(final SimpleInterval interval, final SAMSequenceDictionary dictionary)\\n    public static List<SimpleInterval> cutToShards(Iterable<SimpleInterval> intervals, int shardSize)\\n    public static int shardIndex(int oneBasedOffset, int shardSize)\\n    public static int beginOfShard(int shardIndex, int shardSize)\\n    public static int endOfShard(int shardIndex, int shardSize)\\n    public static List<SimpleInterval> getSpanningIntervals(final List<? extends Locatable> locations, final SAMSequenceDictionary sequenceDictionary)\\n    public static List<Locatable> combineAndSortBreakpoints(final List<T> unsortedLocatables1, final List<T> unsortedLocatables2, final SAMSequenceDictionary dictionary)\\n    public static void validateNoOverlappingIntervals(List<T> locatables)\\n    public static List<T> sortLocatablesBySequenceDictionary(final Collection<T> locatables, final SAMSequenceDictionary dictionary)\\n    public static Map<T, List<U>> createOverlapMap(final List<T> keys, final List<U> vals, final SAMSequenceDictionary dictionary)\\n    public static Comparator<T> getDictionaryOrderComparator(final SAMSequenceDictionary dictionary)\\n    public static boolean isReciprocalOverlap(final SimpleInterval interval1, final SimpleInterval interval2, final double reciprocalOverlapThreshold)\\n    public static Map<String, List<SimpleInterval>> sortAndMergeIntervals(final Collection<SimpleInterval> input, final SAMSequenceDictionary dictionary, final IntervalMergingRule rule)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/IntervalUtils.java', 'methodCount': 55, 'lineCount': 61}, {'fieldCount': 6, 'classCount': 1, 'methodsOnly': 'public class BucketUtils {\\n    public static final String GCS_PREFIX = GoogleCloudStorageFileSystem.SCHEME + \"://\";;\\n    public static final String HTTP_PREFIX = HttpFileSystemProvider.SCHEME + \"://\";;\\n    public static final String HTTPS_PREFIX = HttpsFileSystemProvider.SCHEME + \"://\";;\\n    public static final String HDFS_SCHEME = \"hdfs\";;\\n    public static final String HDFS_PREFIX = HDFS_SCHEME + \"://\";;\\n    // slashes omitted since hdfs paths seem to only have 1 slash which would be weirder to include than no slashes\\npublic static final String FILE_PREFIX = \"file:\";;\\n    private BucketUtils();\\n    public static boolean isGcsUrl(final String path)\\n    public static boolean isGcsUrl(final GATKPath pathSpec)\\n    public static boolean isEligibleForPrefetching(final GATKPath pathSpec)\\n    public static boolean isEligibleForPrefetching(final java.nio.file.Path path)\\n    private static boolean isEligibleForPrefetching(final String scheme)\\n    public static boolean isHttpUrl(String path)\\n    public static boolean isHadoopUrl(String path)\\n    public static boolean isRemoteStorageUrl(String path)\\n    public static String makeFilePathAbsolute(String path)\\n    public static InputStream openFile(String path)\\n    public static OutputStream createFile(String path)\\n    public static void copyFile(String sourcePath, String destPath) throws IOException\\n    public static void deleteFile(String pathToDelete) throws IOException\\n    public static String getTempFilePath(String prefix, String extension)\\n    public static String randomRemotePath(String stagingLocation, String prefix, String suffix)\\n    public static boolean fileExists(String path)\\n    public static long fileSize(String path) throws IOException\\n    public static long dirSize(final GATKPath pathSpecifier)\\n    public static boolean isFileUrl(String path)\\n    public static String getBucket(String path)\\n    public static String getPathWithoutBucket(String path)\\n    public static void setGlobalNIODefaultOptions(int maxReopens, String requesterProject)\\n    public static java.nio.file.Path getPathOnGcs(String gcsUrl)\\n    public static CloudStorageConfiguration getCloudStorageConfiguration(int maxReopens, String requesterProject)\\n    private static StorageOptions.Builder setGenerousTimeouts(StorageOptions.Builder builder)\\n    public static java.nio.file.FileSystem getAuthenticatedGcs(String projectId, String bucket, byte[] credentials) throws IOException\\n    public static SeekableByteChannel addPrefetcher(final int bufferSizeMB, final SeekableByteChannel channel)\\n    public static Function<SeekableByteChannel, SeekableByteChannel> getPrefetchingWrapper(final int cloudPrefetchBuffer)\\n    public static String createSignedUrlToGcsObject(String path, final long hoursToLive)\\n    public static String bucketPathToPublicHttpUrl(String path)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/gcs/BucketUtils.java', 'methodCount': 31, 'lineCount': 39}], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 8}\n",
      "{}\n",
      "/**\n",
      "     * {@code referenceSequenceDictionary} is required because 2bit Broadcast references currently order their\n",
      "     * sequence dictionaries in a scrambled order, see https://github.com/broadinstitute/gatk/issues/2037.\n",
      "     */\n",
      "public static void writeVCF(final List<VariantContext> localVariants, final String vcfFileName, final SAMSequenceDictionary referenceSequenceDictionary, final Set<VCFHeaderLine> defaultToolVCFHeaderLines, final Logger logger)\n",
      "**************************************\n",
      "/**\n",
      " * {@code referenceSequenceDictionary} is required because 2bit Broadcast references currently order their\n",
      " * sequence dictionaries in a scrambled order, see https://github.com/broadinstitute/gatk/issues/2037.\n",
      " */\n",
      "public static void writeVCF(final List<VariantContext> localVariants, final String vcfFileName, final SAMSequenceDictionary referenceSequenceDictionary, final Set<VCFHeaderLine> defaultToolVCFHeaderLines, final Logger logger) {\n",
      "    final List<VariantContext> sortedVariantsList = sortVariantsByCoordinate(localVariants, referenceSequenceDictionary);\n",
      "    if (logger != null)\n",
      "        logNumOfVarByTypes(sortedVariantsList, logger);\n",
      "    writeVariants(vcfFileName, sortedVariantsList, referenceSequenceDictionary, defaultToolVCFHeaderLines);\n",
      "}\n",
      "3406\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1239/1243 [43:10<00:09,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57625\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1240/1243 [43:12<00:06,  2.30s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57779\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1241/1243 [43:14<00:04,  2.33s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "57834\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1242/1243 [43:17<00:02,  2.48s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58158\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing rows: 100%|| 1243/1243 [43:19<00:00,  2.09s/it]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "58159\n",
      "{'classCount': 2, 'imports': [{'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public class MetricsUtils {\\n    private MetricsUtils();\\n    public static void saveMetrics(final MetricsFile<?, ?> metricsFile, String metricsOutputPath)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/metrics/MetricsUtils.java', 'methodCount': 2, 'lineCount': 4}, {'fieldCount': 14, 'classCount': 1, 'methodsOnly': 'public class QualityYieldMetrics {\\n    private static final long serialVersionUID = 1;;\\n    private boolean useOriginalQualities;;\\n    private boolean finalized;;\\n    public QualityYieldMetrics setUseOriginalQualities(final boolean useOriginalQualities)\\n    public int TOTAL_READS;;\\n    public int PF_READS;;\\n    public int READ_LENGTH;;\\n    public long TOTAL_BASES;;\\n    public long PF_BASES;;\\n    public long Q20_BASES;;\\n    public long PF_Q20_BASES;;\\n    public long Q30_BASES;;\\n    public long PF_Q30_BASES;;\\n    public long Q20_EQUIVALENT_YIELD;;\\n    public long PF_Q20_EQUIVALENT_YIELD;;\\n    public QualityYieldMetrics addRead(final GATKRead read)\\n    public QualityYieldMetrics combine(final QualityYieldMetrics that)\\n    public QualityYieldMetrics finish()\\n    public static String getUniqueNameSuffix()\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/metrics/QualityYieldMetrics.java', 'methodCount': 5, 'lineCount': 21}, {'fieldCount': 2, 'classCount': 1, 'methodsOnly': 'public class QualityYieldMetricsArgumentCollection {\\n    public static final long serialVersionUID = 1L;;\\n    @Argument(doc = \"If available in the OQ tag, use the original quality scores \" + \"as inputs instead of the quality scores in the QUAL field.\", shortName = StandardArgumentDefinitions.USE_ORIGINAL_QUALITIES_SHORT_NAME, fullName = StandardArgumentDefinitions.USE_ORIGINAL_QUALITIES_LONG_NAME)\\npublic boolean useOriginalQualities = true;;\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/metrics/QualityYieldMetricsArgumentCollection.java', 'methodCount': 0, 'lineCount': 4}, {'fieldCount': 0, 'classCount': 1, 'methodsOnly': 'public interface GATKRead {\\n     String getName()\\n     int getFlags()\\n     void setName(final String name)\\n     int getLength()\\n     boolean isEmpty()\\n     void setPosition(final String contig, final int start)\\n     void setPosition(final Locatable locatable)\\n     String getAssignedContig()\\n     int getAssignedStart()\\n     int getUnclippedStart()\\n     int getUnclippedEnd()\\n     int getSoftStart()\\n     int getSoftEnd()\\n     int getAdaptorBoundary()\\n     String getMateContig()\\n     int getMateStart()\\n     void setMatePosition(final String contig, final int start)\\n     void setMatePosition(final Locatable locatable)\\n     int getFragmentLength()\\n     void setFragmentLength(final int fragmentLength)\\n     int getMappingQuality()\\n     void setMappingQuality(final int mappingQuality)\\n     byte[] getBases()\\n     byte[] getBasesNoCopy()\\n     int copyBases(final int offset, final byte[] destination, final int destinationOffset, final int maxLength)\\n     int copyBaseQualities(final int offset, final byte[] destination, final int destinationOffset, final int maxLength)\\n     boolean hasBases()\\n     boolean hasBaseQualities()\\n     byte getBase(final int i)\\n     String getBasesString()\\n     void setBases(final byte[] bases)\\n     byte[] getBaseQualities()\\n     byte[] getBaseQualitiesNoCopy()\\n     int getBaseQualityCount()\\n     byte getBaseQuality(final int i)\\n     void setBaseQualities(final byte[] baseQualities)\\n     Cigar getCigar()\\n     List<CigarElement> getCigarElements()\\n     CigarElement getCigarElement(final int i)\\n     int numCigarElements()\\n     void setCigar(final Cigar cigar)\\n     void setCigar(final String cigarString)\\n     String getReadGroup()\\n     void setReadGroup(final String readGroupID)\\n     boolean isPaired()\\n     void setIsPaired(final boolean isPaired)\\n     boolean isProperlyPaired()\\n     void setIsProperlyPaired(final boolean isProperlyPaired)\\n     boolean isUnmapped()\\n     void setIsUnmapped()\\n     boolean isUnplaced()\\n     void setIsUnplaced()\\n     boolean mateIsUnmapped()\\n     void setMateIsUnmapped()\\n     boolean mateIsUnplaced()\\n     void setMateIsUnplaced()\\n     boolean isReverseStrand()\\n     void setIsReverseStrand(final boolean isReverseStrand)\\n     boolean mateIsReverseStrand()\\n     void setMateIsReverseStrand(final boolean mateIsReverseStrand)\\n     boolean isFirstOfPair()\\n     void setIsFirstOfPair()\\n     boolean isSecondOfPair()\\n     void setIsSecondOfPair()\\n     boolean isSecondaryAlignment()\\n     void setIsSecondaryAlignment(final boolean isSecondaryAlignment)\\n     boolean isSupplementaryAlignment()\\n     void setIsSupplementaryAlignment(final boolean isSupplementaryAlignment)\\n     Optional<SamPairUtil.PairOrientation> getPairOrientation()\\n     boolean failsVendorQualityCheck()\\n     void setFailsVendorQualityCheck(final boolean failsVendorQualityCheck)\\n     boolean isDuplicate()\\n     void setIsDuplicate(final boolean isDuplicate)\\n     boolean hasAttribute(final String attributeName)\\n     Integer getAttributeAsInteger(final String attributeName)\\n     Float getAttributeAsFloat(final String attributeName)\\n     String getAttributeAsString(final String attributeName)\\n     byte[] getAttributeAsByteArray(final String attributeName)\\n     Object getTransientAttribute(final Object key)\\n     Optional<T> getOptionalTransientAttribute(final Object key, final Class<T> clazz)\\n     void setAttribute(final String attributeName, final Integer attributeValue)\\n     void setAttribute(final String attributeName, final String attributeValue)\\n     void setAttribute(final String attributeName, final byte[] attributeValue)\\n    public void setTransientAttribute(final Object key, final Object value)\\n     void clearAttribute(final String attributeName)\\n     void clearAttributes()\\n     void clearTransientAttribute(final String attributeName)\\n     GATKRead copy()\\n     GATKRead deepCopy()\\n     SAMRecord convertToSAMRecord(final SAMFileHeader header)\\n     String getSAMString()\\n    public void reverseComplement()\\n     String commonToString()\\n     void hardClipAttributes(final int newStart, final int newLength, int originalLength)\\n}\\n', 'filePath': '/home/ubuntu/repos/broadinstitute,gatk/src/main/java/org/broadinstitute/hellbender/utils/read/GATKRead.java', 'methodCount': 94, 'lineCount': 96}], 'asteriskCount': 0, 'importCount': 2, 'javaImportCount': 2}\n",
      "{'QualityYieldMetrics': 'public class QualityYieldMetrics {\\n    private static final long serialVersionUID = 1;;\\n    private boolean useOriginalQualities;;\\n    private boolean finalized;;\\n    public QualityYieldMetrics setUseOriginalQualities(final boolean useOriginalQualities)\\n    public int TOTAL_READS;;\\n    public int PF_READS;;\\n    public int READ_LENGTH;;\\n    public long TOTAL_BASES;;\\n    public long PF_BASES;;\\n    public long Q20_BASES;;\\n    public long PF_Q20_BASES;;\\n    public long Q30_BASES;;\\n    public long PF_Q30_BASES;;\\n    public long Q20_EQUIVALENT_YIELD;;\\n    public long PF_Q20_EQUIVALENT_YIELD;;\\n    public QualityYieldMetrics addRead(final GATKRead read)\\n    public QualityYieldMetrics combine(final QualityYieldMetrics that)\\n    public QualityYieldMetrics finish()\\n    public static String getUniqueNameSuffix()\\n}\\n'}\n",
      "/**\n",
      "     * Do the actual metrics collection on the provided RDD.\n",
      "     * @param filteredReads The reads to be analyzed for this collector.\n",
      "     * @param samHeader The SAMFileHeader associated with the reads in the input RDD.\n",
      "     */\n",
      "public void collectMetrics(final JavaRDD<GATKRead> filteredReads, final SAMFileHeader samHeader)\n",
      "**************************************\n",
      "@Override\n",
      "public void collectMetrics(final JavaRDD<GATKRead> filteredReads, final SAMFileHeader samHeader) {\n",
      "    final QualityYieldMetrics metrics = filteredReads.aggregate(new QualityYieldMetrics().setUseOriginalQualities(args.useOriginalQualities), (hgp, read) -> hgp.addRead(read), (hgp1, hgp2) -> hgp1.combine(hgp2)).finish();\n",
      "    metricsFile.addMetric(metrics);\n",
      "}\n",
      "2024\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "\"\"\"(\"\"from tenacity import (\n",
    "    retry,\n",
    "    stop_after_attempt,\n",
    "    wait_random_exponential,\n",
    "    wait_exponential\n",
    ")\n",
    "gpt_outputs = []\n",
    "\n",
    "@retry(\n",
    "    stop=stop_after_attempt(15),\n",
    "    wait=wait_exponential(multiplier=2, min=2, max=150),\n",
    ")\n",
    "def get_gpt_response(index):\n",
    "  global tracker\n",
    "  completion = openai.ChatCompletion.create(\n",
    "    model=\"gpt-3.5-turbo\",\n",
    "    max_tokens = 4096-600-int(data.loc[index]['Token Length']),\n",
    "    messages=[\n",
    "      {\"role\": \"user\", \"content\": f\"\"\"I will provide you with a prompt to a function below these instructions. You will output exactly as follows, with the list as well. Text encased in <like this> will be replaced by your response, and text encased in (like this) is just a description for the response that you do not need to type up:\n",
    "  (a) <Boolean> (Is it bioinformatics related?)\n",
    "  (b) <words> (Give a list of 5 keywords of why it is bioinformatics related)\n",
    "  (c) <integer> (Your confidence from 0 to 100 that your response in A is accurate, so for example, if you believe strongly that it is not bioinformatics related, you should also rate a high confidence level)\n",
    "  The code must explicitly reference some bioinformatics methodology, terminology, or process. For example, an AVL Tree would not be a valid bioinformatics function, while a FASTQ processor would. The keywords are defined as important words that allowed you to make the determination that the function is bioinformatics related. The confidence should be your estimate of how confident you are of your responses.\n",
    "\n",
    "  Make sure that in your response is explicitly as follows in the directions. Part A should only be one word and a boolean, either True or False. Part B should only be 5 words, no additional information, Part C should only be a single integer, from 0 to 100, it is a measure of your confidence in your response to Part A.\n",
    "\n",
    "  After selecting keywords, please reverify that the words you used to make the decision for Part A is actually bioinformatics related.\n",
    "  Again, as clarification, I will be providing the function. \n",
    "\n",
    "  The responses should be formatted as a list:\n",
    "  Entry 1: The response to part A converted into a string\n",
    "  Entry 2: A list of 5 words which are strings from the response to Part B\n",
    "  Entry 3: The integer response to part C converted to a string\n",
    "  Therefore, your output should follow this guideline. This will be your only output, there should be no additional outputs beyond the one highlighted in this prompt.\n",
    "\n",
    "  Prompt begins here:\n",
    "  {data.loc[index]['prompt']}\n",
    "\n",
    "  Prompt ends here.\n",
    "  Give the output to the above code encased in \"Prompt begins here:\" and \"Prompt ends here.\" Your keyword search should only encompass the words in the prompt, and ensure that keywords are related to bioinformatics, not statistics.\n",
    "  \"\"\"}]\n",
    "  )\n",
    "  print(index)\n",
    "  if tracker % 100 == 0 & tracker != 0:\n",
    "    temp_df = data\n",
    "    temp_df['GPT Output'] = gpt_outputs\n",
    "    temp_df.to_csv(f'checkpoint_{tracker % 100}_gpt_outputs.csv')\n",
    "    print(f'saved checkpoint {tracker % 100}')\n",
    "  tracker += 1\n",
    "  if tracker % 5 == 0:\n",
    "    print(data.loc[index]['prompt'])\n",
    "    print(data.loc[index]['Token Length'])\n",
    "  return completion.choices[0].message['content']\n",
    "\n",
    "for index, row in tqdm(data.iterrows(), total=len(data), desc=\"Processing rows\"):\n",
    "  gpt_outputs.append(get_gpt_response(index))\"\")\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1243\n"
     ]
    }
   ],
   "source": [
    "print(len(gpt_outputs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1243, 24)\n",
      "Index(['totalNumLines', 'imports', 'numParameters', 'signature', 'lineStart',\n",
      "       'filePath', 'numChar', 'params', 'addImports', 'lineEnd', 'content',\n",
      "       'numCommentChars', 'numCommentLines', 'intraClassFieldsUsed',\n",
      "       'function_id', 'parentClass', 'comments', 'packageName', 'returnType',\n",
      "       'hash', 'prompt', 'good_length', 'Token Length', 'GPT Output'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "data['GPT Output'] = gpt_outputs\n",
    "print(data.shape)\n",
    "print(data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "71 ['False', ['clean', 'RAMPART', 'job', 'directory', 'delete'], '80']\n",
      "86 ['False', ['instance', 'service', 'identifier', 'name', 'create'], '50']\n",
      "87 ['False', ['null', 'instance', 'service', 'executor', 'injected'], '70']\n",
      "90 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '90']\n",
      "523 ['False', ['removeChange', 'colour change', 'node', 'editing', 'range'], '60']\n",
      "540 ['False', ['tree', 'type', 'trait', 'set', 'value'], '70']\n",
      "580 ['False', ['RpowSteadyN', 'steady', 'symmetric', 'index', 'known'], '90']\n",
      "1230 ['False', ['set', 'matrix', 'range', 'i', 'j'], '70']\n",
      "1665 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '85']\n",
      "1671 ['False', ['JComboBox', 'Mode', 'TRANSFORMING', 'PICKING', 'ANNOTATING'], '70']\n",
      "2805 ['False', ['Interval1D', 'intersects', 'contains', 'union', 'overlapCenter'], '90']\n",
      "2827 ['False', ['Interval1D', 'compareTo', 'Interval'], '80']\n",
      "2993 ['False', ['equality', 'list', 'integer', 'compare', 'size'], '60']\n",
      "4458 [\"False\", [\"None\"], \"80\"]\n",
      "4992 ['False', ['restoreProps', 'session file', 'ExpressionFileReaderTask', 'Map', 'simpleName'], '75']\n",
      "5437 ['False', ['JFrame', 'JDialog', 'getWindowInstance', 'Component', 'getParent'], '80']\n",
      "6309 ['False', ['OutputStream', 'close', 'flush', 'uncloseable', 'Java'], '90']\n",
      "7759 [\"False\", [\"java\", \"Object\", \"hashCode\", \"prime\", \"result\"], \"70\"]\n",
      "7782 ['False', ['n/a', 'n/a', 'n/a', 'n/a', 'n/a'], '50']\n",
      "7788 ['False', ['SQL', 'resultset', 'SQLException', 'stmt', 'closeQuery'], '90']\n",
      "7831 ['False', ['hashcode', 'aaPosition', 'geneID', 'hgvs', 'residues'], '80']\n",
      "8171 ['False', ['statistics', 'distribution', 'CDF', 'evaluated', 'tail'], '95']\n",
      "8198 ['False', ['getInitialDomain', 'probability', 'critical value', 'inverseCumulativeProbability', 'domain value'], '90']\n",
      "8209 ['False', ['math', 'probability', 'root', 'critical', 'value'], '70']\n",
      "8300 ['False', ['preconditions', 'binomial', 'IllegalArgumentException', 'size', 'subsets'], '85']\n",
      "8326 ['False', ['Number', 'Scale', 'Power', 'Double', 'Multiply'], '80']\n",
      "8417 ['False', ['verifyIterationCount', 'IllegalArgumentException', 'iteration', 'limits', 'MathRuntimeException'], '85']\n",
      "8564 ['False', ['string', 'characters', 'newline', 'excluded', 'substring'], '80']\n",
      "8585 ['False', ['JSONStringer', 'JSON', 'StringBuilder', 'indentation', 'object'], '90']\n",
      "8595 ['False', ['beforeKey', 'JSONException', 'context', 'Scope', 'replaceTop'], '70']\n",
      "9108 ['False', ['log', 'scaling', 'factor', 'nodes', 'pattern'], '90']\n",
      "9110 ['False', ['None'], '70']\n",
      "9158 ['False', ['setCurrentNodePartials', 'nodeIndex', 'partials', 'category', 'partitions'], '75']\n",
      "9166 ['False', ['None'], '90']\n",
      "9945 ['False', ['random', 'permutation', 'int', 'array', 'uniform'], '70']\n",
      "9946 ['False', ['array', 'shuffle', 'integer', 'length', 'nextInt'], '70']\n",
      "9947 ['False', ['shuffle', 'array', 'numberOfShuffles', 'int', 'temp'], '85']\n",
      "9948 ['False', ['array', 'shuffled', 'indices', 'length', 'shuffle'], '70']\n",
      "9949 ['False', ['array', 'integer', 'permuted', 'uniform', 'random'], '80']\n",
      "9988 ['False', ['compute', 'Epsilon', 'scratch', 'double', 'methodology'], '60']\n",
      "10001 ['False', ['methodology', 'frequency', 'object', 'sortedList', 'incremented'], '80']\n",
      "10033 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '90']\n",
      "10048 ['False', ['proportion', 'values', 'indices', 'interval', 'sorting'], '75']\n",
      "10291 ['False', ['reserved', 'invalid', 'name', 'exit', 'Log'], '95']\n",
      "10827 ['False', ['backwardSubstitution', 'double[]', 'answer', 'rows', 'i'], '90']\n",
      "10832 ['False', ['forwardSubstitution', 'double[]', 'answer', 'permutation', 'rows'], '80']\n",
      "10834 ['False', ['initialize', 'double', 'constructor', 'components', 'rows'], '60']\n",
      "10836 ['False', ['none', 'related', 'bioinformatics', 'terms', 'found'], '80']\n",
      "10846 ['False', ['diagonal', 'matrix', 'singular', 'values', 'DoubleMatrix2D'], '90']\n",
      "10854 ['False', ['position', 'largest', 'pivot', 'step', 'answer'], '50']\n",
      "10860 ['False', ['rows', 'swap', 'double', 'length', 'temp'], '75']\n",
      "10882 ['False', ['representation', 'vector', 'components', 'separator', 'string'], '75']\n",
      "10891 [\"False\", [\"None\"], \"75\"]\n",
      "10927 ['False', ['array', 'manipulation', 'matrix', 'operation', 'numeric'], '60']\n",
      "10937 ['False', ['double', 'receiver', 'components', 'copy', 'answer'], '35']\n",
      "10997 ['False', ['packages', 'classes', 'search', 'BEASTClassLoader', 'find'], '50']\n",
      "10998 ['False', ['package', 'classes', 'interface', 'list', 'found'], '70']\n",
      "11125 ['False', ['services', 'packages', 'class', 'identifying', 'set'], '70']\n",
      "11132 ['False', ['addService','service','className','packageName','BEASTClassLoader'], '60']\n",
      "13708 ['False', ['weighted', 'Z-score', 'Whitlock', 'sample', 'size'], '60']\n",
      "13787 ['False', ['compareNaN', 'Double.NaN', 'value'], '75']\n",
      "15158 ['False', ['java', 'file', 'close', 'buffer', 'write'], '90']\n",
      "15891 ['False', ['class', 'imports', 'methods', 'Bioinformatics', 'terminology'], '75']\n",
      "16090 ['False', ['column', 'index', 'header', 'LinkedList', 'Map'], '80']\n",
      "16785 ['False', ['remove', 'message', 'board', 'iterator', 'panel'], '60']\n",
      "17099 [\"False\", [\"Java\", \"JLabel\", \"file type\", \"label\", \"font\"], \"70\"]\n",
      "20515 ['False', ['sublist', 'size', 'cnt', 'clazz', 'anOriginalList'], '80']\n",
      "20516 ['False', ['sublist', 'boolean', 'empty', 'originalList', 'clazz'], '65']\n",
      "20519 ['False', ['sublist', 'object', 'index', 'input', 'int'], '70']\n",
      "20756 ['False', ['valid', 'sub', 'samples', 'number', 'methodology'], '75']\n",
      "20787 ['False', ['sub', 'sample', 'index', 'string', 'value'], '70']\n",
      "21711 ['False', ['collections', 'concatenation', 'stream', 'null', 'javaImportCount'], '90']\n",
      "21796 ['False', ['acquires', 'mutex', 'Exception', 'try', 'RuntimeException'], '90']\n",
      "21797 ['False', ['None'], '100']\n",
      "21829 ['False', ['getters', 'setters', 'code', 'methods', 'String'], '90']\n",
      "21839 ['False', ['column', 'type', 'coded', 'specified', 'found'], '90']\n",
      "21840 ['False', ['column', 'size', 'specified', 'found', 'method'], '80']\n",
      "21841 ['False', ['converted', 'column', 'type', 'specified', 'found'], '85']\n",
      "21899 ['False', ['cache', 'load', 'object', 'key', 'present'], '85']\n",
      "21908 ['False', ['cache', 'memory', 'save', 'objects', 'thread'], '80']\n",
      "22063 ['False', ['scale', 'values', 'maximum', 'range', 'stream'], '50']\n",
      "22082 ['False', ['Properties', 'ClassLoader', 'InputStream', 'IOException', 'IllegalArgumentException'], '75']\n",
      "22185 ['False', ['thread', 'interrupted', 'value', 'produced', 'join'], '50']\n",
      "23183 ['False', ['method', 'stream', 'output', 'file', 'close'], '80']\n",
      "23197 ['False', ['command', 'line', 'flag', 'check', 'case-sensitive'], '50']\n",
      "23198 ['False', ['parse', 'command-line', 'arguments', 'retrieve', 'format'], '70']\n",
      "23265 ['False', ['prepare', 'renderer', 'data', 'model', 'component'], '95']\n",
      "23295 ['False', ['add', 'user', 'refinement', 'parameter', 'map'], '75']\n",
      "23302 ['False', ['unique', '64 bit', 'hash key', 'constant time', 'database'], '85']\n",
      "24021 ['False', ['charge', 'value', 'string', 'formatted', 'temp'], '75']\n",
      "25218 ['False', ['GUI', 'settings', 'parameters', 'display', 'ThermoRawFileParserParameters'], '70']\n",
      "25752 ['False', ['magic', 'currently', 'supported', 'CMS', 'files'], '70']\n",
      "25958 ['False', ['file', 'delete', 'folder', 'Java', 'JOptionPane'], '90']\n",
      "26039 ['False', ['dialog', 'installEscapeCloseOperation', 'JDialog', 'WindowEvent', 'root'], '75']\n",
      "26177 [\"False\", [\"update\", \"item\", \"index\", \"popup\", \"visible\"], \"50\"]\n",
      "26245 ['False', ['GUI', 'filters', 'proteinCorner', 'ScrollPaneConstants', 'viewport'], '90']\n",
      "26303 ['False', ['DateChooser', 'Calendar', 'JLabel', 'SimpleDateFormat', 'JFrame'], '50']\n",
      "27151 ['False', ['GUI', 'PepNovo', 'parameters', 'populate', 'settings'], '90']\n",
      "27401 ['False', ['ListModel', 'checked', 'items', 'biological', 'data'], '60']\n",
      "27402 ['False', ['java', 'list', 'model', 'check', 'item'], '70']\n",
      "27562 ['False', ['function', 'setTipOfTheDay', 'list', 'setVisible', 'isEmpty'], '90']\n",
      "27577 ['False', ['GUI', 'button', 'action', 'event', 'frame'], '85']\n",
      "27651 ['False', ['TableModel', 'data', 'sortable', 'TableSorter', 'sort'], '90']\n",
      "27815 ['False', ['n/a', 'n/a', 'n/a', 'n/a', 'n/a'], '50']\n",
      "27824 ['False', ['sorted', 'array', 'double', 'median', 'input'], '70']\n",
      "27829 ['False', ['percentile', 'list', 'values', 'interpolation', 'median'], '50']\n",
      "27851 [\"False\", [\"exclude\", \"null\", \"values\", \"double\", \"array\"], \"50\"]\n",
      "27925 ['False', ['Java', 'progress', 'mutex', 'WaitingHandler', 'CLI'], '80']\n",
      "27965 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '70']\n",
      "28021 ['False', ['directories', '*.pkl.spo', 'filter', 'extension', 'FileFilterUtils'], '30']\n",
      "28046 ['False', ['serialized', 'file', 'reading', 'class', 'unknown'], '20']\n",
      "28171 ['False', ['Semaphore', 'mutex', 'nPermits', 'acquire', 'release'], '75']\n",
      "28202 ['False', ['FTP', 'server', 'connection', 'IOException', 'close'], '75']\n",
      "28211 ['False', ['FTP', 'file', 'Ascii', 'IOException', 'read'], '80']\n",
      "28212 ['False', ['FTP', 'server', 'get', 'Binary', 'mode'], '60']\n",
      "28286 ['False', ['deleteDir', 'subdirectories', 'directory', 'deletion', 'successful'], '70']\n",
      "28381 ['False', ['None'], '90']\n",
      "28709 ['False', ['execute', 'IAtom', 'IAtomContainer', 'SortedMap', 'IBond'], '70']\n",
      "28711 ['False', ['execute', 'IAtom', 'IAtomContainer', 'SortedMap', 'IBond'], '60']\n",
      "29583 ['False', ['permutation', 'lexicographically', 'rank', 'int', 'counter'], '50']\n",
      "29584 ['False', ['Lexicographically', 'Permutation', 'Rank', 'Int', 'Array'], '50']\n",
      "29592 ['False', ['atom', 'reaction', 'permutationMap', 'IAtom', 'IAtomContainer'], '80']\n",
      "29651 [\"False\", [\"synchronized\", \"Map\", \"descending\", \"order\", \"entry\"], \"50\"]\n",
      "29652 ['False', ['methodology', 'terminology', 'process', 'synchronized', 'LinkedList'], '70']\n",
      "29653 [\"False\", [\"map\", \"intValueInDescendingOrder\", \"LinkedList\", \"sort\", \"LinkedHashMap\"], \"70\"]\n",
      "29654 ['False', ['synchronized', 'Ascending Order', 'Map', 'Double', 'LinkedList'], '85']\n",
      "29655 ['False', ['Map', 'Descending', 'Order', 'Double', 'Synchronized'], '80']\n",
      "29713 ['False', ['formatMDLInt', 'connection table', 'int', 'String', 'NumberFormat'], '75']\n",
      "29757 ['False', ['chem', 'model', 'IAtomContainer', 'builder', 'newInstance'], '70']\n",
      "29879 ['False', ['none', 'bioinformatics', 'related', 'keywords', 'found'], '0']\n",
      "29907 ['False', ['print', 'Atoms', 'molecules', 'IAtomContainer', 'mol'], '75']\n",
      "29908 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '60']\n",
      "30209 ['False', ['none'], '90']\n",
      "30239 ['False', ['coordinates', 'points', 'bond', 'distance', 'vertical'], '70']\n",
      "30242 ['False', ['closest', 'atom', 'molecule', 'coordinates', 'searched'], '80']\n",
      "30245 ['False', ['coordinates', 'molecule', 'atom', 'closest', 'searched'], '60']\n",
      "30288 ['False', ['None'], '80']\n",
      "30366 ['False', ['Point2d', 'center', 'hull', 'length', 'return'], '80']\n",
      "30521 ['False', ['Java', 'graphics', 'layout', 'IAtomContainer', 'atoms'], '90']\n",
      "30547 [\"False\", [\"vector\", \"position\", \"annotation\", \"switch\", \"return\"], \"70\"]\n",
      "30548 ['False', ['none', 'relevant', 'bioinformatics', 'methodology', 'terminology'], '80']\n",
      "30549 ['False', ['methodology', 'process', 'terminology', 'FASTQ', 'processor'], '80']\n",
      "30710 ['False', ['atom', 'atomic number', 'positive integer value', 'access', 'given'], '70']\n",
      "31415 ['False', ['BitSet', 'CDKException', 'constrains', 'graph', 'parsing'], '80']\n",
      "31424 ['False', ['BitSet', 'isContainedIn', 'sourceBitSet', 'targetBitSet', 'clone'], '70']\n",
      "31580 ['False', ['compare', 'Map', 'size', 'sortOrder'], '70']\n",
      "31859 [\"False\", [\"method\", \"PicardException\", \"setOrCheckLane\", \"exception\", \"lane\"], \"60\"]\n",
      "32552 ['False', ['array', 'doubles', 'sorted', 'median', 'empty'], '50']\n",
      "32558 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '90']\n",
      "32560 ['False', [], '70']\n",
      "32564 ['False', ['array', 'smallest', 'value', 'stored', 'min'], '50']\n",
      "32565 ['False', ['java', 'method', 'array', 'byte', 'value'], '80']\n",
      "32567 ['False', ['function', 'array', 'element', 'index', 'smallest'], '80']\n",
      "32568 ['False', ['None'], '100']\n",
      "32577 ['False', ['arrays', 'length', 'product', 'calculate', 'double'], '70']\n",
      "32578 ['False', ['sum', 'arrays', 'calculates', 'third', 'double'], '80']\n",
      "32579 ['False', ['none'], '90']\n",
      "32730 ['False', ['categories', 'doc', 'index', 'map', 'group'], '70']\n",
      "32805 ['False', ['Properties', 'on-disk', 'Properties', 'class', 'loader'], '70']\n",
      "34156 ['False', ['Base 64', 'encode', 'byte array', 'data', 'String'], '80']\n",
      "34409 ['False', ['ImageIcon', 'URL', 'getResource', 'null', 'find'], '80']\n",
      "34411 ['False', ['None'], '90']\n",
      "34412 ['False', ['None'], '75']\n",
      "35046 ['False', ['createTransferable', 'object', 'type', 'cast', 'implements'], '80']\n",
      "36459 ['False', ['center', 'position', 'features', 'getFeatureCenteredAfter', 'idx'], '70']\n",
      "36718 ['False', ['None'], '50']\n",
      "36960 ['False', ['JSON', 'Web', 'Token', 'String', 'payload'], '75']\n",
      "36971 ['False', ['capitalized', 'words', 'input', 'string', 'series'], '90']\n",
      "37029 [\"False\", [\"web\", \"browser\", \"URL\", \"IOException\", \"Desktop\"], \"50\"]\n",
      "37030 ['False', ['web', 'browser', 'url', 'IOError', 'Desktop'], '60']\n",
      "37222 ['False', ['URL', 'encode', 'string', 'URI', 'path'], '75']\n",
      "37223 ['False', ['string', 'decode', 'percent-encoded', 'fast path', 'methodology'], '90']\n",
      "37225 [\"False\", [\"Java\", \"Stringbuilder\", \"Decoded\", \"Percent-encoded\", \"Hex\"], \"50\"]\n",
      "41106 ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'], '80']\n",
      "41120 ['False', ['slider', 'unit', 'direction', 'thumb', 'value'], '65']\n",
      "41655 ['False', ['zip', 'jar', 'file', 'entries', 'handleItem'], '90']\n",
      "41656 ['False', ['directory', 'filesystem', 'classes', 'package', 'path'], '90']\n",
      "41736 ['False', ['byte', 'input', 'buffer', 'read', 'exception'], '90']\n",
      "42078 ['False', ['NA'], '90']\n",
      "42683 ['False', ['JexlVCMatchExp', 'expression', 'evaluate', 'Jexl', 'context'], '80']\n",
      "43927 ['False', ['not', 'bioinformatics', 'related', 'methodology', 'terminology'], '80']\n",
      "44108 ['False', ['minimum', 'interval', 'tree', 'earliest', 'null'], '85']\n",
      "44111 ['False', ['None'], '80']\n",
      "44177 ['False', ['none'], '90']\n",
      "44285 ['False', ['hasNext', 'doneAdding', 'cleanedUp', 'ramValues', 'priorityQueue'], '70']\n",
      "44428 ['False', ['java', 'iterator', 'object', 'exception', 'IterableOnceIterator'], '80']\n",
      "44540 ['False', ['ASCII', 'bytes', 'write', 'string', 'length'], '75']\n",
      "44542 ['False', ['byte', 'short', 'value', 'writing', 'method'], '70']\n",
      "44543 ['False', ['writeUShort', 'unsigned', 'byteBuffer', 'val', 'IllegalArgumentException'], '90']\n",
      "44544 ['False', ['writing', 'unsigned', '32-bit', 'val', 'byteBuffer'], '50']\n",
      "44552 [\"False\", [\"String\", \"length\", \"devourNull\", \"readInt\", \"readString\"], \"80\"]\n",
      "45111 ['False', ['code', 'BitCodeWithPrefix', 'String', 'binary', 'buffer'], '60']\n",
      "46749 ['False', ['decode', 'attributes', 'bases', 'qualities', 'binary'], '90']\n",
      "46758 ['False', ['methodology', 'terminology', 'process', 'byte', 'data'], '50']\n",
      "47415 ['False', ['legal', 'size', 'elements', 'load', 'factor'], '85']\n",
      "47751 ['False', ['increment', 'table', 'exception', 'row', 'column'], '75']\n",
      "48097 ['False', ['class', 'constructor', 'public', 'synthetic', 'interface'], '90']\n",
      "48359 ['False', ['None'], '90']\n",
      "48374 ['False', ['cache', 'precomputed', 'expandCache', 'compute', 'synchronized'], '95']\n",
      "48712 ['False', ['super-category', 'custom-order', 'categories', 'doc index', 'getGroupMap'], '80']\n",
      "48725 ['False', ['freemarker', 'WorkUnit', 'template', 'output', 'File'], '10']\n",
      "50466 ['False', ['print', 'matrix', 'human', 'readable', 'form'], '80']\n",
      "50517 ['False', ['print', 'debugging', 'name', 'matrix', 'values'], '70']\n",
      "50670 ['False', ['methodCount', 'GATKException', 'String', 'log', 'suitable'], '85']\n",
      "55220 ['False', ['graph', 'vertices', 'edges', 'reference', 'source'], '70']\n",
      "56604 ['False', ['compute', 'allele', 'frequencies', 'counts', 'number'], '80']\n",
      "206\n"
     ]
    }
   ],
   "source": [
    "track = 0\n",
    "for index, row in data.iterrows():\n",
    "    if row['GPT Output'][2] == 'F':\n",
    "        track +=1\n",
    "        print(index, row['GPT Output'])\n",
    "\n",
    "print(track)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n",
      "['False', ['clean', 'RAMPART', 'job', 'directory', 'delete'], '80']\n",
      "['False', ['clean', 'RAMPART', 'job', 'directory', 'delete'], '80']\n",
      "False\n",
      "['True', ['clean', 'RAMPART', 'job', 'directory', 'delete'], '80']\n",
      "['False', ['clean', 'RAMPART', 'job', 'directory', 'delete'], '80']\n",
      "71       ['False', ['clean', 'RAMPART', 'job', 'directo...\n",
      "86       ['False', ['instance', 'service', 'identifier'...\n",
      "87       ['False', ['null', 'instance', 'service', 'exe...\n",
      "90       ['False', ['N/A', 'N/A', 'N/A', 'N/A', 'N/A'],...\n",
      "91       ['True', ['services', 'list', 'iterator', 'loa...\n",
      "                               ...                        \n",
      "57625    ['True', ['QNameAndInterval', 'template', 'nam...\n",
      "57779    ['True', ['K-S calculation', 'windows', 'evide...\n",
      "57834    ['True', ['contig', 'configurations', 'mapping...\n",
      "58158    ['True', ['QualityYieldMetrics', 'GATKRead', '...\n",
      "58159    [True, ['bioinformatics', 'RDD', 'SAMFileHeade...\n",
      "Name: GPT Output, Length: 1243, dtype: object\n",
      "1243\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_702009/596899551.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  run_on_this.loc[71]['GPT Output'] = gpt_value\n",
      "/tmp/ipykernel_702009/596899551.py:20: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  run_on_this.loc[location]['GPT Output'] = str(gpt_value)\n"
     ]
    }
   ],
   "source": [
    "import ast\n",
    "num_edit = [71, 7831, 9945, 10997, 26245, 27401, 28171, 28709, 28711, 30242, 30245, 30710, 34156, 37223, 50466, 56604]\n",
    "run_on_this = data\n",
    "print(len(num_edit))\n",
    "print(run_on_this.loc[71]['GPT Output'])\n",
    "gpt_value = run_on_this.loc[71]['GPT Output']\n",
    "print(gpt_value)\n",
    "gpt_value = ast.literal_eval(gpt_value)\n",
    "print(gpt_value[0])\n",
    "gpt_value[0] = 'True'\n",
    "print(gpt_value)\n",
    "gpt_value = str(gpt_value)\n",
    "run_on_this.loc[71]['GPT Output'] = gpt_value\n",
    "print(run_on_this.loc[71]['GPT Output'])\n",
    "\n",
    "for location in num_edit:\n",
    "    gpt_value = run_on_this.loc[location]['GPT Output']\n",
    "    gpt_value = ast.literal_eval(gpt_value)\n",
    "    gpt_value[0] = 'True'\n",
    "    run_on_this.loc[location]['GPT Output'] = str(gpt_value)\n",
    "print(run_on_this['GPT Output'])\n",
    "print(len(run_on_this['GPT Output']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'run_on_this' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[16], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m run_on_this\u001b[39m.\u001b[39mto_json(\u001b[39m'\u001b[39m\u001b[39m/home/ubuntu/Bio-Code-Eval/Bio-Code-Eval-new/final_java_functions_for_testing.json\u001b[39m\u001b[39m'\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'run_on_this' is not defined"
     ]
    }
   ],
   "source": [
    "run_on_this.to_json('/home/ubuntu/Bio-Code-Eval/Bio-Code-Eval-new/final_java_functions_for_testing.json')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index([   71,    86,    87,    90,    91,   523,   540,   544,   554,   569,\n",
      "       ...\n",
      "       57280, 57281, 57304, 57319, 57461, 57625, 57779, 57834, 58158, 58159],\n",
      "      dtype='int64', length=1243)\n",
      "{'classCount': 0, 'imports': [], 'asteriskCount': 0, 'importCount': 0, 'javaImportCount': 4}\n",
      "{}\n",
      "/**\n",
      "     * Returns a list of the requested services\n",
      "     * @return A list of services\n",
      "     */\n",
      "public List<T> getServices()\n",
      "**************************************\n",
      "public List<T> getServices() {\n",
      "    Iterator<T> it = loader.iterator();\n",
      "    List<T> serviceList = new ArrayList<>();\n",
      "    while (it.hasNext()) {\n",
      "        serviceList.add(it.next());\n",
      "    }\n",
      "    return serviceList;\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "files = '/home/ubuntu/Bio-Code-Eval/Bio-Code-Eval-new/2-filter-and-annotate/final functions for test cases/final_java_functions_for_testing.json'\n",
    "df = pd.read_json(files)\n",
    "print(df.index)\n",
    "print(df.loc[91]['prompt'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
